<!doctype html>
<html>
<head>
<meta charset='UTF-8'><meta name='viewport' content='width=device-width initial-scale=1'>
<meta charset="UTF-8" content="">
<meta name="author" content="">
<meta name="description" content="">
<style type='text/css'>html {overflow-x: initial !important;}:root { --bg-color: #ffffff; --text-color: #333333; --select-text-bg-color: #B5D6FC; --select-text-font-color: auto; --monospace: "Lucida Console",Consolas,"Courier",monospace; --title-bar-height: 20px; }
.mac-os-11 { --title-bar-height: 28px; }
html { font-size: 14px; background-color: var(--bg-color); color: var(--text-color); font-family: "Helvetica Neue", Helvetica, Arial, sans-serif; -webkit-font-smoothing: antialiased; }
h1, h2, h3, h4, h5 { white-space: pre-wrap; }
body { margin: 0px; padding: 0px; height: auto; inset: 0px; font-size: 1rem; line-height: 1.42857; overflow-x: hidden; background: inherit; }
iframe { margin: auto; }
a.url { word-break: break-all; }
a:active, a:hover { outline: 0px; }
.in-text-selection, ::selection { text-shadow: none; background: var(--select-text-bg-color); color: var(--select-text-font-color); }
#write { margin: 0px auto; height: auto; width: inherit; word-break: normal; overflow-wrap: break-word; position: relative; white-space: normal; overflow-x: visible; padding-top: 36px; }
#write.first-line-indent p { text-indent: 2em; }
#write.first-line-indent li p, #write.first-line-indent p * { text-indent: 0px; }
#write.first-line-indent li { margin-left: 2em; }
.for-image #write { padding-left: 8px; padding-right: 8px; }
body.typora-export { padding-left: 30px; padding-right: 30px; }
.typora-export .footnote-line, .typora-export li, .typora-export p { white-space: pre-wrap; }
.typora-export .task-list-item input { pointer-events: none; }
@media screen and (max-width: 500px) {
  body.typora-export { padding-left: 0px; padding-right: 0px; }
  #write { padding-left: 20px; padding-right: 20px; }
}
#write li > figure:last-child { margin-bottom: 0.5rem; }
#write ol, #write ul { position: relative; }
img { max-width: 100%; vertical-align: middle; image-orientation: from-image; }
button, input, select, textarea { color: inherit; font: inherit; }
input[type="checkbox"], input[type="radio"] { line-height: normal; padding: 0px; }
*, ::after, ::before { box-sizing: border-box; }
#write h1, #write h2, #write h3, #write h4, #write h5, #write h6, #write p, #write pre { width: inherit; position: relative; }
#write svg h1, #write svg h2, #write svg h3, #write svg h4, #write svg h5, #write svg h6, #write svg p { position: static; }
p { line-height: inherit; }
h1, h2, h3, h4, h5, h6 { break-after: avoid-page; break-inside: avoid; orphans: 4; }
p { orphans: 4; }
h1 { font-size: 2rem; }
h2 { font-size: 1.8rem; }
h3 { font-size: 1.6rem; }
h4 { font-size: 1.4rem; }
h5 { font-size: 1.2rem; }
h6 { font-size: 1rem; }
.md-math-block, .md-rawblock, h1, h2, h3, h4, h5, h6, p { margin-top: 1rem; margin-bottom: 1rem; }
.hidden { display: none; }
.md-blockmeta { color: rgb(204, 204, 204); font-weight: 700; font-style: italic; }
a { cursor: pointer; }
sup.md-footnote { padding: 2px 4px; background-color: rgba(238, 238, 238, 0.7); color: rgb(85, 85, 85); border-radius: 4px; cursor: pointer; }
sup.md-footnote a, sup.md-footnote a:hover { color: inherit; text-transform: inherit; text-decoration: inherit; }
#write input[type="checkbox"] { cursor: pointer; width: inherit; height: inherit; }
figure { overflow-x: auto; margin: 1.2em 0px; max-width: calc(100% + 16px); padding: 0px; }
figure > table { margin: 0px; }
thead, tr { break-inside: avoid; break-after: auto; }
thead { display: table-header-group; }
table { border-collapse: collapse; border-spacing: 0px; width: 100%; overflow: auto; break-inside: auto; text-align: left; }
table.md-table td { min-width: 32px; }
.CodeMirror-gutters { border-right: 0px; background-color: inherit; }
.CodeMirror-linenumber { user-select: none; }
.CodeMirror { text-align: left; }
.CodeMirror-placeholder { opacity: 0.3; }
.CodeMirror pre { padding: 0px 4px; }
.CodeMirror-lines { padding: 0px; }
div.hr:focus { cursor: none; }
#write pre { white-space: pre-wrap; }
#write.fences-no-line-wrapping pre { white-space: pre; }
#write pre.ty-contain-cm { white-space: normal; }
.CodeMirror-gutters { margin-right: 4px; }
.md-fences { font-size: 0.9rem; display: block; break-inside: avoid; text-align: left; overflow: visible; white-space: pre; background: inherit; position: relative !important; }
.md-fences-adv-panel { width: 100%; margin-top: 10px; text-align: center; padding-top: 0px; padding-bottom: 8px; overflow-x: auto; }
#write .md-fences.mock-cm { white-space: pre-wrap; }
.md-fences.md-fences-with-lineno { padding-left: 0px; }
#write.fences-no-line-wrapping .md-fences.mock-cm { white-space: pre; overflow-x: auto; }
.md-fences.mock-cm.md-fences-with-lineno { padding-left: 8px; }
.CodeMirror-line, twitterwidget { break-inside: avoid; }
svg { break-inside: avoid; }
.footnotes { opacity: 0.8; font-size: 0.9rem; margin-top: 1em; margin-bottom: 1em; }
.footnotes + .footnotes { margin-top: 0px; }
.md-reset { margin: 0px; padding: 0px; border: 0px; outline: 0px; vertical-align: top; background: 0px 0px; text-decoration: none; text-shadow: none; float: none; position: static; width: auto; height: auto; white-space: nowrap; cursor: inherit; -webkit-tap-highlight-color: transparent; line-height: normal; font-weight: 400; text-align: left; box-sizing: content-box; direction: ltr; }
li div { padding-top: 0px; }
blockquote { margin: 1rem 0px; }
li .mathjax-block, li p { margin: 0.5rem 0px; }
li blockquote { margin: 1rem 0px; }
li { margin: 0px; position: relative; }
blockquote > :last-child { margin-bottom: 0px; }
blockquote > :first-child, li > :first-child { margin-top: 0px; }
.footnotes-area { color: rgb(136, 136, 136); margin-top: 0.714rem; padding-bottom: 0.143rem; white-space: normal; }
#write .footnote-line { white-space: pre-wrap; }
@media print {
  body, html { border: 1px solid transparent; height: 99%; break-after: avoid; break-before: avoid; font-variant-ligatures: no-common-ligatures; }
  #write { margin-top: 0px; border-color: transparent !important; padding-top: 0px !important; padding-bottom: 0px !important; }
  .typora-export * { -webkit-print-color-adjust: exact; }
  .typora-export #write { break-after: avoid; }
  .typora-export #write::after { height: 0px; }
  .is-mac table { break-inside: avoid; }
  #write > p:nth-child(1) { margin-top: 0px; }
  .typora-export-show-outline .typora-export-sidebar { display: none; }
  figure { overflow-x: visible; }
}
.footnote-line { margin-top: 0.714em; font-size: 0.7em; }
a img, img a { cursor: pointer; }
pre.md-meta-block { font-size: 0.8rem; min-height: 0.8rem; white-space: pre-wrap; background: rgb(204, 204, 204); display: block; overflow-x: hidden; }
p > .md-image:only-child:not(.md-img-error) img, p > img:only-child { display: block; margin: auto; }
#write.first-line-indent p > .md-image:only-child:not(.md-img-error) img { left: -2em; position: relative; }
p > .md-image:only-child { display: inline-block; width: 100%; }
#write .MathJax_Display { margin: 0.8em 0px 0px; }
.md-math-block { width: 100%; }
.md-math-block:not(:empty)::after { display: none; }
.MathJax_ref { fill: currentcolor; }
[contenteditable="true"]:active, [contenteditable="true"]:focus, [contenteditable="false"]:active, [contenteditable="false"]:focus { outline: 0px; box-shadow: none; }
.md-task-list-item { position: relative; list-style-type: none; }
.task-list-item.md-task-list-item { padding-left: 0px; }
.md-task-list-item > input { position: absolute; top: 0px; left: 0px; margin-left: -1.2em; margin-top: calc(1em - 10px); border: none; }
.math { font-size: 1rem; }
.md-toc { min-height: 3.58rem; position: relative; font-size: 0.9rem; border-radius: 10px; }
.md-toc-content { position: relative; margin-left: 0px; }
.md-toc-content::after, .md-toc::after { display: none; }
.md-toc-item { display: block; color: rgb(65, 131, 196); }
.md-toc-item a { text-decoration: none; }
.md-toc-inner:hover { text-decoration: underline; }
.md-toc-inner { display: inline-block; cursor: pointer; }
.md-toc-h1 .md-toc-inner { margin-left: 0px; font-weight: 700; }
.md-toc-h2 .md-toc-inner { margin-left: 2em; }
.md-toc-h3 .md-toc-inner { margin-left: 4em; }
.md-toc-h4 .md-toc-inner { margin-left: 6em; }
.md-toc-h5 .md-toc-inner { margin-left: 8em; }
.md-toc-h6 .md-toc-inner { margin-left: 10em; }
@media screen and (max-width: 48em) {
  .md-toc-h3 .md-toc-inner { margin-left: 3.5em; }
  .md-toc-h4 .md-toc-inner { margin-left: 5em; }
  .md-toc-h5 .md-toc-inner { margin-left: 6.5em; }
  .md-toc-h6 .md-toc-inner { margin-left: 8em; }
}
a.md-toc-inner { font-size: inherit; font-style: inherit; font-weight: inherit; line-height: inherit; }
.footnote-line a:not(.reversefootnote) { color: inherit; }
.reversefootnote { font-family: ui-monospace, sans-serif; }
.md-attr { display: none; }
.md-fn-count::after { content: "."; }
code, pre, samp, tt { font-family: var(--monospace); }
kbd { margin: 0px 0.1em; padding: 0.1em 0.6em; font-size: 0.8em; color: rgb(36, 39, 41); background: rgb(255, 255, 255); border: 1px solid rgb(173, 179, 185); border-radius: 3px; box-shadow: rgba(12, 13, 14, 0.2) 0px 1px 0px, rgb(255, 255, 255) 0px 0px 0px 2px inset; white-space: nowrap; vertical-align: middle; }
.md-comment { color: rgb(162, 127, 3); opacity: 0.6; font-family: var(--monospace); }
code { text-align: left; vertical-align: initial; }
a.md-print-anchor { white-space: pre !important; border-width: initial !important; border-style: none !important; border-color: initial !important; display: inline-block !important; position: absolute !important; width: 1px !important; right: 0px !important; outline: 0px !important; background: 0px 0px !important; text-decoration: initial !important; text-shadow: initial !important; }
.os-windows.monocolor-emoji .md-emoji { font-family: "Segoe UI Symbol", sans-serif; }
.md-diagram-panel > svg { max-width: 100%; }
[lang="flow"] svg, [lang="mermaid"] svg { max-width: 100%; height: auto; }
[lang="mermaid"] .node text { font-size: 1rem; }
table tr th { border-bottom: 0px; }
video { max-width: 100%; display: block; margin: 0px auto; }
iframe { max-width: 100%; width: 100%; border: none; }
.highlight td, .highlight tr { border: 0px; }
mark { background: rgb(255, 255, 0); color: rgb(0, 0, 0); }
.md-html-inline .md-plain, .md-html-inline strong, mark .md-inline-math, mark strong { color: inherit; }
.md-expand mark .md-meta { opacity: 0.3 !important; }
mark .md-meta { color: rgb(0, 0, 0); }
@media print {
  .typora-export h1, .typora-export h2, .typora-export h3, .typora-export h4, .typora-export h5, .typora-export h6 { break-inside: avoid; }
}
.md-diagram-panel .messageText { stroke: none !important; }
.md-diagram-panel .start-state { fill: var(--node-fill); }
.md-diagram-panel .edgeLabel rect { opacity: 1 !important; }
.md-fences.md-fences-math { font-size: 1em; }
.md-fences-advanced:not(.md-focus) { padding: 0px; white-space: nowrap; border: 0px; }
.md-fences-advanced:not(.md-focus) { background: inherit; }
.mermaid-svg { margin: auto; }
.typora-export-show-outline .typora-export-content { max-width: 1440px; margin: auto; display: flex; flex-direction: row; }
.typora-export-sidebar { width: 300px; font-size: 0.8rem; margin-top: 80px; margin-right: 18px; }
.typora-export-show-outline #write { --webkit-flex: 2; flex: 2 1 0%; }
.typora-export-sidebar .outline-content { position: fixed; top: 0px; max-height: 100%; overflow: hidden auto; padding-bottom: 30px; padding-top: 60px; width: 300px; }
@media screen and (max-width: 1024px) {
  .typora-export-sidebar, .typora-export-sidebar .outline-content { width: 240px; }
}
@media screen and (max-width: 800px) {
  .typora-export-sidebar { display: none; }
}
.outline-content li, .outline-content ul { margin-left: 0px; margin-right: 0px; padding-left: 0px; padding-right: 0px; list-style: none; overflow-wrap: anywhere; }
.outline-content ul { margin-top: 0px; margin-bottom: 0px; }
.outline-content strong { font-weight: 400; }
.outline-expander { width: 1rem; height: 1.42857rem; position: relative; display: table-cell; vertical-align: middle; cursor: pointer; padding-left: 4px; }
.outline-expander::before { content: ""; position: relative; font-family: Ionicons; display: inline-block; font-size: 8px; vertical-align: middle; }
.outline-item { padding-top: 3px; padding-bottom: 3px; cursor: pointer; }
.outline-expander:hover::before { content: ""; }
.outline-h1 > .outline-item { padding-left: 0px; }
.outline-h2 > .outline-item { padding-left: 1em; }
.outline-h3 > .outline-item { padding-left: 2em; }
.outline-h4 > .outline-item { padding-left: 3em; }
.outline-h5 > .outline-item { padding-left: 4em; }
.outline-h6 > .outline-item { padding-left: 5em; }
.outline-label { cursor: pointer; display: table-cell; vertical-align: middle; text-decoration: none; color: inherit; }
.outline-label:hover { text-decoration: underline; }
.outline-item:hover { border-color: rgb(245, 245, 245); background-color: var(--item-hover-bg-color); }
.outline-item:hover { margin-left: -28px; margin-right: -28px; border-left: 28px solid transparent; border-right: 28px solid transparent; }
.outline-item-single .outline-expander::before, .outline-item-single .outline-expander:hover::before { display: none; }
.outline-item-open > .outline-item > .outline-expander::before { content: ""; }
.outline-children { display: none; }
.info-panel-tab-wrapper { display: none; }
.outline-item-open > .outline-children { display: block; }
.typora-export .outline-item { padding-top: 1px; padding-bottom: 1px; }
.typora-export .outline-item:hover { margin-right: -8px; border-right: 8px solid transparent; }
.typora-export .outline-expander::before { content: "+"; font-family: inherit; top: -1px; }
.typora-export .outline-expander:hover::before, .typora-export .outline-item-open > .outline-item > .outline-expander::before { content: "−"; }
.typora-export-collapse-outline .outline-children { display: none; }
.typora-export-collapse-outline .outline-item-open > .outline-children, .typora-export-no-collapse-outline .outline-children { display: block; }
.typora-export-no-collapse-outline .outline-expander::before { content: "" !important; }
.typora-export-show-outline .outline-item-active > .outline-item .outline-label { font-weight: 700; }
.md-inline-math-container mjx-container { zoom: 0.95; }
mjx-container { break-inside: avoid; }
.md-alert.md-alert-note { border-left-color: rgb(9, 105, 218); }
.md-alert.md-alert-important { border-left-color: rgb(130, 80, 223); }
.md-alert.md-alert-warning { border-left-color: rgb(154, 103, 0); }
.md-alert.md-alert-tip { border-left-color: rgb(31, 136, 61); }
.md-alert.md-alert-caution { border-left-color: rgb(207, 34, 46); }
.md-alert { padding: 0px 1em; margin-bottom: 16px; color: inherit; border-left: 0.25em solid rgb(0, 0, 0); }
.md-alert-text-note { color: rgb(9, 105, 218); }
.md-alert-text-important { color: rgb(130, 80, 223); }
.md-alert-text-warning { color: rgb(154, 103, 0); }
.md-alert-text-tip { color: rgb(31, 136, 61); }
.md-alert-text-caution { color: rgb(207, 34, 46); }
.md-alert-text { font-size: 0.9rem; font-weight: 700; }
.md-alert-text svg { fill: currentcolor; position: relative; top: 0.125em; margin-right: 1ch; overflow: visible; }
.md-alert-text-container::after { content: attr(data-text); text-transform: capitalize; pointer-events: none; margin-right: 1ch; }


.CodeMirror { height: auto; }
.CodeMirror.cm-s-inner { background: inherit; }
.CodeMirror-scroll { overflow: auto hidden; z-index: 3; }
.CodeMirror-gutter-filler, .CodeMirror-scrollbar-filler { background-color: rgb(255, 255, 255); }
.CodeMirror-gutters { border-right: 1px solid rgb(221, 221, 221); background: inherit; white-space: nowrap; }
.CodeMirror-linenumber { padding: 0px 3px 0px 5px; text-align: right; color: rgb(153, 153, 153); }
.cm-s-inner .cm-keyword { color: rgb(119, 0, 136); }
.cm-s-inner .cm-atom, .cm-s-inner.cm-atom { color: rgb(34, 17, 153); }
.cm-s-inner .cm-number { color: rgb(17, 102, 68); }
.cm-s-inner .cm-def { color: rgb(0, 0, 255); }
.cm-s-inner .cm-variable { color: rgb(0, 0, 0); }
.cm-s-inner .cm-variable-2 { color: rgb(0, 85, 170); }
.cm-s-inner .cm-variable-3 { color: rgb(0, 136, 85); }
.cm-s-inner .cm-string { color: rgb(170, 17, 17); }
.cm-s-inner .cm-property { color: rgb(0, 0, 0); }
.cm-s-inner .cm-operator { color: rgb(152, 26, 26); }
.cm-s-inner .cm-comment, .cm-s-inner.cm-comment { color: rgb(170, 85, 0); }
.cm-s-inner .cm-string-2 { color: rgb(255, 85, 0); }
.cm-s-inner .cm-meta { color: rgb(85, 85, 85); }
.cm-s-inner .cm-qualifier { color: rgb(85, 85, 85); }
.cm-s-inner .cm-builtin { color: rgb(51, 0, 170); }
.cm-s-inner .cm-bracket { color: rgb(153, 153, 119); }
.cm-s-inner .cm-tag { color: rgb(17, 119, 0); }
.cm-s-inner .cm-attribute { color: rgb(0, 0, 204); }
.cm-s-inner .cm-header, .cm-s-inner.cm-header { color: rgb(0, 0, 255); }
.cm-s-inner .cm-quote, .cm-s-inner.cm-quote { color: rgb(0, 153, 0); }
.cm-s-inner .cm-hr, .cm-s-inner.cm-hr { color: rgb(153, 153, 153); }
.cm-s-inner .cm-link, .cm-s-inner.cm-link { color: rgb(0, 0, 204); }
.cm-negative { color: rgb(221, 68, 68); }
.cm-positive { color: rgb(34, 153, 34); }
.cm-header, .cm-strong { font-weight: 700; }
.cm-del { text-decoration: line-through; }
.cm-em { font-style: italic; }
.cm-link { text-decoration: underline; }
.cm-error { color: red; }
.cm-invalidchar { color: red; }
.cm-constant { color: rgb(38, 139, 210); }
.cm-defined { color: rgb(181, 137, 0); }
div.CodeMirror span.CodeMirror-matchingbracket { color: rgb(0, 255, 0); }
div.CodeMirror span.CodeMirror-nonmatchingbracket { color: rgb(255, 34, 34); }
.cm-s-inner .CodeMirror-activeline-background { background: inherit; }
.CodeMirror { position: relative; overflow: hidden; }
.CodeMirror-scroll { height: 100%; outline: 0px; position: relative; box-sizing: content-box; background: inherit; }
.CodeMirror-sizer { position: relative; }
.CodeMirror-gutter-filler, .CodeMirror-hscrollbar, .CodeMirror-scrollbar-filler, .CodeMirror-vscrollbar { position: absolute; z-index: 6; display: none; outline: 0px; }
.CodeMirror-vscrollbar { right: 0px; top: 0px; overflow: hidden; }
.CodeMirror-hscrollbar { bottom: 0px; left: 0px; overflow: auto hidden; }
.CodeMirror-scrollbar-filler { right: 0px; bottom: 0px; }
.CodeMirror-gutter-filler { left: 0px; bottom: 0px; }
.CodeMirror-gutters { position: absolute; left: 0px; top: 0px; padding-bottom: 10px; z-index: 3; overflow-y: hidden; }
.CodeMirror-gutter { white-space: normal; height: 100%; box-sizing: content-box; padding-bottom: 30px; margin-bottom: -32px; display: inline-block; }
.CodeMirror-gutter-wrapper { position: absolute; z-index: 4; background: 0px 0px !important; border: none !important; }
.CodeMirror-gutter-background { position: absolute; top: 0px; bottom: 0px; z-index: 4; }
.CodeMirror-gutter-elt { position: absolute; cursor: default; z-index: 4; }
.CodeMirror-lines { cursor: text; }
.CodeMirror pre { border-radius: 0px; border-width: 0px; background: 0px 0px; font-family: inherit; font-size: inherit; margin: 0px; white-space: pre; overflow-wrap: normal; color: inherit; z-index: 2; position: relative; overflow: visible; }
.CodeMirror-wrap pre { overflow-wrap: break-word; white-space: pre-wrap; word-break: normal; }
.CodeMirror-code pre { border-right: 30px solid transparent; width: fit-content; }
.CodeMirror-wrap .CodeMirror-code pre { border-right: none; width: auto; }
.CodeMirror-linebackground { position: absolute; inset: 0px; z-index: 0; }
.CodeMirror-linewidget { position: relative; z-index: 2; overflow: auto; }
.CodeMirror-wrap .CodeMirror-scroll { overflow-x: hidden; }
.CodeMirror-measure { position: absolute; width: 100%; height: 0px; overflow: hidden; visibility: hidden; }
.CodeMirror-measure pre { position: static; }
.CodeMirror div.CodeMirror-cursor { position: absolute; visibility: hidden; border-right: none; width: 0px; }
.CodeMirror div.CodeMirror-cursor { visibility: hidden; }
.CodeMirror-focused div.CodeMirror-cursor { visibility: inherit; }
.cm-searching { background: rgba(255, 255, 0, 0.4); }
span.cm-underlined { text-decoration: underline; }
span.cm-strikethrough { text-decoration: line-through; }
.cm-tw-syntaxerror { color: rgb(255, 255, 255); background-color: rgb(153, 0, 0); }
.cm-tw-deleted { text-decoration: line-through; }
.cm-tw-header5 { font-weight: 700; }
.cm-tw-listitem:first-child { padding-left: 10px; }
.cm-tw-box { border-style: solid; border-right-width: 1px; border-bottom-width: 1px; border-left-width: 1px; border-color: inherit; border-top-width: 0px !important; }
.cm-tw-underline { text-decoration: underline; }
@media print {
  .CodeMirror div.CodeMirror-cursor { visibility: hidden; }
}


:root {

    /********************************* Text alignment 文本对齐方式 *************************************
     **************************************************************************************************
     * - left:         Align Left         左对齐 
     * - right:        Align right        右对齐
     * - center:       Align center       居中对齐
     * - justify:      Align justify      两端对齐
     * - justify-all:  Align justify-all  全部两端对齐
     * - start:        Align start        行首对齐
     * - end:          Align end          行尾对齐
     **************************************************************************************************/
    --text-align: justify;

    /************************* Interactive animation configuration 交互动画配置 ************************
     **************************************************************************************************
     ** 1.To use interactive animations, first set --use-dynamic-effect` to 1, Then adjust other 
     **   configuration items as needed.
     ** 2.To turn off all interactive animations, set `--use-dynamic-effect` to 0
     **                                           set `--list-ol-text-style` to italic
     *
     ** 1.若要使用交互动画, 请先将 `--use-dynamic-effect` 设置为 1, 然后根据需要调整其他配置项
     ** 2.若要关闭全部交互动画, 请将 `--use-dynamic-effect` 设置为 0  
     **                            `--list-ol-text-style` 设置为 italic
     *
     * 其余配置项请根据需要自行配置
     * - 0: Close 关闭
     * - 1: Open 开启
     **************************************************************************************************/
    /* Use/Close interactive animations 使用/关闭交互动画 */
    --use-dynamic-effect: 1;

    /* H1 interactive effect 一级标题交互效果 */
    --h1-hover-effect: 1;

    /* H2 interactive effect 二级标题交互效果 */
    --h2-after-effect: 1;

    /* H3 H4 H5 H6 interactive effect 三,四,五,六级标题交互效果 */
    --h3-6-hover-effect: 1;

    /* Paragraph interactive effect 段落交互效果 */
    --p-hover-effect: 1;

    /* Image interactive effect 图片交互效果 */
    --img-hover-effect: 1;

    /* Blockquote interactive effect 引用块交互效果 */
    --blockquote-hover-effect: 1;
    
    /* Alert interactive effect 提示框交互效果 */
    --alert-hover-effect: 1;
    
    /* Strong text interactive effect 粗体文本交互效果 */
    --strong-hover-effect: 1;

    /* Italic text interactive effect 斜体文本交互效果 */
    --em-hover-effect: 1;

    /* Underline interactive effect 下划线交互效果 */
    --u-hover-effect: 1;

    /* Marked text interactive effect 高亮文本交互效果 */
    --mark-hover-effect: 1;

    /* Deleted text interactive effect 删除线交互效果 */
    --del-hover-effect: 1;

    /* Inline code block interactive effect 内联代码块交互效果 */
    --code-hover-effect: 1;

    /* Code block interactive effect 代码块交互效果 */
    --fence-hover-effect: 1;
    
    /* List interactive effect 列表交互效果
     * The following two values ​​need to be changed synchronously
     * 下面两个值需同步更改 0-italic or 1-normal */
    --list-marker-effect: 1;
    --list-ol-text-style: normal;
    
    /* Divider interactive effect 分割线交互效果 */
    --hr-hover-effect: 1;
    
    /* Table interactive effect 表格交互效果 */
    --table-hover-effect: 1;

    /*************************** Theme color configuration 主题颜色配置 ********************************
     **************************************************************************************************
     * - 
     **************************************************************************************************/
     --body-text-color: #282c34;
     --content-write-bg-color: #fefefe;
     --selection-color: #def2e8;

     --write-h1-before-bg: linear-gradient(90deg, rgba(152, 251, 152, 0), #e8f4ee, rgba(152, 251, 152, 0));

     --write-h2-after-bg: linear-gradient(90deg, rgba(106, 158, 69, 0), #3ea174, rgba(106, 158, 69, 0));

     --p-hover-bg-color: rgba(123, 181, 147, calc(0.1 * var(--p-hover-effect) * var(--use-dynamic-effect)));
     
     --ul-marker-color: #087725;
     --ulul-marker-color: #087725;
     --ululul-marker-color: #087725;

     --ol-hover-marker-color:rgb(0, calc(100 * var(--list-marker-effect) * var(--use-dynamic-effect)), 0);

     --blockquote-border-color: #2f845e;
     --blockquote-bg-color: #e8f4eb;
     --blockquote-hover-border-color: #2f845e;
     --blockquote-thead-bg-color: #3b9457;
     --blockquote-thead-text-color: #3b3b3b4e;
     --blockquote-td-bg-color: #fafafa;

     --alert-tip-border-color: #1F883D;
     --alert-tip-bg-color: #e8f4eb;
     --alert-tip-hover-border-color: #1F883D;
     --alert-tip-thead-bg-color: #3b9457;
     --alert-tip-thead-text-color: #fafbfa;

     --alert-caution-border-color: #CF222E;
     --alert-caution-bg-color: #f5ecec;
     --alert-caution-hover-border-color: #CF222E;
     --alert-caution-hover-bg-color: #cf222d0a;
     --alter-caution-selection-bg-color: #cf222d17;
     --alert-caution-text-color: #CF222E;
     --alert-caution-thead-bg-color: #d83c46;
     --alert-caution-thead-text-color: #fafbfa;
     --alert-caution-td-hover-bg-color: #f5ecec;
     --alert-caution-list-item-marker-color: #CF222E;

     --alert-warning-border-color: #9A6700;
     --alert-warning-bg-color: #f4f1e9;
     --alert-warning-hover-border-color: #9A6700;
     --alert-warning-hover-bg-color: #9a67000a;
     --alert-warning-selection-bg-color: #9a670017;
     --alert-warning-text-color: #9A6700;
     --alert-warning-thead-bg-color: #b7831c;
     --alert-warning-thead-text-color: #fafbfa;
     --alert-warning-list-item-marker-color: #9A6700;

     --alert-important-border-color: #8250DF;
     --alert-important-bg-color: #eae6f2;
     --alert-important-hover-border-color: #8250DF;
     --alert-important-hover-bg-color: #8250df0a;
     --alert-important-selection-bg-color: #8250df17;
     --alert-important-text-color: #8250DF;
     --alert-important-thead-bg-color: #8250DF;
     --alert-important-thead-text-color: #fafbfa;
     --alert-important-list-item-marker-color: #8250DF;

     --alert-note-border-color: #0969DA;
     --alert-note-bg-color: #e5ebf1;
     --alert-note-hover-border-color: #0969DA;
     --alert-note-hover-bg-color: #0969da0a;
     --alert-note-selection-bg-color: #0969da17;
     --alert-note-text-color: #0969DA;
     --alert-note-thead-bg-color: #0969DA;
     --alert-note-thead-text-color: #fafbfa;
     --alert-note-list-item-marker-color: #0969DA;

     --alert-code-bg-color: #67676716;
     --alert-code-3circle-1: #fc625d;
     --alert-code-3circle-2: #fdbc40;
     --alert-code-3circle-3: #34c759;
     --alert-td-bg-color: #fafafa;

     --hr-bg-image: linear-gradient(90deg, #5ea67c, #2f845e, #5ea67c);
     --hr-box-shadow: 0 4px 6px rgba(0, 0, 0, 0.15);

     --table-border-color: #e0e0e0;
     --table-td-hover-bg: rgba(239, 245, 242, calc(0.7 * var(--table-hover-effect))) calc(var(--table-hover-effect) * 100%);
     --blockquote-td-hover-bg: rgba(239, 245, 242, calc(0.7 * var(--table-hover-effect))) calc(var(--table-hover-effect) * 100%);
     --alert-tip-td-hover-bg: rgba(239, 245, 242, calc(0.7 * var(--table-hover-effect))) calc(var(--table-hover-effect) * 100%);
     --alert-caution-td-hover-bg: rgba(245, 236, 236, calc(0.7 * var(--table-hover-effect))) calc(var(--table-hover-effect) * 100%);
     --alert-warning-td-hover-bg: rgba(245, 241, 236, calc(0.7 * var(--table-hover-effect))) calc(var(--table-hover-effect) * 100%);
     --alert-important-td-hover-bg: rgba(236, 236, 245, calc(0.7 * var(--table-hover-effect))) calc(var(--table-hover-effect) * 100%);
     --alert-note-td-hover-bg: rgba(236, 236, 245, calc(0.7 * var(--table-hover-effect))) calc(var(--table-hover-effect) * 100%);

     --table-thead-bg-color: #eff5f2;
     --table-thead-text-color: #06621e;

     --table-tbody-border-color: #e8f4ee;
     --table-edit-bg-color: #ffffff;
     --table-edit-hover-bg-color: #e8f4ee;

     --table-webkit-box-shadow-color1: rgba(0, 0, 0, 0.16);
     --table-webkit-box-shadow-color2: rgba(0, 0, 0, 0.12);

     --table-reset-bg-color: #ffffff;
     --table-reset-input-color: #282c34;
     --table-some-border-top-color: #eee;

     --table-meta-block-bg-color: #eaf3ec;
     --table-meta-block-color: #3f8a5f;

     --task-list-border-color: #737373;
     --task-list-bg-color: #ffffff;
     --task-list-input-after-border-color: #fff;
     --task-list-input-checked-bg-color: #399b6f;
     --task-list-done-color: #999;

     --math-html-webkit-box-shadow-color1: rgba(0, 0, 0, 0.16);
     --math-html-webkit-box-shadow-color2: rgba(0, 0, 0, 0.12);
     --math-html-rawblock-bg-color: #f5f6f7;
     --math-html-rawblock-control-bg-color: #f5f6f7;
     --math-html-rawblock-tooltip-bg-color: #f5f6f7;
     --math-html-rawblock-input-color: #282c34;
     --math-html-rawblock-gutter-bg-color: #f5f6f7;
     --math-html-rawblock-selected-bg-color: #bac6e9;
     --math-html-rawblock-selected-bg-color2: #def2e8;

     --strong-color: #06621e;
     --em-color: #000000;
     --em-hover-color: #006400;
     --u-border-color: rgb(9, 156, 70);

     --code-bg-color: rgba(221, 235, 213, 0.5);
     --code-color: #359d09;
     --code-hover-bg-color: rgba(221, 235, 213, calc(0.5 + 0.5 * var(--code-hover-effect) * var(--use-dynamic-effect)));

     --del-color: rgba(50, 100, 50, 0.5);
     --del-text-decoration-color:rgba(76, 175, 80, 0.7);
     --del-hover-color: rgba(50, 100, 50, calc(0.5 + 0.5 * var(--del-hover-effect) * var(--use-dynamic-effect)));
     --del-hover-text-decoration-color: rgba(76, 175, 80, calc(0.7 + 0.3 * var(--del-hover-effect) * var(--use-dynamic-effect)));

     --mark-bg-color: #aee6c3;
     --mark-hover-bg-color: rgba(174, 230, 195, calc(1 - 0.5 * var(--mark-hover-effect) * var(--use-dynamic-effect)));

     --a-color: #0c8f94;
     --a-hover-color: #02502a;

     --code-fences-color: #2f4f4f;
     --code-fences-bg-color: #ffffff;
     --code-fences-box-shadow: 0 1px 3px rgba(0, 0, 0, 0.15);
     --code-fences-before-bg: #015e19;
     --code-fences-before-box-shadow: 20px 0 #068f28, 40px 0 rgb(116, 219, 38);
     --code-fences-3circle-1: #fc625d;
     --code-fences-3circle-2: #fdbc40;
     --code-fences-3circle-3: #34c759;

     --cm-s-inner-linenumber-color: #2f845e;

     --scrollbar-track-bg: #fafafa;
     --scrollbar-thumb-bg: linear-gradient(to top right, #4aad23, #1b5b3e) !important;

     --quick-open-bg-color: #fff;
     --quick-open-color: #282c34;
     --quick-open-item-active-bg-color: #bac6e9;
     --quick-open-category-border-color: #eee;

     --modal-dialog-webkit-box-shadow: 0 2px 5px 0 rgba(0, 0, 0, 0.16), 0 2px 10px 0 rgba(0, 0, 0, 0.12);
     --modal-dialog-box-shadow: 0 2px 5px 0 rgba(0, 0, 0, 0.16), 0 2px 10px 0 rgba(0, 0, 0, 0.12);
     --modal-dialog-bg-color: #f5f6f7;
     --modal-dialog-btn-default-hover-bg-color: #bac6e9;
     
     --preferences-color: #282c34;
     --preference-sidebar-nav-group-item-active-color: #efefef;
     --preference-sidebar-nav-group-item-active-bg-color: #2f845e;

     --html-for-mac-sidebar-webkit-box-shadow: 0 2px 5px 0 rgba(0, 0, 0, 0.16), 0 2px 10px 0 rgba(0, 0, 0, 0.12);
     --html-for-mac-sidebar-box-shadow: 0 2px 5px 0 rgba(0, 0, 0, 0.16), 0 2px 10px 0 rgba(0, 0, 0, 0.12);

     --notification-bg-color: #fff;
     --notification-hover-bg-color: #bac6e9;
     --menu-sidebar-color: #282c34;
     --menu-sidebar-list-border-color: #2f353a;
     --menu-sidebar-list-bg-color: #2f353a;
     --menu-sidebar-li-a-color: #e3dede;
     --menu-sidebar-li-a-active-hover-color: #3e4249;
     --megamenu-content-color: #000;

     --sidebar-bg-color: #2f845e;
     --sidebar-content-file-list-item-border-color: #eee;
     --sidebar-content-file-list-item-active-bg-color: rgba(66, 185, 131, 0.1);
     --sidebar-content-file-list-item-active-border-color: #2f845e;
     --sidebar-content-file-list-item-search-pannel-border-color: #eee;
     --sidebar-content-file-list-item-search-pannel-search-option-btn-bg-color: #fff;
     --file-tree-node-hover-border-color: #2f845e;
     --file-tree-node-hover-bg-color: rgba(66, 185, 131, 0.1);
     --file-tree-node-active-border-color: #2f845e;
     --file-tree-node-active-bg-color: rgba(66, 185, 131, 0.1);
     --file-library-file-node-hover-border-color: #2f845e;
     --file-library-file-node-hover-bg-color: rgba(66, 185, 131, 0.1);
     
     --outline-item-hover-bg-color: #e8f4ee;
     --sidebar-footer-border-color: #eee;
     --sidebar-footer-hover-bg-color: #caebdb;
     --sidebar-files-menu-webkit-box-shadow: 0 2px 5px 0 rgba(0, 0, 0, 0.16), 0 2px 10px 0 rgba(0, 0, 0, 0.12);
     --sidebar-files-menu-box-shadow: 0 2px 5px 0 rgba(0, 0, 0, 0.16), 0 2px 10px 0 rgba(0, 0, 0, 0.12);

     --dropdown-menu-border-color: #eee;
     --dropdown-menu-li-a-color: #282c34;
     --dropdown-menu-li-a-hover-color: #e8f4ee;
     --dropdown-menu-divider-border-color: #fff;

     --toc-bg-color: #fff;
     --toc-webkit-box-shadow: 0 2px 5px 0 rgba(0, 0, 0, 0.16), 0 2px 10px 0 rgba(0, 0, 0, 0.12);
     --toc-box-shadow: 0 2px 5px 0 rgba(0, 0, 0, 0.16), 0 2px 10px 0 rgba(0, 0, 0, 0.12);
     --toc-hover-bg-color: #caebdb;

     --auto-suggest-container-bg-color: #d4eee1;

     --source-color: #6a6a6a;
     --source-cm-header-color: #428bca;
     --source-cm-atom-color: #777777;
}

content #write {
    background-color: var(--content-write-bg-color);
    max-width: 90%;
    margin: 0 auto;
    padding-bottom: 7.5rem;
}

content #write p {
    word-spacing: 0.1rem;
    line-height: 1.9rem;
    margin: 3px;
}

content #write ol li {
    padding-left: 0.5rem;
}

content #write>h5.md-focus:before {
    top: 2px;
}

::-moz-selection {
    background-color: var(--selection-color);
}

::selection {
    background-color: var(--selection-color);
}

/* h1-h3 */
#write h1,
#write h2,
#write h3 {
    font-style: normal;
}

#write h1,
#write h2 {
    text-align: center;
    margin-top: 2rem;
    margin-bottom: 1rem;
}

#write h1 {
    font-size: 2.4rem;
    position: relative;
    overflow: hidden;
    transition: transform 0.4s ease;
}

#write h1:before {
    content: "";
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background: var(--write-h1-before-bg);
    transition: transform 0.4s ease;
    transform: scaleX(0);
    transform-origin: center;
    z-index: -1;
    display: calc(var(--h1-hover-effect) * var(--use-dynamic-effect));
}

#write h1:hover {
    transform: scale(calc(1 + 0.1 * var(--h1-hover-effect) * var(--use-dynamic-effect)));
}

#write h1:hover:before {
    transform: scaleX(calc(var(--h1-hover-effect) * var(--use-dynamic-effect)));
}

#write h2 {
    font-size: 2rem;
    position: relative;
    padding-bottom: 0rem;
    font-weight: bold;
    margin-bottom: 0rem;
}

#write h2:after {
    content: "";
    display: block;
    width: 30rem;
    height: 2.5px;
    margin: 0.5rem auto 0;
    background: var(--write-h2-after-bg);
    border-radius: 10px;
    transform: scaleX(0);
    transition: transform 0.4s ease, background 0.4s ease;
}

#write h2:hover:after {
    transform: scaleX(calc(1 * var(--h2-after-effect) * var(--use-dynamic-effect)));
}

#write h3 {
    font-size: 1.8rem;
}

#write h4 {
    font-size: 1.5rem;
}

#write h5 {
    font-size: 1.3rem;
}

#write h6 {
    font-size: 1.2rem;
}

#write h3,
#write h4,
#write h5,
#write h6 {
    position: relative;
    margin-top: 1.0rem;
    margin-bottom: 0.5rem;
    transition: transform 0.3s ease;
}

#write h3:hover,
#write h4:hover,
#write h5:hover,
#write h6:hover {
    transform: scale(calc(1 + 0.01 * var(--h3-6-hover-effect) * var(--use-dynamic-effect)));
}

/* image */
img {
    border-radius: 10px;
    max-width: 100%;
    display: block;
    margin: 1em auto;
    transition: transform 0.15s ease, border-radius 0.15s ease;
}

img:hover {
    transform: scale(calc(1 + 0.01 * var(--img-hover-effect) * var(--use-dynamic-effect)));
    border-radius: calc(10px - 10px * var(--img-hover-effect) * var(--use-dynamic-effect));
}

/* paragraph */
p {
    font-size: 1rem;
    padding: 0 0.3rem;
    transition: all 0.25s ease;
}

p:not(:has(img)):hover {
    background-color: var(--p-hover-bg-color);
}

p.md-toc-content:hover {
    padding-left: 20px;
}

/* ul ol */
ul,
ol {
    padding-left: 1.5rem;
}

ul:first-child,
ol:first-child {
    margin-top: 0;
}

ul:last-child,
ol:last-child {
    margin-bottom: 0;
}

@keyframes bounce {

    0%,
    100% {
        transform: scale(1);
    }

    50% {
        transform: scale(calc(1 + 0.3 * var(--list-marker-effect) * var(--use-dynamic-effect)));
    }
}

ul .md-list-item::marker {
    list-style-type: disc;
    font-size: 27px !important;
    color: var(--ul-marker-color);
    transition: transform 0.3s;
}

ul .md-list-item:hover::marker {
    animation: bounce 0.6s ease-in-out;
}

ul ul .md-list-item::marker {
    list-style-type: circle;
    font-size: 27px;
    color: var(--ulul-marker-color);
    font-weight: bold;
    transition: transform 0.3s;
}

ul ul .md-list-item:hover::marker {
    animation: bounce 0.6s ease-in-out;
}

ul ul ul .md-list-item::marker {
    list-style-type: square;
    font-size: 27px;
    color: var(--ululul-marker-color);
    transition: transform 0.3s;
}

ul ul ul .md-list-item:hover::marker {
    animation: bounce 0.6s ease-in-out;
}

ol .md-list-item::marker {
    font-family: "Times New Roman", Times, serif;
    font-size: 20px;
    font-style: italic;
    transition: font-weight 0.2s ease, font-style 0.2s ease, color 0.2s ease, font-size 0.2s ease;
}

ol .md-list-item:hover::marker {
    font-style: var(--list-ol-text-style);
    font-weight: calc(500 + 100 * var(--list-marker-effect) * var(--use-dynamic-effect));
    font-size: calc(20px + 3px * var(--list-marker-effect) * var(--use-dynamic-effect));
    color: var(--ol-hover-marker-color);
}

/* blockquote */
blockquote {
    padding: 0.5rem 1rem;
    border-left: 8px solid var(--blockquote-border-color);
    background-color: var(--blockquote-bg-color);
    border-radius: 7px;
    transition: box-shadow 0.3s ease, border-radius 0.3s ease, border-left 0.3s ease;
}

blockquote p {
    margin: 0.1rem 0;
}

blockquote:hover {
    border-left: calc(8px + 1px * var(--blockquote-hover-effect) * var(--use-dynamic-effect)) solid var(--blockquote-hover-border-color);
    box-shadow: calc(0px + 1px * var(--blockquote-hover-effect) * var(--use-dynamic-effect)) calc(1px + 1px * var(--blockquote-hover-effect) * var(--use-dynamic-effect)) calc(3px + 3px * var(--blockquote-hover-effect) * var(--use-dynamic-effect)) rgba(0, 0, 0, calc(0.15 * var(--blockquote-hover-effect) * var(--use-dynamic-effect)));
    border-radius: calc(7px + 8px * var(--blockquote-hover-effect) * var(--use-dynamic-effect));
}

blockquote table thead,
blockquote .md-table-fig .md-table thead {
    background-color: var(--blockquote-thead-bg-color);
    color: var(--blockquote-thead-text-color);
}

blockquote td {
    background-color: var(--blockquote-td-bg-color);
}

.md-alert.md-alert-tip {
    padding: 0.5rem 1rem;
    border-left: 8px solid var(--alert-tip-border-color);
    background-color: var(--alert-tip-bg-color);
    border-radius: 7px;
    transition: box-shadow 0.3s ease, border-radius 0.3s ease, border-left 0.3s ease;
}

.md-alert.md-alert-tip:hover {
    border-left: calc(8px + 1px * var(--alert-hover-effect) * var(--use-dynamic-effect)) solid var(--alert-tip-hover-border-color);
    box-shadow: calc(0px + 1px * var(--alert-hover-effect) * var(--use-dynamic-effect)) calc(1px + 1px * var(--alert-hover-effect) * var(--use-dynamic-effect)) calc(3px + 3px * var(--alert-hover-effect) * var(--use-dynamic-effect)) rgba(0, 0, 0, calc(0.15 * var(--alert-hover-effect) * var(--use-dynamic-effect)));
    border-radius: calc(7px + 8px * var(--alert-hover-effect) * var(--use-dynamic-effect));
}

.md-alert.md-alert-tip table thead,
.md-alert.md-alert-tip .md-table-fig .md-table thead {
    background-color: var(--alert-tip-thead-bg-color);
    color: var(--alert-tip-thead-text-color);
}

.md-alert.md-alert-caution {
    padding: 0.5rem 1rem;
    border-left: 8px solid var(--alert-caution-border-color);
    background-color: var(--alert-caution-bg-color);
    border-radius: 7px;
    transition: box-shadow 0.3s ease, border-radius 0.3s ease, border-left 0.3s ease;
}

.md-alert.md-alert-caution:hover {
    border-left: calc(8px + 1px * var(--alert-hover-effect) * var(--use-dynamic-effect)) solid var(--alert-caution-hover-border-color);
    box-shadow: calc(0px + 1px * var(--alert-hover-effect) * var(--use-dynamic-effect)) calc(1px + 1px * var(--alert-hover-effect) * var(--use-dynamic-effect)) calc(3px + 3px * var(--alert-hover-effect) * var(--use-dynamic-effect)) rgba(0, 0, 0, calc(0.15 * var(--alert-hover-effect) * var(--use-dynamic-effect)));
    border-radius: calc(7px + 8px * var(--alert-hover-effect) * var(--use-dynamic-effect));
}

.md-alert.md-alert-caution p:hover {
    background-color: var(--alert-caution-hover-bg-color);
}

.md-alert.md-alert-caution ::selection {
    background-color: var(--alter-caution-selection-bg-color);
}

.md-alert.md-alert-caution strong,
.md-alert.md-alert-caution em,
.md-alert.md-alert-caution code,
.md-alert.md-alert-caution a {
    color: var(--alert-caution-text-color);
}

.md-alert.md-alert-caution table thead,
.md-alert.md-alert-caution .md-table-fig .md-table thead {
    background-color: var(--alert-caution-thead-bg-color);
    color: var(--alert-caution-thead-text-color);
}

.md-alert.md-alert-caution td:hover {
    background-color: var(--alert-caution-td-hover-bg-color);
}

.md-alert.md-alert-caution ul .md-list-item::marker,
.md-alert.md-alert-caution ul ul .md-list-item::marker,
.md-alert.md-alert-caution ul ul .md-list-item::marker,
.md-alert.md-alert-caution ol .md-list-item:hover::marker {
    color: var(--alert-caution-list-item-marker-color);
}

.md-alert.md-alert-warning {
    padding: 0.5rem 1rem;
    border-left: 8px solid var(--alert-warning-border-color);
    background-color: var(--alert-warning-bg-color);
    border-radius: 7px;
    transition: box-shadow 0.3s ease, border-radius 0.3s ease, border-left 0.3s ease;
}

.md-alert.md-alert-warning:hover {
    border-left: calc(8px + 1px * var(--alert-hover-effect) * var(--use-dynamic-effect)) solid var(--alert-warning-hover-border-color);
    box-shadow: calc(0px + 1px * var(--alert-hover-effect) * var(--use-dynamic-effect)) calc(1px + 1px * var(--alert-hover-effect) * var(--use-dynamic-effect)) calc(3px + 3px * var(--alert-hover-effect) * var(--use-dynamic-effect)) rgba(0, 0, 0, calc(0.15 * var(--alert-hover-effect) * var(--use-dynamic-effect)));
    border-radius: calc(7px + 8px * var(--alert-hover-effect) * var(--use-dynamic-effect));
}

.md-alert.md-alert-warning p:hover {
    background-color: var(--alert-warning-hover-bg-color);
}

.md-alert.md-alert-warning ::selection {
    background-color: var(--alert-warning-selection-bg-color);
}

.md-alert.md-alert-warning strong,
.md-alert.md-alert-warning em,
.md-alert.md-alert-warning code,
.md-alert.md-alert-warning a {
    color: var(--alert-warning-text-color);
}

.md-alert.md-alert-warning table thead,
.md-alert.md-alert-warning .md-table-fig .md-table thead {
    background-color: var(--alert-warning-thead-bg-color);
    color: var(--alert-warning-thead-text-color);
}

.md-alert.md-alert-warning ul .md-list-item::marker,
.md-alert.md-alert-warning ul ul .md-list-item::marker,
.md-alert.md-alert-warning ul ul .md-list-item::marker,
.md-alert.md-alert-warning ol .md-list-item:hover::marker {
    color: var(--alert-warning-list-item-marker-color);
}

.md-alert.md-alert-important {
    padding: 0.5rem 1rem;
    border-left: 8px solid var(--alert-important-border-color);
    background-color: var(--alert-important-bg-color);
    border-radius: 7px;
    transition: box-shadow 0.3s ease, border-radius 0.3s ease, border-left 0.3s ease;
}

.md-alert.md-alert-important:hover {
    border-left: calc(8px + 1px * var(--alert-hover-effect) * var(--use-dynamic-effect)) solid var(--alert-important-hover-border-color);
    box-shadow: calc(0px + 1px * var(--alert-hover-effect) * var(--use-dynamic-effect)) calc(1px + 1px * var(--alert-hover-effect) * var(--use-dynamic-effect)) calc(3px + 3px * var(--alert-hover-effect) * var(--use-dynamic-effect)) rgba(0, 0, 0, calc(0.15 * var(--alert-hover-effect) * var(--use-dynamic-effect)));
    border-radius: calc(7px + 8px * var(--alert-hover-effect) * var(--use-dynamic-effect));
}

.md-alert.md-alert-important p:hover {
    background-color: var(--alert-important-hover-bg-color);
}

.md-alert.md-alert-important ::selection {
    background-color: var(--alert-important-selection-bg-color);
}

.md-alert.md-alert-important strong,
.md-alert.md-alert-important em,
.md-alert.md-alert-important code,
.md-alert.md-alert-important a {
    color: var(--alert-important-text-color);
}

.md-alert.md-alert-important table thead,
.md-alert.md-alert-important .md-table-fig .md-table thead {
    background-color: var(--alert-important-thead-bg-color);
    color: var(--alert-important-thead-text-color);
}

.md-alert.md-alert-important ul .md-list-item::marker,
.md-alert.md-alert-important ul ul .md-list-item::marker,
.md-alert.md-alert-important ul ul .md-list-item::marker,
.md-alert.md-alert-important ol .md-list-item:hover::marker {
    color: var(--alert-important-list-item-marker-color);
}

.md-alert.md-alert-note {
    padding: 0.5rem 1rem;
    border-left: 8px solid var(--alert-note-border-color);
    background-color: var(--alert-note-bg-color);
    border-radius: 7px;
    transition: box-shadow 0.3s ease, border-radius 0.3s ease, border-left 0.3s ease;
}

.md-alert.md-alert-note:hover {
    border-left: calc(8px + 1px * var(--alert-hover-effect) * var(--use-dynamic-effect)) solid var(--alert-note-hover-border-color);
    box-shadow: calc(0px + 1px * var(--alert-hover-effect) * var(--use-dynamic-effect)) calc(1px + 1px * var(--alert-hover-effect) * var(--use-dynamic-effect)) calc(3px + 3px * var(--alert-hover-effect) * var(--use-dynamic-effect)) rgba(0, 0, 0, calc(0.15 * var(--alert-hover-effect) * var(--use-dynamic-effect)));
    border-radius: calc(7px + 8px * var(--alert-hover-effect) * var(--use-dynamic-effect));
}

.md-alert.md-alert-note p:hover {
    background-color: var(--alert-note-hover-bg-color);
}  

.md-alert.md-alert-note ::selection {
    background-color: var(--alert-note-selection-bg-color);
}

.md-alert.md-alert-note strong,
.md-alert.md-alert-note em,
.md-alert.md-alert-note code,
.md-alert.md-alert-note a {
    color: var(--alert-note-text-color);
}

.md-alert.md-alert-note table thead,
.md-alert.md-alert-note .md-table-fig .md-table thead {
    background-color: var(--alert-note-thead-bg-color);
    color: var(--alert-note-thead-text-color);
}

.md-alert.md-alert-note ul .md-list-item::marker,
.md-alert.md-alert-note ul ul .md-list-item::marker,
.md-alert.md-alert-note ul ul .md-list-item::marker,
.md-alert.md-alert-note ol .md-list-item:hover::marker {
    color: var(--alert-note-list-item-marker-color);
}

.md-alert p {
    margin: 0.1rem 0;
}

blockquote code,
.md-alert.md-alert-tip code,
.md-alert.md-alert-caution code,
.md-alert.md-alert-warning code,
.md-alert.md-alert-important code,
.md-alert.md-alert-note code {
    background-color: var(--alert-code-bg-color);
}

.md-alert.md-alert-tip .md-fences::before,
.md-alert.md-alert-caution .md-fences::before,
.md-alert.md-alert-warning .md-fences::before,
.md-alert.md-alert-important .md-fences::before,
.md-alert.md-alert-note .md-fences::before {
    background: var(--alert-code-3circle-1);
    box-shadow: 20px 0 var(--alert-code-3circle-2), 40px 0 var(--alert-code-3circle-3);
}

blockquote td,
.md-alert.md-alert-tip td,
.md-alert.md-alert-caution td,
.md-alert.md-alert-warning td,
.md-alert.md-alert-important td,
.md-alert.md-alert-note td {
    background-color: var(--alert-td-bg-color);
}

.md-alert-text-container {
    font-size: 20px;
}

/* horizontal line */
hr {
    position: relative;
    margin-top: 2rem;
    margin-bottom: 2rem;
    border: none;
    border-top: 2px solid transparent;
    background-image: var(--hr-bg-image);
    height: 1.5px;
    overflow: visible;
    color: #abb2bf;
    opacity: 0.8;
    border-radius: 5px;
    box-shadow: var(--hr-box-shadow);
    animation: dashAnimation 0.5s ease-in-out;

    transition: transform 0.3s ease, background-position 0.3s ease, height 0.3s ease;
    background-size: 200% 100%;
    background-position: 0 0;
}

hr:hover {
    background-position: calc(100% * var(--hr-hover-effect) * var(--use-dynamic-effect)), 0;
    transform: scale(calc(1 + 0.02 * var(--hr-hover-effect) * var(--use-dynamic-effect)));
    height: calc(1.5px + 1px * var(--hr-hover-effect) * var(--use-dynamic-effect));
}

@keyframes dashAnimation {
    0% {
        width: calc(100% - 100% * var(--hr-hover-effect) * var(--use-dynamic-effect));
        opacity: calc(1 - 1 * var(--hr-hover-effect) * var(--use-dynamic-effect));
    }

    100% {
        width: 100%;
        opacity: 1;
    }
}

/* table */
table,
.md-table-fig .md-table {
    border: 1px solid var(--table-border-color);
    border-radius: 10px;
    border-collapse: separate;
    overflow: hidden;
}

table th,
table td,
.md-table-fig .md-table th,
.md-table-fig .md-table td {
    padding: 10px 16px;
    border-top: 1px solid var(--table-border-color);
    border: 0.5px solid var(--table-border-color);
    transition: all 0.3s ease;
}

table th,
.md-table-fig .md-table th {
    border-bottom: 1px solid var(--table-border-color);
    border-top: 0;
}

table td:hover,
.md-table-fig .md-table td:hover {
    background: var(--table-td-hover-bg);
    transition: all calc(0.3s * var(--table-hover-effect)) ease;
}

blockquote td:hover {
    background: var(--blockquote-td-hover-bg);
}

.md-alert.md-alert-tip td:hover {
    background: var(--alert-tip-td-hover-bg);
}

.md-alert.md-alert-caution td:hover {
    background: var(--alert-caution-td-hover-bg);
}

.md-alert.md-alert-warning td:hover {
    background: var(--alert-warning-td-hover-bg);
}

.md-alert.md-alert-important td:hover {
    background: var(--alert-important-td-hover-bg);
}

.md-alert.md-alert-note td:hover {
    background: var(--alert-note-td-hover-bg);
}

table thead,
.md-table-fig .md-table thead {
    background-color: var(--table-thead-bg-color);
    color: var(--table-thead-text-color);
}

table tbody+tbody,
.md-table-fig .md-table tbody+tbody {
    border-top: 2px solid var(--table-tbody-border-color);
}

.md-table-fig .md-table-edit {
    background-color: var(--table-edit-bg-color);
}

.md-table-fig .md-table-edit .btn-default:hover {
    background-color: var(--table-edit-hover-bg-color);
}

.md-table-fig .md-table-edit .md-table-resize-popover {
    -webkit-box-shadow: 0 2px 5px 0 var(--table-webkit-box-shadow-color1), 0 2px 10px 0 var(--table-webkit-box-shadow-color2);
    box-shadow: 0 2px 5px 0 var(--table-webkit-box-shadow-color1), 0 2px 10px 0 var(--table-webkit-box-shadow-color2);
    border: none;
    padding: 0;
    background-color: transparent;
}

.md-table-fig .md-table-edit .md-table-resize-popover .arrow {
    top: -10px;
}

.md-table-fig .md-table-edit .md-table-resize-popover .arrow::after {
    border-bottom-color: var(--table-border-color);
}

.md-table-fig .md-table-edit .md-table-resize-popover .md-reset {
    border-radius: 10px;
    background-color: var(--table-reset-bg-color);
}

.md-table-fig .md-table-edit .md-table-resize-popover .md-reset input {
    color: var(--table-reset-input-color) !important;
}

.md-table-fig .md-table-edit .md-table-resize-popover .md-reset input#md-grid-width,
.md-table-fig .md-table-edit .md-table-resize-popover .md-reset input#md-grid-height {
    text-align: center;
}

.md-table-fig .md-table-edit .md-table-resize-popover .popover-title {
    border-top: 1px solid var(--table-some-border-top-color);
}

/* yaml front matter */
#write pre.md-meta-block {
    background: var(--table-meta-block-bg-color);
    color: var(--table-meta-block-color);
    border: 0;
    border-radius: 6px;
    line-height: 1.3rem;
    padding: 0.5rem 1.5rem;
    margin-top: -1rem;
    margin-bottom: 2rem;
}

/* table of contents */
.md-toc {
    margin-bottom: 1.25rem;
}

/* task list */
.md-task-list-item>input::before {
    content: "";
    position: absolute;
    top: -1.5px;
    left: -3.5px;
    border-radius: 50%;
    width: 20px;
    height: 20px;
    border: 1px solid var(--task-list-border-color);
    background: var(--task-list-bg-color);
    -webkit-transition: background-color 200ms ease-in-out;
    transition: background-color 200ms ease-in-out;
}

.md-task-list-item>input::after {
    content: "";
    position: absolute;
    top: 5px;
    left: 1px;
    width: 0.6rem;
    height: 0.3rem;
    border:2px solid var(--task-list-input-after-border-color);
    border-top: 0;
    border-right: 0;
    -webkit-transform: rotate(-45deg);
    transform: rotate(-45deg);
    opacity: 0;
    -webkit-transition: opacity 200ms ease-in-out;
    transition: opacity 200ms ease-in-out;
}

.md-task-list-item>input:checked::before {
    background-color: var(--task-list-input-checked-bg-color);
    border: none;
}

.md-task-list-item>input:checked::after {
    opacity: 1;
}

.md-task-list-item.task-list-done {
    text-decoration: line-through;
    color: var(--task-list-done-color);
}

/* footnote */
.footnotes {
    font-size: 1rem;
}

/* math,html block common */
.mathjax-block[contenteditable="false"]:active,
.mathjax-block[contenteditable="false"]:focus,
.md-htmlblock[contenteditable="false"]:active,
.md-htmlblock[contenteditable="false"]:focus {
    -webkit-box-shadow: 0 2px 5px 0 var(--math-html-webkit-box-shadow-color1), 0 2px 10px 0 var(--math-html-webkit-box-shadow-color2);
    box-shadow: 0 2px 5px 0 var(--math-html-webkit-box-shadow-color1), 0 2px 10px 0 var(--math-html-webkit-box-shadow-color2);
}

.mathjax-block:hover .md-rawblock-container,
.mathjax-block:hover .md-rawblock-tooltip,
.md-htmlblock:hover .md-rawblock-container,
.md-htmlblock:hover .md-rawblock-tooltip {
    background-color: var(--math-html-rawblock-bg-color);
    -webkit-animation: showRawBlockTooltip 0s linear;
    animation: showRawBlockTooltip 0s linear;
}

.mathjax-block:hover .md-rawblock-container,
.md-htmlblock:hover .md-rawblock-container {
    border-radius: 6px 0 6px 6px;
}

.mathjax-block .md-rawblock-control,
.md-htmlblock .md-rawblock-control {
    background-color: var(--math-html-rawblock-control-bg-color);
}

.mathjax-block .md-rawblock-tooltip,
.md-htmlblock .md-rawblock-tooltip {
    background-color: var(--math-html-rawblock-tooltip-bg-color);
    -webkit-animation: showRawBlockTooltip 0s linear;
    animation: showRawBlockTooltip 0s linear;
    border-radius: 4px 4px 0 0;
}

/* math block */
.mathjax-block .md-mathblock-panel .code-tooltip[contenteditable="false"] {
    -webkit-box-shadow: none;
    box-shadow: none;
}

.mathjax-block .md-mathblock-panel .md-rawblock-before {
    padding-top: 6px;
    border-top-left-radius: 6px;
}

.mathjax-block .md-mathblock-panel .md-rawblock-after {
    padding-bottom: 6px;
}

.mathjax-block .md-mathblock-panel .md-rawblock-input .cm-s-inner {
    color: var(--math-html-rawblock-input-color);
}

.mathjax-block .md-mathblock-panel .md-rawblock-input .cm-s-inner .CodeMirror-lines {
    color: var(--math-html-rawblock-input-color);
}

.mathjax-block .md-mathblock-panel .md-rawblock-input .cm-s-inner .CodeMirror-gutters {
    background-color: var(--math-html-rawblock-gutter-bg-color);
}

.mathjax-block .md-mathblock-panel .md-rawblock-input .cm-s-inner .CodeMirror-selected,
.mathjax-block .md-mathblock-panel .md-rawblock-input .cm-s-inner .CodeMirror-selectedtext {
    background-color: var(--math-html-rawblock-selected-bg-color) !important;
}

.mathjax-block .md-math-container {
    padding-top: 10px;
    padding-bottom: 10px;
}

/* html block */
.md-htmlblock .md-htmlblock-panel {
    border-radius: 10px;
    border-top-right-radius: 0;
    padding-left: 6px;
    padding-right: 6px;
}

.md-htmlblock .md-htmlblock-panel .md-rawblock-input {
    padding-top: 14px;
    padding-bottom: 10px;
}

.md-htmlblock .md-htmlblock-panel .md-rawblock-input .cm-s-inner .CodeMirror-selected,
.md-htmlblock .md-htmlblock-panel .md-rawblock-input .cm-s-inner .CodeMirror-selectedtext {
    background-color: var(--math-html-rawblock-selected-bg-color2) !important;
}

/* strong */
strong {
    font-weight: bold;
    color: var(--strong-color);
    font-size: 1.05rem;
    transition: font-size 0.3s ease, padding 0.3s ease;
    display: inline-block;
    margin: 0 2px;
}

strong:hover {
    font-size: calc(1.05rem + 0.05rem * var(--strong-hover-effect) * var(--use-dynamic-effect));
    padding: calc(0px + 3px * var(--strong-hover-effect) * var(--use-dynamic-effect));
}

/* emphasis */
em {
    background-color: inherit;
    color: var(--em-color);
    transition: all 0.3s ease;
    padding-left: 1px;
    padding-right: 1px;
}

em:hover {
    color: var(--em-hover-color);
    font-weight: calc(500 + 100 * var(--em-hover-effect) * var(--use-dynamic-effect));
    padding-left: calc(1px + 2px * var(--em-hover-effect) * var(--use-dynamic-effect));
    padding-right: calc(1px + 2px * var(--em-hover-effect) * var(--use-dynamic-effect));
}

/* underline */
u {
    background-color: inherit;
    color: inherit;
    text-decoration: none;
    border-bottom: 2px solid var(--u-border-color);
    padding-bottom: 2.5px;
    transition: all 0.3s ease;
}

u:hover {
    border: calc(2px * var(--u-hover-effect) * var(--use-dynamic-effect)) solid var(--u-border-color);
    padding: calc(0px + 4px * var(--u-hover-effect) * var(--use-dynamic-effect));
    margin: calc(0px + 4px * var(--u-hover-effect) * var(--use-dynamic-effect));
    border-radius: calc(0px + 7px * var(--u-hover-effect) * var(--use-dynamic-effect));
    border-bottom: 2px solid var(--u-border-color);
    padding-bottom: 2.5px;
}

/* code */
h1 code,
h2 code,
h3 code,
h4 code,
h5 code,
h6 code {
    font-size: inherit;
}

code {
    background-color: var(--code-bg-color);
    color: var(--code-color);
    font-family: "Fira Code", Consolas, "Lucida Console", "Courier", monospace, "Helvetica Neue", Helvetica, Arial, sans-serif;
    font-size: 1rem;
    font-weight: 550;
    margin: 0 2px;
    padding: 3px 3px 1px;
    border-radius: 7px;
    transition: all 0.3s ease;
    z-index: -1;
}

code:hover {
    background-color: var(--code-hover-bg-color);
    font-weight: calc(550 + 150 * var(--code-hover-effect) * var(--use-dynamic-effect));
    padding: 3px 3px calc(1px + 2px * var(--code-hover-effect) * var(--use-dynamic-effect));
}

/* strike */
del {
    color: var(--del-color);
    text-decoration: line-through;
    text-decoration-color: var(--del-text-decoration-color);
    text-decoration-thickness: 1.5px;
    font-style: italic;
    opacity: 0.8;
    transition: color 0.3s ease, opacity 0.3s ease;
}

del:hover {
    color: var(--del-hover-color);
    opacity: calc(0.8 + 0.2 * var(--del-hover-effect) * var(--use-dynamic-effect));
    text-decoration-color: var(--del-hover-text-decoration-color);
}

/* hightlight */
mark {
    font-weight: 500;
    background-color: var(--mark-bg-color);
    color: inherit;
    margin: 0 2px;
    padding: 3px 3px 2.5px;
    border-radius: 0px;
    transition: all 0.3s ease
}

mark:hover {
    border-radius: calc(0px + 7px * var(--mark-hover-effect) * var(--use-dynamic-effect));
    background-color: var(--mark-hover-bg-color);
    padding: 3px 3px calc(2.5px + 1.5px * var(--mark-hover-effect) * var(--use-dynamic-effect));
}

/* hyperlink */
a {
    background-color: inherit;
    color: var(--a-color);
    text-decoration: none;
    transition: color 0.3s ease;
}

a:hover {
    text-decoration: underline;
    color: var(--a-hover-color);
}

a img {
    border: none;
}

/* code fences */
.md-fences {
    font-family: "Fira Code", Consolas, "Lucida Console", "Courier", monospace, "Helvetica Neue", Helvetica, Arial, sans-serif;
    color: var(--code-fences-color);
    background-color: var(--code-fences-bg-color);
    padding: 17px;
    padding-top: 32px;
    border-radius: 10px;
    box-shadow: var(--code-fences-box-shadow);
    transition: transform 0.3s ease, box-shadow 0.3s ease;
    position: relative;
}

.md-fences::before {
    background: var(--code-fences-before-bg);
    border-radius: 50%;
    box-shadow: var(--code-fences-before-box-shadow);
    content: ' ';
    height: 12px;
    width: 12px;
    position: absolute;
    left: 13px;
    top: 12px;
    transition: all 0.3s ease;
}

/* 1. Scale animation */
/* @keyframes bounce-code {
    0%, 100% {
        transform: scale(1);
    }
    50% {
        transform: scale(1.01); 
    }
} */

/* 2. Displacement Animation */
@keyframes bounce-code {

    0%,
    100% {
        width: 100%;
        height: 100%;
        transform: translate(0, 0);
    }

    50% {
        transform: translate(calc(-3px * var(--fence-hover-effect) * var(--use-dynamic-effect)), calc(-3px * var(--fence-hover-effect) * var(--use-dynamic-effect)));
    }
}

.md-fences:hover {
    animation: bounce-code 0.5s ease-in-out;
    box-shadow: 0 calc(1px + 1px * var(--fence-hover-effect) * var(--use-dynamic-effect)) calc(3px + 1px * var(--fence-hover-effect) * var(--use-dynamic-effect)) rgba(0, 0, 0, calc(0.15 + 0.05 * var(--fence-hover-effect) * var(--use-dynamic-effect)));
}

.md-fences:hover::before {
    background: var(--code-fences-3circle-1);
    box-shadow: 20px 0 var(--code-fences-3circle-2), 40px 0 var(--code-fences-3circle-3);
}

.code-tooltip[contenteditable="false"] {
    border-radius: 7px;
}

.cm-s-inner .CodeMirror-linenumber {
    margin-left: 7px;
    margin-right: 3px;
    color: var(--cm-s-inner-linenumber-color);
}

/* 关键字 */
.cm-s-inner .cm-keyword {
    color: #222fa7;
    font-weight: bold;
}

/* 操作符 */
.cm-s-inner .cm-operator {
    color: #006400;
}

/* 变量 */
.cm-s-inner .cm-variable {
    color: #239d37;
    font-weight: bold;
}

/* 函数参数 */
.cm-s-inner .cm-variable-2 {
    color: #108979;
}

/* 类名 */
.cm-s-inner .cm-variable-3 {
    color: #183da2;
    font-weight: bold;
    font-style: italic;
}

/* 内建函数 */
.cm-s-inner .cm-builtin {
    color: #006400;
    font-style: italic;
}

/* 常量 */
.cm-s-inner .cm-atom {
    color: #3b8f8f;
}

/* 数字 */
.cm-s-inner .cm-number {
    color: #3e19cf;
}

/* 函数定义 */
.cm-s-inner .cm-def {
    color: #10623e;
    font-weight: bold;
}

/* 字符串 */
.cm-s-inner .cm-string {
    color: #0264bf;
    font-style: italic;
}

/* 模板字符串 */
.cm-s-inner .cm-string-2 {
    color: #6b8e23;
}

/* 注释 */
.cm-s-inner .cm-comment {
    color: #6a737d;
    font-style: italic;
}

/* HTML 标签 */
.cm-s-inner .cm-tag {
    color: #042c99;
}

/* HTML 标签属性 */
.cm-s-inner .cm-attribute {
    color: #2b34e3;
}

/* CSS 属性 */
.cm-s-inner .cm-property {
    color: #2e8b57;
}

/* 选择器 */
.cm-s-inner .cm-qualifier {
    color: #2e8b57;
}

/* 元信息 */
.cm-s-inner .cm-meta {
    color: #099f8b;
    font-weight: bold;
    font-style: italic;
}

/* 链接 */
.cm-s-inner .cm-link {
    color: #1565c0;
    text-decoration: underline;
}

/* 错误信息 */
.cm-s-inner .cm-error {
    color: #d32f2f;
    background-color: #fdd;
}

/* 布尔值 */
.cm-s-inner .cm-boolean {
    color: #228b22;
    font-weight: bold;
}

/* 类型 */
.cm-s-inner .cm-type {
    color: #1565c0;
    font-weight: bold;
}

/* 标题 */
.cm-s-inner .cm-header {
    color: #2e8b57;
    font-weight: bold;
}

/* 引用 */
.cm-s-inner .cm-quote {
    color: #3503a8;
    font-style: italic;
}

/* scrollbar */
::-webkit-scrollbar {
    width: 8px;
}

::-webkit-scrollbar-track {
    background: var(--scrollbar-track-bg);
}

::-webkit-scrollbar-thumb {
    background: var(--scrollbar-thumb-bg);
    border-radius: 5px;
}

/* find dialog */
#typora-quick-open {
    background-color: var(--quick-open-bg-color);
    color: var(--quick-open-color);
}

#typora-quick-open .typora-quick-open-item.active {
    background-color: var(--quick-open-item-active-bg-color) !important;
}

#typora-quick-open .ty-quick-open-category-title {
    border-top: 1px solid var(--quick-open-category-border-color) !important;
}

/* modal dialog */
#common-dialog .modal-dialog .modal-content {
    -webkit-box-shadow: var(--modal-dialog-webkit-box-shadow);
    box-shadow: var(--modal-dialog-box-shadow);
    background-color: var(--modal-dialog-bg-color);
    border: none;
    border-radius: 6px;
}

#common-dialog .modal-dialog .modal-content .modal-footer .btn-default:hover {
    background-color: var(--modal-dialog-btn-default-hover-bg-color) !important;
}

/* preferences */
.ty-preferences {
    font-family: "Fira Code", Consolas, "Lucida Console", "Courier", monospace, "Helvetica Neue", Helvetica, Arial, sans-serif;
    color: var(--preferences-color);
}

.ty-preferences .window-header h2 {
    font-size: 1.6rem;
}

.ty-preferences .sidebar {
    font-size: 1.4rem;
}

.ty-preferences .sidebar .nav-group-item.active,
.ty-preferences .sidebar .nav-group-item:active {
    color: var(--preference-sidebar-nav-group-item-active-color);
    background-color: var(--preference-sidebar-nav-group-item-active-bg-color);
}

/* sidebar includes file-tree,articles and outline*/
.html-for-mac #typora-sidebar {
    -webkit-box-shadow: var(--html-for-mac-sidebar-webkit-box-shadow);
    box-shadow: var(--html-for-mac-sidebar-box-shadow);
}

/* focus mode/typewriter mode notification  */
#md-notification {
    background-color: var(--notification-bg-color) !important;
}

#md-notification #ty-surpress-mode-warning-close-btn:hover {
    background-color: var(--notification-hover-bg-color) !important;
}

/* Unibody sidebar */
#megamenu-menu-sidebar {
    color: var(--menu-sidebar-color);
    font-size: 1rem;
}

#megamenu-menu-sidebar #megamenu-menu-list {
    border-top-color: var(--menu-sidebar-list-border-color);
    background-color: var(--menu-sidebar-list-bg-color);
}

#megamenu-menu-sidebar #megamenu-menu-list li a {
    color: var(--menu-sidebar-li-a-color) !important;
}

#megamenu-menu-sidebar #megamenu-menu-list li a.active,
#megamenu-menu-sidebar #megamenu-menu-list li a:hover {
    background-color: var(--menu-sidebar-li-a-active-hover-color) !important;
}

#megamenu-menu-sidebar #megamenu-menu-list li a#m-saved {
    background-color: transparent !important;
}

.megamenu-opened .megamenu-menu {
    left: 0;
}

#megamenu-content {
    color: var(--megamenu-content-color);
}

/* sidebar */
#typora-sidebar .info-panel-tab-wrapper .info-panel-tab .info-panel-tab-border {
    background-color: var(--sidebar-bg-color);
    border-radius: 4px;
}

#typora-sidebar .sidebar-osx-tab .sidebar-tabs {
    border-bottom-color: transparent;
}

#typora-sidebar #sidebar-content .file-list-item {
    border-bottom: 1px solid var(--sidebar-content-file-list-item-border-color);
}

#typora-sidebar #sidebar-content .file-list-item.active {
    background-color: var(--sidebar-content-file-list-item-active-bg-color);
    border-left: 4px solid var(--sidebar-content-file-list-item-active-border-color);
}

#typora-sidebar #sidebar-content .ty-sidebar-search-panel {
    border-bottom: 1px solid var(--sidebar-content-file-list-item-search-pannel-border-color);
}

#typora-sidebar #sidebar-content .ty-sidebar-search-panel .searchpanel-search-option-btn {
    background-color: var(--sidebar-content-file-list-item-search-pannel-search-option-btn-bg-color);
}

#typora-sidebar #sidebar-content .sidebar-content-content .file-tree-node:not(.file-node-root):hover>.file-node-background {
    border-left: 4px solid var(--file-tree-node-hover-border-color);
    background-color: var(--file-tree-node-hover-bg-color);
}

#typora-sidebar #sidebar-content .sidebar-content-content .file-tree-node.active>.file-node-background {
    border-color: var(--file-tree-node-active-border-color);
    background-color: var(--file-tree-node-active-bg-color);
}

#typora-sidebar #sidebar-content .sidebar-content-content #file-library-list-children .file-library-file-node:hover {
    border-left: 4px solid var(--file-library-file-node-hover-border-color);
    background-color: var(--file-library-file-node-hover-bg-color);
}

#typora-sidebar #sidebar-content #outline-content .no-collapse-outline .outline-item {
    line-height: 1.375rem;
    font-size: 1rem;
}

#typora-sidebar #sidebar-content #outline-content .outline-expander:before {
    color: inherit;
    font-size: 14px;
    top: auto;
    content: "\f0da";
    font-family: FontAwesome;
}

#typora-sidebar #sidebar-content #outline-content .outline-expander:hover:before,
#typora-sidebar #sidebar-content #outline-content .outline-item-open>.outline-item>.outline-expander:before {
    content: "\f0d7";
}

#typora-sidebar #sidebar-content #outline-content .outline-item:hover {
    background-color: var(--outline-item-hover-bg-color) !important;
}

#typora-sidebar #ty-sidebar-footer {
    border-top: 1px solid var(--sidebar-footer-border-color);
}

#typora-sidebar #ty-sidebar-footer .sidebar-footer-item:hover {
    background-color: var(--sidebar-footer-hover-bg-color) !important;
}

#typora-sidebar #ty-sidebar-footer #sidebar-files-menu {
    -webkit-box-shadow: var(--sidebar-files-menu-webkit-box-shadow);
    box-shadow: var(--sidebar-files-menu-box-shadow);
}

/* dropdown menu */
.dropdown-menu>.show+.menuitem-group-label.show {
    border-top: 1px solid var(--dropdown-menu-border-color) !important;
}

.dropdown-menu li a {
    color: var(--dropdown-menu-li-a-color) !important;
}

.dropdown-menu li a:hover {
    background-color: var(--dropdown-menu-li-a-hover-color) !important;
}

.dropdown-menu .divider {
    border-color: var(--dropdown-menu-divider-border-color);
}

/* sidebar-files-menu,toc-dropmenu,suggest common */
#sidebar-files-menu,
#toc-dropmenu,
.auto-suggest-container {
    background-color: var(--toc-bg-color);
    -webkit-box-shadow: var(--toc-webkit-box-shadow);
    box-shadow: var(--toc-box-shadow);
    border: none;
}

#toc-dropmenu .outline-item:hover {
    background-color: var(--toc-hover-bg-color) !important;
}

/* emoji,fences,word suggest */
.auto-suggest-container {
    padding-top: 4px;
    padding-bottom: 4px;
}

.auto-suggest-container li.active,
.auto-suggest-container li:hover {
    background-color: var(--auto-suggest-container-bg-color) !important;
}

/* source code mode */
#typora-source {
    font-family: "Fira Code", Consolas, "Lucida Console", "Courier", monospace, "Helvetica Neue", Helvetica, Arial, sans-serif;
    color: var(--source-color);
}

#typora-source .cm-s-typora-default .cm-header,
#typora-source .cm-s-typora-default .cm-property,
#typora-source .CodeMirror.cm-s-typora-default div.CodeMirror-cursor {
    color: var(--source-cm-header-color);
}

#typora-source .cm-s-typora-default .cm-atom,
#typora-source .cm-s-typora-default .cm-number {
    color: var(--source-cm-atom-color);
}

#typora-source .CodeMirror-selectedtext {
    background-color: transparent !important;
}

.typora-node .file-list-item-parent-loc,
.typora-node .file-list-item-time,
.typora-node .file-list-item-summary {
    font-family: -apple-system, BlinkMacSystemFont, "Helvetica Neue", Helvetica, "PingFang SC", "Microsoft YaHei", sans-serif;
}

@media print {
    body {
        zoom: 70%;
        margin: 0rem 1rem;
        padding: 1rem;
    }

    table,
    .md-table-fig .md-table {
        border: none;
        border-radius: 10px;
        border-collapse: separate;
        box-shadow: 0 1px 3px rgba(0, 0, 0, 0.15); 
    }

    table th,
    table td,
    .md-table-fig .md-table th,
    .md-table-fig .md-table td {
        padding: 10px 16px;
        border-top: none;
        border: none;
    }

    table th,
    .md-table-fig .md-table th {
        background-color: var(--table-thead-bg-color);
        color: var(--table-thead-text-color);
    }

    blockquote table th,
    blockquote .md-table-fig .md-table th {
        background-color: var(--blockquote-thead-bg-color);
        color: var(--blockquote-thead-text-color);
    }

    .md-alert.md-alert-tip table th,
    .md-alert.md-alert-tip .md-table-fig .md-table th {
        background-color: var(--alert-tip-thead-bg-color);
        color: var(--alert-tip-thead-text-color);
    }

    .md-alert.md-alert-caution table th,
    .md-alert.md-alert-caution .md-table-fig .md-table th {
        background-color: var(--alert-caution-thead-bg-color);
        color: var(--alert-caution-thead-text-color);
    }

    .md-alert.md-alert-warning table th,
    .md-alert.md-alert-warning .md-table-fig .md-table th {
        background-color: var(--alert-warning-thead-bg-color);
        color: var(--alert-warning-thead-text-color);
    }

    .md-alert.md-alert-important table th,
    .md-alert.md-alert-important .md-table-fig .md-table th {
        background-color: var(--alert-important-thead-bg-color);
        color: var(--alert-important-thead-text-color);
    }

    .md-alert.md-alert-note table th,
    .md-alert.md-alert-note .md-table-fig .md-table th {
        background-color: var(--alert-note-thead-bg-color);
        color: var(--alert-note-thead-text-color);
    }

    :where(table, .md-table-fig .md-table) {
        & tr:nth-child(even) {
            background-color: var(--table-row-even-bg-color, #eff5f262);
        }
        & tr:nth-child(odd) {
            background-color: var(--table-row-odd-bg-color, #ffffff);
        }
    }

    hr {
        position: relative;
        margin: 1.2rem auto;
        border: none;
        background-color: #2f845e;
        height: 2.5px;
        width: 1024px !important;
        opacity: 0.8;
        border-radius: 1.75px;
    }
}

@media only screen and (min-width: 1400px) {
    #write {
        max-width: 1024px;
    }

    hr {
        position: relative;
        margin: 2rem auto;
        border: none;
        background-color: #2f845e;
        height: 3px;
        opacity: 0.8;
        border-radius: 1.5px;
    }

    .md-fences:hover {
        animation: none;
    }
}

@media only screen and (min-width: 1800px) {
    #write {
        max-width: 1200px;
    }

    hr {
        position: relative;
        margin: 2rem auto;
        border: none;
        background-color: #2f845e;
        height: 3px;
        opacity: 0.8;
        border-radius: 1.5px;
    }

    .md-fences:hover {
        animation: none;
    }
}

html {
    font-size: 17px;
}



mjx-container[jax="SVG"] {
  direction: ltr;
}

mjx-container[jax="SVG"] > svg {
  overflow: visible;
  min-height: 1px;
  min-width: 1px;
}

mjx-container[jax="SVG"] > svg a {
  fill: blue;
  stroke: blue;
}

mjx-assistive-mml {
  position: absolute !important;
  top: 0px;
  left: 0px;
  clip: rect(1px, 1px, 1px, 1px);
  padding: 1px 0px 0px 0px !important;
  border: 0px !important;
  display: block !important;
  width: auto !important;
  overflow: hidden !important;
  -webkit-touch-callout: none;
  -webkit-user-select: none;
  -khtml-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

mjx-assistive-mml[display="block"] {
  width: 100% !important;
}

mjx-container[jax="SVG"][display="true"] {
  display: block;
  text-align: center;
  margin: 1em 0;
}

mjx-container[jax="SVG"][display="true"][width="full"] {
  display: flex;
}

mjx-container[jax="SVG"][justify="left"] {
  text-align: left;
}

mjx-container[jax="SVG"][justify="right"] {
  text-align: right;
}

g[data-mml-node="merror"] > g {
  fill: red;
  stroke: red;
}

g[data-mml-node="merror"] > rect[data-background] {
  fill: yellow;
  stroke: none;
}

g[data-mml-node="mtable"] > line[data-line], svg[data-table] > g > line[data-line] {
  stroke-width: 70px;
  fill: none;
}

g[data-mml-node="mtable"] > rect[data-frame], svg[data-table] > g > rect[data-frame] {
  stroke-width: 70px;
  fill: none;
}

g[data-mml-node="mtable"] > .mjx-dashed, svg[data-table] > g > .mjx-dashed {
  stroke-dasharray: 140;
}

g[data-mml-node="mtable"] > .mjx-dotted, svg[data-table] > g > .mjx-dotted {
  stroke-linecap: round;
  stroke-dasharray: 0,140;
}

g[data-mml-node="mtable"] > g > svg {
  overflow: visible;
}

[jax="SVG"] mjx-tool {
  display: inline-block;
  position: relative;
  width: 0;
  height: 0;
}

[jax="SVG"] mjx-tool > mjx-tip {
  position: absolute;
  top: 0;
  left: 0;
}

mjx-tool > mjx-tip {
  display: inline-block;
  padding: .2em;
  border: 1px solid #888;
  font-size: 70%;
  background-color: #F8F8F8;
  color: black;
  box-shadow: 2px 2px 5px #AAAAAA;
}

g[data-mml-node="maction"][data-toggle] {
  cursor: pointer;
}

mjx-status {
  display: block;
  position: fixed;
  left: 1em;
  bottom: 1em;
  min-width: 25%;
  padding: .2em .4em;
  border: 1px solid #888;
  font-size: 90%;
  background-color: #F8F8F8;
  color: black;
}

foreignObject[data-mjx-xml] {
  font-family: initial;
  line-height: normal;
  overflow: visible;
}

mjx-container[jax="SVG2"] path[data-c], mjx-container[jax="SVG2"] use[data-c] {
  stroke-width: 3;
}

g[data-mml-node="xypic"] path {
  stroke-width: inherit;
}

.MathJax g[data-mml-node="xypic"] path {
  stroke-width: inherit;
}
mjx-container[jax="SVG"] path[data-c], mjx-container[jax="SVG"] use[data-c] {
							stroke-width: 0;
						}
</style><title>ChipSeq_and_CUT&amp;Tag_upstream</title>
</head>
<body class='typora-export os-windows typora-export-show-outline typora-export-collapse-outline'><div class='typora-export-content'>
<div class="typora-export-sidebar"><div class="outline-content"><li class="outline-item-wrapper outline-h1 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#perface">perface</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h1"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#basic-knowledge">Basic knowledge</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h2"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#the-workflow-of-illumina-ngs">The workflow of Illumina NGS</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#step-1-library-preparation">Step 1. Library preparation</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#step-2-cluster-generation">Step 2. Cluster generation</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#step-3-sequencing">Step 3. Sequencing</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#single-read-and-paired-end">Single-read and Paired-End</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h4 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#单端测序single-read">单端测序(Single-read)</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h4 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#双端测序paired-end">双端测序(Paired-End)</a></div><ul class="outline-children"></ul></li></ul></li></ul></li><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#what-are-reads-contigs-and-scaffold">What are reads, contigs and scaffold?</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#chip-seq-theory">ChIP-seq theory</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#cuttag-theory">CUT&amp;TAG theory</a></div><ul class="outline-children"></ul></li></ul></li><li class="outline-item-wrapper outline-h1"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#various-types-of-file-structure">Various types of file structure</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#0-base-or-1-base">0-base or 1-base？</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#cigar-字符串">CIGAR 字符串</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#sra">SRA</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#fastq">FASTQ</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#structure-of-fastq">Structure of FASTQ</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#fastq-sequence-identifier">FASTQ Sequence Identifier</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#phred33-and-phred64-encoding">phred33 and phred64 encoding</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#view-basic-information-of-fastq">View basic information of FASTQ</a></div><ul class="outline-children"></ul></li></ul></li><li class="outline-item-wrapper outline-h2"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#sam">SAM</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#sam-header">SAM Header</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#sam-alignment">SAM Alignment</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#flag值">FLAG值</a></div><ul class="outline-children"></ul></li></ul></li><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#bam">BAM</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#cram">CRAM</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#bed">BED</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#基本-bed-格式bed3">基本 BED 格式（BED3）</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#扩展-bed-格式bed6">扩展 BED 格式（BED6）</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#完整-bed-格式bed12">完整 BED 格式（BED12）</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#bedpe-格式">BEDPE 格式</a></div><ul class="outline-children"></ul></li></ul></li><li class="outline-item-wrapper outline-h2"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#gffgtf">GFF/GTF</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#gff">GFF</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#gtf">GTF</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#different-from-gff-of-gtf">Different from GFF of GTF</a></div><ul class="outline-children"></ul></li></ul></li><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#vcf">VCF</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#chromsizes">chrom.sizes</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#bedgraph-and-bigwig">Bedgraph and BigWIG</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#bedgraph">Bedgraph </a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#bigwig">BigWIG</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#different-from-bedgraph-and-bigwig">Different from Bedgraph and BigWIG</a></div><ul class="outline-children"></ul></li></ul></li></ul></li><li class="outline-item-wrapper outline-h1 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#environment-set-up">Environment set up</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h1"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#convert-sra-to-fastq">Convert SRA to FASTQ</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#parallel-fastq-dump">parallel-fastq-dump</a></div><ul class="outline-children"></ul></li></ul></li><li class="outline-item-wrapper outline-h1"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#qc-report-and-data-cleaning---trim-galore">QC Report and Data cleaning—trim-galore</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#qc-report------fastqc">QC Report——fastqc</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#stucture-of-qc-report">Stucture of QC report</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#basic-statistics">Basic Statistics</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#per-base-sequence-quality">Per base sequence quality</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#per-tile-sequence-quality">Per tile sequence quality</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#per-sequence-quality-scores">Per sequence quality scores</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#per-base-sequence-content">Per base sequence content</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#per-base-sequence-content-in-cuttag">Per base sequence content in CUT&amp;Tag</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#per-sequence-gc-conten">Per sequence GC conten</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#per-base-n-content">Per base N content</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#sequence-length-distribution">Sequence Length Distribution</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#sequence-duplication-level">Sequence Duplication Level</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#overrepresented-sequences">Overrepresented sequences</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#adapter-content">Adapter Content</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#overrepresented-kmers">Overrepresented Kmers</a></div><ul class="outline-children"></ul></li></ul></li><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#fastqc是如何检测adapter的呢">fastqc是如何检测adapter的呢？</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#data-cleaning---cutadapt">Data cleaning—cutadapt</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#theory">Theory</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#parameters">parameters</a></div><ul class="outline-children"></ul></li></ul></li></ul></li><li class="outline-item-wrapper outline-h1"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#reads-alignment">Reads Alignment</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h2"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#alignment-theory">Alignment theory</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#如何评估序列相似">如何评估序列相似？</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#常见的dna替换记分矩阵">常见的DNA替换记分矩阵</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#what-is-the-alignment">What is the alignment?</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#global-alignment">Global Alignment</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#local-alignment">Local Alignment</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#different-of-global-alignment-and-local-alignment-in-reads-alignment">Different of Global Alignment and Local Alignment in reads alignment</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#一致度identity和相似度similarity">一致度(identity)和相似度(similarity)</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#blast">BLAST</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#different-of-dnaseq-alignment-and-rnaseq-alignment">Different of DNAseq alignment and RNAseq alignment</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#method-of-dnaseq-alignment------seed-and-extend">Method of DNAseq alignment——seed and extend</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#different-of-signle-read-alignment-and-paired-end-alignment">Different of Signle-read alignment and Paired-End alignment</a></div><ul class="outline-children"></ul></li></ul></li><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#software-selection-of-alignment">Software selection of Alignment</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#bowtie2">Bowtie2</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#bowtie2-preparation">Bowtie2 preparation</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#bowtie2-parameters">Bowtie2 parameters</a></div><ul class="outline-children"></ul></li></ul></li><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#use-the-ecoli-genome-to-normalize-data-optionalcuttagstep1">Use the E.coli genome to normalize data (optional/CUT&amp;Tag)(step1)</a></div><ul class="outline-children"></ul></li></ul></li><li class="outline-item-wrapper outline-h1"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#samtools">samtools</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#samtools-sort">Samtools sort</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#samtools-idxstats------view-the-alignment-with-chromosome">Samtools idxstats——view the alignment with chromosome</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#samtools-flagstat------view-the-alignment-results">Samtools flagstat——view the alignment results</a></div><ul class="outline-children"></ul></li></ul></li><li class="outline-item-wrapper outline-h1 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#convert-sam-to-bamcram">Convert SAM to BAM/CRAM</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h1"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#remove-duplicates">Remove duplicates</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#remove-duplicates-theory">Remove duplicates theory</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#patterned-flowcell">Patterned Flowcell?</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#传统未图案化流动池non-patterned-flowcell">传统未图案化流动池(Non-Patterned Flowcell)</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#图案化流动池patterned-flowcell">图案化流动池(Patterned Flowcell)</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#patterned-flowcell与光学重复检测的关系">Patterned Flowcell与光学重复检测的关系</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#如何查看自己数据是否使用patterned-flowcell">如何查看自己数据是否使用Patterned Flowcell？</a></div><ul class="outline-children"></ul></li></ul></li><li class="outline-item-wrapper outline-h2"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#software-selection-of-remove-duplicates">Software selection of remove duplicates</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#samtools-rmdup">samtools rmdup</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#picard-markduplicates">Picard MarkDuplicates</a></div><ul class="outline-children"></ul></li></ul></li><li class="outline-item-wrapper outline-h2"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#picard-markduplicates-2">Picard MarkDuplicates</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#picard-markduplicates-parameters">Picard MarkDuplicates parameters</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#remove-duplicates-strategy">Remove duplicates strategy</a></div><ul class="outline-children"></ul></li></ul></li><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#remove-duplicates-in-cuttag">Remove duplicates in CUT&amp;Tag</a></div><ul class="outline-children"></ul></li></ul></li><li class="outline-item-wrapper outline-h1"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#assess-mapped-fragment-size-distributionoptial">Assess mapped fragment size distribution(optial)</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#what-is-the-fragment-">What is the fragment ?</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#tn5-转座酶偏好性">Tn5 转座酶偏好性</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#extract-fragment-length-and-count">Extract fragment length and count</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#draw-picture">Draw picture</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#python">python</a></div><ul class="outline-children"></ul></li></ul></li></ul></li><li class="outline-item-wrapper outline-h1"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#alignment-filtering">Alignment filtering</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h2"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#alignment-filtering-with-mapq-optinal">Alignment filtering with MAPQ (optinal)</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#filtering-with-mapq">Filtering with MAPQ</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#view-the-reads-before-and-after-filtering">View the reads before and after filtering</a></div><ul class="outline-children"></ul></li></ul></li><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#alignment-filtering-with-mapped-read-pairs">Alignment filtering with Mapped Read Pairs</a></div><ul class="outline-children"></ul></li></ul></li><li class="outline-item-wrapper outline-h1"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#bedtools">Bedtools</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#bedtools-introduction">Bedtools introduction</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#bedtools-genomecov">Bedtools genomecov</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#bedtools-intersect">Bedtools intersect</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#bedtools-bamtobed">Bedtools bamtobed</a></div><ul class="outline-children"></ul></li></ul></li><li class="outline-item-wrapper outline-h1"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#convert-bam-to-bed-and-bed-operate">Convert BAM to BED and BED Operate</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#convert-bam-to-bed-with-bedtools">Convert BAM to BED with bedtools</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#keep-the-read-pairs-that-are-on-the-same-chromosome-and-fragment-length-less-than-1000bpoptial">Keep the read pairs that are on the same chromosome and fragment length less than 1000bp（optial）</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#extract-the-fragment-related-columns">Extract the fragment related columns</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#evaluate-the-reproducibility-of-replicate-samplesoptial">Evaluate the reproducibility of replicate samples(optial)</a></div><ul class="outline-children"></ul></li></ul></li><li class="outline-item-wrapper outline-h1"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#convert-bed-to-bedgraph">Convert BED to Bedgraph</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#convert-bed-to-bedgraph-2">Convert BED to Bedgraph</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#use-the-ecoli-genome-to-normalize-data-optionalcuttagstep2">Use the E.coli genome to normalize data (optional/CUT&amp;Tag)(step2)</a></div><ul class="outline-children"></ul></li></ul></li><li class="outline-item-wrapper outline-h1"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#peak-calling">Peak calling</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#macs2">macs2</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#seacrfor-cuttag-and-cutrun">SEACR(For CUT&amp;Tag and CUT&amp;RUN)</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#seacr-theory">SEACR Theory</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#seacr-选项">SEACR 选项</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#seacr安装与用法">SEACR安装与用法</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#seacr网页版">SEACR网页版</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#structure-of-seacr-bed-file">Structure of SEACR bed file</a></div><ul class="outline-children"></ul></li></ul></li><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#number-of-peaks">Number of peaks</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#comparing-the-same-peaks-on-replicate-sample-data-setsoptial">Comparing the same peaks on replicate sample data sets(optial)</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#calculates-frips-fraction-of-reads-in-peaks">Calculates FRiPs (Fraction of Reads in Peaks)</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#sorting-file-procedure">Sorting file procedure</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#code-with-sorting-file-and-calcuating-frips">Code with Sorting file and calcuating FRIPs </a></div><ul class="outline-children"></ul></li></ul></li></ul></li><li class="outline-item-wrapper outline-h1 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#genome-browser">Genome browser</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h1"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#normalization-methods">Normalization methods</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#rpkm">RPKM</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#cpm">CPM</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#bpm">BPM</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#rpgc">RPGC</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#chipseqspikeinfree">ChIPseqSpikeInFree</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#chipseqspikeinfree-perface">ChIPseqSpikeInFree perface</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#chipseqspikeinfree-code">ChIPseqSpikeInFree code</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#chipseqspikeinfree-网页版">ChIPseqSpikeInFree 网页版</a></div><ul class="outline-children"></ul></li></ul></li></ul></li><li class="outline-item-wrapper outline-h1"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#deeptools">Deeptools</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#deeptools-installation">Deeptools installation</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#bamcoverage">BamCoverage</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#computematrix">computeMatrix</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#computematrix-perface">computeMatrix perface</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#computematrix-scale-regions">computeMatrix scale-regions</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#computematrix-referencepoint">computeMatrix referencePoint</a></div><ul class="outline-children"></ul></li></ul></li><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#plotheatmap">plotHeatmap</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#deepblue">DeepBlue </a></div><ul class="outline-children"></ul></li></ul></li><li class="outline-item-wrapper outline-h1"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#convert-bam-to-bigwig-and-bigwig-operate">Convert Bam to BigWIG and BigWIG Operate</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h2 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#convert-bam-to-bigwig">Convert Bam to BigWIG </a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h2"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#heatmap-visualization">Heatmap visualization</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#heatmap-over-transcription-units">Heatmap over transcription units</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#heatmap-on-cuttag-peaks">Heatmap on CUT&amp;Tag peaks</a></div><ul class="outline-children"></ul></li></ul></li></ul></li><li class="outline-item-wrapper outline-h1"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#differential-analysis">Differential analysis</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h2"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#deseq2">DESeq2</a></div><ul class="outline-children"><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#deseq2-perface">DEseq2 perface</a></div><ul class="outline-children"></ul></li><li class="outline-item-wrapper outline-h3 outline-item-single"><div class="outline-item"><span class="outline-expander"></span><a class="outline-label" href="#code">Code</a></div><ul class="outline-children"></ul></li></ul></li></ul></li></div></div><div id='write'  class=''><p>&nbsp;</p><h1 id='perface'><span>perface</span></h1><ul><li><p><a href='https://www.protocols.io/view/cut-amp-tag-data-processing-and-analysis-tutorial-e6nvw93x7gmk/v1'><span>CUT&amp;TAG数据分析流程</span></a></p></li></ul><p><img src="./picture/086.png" width="1200"/></p><p><span>&lt;img src=&quot;./picture/087.png&quot; width=&quot;700&quot;/</span></p><p>&nbsp;</p><hr /><h1 id='basic-knowledge'><span>Basic knowledge</span></h1><h2 id='the-workflow-of-illumina-ngs'><span>The workflow of Illumina NGS</span></h2><p><strong><span>参考：</span></strong><span> </span><a href='https://www.jianshu.com/p/077911bf772d'><span>基础——illumina测序原理与细节</span></a></p><h3 id='step-1-library-preparation'><span>Step 1. Library preparation</span></h3><table><tr>
  <td><img src="./picture/068.png" width="100%" /></td>
  <td><img src="./picture/069.png" width="100%" /></td>
</tr></table><ul><li><p><strong><span>DNA fragmentation</span></strong><span>：genomic DNA becomes DNA fragment with 200-500bp in length</span></p></li><li><p><strong><span>End Repair</span></strong><span>: Blunt-ended fragments are generated</span></p></li><li><p><strong><span>A-tailing</span></strong><span>: 使用</span><strong><span>Klenow酶</span></strong><span>来给</span><strong><span>3&#39;末端</span></strong><span>添加一个突出的</span><strong><span>A</span></strong><span>，用于连接 </span><strong><span>Adapters</span></strong><span> 突出的</span><strong><span>T</span></strong></p></li><li><p><strong><span>Adapter Ligation</span></strong><span>: </span></p><figure class='table-figure'><table><thead><tr><th style='text-align:left;' ><span>区域</span></th><th style='text-align:left;' ><span>功能</span></th></tr></thead><tbody><tr><td style='text-align:left;' ><strong><span>P5/P7</span></strong></td><td style='text-align:left;' ><span>与 Flowcell 上的寡核苷酸互补结合，用于桥式PCR扩增和测序引物结合。</span></td></tr><tr><td style='text-align:left;' ><strong><span>Rd1 SP/Rd2 SP</span></strong></td><td style='text-align:left;' ><span>Sequencing Primer 测序引物</span></td></tr><tr><td style='text-align:left;' ><strong><span>Barcode/Index</span></strong></td><td style='text-align:left;' ><span>用于区分不同样本。部分平台在P5 Adapter上增加第二个Index（</span><strong><span>双索引设计</span></strong><span>）</span></td></tr></tbody></table></figure></li></ul><p>&nbsp;</p><div class="md-alert md-alert-tip tip"><p><span class='md-alert-text md-alert-text-tip'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M8 1.5c-2.363 0-4 1.69-4 3.75 0 .984.424 1.625.984 2.304l.214.253c.223.264.47.556.673.848.284.411.537.896.621 1.49a.75.75 0 0 1-1.484.211c-.04-.282-.163-.547-.37-.847a8.456 8.456 0 0 0-.542-.68c-.084-.1-.173-.205-.268-.32C3.201 7.75 2.5 6.766 2.5 5.25 2.5 2.31 4.863 0 8 0s5.5 2.31 5.5 5.25c0 1.516-.701 2.5-1.328 3.259-.095.115-.184.22-.268.319-.207.245-.383.453-.541.681-.208.3-.33.565-.37.847a.751.751 0 0 1-1.485-.212c.084-.593.337-1.078.621-1.489.203-.292.45-.584.673-.848.075-.088.147-.173.213-.253.561-.679.985-1.32.985-2.304 0-2.06-1.637-3.75-4-3.75ZM5.75 12h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5ZM6 15.25a.75.75 0 0 1 .75-.75h2.5a.75.75 0 0 1 0 1.5h-2.5a.75.75 0 0 1-.75-.75Z"></path></svg>Tip</span><br></p><p><strong><span>双Index设计</span></strong></p><ul><li><p><strong><span>Paire-End</span></strong><span> 区分两个方向的 reads</span></p></li><li><p><span>单Index在测序中可能因化学错误或交叉污染导致Index误读。双Index需同时匹配i5+i7才认定样本来源，错误率降低至&lt;0.1%</span></p></li><li><p><span>通过组合i5和i7，可生成 </span><strong><span>指数级增长的样本容量</span></strong></p></li><li><p><span>双索引可与</span><strong><span>UMI（Unique Molecular Identifier</span></strong><span>）联用，进一步消除PCR重复偏差</span></p></li></ul></div><p>&nbsp;</p><h3 id='step-2-cluster-generation'><span>Step 2. Cluster generation</span></h3><p>&nbsp;</p><div class="md-alert md-alert-note note"><p><span class='md-alert-text md-alert-text-note'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M0 8a8 8 0 1 1 16 0A8 8 0 0 1 0 8Zm8-6.5a6.5 6.5 0 1 0 0 13 6.5 6.5 0 0 0 0-13ZM6.5 7.75A.75.75 0 0 1 7.25 7h1a.75.75 0 0 1 .75.75v2.75h.25a.75.75 0 0 1 0 1.5h-2a.75.75 0 0 1 0-1.5h.25v-2h-.25a.75.75 0 0 1-.75-.75ZM8 6a1 1 0 1 1 0-2 1 1 0 0 1 0 2Z"></path></svg>Note</span><br></p><p><strong><span>Flowcell</span></strong></p><p><img src="./picture/061.png" width="700"/></p><p><strong><span>Flowcell</span></strong><span>，我们也可以把它叫做</span><strong><span>芯片</span></strong><span>，像一个载玻片，上面有一行字母，是这个 </span><strong><span>Flowcell</span></strong><span>的</span><strong><span>编号</span></strong></p><ul><li><p><span>中间的通道，我们把它们叫做 </span><strong><span>lane</span></strong><span>，这里就是我们</span><strong><span>测序反应发生的地方</span></strong></p></li><li><p><strong><span>lane</span></strong><span> 又被分成了好多行，每一行我们把它叫做 </span><strong><span>swath</span></strong></p></li><li><p><span>继续放大！每行 </span><strong><span>swath</span></strong><span> 又被分成好多小格子，每个小格叫做 </span><strong><span>tile</span></strong></p></li><li><p><strong><span>tile</span></strong><span>里的DNA片段被扩增就是</span><strong><span>簇（Cluster）</span></strong><span>生成，一个</span><strong><span>Cluster</span></strong><span>对应 fastq 中一条 </span><strong><span>read</span></strong></p></li></ul></div><p>&nbsp;</p><p><strong><span>Flowcell</span></strong><span>，这个结构的底部有很多的多核苷酸链，这些多核苷酸链是与我们的</span><strong><span>p5和p7 adapter</span></strong><span>互补的。</span></p><table><tr>
  <td><img src="./picture/071.png" width="100%" /></td>
  <td><img src="./picture/072.png" width="100%" /></td>
</tr></table><ul><li><p><span>P5 adapter 序列会和 Flowcell 上面的P5 adapter互补序列结合</span></p></li><li><p><span>DNA的合成后，由于 Flowcell 上面的 adapter互补序列是和 Flowcell</span><strong><span>共价连接</span></strong><span>的（黄色圆圈标出）</span></p></li><li><p><span>使用碱性溶液冲洗 Flowcelll 时，靠</span><strong><span>共价键连接在 Flowcell 上的链并不会被冲洗掉</span></strong><span>，反之</span><strong><span>另一条链就被洗掉</span></strong><span>了，最后就成了最右边的那种情况。</span></p><p>&nbsp;</p></li></ul><table><tr>
  <td><img src="./picture/073.png" width="100%" /></td>
  <td><img src="./picture/074.png" width="100%" /></td>
</tr></table><ul><li><p><span>P7 adapter 序列会和 Flowcell 上面的P7 adapter互补序列结合</span></p></li><li><p><span>形成一个桥，进行</span><strong><span>桥式PCR</span></strong><span>，DNA合成后用碱性溶液就能使DNA双链解链，</span></p></li><li><p><strong><span>簇生成（Cluster generation）</span></strong><span>：如此重复进行</span><strong><span>桥式PCR</span></strong><span>，最终就会形成一个</span><strong><span>局部簇（localized cluster）</span></strong></p><p><span>单个碱基的荧光终究是比较弱的，但是很多个碱基的荧光聚在一起就会很明显了</span></p><p>&nbsp;</p></li></ul><h3 id='step-3-sequencing'><span>Step 3. Sequencing</span></h3><table><tr>
  <td><img src="./picture/075.png" width="100%" /></td>
  <td><img src="./picture/076.png" width="100%" /></td>
</tr></table><ul><li><p><span>连接在P5上面的序列洗掉，就成了上图左边的情况，然后我们让read1的primer结合上去，边合成边测序，记住，这里一簇DNA有很多这样的链，所以你看到的荧光信号会很强</span></p></li><li><p><strong><span>双端测序</span></strong><span>: 同样的我们再进行一次桥式PCR，这时候Flowcell上的P5接头也会“长出”一些序列，然后我们这次洗掉连接在P7上的序列，上下的就只有P5连接的序列了，再把Read2的primer加进去就能测得Read2了</span></p></li></ul><p>&nbsp;</p><h3 id='single-read-and-paired-end'><span>Single-read and Paired-End</span></h3><p><img src="./picture/001.png" width="600"/></p><p><strong><span>单端测序</span></strong><span>只从一侧读，而</span><strong><span>双端测序</span></strong><span>是两头同时读然后拼接</span></p><p>&nbsp;</p><h4 id='单端测序single-read'><span>单端测序(Single-read)</span></h4><ul><li><p><strong><span>简写:SE50</span></strong><span> 只测一个方向，且read长度固定为50 bp</span></p></li><li><p><strong><span>Single-read</span></strong><span>首先将DNA样本进行片段化处理形成200-500bp的片段，引物 sequence 连接到DNA片段的一端，然后末端加上adapter，将片段固定在Flowcell上生成DNA簇，上机测序单端读取 sequence 。该方式建库简单，操作步骤少，常用于小基因组、转录组、宏基因组测序。</span></p></li><li><p><span>测序的质量会随着测序的进行而下降，所以 </span><strong><span>reads 约往后面越不准确</span></strong><span>，</span><strong><span>单端测序下游质量会很差</span></strong><span>，所以就引入了双端测序可以大大提高测序的准确率。</span></p></li></ul><p>&nbsp;</p><h4 id='双端测序paired-end'><span>双端测序(Paired-End)</span></h4><ul><li><p><strong><span>简写:25x25 PE</span></strong><span> 每个片段会从两个方向进行测序，每个read长度为25 bp</span></p></li><li><p><strong><span>Paired-end</span></strong><span>文库制备，指在构建待测DNA文库时在两端的adapter上都加上测序引物结合位点，在第一轮测序完成后，去除第一轮测序的模板链，用对读测序模块(Paired-End Module)引导互补链在原位置再生和扩增，以达到第二轮测序所用的模板量，进行第二轮互补链的合成测序。</span></p><p>&nbsp;</p></li></ul><hr /><h2 id='what-are-reads-contigs-and-scaffold'><span>What are reads, contigs and scaffold?</span></h2><p><img src="./picture/050.png" width="700"/></p><ul><li><p><strong><span>Reads</span></strong><span>是测序仪直接产生的短片段 DNA  sequence 。它们是基因组 sequence 的基本单位。</span></p></li><li><p><strong><span>Contig</span></strong><span>是通过将重叠的 </span><strong><span>reads</span></strong><span> 拼接起来，组装而成的连续 DNA  sequence ，</span><strong><span>Contig</span></strong><span> 不包含缺口，是从 </span><strong><span>reads</span></strong><span> 直接拼接得来的最连续的部分</span></p></li><li><p><strong><span>Scaffold</span></strong><span> 是更高层次的组装结构，它将多个 </span><strong><span>contig</span></strong><span> 连接在一起，可能使用额外信息（如 mate-pair reads 或物理图谱数据）来排列和定向这些 </span><strong><span>contig</span></strong><span>。</span><strong><span>Scaffold</span></strong><span> 之间可能有未知的区域（这些区域用 N 表示）。</span></p><p>&nbsp;</p></li></ul><hr /><h2 id='chip-seq-theory'><span>ChIP-seq theory</span></h2><ul><li><p><span>使用甲醛将目标蛋白（组蛋白，转录因子等）与染色质交联固定起来</span>
<img src="./picture/004.png" width="500"/></p></li><li><p><span>从细胞裂解液分离基因组DNA，通过超声打断DNA为一定长度的小片段</span>
<img src="./picture/005.png"  width="500"/></p></li><li><p><span>添加与目标蛋白质特异的抗体，该抗体会与目标蛋白形成免疫结合复合体沉淀（靶蛋白 + 抗体 + 靶蛋白结合的DNA），收集这些沉淀</span>
<img src="./picture/006.png"  width="500"/></p></li><li><p><span>去交联，分开蛋白与DNA，纯化DNA即可得到染色质免疫沉淀的DNA样本</span>
<img src="./picture/007.png"  width="500"/></p></li><li><p><span>建立好文库，用测序仪进行测序</span>
<img src="./picture/008.png"  width="500"/></p><p>&nbsp;</p></li></ul><hr /><h2 id='cuttag-theory'><span>CUT&amp;TAG theory</span></h2><ul><li><p><span>首先，特异性抗体和靶标蛋白孵育结合；加入Tn5转座酶—Protein A复合物，其中Tn5转座酶两端已装载好建库adapter引物</span></p></li><li><p><span>特异性抗体和Protein A相结合，转座酶Tn5被“拉”到转录因子附近；转座酶Tn5一次性完成附近DNA打断并完成NGS加adapter过程，可直接建库</span></p></li></ul><p><img src= "./picture/010.png" width="500"/></p><p>&nbsp;</p><ul><li><p><span>对比ChIP-seq，CUT&amp;TAG免去了甲醛交联、超声破碎和免疫共沉淀的过程</span></p></li><li><p><span>这样既节省了初始实验材料，需求的细胞量较少，又提高了信噪比、提升了实验重复性</span></p></li></ul><p><img src="./picture/011.png" width="500"/></p><p>&nbsp;</p><hr /><h1 id='various-types-of-file-structure'><span>Various types of file structure</span></h1><h2 id='0-base-or-1-base'><span>0-base or 1-base？</span></h2><ul><li><p><a href='https://genome-blog.gi.ucsc.edu/blog/2016/12/12/the-ucsc-genome-browser-coordinate-counting-systems/'><span>详见</span></a></p></li><li><p><span>在</span><strong><span>0-based</span></strong><span>坐标系中，起始位置从 0 开始计数，范围是 </span><strong><span>[start, end)</span></strong><span> ，即 </span><strong><span>end 位置不包含在区间内</span></strong><span>。</span></p></li></ul><ul><li><p><span>在</span><strong><span>1-based</span></strong><span>坐标系中，起始位置从 1 开始计数，范围是 </span><strong><span>[start, end]</span></strong><span> ，即 </span><strong><span>end 位置包含在区间内</span></strong><span>。</span></p></li></ul><figure class='table-figure'><table><thead><tr><th><span>文件格式</span></th><th><span>坐标系统</span></th><th><span>说明</span></th></tr></thead><tbody><tr><td><strong><span>BED (UCSC)</span></strong></td><td><strong><span>0-based</span></strong></td><td><code>[start, end)</code></td></tr><tr><td><strong><span>GTF/GFF (Ensembl)</span></strong></td><td><strong><span>1-based</span></strong></td><td><code>[start, end]</code></td></tr><tr><td><strong><span>VCF (Ensembl)</span></strong></td><td><strong><span>1-based</span></strong></td><td><span>变异位点坐标从 1 开始，不是区间，而是单个碱基的位置</span></td></tr><tr><td><strong><span>BAM/SAM</span></strong></td><td><strong><span>1-based</span></strong></td><td><span>比对起始位置为 1-based，CIGAR 操作的结果确定终止位置</span></td></tr><tr><td><strong><span>LiftOver</span></strong></td><td><strong><span>0-based</span></strong></td><td><span>输入文件通常是 0-based（如 BED）</span></td></tr><tr><td><strong><span>CrossMap</span></strong></td><td><strong><span>根据输入格式</span></strong></td><td><span>支持多种格式，如 BED（0-based）、GTF/GFF/VCF（1-based）</span></td></tr></tbody></table></figure><p>&nbsp;</p><hr /><h2 id='cigar-字符串'><span>CIGAR 字符串</span></h2><p><span>CIGAR（Compact Idiosyncratic Gapped Alignment Report） 字符串是一种用于描述比对过程中 sequence 与参考基因组之间关系的格式。CIGAR 字符串以紧凑的方式表示了每个比对中存在的匹配、缺失、插入等操作</span></p><ul><li><p><strong><span>M</span></strong><span>: 匹配（Match）。表示 sequence 和参考 sequence 在这一段上的匹配或者不匹配碱基。通常在 DNA-seq 中仅表示匹配碱基，而在 RNA-seq 中可能还表示不匹配。</span></p></li><li><p><strong><span>I</span></strong><span>: 插入（Insertion）。表示 sequence 相对于参考基因组有插入碱基。</span></p></li><li><p><strong><span>D</span></strong><span>: 缺失（Deletion）。表示 sequence 相对于参考基因组有缺失碱基。</span></p></li><li><p><strong><span>N</span></strong><span>: 跳过区域（Skipped region）。通常用于 RNA-seq 数据，表示剪接事件，即读取的 sequence 跳过了参考基因组中的一段内含子。</span></p></li><li><p><strong><span>S</span></strong><span>: 软剪切（Soft clipping）。表示读取的 sequence 中有一部分没有比对到参考基因组（这些碱基仍保留在比对的 sequence 中）。</span></p></li><li><p><strong><span>H</span></strong><span>: 硬剪切（Hard clipping）。与软剪切类似，但这些碱基在输出文件中被完全删除，不再保留。</span></p></li><li><p><strong><span>P</span></strong><span>: 填充（Padding）。用于在某些比对中添加填充符号（一般用于对齐中间的空隙）。</span></p></li><li><p><strong><span>=</span></strong><span>:  sequence 匹配（Sequence match）。表示 sequence 与参考基因组完全匹配的部分。</span></p></li><li><p><strong><span>X</span></strong><span>:  sequence 不匹配（Sequence mismatch）。表示 sequence 与参考基因</span></p><p>&nbsp;</p></li></ul><div class="md-alert md-alert-note note"><p><span class='md-alert-text md-alert-text-note'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M0 8a8 8 0 1 1 16 0A8 8 0 0 1 0 8Zm8-6.5a6.5 6.5 0 1 0 0 13 6.5 6.5 0 0 0 0-13ZM6.5 7.75A.75.75 0 0 1 7.25 7h1a.75.75 0 0 1 .75.75v2.75h.25a.75.75 0 0 1 0 1.5h-2a.75.75 0 0 1 0-1.5h.25v-2h-.25a.75.75 0 0 1-.75-.75ZM8 6a1 1 0 1 1 0-2 1 1 0 0 1 0 2Z"></path></svg>Note</span><br></p><p><strong><span>例如:10M1I5M1D4M</span></strong></p><ul><li><p><span>M表示前 10 个碱基与参考基因组匹配</span></p></li><li><p><span>1I表示接下来有 1 个碱基是插入的</span></p></li><li><p><span>5M表示接下来的 5 个碱基与参考基因组匹配</span></p></li><li><p><span>1D表示在参考基因组中缺失了 1 个碱基，这个碱基在比对的 sequence 中不存在</span></p></li><li><p><span>4M表示最后的 4 个碱基与参考基因组匹配组不匹配的部分</span></p></li></ul></div><p>&nbsp;</p><hr /><h2 id='sra'><span>SRA</span></h2><p><span>SRA 是 NCBI 及其他数据库（如 EBI、DDBJ）用于存储高通量测序数据的一种特殊格式，采用 </span><strong><span>Spot-based 存储方式</span></strong><span>。SRA 文件包含 </span><strong><span>元数据（Metadata）、测序数据（Reads）、质量值（Quality Scores）</span></strong><span> 等信息，并采用 </span><strong><span>压缩存储</span></strong><span> 以减少文件体积。</span></p><p>&nbsp;</p><p><span>一个 </span><code>.sra</code><span> 文件通常包含：</span></p><figure class='table-figure'><table><thead><tr><th><strong><span>部分</span></strong></th><th><strong><span>说明</span></strong></th></tr></thead><tbody><tr><td><strong><span>Metadata（元数据）</span></strong></td><td><span>记录实验信息，如测序仪型号、文库构建方法、生物样本信息等。</span></td></tr><tr><td><strong><span>Reads（测序数据）</span></strong></td><td><span>存储实际的Reads，单端数据包含 1 条 Read，双端数据包含 2 条 Read。</span></td></tr><tr><td><strong><span>Quality Scores（质量值）</span></strong></td><td><span>记录每个碱基的质量评分（Phred Score），用于评估测序准确性。</span></td></tr><tr><td><strong><span>Index（索引）</span></strong></td><td><strong><span>Spot ID</span></strong><span>。用于快速检索。</span></td></tr></tbody></table></figure><div class="md-alert md-alert-tip tip"><p><span class='md-alert-text md-alert-text-tip'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M8 1.5c-2.363 0-4 1.69-4 3.75 0 .984.424 1.625.984 2.304l.214.253c.223.264.47.556.673.848.284.411.537.896.621 1.49a.75.75 0 0 1-1.484.211c-.04-.282-.163-.547-.37-.847a8.456 8.456 0 0 0-.542-.68c-.084-.1-.173-.205-.268-.32C3.201 7.75 2.5 6.766 2.5 5.25 2.5 2.31 4.863 0 8 0s5.5 2.31 5.5 5.25c0 1.516-.701 2.5-1.328 3.259-.095.115-.184.22-.268.319-.207.245-.383.453-.541.681-.208.3-.33.565-.37.847a.751.751 0 0 1-1.485-.212c.084-.593.337-1.078.621-1.489.203-.292.45-.584.673-.848.075-.088.147-.173.213-.253.561-.679.985-1.32.985-2.304 0-2.06-1.637-3.75-4-3.75ZM5.75 12h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5ZM6 15.25a.75.75 0 0 1 .75-.75h2.5a.75.75 0 0 1 0 1.5h-2.5a.75.75 0 0 1-.75-.75Z"></path></svg>Tip</span><br></p><p><strong><span>Spot ID</span></strong><span> 是 SRA 数据中的核心索引，表示 </span><strong><span>一条完整的测序记录</span></strong><span>（可能包含单端或双端的 reads）</span></p></div><p>&nbsp;</p><p><span>对于 </span><strong><span>双端测序（Paired-End Sequencing）</span></strong><span>，SRA 文件中的</span><strong><span>Index（索引）</span></strong><span>、</span><strong><span>测序数据（Reads）、质量值（Quality Scores）</span></strong><span>可能如下：</span></p><figure class='table-figure'><table><thead><tr><th><strong><span>Spot ID</span></strong></th><th><strong><span>Read 1 (Forward Read)</span></strong></th><th><strong><span>Read 2 (Reverse Read)</span></strong></th><th><strong><span>质量值（Q Score）</span></strong></th></tr></thead><tbody><tr><td><span>1</span></td><td><span>ATCGTACGATCGA</span></td><td><span>TAGCTAGCTAGCTAGC</span></td><td><span>40 38 37 35 ...</span></td></tr><tr><td><span>2</span></td><td><span>GCTAGCTAGCTAGC</span></td><td><span>CGTACGTACGTAGCTA</span></td><td><span>39 37 36 35 ...</span></td></tr><tr><td><span>3</span></td><td><span>TACGATCGATCGTA</span></td><td><span>ATCGTAGCTAGCTAGC</span></td><td><span>38 36 34 33 ...</span></td></tr></tbody></table></figure><p>&nbsp;</p><hr /><h2 id='fastq'><span>FASTQ</span></h2><h3 id='structure-of-fastq'><span>Structure of FASTQ</span></h3><p><span>对于每个通过质控参数的</span><strong><span>cluster</span></strong><span>，一个 read 被写入相应样本的 R1 FASTQ 文件，而对于双端测序运行，另外一个 read 也被写入该样本的 R2 FASTQ 文件。  </span></p><ul><li><p><strong><span>FASTQ 文件中的每个条目包含 4 行</span></strong></p><p>&nbsp;</p></li></ul><p><img src="./picture/012.png" width="800"/></p><ul><li><ul><li><p><strong><span>Sequence Identifier</span></strong><span>（</span><code>@</code><span> 开头），包含有关测序运行和Cluster的信息。 该行的具体内容会因使用的 BCL 到 FASTQ 转换软件而不同。</span></p></li><li><p><span>** sequence **（碱基信号；A、C、T、G 和 N）。</span></p></li><li><p><strong><span>分隔符</span></strong><span>，仅为一个加号（</span><code>+</code><span>）。</span></p></li><li><p><strong><span>质量值</span></strong><span>，表示每个碱基的测序质量，使用 </span><strong><span>Phred +33</span></strong><span> 编码的 ASCII 字符。</span></p></li></ul></li></ul><p>&nbsp;</p><h3 id='fastq-sequence-identifier'><span>FASTQ Sequence Identifier</span></h3><ul><li><p><strong><span>Illumina 新格式 (HiSeq/NovaSeq)</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang=""><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang=""><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><span><span>​</span>x</span></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation"><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">@HISEQ:001:ABC123:1:1101:2071:3156 1:N:0:ATCACG</span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">@&lt;instrument&gt;:&lt;run_number&gt;:&lt;flowcell_ID&gt;:&lt;lane&gt;:&lt;tile&gt;:&lt;x_pos&gt;:&lt;y_pos&gt; &lt;read&gt;:&lt;is_filtered&gt;:&lt;control_number&gt;:&lt;index&gt;</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 66px;"></div><div class="CodeMirror-gutters" style="display: none; height: 66px;"></div></div></div></pre><figure class='table-figure'><table><thead><tr><th style='text-align:left;' ><span>名称</span></th><th><span>字段</span></th><th><span>说明</span></th></tr></thead><tbody><tr><td style='text-align:left;' ><strong><span>instrument</span></strong></td><td><code>HISEQ</code></td><td><span>测序仪名称（仪器 ID）</span></td></tr><tr><td style='text-align:left;' ><strong><span>run_number</span></strong></td><td><code>001</code></td><td><span>测序运行编号</span></td></tr><tr><td style='text-align:left;' ><strong><span>flowcell_ID</span></strong></td><td><code>ABC123</code></td><td><span>Flowcell ID</span></td></tr><tr><td style='text-align:left;' ><strong><span>lane</span></strong></td><td><code>1</code></td><td><span>测序 lane 号</span></td></tr><tr><td style='text-align:left;' ><strong><span>tile</span></strong></td><td><code>1101</code></td><td><span>Tile 编号</span></td></tr><tr><td style='text-align:left;' ><strong><span>x_pos</span></strong></td><td><code>2071</code></td><td><span>x 坐标（像素级别）</span></td></tr><tr><td style='text-align:left;' ><strong><span>y_pos</span></strong></td><td><code>3156</code></td><td><span>y 坐标（像素级别）</span></td></tr><tr><td style='text-align:left;' ><strong><span>read</span></strong></td><td><code>1</code></td><td><span>Read 编号（1=Read1，2=Read2）</span></td></tr><tr><td style='text-align:left;' ><strong><span>is_filtered</span></strong></td><td><code>N</code></td><td><span>是否通过质量控制（N=通过，Y=未通过）</span></td></tr><tr><td style='text-align:left;' ><strong><span>control_number</span></strong></td><td><code>0</code></td><td><span>控制位（一般为 0）</span></td></tr><tr><td style='text-align:left;' ><strong><span>index</span></strong></td><td><code>ATCACG</code></td><td><span>样本的 Index（barcode）sequence</span></td></tr></tbody></table></figure><p>&nbsp;</p></li><li><p><strong><span>Illumina 旧格式 (Solexa)</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang=""><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang=""><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation"><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">@HWI-ST1234_0012:4:1101:2071:3156#ATCACG/1</span></pre></div></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 22px;"></div><div class="CodeMirror-gutters" style="display: none; height: 22px;"></div></div></div></pre><figure class='table-figure'><table><thead><tr><th><strong><span>字段</span></strong></th><th><strong><span>说明</span></strong></th></tr></thead><tbody><tr><td><code>HWI-ST1234_0012</code></td><td><span>测序仪 ID</span></td></tr><tr><td><code>4</code></td><td><span>测序 lane 号</span></td></tr><tr><td><code>1101</code></td><td><span>Tile 编号</span></td></tr><tr><td><code>2071</code></td><td><span>x 坐标（像素级别）</span></td></tr><tr><td><code>3156</code></td><td><span>y 坐标（像素级别）</span></td></tr><tr><td><code>#ATCACG</code></td><td><span>Index（barcode）sequence</span></td></tr><tr><td><code>/1</code></td><td><span>Read 编号（1=Read1，2=Read2）</span></td></tr></tbody></table></figure></li></ul><p>&nbsp;</p><p>&nbsp;</p><h3 id='phred33-and-phred64-encoding'><span>phred33 and phred64 encoding</span></h3><figure class='table-figure'><table><thead><tr><th><span>编码方式</span></th><th><span>质量值范围</span></th><th><span>ASCII 字符范围</span></th><th><span>适用测序仪</span></th></tr></thead><tbody><tr><td><strong><span>Phred+33</span></strong></td><td><span>33 - 74</span></td><td><code>!</code><span> 到 </span><code>J</code></td><td><span>Illumina 1.8 及更新版本（当前主流）</span></td></tr><tr><td><strong><span>Phred+64</span></strong></td><td><span>64 - 104</span></td><td><code>@</code><span> 到 </span><code>h</code></td><td><span>早期 Illumina 1.3 - 1.7 版本</span></td></tr></tbody></table></figure><p><span>Phred 质量分数越高，表示测序碱基的可靠性越高。</span></p><p>&nbsp;</p><h3 id='view-basic-information-of-fastq'><span>View basic information of FASTQ</span></h3><p><span>获取 sequence 标识符，其中包含有关测序运行和cluster的信息，有些包含测序平台信息：</span></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation"><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">zcat &lt;xxx.fastq.gz&gt; | <span class="cm-builtin">grep</span> <span class="cm-string">"^@"</span> | head <span class="cm-attribute">-n</span> <span class="cm-number">5</span></span></pre></div></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 22px;"></div><div class="CodeMirror-gutters" style="display: none; height: 22px;"></div></div></div></pre><p>&nbsp;</p><hr /><h2 id='sam'><span>SAM</span></h2><ul><li><p><a href='https://samtools.github.io/hts-specs/SAMtags.pdf'><span>SAM文件详细指南</span></a></p></li><li><p><strong><span>SAM</span></strong><span> 文件是 Sequence Alignment/Map 格式的简称，从名字就可以看出，该文件格式设计初衷就是为了记录比对结果的。 但是由于其在记录的比对结果的同时，其实也记录了 sequence 本身，因此也可以作为测序数据的存储格式。</span></p></li><li><p><strong><span>SAM</span></strong><span>文件是一种平面文件，也就是没有经过压缩，因此一般占用的存储空间相对较大。因此实际工作中，一般不以</span><strong><span>SAM</span></strong><span>文件格式作为存储形式，而是采用其压缩格式</span><strong><span>BAM</span></strong><span>形式。</span></p></li><li><p><strong><span>SAM</span></strong><span> is a tab-delimited format with two main sections: the </span><strong><span>Header</span></strong><span> and the </span><strong><span>alignment section</span></strong><span>.</span></p></li></ul><p>&nbsp;</p><h3 id='sam-header'><span>SAM Header</span></h3><p><strong><span>SAM Header</span></strong><span> 以 </span><code>@</code><span> 开头，包含元数据，如参考基因组信息、样本信息、比对工具等</span></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="text"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="text"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">@HD &nbsp;  VN:1.4 &nbsp;  SO:coordinate</span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">@SQ &nbsp;  SN:CHROMOSOME_I &nbsp;  LN:15072423</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">UR:ftp://ftp.ncbi.nlm.nih.gov/genbank/genomes/Eukaryotes/invertebrates/Caenorhabditis_elegans/</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">WBcel215/Primary_Assembly/assembled_chromosomes/FASTA/chrI.fa.gz &nbsp;  AS:ce10 &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">SP:Caenorhabditis elegans</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">@SQ &nbsp;  SN:CHROMOSOME_II &nbsp;  LN:15279345 &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">UR:ftp://ftp.ncbi.nlm.nih.gov/genbank/genomes/Eukaryotes/invertebrates/Caenorhabditis_elegans/</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">WBcel215/Primary_Assembly/assembled_chromosomes/FASTA/chrII.fa.gz &nbsp; &nbsp; AS:ce10 &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">SP:Caenorhabditis elegans &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">@RG &nbsp;  ID:1 &nbsp;  PL:ILLUMINA &nbsp;  LB:C_ele_05 &nbsp;  DS:WGS of C elegans &nbsp;  PG:BamIndexDecoder</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">@PG &nbsp;  ID:bwa &nbsp;  PN:bwa &nbsp;  VN:0.5.10-tpx</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 284px;"></div><div class="CodeMirror-gutters" style="display: none; height: 284px;"></div></div></div></pre><figure class='table-figure'><table><thead><tr><th><span>标签</span></th><th><span>含义</span></th><th><span>解释</span></th></tr></thead><tbody><tr><td><strong><span>@HD(Header)</span></strong></td><td><strong><span>文件头信息</span></strong></td><td><strong><span>记录 SAM 版本和排序方式</span></strong></td></tr><tr><td><span>VN:1.4</span></td><td><span>版本号</span></td><td><span>SAM 文件格式版本</span></td></tr><tr><td><span>SO:coordinate</span></td><td><span>排序方式</span></td><td><span>按基因组坐标排序</span></td></tr><tr><td><strong><span>@SQ(Sequence Dictionary)</span></strong></td><td><strong><span>参考 sequence </span></strong></td><td><strong><span>记录基因组染色体信息</span></strong></td></tr><tr><td><span>SN:CHROMOSOME_I</span></td><td><span>sequence 名称</span></td><td><span>染色体 I</span></td></tr><tr><td><span>LN:15072423</span></td><td><span>sequence 长度</span></td><td><span>15,072,423 bp</span></td></tr><tr><td><span>UR:ftp://...</span></td><td><span>参考 sequence 来源</span></td><td><span>基因组文件的 FTP 地址</span></td></tr><tr><td><span>AS:ce10</span></td><td><span>基因组版本</span></td><td><span>ce10</span></td></tr><tr><td><span>SP:Caenorhabditis elegans</span></td><td><span>物种名称</span></td><td><span>秀丽隐杆线虫</span></td></tr><tr><td><strong><span>@RG(Read Group)</span></strong></td><td><strong><span>读组</span></strong></td><td><strong><span>记录测序数据的信息</span></strong></td></tr><tr><td><span>ID:1</span></td><td><span>读组 ID</span></td><td><span>读组编号为 1</span></td></tr><tr><td><span>PL:ILLUMINA</span></td><td><span>测序平台</span></td><td><span>ILLUMINA</span></td></tr><tr><td><span>LB:C_ele_05</span></td><td><span>文库 ID</span></td><td><span>C_ele_05</span></td></tr><tr><td><span>DS:WGS of C elegans</span></td><td><span>描述信息</span></td><td><span>秀丽隐杆线虫全基因组测序</span></td></tr><tr><td><span>PG:BamIndexDecoder</span></td><td><span>处理工具</span></td><td><span>BamIndexDecoder</span></td></tr><tr><td><strong><span>@PG(Program)</span></strong></td><td><strong><span>处理程序</span></strong></td><td><strong><span>记录比对软件信息</span></strong></td></tr><tr><td><span>ID:bwa</span></td><td><span>程序 ID</span></td><td><span>bwa</span></td></tr><tr><td><span>PN:bwa</span></td><td><span>程序名称</span></td><td><span>bwa</span></td></tr><tr><td><span>VN:0.5.10-tpx</span></td><td><span>版本号</span></td><td><span>bwa 版本 0.5.10-tpx</span></td></tr></tbody></table></figure><p>&nbsp;</p><h3 id='sam-alignment'><span>SAM Alignment</span></h3><p><strong><span>SAM Alignment</span></strong><span>  文件包含测序比对的结果，每一行代表一条</span><strong><span>read</span></strong><span>的比对信息，字段以 (</span><code>\t</code><span>) 分隔。</span></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang=""><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang=""><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">3658435 &nbsp;  145 &nbsp;  CHROMOSOME_I &nbsp;  1 &nbsp;  0 &nbsp;  100M &nbsp;  CHROMOSOME_II &nbsp;  2716898 &nbsp;  0 &nbsp; &nbsp;</span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">GCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCT</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">AAGCCT &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">@CCC?:CCCCC@CCCEC&gt;AFDFDBEGHEAHCIGIHHGIGEGJGGIIIHFHIHGF@HGGIGJJJJJIJJJJJJJJJJJJJJJJJJJJJHHHHHFF</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">FFFCCC &nbsp;  RG:Z:1 &nbsp;  NH:i:1 &nbsp;  NM:i:0</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">5482659 &nbsp;  65 &nbsp;  CHROMOSOME_I &nbsp;  1 &nbsp;  0 &nbsp;  100M &nbsp;  CHROMOSOME_II &nbsp;  11954696 &nbsp;  0 &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">GCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCT</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">AAGCCT &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">CCCFFFFFHHGHGJJGIJHIJIJJJJJIJJJJJIJJGIJJJJJIIJIIJFJJJJJFIJJJJIIIIGIIJHHHHDEEFFFEEEEEDDDDCDCCCA</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">AA?CC: &nbsp;  RG:Z:1 &nbsp;  NH:i:1 &nbsp;  NM:i:0</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 240px;"></div><div class="CodeMirror-gutters" style="display: none; height: 240px;"></div></div></div></pre><figure class='table-figure'><table><thead><tr><th><strong><span>字段</span></strong></th><th><strong><span>含义</span></strong></th><th><strong><span>解释</span></strong></th></tr></thead><tbody><tr><td><span>3658435</span></td><td><span>QNAME</span></td><td><span>read的查询名称</span></td></tr><tr><td><span>145</span></td><td><span>FLAG</span></td><td><span>表示比对特性(如正向/反向、是否配对等)  （二进制字符）</span></td></tr><tr><td><span>CHROMOSOME_I</span></td><td><span>RNAME</span></td><td><span>参考 sequence 的名称 (染色体编号)</span></td></tr><tr><td><span>1</span></td><td><span>POS</span></td><td><span>read的起始位置 (1-based)</span></td></tr><tr><td><span>0</span></td><td><span>MAPQ</span></td><td><span>比对质量分数(0表示未评估)</span></td></tr><tr><td><span>100M</span></td><td><span>CIGAR</span></td><td><span>CIGAR 字符串，比对描述符(如100M表示100bp完美匹配)</span></td></tr><tr><td><span>CHROMOSOME_II</span></td><td><span>RNEXT</span></td><td><span>配对read比对到的参考 sequence 名称</span></td></tr><tr><td><span>2716898</span></td><td><span>PNEXT</span></td><td><span>配对 read 的起始位置</span></td></tr><tr><td><span>0</span></td><td><span>TLEN</span></td><td><span>片段长度 (若无用 0 代替)</span></td></tr><tr><td><span>GCCTA...</span></td><td><span>SEQ</span></td><td><span>读取的碱基 sequence</span></td></tr><tr><td><span>@CCC?...</span></td><td><span>QUAL</span></td><td><span>碱基质量值（ASCII 编码）</span></td></tr><tr><td><span>RG:Z:1    NH:i:1    NM:i:0</span></td><td><span>可选字段</span></td><td><span>可选字段(如RG:Z:1表示读组1，NH:i:1表示比对到1个位置，NM:i:0无错配)</span></td></tr></tbody></table></figure><p>&nbsp;</p><h3 id='flag值'><span>FLAG值</span></h3><p><strong><span>FLAG值</span></strong><span>：在 </span><strong><span>SAM/BAM</span></strong><span> 格式中，</span><code>FLAG</code><span> 是一个 </span><strong><span>位掩码（bitmask）</span></strong><span>，用不同的二进制位表示不同的比对属性。每个位的含义如下（</span><strong><span>从右到左</span></strong><span>，最低位是 </span><code>2^0</code><span>）：</span></p><figure class='table-figure'><table><thead><tr><th style='text-align:left;' ><span>位（Bit）</span></th><th style='text-align:left;' ><span>值（2^n）</span></th><th style='text-align:left;' ><span>含义（SAM FLAG 定义）</span></th></tr></thead><tbody><tr><td style='text-align:left;' ><span>0 (LSB)</span></td><td style='text-align:left;' ><span>1</span></td><td style='text-align:left;' ><span>模板有多个比对（read 比对到多个位置）</span></td></tr><tr><td style='text-align:left;' ><span>1</span></td><td style='text-align:left;' ><span>2</span></td><td style='text-align:left;' ><span>每个片段都正确比对（paired-end 比对时使用）</span></td></tr><tr><td style='text-align:left;' ><span>2</span></td><td style='text-align:left;' ><span>4</span></td><td style='text-align:left;' ><span>该 read 未比对到参考 sequence</span></td></tr><tr><td style='text-align:left;' ><span>3</span></td><td style='text-align:left;' ><span>8</span></td><td style='text-align:left;' ><span>该 read 的 </span><strong><span>mate（指paired-end另一条read）</span></strong><span>未比对到参考 sequence</span></td></tr><tr><td style='text-align:left;' ><span>4</span></td><td style='text-align:left;' ><span>16</span></td><td style='text-align:left;' ><span>该 read 是反向比对（比对到负链）</span></td></tr><tr><td style='text-align:left;' ><span>5</span></td><td style='text-align:left;' ><span>32</span></td><td style='text-align:left;' ><span>该 read 的 mate 是反向比对</span></td></tr><tr><td style='text-align:left;' ><span>6</span></td><td style='text-align:left;' ><span>64</span></td><td style='text-align:left;' ><span>这是 read 1（在 paired-end 测序中）</span></td></tr><tr><td style='text-align:left;' ><span>7 (MSB)</span></td><td style='text-align:left;' ><span>128</span></td><td style='text-align:left;' ><span>这是 read 2（在 paired-end 测序中）</span></td></tr></tbody></table></figure><p>&nbsp;</p><p><strong><span>FLAG=145 的二进制解析</span></strong></p><p><code>FLAG=145=1+16+128</code><span>，二进制是 </span><code>10010001</code><span>：顺序：</span><strong><span>从右到左</span></strong></p><ul><li><p><strong><span>Bit 0 (1)</span></strong><span>：</span><code>1</code><span> → 多比对</span></p></li><li><p><strong><span>Bit 1 (2)</span></strong><span>：</span><code>0</code><span> → single-read或paired-end中有read未正确比对</span></p></li><li><p><strong><span>Bit 2 (4)</span></strong><span>：</span><code>0</code><span> →该 read 本身比对成功</span></p></li><li><p><strong><span>Bit 3 (8)</span></strong><span>：</span><code>0</code><span> → mate 比对成功</span></p></li><li><p><strong><span>Bit 4 (16)</span></strong><span>：</span><code>1</code><span> → 该 read 是 反向比对</span></p></li><li><p><strong><span>Bit 5 (32)</span></strong><span>：</span><code>0</code><span>→ mate 是正向比对</span></p></li><li><p><strong><span>Bit 6 (64)</span></strong><span>：</span><code>0</code><span>→ 这不是 paired-end 测序中的 read 1</span></p></li><li><p><strong><span>Bit 7 (128)</span></strong><span>：</span><code>1</code><span> → 这是 paired-end 测序中的 read 2</span></p></li></ul><p>&nbsp;</p><hr /><h2 id='bam'><span>BAM</span></h2><p><strong><span>BAM</span></strong><span> 文件是通过 bgzip压缩过的</span><strong><span>SAM</span></strong><span>文件。因此二者记录的信息本质是一样的。bgzip 对文件的压缩，其结果使文件被压缩成了一系列的&#39;BGZF block&#39;单元，默认情况下，每个单元大小不超过64K。除了节省存储空间之外，另一个好处就是可以通过建立索引加速查询</span></p><p><img src="./picture/036.png" width="600"/></p><p><span>上述表格中基本信息与</span><strong><span>SAM</span></strong><span>文件的内容存在对应关系，但其每个记录的类型存在转化关系，以seq信息为例，在</span><strong><span>SAM</span></strong><span>文件中是以</span><strong><span>字符串</span></strong><span>形式存储的，在</span><strong><span>BAM</span></strong><span>文件中，该项则是将</span><strong><span>字符串集</span></strong><span>转化为了[0,15]的</span><strong><span>整数</span></strong><span>进行存储。</span></p><p>&nbsp;</p><hr /><h2 id='cram'><span>CRAM</span></h2><ul><li><p><strong><span>CRAM</span></strong><span> 文件是一种用于存储 DNA  sequence 比对数据的压缩格式，类似于 </span><strong><span>BAM</span></strong><span> 文件，为</span><strong><span>BAM</span></strong><span>的高压缩格式，使得</span><strong><span>文件体积更小</span></strong><span>，从而节省存储空间和传输时间</span></p></li><li><p><strong><span>CRAM</span></strong><span>一定会取代</span><strong><span>BAM</span></strong><span>。我想这必将很大程度上解决NGS(Next-Generation Sequencing)数据存储的问题</span></p></li></ul><ul><li><p><strong><span>CRAM</span></strong><span>大小：</span></p><p><img src="./picture/037.png"></p></li></ul><ul><li><p><span>CRAM 格式</span><strong><span>依赖参考 sequence 进行压缩和解压缩</span></strong><span>，进行转换时要</span><strong><span>指定参考 sequence 文件</span></strong></p></li></ul><p>&nbsp;</p><hr /><h2 id='bed'><span>BED</span></h2><h3 id='基本-bed-格式bed3'><span>基本 BED 格式（BED3）</span></h3><figure class='table-figure'><table><thead><tr><th><span>Chromosome</span></th><th><span>Start</span></th><th><span>End</span></th></tr></thead><tbody><tr><td><span>chr1</span></td><td><span>1000</span></td><td><span>1050</span></td></tr><tr><td><span>chr2</span></td><td><span>2000</span></td><td><span>2050</span></td></tr><tr><td><span>chr3</span></td><td><span>3000</span></td><td><span>3050</span></td></tr></tbody></table></figure><ul><li><p><strong><span>chrom</span></strong><span>：染色体名称。</span></p></li><li><p><strong><span>start</span></strong><span>：区间的起始位置（0-based）</span></p></li><li><p><strong><span>end</span></strong><span>：区间的结束位置（0-based）</span></p></li></ul><p>&nbsp;</p><h3 id='扩展-bed-格式bed6'><span>扩展 BED 格式（BED6）</span></h3><figure class='table-figure'><table><thead><tr><th><span>Chromosome</span></th><th><span>Start</span></th><th><span>End</span></th><th><span>Name</span></th><th><span>Score</span></th><th><span>Strand</span></th></tr></thead><tbody><tr><td><span>chr1</span></td><td><span>1000</span></td><td><span>1050</span></td><td><span>gene1</span></td><td><span>100</span></td><td><span>+</span></td></tr><tr><td><span>chr2</span></td><td><span>2000</span></td><td><span>2050</span></td><td><span>gene2</span></td><td><span>200</span></td><td><span>-</span></td></tr><tr><td><span>chr3</span></td><td><span>3000</span></td><td><span>3050</span></td><td><span>gene3</span></td><td><span>300</span></td><td><span>+</span></td></tr></tbody></table></figure><ul><li><p><strong><span>name</span></strong><span>：区间的名称或ID（如基因名称、 sequence ID等）。</span></p></li><li><p><strong><span>score</span></strong><span>：得分（通常为0-1000之间的整数），可以表示某种权重或置信度。</span></p></li><li><p><strong><span>strand</span></strong><span>：链方向，+ 表示正链，- 表示负链</span></p></li></ul><p>&nbsp;</p><h3 id='完整-bed-格式bed12'><span>完整 BED 格式（BED12）</span></h3><p><span>对于描述更复杂的区间，如包含多个</span><strong><span>“块”（block）</span></strong><span>的区间（如跨多个外显子的转录本）</span></p><figure class='table-figure'><table><thead><tr><th><span>Chromosome</span></th><th><span>Start Position</span></th><th><span>End Position</span></th><th><span>Name</span></th><th><span>Score</span></th><th><span>Strand</span></th><th><span>Thick Start</span></th><th><span>Thick End</span></th><th><span>Item RGB</span></th><th><span>Block Count</span></th><th><span>Block Sizes</span></th><th><span>Block Starts</span></th></tr></thead><tbody><tr><td><span>chr1</span></td><td><span>1000</span></td><td><span>1050</span></td><td><span>gene1</span></td><td><span>100</span></td><td><span>+</span></td><td><span>1000</span></td><td><span>1050</span></td><td><span>255,0,0</span></td><td><span>2</span></td><td><span>50,30</span></td><td><span>0,70</span></td></tr><tr><td><span>chr2</span></td><td><span>2000</span></td><td><span>2100</span></td><td><span>gene2</span></td><td><span>200</span></td><td><span>-</span></td><td><span>2000</span></td><td><span>2100</span></td><td><span>0,255,0</span></td><td><span>3</span></td><td><span>100,50,30</span></td><td><span>0,100,170</span></td></tr><tr><td><span>chr3</span></td><td><span>3000</span></td><td><span>3150</span></td><td><span>gene3</span></td><td><span>300</span></td><td><span>+</span></td><td><span>3000</span></td><td><span>3150</span></td><td><span>0,0,255</span></td><td><span>1</span></td><td><span>150</span></td><td><span>0</span></td></tr></tbody></table></figure><ul><li><p><strong><span>thickStart 和 thickEnd</span></strong><span>：表示区间中加粗显示的部分（例如 CDS 区间）。</span></p></li><li><p><strong><span>itemRgb</span></strong><span>：表示颜色信息，用于可视化（R,G,B格式）。</span></p></li><li><p><strong><span>blockCount</span></strong><span>：区间中“块”的数量（如外显子数量）。</span></p></li><li><p><strong><span>blockSizes</span></strong><span>：每个“块”的大小，以逗号分隔。</span></p></li><li><p><strong><span>blockStarts</span></strong><span>：指示每个“块”（如外显子）在基因组上的起始位置。</span></p></li></ul><p>&nbsp;</p><h3 id='bedpe-格式'><span>BEDPE 格式</span></h3><p><span>BEDPE 格式（BED Paired-End）是用来描述 </span><strong><span>paired-end</span></strong><span>  reads在基因组上位置的格式，它包含两个读取对的起始和结束位置等信息。通常用于分析测序数据中成对读取的配对和片段长度等。</span></p><figure class='table-figure'><table><thead><tr><th><span>Chromosome1</span></th><th><span>Start1</span></th><th><span>End1</span></th><th><span>Chromosome2</span></th><th><span>Start2</span></th><th><span>End2</span></th><th><span>Name</span></th><th><span>Score</span></th><th><span>Strand1</span></th><th><span>Strand2</span></th></tr></thead><tbody><tr><td><span>chr1</span></td><td><span>1000</span></td><td><span>1050</span></td><td><span>chr1</span></td><td><span>2000</span></td><td><span>2050</span></td><td><span>gene1</span></td><td><span>100</span></td><td><span>+</span></td><td><span>-</span></td></tr><tr><td><span>chr2</span></td><td><span>3000</span></td><td><span>3050</span></td><td><span>chr2</span></td><td><span>4000</span></td><td><span>4050</span></td><td><span>gene2</span></td><td><span>200</span></td><td><span>-</span></td><td><span>+</span></td></tr><tr><td><span>chr3</span></td><td><span>5000</span></td><td><span>5050</span></td><td><span>chr3</span></td><td><span>6000</span></td><td><span>6050</span></td><td><span>gene3</span></td><td><span>300</span></td><td><span>+</span></td><td><span>-</span></td></tr></tbody></table></figure><div class="md-alert md-alert-important important"><p><span class='md-alert-text md-alert-text-important'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M0 1.75C0 .784.784 0 1.75 0h12.5C15.216 0 16 .784 16 1.75v9.5A1.75 1.75 0 0 1 14.25 13H8.06l-2.573 2.573A1.458 1.458 0 0 1 3 14.543V13H1.75A1.75 1.75 0 0 1 0 11.25Zm1.75-.25a.25.25 0 0 0-.25.25v9.5c0 .138.112.25.25.25h2a.75.75 0 0 1 .75.75v2.19l2.72-2.72a.749.749 0 0 1 .53-.22h6.5a.25.25 0 0 0 .25-.25v-9.5a.25.25 0 0 0-.25-.25Zm7 2.25v2.5a.75.75 0 0 1-1.5 0v-2.5a.75.75 0 0 1 1.5 0ZM9 9a1 1 0 1 1-2 0 1 1 0 0 1 2 0Z"></path></svg>Important</span><br></p><p><strong><span>得分列默认使用两端比对质量的最小值</span></strong></p></div><p>&nbsp;</p><hr /><h2 id='gffgtf'><span>GFF/GTF</span></h2><ul><li><p><a href='https://blog.csdn.net/u011262253/article/details/89363809'><span>基因组注释文件(GFF,GTF)下载的五种方法</span></a></p><p>&nbsp;</p></li></ul><h3 id='gff'><span>GFF</span></h3><ul><li><p><a href='https://gmod.org/wiki/GFF3'><span>GFF</span></a><span> (General Feature Format Version 3) 是一种用于描述基因组 </span><strong><span>feature</span></strong><span>（如基因、外显子、转录本等）的文本格式。它能够有效地表示生物 sequence 中的不同特征</span></p></li><li><p><span>现在我们所使用的大部分都是第三版（</span><code>GFF3</code><span>）</span></p></li></ul><figure class='table-figure'><table><thead><tr><th><span>seqid</span></th><th><span>source</span></th><th><span>type</span></th><th><span>start</span></th><th><span>end</span></th><th><span>score</span></th><th><span>strand</span></th><th><span>phase</span></th><th><span>attributes</span></th></tr></thead><tbody><tr><td><span>chr1</span></td><td><span>.</span></td><td><span>gene</span></td><td><span>1000</span></td><td><span>5000</span></td><td><span>.</span></td><td><span>+</span></td><td><span>.</span></td><td><span>ID=gene1;Name=Gene1</span></td></tr><tr><td><span>chr1</span></td><td><span>.</span></td><td><span>mRNA</span></td><td><span>1000</span></td><td><span>5000</span></td><td><span>.</span></td><td><span>+</span></td><td><span>.</span></td><td><span>ID=mRNA1;Parent=gene1</span></td></tr><tr><td><span>chr1</span></td><td><span>.</span></td><td><span>exon</span></td><td><span>1000</span></td><td><span>2000</span></td><td><span>.</span></td><td><span>+</span></td><td><span>0</span></td><td><span>ID=exon1;Parent=mRNA1</span></td></tr><tr><td><span>chr1</span></td><td><span>.</span></td><td><span>exon</span></td><td><span>3000</span></td><td><span>5000</span></td><td><span>.</span></td><td><span>+</span></td><td><span>0</span></td><td><span>ID=exon2;Parent=mRNA1</span></td></tr></tbody></table></figure><figure class='table-figure'><table><thead><tr><th><span>列名</span></th><th><span>描述</span></th></tr></thead><tbody><tr><td><span>seqid</span></td><td><span>参考 sequence 的 ID</span></td></tr><tr><td><span>source</span></td><td><span>该特征的来源，未知时用 </span><code>.</code><span> 代替</span></td></tr><tr><td><span>type</span></td><td><span>特征类型，如 </span><code>gene</code><span>、</span><code>exon</code><span>、</span><code>CDS</code><span> 等，建议使用 SO 术语</span></td></tr><tr><td><span>start</span></td><td><span>起始坐标（1-based）</span></td></tr><tr><td><span>end</span></td><td><span>终止坐标</span></td></tr><tr><td><span>score</span></td><td><span>量化得分，空值用 </span><code>.</code><span> 代替</span></td></tr><tr><td><span>strand</span></td><td><code>+</code><span> 表示正链，</span><code>-</code><span> 表示负链，</span><code>.</code><span> 表示无方向要求</span></td></tr><tr><td><span>phase</span></td><td><span>编码相位，CDS 需要此列，当前 CDS 片段的起始碱基在密码子中的偏移量。0&quot;、&quot;1 &quot;或 &quot;2 &quot;中的一个。 0 &quot;表示特征的第一个碱基是密码子的第一个碱基，&quot;1 &quot;表示第二个碱基是密码子的第一个碱基，以此类推</span></td></tr><tr><td><span>attributes</span></td><td><span>属性列表，格式为 </span><code>key=value</code><span>，不同属性用 </span><code>;</code><span> 分隔</span></td></tr></tbody></table></figure><p>&nbsp;</p><h3 id='gtf'><span>GTF</span></h3><ul><li><p><code>GTF</code><span>(gene transfer format)，主要是用来对基因进行注释</span></p></li><li><p><span>当前所广泛使用的GTF格式为第二版（</span><code>GTF2</code><span>）</span></p></li></ul><figure class='table-figure'><table><thead><tr><th><span>seqname</span></th><th><span>source</span></th><th><span>feature</span></th><th><span>start</span></th><th><span>end</span></th><th><span>score</span></th><th><span>strand</span></th><th><span>frame</span></th><th><span>attributes</span></th></tr></thead><tbody><tr><td><span>AB000381</span></td><td><span>Twinscan</span></td><td><span>CDS</span></td><td><span>700</span></td><td><span>707</span></td><td><span>.</span></td><td><span>+</span></td><td><span>2</span></td><td><span>gene_id &quot;AB000381.000&quot;; transcript_id &quot;AB000381.000.1&quot;;</span></td></tr><tr><td><span>AB000381</span></td><td><span>Twinscan</span></td><td><span>exon</span></td><td><span>900</span></td><td><span>1000</span></td><td><span>.</span></td><td><span>+</span></td><td><span>.</span></td><td><span>gene_id &quot;AB000381.000&quot;; transcript_id &quot;AB000381.000.1&quot;;</span></td></tr><tr><td><span>AB000381</span></td><td><span>Twinscan</span></td><td><span>start_codon</span></td><td><span>380</span></td><td><span>382</span></td><td><span>.</span></td><td><span>+</span></td><td><span>0</span></td><td><span>gene_id &quot;AB000381.000&quot;; transcript_id &quot;AB000381.000.1&quot;;</span></td></tr><tr><td><span>AB000381</span></td><td><span>Twinscan</span></td><td><span>stop_codon</span></td><td><span>708</span></td><td><span>710</span></td><td><span>.</span></td><td><span>+</span></td><td><span>0</span></td><td><span>gene_id &quot;AB000381.000&quot;; transcript_id &quot;AB000381.000.1&quot;;</span></td></tr></tbody></table></figure><figure class='table-figure'><table><thead><tr><th><span>列名</span></th><th><span>描述</span></th></tr></thead><tbody><tr><td><span>seqname</span></td><td><span>sequence  ID，通常是染色体 ID 或 Contig ID</span></td></tr><tr><td><span>source</span></td><td><span>该特征的来源，未知时用 </span><code>.</code><span> 代替</span></td></tr><tr><td><span>type</span></td><td><span>特征类型，特征类型，如 </span><code>CDS</code><span>、</span><code>start_codon</code><span>(起始密码子)、</span><code>stop_codon</code><span>（终止密码子） （</span><strong><span>这一列必须注明</span></strong><span>）</span></td></tr><tr><td><span>start</span></td><td><span>起始坐标（1-based）</span></td></tr><tr><td><span>end</span></td><td><span>终止坐标</span></td></tr><tr><td><span>score</span></td><td><span>量化得分，空值用 </span><code>.</code><span> 代替</span></td></tr><tr><td><span>strand</span></td><td><code>+</code><span> 表示正链，</span><code>-</code><span> 表示负链，</span><code>.</code><span> 表示无方向要求</span></td></tr><tr><td><span>frame</span></td><td><span>读码框（0、1 或 2，仅适用于 </span><code>CDS</code><span>）</span></td></tr><tr><td><span>attributes</span></td><td><span>其他信息（键值对，</span><code>key &quot;value&quot;;</code><span> 格式）</span></td></tr></tbody></table></figure><p>&nbsp;</p><h3 id='different-from-gff-of-gtf'><span>Different from GFF of GTF</span></h3><figure class='table-figure'><table><thead><tr><th><span>对比项</span></th><th><strong><code>GFF3</code></strong></th><th><strong><code>GTF2</code></strong></th></tr></thead><tbody><tr><td><span>type/feature</span></td><td><span>必须注明</span></td><td><span>可以是任意名称</span></td></tr><tr><td><span>attributes</span></td><td><code>key=value</code></td><td><code>key &quot;value&quot;</code></td></tr><tr><td><span>attributes 层级</span></td><td><span>明确层级关系（gene → mRNA → exon/CDS）</span></td><td><span>层级关系相对不严格</span></td></tr><tr><td><span>支持的特征</span></td><td><span>更丰富（gene, mRNA, exon, CDS, ncRNA, repeat_region 等）</span></td><td><span>主要针对转录组数据（gene, transcript, exon, CDS）</span></td></tr></tbody></table></figure><p>&nbsp;</p><hr /><h2 id='vcf'><span>VCF</span></h2><ul><li><p><a href='https://samtools.github.io/hts-specs/VCFv4.5.pdf'><span>VCF</span></a><span> (Variant Call Format) 是一种用于存储变异信息（如 SNP、插入、缺失等）的文本格式，通常用于表示基因组 sequence 的变化。</span></p></li></ul><figure class='table-figure'><table><thead><tr><th><span>#CHROM</span></th><th><span>POS</span></th><th><span>ID</span></th><th><span>REF</span></th><th><span>ALT</span></th><th><span>QUAL</span></th><th><span>FILTER</span></th><th><span>INFO</span></th><th><span>FORMAT</span></th><th><span>SAMPLE1</span></th><th><span>SAMPLE2</span></th></tr></thead><tbody><tr><td><span>chr1</span></td><td><span>1000</span></td><td><span>.</span></td><td><span>A</span></td><td><span>G</span></td><td><span>99</span></td><td><span>PASS</span></td><td><span>DP=20</span></td><td><span>GT:DP</span></td><td><span>0/1:10</span></td><td><span>0/0:20</span></td></tr><tr><td><span>chr1</span></td><td><span>1500</span></td><td><span>.</span></td><td><span>C</span></td><td><span>T</span></td><td><span>99</span></td><td><span>PASS</span></td><td><span>DP=30</span></td><td><span>GT:DP</span></td><td><span>0/1:15</span></td><td><span>0/1:15</span></td></tr></tbody></table></figure><figure class='table-figure'><table><thead><tr><th><span>字段</span></th><th><span>含义</span></th></tr></thead><tbody><tr><td><strong><span>#CHROM</span></strong></td><td><span>染色体名称。</span></td></tr><tr><td><strong><span>POS</span></strong></td><td><span>变异在染色体上的位置（1-based）。</span></td></tr><tr><td><strong><span>ID</span></strong></td><td><span>变异的标识符（如果没有则为 </span><code>.</code><span>）。</span></td></tr><tr><td><strong><span>REF</span></strong></td><td><span>参考 sequence 中的碱基。</span></td></tr><tr><td><strong><span>ALT</span></strong></td><td><span>变异碱基（可以有多个）。</span></td></tr><tr><td><strong><span>QUAL</span></strong></td><td><span>变异质量分数。</span></td></tr><tr><td><strong><span>FILTER</span></strong></td><td><span>过滤信息，表示变异是否通过了质量过滤。</span></td></tr><tr><td><strong><span>INFO</span></strong></td><td><span>附加信息字段，如覆盖深度 </span><code>DP</code><span>。</span></td></tr><tr><td><strong><span>FORMAT</span></strong></td><td><span>样本信息的格式。</span></td></tr><tr><td><strong><span>GT</span></strong></td><td><span>基因型（Genotype），表示样本的等位基因组合，例如 0/0（纯合参考）、0/1（杂合）或 1/1（纯合变异）。</span></td></tr><tr><td><strong><span>DP</span></strong></td><td><span>覆盖深度（Depth of Coverage），表示该位置的测序读取数量。</span></td></tr><tr><td><strong><span>其他信息</span></strong></td><td><span>可能还包括如 </span><strong><span>GQ</span></strong><span>（基因型质量）、</span><strong><span>PL</span></strong><span>（后验概率）等。</span></td></tr><tr><td><strong><span>SAMPLE1</span></strong></td><td><span>表示第一个样本的信息，对应 FORMAT 字段定义。</span></td></tr><tr><td><strong><span>SAMPLE2</span></strong></td><td><span>表示第二个样本的信息，对应FORMAT 字段定义。</span></td></tr></tbody></table></figure><p>&nbsp;</p><hr /><h2 id='chromsizes'><span>chrom.sizes</span></h2><p><span>染色体长度信息文件(</span><strong><span>chrom.sizes</span></strong><span>)，该文件保存了基因组中的染色体名称已经对应的长度</span></p><figure class='table-figure'><table><thead><tr><th><span>Chromosome</span></th><th><span>Size</span></th></tr></thead><tbody><tr><td><span>chr1</span></td><td><span>248956422</span></td></tr><tr><td><span>chr2</span></td><td><span>242193529</span></td></tr><tr><td><span>chrX</span></td><td><span>156040895</span></td></tr><tr><td><span>chrY</span></td><td><span>57227415</span></td></tr><tr><td><span>chrM</span></td><td><span>16569</span></td></tr></tbody></table></figure><p>&nbsp;</p><p><span>There are three ways to obtain the </span><strong><span>chrom.sizes</span></strong><span> file</span></p><ol start='' ><li><p><span>通过bam文件获取chrom.sizes(</span><strong><span>recommend</span></strong><span>)</span></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation"><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">samtools view <span class="cm-attribute">-H</span> &lt;bam&gt; | &nbsp;<span class="cm-builtin">grep</span> <span class="cm-string">'^@SQ'</span> | <span class="cm-builtin">awk</span> <span class="cm-attribute">-F</span><span class="cm-string">'\t'</span> <span class="cm-string">'{split($3,a,":"); gsub("SN:","",$2); print $2"\t"a[2]}'</span></span></pre></div></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 22px;"></div><div class="CodeMirror-gutters" style="display: none; height: 22px;"></div></div></div></pre><p>&nbsp;</p></li><li><p><span>从UCSC下载，适用于UCSC数据库中已有的物种</span></p><p><a href='https://hgdownload.soe.ucsc.edu/downloads.html' target='_blank' class='url'>https://hgdownload.soe.ucsc.edu/downloads.html</a><span>  找到比对时的index对应的chrom.sizes下载</span></p><div class="md-alert md-alert-warning warning"><p><span class='md-alert-text md-alert-text-warning'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M6.457 1.047c.659-1.234 2.427-1.234 3.086 0l6.082 11.378A1.75 1.75 0 0 1 14.082 15H1.918a1.75 1.75 0 0 1-1.543-2.575Zm1.763.707a.25.25 0 0 0-.44 0L1.698 13.132a.25.25 0 0 0 .22.368h12.164a.25.25 0 0 0 .22-.368Zm.53 3.996v2.5a.75.75 0 0 1-1.5 0v-2.5a.75.75 0 0 1 1.5 0ZM9 11a1 1 0 1 1-2 0 1 1 0 0 1 2 0Z"></path></svg>Warning</span><br></p><p><strong><span>该方法获取的染色体格式(ucsc格式)可能与bed文件不一样</span></strong></p><p>例如：chrom.sizes中：</p><p><img src="./picture/051.png" width="200"/></p><p>而bed文件中：</p><p><img src="./picture/052.png" width="250"/></p><ul><li><p><span>可以发现一个使用的是裸染色体编号（如 1、2），而另一个使用的是带前缀的染色体名称（如 chr1、chr2）。</span></p></li><li><p><span>为了使 </span><code>bedtools genomecov</code><span> 正常运行，必须保证 BED 文件和 .chrom.sizes 文件的染色体名称一致。</span></p></li><li><p><span>可以通过简单的命令修改 .chrom.sizes 文件，将 chr 前缀去掉</span></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.98999px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation"><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># sed 's/^chr//'：去除每一行开头的 chr 前缀</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-builtin">sed</span> <span class="cm-string">'s/^chr//'</span> xxx.chrom.sizes &gt; xxx.fixed.chrom.sizes</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 44px;"></div><div class="CodeMirror-gutters" style="display: none; height: 44px;"></div></div></div></pre></li></ul></div><p>&nbsp;</p></li><li><p><span>通过fasta文件获取chrom.sizes</span></p><p><code>samtools faidx</code><span>命令可以获取fasta文件中的序列长度信息，从其生成的后缀为fai的文件中可以获得chrom.sizes文件。</span></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation"><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">samtools faidx xxx.fa</span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-builtin">cut</span> <span class="cm-attribute">-f1</span>,2 xxx.fa.fai &gt; xxx.chrom.sizes</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 66px;"></div><div class="CodeMirror-gutters" style="display: none; height: 66px;"></div></div></div></pre><p>&nbsp;</p></li></ol><hr /><h2 id='bedgraph-and-bigwig'><span>Bedgraph and BigWIG</span></h2><h3 id='bedgraph'><span>Bedgraph </span></h3><ul><li><p><span>Bedgraph 是一种 </span><strong><span>用于表示基因组上连续信号轨迹</span></strong><span> 的</span><strong><span>文本</span></strong><span>格式</span></p></li><li><p><span>BigWig 是从 bed/bam 文件统计每个</span><strong><span>区间</span></strong><span>的 read 覆盖度（coverage）得来的</span></p></li></ul><figure class='table-figure'><table><thead><tr><th><span>Chromosome</span></th><th><span>Start</span></th><th><span>End</span></th><th><span>Coverage</span></th></tr></thead><tbody><tr><td><span>chr1</span></td><td><span>1000</span></td><td><span>2000</span></td><td><span>5</span></td></tr><tr><td><span>chr1</span></td><td><span>2000</span></td><td><span>3000</span></td><td><span>8</span></td></tr><tr><td><span>chr2</span></td><td><span>1500</span></td><td><span>2500</span></td><td><span>10</span></td></tr></tbody></table></figure><ul><li><p><strong><span>start</span></strong><span>: 基因组上</span><strong><span>区间</span></strong><span>的起始坐标。</span></p></li><li><p><strong><span>end</span></strong><span>: 基因组上</span><strong><span>区间</span></strong><span>的结束坐标。</span></p></li><li><p><strong><span>Coverage</span></strong><span>: 表示该区域的平均覆盖度。</span></p></li></ul><p>&nbsp;</p><div class="md-alert md-alert-tip tip"><p><span class='md-alert-text md-alert-text-tip'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M8 1.5c-2.363 0-4 1.69-4 3.75 0 .984.424 1.625.984 2.304l.214.253c.223.264.47.556.673.848.284.411.537.896.621 1.49a.75.75 0 0 1-1.484.211c-.04-.282-.163-.547-.37-.847a8.456 8.456 0 0 0-.542-.68c-.084-.1-.173-.205-.268-.32C3.201 7.75 2.5 6.766 2.5 5.25 2.5 2.31 4.863 0 8 0s5.5 2.31 5.5 5.25c0 1.516-.701 2.5-1.328 3.259-.095.115-.184.22-.268.319-.207.245-.383.453-.541.681-.208.3-.33.565-.37.847a.751.751 0 0 1-1.485-.212c.084-.593.337-1.078.621-1.489.203-.292.45-.584.673-.848.075-.088.147-.173.213-.253.561-.679.985-1.32.985-2.304 0-2.06-1.637-3.75-4-3.75ZM5.75 12h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5ZM6 15.25a.75.75 0 0 1 .75-.75h2.5a.75.75 0 0 1 0 1.5h-2.5a.75.75 0 0 1-.75-.75Z"></path></svg>Tip</span><br></p><p><span>Calculate </span><strong><span>Coverage</span></strong><span>：</span></p><div contenteditable="false" spellcheck="false" class="mathjax-block md-end-block md-math-block md-rawblock" id="mathjax-n1228" cid="n1228" mdtype="math_block" data-math-tag-before="0" data-math-tag-after="0" data-math-labels="[]"><div class="md-math-container"><mjx-container class="MathJax" jax="SVG" display="true" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="34.718ex" height="5.285ex" role="img" focusable="false" viewBox="0 -1426 15345.3 2336" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -2.059ex;"><defs><path id="MJX-1-TEX-N-43" d="M56 342Q56 428 89 500T174 615T283 681T391 705Q394 705 400 705T408 704Q499 704 569 636L582 624L612 663Q639 700 643 704Q644 704 647 704T653 705H657Q660 705 666 699V419L660 413H626Q620 419 619 430Q610 512 571 572T476 651Q457 658 426 658Q322 658 252 588Q173 509 173 342Q173 221 211 151Q232 111 263 84T328 45T384 29T428 24Q517 24 571 93T626 244Q626 251 632 257H660L666 251V236Q661 133 590 56T403 -21Q262 -21 159 83T56 342Z"></path><path id="MJX-1-TEX-N-6F" d="M28 214Q28 309 93 378T250 448Q340 448 405 380T471 215Q471 120 407 55T250 -10Q153 -10 91 57T28 214ZM250 30Q372 30 372 193V225V250Q372 272 371 288T364 326T348 362T317 390T268 410Q263 411 252 411Q222 411 195 399Q152 377 139 338T126 246V226Q126 130 145 91Q177 30 250 30Z"></path><path id="MJX-1-TEX-N-76" d="M338 431Q344 429 422 429Q479 429 503 431H508V385H497Q439 381 423 345Q421 341 356 172T288 -2Q283 -11 263 -11Q244 -11 239 -2Q99 359 98 364Q93 378 82 381T43 385H19V431H25L33 430Q41 430 53 430T79 430T104 429T122 428Q217 428 232 431H240V385H226Q187 384 184 370Q184 366 235 234L286 102L377 341V349Q377 363 367 372T349 383T335 385H331V431H338Z"></path><path id="MJX-1-TEX-N-65" d="M28 218Q28 273 48 318T98 391T163 433T229 448Q282 448 320 430T378 380T406 316T415 245Q415 238 408 231H126V216Q126 68 226 36Q246 30 270 30Q312 30 342 62Q359 79 369 104L379 128Q382 131 395 131H398Q415 131 415 121Q415 117 412 108Q393 53 349 21T250 -11Q155 -11 92 58T28 218ZM333 275Q322 403 238 411H236Q228 411 220 410T195 402T166 381T143 340T127 274V267H333V275Z"></path><path id="MJX-1-TEX-N-72" d="M36 46H50Q89 46 97 60V68Q97 77 97 91T98 122T98 161T98 203Q98 234 98 269T98 328L97 351Q94 370 83 376T38 385H20V408Q20 431 22 431L32 432Q42 433 60 434T96 436Q112 437 131 438T160 441T171 442H174V373Q213 441 271 441H277Q322 441 343 419T364 373Q364 352 351 337T313 322Q288 322 276 338T263 372Q263 381 265 388T270 400T273 405Q271 407 250 401Q234 393 226 386Q179 341 179 207V154Q179 141 179 127T179 101T180 81T180 66V61Q181 59 183 57T188 54T193 51T200 49T207 48T216 47T225 47T235 46T245 46H276V0H267Q249 3 140 3Q37 3 28 0H20V46H36Z"></path><path id="MJX-1-TEX-N-61" d="M137 305T115 305T78 320T63 359Q63 394 97 421T218 448Q291 448 336 416T396 340Q401 326 401 309T402 194V124Q402 76 407 58T428 40Q443 40 448 56T453 109V145H493V106Q492 66 490 59Q481 29 455 12T400 -6T353 12T329 54V58L327 55Q325 52 322 49T314 40T302 29T287 17T269 6T247 -2T221 -8T190 -11Q130 -11 82 20T34 107Q34 128 41 147T68 188T116 225T194 253T304 268H318V290Q318 324 312 340Q290 411 215 411Q197 411 181 410T156 406T148 403Q170 388 170 359Q170 334 154 320ZM126 106Q126 75 150 51T209 26Q247 26 276 49T315 109Q317 116 318 175Q318 233 317 233Q309 233 296 232T251 223T193 203T147 166T126 106Z"></path><path id="MJX-1-TEX-N-67" d="M329 409Q373 453 429 453Q459 453 472 434T485 396Q485 382 476 371T449 360Q416 360 412 390Q410 404 415 411Q415 412 416 414V415Q388 412 363 393Q355 388 355 386Q355 385 359 381T368 369T379 351T388 325T392 292Q392 230 343 187T222 143Q172 143 123 171Q112 153 112 133Q112 98 138 81Q147 75 155 75T227 73Q311 72 335 67Q396 58 431 26Q470 -13 470 -72Q470 -139 392 -175Q332 -206 250 -206Q167 -206 107 -175Q29 -140 29 -75Q29 -39 50 -15T92 18L103 24Q67 55 67 108Q67 155 96 193Q52 237 52 292Q52 355 102 398T223 442Q274 442 318 416L329 409ZM299 343Q294 371 273 387T221 404Q192 404 171 388T145 343Q142 326 142 292Q142 248 149 227T179 192Q196 182 222 182Q244 182 260 189T283 207T294 227T299 242Q302 258 302 292T299 343ZM403 -75Q403 -50 389 -34T348 -11T299 -2T245 0H218Q151 0 138 -6Q118 -15 107 -34T95 -74Q95 -84 101 -97T122 -127T170 -155T250 -167Q319 -167 361 -139T403 -75Z"></path><path id="MJX-1-TEX-N-3D" d="M56 347Q56 360 70 367H707Q722 359 722 347Q722 336 708 328L390 327H72Q56 332 56 347ZM56 153Q56 168 72 173H708Q722 163 722 153Q722 140 707 133H70Q56 140 56 153Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="mtext"><use data-c="43" xlink:href="#MJX-1-TEX-N-43"></use><use data-c="6F" xlink:href="#MJX-1-TEX-N-6F" transform="translate(722,0)"></use><use data-c="76" xlink:href="#MJX-1-TEX-N-76" transform="translate(1222,0)"></use><use data-c="65" xlink:href="#MJX-1-TEX-N-65" transform="translate(1750,0)"></use><use data-c="72" xlink:href="#MJX-1-TEX-N-72" transform="translate(2194,0)"></use><use data-c="61" xlink:href="#MJX-1-TEX-N-61" transform="translate(2586,0)"></use><use data-c="67" xlink:href="#MJX-1-TEX-N-67" transform="translate(3086,0)"></use><use data-c="65" xlink:href="#MJX-1-TEX-N-65" transform="translate(3586,0)"></use></g><g data-mml-node="mo" transform="translate(4307.8,0)"><use data-c="3D" xlink:href="#MJX-1-TEX-N-3D"></use></g><g data-mml-node="mfrac" transform="translate(5363.6,0)"><g data-mml-node="mtext" transform="translate(2388.6,676)"><text data-variant="normal" transform="scale(1,-1)" font-size="852.2px" font-family="serif">总</text><text data-variant="normal" transform="translate(867.4,0) scale(1,-1)" font-size="852.2px" font-family="serif">测</text><text data-variant="normal" transform="translate(1734.9,0) scale(1,-1)" font-size="852.2px" font-family="serif">序</text><text data-variant="normal" transform="translate(2602.3,0) scale(1,-1)" font-size="852.2px" font-family="serif">碱</text><text data-variant="normal" transform="translate(3469.7,0) scale(1,-1)" font-size="852.2px" font-family="serif">基</text><text data-variant="normal" transform="translate(4337.2,0) scale(1,-1)" font-size="852.2px" font-family="serif">数</text></g><g data-mml-node="mtext" transform="translate(220,-710)"><text data-variant="normal" transform="scale(1,-1)" font-size="852.2px" font-family="serif">目</text><text data-variant="normal" transform="translate(867.4,0) scale(1,-1)" font-size="852.2px" font-family="serif">标</text><text data-variant="normal" transform="translate(1734.9,0) scale(1,-1)" font-size="852.2px" font-family="serif">基</text><text data-variant="normal" transform="translate(2602.3,0) scale(1,-1)" font-size="852.2px" font-family="serif">因</text><text data-variant="normal" transform="translate(3469.7,0) scale(1,-1)" font-size="852.2px" font-family="serif">组</text><text data-variant="normal" transform="translate(4337.2,0) scale(1,-1)" font-size="852.2px" font-family="serif">区</text><text data-variant="normal" transform="translate(5204.6,0) scale(1,-1)" font-size="852.2px" font-family="serif">域</text><text data-variant="normal" transform="translate(6072,0) scale(1,-1)" font-size="852.2px" font-family="serif">的</text><text data-variant="normal" transform="translate(6939.5,0) scale(1,-1)" font-size="852.2px" font-family="serif">碱</text><text data-variant="normal" transform="translate(7806.9,0) scale(1,-1)" font-size="852.2px" font-family="serif">基</text><text data-variant="normal" transform="translate(8674.3,0) scale(1,-1)" font-size="852.2px" font-family="serif">数</text></g><rect width="9741.8" height="60" x="120" y="220"></rect></g></g></g></svg><mjx-assistive-mml unselectable="on" display="block"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><mtext>Coverage</mtext><mo>=</mo><mfrac><mtext>总测序碱基数</mtext><mtext>目标基因组区域的碱基数</mtext></mfrac></math></mjx-assistive-mml></mjx-container></div></div><ul><li><p><strong><span>总测序碱基数</span></strong><span>：该区域测序 reads 的总碱基数（测序深度乘以读长）</span></p></li><li><p><strong><span>目标基因组区域的碱基数</span></strong><span>：该区域的碱基总长度。</span></p></li></ul><p><span>Example of calculate </span><strong><span>Coverage</span></strong><span>  ：</span></p><ul><li><p><span>该区域存在50 条reads</span></p></li><li><p><span>每条 reads 长度为 150 bp</span></p></li><li><p><span>目标基因组区域长度为 1000 bp</span></p></li></ul><div contenteditable="false" spellcheck="false" class="mathjax-block md-end-block md-math-block md-rawblock" id="mathjax-n1242" cid="n1242" mdtype="math_block" data-math-tag-before="0" data-math-tag-after="0" data-math-labels="[]"><div class="md-math-container"><mjx-container class="MathJax" jax="SVG" display="true" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="27.461ex" height="4.638ex" role="img" focusable="false" viewBox="0 -1342 12137.6 2050" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -1.602ex;"><defs><path id="MJX-2-TEX-N-43" d="M56 342Q56 428 89 500T174 615T283 681T391 705Q394 705 400 705T408 704Q499 704 569 636L582 624L612 663Q639 700 643 704Q644 704 647 704T653 705H657Q660 705 666 699V419L660 413H626Q620 419 619 430Q610 512 571 572T476 651Q457 658 426 658Q322 658 252 588Q173 509 173 342Q173 221 211 151Q232 111 263 84T328 45T384 29T428 24Q517 24 571 93T626 244Q626 251 632 257H660L666 251V236Q661 133 590 56T403 -21Q262 -21 159 83T56 342Z"></path><path id="MJX-2-TEX-N-6F" d="M28 214Q28 309 93 378T250 448Q340 448 405 380T471 215Q471 120 407 55T250 -10Q153 -10 91 57T28 214ZM250 30Q372 30 372 193V225V250Q372 272 371 288T364 326T348 362T317 390T268 410Q263 411 252 411Q222 411 195 399Q152 377 139 338T126 246V226Q126 130 145 91Q177 30 250 30Z"></path><path id="MJX-2-TEX-N-76" d="M338 431Q344 429 422 429Q479 429 503 431H508V385H497Q439 381 423 345Q421 341 356 172T288 -2Q283 -11 263 -11Q244 -11 239 -2Q99 359 98 364Q93 378 82 381T43 385H19V431H25L33 430Q41 430 53 430T79 430T104 429T122 428Q217 428 232 431H240V385H226Q187 384 184 370Q184 366 235 234L286 102L377 341V349Q377 363 367 372T349 383T335 385H331V431H338Z"></path><path id="MJX-2-TEX-N-65" d="M28 218Q28 273 48 318T98 391T163 433T229 448Q282 448 320 430T378 380T406 316T415 245Q415 238 408 231H126V216Q126 68 226 36Q246 30 270 30Q312 30 342 62Q359 79 369 104L379 128Q382 131 395 131H398Q415 131 415 121Q415 117 412 108Q393 53 349 21T250 -11Q155 -11 92 58T28 218ZM333 275Q322 403 238 411H236Q228 411 220 410T195 402T166 381T143 340T127 274V267H333V275Z"></path><path id="MJX-2-TEX-N-72" d="M36 46H50Q89 46 97 60V68Q97 77 97 91T98 122T98 161T98 203Q98 234 98 269T98 328L97 351Q94 370 83 376T38 385H20V408Q20 431 22 431L32 432Q42 433 60 434T96 436Q112 437 131 438T160 441T171 442H174V373Q213 441 271 441H277Q322 441 343 419T364 373Q364 352 351 337T313 322Q288 322 276 338T263 372Q263 381 265 388T270 400T273 405Q271 407 250 401Q234 393 226 386Q179 341 179 207V154Q179 141 179 127T179 101T180 81T180 66V61Q181 59 183 57T188 54T193 51T200 49T207 48T216 47T225 47T235 46T245 46H276V0H267Q249 3 140 3Q37 3 28 0H20V46H36Z"></path><path id="MJX-2-TEX-N-61" d="M137 305T115 305T78 320T63 359Q63 394 97 421T218 448Q291 448 336 416T396 340Q401 326 401 309T402 194V124Q402 76 407 58T428 40Q443 40 448 56T453 109V145H493V106Q492 66 490 59Q481 29 455 12T400 -6T353 12T329 54V58L327 55Q325 52 322 49T314 40T302 29T287 17T269 6T247 -2T221 -8T190 -11Q130 -11 82 20T34 107Q34 128 41 147T68 188T116 225T194 253T304 268H318V290Q318 324 312 340Q290 411 215 411Q197 411 181 410T156 406T148 403Q170 388 170 359Q170 334 154 320ZM126 106Q126 75 150 51T209 26Q247 26 276 49T315 109Q317 116 318 175Q318 233 317 233Q309 233 296 232T251 223T193 203T147 166T126 106Z"></path><path id="MJX-2-TEX-N-67" d="M329 409Q373 453 429 453Q459 453 472 434T485 396Q485 382 476 371T449 360Q416 360 412 390Q410 404 415 411Q415 412 416 414V415Q388 412 363 393Q355 388 355 386Q355 385 359 381T368 369T379 351T388 325T392 292Q392 230 343 187T222 143Q172 143 123 171Q112 153 112 133Q112 98 138 81Q147 75 155 75T227 73Q311 72 335 67Q396 58 431 26Q470 -13 470 -72Q470 -139 392 -175Q332 -206 250 -206Q167 -206 107 -175Q29 -140 29 -75Q29 -39 50 -15T92 18L103 24Q67 55 67 108Q67 155 96 193Q52 237 52 292Q52 355 102 398T223 442Q274 442 318 416L329 409ZM299 343Q294 371 273 387T221 404Q192 404 171 388T145 343Q142 326 142 292Q142 248 149 227T179 192Q196 182 222 182Q244 182 260 189T283 207T294 227T299 242Q302 258 302 292T299 343ZM403 -75Q403 -50 389 -34T348 -11T299 -2T245 0H218Q151 0 138 -6Q118 -15 107 -34T95 -74Q95 -84 101 -97T122 -127T170 -155T250 -167Q319 -167 361 -139T403 -75Z"></path><path id="MJX-2-TEX-N-3D" d="M56 347Q56 360 70 367H707Q722 359 722 347Q722 336 708 328L390 327H72Q56 332 56 347ZM56 153Q56 168 72 173H708Q722 163 722 153Q722 140 707 133H70Q56 140 56 153Z"></path><path id="MJX-2-TEX-N-35" d="M164 157Q164 133 148 117T109 101H102Q148 22 224 22Q294 22 326 82Q345 115 345 210Q345 313 318 349Q292 382 260 382H254Q176 382 136 314Q132 307 129 306T114 304Q97 304 95 310Q93 314 93 485V614Q93 664 98 664Q100 666 102 666Q103 666 123 658T178 642T253 634Q324 634 389 662Q397 666 402 666Q410 666 410 648V635Q328 538 205 538Q174 538 149 544L139 546V374Q158 388 169 396T205 412T256 420Q337 420 393 355T449 201Q449 109 385 44T229 -22Q148 -22 99 32T50 154Q50 178 61 192T84 210T107 214Q132 214 148 197T164 157Z"></path><path id="MJX-2-TEX-N-30" d="M96 585Q152 666 249 666Q297 666 345 640T423 548Q460 465 460 320Q460 165 417 83Q397 41 362 16T301 -15T250 -22Q224 -22 198 -16T137 16T82 83Q39 165 39 320Q39 494 96 585ZM321 597Q291 629 250 629Q208 629 178 597Q153 571 145 525T137 333Q137 175 145 125T181 46Q209 16 250 16Q290 16 318 46Q347 76 354 130T362 333Q362 478 354 524T321 597Z"></path><path id="MJX-2-TEX-N-D7" d="M630 29Q630 9 609 9Q604 9 587 25T493 118L389 222L284 117Q178 13 175 11Q171 9 168 9Q160 9 154 15T147 29Q147 36 161 51T255 146L359 250L255 354Q174 435 161 449T147 471Q147 480 153 485T168 490Q173 490 175 489Q178 487 284 383L389 278L493 382Q570 459 587 475T609 491Q630 491 630 471Q630 464 620 453T522 355L418 250L522 145Q606 61 618 48T630 29Z"></path><path id="MJX-2-TEX-N-31" d="M213 578L200 573Q186 568 160 563T102 556H83V602H102Q149 604 189 617T245 641T273 663Q275 666 285 666Q294 666 302 660V361L303 61Q310 54 315 52T339 48T401 46H427V0H416Q395 3 257 3Q121 3 100 0H88V46H114Q136 46 152 46T177 47T193 50T201 52T207 57T213 61V578Z"></path><path id="MJX-2-TEX-N-37" d="M55 458Q56 460 72 567L88 674Q88 676 108 676H128V672Q128 662 143 655T195 646T364 644H485V605L417 512Q408 500 387 472T360 435T339 403T319 367T305 330T292 284T284 230T278 162T275 80Q275 66 275 52T274 28V19Q270 2 255 -10T221 -22Q210 -22 200 -19T179 0T168 40Q168 198 265 368Q285 400 349 489L395 552H302Q128 552 119 546Q113 543 108 522T98 479L95 458V455H55V458Z"></path><path id="MJX-2-TEX-N-2E" d="M78 60Q78 84 95 102T138 120Q162 120 180 104T199 61Q199 36 182 18T139 0T96 17T78 60Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="mtext"><use data-c="43" xlink:href="#MJX-2-TEX-N-43"></use><use data-c="6F" xlink:href="#MJX-2-TEX-N-6F" transform="translate(722,0)"></use><use data-c="76" xlink:href="#MJX-2-TEX-N-76" transform="translate(1222,0)"></use><use data-c="65" xlink:href="#MJX-2-TEX-N-65" transform="translate(1750,0)"></use><use data-c="72" xlink:href="#MJX-2-TEX-N-72" transform="translate(2194,0)"></use><use data-c="61" xlink:href="#MJX-2-TEX-N-61" transform="translate(2586,0)"></use><use data-c="67" xlink:href="#MJX-2-TEX-N-67" transform="translate(3086,0)"></use><use data-c="65" xlink:href="#MJX-2-TEX-N-65" transform="translate(3586,0)"></use></g><g data-mml-node="mo" transform="translate(4307.8,0)"><use data-c="3D" xlink:href="#MJX-2-TEX-N-3D"></use></g><g data-mml-node="mfrac" transform="translate(5363.6,0)"><g data-mml-node="mrow" transform="translate(220,676)"><g data-mml-node="mn"><use data-c="35" xlink:href="#MJX-2-TEX-N-35"></use><use data-c="30" xlink:href="#MJX-2-TEX-N-30" transform="translate(500,0)"></use></g><g data-mml-node="mo" transform="translate(1222.2,0)"><use data-c="D7" xlink:href="#MJX-2-TEX-N-D7"></use></g><g data-mml-node="mn" transform="translate(2222.4,0)"><use data-c="31" xlink:href="#MJX-2-TEX-N-31"></use><use data-c="35" xlink:href="#MJX-2-TEX-N-35" transform="translate(500,0)"></use><use data-c="30" xlink:href="#MJX-2-TEX-N-30" transform="translate(1000,0)"></use></g></g><g data-mml-node="mn" transform="translate(1081.2,-686)"><use data-c="31" xlink:href="#MJX-2-TEX-N-31"></use><use data-c="30" xlink:href="#MJX-2-TEX-N-30" transform="translate(500,0)"></use><use data-c="30" xlink:href="#MJX-2-TEX-N-30" transform="translate(1000,0)"></use><use data-c="30" xlink:href="#MJX-2-TEX-N-30" transform="translate(1500,0)"></use></g><rect width="3922.4" height="60" x="120" y="220"></rect></g><g data-mml-node="mo" transform="translate(9803.8,0)"><use data-c="3D" xlink:href="#MJX-2-TEX-N-3D"></use></g><g data-mml-node="mn" transform="translate(10859.6,0)"><use data-c="37" xlink:href="#MJX-2-TEX-N-37"></use><use data-c="2E" xlink:href="#MJX-2-TEX-N-2E" transform="translate(500,0)"></use><use data-c="35" xlink:href="#MJX-2-TEX-N-35" transform="translate(778,0)"></use></g></g></g></svg><mjx-assistive-mml unselectable="on" display="block"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><mtext>Coverage</mtext><mo>=</mo><mfrac><mrow><mn>50</mn><mo>×</mo><mn>150</mn></mrow><mn>1000</mn></mfrac><mo>=</mo><mn>7.5</mn></math></mjx-assistive-mml></mjx-container></div></div></div><p>&nbsp;</p><h3 id='bigwig'><span>BigWIG</span></h3><ul><li><p><span>BigWig 类似于Bedgraph，是一种 </span><strong><span>用于表示基因组上连续信号轨迹</span></strong><span> 的</span><strong><span>压缩二进制文件</span></strong><span>格式。</span></p></li><li><p><span>BigWig 是从allignment后的 BAM 文件统计每个</span><strong><span>Blocks/bin</span></strong><span>的 read 覆盖度（coverage）得来的</span></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="pgsql"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="pgsql"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-operator">+--------------------------+</span> &nbsp; &nbsp; &nbsp; <span class="cm-operator">+---------------------+</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-operator">|</span> &nbsp; <span class="cm-keyword">Header</span> <span class="cm-bracket">(</span>文件头部<span class="cm-bracket">)</span> &nbsp; &nbsp; &nbsp;<span class="cm-operator">|-----&gt;</span> <span class="cm-operator">|</span> &nbsp; <span class="cm-keyword">Data</span> Blocks <span class="cm-bracket">(</span>数据块<span class="cm-bracket">)</span><span class="cm-operator">|</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-operator">+--------------------------+</span> &nbsp; &nbsp; &nbsp; <span class="cm-operator">+---------------------+</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-operator">|</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;  v</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-operator">+-----------------------+</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-operator">|</span> &nbsp; <span class="cm-keyword">Index</span> <span class="cm-bracket">(</span>索引<span class="cm-bracket">)</span> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-operator">|</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-operator">+-----------------------+</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 197px;"></div><div class="CodeMirror-gutters" style="display: none; height: 197px;"></div></div></div></pre><p><strong><span>Header</span></strong><span>：包含文件版本、染色体信息等元数据。</span></p><p><strong><span>Data Blocks</span></strong><span>：存储各个 </span><strong><span>bin</span></strong><span> 的信号值数据。</span></p><p><strong><span>Index</span></strong><span>：记录数据块的位置，支持高效的随机访问。</span></p><p>&nbsp;</p></li></ul><h3 id='different-from-bedgraph-and-bigwig'><span>Different from Bedgraph and BigWIG</span></h3><figure class='table-figure'><table><thead><tr><th><span>特性</span></th><th><strong><span>bedGraph</span></strong></th><th><strong><span>BigWig</span></strong></th></tr></thead><tbody><tr><td><strong><span>格式</span></strong></td><td><span>文本格式</span></td><td><span>二进制格式</span></td></tr><tr><td><strong><span>存储效率</span></strong></td><td><span>存储效率低，文件较大</span></td><td><span>存储效率高，文件较小</span></td></tr><tr><td><strong><span>访问速度</span></strong></td><td><span>加载和查询速度慢</span></td><td><span>支持高效的随机访问，查询快速</span></td></tr><tr><td><strong><span>查询效率</span></strong></td><td><span>无法直接查询，只能加载整个文件</span></td><td><span>可以快速查询指定区域</span></td></tr><tr><td><strong><span>可读性</span></strong></td><td><span>可用文本编辑器查看</span></td><td><span>需要特定工具转换或查看</span></td></tr></tbody></table></figure><p>&nbsp;</p><hr /><h1 id='environment-set-up'><span>Environment set up</span></h1><p><span>使用conda进行环境的管理</span></p><ul><li><p><span>创建环境：</span><code>conda create --name &lt;environment_name&gt;</code></p></li><li><p><span>激活环境：</span><code>conda activate &lt;environment_name&gt;</code></p></li><li><p><span>退出环境：</span><code>conda deactivate</code></p></li><li><p><span>查看当前所有的环境： </span><code>conda env list</code></p></li><li><p><span>删除环境： </span><code>conda env remove --name &lt;environment_name&gt;</code></p></li></ul><p>&nbsp;</p><hr /><h1 id='convert-sra-to-fastq'><span>Convert SRA to FASTQ</span></h1><h2 id='parallel-fastq-dump'><span>parallel-fastq-dump</span></h2><ul><li><p><code>fastq-dump</code><span> 是 NCBI SRA Toolkit(</span><a href='https://github.com/ncbi/sra-tools'><strong><span>sratoolkit</span></strong></a><span>)中的一个工具，用于Convert SSR to FASTQ</span></p></li><li><p><span>NCBI </span><code>fastq-dump</code><span> can be very slow sometimes, even if you have the resources (network, IO, CPU) to go faster</span></p></li><li><p><a href='https://github.com/rvalieris/parallel-fastq-dump'><strong><span>parallel-fastq-dump</span></strong></a><span> 是 </span><code>fastq-dump</code><span> 的并行版本，它利用 GNU parallel 来加速 SRA 数据的转换，适合大规模数据处理</span></p></li><li><p><strong><span>Software installation:</span></strong><span> </span><code>conda install parallel-fastq-dump &#39;sra-tools&gt;=3.0.0&#39;</code></p><p><strong><span>Important</span></strong><span>: Make sure the sra-tools package being installed is </span><strong><span>a recent version(&gt;=2.10.0)</span></strong><span> to guarantee compatibility with NCBI servers</span></p></li></ul><p>&nbsp;</p><p><strong><span>parameters</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">parallel-fastq-dump </span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--sra-id</span> <span class="cm-comment"># 输入SRA文件名</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--threads</span> <span class="cm-comment"># 用的核数</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--outdir</span> <span class="cm-comment"># 输出文件夹</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--split-files</span> <span class="cm-comment"># 文件是否切割来运行</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--gzip</span> <span class="cm-comment"># 输出文件是否为压缩包</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--minSpotId</span> <span class="cm-comment"># Minimum spot id (default: 1)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--maxSpotId</span> &nbsp;<span class="cm-comment"># Maximum spot id (default: None)</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 175px;"></div><div class="CodeMirror-gutters" style="display: none; height: 175px;"></div></div></div></pre><p><strong><span>Example</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">parallel-fastq-dump </span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--sra-id</span> IP_H3K27ac_Vehicle \ </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--threads</span> <span class="cm-number">11</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--outdir</span> fastq.data/ \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--split-files</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--gzipconc</span> \</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 131px;"></div><div class="CodeMirror-gutters" style="display: none; height: 131px;"></div></div></div></pre><div class="md-alert md-alert-caution caution"><p><span class='md-alert-text md-alert-text-caution'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M4.47.22A.749.749 0 0 1 5 0h6c.199 0 .389.079.53.22l4.25 4.25c.141.14.22.331.22.53v6a.749.749 0 0 1-.22.53l-4.25 4.25A.749.749 0 0 1 11 16H5a.749.749 0 0 1-.53-.22L.22 11.53A.749.749 0 0 1 0 11V5c0-.199.079-.389.22-.53Zm.84 1.28L1.5 5.31v5.38l3.81 3.81h5.38l3.81-3.81V5.31L10.69 1.5ZM8 4a.75.75 0 0 1 .75.75v3.5a.75.75 0 0 1-1.5 0v-3.5A.75.75 0 0 1 8 4Zm0 8a1 1 0 1 1 0-2 1 1 0 0 1 0 2Z"></path></svg>Caution</span><br></p><p><strong><span>如果一个 srr 文件转化为两个 fastq 文件，说明该数据为双端测序</span></strong></p></div><p>&nbsp;</p><hr /><h1 id='qc-report-and-data-cleaning---trim-galore'><span>QC Report and Data cleaning—trim-galore</span></h1><p><code>trim-galore</code><span>有2个小软件，分别是 </span><code>cutadapt</code><span> 用于</span><strong><span>去除 adapter </span></strong><span>及</span><strong><span>质量过滤</span></strong><span>，</span><code>fastqc</code><span> 用于查看</span><strong><span>数据质量分布</span></strong></p><ul><li><p><strong><span>Software installation</span></strong><code>conda install bioconda::trim-galore</code></p></li></ul><ul><li><p><strong><span>suggestion</span></strong><span>: 建议使用 </span><strong><span>Python3</span></strong><span> 环境，启用多线程（ </span><code>trim-galore</code><span> 中的 </span><code>cutadapt</code><span> 在 python2 不支持多线程操作）</span></p></li></ul><p>&nbsp;</p><hr /><h2 id='qc-report------fastqc'><span>QC Report——fastqc</span></h2><div class="md-alert md-alert-caution caution"><p><span class='md-alert-text md-alert-text-caution'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M4.47.22A.749.749 0 0 1 5 0h6c.199 0 .389.079.53.22l4.25 4.25c.141.14.22.331.22.53v6a.749.749 0 0 1-.22.53l-4.25 4.25A.749.749 0 0 1 11 16H5a.749.749 0 0 1-.53-.22L.22 11.53A.749.749 0 0 1 0 11V5c0-.199.079-.389.22-.53Zm.84 1.28L1.5 5.31v5.38l3.81 3.81h5.38l3.81-3.81V5.31L10.69 1.5ZM8 4a.75.75 0 0 1 .75.75v3.5a.75.75 0 0 1-1.5 0v-3.5A.75.75 0 0 1 8 4Zm0 8a1 1 0 1 1 0-2 1 1 0 0 1 0 2Z"></path></svg>Caution</span><br></p><p><strong><span>生成质控报告，可以先做质控报告，如果都合格就没必要做 Data cleaning</span></strong></p></div><p>&nbsp;</p><p><strong><span>parameters</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">fastqc <span class="cm-comment"># 输入文件名</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==input options==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-f</span> &lt;bam/sam/bam_mapped/sam_mapped/fastq&gt; <span class="cm-comment"># 强制指定输入文件格式（FastQC 默认会自动检测格式，一般不需要使用此选项）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--casava</span> <span class="cm-comment"># 处理 CASAVA 生成的 FASTQ 文件</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--nofilter</span>：<span class="cm-comment"># 如果使用 --casava 选项，默认会去除低质量 reads，但加上 --nofilter 选项后，将保留所有 reads 进行分析</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--nano</span>：<span class="cm-comment"># 用于处理三代测序（Nanopore）测序数据，直接从 .fast5 格式的文件中提取序列</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==ourput options==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-o</span> &lt;path&gt; <span class="cm-comment"># 输出文件名</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--extract</span> <span class="cm-comment"># 解压缩输出文件（deafult）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" class="cm-tab-wrap-hack" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--noextract</span> <span class="cm-comment"># 不解压缩输出文件<span class="cm-tab" role="presentation" cm-text="	">  </span></span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==Computing Optimization and Environment==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-t</span> &lt;int&gt; &nbsp;<span class="cm-comment"># 线程数，每个线程占用 250MB 内存，所以不建议超过可用内存的上限</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-j</span> &lt;path&gt; <span class="cm-comment"># 指定Java路径（FastQC 依赖 Java 运行）。如果不提供，FastQC 会使用系统默认的 Java </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># == Analysis parameters ==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-k</span> &lt;int&gt; <span class="cm-comment"># 指定Kmer内容模块中要查找的 Kmer长度。指定的Kmer长度必须在2到10之间(deafult:5）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--min_length</span> &lt;int&gt; <span class="cm-comment"># 设置要在报告中显示 reads 的最小长度</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-l</span> &lt;file&gt; </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 指定一个文件，其中包含一组标准将用于确定各种模块的FALL/错误限制，此文件还可用于有选择地从输出中删除一些模块</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; <span class="cm-comment"># 该文件的格式需要与 Configuration 文件夹中的默认 limits.txt 文件相匹配</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--nogroup</span> </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 禁用碱基分组，对于 &gt;50 bp 的 reads，FastQC 默认会对碱基进行分组以减少计算量。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 使用此选项，每个碱基都会单独分析.</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 纯在非常长的read上使用此选项，将导致 fastqc 崩溃，并且您的图表最终可能会变得非常大</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">#== Aptamer &amp; Contamination Sequence Analysis ==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-a</span> &lt;file&gt; <span class="cm-comment"># 指定一个文件(包含特定lst of adapter): 分析特定的 adapter污染情况. 不指定：使用常见四种adapter</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-c</span> &lt;file&gt; <span class="cm-comment"># 指定一个文件(包含特定lst of 污染物)，分析特定的污染物污染情况 </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">#== other options ==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-q</span> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 静默模式，不显示进度信息，仅报告错误。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-d</span> &lt;path&gt; <span class="cm-comment"># 指定用于存放临时文件的目录（deafult:使用. /tmp）</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 1093px;"></div><div class="CodeMirror-gutters" style="display: none; height: 1093px;"></div></div></div></pre><p>&nbsp;</p><div class="md-alert md-alert-tip tip"><p><span class='md-alert-text md-alert-text-tip'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M8 1.5c-2.363 0-4 1.69-4 3.75 0 .984.424 1.625.984 2.304l.214.253c.223.264.47.556.673.848.284.411.537.896.621 1.49a.75.75 0 0 1-1.484.211c-.04-.282-.163-.547-.37-.847a8.456 8.456 0 0 0-.542-.68c-.084-.1-.173-.205-.268-.32C3.201 7.75 2.5 6.766 2.5 5.25 2.5 2.31 4.863 0 8 0s5.5 2.31 5.5 5.25c0 1.516-.701 2.5-1.328 3.259-.095.115-.184.22-.268.319-.207.245-.383.453-.541.681-.208.3-.33.565-.37.847a.751.751 0 0 1-1.485-.212c.084-.593.337-1.078.621-1.489.203-.292.45-.584.673-.848.075-.088.147-.173.213-.253.561-.679.985-1.32.985-2.304 0-2.06-1.637-3.75-4-3.75ZM5.75 12h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5ZM6 15.25a.75.75 0 0 1 .75-.75h2.5a.75.75 0 0 1 0 1.5h-2.5a.75.75 0 0 1-.75-.75Z"></path></svg>Tip</span><br></p><p><code>--casava</code></p><p><span>高通量测序 (如 Illumina HiSeqTM/MiseqTM) 得到的原始图像数据文件经 </span><strong><span>CASAVA</span></strong><span> 软件 Base Calling 分析转化为原始 Sequenced Reads，我们称之为 Raw Data 或 Raw Reads，结果以 FASTQ 文件格式存储</span></p><p>&nbsp;</p><p><strong><span>CASAVA</span></strong><span> 软件 Base Calling 会将</span><strong><span>同一样本</span></strong><span>输出</span><strong><span>多个 FASTQ 文件</span></strong></p><p>&nbsp;</p><p><span>FastQC 提供了 </span><code>--casava</code><span> 选项，专门用于处理 CASAVA 生成的 FASTQ 文件 </span></p><ul><li><p><span>将自动识别相同样本的不同 fastq 文件，视为 </span><strong><span>同一数据集</span></strong><span> 进行分析</span></p></li><li><p><span>并</span><strong><span>去除低质量的 reads</span></strong><span>（CASAVA 在 </span><strong><span>FASTQ Sequence Identifier</span></strong><span> 里标记了 </span><code>Y</code><span> 的 reads（</span><a href='#fastq-sequence-identifier'><span>详见点此</span></a><span>）会被去除）</span></p><p><span>使用</span><code>--nofilter</code><span>则保留所有 reads</span></p></li></ul></div><p>&nbsp;</p><p><strong><span>Example</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation"><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">fastqc IP_H3K27ac_Vehicle_1_trimmed.fq.gz <span class="cm-attribute">-o</span> fatqc/ &nbsp;<span class="cm-attribute">-t</span> <span class="cm-number">12</span> </span></pre></div></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 22px;"></div><div class="CodeMirror-gutters" style="display: none; height: 22px;"></div></div></div></pre><p>&nbsp;</p><hr /><h2 id='stucture-of-qc-report'><span>Stucture of QC report</span></h2><p><a href='https://mugenomicscore.missouri.edu/PDF/FastQC_Manual.pdf'><span>QC report详情</span></a></p><p><a href='https://www.bioinformatics.babraham.ac.uk/projects/fastqc/bad_sequence_fastqc.html'><span>QC report reference in CUT&amp;Tag</span></a></p><p>&nbsp;</p><h3 id='basic-statistics'><span>Basic Statistics</span></h3><p><img src= "./picture/013.png" width="700"/></p><ul><li><p><strong><span>%GC</span></strong><span>：是我们需要重点关注的一个指标，这个值表示的是整体 sequence 中的GC含量</span></p></li><li><p><strong><span>%GC</span></strong><span> 这个数值一般是物种特异的，比如 </span><strong><span>Human 约 42%</span></strong><span>， </span><strong><span>mouse 约 41%</span></strong><span>，</span><strong><span>Rat 约 42%</span></strong><span> </span></p></li><li><p><span>如果测序原始数据的GC含量远远偏离这个比例，说明测序数据存在一定偏好性，如果直接用测序数据，会影响后续的CNV和变异检测的分析。</span></p></li></ul><p>&nbsp;</p><hr /><h3 id='per-base-sequence-quality'><span>Per base sequence quality</span></h3><p><img src= "./picture/014.png" width="900"/></p><ul><li><p><span>箱线图应该位于绿色部分就是合格的，黄色部分出</span><strong><span>FALL</span></strong><span>，红色部分</span><strong><span>WARN</span></strong></p></li><li><p><span>这是 read length = 100 的数据</span></p><ul><li><p><strong><span>横轴</span></strong><span>为碱基在</span><strong><span>read 的位置</span></strong></p></li><li><p><strong><span>纵轴</span></strong><span>是 </span><strong><span>quality</span></strong><span>。</span><mjx-container class="MathJax" jax="SVG" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="24.299ex" height="2.278ex" role="img" focusable="false" viewBox="0 -750 10740.1 1006.9" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -0.581ex;"><defs><path id="MJX-22-TEX-N-71" d="M33 218Q33 308 95 374T236 441H246Q330 441 381 372L387 364Q388 364 404 403L420 442H457V156Q457 -132 458 -134Q462 -142 470 -145Q491 -148 519 -148H535V-194H527L504 -193Q480 -192 453 -192T415 -191Q312 -191 303 -194H295V-148H311Q339 -148 360 -145Q369 -141 371 -135T373 -106V-41V49Q313 -11 236 -11Q154 -11 94 53T33 218ZM376 300Q346 389 278 401Q275 401 269 401T261 402Q211 400 171 350T131 214Q131 137 165 82T253 27Q296 27 328 54T376 118V300Z"></path><path id="MJX-22-TEX-N-75" d="M383 58Q327 -10 256 -10H249Q124 -10 105 89Q104 96 103 226Q102 335 102 348T96 369Q86 385 36 385H25V408Q25 431 27 431L38 432Q48 433 67 434T105 436Q122 437 142 438T172 441T184 442H187V261Q188 77 190 64Q193 49 204 40Q224 26 264 26Q290 26 311 35T343 58T363 90T375 120T379 144Q379 145 379 161T380 201T380 248V315Q380 361 370 372T320 385H302V431Q304 431 378 436T457 442H464V264Q464 84 465 81Q468 61 479 55T524 46H542V0Q540 0 467 -5T390 -11H383V58Z"></path><path id="MJX-22-TEX-N-61" d="M137 305T115 305T78 320T63 359Q63 394 97 421T218 448Q291 448 336 416T396 340Q401 326 401 309T402 194V124Q402 76 407 58T428 40Q443 40 448 56T453 109V145H493V106Q492 66 490 59Q481 29 455 12T400 -6T353 12T329 54V58L327 55Q325 52 322 49T314 40T302 29T287 17T269 6T247 -2T221 -8T190 -11Q130 -11 82 20T34 107Q34 128 41 147T68 188T116 225T194 253T304 268H318V290Q318 324 312 340Q290 411 215 411Q197 411 181 410T156 406T148 403Q170 388 170 359Q170 334 154 320ZM126 106Q126 75 150 51T209 26Q247 26 276 49T315 109Q317 116 318 175Q318 233 317 233Q309 233 296 232T251 223T193 203T147 166T126 106Z"></path><path id="MJX-22-TEX-N-6C" d="M42 46H56Q95 46 103 60V68Q103 77 103 91T103 124T104 167T104 217T104 272T104 329Q104 366 104 407T104 482T104 542T103 586T103 603Q100 622 89 628T44 637H26V660Q26 683 28 683L38 684Q48 685 67 686T104 688Q121 689 141 690T171 693T182 694H185V379Q185 62 186 60Q190 52 198 49Q219 46 247 46H263V0H255L232 1Q209 2 183 2T145 3T107 3T57 1L34 0H26V46H42Z"></path><path id="MJX-22-TEX-N-69" d="M69 609Q69 637 87 653T131 669Q154 667 171 652T188 609Q188 579 171 564T129 549Q104 549 87 564T69 609ZM247 0Q232 3 143 3Q132 3 106 3T56 1L34 0H26V46H42Q70 46 91 49Q100 53 102 60T104 102V205V293Q104 345 102 359T88 378Q74 385 41 385H30V408Q30 431 32 431L42 432Q52 433 70 434T106 436Q123 437 142 438T171 441T182 442H185V62Q190 52 197 50T232 46H255V0H247Z"></path><path id="MJX-22-TEX-N-74" d="M27 422Q80 426 109 478T141 600V615H181V431H316V385H181V241Q182 116 182 100T189 68Q203 29 238 29Q282 29 292 100Q293 108 293 146V181H333V146V134Q333 57 291 17Q264 -10 221 -10Q187 -10 162 2T124 33T105 68T98 100Q97 107 97 248V385H18V422H27Z"></path><path id="MJX-22-TEX-N-79" d="M69 -66Q91 -66 104 -80T118 -116Q118 -134 109 -145T91 -160Q84 -163 97 -166Q104 -168 111 -168Q131 -168 148 -159T175 -138T197 -106T213 -75T225 -43L242 0L170 183Q150 233 125 297Q101 358 96 368T80 381Q79 382 78 382Q66 385 34 385H19V431H26L46 430Q65 430 88 429T122 428Q129 428 142 428T171 429T200 430T224 430L233 431H241V385H232Q183 385 185 366L286 112Q286 113 332 227L376 341V350Q376 365 366 373T348 383T334 385H331V431H337H344Q351 431 361 431T382 430T405 429T422 429Q477 429 503 431H508V385H497Q441 380 422 345Q420 343 378 235T289 9T227 -131Q180 -204 113 -204Q69 -204 44 -177T19 -116Q19 -89 35 -78T69 -66Z"></path><path id="MJX-22-TEX-N-3D" d="M56 347Q56 360 70 367H707Q722 359 722 347Q722 336 708 328L390 327H72Q56 332 56 347ZM56 153Q56 168 72 173H708Q722 163 722 153Q722 140 707 133H70Q56 140 56 153Z"></path><path id="MJX-22-TEX-N-2212" d="M84 237T84 250T98 270H679Q694 262 694 250T679 230H98Q84 237 84 250Z"></path><path id="MJX-22-TEX-N-31" d="M213 578L200 573Q186 568 160 563T102 556H83V602H102Q149 604 189 617T245 641T273 663Q275 666 285 666Q294 666 302 660V361L303 61Q310 54 315 52T339 48T401 46H427V0H416Q395 3 257 3Q121 3 100 0H88V46H114Q136 46 152 46T177 47T193 50T201 52T207 57T213 61V578Z"></path><path id="MJX-22-TEX-N-30" d="M96 585Q152 666 249 666Q297 666 345 640T423 548Q460 465 460 320Q460 165 417 83Q397 41 362 16T301 -15T250 -22Q224 -22 198 -16T137 16T82 83Q39 165 39 320Q39 494 96 585ZM321 597Q291 629 250 629Q208 629 178 597Q153 571 145 525T137 333Q137 175 145 125T181 46Q209 16 250 16Q290 16 318 46Q347 76 354 130T362 333Q362 478 354 524T321 597Z"></path><path id="MJX-22-TEX-N-D7" d="M630 29Q630 9 609 9Q604 9 587 25T493 118L389 222L284 117Q178 13 175 11Q171 9 168 9Q160 9 154 15T147 29Q147 36 161 51T255 146L359 250L255 354Q174 435 161 449T147 471Q147 480 153 485T168 490Q173 490 175 489Q178 487 284 383L389 278L493 382Q570 459 587 475T609 491Q630 491 630 471Q630 464 620 453T522 355L418 250L522 145Q606 61 618 48T630 29Z"></path><path id="MJX-22-TEX-N-6F" d="M28 214Q28 309 93 378T250 448Q340 448 405 380T471 215Q471 120 407 55T250 -10Q153 -10 91 57T28 214ZM250 30Q372 30 372 193V225V250Q372 272 371 288T364 326T348 362T317 390T268 410Q263 411 252 411Q222 411 195 399Q152 377 139 338T126 246V226Q126 130 145 91Q177 30 250 30Z"></path><path id="MJX-22-TEX-N-67" d="M329 409Q373 453 429 453Q459 453 472 434T485 396Q485 382 476 371T449 360Q416 360 412 390Q410 404 415 411Q415 412 416 414V415Q388 412 363 393Q355 388 355 386Q355 385 359 381T368 369T379 351T388 325T392 292Q392 230 343 187T222 143Q172 143 123 171Q112 153 112 133Q112 98 138 81Q147 75 155 75T227 73Q311 72 335 67Q396 58 431 26Q470 -13 470 -72Q470 -139 392 -175Q332 -206 250 -206Q167 -206 107 -175Q29 -140 29 -75Q29 -39 50 -15T92 18L103 24Q67 55 67 108Q67 155 96 193Q52 237 52 292Q52 355 102 398T223 442Q274 442 318 416L329 409ZM299 343Q294 371 273 387T221 404Q192 404 171 388T145 343Q142 326 142 292Q142 248 149 227T179 192Q196 182 222 182Q244 182 260 189T283 207T294 227T299 242Q302 258 302 292T299 343ZM403 -75Q403 -50 389 -34T348 -11T299 -2T245 0H218Q151 0 138 -6Q118 -15 107 -34T95 -74Q95 -84 101 -97T122 -127T170 -155T250 -167Q319 -167 361 -139T403 -75Z"></path><path id="MJX-22-TEX-N-2061" d=""></path><path id="MJX-22-TEX-N-28" d="M94 250Q94 319 104 381T127 488T164 576T202 643T244 695T277 729T302 750H315H319Q333 750 333 741Q333 738 316 720T275 667T226 581T184 443T167 250T184 58T225 -81T274 -167T316 -220T333 -241Q333 -250 318 -250H315H302L274 -226Q180 -141 137 -14T94 250Z"></path><path id="MJX-22-TEX-I-1D45D" d="M23 287Q24 290 25 295T30 317T40 348T55 381T75 411T101 433T134 442Q209 442 230 378L240 387Q302 442 358 442Q423 442 460 395T497 281Q497 173 421 82T249 -10Q227 -10 210 -4Q199 1 187 11T168 28L161 36Q160 35 139 -51T118 -138Q118 -144 126 -145T163 -148H188Q194 -155 194 -157T191 -175Q188 -187 185 -190T172 -194Q170 -194 161 -194T127 -193T65 -192Q-5 -192 -24 -194H-32Q-39 -187 -39 -183Q-37 -156 -26 -148H-6Q28 -147 33 -136Q36 -130 94 103T155 350Q156 355 156 364Q156 405 131 405Q109 405 94 377T71 316T59 280Q57 278 43 278H29Q23 284 23 287ZM178 102Q200 26 252 26Q282 26 310 49T356 107Q374 141 392 215T411 325V331Q411 405 350 405Q339 405 328 402T306 393T286 380T269 365T254 350T243 336T235 326L232 322Q232 321 229 308T218 264T204 212Q178 106 178 102Z"></path><path id="MJX-22-TEX-N-29" d="M60 749L64 750Q69 750 74 750H86L114 726Q208 641 251 514T294 250Q294 182 284 119T261 12T224 -76T186 -143T145 -194T113 -227T90 -246Q87 -249 86 -250H74Q66 -250 63 -250T58 -247T55 -238Q56 -237 66 -225Q221 -64 221 250T66 725Q56 737 55 738Q55 746 60 749Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="mtext"><use data-c="71" xlink:href="#MJX-22-TEX-N-71"></use><use data-c="75" xlink:href="#MJX-22-TEX-N-75" transform="translate(528,0)"></use><use data-c="61" xlink:href="#MJX-22-TEX-N-61" transform="translate(1084,0)"></use><use data-c="6C" xlink:href="#MJX-22-TEX-N-6C" transform="translate(1584,0)"></use><use data-c="69" xlink:href="#MJX-22-TEX-N-69" transform="translate(1862,0)"></use><use data-c="74" xlink:href="#MJX-22-TEX-N-74" transform="translate(2140,0)"></use><use data-c="79" xlink:href="#MJX-22-TEX-N-79" transform="translate(2529,0)"></use></g><g data-mml-node="mo" transform="translate(3334.8,0)"><use data-c="3D" xlink:href="#MJX-22-TEX-N-3D"></use></g><g data-mml-node="mo" transform="translate(4390.6,0)"><use data-c="2212" xlink:href="#MJX-22-TEX-N-2212"></use></g><g data-mml-node="mn" transform="translate(5168.6,0)"><use data-c="31" xlink:href="#MJX-22-TEX-N-31"></use><use data-c="30" xlink:href="#MJX-22-TEX-N-30" transform="translate(500,0)"></use></g><g data-mml-node="mo" transform="translate(6390.8,0)"><use data-c="D7" xlink:href="#MJX-22-TEX-N-D7"></use></g><g data-mml-node="msub" transform="translate(7391,0)"><g data-mml-node="mi"><use data-c="6C" xlink:href="#MJX-22-TEX-N-6C"></use><use data-c="6F" xlink:href="#MJX-22-TEX-N-6F" transform="translate(278,0)"></use><use data-c="67" xlink:href="#MJX-22-TEX-N-67" transform="translate(778,0)"></use></g><g data-mml-node="TeXAtom" transform="translate(1311,-241.4) scale(0.707)" data-mjx-texclass="ORD"><g data-mml-node="mn"><use data-c="31" xlink:href="#MJX-22-TEX-N-31"></use><use data-c="30" xlink:href="#MJX-22-TEX-N-30" transform="translate(500,0)"></use></g></g></g><g data-mml-node="mo" transform="translate(9459.1,0)"><use data-c="2061" xlink:href="#MJX-22-TEX-N-2061"></use></g><g data-mml-node="mo" transform="translate(9459.1,0)"><use data-c="28" xlink:href="#MJX-22-TEX-N-28"></use></g><g data-mml-node="mi" transform="translate(9848.1,0)"><use data-c="1D45D" xlink:href="#MJX-22-TEX-I-1D45D"></use></g><g data-mml-node="mo" transform="translate(10351.1,0)"><use data-c="29" xlink:href="#MJX-22-TEX-N-29"></use></g></g></g></svg><mjx-assistive-mml unselectable="on" display="inline"><math xmlns="http://www.w3.org/1998/Math/MathML"><mtext>quality</mtext><mo>=</mo><mo>−</mo><mn>10</mn><mo>×</mo><msub><mi>log</mi><mrow data-mjx-texclass="ORD"><mn>10</mn></mrow></msub><mo data-mjx-texclass="NONE">⁡</mo><mo stretchy="false">(</mo><mi>p</mi><mo stretchy="false">)</mo></math></mjx-assistive-mml></mjx-container><script type="math/tex">\text{quality} = -10 \times \log_{10}(p)</script><span>，p为测错的概率。</span></p></li><li><p><span>根据 quality 给出质量结果：正常区间（28 - 40），警告区间（20-28），错误区间（0-20）</span></p></li></ul></li></ul><p>&nbsp;</p><p><strong><span>标准：</span></strong><span> </span></p><ul><li><p><span>如果任何一个位置的下四分位数小于10或者中位数小于25，会显示</span><strong><span>FALL</span></strong><span>；</span></p></li><li><p><span>如果任何一个位置的下四分位数小于5或者中位数小于20，会显示</span><strong><span>WARN</span></strong></p><p>&nbsp;</p></li></ul><p><strong><span>注意：</span></strong><span> 在FastQ文件中编码质量分数有许多不同的方法。FastQC尝试自动确定使用了哪种编码方法，但在一些非常有限的数据集中，它可能会猜错，图表的标题将描述FastQC认为您的文件使用的编码。</span></p><p>&nbsp;</p><hr /><h3 id='per-tile-sequence-quality'><span>Per tile sequence quality</span></h3><p><img src= "./picture/015.png" width="700"/></p><ul><li><p><span>横轴为碱基在 read 的位置，纵轴是</span><strong><span>tail 编号</span></strong></p><div class="md-alert md-alert-tip tip"><p><span class='md-alert-text md-alert-text-tip'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M8 1.5c-2.363 0-4 1.69-4 3.75 0 .984.424 1.625.984 2.304l.214.253c.223.264.47.556.673.848.284.411.537.896.621 1.49a.75.75 0 0 1-1.484.211c-.04-.282-.163-.547-.37-.847a8.456 8.456 0 0 0-.542-.68c-.084-.1-.173-.205-.268-.32C3.201 7.75 2.5 6.766 2.5 5.25 2.5 2.31 4.863 0 8 0s5.5 2.31 5.5 5.25c0 1.516-.701 2.5-1.328 3.259-.095.115-.184.22-.268.319-.207.245-.383.453-.541.681-.208.3-.33.565-.37.847a.751.751 0 0 1-1.485-.212c.084-.593.337-1.078.621-1.489.203-.292.45-.584.673-.848.075-.088.147-.173.213-.253.561-.679.985-1.32.985-2.304 0-2.06-1.637-3.75-4-3.75ZM5.75 12h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5ZM6 15.25a.75.75 0 0 1 .75-.75h2.5a.75.75 0 0 1 0 1.5h-2.5a.75.75 0 0 1-.75-.75Z"></path></svg>Tip</span><br></p><p><strong><span>tile编号</span></strong><span>通常用于识别Illumina测序仪上的特定物理位置。每个</span><strong><span>tile</span></strong><span>代表</span><strong><span>flowcell</span></strong><span>的最小单位 </span><a href='#Flowcell'><span>详情点此</span></a></p></div></li><li><p><span>图中的颜色是从</span><strong><span>冷色调</span></strong><span>到</span><strong><span>暖色调</span></strong><span>的渐变</span></p><ul><li><p><strong><span>冷色调</span></strong><span>表示这个 </span><strong><span>tile</span></strong><span> 在这个位置上的质量值高于所有 </span><strong><span>tiles</span></strong><span> 在这个位置上的平均质量值</span></p></li><li><p><strong><span>暖色调</span></strong><span>表示这个 </span><strong><span>tile</span></strong><span> 的在这个位置上的质量值比其它 </span><strong><span>tiles</span></strong><span> 要差</span></p></li></ul></li></ul><ul><li><p><span>一个非常好的结果，整张图都应该是蓝色</span></p></li><li><p><span>如果出现质量问题可能是短暂的，如有气泡产生，也可能是长期的，如在某一小孔中存在残骸，问题不大</span></p></li></ul><p>&nbsp;</p><p><strong><span>标准：</span></strong></p><ul><li><p><span>如果任何 </span><strong><span>tile 的平均质量值</span></strong><span>与这个位置上</span><strong><span>所有tiles的平均质量值</span></strong><span>相差</span><strong><span>2</span></strong><span>以上会显</span><strong><span>FALL</span></strong></p></li><li><p><span>如果任何 </span><strong><span>tile 的平均质量值</span></strong><span>与这个位置上</span><strong><span>所有 tiles 的平均质量值</span></strong><span>相差</span><strong><span>5</span></strong><span>以上会显</span><strong><span>WARN</span></strong></p></li></ul><p>&nbsp;</p><p><strong><span>不好的结果：</span></strong></p><p><img src= "./picture/016.png" width="600"/></p><p>&nbsp;</p><hr /><h3 id='per-sequence-quality-scores'><span>Per sequence quality scores</span></h3><p><img src= "./picture/017.png" width="700"/></p><ul><li><p><span>x轴是平均碱基质量值，y轴是reads数。</span></p></li></ul><p>&nbsp;</p><p><strong><span>标准：</span></strong><span> </span></p><ul><li><p><span>如果最高峰的质量值小于27( 错误率0.2%)则会显示</span><strong><span>FALL</span></strong></p></li><li><p><span>如果最高峰的质量值小于20(错误率 1 %) 则会显示</span><strong><span>WARN</span></strong></p><p>&nbsp;</p></li></ul><div class="md-alert md-alert-important important"><p><span class='md-alert-text md-alert-text-important'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M0 1.75C0 .784.784 0 1.75 0h12.5C15.216 0 16 .784 16 1.75v9.5A1.75 1.75 0 0 1 14.25 13H8.06l-2.573 2.573A1.458 1.458 0 0 1 3 14.543V13H1.75A1.75 1.75 0 0 1 0 11.25Zm1.75-.25a.25.25 0 0 0-.25.25v9.5c0 .138.112.25.25.25h2a.75.75 0 0 1 .75.75v2.19l2.72-2.72a.749.749 0 0 1 .53-.22h6.5a.25.25 0 0 0 .25-.25v-9.5a.25.25 0 0 0-.25-.25Zm7 2.25v2.5a.75.75 0 0 1-1.5 0v-2.5a.75.75 0 0 1 1.5 0ZM9 9a1 1 0 1 1-2 0 1 1 0 0 1 2 0Z"></path></svg>Important</span><br></p><p><strong><span>低质量的会有两或多个峰</span></strong></p></div><p>&nbsp;</p><div class="md-alert md-alert-warning warning"><p><span class='md-alert-text md-alert-text-warning'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M6.457 1.047c.659-1.234 2.427-1.234 3.086 0l6.082 11.378A1.75 1.75 0 0 1 14.082 15H1.918a1.75 1.75 0 0 1-1.543-2.575Zm1.763.707a.25.25 0 0 0-.44 0L1.698 13.132a.25.25 0 0 0 .22.368h12.164a.25.25 0 0 0 .22-.368Zm.53 3.996v2.5a.75.75 0 0 1-1.5 0v-2.5a.75.75 0 0 1 1.5 0ZM9 11a1 1 0 1 1-2 0 1 1 0 0 1 2 0Z"></path></svg>Warning</span><br></p><p><span>如果在一次运行中有</span><strong><span>相当大比例</span></strong><span>的 sequence 整体质量较低，那么这可能表明某种</span><strong><span>系统问题</span></strong><span>——可能是系统部分运行(例如</span><strong><span>one end of a flowcell</span></strong><span>)</span></p><p><strong><span>One end of a flow cell</span></strong><span>：指</span><strong><span>flow cell</span></strong><span> 的某一端出现 </span><strong><span>较多低质量 reads</span></strong></p><ul><li><p><strong><span>局部成像问题</span></strong><span>：荧光信号在该区域的检测不佳，导致测序质量下降。</span></p></li><li><p><strong><span>试剂分布不均</span></strong><span>：某些区域可能受到试剂浓度或分布的影响，导致质量偏差。</span></p></li><li><p><strong><span>污染或残留</span></strong><span>：该区域可能有残留试剂、气泡、或者堵塞，影响测序效果。</span></p></li><li><p><strong><span>激光或光学问题</span></strong><span>：Illumina 通过光学检测信号，如果某一端的光学系统异常，也可能导致该区域的质量下降。</span></p></li></ul></div><p>&nbsp;</p><hr /><h3 id='per-base-sequence-content'><span>Per base sequence content</span></h3><p><img src= "./picture/018.png" width="700"/></p><p><strong><span>横坐标</span></strong><span>代表 read 的</span><strong><span>碱基位置</span></strong><span>, </span><strong><span>纵坐标</span></strong><span>代表</span><strong><span>该碱基百分比</span></strong></p><ul><li><p><span>一个完全随机的文库内每个位置上 4 种碱基的比例应该大致相同，因此图中的四条线应该相互平行且接近</span></p></li></ul><ul><li><p><span>在 </span><strong><span>reads 开头</span></strong><span>出现碱基组成偏离往往是我们的建库操作造成的</span></p></li></ul><div class="md-alert md-alert-tip tip"><p><span class='md-alert-text md-alert-text-tip'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M8 1.5c-2.363 0-4 1.69-4 3.75 0 .984.424 1.625.984 2.304l.214.253c.223.264.47.556.673.848.284.411.537.896.621 1.49a.75.75 0 0 1-1.484.211c-.04-.282-.163-.547-.37-.847a8.456 8.456 0 0 0-.542-.68c-.084-.1-.173-.205-.268-.32C3.201 7.75 2.5 6.766 2.5 5.25 2.5 2.31 4.863 0 8 0s5.5 2.31 5.5 5.25c0 1.516-.701 2.5-1.328 3.259-.095.115-.184.22-.268.319-.207.245-.383.453-.541.681-.208.3-.33.565-.37.847a.751.751 0 0 1-1.485-.212c.084-.593.337-1.078.621-1.489.203-.292.45-.584.673-.848.075-.088.147-.173.213-.253.561-.679.985-1.32.985-2.304 0-2.06-1.637-3.75-4-3.75ZM5.75 12h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5ZM6 15.25a.75.75 0 0 1 .75-.75h2.5a.75.75 0 0 1 0 1.5h-2.5a.75.75 0 0 1-.75-.75Z"></path></svg>Tip</span><br></p><p><span>比如GBS文库中的</span><strong><span>条形码</span></strong><span>和</span><strong><span>酶切位点</span></strong><span>会影响测序reads的开头部分，这可能会在质量控制报告中看到碱基组成的偏离。以下是进一步解释：</span></p><ul><li><p><strong><span>条形码（Barcode）</span></strong><span>：每个样本使用不同的 Barcode sequence，这些 Barcode 位于测序 reads 的开头。由于 Barcode sequence 的多样性，reads 开头的碱基组成不是均一的。</span></p></li><li><p><strong><span>酶切位点</span></strong><span>：限制性酶在特定 sequence 处切割 DNA，因此这些酶切位点的碱基组成是固定的。如果这些位点在测序 reads 的开头，可能会导致碱基组成偏离</span></p></li></ul></div><ul><li><p><strong><span>reads 结尾</span></strong><span>出现的碱基组成偏离，往往是测序 </span><strong><span>adapter</span></strong><span> 的污染造成的，</span><strong><span>adapter</span></strong><span> 可能没有完全从 reads 中剪切掉，导致 reads 结尾包含</span><strong><span>adapter</span></strong><span>。</span></p></li><li><p><span>在 sequence 前12~15个碱基，可以</span><strong><span>使用剪切软件（cutadapt）切除</span></strong><span>，切除多少个需要看你的数据，比如上图的结果显示就可以切15个或19个。</span></p></li></ul><p>&nbsp;</p><p><strong><span>标准：</span></strong><span> </span></p><ul><li><p><span>如果任何一个位置上的 A 和 T 之间或者 G 和 C 之间的比例相差 10 % 以上则报</span><strong><span>FALL</span></strong></p></li><li><p><span>任何一个位置上的 A 和 T 之间或者 G 和 C 之间的比例相差 20 % 以上则报</span><strong><span>WARN</span></strong></p></li></ul><p>&nbsp;</p><hr /><h3 id='per-base-sequence-content-in-cuttag'><span>Per base sequence content in CUT&amp;Tag</span></h3><p>&nbsp;</p><p><img src="./picture/019.png" width="900"/></p><ul><li><p><strong><span>read</span></strong><span> 在开始时不一致的 sequence 内容是 CUT&amp;Tag 读长的常见现象。WARN的 </span><strong><span>Per base seuqnence</span></strong><span>  内容并不意味着您的数据失败。</span></p><ul><li><p><span>这可能是由于 </span><strong><span>Tn5 偏好性</span></strong><span>。</span></p></li><li><p><span>您可能检测到的是 </span><strong><span>read 前 10bp</span></strong><span>左右呈现为锯齿状。如果是这样，这是正常的，不会影响 </span><strong><span>alignment</span></strong><span> 或 </span><strong><span>peak calling</span></strong><span> 。</span><strong><span>在任何情况下，我们都不建议进行修剪</span></strong><span>，接下来通过</span><strong><span>修改 bowtie2 参数无需修剪</span></strong><span>即可提供准确的映射信息</span></p></li></ul></li></ul><p>&nbsp;</p><hr /><h3 id='per-sequence-gc-conten'><span>Per sequence GC conten</span></h3><p><img src="./picture/020.png" width="700"/></p><ul><li><p><span>这部分表示统计每个 sequence  GC 含量的频数。</span></p></li><li><p><strong><span>红色</span></strong><span>线是</span><strong><span>实际值</span></strong><span>，</span><strong><span>蓝色</span></strong><span>线是</span><strong><span>理论值</span></strong><span>。</span></p></li><li><p><span>峰值位点对应着总体 GC 含量</span></p></li><li><p><span>在一个正常的随机文库中，GC 含量的分布应接近正态分布，且中心的峰值和所测基因组的 GC 含量一致</span></p></li><li><p><span>由于软件并不知道所测物种真实的 GC 含量，图中的理论分布是基于所测数据计算得来的；</span><strong><span>红色线于蓝色线越接近越好</span></strong><span>，越接近测序越好，越可靠</span></p></li></ul><p>&nbsp;</p><p><strong><span>不好的结果：</span></strong></p><p><img src="./picture/021.png"  width="700"/></p><ul><li><p><span>如果出现不正常的尖峰分布 (如上图)，则说明文库可能有污染 (如果是 </span><strong><span>adapter</span></strong><span> 的污染，那么在 </span><strong><span>Overrepresented Sequences</span></strong><span> 那部分结果还会得到提示)，或者存在其它形式的偏选</span></p></li></ul><p>&nbsp;</p><p><strong><span>标准：</span></strong></p><ul><li><p><span>如果偏离理论分布的 reads 数超过总 reads 数的 15 % 则报</span><strong><span>FALL</span></strong></p></li><li><p><span>如果偏离理论分布的 reads 数超过总 reads 数的 30 % 则报</span><strong><span>WARN</span></strong><span>。</span></p></li></ul><p>&nbsp;</p><hr /><h3 id='per-base-n-content'><span>Per base N content</span></h3><p><img src="./picture/022.png" width="700"/></p><ul><li><p><span>在测序仪工作过程中，如果不能正常完成某个 </span><strong><span>base calling</span></strong><span> (测序仪对每个测序碱基进行检测和识别的过程)，将会以 N 来表示这个位置的碱基，而不是 A、T、C、G</span></p></li><li><p><span>出现一定比例的Ns在测序数据中是一个常见现象，通常反映了测序过程中存在某些问题。这种现象的原因可以分为两类，具体解释如下：</span></p><ul><li><p><strong><span>普遍的质量丢失 (General Loss of Quality)</span></strong></p><p><strong><span>原因</span></strong><span>：在整个测序过程中，如果测序仪由于多种原因（如化学反应不佳、光学系统问题或 flowcell 污染等）无法准确地</span><strong><span>base calling</span></strong><span>，可能会导致整个测序文库的质量下降。这样的质量下降通常表现为</span><strong><span>较高比例的“N”出现在reads的不同位置</span></strong><span>。</span></p><p>&nbsp;</p><p><strong><span>判断方法</span></strong><span>：这种情况通常会影响整个文库的测序质量，具体可以通过查看质量控制报告中的多项指标来综合判断，如 </span><strong><span>Per Sequence Quality Scores</span></strong><span>  和 </span><strong><span>Per Base Quality Scores</span></strong><span> 等。这些指标可以帮助识别是否存在普遍的质量丢失。</span></p></li></ul><p>&nbsp;</p><ul><li><p><strong><span>Reads 开头出现较高比例的 N</span></strong></p><p><strong><span>原因</span></strong><span>：整体测序质量较高的文库中，</span><strong><span>reads 开头碱基组成偏差导致</span></strong><span>。例如 </span><strong><span>GBS 文库</span></strong><span>，reads 开头通常包含</span><strong><span>固定 sequence </span></strong><span>(如酶切位点和Adapters），这些 sequence 在所有 reads 中都是相同的。然而，在初始测序循环中，由于碱基组成过于单一，测序仪可能无法正确识别荧光信号，导致 base calling 失败，从而在 reads 的开头填充 “N” </span></p><p>&nbsp;</p><p><strong><span>判断方法</span></strong><span>：这种现象可以通过结合 </span><strong><span>Per Base Sequence Content</span></strong><span> 的结果来判断。在 </span><strong><span>Per Base Sequence Content</span></strong><span> 中，如果看到</span><strong><span>开头几位的碱基质量评分较低</span></strong><span>，则可以推测这些位置的 N 是由于碱基组成偏差导致的。</span></p></li></ul><p>&nbsp;</p></li></ul><ul><li><p><strong><span>该N的含量越小越好，为0是最最最理想的。</span></strong></p><p>&nbsp;</p></li></ul><p><strong><span>标准：</span></strong></p><ul><li><p><span>N 的比例大于 5 % 则报</span><strong><span>FALL</span></strong></p></li><li><p><span>N 的比例大于大于 20 % 则报</span><strong><span>WARN</span></strong><span>。</span></p></li></ul><p>&nbsp;</p><hr /><h3 id='sequence-length-distribution'><span>Sequence Length Distribution</span></h3><p><strong><span>Data cleaning 前</span></strong><span>：</span></p><p><img src="./picture/023.png" width="800"/></p><ul><li><p><span>该部分是 sequence 长度分布情况，上图例子显示，整个数据 sequence 的长度分布于100~102bp之间。</span></p></li><li><p><span>测序仪出来的原始 reads 通常是均一长度的，但经过</span><strong><span>Data cleaning 软件等处理过的数据</span></strong><span>则不然。</span></p><ul><li><p><strong><span>adapter 去除</span></strong></p><p><span>在测序过程中，reads 可能包含 adapter。如果 adapter 没有被去除，Data cleaning 软件会识别并去除这些 adapter。</span></p><p><span>adapter 的去除可能导致 reads 长度不再一致，特别是在测序片段较短的情况下，去除 adapter 后剩余的 sequence 长度会变短。</span></p></li></ul><p>&nbsp;</p><ul><li><p><strong><span>低质量碱基的剪切</span></strong></p><p><span>Data cleaning 软件通常会检测并去除 reads 中质量较低的碱基，尤其是 reads 的末端。如果某些碱基的质量低于设定的阈值，这些碱基会被剪掉，从而导致 reads 的长度缩短。由于不同 reads 的低质量区段长度不同，处理后的 reads 长度可能会有所差异。</span></p></li></ul><p>&nbsp;</p><ul><li><p><strong><span>去除污染 sequence </span></strong></p><p><span>在某些情况下，Data cleaning 软件可能会识别并去除污染 sequence 或低复杂度 sequence （如多次重复的碱基）。这些 sequence 的去除同样可能导致 reads长度的变化。</span></p><p>&nbsp;</p></li><li><p><strong><span>质量过滤</span></strong></p><p><span>Data cleaning 软件可能会完全丢弃一些长度过短或质量过低的 reads。这意味着在处理后，剩余的 reads 可能存在不同的长度.</span></p></li></ul><p>&nbsp;</p></li></ul><p>&nbsp;</p><p><strong><span>Data cleaning 后</span></strong><span>：</span></p><p><img src="./picture/024.png" width="700"/></p><p>&nbsp;</p><p><strong><span>标准：</span></strong></p><ul><li><p><span>当 reads 长度不一致时报</span><strong><span>FALL</span></strong></p></li><li><p><span>当有长度为 0 的 reads 时则报</span><strong><span>WARN</span></strong><span>。</span></p></li></ul><p>&nbsp;</p><hr /><h3 id='sequence-duplication-level'><span>Sequence Duplication Level</span></h3><p><img src="./picture/025.png" width="700"/></p><ul><li><p><strong><span>横坐标</span></strong><span>为</span><strong><span>重复（duplication）的次数</span></strong></p></li><li><p><strong><span>纵坐标</span></strong><span>为 </span><strong><span>reads 的数目</span></strong></p></li><li><p><span>为了减少该模块的内存需求，只分析每个文件中前200,000个 sequences 。</span></p><p><span>分析每个文件中的前200,000个 sequence ，但这应该足以对整个文件的重复程度有一个好的印象</span></p><p>&nbsp;</p></li></ul><p><strong><span>低重复度</span></strong><span>：在一个多样性的文库中，大多数 sequence 只会出现一次。这通常意味着测序覆盖度高，即测序数据很好地覆盖了目标 sequence 的不同区域</span></p><p><strong><span>高重复度</span></strong><span>：如果某些 sequence 多次出现，可能表明数据存在偏差，例如PCR过度扩增。高重复度通常意味着测序过程中对某些 sequence 进行了过度扩增</span></p><p>&nbsp;</p><p><strong><span>标准</span></strong></p><ul><li><p><span>如果 duplication 占总数的20%以上，该模块将发出经</span><strong><span>FALL</span></strong></p></li><li><p><span>如果 duplication 占总数的50%以上，该模块将发出一个</span><strong><span>WARN</span></strong></p><p>&nbsp;</p></li></ul><div class="md-alert md-alert-tip tip"><p><span class='md-alert-text md-alert-text-tip'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M8 1.5c-2.363 0-4 1.69-4 3.75 0 .984.424 1.625.984 2.304l.214.253c.223.264.47.556.673.848.284.411.537.896.621 1.49a.75.75 0 0 1-1.484.211c-.04-.282-.163-.547-.37-.847a8.456 8.456 0 0 0-.542-.68c-.084-.1-.173-.205-.268-.32C3.201 7.75 2.5 6.766 2.5 5.25 2.5 2.31 4.863 0 8 0s5.5 2.31 5.5 5.25c0 1.516-.701 2.5-1.328 3.259-.095.115-.184.22-.268.319-.207.245-.383.453-.541.681-.208.3-.33.565-.37.847a.751.751 0 0 1-1.485-.212c.084-.593.337-1.078.621-1.489.203-.292.45-.584.673-.848.075-.088.147-.173.213-.253.561-.679.985-1.32.985-2.304 0-2.06-1.637-3.75-4-3.75ZM5.75 12h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5ZM6 15.25a.75.75 0 0 1 .75-.75h2.5a.75.75 0 0 1 0 1.5h-2.5a.75.75 0 0 1-.75-.75Z"></path></svg>Tip</span><br></p><p><strong><span>为什么 ChIP-seq 中 duplication 会过多？</span></strong></p><ul><li><p><span>在 ChIP-seq 实验中，样本量通常较少，因此需要进行多轮PCR扩增来生成足够的DNA用于测序</span></p></li><li><p><span>有可能某个转录因子在基因组中的结合位点非常有限，并且这些位点非常集中，那么使用针对该转录因子的抗体进行 ChIP-seq 实验时，测序结果中会有很多 duplication</span></p></li></ul></div><p>&nbsp;</p><hr /><h3 id='overrepresented-sequences'><span>Overrepresented sequences</span></h3><p><img src="./picture/026.png" width="800"/></p><p>&nbsp;</p><ul><li><p><span>它展示了长度至少</span><strong><span>20 bp</span></strong><span>，数量占总数</span><strong><span>0.1%以上</span></strong><span>的 reads sequences，它可以帮助判断污染(比如：载体、接头序列)。</span></p></li><li><p><span>一个正常的高通量文库将包含一组不同的 sequence，没有单个 sequence 占整体的一小部分。发现单个 sequence 在集合中被Overrepresented，要么意味着它具有高度的生物学意义，要么表明文库被污染了</span></p><ul><li><p><span>对于每一个 Overrepresented sequence，程序将在一个常见污染物的数据库或接头数据库中寻找匹配，并报告它找到的最佳匹配</span></p></li><li><p><span>如果不是，可以复制序列去 blast</span></p><p>&nbsp;</p></li></ul></li></ul><ul><li><p><span>为了节省内存，只分析每个文件中前200,000个 sequence 。</span></p><p><span>因此，由于某种原因，该模块可能会错过一个被 Overrepresented 但未出现在文件开头的 sequence。</span></p><p>&nbsp;</p></li><li><p><span>由于 Overrepresented sequences 检测需要在序列的整个长度上进行精确的序列匹配，因此任何</span><strong><span>长度超过75bp的读取都被截断为50bp</span></strong></p></li></ul><p>&nbsp;</p><p><strong><span>标准</span></strong></p><ul><li><p><span>当发现 Overrepresented reads 超过总 reads 0.1% 的reads时报</span><strong><span>FALL</span></strong></p></li><li><p><span>当发现 Overrepresented reads 超过总 reads 1% 的reads时报</span><strong><span>WARN</span></strong></p></li></ul><p>&nbsp;</p><hr /><h3 id='adapter-content'><span>Adapter Content</span></h3><p><strong><span>没去除adapter</span></strong></p><p><img src="./picture/027.png" width="700"/></p><ul><li><p><span>上图中 adapter 没有去除，在后续分析的时候需要先使用 Data cleaning 软件进行去接头。</span></p></li></ul><p>&nbsp;</p><p><strong><span>去除adapter后</span></strong></p><p><img src="./picture/028.png" width="700"/></p><ul><li><p><span>当时</span><code>fastqc</code><span>分析的时候 </span><code>-a</code><span> 选项没有内容，则默认使用图例中的四种通用adapter序列进行统计</span></p></li></ul><p>&nbsp;</p><hr /><h3 id='overrepresented-kmers'><span>Overrepresented Kmers</span></h3><p><img src="./picture/029.png" width="700"/></p><ul><li><p><span>k bp 的短序列在 reads 中大量出现，其频率高于统计期望的话，将其记为</span><strong><span>over-represented k-mer</span></strong><span>。</span></p></li><li><p><code>-k</code><span>可以指定 Kmer 内容模块中要查找的 Kmer 长度。指定的 Kmer 长度必须在</span><code>2-10</code><span>之间 (deafult:</span><code>k = 5</code><span>) </span></p></li><li><p><span>出现频率总体上3倍于期望或是在某位置上5倍于期望的 kmer 被认为是 over-represented</span></p></li><li><p><strong><span>over-represented kmer</span></strong><span> 有可能是</span><strong><span>污染</span></strong><span>，也可能是与</span><strong><span>转录因子的 DNA 结合位点</span></strong><span>相关</span></p></li></ul><p>&nbsp;</p><p><strong><span>标准</span></strong></p><ul><li><p><span>当有出现频率总体上3倍于期望或是在某位置上5倍于期望的k-mer时，报</span><strong><span>WARN</span></strong></p></li><li><p><span>当有出现频率在某位置上10倍于期望的k-mer时报</span><strong><span>FAIL</span></strong></p></li></ul><p>&nbsp;</p><p><strong><span>ChIP-seq 中的Kmers</span></strong></p><p><img src="./picture/030.png" width="700"/></p><ul><li><p><span>对于 ChIP-seq 数据来说，这一指标出现</span><strong><span>FAIL</span></strong><span>是意料之中的，而且实际上是数据中信号良好的标志</span></p></li></ul><p>&nbsp;</p><hr /><h2 id='fastqc是如何检测adapter的呢'><span>fastqc是如何检测adapter的呢？</span></h2><p><strong><span>参考：</span></strong><a href='https://www.jianshu.com/p/077911bf772d'><span>基础——illumina测序原理与细节</span></a></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="test"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="test"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation"><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">Illumina Universal Adapter &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;  AGATCGGAAGAG</span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">Illumina Small RNA Adapter &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;  ATGGAATTCTCG</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">Nextera Transposase Sequence &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;  CTGTCTCTTATA</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">SOLID Small RNA Adapter &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; CGCCTTGGCCGT</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 87px;"></div><div class="CodeMirror-gutters" style="display: none; height: 87px;"></div></div></div></pre><p><strong><span>为什么这几个序列就能帮我们判断adapter的有无，按照我们的理解来说不应该adapter序列会很多样吗？</span></strong></p><p><strong><span>这几个序列到底在adapter的哪里？</span></strong></p><ul><li><p><span>这几个序列在测序</span><strong><span>引物的3’端</span></strong></p></li><li><p><span>由于基本上所有的</span><strong><span>引物3&#39;端</span></strong><span>都是这些序列，所以我们可以通过这几个序列判断adapter</span></p></li><li><p><span>无论什么测序，都必须连接在这些</span><strong><span>引物3&#39;端</span></strong><span>，所以才能检测到adapter</span></p></li></ul><p><img src="./picture/078.png" width="900"/></p><p>&nbsp;</p><hr /><h2 id='data-cleaning---cutadapt'><span>Data cleaning—cutadapt</span></h2><h3 id='theory'><span>Theory</span></h3><ul><li><p><a href='https://github.com/FelixKrueger/TrimGalore/blob/master/Docs/Trim_Galore_User_Guide.md'><span>cutadapt</span></a></p></li><li><p><strong><span>Step 1: Quality Trimming</span></strong><span>——去除 reads 3’端的低质量碱基</span></p><p><span>illumina平台的测序数据，通常3’端质量较差。</span><code>trim_galore</code><span> 首先会过滤掉3”端的低质量碱基.</span></p></li><li><p><strong><span>Step 2: Adapter Trimming</span></strong><span>——去除 adapter</span></p></li><li><p><strong><span>Step 3: Removing Short Sequences</span></strong><span>——去除长度太短的 reads。默认情况下，如果序列长度少于 20bp，这条序列会被丢掉。</span></p></li></ul><div class="md-alert md-alert-important important"><p><span class='md-alert-text md-alert-text-important'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M0 1.75C0 .784.784 0 1.75 0h12.5C15.216 0 16 .784 16 1.75v9.5A1.75 1.75 0 0 1 14.25 13H8.06l-2.573 2.573A1.458 1.458 0 0 1 3 14.543V13H1.75A1.75 1.75 0 0 1 0 11.25Zm1.75-.25a.25.25 0 0 0-.25.25v9.5c0 .138.112.25.25.25h2a.75.75 0 0 1 .75.75v2.19l2.72-2.72a.749.749 0 0 1 .53-.22h6.5a.25.25 0 0 0 .25-.25v-9.5a.25.25 0 0 0-.25-.25Zm7 2.25v2.5a.75.75 0 0 1-1.5 0v-2.5a.75.75 0 0 1 1.5 0ZM9 9a1 1 0 1 1-2 0 1 1 0 0 1 2 0Z"></path></svg>Important</span><br></p><p><strong><span>如果对双端测序中的一个文件单独分析，不建议删除太短的reads</span></strong></p><ul><li><p><span>因为这可能会破坏 </span><strong><span>成对read</span></strong><span> 的顺序，而这种顺序对许多 </span><strong><span>alignment</span></strong><span> 软件很重要</span></p></li><li><p><code>Trim Galore</code><span>对双端测序修剪后验证 </span><strong><span>成对read</span></strong><span> 的完整性，如果其中一个 </span><strong><span>read</span></strong><span> 变得太短，整个</span><strong><span>成对read</span></strong><span>会被移除</span></p></li><li><p><code>Trim Galore</code><span>的</span><code>--retain_unpaired</code><span>：如果</span><strong><span>只有一个read</span></strong><span>合格，这个</span><strong><span>read</span></strong><span>会被保存到单独的文件中以便</span><strong><span>单端alignment</span></strong><span>。这可以确保你最终使用的数据是高质量的。</span></p></li></ul></div><p>&nbsp;</p><ul><li><p><span>Step 4: </span><strong><span>Specialised Trimming</span></strong><span> </span><strong><span>(专业修整)</span></strong></p><ul><li><p><span>Hard-trimming to leave bases at the 5&#39;-end</span></p><p><code>--hardtrim5</code><span> </span><code>&lt;int&gt;</code></p><p><span>这个选项会硬性截短序列，只保留从 5&#39; 端开始的指定数量（int 个碱基）。也就是说，无论序列原来的长度是多少，只会保留序列最前面的 int 个碱基，其余部分会被完全删除。生成硬修整 FastQ 文件</span></p><p>&nbsp;</p></li><li><p><span>Hard-trimming to leave bases at the 3&#39;-end</span></p><p><code>--hardtrim3</code><span> </span><code>&lt;int&gt;</code></p><p><span>这个选项会硬性截短序列，只保留从 3&#39; 端开始的指定数量（int 个碱基）。也就是说，无论序列原来的长度是多少，只会保留序列最前面的 int 个碱基，其余部分会被完全删除。并生成硬修整 FastQ 文件</span></p><p>&nbsp;</p></li><li><p><span>小鼠表观遗传学时钟修整</span></p><p><code>--clock</code></p><p><a href='https://genomebiology.biomedcentral.com/articles/10.1186/s13059-017-1203-5'><span>详见</span></a></p><ul><li><p><span>在DNA甲基化测序中，PCR扩增不可避免，UMI（唯一分子标识符）的引入可以帮助去除这些PCR重复，使得测序数据更准确地反映真实的甲基化水平</span></p></li><li><p><span>选项用于处理UMI测序数据，特别是用于鼠类表观遗传钟分析。它会识别并移除序列中的UMI和固定的非测序片段，把这些信息记录到每条序列的ID中，然后生成修剪后的文件。</span></p></li><li><p><span>这些文件随后需要进一步修剪掉尾端的15个碱基，以确保所有UMI和固定序列都被移除。之后，数据可以进行allignment，以便于后续的DNA甲基化分析。</span></p></li></ul></li></ul></li></ul><p>&nbsp;</p><ul><li><p><strong><span>Trimming whether or not?</span></strong></p><ul><li><p><strong><span>不需要trimming：</span></strong></p><p><span>对于25x25 PE测序，由于每个read长度为25 bp，在这种情况下</span></p><ul><li><p><span>adapter序列不会出现在read的末端，因为测序长度不足以覆盖整个插入片段并延伸到adapter。</span></p></li><li><p><span>read长度也比较短，3&#39;端的质量也比较高</span></p></li></ul><p><span>因此，也就不需要进行read修剪。</span></p></li><li><p><strong><span>需要trimming：</span></strong></p><p><span>对于read长度更长的测序（例如100 bp或更长），测序的长度可能接近或超过插入read的长度。这意味着adapter序列可能出现在read的末端，且3&#39;端的质量也比较差。可能会干扰allignment和后续分析。因此，需要进行read修剪</span></p></li><li><p><strong><span>使用</span><code>--local</code><span>模式进行比对：</span></strong></p><p><span>对于经过</span><strong><span>trimmed read</span></strong><span>，建议使用</span><code>--local</code><span>。</span><code>--local</code><span>模式允许比对器在read的一部分上寻找匹配，而不要求整个read都参与allignment。这非常适合trimmed read.</span></p><p>&nbsp;</p></li></ul></li></ul><h3 id='parameters'><span>parameters</span></h3><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">trim_galore &nbsp;<span class="cm-comment"># 输入文件名</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==输入选项== &nbsp;</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--paired</span> </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 用于双端测序数据。需要提供两个文件。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 对于双端测序结果，一对reads中，如果有一个被剔除，那么另一个会被同样抛弃，而不管是否达到标准</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--retain_unpaired</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 如果两个配对端读取中只有一个太短，则较长的读取将写入任一文件或输出文件</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 默认情况下retain_unpaired 是关闭的（OFF）。这意味着如果一对读取中只有一个读取满足长度要求，则整个读取对都会被丢弃</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 当双端测序数据中两个读取的质量差异较大时，其中一个读取可能会由于低质量区域的修剪变得非常短。如果你希望保留较长的那个读取以进行单端分析，可以启用 --retain_unpaired 选项。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-r1</span> &lt;int&gt; <span class="cm-comment"># 指保留R1时的最小长度阈值,这个长度是指在经过质量修剪之后，R1 必须达到的最小长度。如果启用了--retain_unpaired，且R1 满足这个长度要求但R2 太短，则R1 会被保留并写入 unpaired_1.fq 文件中。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-r2</span> &lt;int&gt; <span class="cm-comment"># 指保留R2时的最小长度阈值</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--phred33</span> <span class="cm-comment"># 质量分数为Phred+33编码</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--phred66</span> <span class="cm-comment"># 质量分数为Phred+64编码</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==输出选项==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-o</span> <span class="cm-comment"># 指定输出目录</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--gzip</span> <span class="cm-comment"># 输出文件为解压文件</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--dont_gzip</span> <span class="cm-comment"># 输出文件不为解压文件</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--fastqc</span> <span class="cm-comment"># 修剪完成后运行fastqc</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--fastqc_args</span> <span class="cm-string">"&lt;ARGS&gt;"</span> </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 将额外的参数传递给 FastQC</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 例如：--fastqc_args "--nogroup --outdir /home/"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 参数将自动调用 FastQC，因此不必单独指定。--fastqc</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==性能优化==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--cores</span> &lt;int&gt; </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 指定使用的核心数。建议使用 4 核，以达到性能和效率的最佳平衡 </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 只支持python3以上的版本，Python2只能使用单线程</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==普通修剪选项==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-q</span> &lt;int&gt; <span class="cm-comment"># 设置修剪低质量碱基的阈值，默认为 20（按默认走就行）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--max_length</span> &lt;int&gt; <span class="cm-comment"># 修剪后丢弃长度超过 bp 的读数</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--length</span> &lt;int&gt; <span class="cm-comment"># 设置最小read长度，低于此长度的read将被丢弃（默认20）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--stringency</span> &lt;int&gt;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 如果在读取序列的末端找到了一个至少具有指定长度的adapte序列重叠部分，则认为该序列包含adapte，并从该位置进行修剪</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 默认为1(非常苛刻)可以适度放宽至3到4，因为后一个adapter几乎不可能被测序仪读到</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 较高的 stringency 值可以减少误匹配的修剪，但可能会导致一些实际包含adapter的序列未被修剪；较低的 stringency 值则相反</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-e</span> &lt;int&gt;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 参数用于设置adapte序列匹配时允许的最大错误率（即错配、插入或删除的比例）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 错误率越高，匹配adapte的严格性就越低，反之则越严格。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 默认为0.1</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==特定修剪选项==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--clip_R1</span> &lt;int&gt; <span class="cm-comment"># 指 单端测序 或 双端测序中的R1 5' 端去除指定数量的碱基</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--clip_R2</span> &lt;int&gt; <span class="cm-comment"># 指双端测序中的R2 5' 端去除指定数量的碱基</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 在双端 BS-Seq（双硫化物测序）中，5' 端的碱基可能因为末端修复反应而导致低甲基化偏差。根据M-bias plot section in the Bismark User Guide，适当去除几个位点的碱基</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--three_prime_clip_R1</span> &lt;int&gt; <span class="cm-comment"># 指修剪adapter后，单端测序 或 双端测序中的R1 3' 端去除指定数量的碱基</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--three_prime_clip_R2</span> &lt;int&gt; <span class="cm-comment"># 指修剪adapter后，双端测序中的R2 3' 端去除指定数量的碱基</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--max_n</span> &lt;int&gt; <span class="cm-comment"># 参数用于设置允许序列中最大未确定碱基（N）的数量。如果序列中的未确定碱基数超过了这个指定的阈值，则该序列将被完全过滤掉</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--trim-n</span> <span class="cm-comment"># 从读取的任一侧删除未确定碱基（N）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--2colour</span>/--nextseq &lt;int&gt;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 在 NextSeq 和 NovaSeq 平台上，G 碱基在没有信号的情况下也可能被标记为高质量，这会导致下游分析中的问题</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 详细见：https://sequencing.qcfail.com/articles/illumina-2-colour-chemistry-can-overcall-high-confidence-g-bases/.</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==特定adapter选项==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-a</span> <span class="cm-comment"># 指定特定的adapter</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-a2</span> <span class="cm-comment"># 选项用于指定第二条adapte序列。这在双端测序处理中尤其有用</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 1770px;"></div><div class="CodeMirror-gutters" style="display: none; height: 1770px;"></div></div></div></pre><p>&nbsp;</p><p><strong><span>Example</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==单端测序==</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">trim_galore IP_H3K27ac_Vehicle_1.fastq.gz <span class="cm-attribute">-q</span> <span class="cm-number">25</span> <span class="cm-attribute">--phred33</span> <span class="cm-attribute">--length</span> <span class="cm-number">25</span> <span class="cm-attribute">-e</span> <span class="cm-number">0</span>.1 <span class="cm-attribute">--stringency</span> <span class="cm-number">4</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==双端测序==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">trim_galore \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--paired</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--phred33</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--length</span> <span class="cm-number">50</span>  \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--fastqc</span> KChIP2-OE-230929S_230929S_R1.fastq.gz KChIP2-OE-230929S_230929S_R2.fastq.gz \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--stringency</span> <span class="cm-number">4</span> <span class="cm-attribute">--cores</span> <span class="cm-number">4</span> <span class="cm-attribute">--clip_R1</span> <span class="cm-number">15</span> <span class="cm-attribute">--clip_R2</span> <span class="cm-number">15</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-o</span> trim_galore</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==不进行trim(step1,step2),只去除太短的reads(step3),对于双端测序,可以用到fastp工具==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">fastp \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-i</span> KChIP2-OE-230929S_230929S_R1.fastq.gz \ &nbsp;<span class="cm-comment"># 输入文件1（R1）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-I</span> KChIP2-OE-230929S_230929S_R2.fastq.gz \ <span class="cm-comment"># 输入文件2（R2）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--length_required</span> <span class="cm-number">10</span> \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 最短reads长度要求</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--disable_adapter_trimming</span> \ &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 禁用接头修剪</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--disable_quality_filtering</span> \ &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 禁用质量过滤</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-o</span> out.R1.fastq.gz \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 输出文件1（R1）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-O</span> out.R2.fastq.gz \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 输出文件2（R2）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-w</span> <span class="cm-number">4</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 使用的线程数</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 525px;"></div><div class="CodeMirror-gutters" style="display: none; height: 525px;"></div></div></div></pre><hr /><h1 id='reads-alignment'><span>Reads Alignment</span></h1><h2 id='alignment-theory'><span>Alignment theory</span></h2><h3 id='如何评估序列相似'><span>如何评估序列相似？</span></h3><ul><li><p><span>序列相似有2个指标：</span><strong><span>一致度(identity)</span></strong><span> ，</span><strong><span>相似度(similarity)</span></strong></p></li><li><p><strong><span>identity</span></strong><span>: 如采两个序列(蛋白质或核酸)长度相同，那么它们的一致度定义为他们对应位置上</span><strong><span>相同</span></strong><span>的残基占总长度的百分数。</span></p></li><li><p><strong><span>similarity</span></strong><span>: 如果两个序列(蛋白质或核酸)长度相同，那么它们的相似度定义为他们对应位置上</span><strong><span>相似与相同</span></strong><span>的残基的数目占总长度的百分数。</span></p></li><li><p><strong><span>可为相识与相同的残基？</span></strong><span>：</span>
<strong><span>替换记分矩阵(Substitution Matrix)</span></strong><span> ，反映残基之间相互替换率的替换记分矩阵，它描述了残基两两相似的量化关系。分为</span><strong><span>DNA替换记分矩阵</span></strong><span>和</span><strong><span>蛋白质替换记分矩阵</span></strong><span>。</span></p><p><span>  </span><img src="./picture/067.png" width="700"/></p><p>&nbsp;</p></li></ul><hr /><h3 id='常见的dna替换记分矩阵'><span>常见的DNA替换记分矩阵</span></h3><ul><li><p><strong><span>等价矩阵(unitarymatrix)</span></strong><span> :最简单的替换计分矩阵，其中，相同的碱基之间匹配得分为1，而不同碱基间的替换得分为0。</span></p><p>&nbsp;</p><p><span>由于这种矩阵不含有碱基的物理化学信息，而且不区别对待不同的替换，因此在实际研究中运用较少。</span></p><p><img src="./picture/112.png" width="350"/></p></li></ul><p><span>  </span></p><ul><li><p><strong><span>转换-颠换矩阵(transition-transversionmatrix)</span></strong><span> :碱基的替换发生在嘧啶之间或嘌呤之间，则称为</span><strong><span>转换</span></strong><span>；而如果发生在嘧啶和嘌呤之间，则称为</span><strong><span>颠换</span></strong></p><p>&nbsp;</p><p><span>在进化过程中，转换发生的频率远比颠换高。为了反映这一情况，通常该矩阵中</span><strong><span>转换</span></strong><span>的得分为-1，而</span><strong><span>颠换</span></strong><span>的得分为-5</span></p><p><img src="./picture/113.png" width="350"/></p></li></ul><p><span>  </span></p><ul><li><p><strong><span>BLAST矩阵</span></strong><span> :经过在研究过程中的大量实际对比发现，如果令被比对的两个核苷酸相同时得分为+5，反之为-4，则比对效果较好，这就是被广泛使用的BLAST矩阵。</span></p><p><img src="./picture/114.png" width="350"/></p></li></ul><p><span>  </span></p><hr /><h3 id='what-is-the-alignment'><span>What is the alignment?</span></h3><p><strong><span>序列比对法(alignment)</span></strong><span> :两条序列的alignment就是把 序列s 和 序列t 这两个字符串上下排列起来，在某些位置插入空格(gap)。</span></p><p><span>然后依次比较它们在每一个位置上字符的匹配情况</span></p><p><span>匹配的好，这个位置就会得高分</span></p><p><span>匹配的不好，看看能不能左右错一错，或填上个空位，让附近的位置更好的匹配在一起，从而</span><strong><span>使所有位置的得分之和尽可能的高</span></strong><span>。</span></p><p><span>说白了，就是通过插入空位，让上下两行中尽可能多的一致的和相似的字符对在一起</span></p><p><span>需要使用专门的</span><strong><span>序列比对算法</span></strong><span>。分为</span><strong><span>全局比对</span></strong><span>与</span><strong><span>局部比对</span></strong></p><p><span> </span><img src="./picture/062.png" width="700"/></p><p>&nbsp;</p><hr /><h3 id='global-alignment'><span>Global Alignment</span></h3><ul><li><p><strong><span>Needleman-Wunsch算法</span></strong></p></li><li><p><span>比较复杂，建议看视频：</span><a href='https://www.bilibili.com/video/BV13t411372E?spm_id_from=333.788.videopod.episodes&amp;vd_source=951f1d2399bb1456eb6ac0a86e6efdd1&amp;p=37'><span>视频1</span></a><span>，</span><a href='https://www.bilibili.com/video/BV1qM4y1D7ND?spm_id_from=333.788.videopod.sections&amp;vd_source=951f1d2399bb1456eb6ac0a86e6efdd1'><span>视频2</span></a></p></li></ul><p><span>输入选项有：</span><strong><span>（1）序列p和序列q，（2）替换记分矩阵，（3）空位罚分（下图gap=-5）</span></strong></p><p><span>其中箭头代表</span><strong><span>追溯箭头</span></strong><span>代表其最大值从那个方格来。</span></p><p><img src="./picture/063.png" width="700"/></p><p>&nbsp;</p><p><img src="./picture/064.png" width="700"/></p><p>&nbsp;</p><p><img src="./picture/065.png" width="700"/></p><p>&nbsp;</p><hr /><h3 id='local-alignment'><span>Local Alignment</span></h3><ul><li><p><strong><span>Smith-Waterman算法</span></strong></p></li><li><p><span>比较复杂，建议看视频：</span><a href='https://www.bilibili.com/video/BV13t411372E?spm_id_from=333.788.videopod.episodes&amp;vd_source=951f1d2399bb1456eb6ac0a86e6efdd1&amp;p=38'><span>视频1</span></a></p></li><li><p><span>与全局比对的区别在于，</span><strong><span>其s(i,j)最大值多了个0</span></strong></p></li></ul><p><img src="./picture/132.png" width="700"/></p><p>&nbsp;</p><ul><li><p><strong><span>其得分为矩阵中最大的值：16</span></strong></p></li><li><p><span>追溯箭头不是从右下角到左下角，</span><strong><span>而是从刚刚找到的最大值开始追溯到没有箭头为止</span></strong></p></li></ul><p><img src="./picture/133.png" width="700"/></p><p>&nbsp;</p><hr /><h3 id='different-of-global-alignment-and-local-alignment-in-reads-alignment'><span>Different of Global Alignment and Local Alignment in reads alignment</span></h3><figure class='table-figure'><table><thead><tr><th><span>特征</span></th><th><span>全局比对 (Global Alignment)</span></th><th><span>局部比对 (Local Alignment)</span></th></tr></thead><tbody><tr><td><strong><span>定义</span></strong></td><td><span>将整个查询序列与参考序列进行比对</span></td><td><span>寻找两个序列中的最优匹配片段</span></td></tr><tr><td><strong><span>适用场景</span></strong></td><td><span>高相似度的长序列</span></td><td><span>存在小片段或插入/删除的情况</span></td></tr><tr><td><strong><span>常用算法</span></strong></td><td><span>Needleman-Wunsch, Smith-Waterman</span></td><td><span>Smith-Waterman</span></td></tr><tr><td><strong><span>优势</span></strong></td><td><span>能够检测到插入、缺失等复杂变异，适合高质量的读长较长的数据。</span></td><td><span>在处理质量较差的数据或序列中存在非生物学意义的片段时，比对率更高，能够处理部分匹配的情况。</span></td></tr><tr><td><strong><span>劣势</span></strong></td><td><span>如果读序列中存在低质量区域或adapters等非生物学意义的序列，全局比对可能会产生较低的比对率，计算成本高</span></td><td><span>可能会忽略掉一些复杂的结构变异，尤其是当这些变异涉及较长序列的精确比对时。</span></td></tr></tbody></table></figure><p>&nbsp;</p><hr /><h3 id='一致度identity和相似度similarity'><span>一致度(identity)和相似度(similarity)</span></h3><p><img src="./picture/134.png" width="700"/></p><p>&nbsp;</p><hr /><h3 id='blast'><span>BLAST</span></h3><ul><li><p><span>具体参考</span><a href='https://blog.csdn.net/weixin_46432623/article/details/105520286'><span>轻松理解BLAST搜索原理</span></a><span>  </span><a href='https://www.bilibili.com/video/BV1oY411p7jV/?spm_id_from=333.337.search-card.all.click&amp;vd_source=951f1d2399bb1456eb6ac0a86e6efdd1'><span>视频1</span></a></p></li><li><p><span>使用</span><strong><span>alignment</span></strong><span>去比对数据库的每一条序列，耗时比较长。所以使用</span><strong><span>BLAST算法</span></strong><span>快速从数据库准确找到相似的序列</span></p></li><li><p><strong><span>BLAST算法(Basic Local Alignment Search Tool)</span></strong><span> 采取</span><strong><span>seeding-and-extending</span></strong><span>方法。 首先找到query序列和subject序列间匹配的</span><strong><span>seed</span></strong><span>， 然后沿</span><strong><span>seed</span></strong><span>向两边延伸，产生</span><strong><span>High Scoring Segement Pairs (HSPs)</span></strong><span>， 然后在这些特定区域运用</span><strong><span>Simith-waterman</span></strong><span> 算法，最后评估比对的identify。</span></p></li></ul><p>&nbsp;</p><p><strong><span>BLAST的主要步骤</span></strong></p><ul><li><p><strong><span>分词(Creat a word list)</span></strong><span> </span></p><p><span>BLAST首先将输入的序列分解成固定长度的小片段。</span></p><div class="md-alert md-alert-tip tip"><p><span class='md-alert-text md-alert-text-tip'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M8 1.5c-2.363 0-4 1.69-4 3.75 0 .984.424 1.625.984 2.304l.214.253c.223.264.47.556.673.848.284.411.537.896.621 1.49a.75.75 0 0 1-1.484.211c-.04-.282-.163-.547-.37-.847a8.456 8.456 0 0 0-.542-.68c-.084-.1-.173-.205-.268-.32C3.201 7.75 2.5 6.766 2.5 5.25 2.5 2.31 4.863 0 8 0s5.5 2.31 5.5 5.25c0 1.516-.701 2.5-1.328 3.259-.095.115-.184.22-.268.319-.207.245-.383.453-.541.681-.208.3-.33.565-.37.847a.751.751 0 0 1-1.485-.212c.084-.593.337-1.078.621-1.489.203-.292.45-.584.673-.848.075-.088.147-.173.213-.253.561-.679.985-1.32.985-2.304 0-2.06-1.637-3.75-4-3.75ZM5.75 12h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5ZM6 15.25a.75.75 0 0 1 .75-.75h2.5a.75.75 0 0 1 0 1.5h-2.5a.75.75 0 0 1-.75-.75Z"></path></svg>Tip</span><br></p><p><span>假设你的句子是“GATTACA”，BLAST会拆成这些小片段：&quot;GAT&quot;, &quot;ATT&quot;, &quot;TAC&quot;, &quot;ACA&quot;  </span></p></div></li></ul><ul><li><p><strong><span>种子匹配（Seed Matching）</span></strong></p><p><span>BLAST会把这些小片段和数据库中的索引(index)快速匹配，找到所有完全相同的序列（</span><strong><span>seed匹配</span></strong><span>）。  </span></p><div class="md-alert md-alert-tip tip"><p><span class='md-alert-text md-alert-text-tip'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M8 1.5c-2.363 0-4 1.69-4 3.75 0 .984.424 1.625.984 2.304l.214.253c.223.264.47.556.673.848.284.411.537.896.621 1.49a.75.75 0 0 1-1.484.211c-.04-.282-.163-.547-.37-.847a8.456 8.456 0 0 0-.542-.68c-.084-.1-.173-.205-.268-.32C3.201 7.75 2.5 6.766 2.5 5.25 2.5 2.31 4.863 0 8 0s5.5 2.31 5.5 5.25c0 1.516-.701 2.5-1.328 3.259-.095.115-.184.22-.268.319-.207.245-.383.453-.541.681-.208.3-.33.565-.37.847a.751.751 0 0 1-1.485-.212c.084-.593.337-1.078.621-1.489.203-.292.45-.584.673-.848.075-.088.147-.173.213-.253.561-.679.985-1.32.985-2.304 0-2.06-1.637-3.75-4-3.75ZM5.75 12h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5ZM6 15.25a.75.75 0 0 1 .75-.75h2.5a.75.75 0 0 1 0 1.5h-2.5a.75.75 0 0 1-.75-.75Z"></path></svg>Tip</span><br></p><p><span>这就像在图书馆中快速找到包含“GAT”或“ATT”等关键词的书。</span></p></div></li><li><p><strong><span>扩展匹配（Extension）</span></strong></p><p><span>一旦找到一个匹配的小片段，BLAST会从这个片段出发，尝试向两侧扩展匹配，计算相似性得分，直到得分下降到一定程度为止（</span><strong><span>得分下降阈值</span></strong><span>）</span></p><div class="md-alert md-alert-tip tip"><p><span class='md-alert-text md-alert-text-tip'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M8 1.5c-2.363 0-4 1.69-4 3.75 0 .984.424 1.625.984 2.304l.214.253c.223.264.47.556.673.848.284.411.537.896.621 1.49a.75.75 0 0 1-1.484.211c-.04-.282-.163-.547-.37-.847a8.456 8.456 0 0 0-.542-.68c-.084-.1-.173-.205-.268-.32C3.201 7.75 2.5 6.766 2.5 5.25 2.5 2.31 4.863 0 8 0s5.5 2.31 5.5 5.25c0 1.516-.701 2.5-1.328 3.259-.095.115-.184.22-.268.319-.207.245-.383.453-.541.681-.208.3-.33.565-.37.847a.751.751 0 0 1-1.485-.212c.084-.593.337-1.078.621-1.489.203-.292.45-.584.673-.848.075-.088.147-.173.213-.253.561-.679.985-1.32.985-2.304 0-2.06-1.637-3.75-4-3.75ZM5.75 12h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5ZM6 15.25a.75.75 0 0 1 .75-.75h2.5a.75.75 0 0 1 0 1.5h-2.5a.75.75 0 0 1-.75-.75Z"></path></svg>Tip</span><br></p><p><span>如果发现一本书有一句话和你的句子的一部分相同，就继续往前后查找，看整段话是否更相似。</span></p></div></li><li><p><strong><span>评分系统（Scoring System）</span></strong><span>BLAST会对每次比对结果进行打分  </span></p></li><li><p><strong><span>返回最佳结果（High-Scoring Pair, HSP）</span></strong></p><p><span>BLAST会对所有比对结果进行排序，返回得分最高的片段（</span><strong><span>HSP</span></strong><span>）。这些结果通常用 </span><strong><span>E-value</span></strong><span> 衡量（值越小，匹配越可靠）。</span></p><div class="md-alert md-alert-note note"><p><span class='md-alert-text md-alert-text-note'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M0 8a8 8 0 1 1 16 0A8 8 0 0 1 0 8Zm8-6.5a6.5 6.5 0 1 0 0 13 6.5 6.5 0 0 0 0-13ZM6.5 7.75A.75.75 0 0 1 7.25 7h1a.75.75 0 0 1 .75.75v2.75h.25a.75.75 0 0 1 0 1.5h-2a.75.75 0 0 1 0-1.5h.25v-2h-.25a.75.75 0 0 1-.75-.75ZM8 6a1 1 0 1 1 0-2 1 1 0 0 1 0 2Z"></path></svg>Note</span><br></p><p><strong><span>E-value</span></strong><span>表示在给定数据库规模下，随机出现某个比对得分或更高得分的</span><strong><span>预期次数</span></strong><span>(其不为概率，应大于1)。其公式如下：</span></p><div contenteditable="false" spellcheck="false" class="mathjax-block md-end-block md-math-block md-rawblock" id="mathjax-n1951" cid="n1951" mdtype="math_block" data-math-tag-before="0" data-math-tag-after="0" data-math-labels="[]"><div class="md-math-container"><mjx-container class="MathJax" jax="SVG" display="true" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="19.456ex" height="2.248ex" role="img" focusable="false" viewBox="0 -911.5 8599.3 993.5" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -0.186ex;"><defs><path id="MJX-3-TEX-I-1D438" d="M492 213Q472 213 472 226Q472 230 477 250T482 285Q482 316 461 323T364 330H312Q311 328 277 192T243 52Q243 48 254 48T334 46Q428 46 458 48T518 61Q567 77 599 117T670 248Q680 270 683 272Q690 274 698 274Q718 274 718 261Q613 7 608 2Q605 0 322 0H133Q31 0 31 11Q31 13 34 25Q38 41 42 43T65 46Q92 46 125 49Q139 52 144 61Q146 66 215 342T285 622Q285 629 281 629Q273 632 228 634H197Q191 640 191 642T193 659Q197 676 203 680H757Q764 676 764 669Q764 664 751 557T737 447Q735 440 717 440H705Q698 445 698 453L701 476Q704 500 704 528Q704 558 697 578T678 609T643 625T596 632T532 634H485Q397 633 392 631Q388 629 386 622Q385 619 355 499T324 377Q347 376 372 376H398Q464 376 489 391T534 472Q538 488 540 490T557 493Q562 493 565 493T570 492T572 491T574 487T577 483L544 351Q511 218 508 216Q505 213 492 213Z"></path><path id="MJX-3-TEX-N-3D" d="M56 347Q56 360 70 367H707Q722 359 722 347Q722 336 708 328L390 327H72Q56 332 56 347ZM56 153Q56 168 72 173H708Q722 163 722 153Q722 140 707 133H70Q56 140 56 153Z"></path><path id="MJX-3-TEX-I-1D43E" d="M285 628Q285 635 228 637Q205 637 198 638T191 647Q191 649 193 661Q199 681 203 682Q205 683 214 683H219Q260 681 355 681Q389 681 418 681T463 682T483 682Q500 682 500 674Q500 669 497 660Q496 658 496 654T495 648T493 644T490 641T486 639T479 638T470 637T456 637Q416 636 405 634T387 623L306 305Q307 305 490 449T678 597Q692 611 692 620Q692 635 667 637Q651 637 651 648Q651 650 654 662T659 677Q662 682 676 682Q680 682 711 681T791 680Q814 680 839 681T869 682Q889 682 889 672Q889 650 881 642Q878 637 862 637Q787 632 726 586Q710 576 656 534T556 455L509 418L518 396Q527 374 546 329T581 244Q656 67 661 61Q663 59 666 57Q680 47 717 46H738Q744 38 744 37T741 19Q737 6 731 0H720Q680 3 625 3Q503 3 488 0H478Q472 6 472 9T474 27Q478 40 480 43T491 46H494Q544 46 544 71Q544 75 517 141T485 216L427 354L359 301L291 248L268 155Q245 63 245 58Q245 51 253 49T303 46H334Q340 37 340 35Q340 19 333 5Q328 0 317 0Q314 0 280 1T180 2Q118 2 85 2T49 1Q31 1 31 11Q31 13 34 25Q38 41 42 43T65 46Q92 46 125 49Q139 52 144 61Q147 65 216 339T285 628Z"></path><path id="MJX-3-TEX-N-22C5" d="M78 250Q78 274 95 292T138 310Q162 310 180 294T199 251Q199 226 182 208T139 190T96 207T78 250Z"></path><path id="MJX-3-TEX-I-1D45A" d="M21 287Q22 293 24 303T36 341T56 388T88 425T132 442T175 435T205 417T221 395T229 376L231 369Q231 367 232 367L243 378Q303 442 384 442Q401 442 415 440T441 433T460 423T475 411T485 398T493 385T497 373T500 364T502 357L510 367Q573 442 659 442Q713 442 746 415T780 336Q780 285 742 178T704 50Q705 36 709 31T724 26Q752 26 776 56T815 138Q818 149 821 151T837 153Q857 153 857 145Q857 144 853 130Q845 101 831 73T785 17T716 -10Q669 -10 648 17T627 73Q627 92 663 193T700 345Q700 404 656 404H651Q565 404 506 303L499 291L466 157Q433 26 428 16Q415 -11 385 -11Q372 -11 364 -4T353 8T350 18Q350 29 384 161L420 307Q423 322 423 345Q423 404 379 404H374Q288 404 229 303L222 291L189 157Q156 26 151 16Q138 -11 108 -11Q95 -11 87 -5T76 7T74 17Q74 30 112 181Q151 335 151 342Q154 357 154 369Q154 405 129 405Q107 405 92 377T69 316T57 280Q55 278 41 278H27Q21 284 21 287Z"></path><path id="MJX-3-TEX-I-1D45B" d="M21 287Q22 293 24 303T36 341T56 388T89 425T135 442Q171 442 195 424T225 390T231 369Q231 367 232 367L243 378Q304 442 382 442Q436 442 469 415T503 336T465 179T427 52Q427 26 444 26Q450 26 453 27Q482 32 505 65T540 145Q542 153 560 153Q580 153 580 145Q580 144 576 130Q568 101 554 73T508 17T439 -10Q392 -10 371 17T350 73Q350 92 386 193T423 345Q423 404 379 404H374Q288 404 229 303L222 291L189 157Q156 26 151 16Q138 -11 108 -11Q95 -11 87 -5T76 7T74 17Q74 30 112 180T152 343Q153 348 153 366Q153 405 129 405Q91 405 66 305Q60 285 60 284Q58 278 41 278H27Q21 284 21 287Z"></path><path id="MJX-3-TEX-I-1D452" d="M39 168Q39 225 58 272T107 350T174 402T244 433T307 442H310Q355 442 388 420T421 355Q421 265 310 237Q261 224 176 223Q139 223 138 221Q138 219 132 186T125 128Q125 81 146 54T209 26T302 45T394 111Q403 121 406 121Q410 121 419 112T429 98T420 82T390 55T344 24T281 -1T205 -11Q126 -11 83 42T39 168ZM373 353Q367 405 305 405Q272 405 244 391T199 357T170 316T154 280T149 261Q149 260 169 260Q282 260 327 284T373 353Z"></path><path id="MJX-3-TEX-N-2212" d="M84 237T84 250T98 270H679Q694 262 694 250T679 230H98Q84 237 84 250Z"></path><path id="MJX-3-TEX-I-1D706" d="M166 673Q166 685 183 694H202Q292 691 316 644Q322 629 373 486T474 207T524 67Q531 47 537 34T546 15T551 6T555 2T556 -2T550 -11H482Q457 3 450 18T399 152L354 277L340 262Q327 246 293 207T236 141Q211 112 174 69Q123 9 111 -1T83 -12Q47 -12 47 20Q47 37 61 52T199 187Q229 216 266 252T321 306L338 322Q338 323 288 462T234 612Q214 657 183 657Q166 657 166 673Z"></path><path id="MJX-3-TEX-I-1D446" d="M308 24Q367 24 416 76T466 197Q466 260 414 284Q308 311 278 321T236 341Q176 383 176 462Q176 523 208 573T273 648Q302 673 343 688T407 704H418H425Q521 704 564 640Q565 640 577 653T603 682T623 704Q624 704 627 704T632 705Q645 705 645 698T617 577T585 459T569 456Q549 456 549 465Q549 471 550 475Q550 478 551 494T553 520Q553 554 544 579T526 616T501 641Q465 662 419 662Q362 662 313 616T263 510Q263 480 278 458T319 427Q323 425 389 408T456 390Q490 379 522 342T554 242Q554 216 546 186Q541 164 528 137T492 78T426 18T332 -20Q320 -22 298 -22Q199 -22 144 33L134 44L106 13Q83 -14 78 -18T65 -22Q52 -22 52 -14Q52 -11 110 221Q112 227 130 227H143Q149 221 149 216Q149 214 148 207T144 186T142 153Q144 114 160 87T203 47T255 29T308 24Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="mi"><use data-c="1D438" xlink:href="#MJX-3-TEX-I-1D438"></use></g><g data-mml-node="mo" transform="translate(1041.8,0)"><use data-c="3D" xlink:href="#MJX-3-TEX-N-3D"></use></g><g data-mml-node="mi" transform="translate(2097.6,0)"><use data-c="1D43E" xlink:href="#MJX-3-TEX-I-1D43E"></use></g><g data-mml-node="mo" transform="translate(3208.8,0)"><use data-c="22C5" xlink:href="#MJX-3-TEX-N-22C5"></use></g><g data-mml-node="mi" transform="translate(3709,0)"><use data-c="1D45A" xlink:href="#MJX-3-TEX-I-1D45A"></use></g><g data-mml-node="mo" transform="translate(4809.2,0)"><use data-c="22C5" xlink:href="#MJX-3-TEX-N-22C5"></use></g><g data-mml-node="mi" transform="translate(5309.4,0)"><use data-c="1D45B" xlink:href="#MJX-3-TEX-I-1D45B"></use></g><g data-mml-node="mo" transform="translate(6131.7,0)"><use data-c="22C5" xlink:href="#MJX-3-TEX-N-22C5"></use></g><g data-mml-node="msup" transform="translate(6631.9,0)"><g data-mml-node="mi"><use data-c="1D452" xlink:href="#MJX-3-TEX-I-1D452"></use></g><g data-mml-node="TeXAtom" transform="translate(499,413) scale(0.707)" data-mjx-texclass="ORD"><g data-mml-node="mo"><use data-c="2212" xlink:href="#MJX-3-TEX-N-2212"></use></g><g data-mml-node="mi" transform="translate(778,0)"><use data-c="1D706" xlink:href="#MJX-3-TEX-I-1D706"></use></g><g data-mml-node="mi" transform="translate(1361,0)"><use data-c="1D446" xlink:href="#MJX-3-TEX-I-1D446"></use></g></g></g></g></g></svg><mjx-assistive-mml unselectable="on" display="block"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><mi>E</mi><mo>=</mo><mi>K</mi><mo>⋅</mo><mi>m</mi><mo>⋅</mo><mi>n</mi><mo>⋅</mo><msup><mi>e</mi><mrow data-mjx-texclass="ORD"><mo>−</mo><mi>λ</mi><mi>S</mi></mrow></msup></math></mjx-assistive-mml></mjx-container></div></div><p><mjx-container class="MathJax" jax="SVG" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="1.729ex" height="1.538ex" role="img" focusable="false" viewBox="0 -680 764 680" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: 0px;"><defs><path id="MJX-23-TEX-I-1D438" d="M492 213Q472 213 472 226Q472 230 477 250T482 285Q482 316 461 323T364 330H312Q311 328 277 192T243 52Q243 48 254 48T334 46Q428 46 458 48T518 61Q567 77 599 117T670 248Q680 270 683 272Q690 274 698 274Q718 274 718 261Q613 7 608 2Q605 0 322 0H133Q31 0 31 11Q31 13 34 25Q38 41 42 43T65 46Q92 46 125 49Q139 52 144 61Q146 66 215 342T285 622Q285 629 281 629Q273 632 228 634H197Q191 640 191 642T193 659Q197 676 203 680H757Q764 676 764 669Q764 664 751 557T737 447Q735 440 717 440H705Q698 445 698 453L701 476Q704 500 704 528Q704 558 697 578T678 609T643 625T596 632T532 634H485Q397 633 392 631Q388 629 386 622Q385 619 355 499T324 377Q347 376 372 376H398Q464 376 489 391T534 472Q538 488 540 490T557 493Q562 493 565 493T570 492T572 491T574 487T577 483L544 351Q511 218 508 216Q505 213 492 213Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="mi"><use data-c="1D438" xlink:href="#MJX-23-TEX-I-1D438"></use></g></g></g></svg><mjx-assistive-mml unselectable="on" display="inline"><math xmlns="http://www.w3.org/1998/Math/MathML"><mi>E</mi></math></mjx-assistive-mml></mjx-container><script type="math/tex"> E </script><span>：期望值，表示随机比对中出现得分不低于 </span><mjx-container class="MathJax" jax="SVG" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="1.459ex" height="1.645ex" role="img" focusable="false" viewBox="0 -705 645 727" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -0.05ex;"><defs><path id="MJX-41-TEX-I-1D446" d="M308 24Q367 24 416 76T466 197Q466 260 414 284Q308 311 278 321T236 341Q176 383 176 462Q176 523 208 573T273 648Q302 673 343 688T407 704H418H425Q521 704 564 640Q565 640 577 653T603 682T623 704Q624 704 627 704T632 705Q645 705 645 698T617 577T585 459T569 456Q549 456 549 465Q549 471 550 475Q550 478 551 494T553 520Q553 554 544 579T526 616T501 641Q465 662 419 662Q362 662 313 616T263 510Q263 480 278 458T319 427Q323 425 389 408T456 390Q490 379 522 342T554 242Q554 216 546 186Q541 164 528 137T492 78T426 18T332 -20Q320 -22 298 -22Q199 -22 144 33L134 44L106 13Q83 -14 78 -18T65 -22Q52 -22 52 -14Q52 -11 110 221Q112 227 130 227H143Q149 221 149 216Q149 214 148 207T144 186T142 153Q144 114 160 87T203 47T255 29T308 24Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="mi"><use data-c="1D446" xlink:href="#MJX-41-TEX-I-1D446"></use></g></g></g></svg><mjx-assistive-mml unselectable="on" display="inline"><math xmlns="http://www.w3.org/1998/Math/MathML"><mi>S</mi></math></mjx-assistive-mml></mjx-container><script type="math/tex"> S </script><span> 的次数。</span></p><p><mjx-container class="MathJax" jax="SVG" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="2.011ex" height="1.545ex" role="img" focusable="false" viewBox="0 -683 889 683" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: 0px;"><defs><path id="MJX-25-TEX-I-1D43E" d="M285 628Q285 635 228 637Q205 637 198 638T191 647Q191 649 193 661Q199 681 203 682Q205 683 214 683H219Q260 681 355 681Q389 681 418 681T463 682T483 682Q500 682 500 674Q500 669 497 660Q496 658 496 654T495 648T493 644T490 641T486 639T479 638T470 637T456 637Q416 636 405 634T387 623L306 305Q307 305 490 449T678 597Q692 611 692 620Q692 635 667 637Q651 637 651 648Q651 650 654 662T659 677Q662 682 676 682Q680 682 711 681T791 680Q814 680 839 681T869 682Q889 682 889 672Q889 650 881 642Q878 637 862 637Q787 632 726 586Q710 576 656 534T556 455L509 418L518 396Q527 374 546 329T581 244Q656 67 661 61Q663 59 666 57Q680 47 717 46H738Q744 38 744 37T741 19Q737 6 731 0H720Q680 3 625 3Q503 3 488 0H478Q472 6 472 9T474 27Q478 40 480 43T491 46H494Q544 46 544 71Q544 75 517 141T485 216L427 354L359 301L291 248L268 155Q245 63 245 58Q245 51 253 49T303 46H334Q340 37 340 35Q340 19 333 5Q328 0 317 0Q314 0 280 1T180 2Q118 2 85 2T49 1Q31 1 31 11Q31 13 34 25Q38 41 42 43T65 46Q92 46 125 49Q139 52 144 61Q147 65 216 339T285 628Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="mi"><use data-c="1D43E" xlink:href="#MJX-25-TEX-I-1D43E"></use></g></g></g></svg><mjx-assistive-mml unselectable="on" display="inline"><math xmlns="http://www.w3.org/1998/Math/MathML"><mi>K</mi></math></mjx-assistive-mml></mjx-container><script type="math/tex"> K </script><span>：与搜索空间维度及特性相关的常数。</span></p><p><mjx-container class="MathJax" jax="SVG" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="1.986ex" height="1.025ex" role="img" focusable="false" viewBox="0 -442 878 453" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -0.025ex;"><defs><path id="MJX-26-TEX-I-1D45A" d="M21 287Q22 293 24 303T36 341T56 388T88 425T132 442T175 435T205 417T221 395T229 376L231 369Q231 367 232 367L243 378Q303 442 384 442Q401 442 415 440T441 433T460 423T475 411T485 398T493 385T497 373T500 364T502 357L510 367Q573 442 659 442Q713 442 746 415T780 336Q780 285 742 178T704 50Q705 36 709 31T724 26Q752 26 776 56T815 138Q818 149 821 151T837 153Q857 153 857 145Q857 144 853 130Q845 101 831 73T785 17T716 -10Q669 -10 648 17T627 73Q627 92 663 193T700 345Q700 404 656 404H651Q565 404 506 303L499 291L466 157Q433 26 428 16Q415 -11 385 -11Q372 -11 364 -4T353 8T350 18Q350 29 384 161L420 307Q423 322 423 345Q423 404 379 404H374Q288 404 229 303L222 291L189 157Q156 26 151 16Q138 -11 108 -11Q95 -11 87 -5T76 7T74 17Q74 30 112 181Q151 335 151 342Q154 357 154 369Q154 405 129 405Q107 405 92 377T69 316T57 280Q55 278 41 278H27Q21 284 21 287Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="mi"><use data-c="1D45A" xlink:href="#MJX-26-TEX-I-1D45A"></use></g></g></g></svg><mjx-assistive-mml unselectable="on" display="inline"><math xmlns="http://www.w3.org/1998/Math/MathML"><mi>m</mi></math></mjx-assistive-mml></mjx-container><script type="math/tex"> m </script><span>：查询序列（query sequence）的长度。</span></p><p><mjx-container class="MathJax" jax="SVG" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="1.357ex" height="1.025ex" role="img" focusable="false" viewBox="0 -442 600 453" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -0.025ex;"><defs><path id="MJX-27-TEX-I-1D45B" d="M21 287Q22 293 24 303T36 341T56 388T89 425T135 442Q171 442 195 424T225 390T231 369Q231 367 232 367L243 378Q304 442 382 442Q436 442 469 415T503 336T465 179T427 52Q427 26 444 26Q450 26 453 27Q482 32 505 65T540 145Q542 153 560 153Q580 153 580 145Q580 144 576 130Q568 101 554 73T508 17T439 -10Q392 -10 371 17T350 73Q350 92 386 193T423 345Q423 404 379 404H374Q288 404 229 303L222 291L189 157Q156 26 151 16Q138 -11 108 -11Q95 -11 87 -5T76 7T74 17Q74 30 112 180T152 343Q153 348 153 366Q153 405 129 405Q91 405 66 305Q60 285 60 284Q58 278 41 278H27Q21 284 21 287Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="mi"><use data-c="1D45B" xlink:href="#MJX-27-TEX-I-1D45B"></use></g></g></g></svg><mjx-assistive-mml unselectable="on" display="inline"><math xmlns="http://www.w3.org/1998/Math/MathML"><mi>n</mi></math></mjx-assistive-mml></mjx-container><script type="math/tex"> n </script><span>：数据库的大小。</span></p><p><mjx-container class="MathJax" jax="SVG" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="1.459ex" height="1.645ex" role="img" focusable="false" viewBox="0 -705 645 727" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -0.05ex;"><defs><path id="MJX-41-TEX-I-1D446" d="M308 24Q367 24 416 76T466 197Q466 260 414 284Q308 311 278 321T236 341Q176 383 176 462Q176 523 208 573T273 648Q302 673 343 688T407 704H418H425Q521 704 564 640Q565 640 577 653T603 682T623 704Q624 704 627 704T632 705Q645 705 645 698T617 577T585 459T569 456Q549 456 549 465Q549 471 550 475Q550 478 551 494T553 520Q553 554 544 579T526 616T501 641Q465 662 419 662Q362 662 313 616T263 510Q263 480 278 458T319 427Q323 425 389 408T456 390Q490 379 522 342T554 242Q554 216 546 186Q541 164 528 137T492 78T426 18T332 -20Q320 -22 298 -22Q199 -22 144 33L134 44L106 13Q83 -14 78 -18T65 -22Q52 -22 52 -14Q52 -11 110 221Q112 227 130 227H143Q149 221 149 216Q149 214 148 207T144 186T142 153Q144 114 160 87T203 47T255 29T308 24Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="mi"><use data-c="1D446" xlink:href="#MJX-41-TEX-I-1D446"></use></g></g></g></svg><mjx-assistive-mml unselectable="on" display="inline"><math xmlns="http://www.w3.org/1998/Math/MathML"><mi>S</mi></math></mjx-assistive-mml></mjx-container><script type="math/tex"> S </script><span>：比对的原始得分（根据得分矩阵计算）。</span></p><p><mjx-container class="MathJax" jax="SVG" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="1.054ex" height="1.025ex" role="img" focusable="false" viewBox="0 -442 466 453" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -0.025ex;"><defs><path id="MJX-29-TEX-I-1D452" d="M39 168Q39 225 58 272T107 350T174 402T244 433T307 442H310Q355 442 388 420T421 355Q421 265 310 237Q261 224 176 223Q139 223 138 221Q138 219 132 186T125 128Q125 81 146 54T209 26T302 45T394 111Q403 121 406 121Q410 121 419 112T429 98T420 82T390 55T344 24T281 -1T205 -11Q126 -11 83 42T39 168ZM373 353Q367 405 305 405Q272 405 244 391T199 357T170 316T154 280T149 261Q149 260 169 260Q282 260 327 284T373 353Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="mi"><use data-c="1D452" xlink:href="#MJX-29-TEX-I-1D452"></use></g></g></g></svg><mjx-assistive-mml unselectable="on" display="inline"><math xmlns="http://www.w3.org/1998/Math/MathML"><mi>e</mi></math></mjx-assistive-mml></mjx-container><script type="math/tex"> e </script><span>：自然对数</span></p></div></li><li><p><strong><span>序列比对(alignment)得出identify(一致度)</span></strong></p><p><span>BLAST会对所有（</span><strong><span>HSP</span></strong><span>)片段与query序列进行alignment, 得出</span><strong><span>identify(一致度)</span></strong></p><p><img src="./picture/141.png" width="700"/></p><p>&nbsp;</p></li></ul><hr /><h3 id='different-of-dnaseq-alignment-and-rnaseq-alignment'><span>Different of DNAseq alignment and RNAseq alignment</span></h3><ul><li><p><strong><span>DNAseq</span></strong><span>：DNAseq的序列通常是</span><strong><span>连续</span></strong><span>的，与参考基因组的比对无需考虑剪接事件。</span></p></li><li><p><strong><span>RNAseq</span></strong><span>：RNAseq比对需要考虑到</span><strong><span>mRNA的剪接情况</span></strong><span>，因此需要使用能够处理剪接事件的工具，如 </span><code>STAR</code><span>、</span><code>HISAT2</code><span> 等。这些工具可以将RNA序列比对到参考基因组的不同外显子上，并处理内含子区域。</span></p></li></ul><p>&nbsp;</p><hr /><h3 id='method-of-dnaseq-alignment------seed-and-extend'><span>Method of DNAseq alignment——seed and extend</span></h3><p><a href='https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4695832'><span>seed-and-extend</span></a><span> </span><strong><span>类似于blast算法</span></strong></p><p><img src="./picture/031.png" width="500"/></p><ol start='' ><li><p><span>通过扫描参考基因组序列，对参考基因组序列建立哈希表，将序列分成一定长度的小片段（k-mer），这种小片段也被称之为</span><strong><span>seed</span></strong><span>。实施基于</span><strong><span>Needleman-Wunsch</span></strong><span>算法或</span><strong><span>Smith-Waterman</span></strong><span>算法的标准动态程序来进行</span><strong><span>seed</span></strong><span>与</span><strong><span>index</span></strong><span>对齐</span></p></li><li><p><span>在目标序列中查找和种子序列相同的片段并标记，以这些标记点为锚点</span><strong><span>向左右最大限度延伸，并且中间不能有gap</span></strong></p></li><li><p><span>将不合条件的舍弃，符合条件的结果将输出保存</span></p></li></ol><ul><li><p><span>不同软件的算法在比对层面（sensitivity and precision）和运算层面（消耗的时间和内存）各异；并且选择的seed长度不同，比对和运算结果也不同</span></p><p><strong><span>短的seed</span></strong><span>可以</span><strong><span>提高allignment的敏感度</span></strong><span>，</span></p><p><strong><span>长的seed</span></strong><span>可以</span><strong><span>提高allignment速度</span></strong></p><p><span>大部分算法都会给allignment结果提供一个质量值（在</span><code>SAM/BAM</code><span>的结果有体现） </span></p></li></ul><p>&nbsp;</p><hr /><h3 id='different-of-signle-read-alignment-and-paired-end-alignment'><span>Different of Signle-read alignment and Paired-End alignment</span></h3><p><strong><span>假设一条DNA序列包含许多重复序列（下图红色）</span></strong></p><ul><li><p><span>假设单端测序片段为重复序列片段，在allignment的时候，有可能映射到多个重复序列，导致结果不准确</span></p></li><li><p><span>而双端测序，需要2端一起匹配上，才会映射，防止映射到多个重复序列</span></p></li></ul><p><img src="./picture/032.png" width="600"/></p><p>&nbsp;</p><hr /><h2 id='software-selection-of-alignment'><span>Software selection of Alignment</span></h2><ul><li><p><strong><span>Bowtie2</span></strong></p><ul><li><p><span>特别适合处理小片段的 DNA 或 RNA 序列,擅长对齐 大约 50 到 100 或 1,000 个字符。适合处理小片段、</span><strong><span>较短的 RNA-seq 或 ChIP-seq</span></strong><span> 数据</span></p></li><li><p><span>内存占用小：对于人类 基因组，其内存占用量通常约为 3.2 GB</span></p></li><li><p><span>不擅长处理长的读长或有复杂拼接结构的序列，或具有大量插入或缺失（Indel）的序列。</span></p></li></ul></li><li><p><strong><span>STAR</span></strong></p><ul><li><p><code>STAR</code><span>是ENCODE官方推荐的RNA-seq比对工具，特别适合 </span><strong><span>RNA-seq</span></strong><span> 数据的处理，尤其是需要检测可变剪接的研究</span></p></li><li><p><span>速度快，但内存消耗较大</span></p></li></ul></li><li><p><strong><span>HISAT2</span></strong></p><ul><li><p><code>HISAT2</code><span> 是基于 </span><code>Bowtie2</code><span> 开发的一款新一代比对工具，特别针对 </span><strong><span>RNA-seq</span></strong><span> 数据设计，同时也能有效处理 DNA-seq 数据。</span></p></li><li><p><span>内存占用少，比 </span><code>Bowtie2</code><span> 处理复杂剪接的能力更强</span></p></li></ul></li></ul><p>&nbsp;</p><hr /><h2 id='bowtie2'><span>Bowtie2</span></h2><h3 id='bowtie2-preparation'><span>Bowtie2 preparation</span></h3><ul><li><p><strong><span>index download</span></strong><span>：</span><a href='https://bowtie-bio.sourceforge.net/bowtie2/index.shtml'><span>Bowtie2 官方网站</span></a></p><p>&nbsp;</p></li><li><p><strong><span>Software installation</span></strong>
<code>conda install bowtie2</code></p><p>&nbsp;</p></li><li><p><strong><span>index准备</span></strong></p><ul><li><p><span>创建一个index文件夹放index，并到其目录下</span></p></li><li><p><span>安装解压程序</span><code>sudo apt-get install unzip</code></p></li><li><p><span>解压index文件</span><code>unzip hg19.zip</code></p></li><li><p><span>查看index路径</span><code>pwd</code></p><p>&nbsp;</p></li></ul></li><li><p><strong><span>Building an Bowtie2 index</span></strong></p><ul><li><p><code>bowtie2-build</code><span> builds a Bowtie index from a set of Fasta file</span></p></li><li><p><code>bowtie2-build</code><span> outputs a set of 6 files with suffixes </span><code>.1.bt2</code><span>, </span><code>.2.bt2</code><span>, </span><code>.3.bt2</code><span>, </span><code>.4.bt2</code><span>, </span><code>.rev.1.bt2</code><span>, and </span><code>.rev.2.bt2</code><span>.  </span></p><p><span>A large index these suffixes will have a </span><code>bt2l</code><span> termination.</span></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==Building a small index==</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">bowtie2-build example/reference/lambda_virus.fa example/index/lambda_virus</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==Building a large index==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">bowtie2-build <span class="cm-attribute">--large-index</span> example/reference/lambda_virus.fa example/index/lambda_virus</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 131px;"></div><div class="CodeMirror-gutters" style="display: none; height: 131px;"></div></div></div></pre><p>&nbsp;</p></li></ul></li><li><p><strong><span>Bowtie2 index inspect</span></strong></p><ul><li><p><code>bowtie2-inspect</code><span> extracts information from a Bowtie 2 index about what kind of index it is and what reference sequences were used to build it</span></p></li><li><p><span>the tool will output a FASTA file containing the sequences of the original references (with all non-A/C/G/T characters converted to Ns)</span></p></li><li><p><span>It can also be used to extract </span><strong><span>just the reference sequence names</span></strong><span> using the </span><code>-n/--names</code><span> option or a </span><strong><span>more verbose summary</span></strong><span> using the </span><code>-s/--summary</code><span> option.</span></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==output result to the hg19.fasta==</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">bowtie2-inspect index/bowtie2/hg19/hg19 <span class="cm-attribute">-o</span> index/bowtie2/hg19/hg19.fasta</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-builtin">grep</span> <span class="cm-string">"^&gt;"</span> index/bowtie2/hg19/hg19.fasta | head <span class="cm-attribute">-n</span> <span class="cm-number">10</span> &nbsp;<span class="cm-comment"># 查看前10条序列的头部信息</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==just the reference sequence names==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">bowtie2-inspect <span class="cm-attribute">--n</span> example/index/lambda_virus</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==more verbose summary==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">bowtie2-inspect <span class="cm-attribute">--summary</span> example/index/lambda_virus </span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 240px;"></div><div class="CodeMirror-gutters" style="display: none; height: 240px;"></div></div></div></pre><p>&nbsp;</p></li></ul></li></ul><h3 id='bowtie2-parameters'><span>Bowtie2 parameters</span></h3><p><a href='https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#the-bowtie2-aligner'><span>Bowtie 2: Manual</span></a></p><p>&nbsp;</p><p><strong><span>Bowtie2 的比对流程分为两个阶段：</span></strong></p><ol start='' ><li><p><strong><span>快速单错配阶段（Fast 1-mismatch phase）</span></strong></p><ul><li><p><span>对每个seed，允许最多 </span><strong><span>1个碱基错配</span></strong><span>。</span></p></li><li><p><span>使用 </span><strong><span>哈希表或FM-index</span></strong><span> 快速定位基因组中的候选位置。</span></p></li><li><p><span>若找到至少一个候选位置，立即终止搜索并报告结果。</span></p></li></ul></li><li><p><strong><span>最优种子扫描阶段（Optimal seed search phase）</span></strong></p><ul><li><p><span>快速阶段未找到任何有效比对，对每个seed进行比对</span></p></li><li><p><span>根据比对得分（如匹配/错配的权重、gap惩罚）选择全局最优解</span></p></li></ul></li></ol><p>&nbsp;</p><p><strong><span>parameters</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">bowtie2</span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==输入=== &nbsp; </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-x</span> &lt;bt2-idx&gt; <span class="cm-comment"># 指定Bowtie2索引文件的路径和前缀（去掉后缀.bt2）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-1</span> &lt;m1&gt; <span class="cm-attribute">-2</span> &lt;m2&gt; &nbsp; <span class="cm-comment"># 指定 paired-end序列文件</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-U</span> &lt;r&gt; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 指定 single-read的序列文件</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--interleaved</span> &lt;i&gt; <span class="cm-comment"># 指定包含paired-end的交错文件 (单个文件，其中 read1 和 read2 交替存储)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==输入选项==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-q</span> <span class="cm-comment"># 输入文件为FASTQ格式（默认）。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-f</span> <span class="cm-comment"># 输入文件为FASTA格式。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--tab5</span><span class="cm-tab" role="presentation" cm-text="	">  </span><span class="cm-comment"># 输入为 TAB5 格式（旧式 Illumina 格式）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--qseq</span><span class="cm-tab" role="presentation" cm-text="	">  </span><span class="cm-comment"># 输入为 Illumina QSEQ 格式</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-c</span>：&lt;m1&gt;, &lt;m2&gt;, &lt;r&gt; </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; <span class="cm-comment"># 当你使用-c时，你不需要指定输入文件，而是将序列直接写在命令中，序列以字符串形式输入</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-r</span> &nbsp; &nbsp; <span class="cm-comment"># 输入文件为无FASTA/FASTQ头部的纯序列格式</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-F</span> k:&lt;int&gt;,i:&lt;int&gt; &nbsp;<span class="cm-comment"># 从连续的FASTA文件（如长序列组装结果）中提取k-mer，并按指定间隔比对（k长度，i间隔）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 主要用于宏基因组或长序列的局部比对分析</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--phred33</span> &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 质量分数为Phred+33编码（默认）。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--phred64</span> &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 质量分数为Phred+64编码。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--int-quals</span> &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 质量分数为空格分隔的整数</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--ignore-quals</span> &nbsp; &nbsp;<span class="cm-comment"># 忽略质量分数</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==BAM文件输入控制==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-b</span> &lt;bam&gt; <span class="cm-comment"># 指定未比对的BAM文件（需按read name排序） </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--align-paired-reads</span> &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 强制将BAM中的reads作为双端处理</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--preserve-tags</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 保留原始BAM的扩展标签</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==序列预处理选项==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-s</span>/--skip &lt;int&gt; &nbsp;<span class="cm-comment"># 跳过输入中的前&lt;int&gt;个读取或配对（默认不跳过）。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-u</span>/--upto &lt;int&gt; &nbsp;<span class="cm-comment"># 在处理前&lt;int&gt;个读取或配对后停止（默认不限制）。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-5</span>/--trim5 &lt;int&gt; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 从读取的5'/左端剪切&lt;int&gt;个碱基（默认0）。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-3</span>/--trim3 &lt;int&gt; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 从读取的3'/右端剪切&lt;int&gt;个碱基（默认0）。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--trim-to</span> [3:|5:]&lt;int&gt; &nbsp; &nbsp;<span class="cm-comment"># 截断reads至&lt;int&gt;长度（默认3'端）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==比对模式==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--end-to-end</span> <span class="cm-comment"># 要求整个读序列进行比对，不允许部分比对（默认开启）。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--very-fast</span> <span class="cm-comment"># 对应于-D 5 -R 1 -N 0 -L 22 -i S,0,2.50。速度非常快。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--fast</span> <span class="cm-comment">#对应于-D 10 -R 2 -N 0 -L 22 -i S,0,2.50。速度较快。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--sensitive</span> <span class="cm-comment">#对应于-D 15 -R 2 -N 0 -L 22 -i S,1,1.15（默认）。灵敏度较高。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--very-sensitive</span> <span class="cm-comment">#对应于-D 20 -R 3 -N 0 -L 20 -i S,1,0.50。灵敏度非常高</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--local</span> <span class="cm-comment"># 允许局部比对，序列末端可能被软剪切</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--very-fast-local</span> <span class="cm-comment"># 对应于-D 5 -R 1 -N 0 -L 25 -i S,1,2.00。速度非常快。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--fast-local</span> <span class="cm-comment"># 对应于-D 10 -R 2 -N 0 -L 22 -i S,1,1.75。速度较快。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--sensitive-local</span> <span class="cm-comment"># 对应于-D 15 -R 2 -N 0 -L 20 -i S,1,0.75（默认）。灵敏度较高。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--very-sensitive-local</span> <span class="cm-comment"># 对应于-D 20 -R 3 -N 0 -L 20 -i S,1,0.50。灵敏度非常高。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==比对核心参数==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-D</span> &lt;int&gt; &nbsp;<span class="cm-comment"># 设置在延长（扩展）比对时，如果连续遇到int次失败，便会放弃继续延长。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-R</span> &lt;int&gt; &nbsp;<span class="cm-comment"># 当遇到重复的seed序列（即多个地方可能与参考序列匹配的seed序列），它会尝试int组不同的seed序列（其他位置）来进行比对。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-N</span> &lt;int&gt; &nbsp;<span class="cm-comment"># 允许seed序列比对中最多有int个碱基错配（mismatch）。can be 0 or 1 (deafult:0)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-L</span> &lt;int&gt; &nbsp;<span class="cm-comment"># 设置seed序列的长度为int个碱基。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-i</span> S,&lt;int1&gt;,&lt;int2&gt; </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 设置seed序列的间隔函数，用于决定seed序列比对的策略。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># S表示间隔函数类型，这里是线性间隔函数（S for "Step" function）。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># int1表示seed序列之间的距离</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># int2这是间隔的增长率。这个值决定了seed序列之间的间隔随着比对过程的推进如何增长。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 具体来说，每提取一个新的seed序列，其位置会在之前seed序列位置的基础上，按照公式 int2 × &lt;current seed length&gt; 向前移动。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 较高的增长率（如 2.50）适用于对速度有较高要求的场景，而较低的增长率（如 0.50）则更适合对准确性要求较高的场合</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--n-ceil</span> &lt;func&gt; &nbsp; <span class="cm-comment"># 允许非ATCG的最大比例（默认L,0,0.15） 详细见下面tips</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--gbar</span> &lt;int&gt; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 禁止在reads两端&lt;int&gt;碱基内出现gap（默认4）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--ignore-quals</span> &nbsp; &nbsp;<span class="cm-comment"># 忽略质量分数</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--nofw</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 不比对正向序列</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--norc</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 不比对反向互补序列</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--no-1mm-upfront</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 禁止快速单错配预比对模式，直接进行最优种子扫描阶段。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 耗时较长</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" class="cm-tab-wrap-hack" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 启用 --no-1mm-upfront 时会忽略 -N 1<span class="cm-tab" role="presentation" cm-text="	">   </span></span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==for paired-end==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-I</span>/--minins &lt;int&gt; <span class="cm-comment"># 最小插入片段长度（默认0）。详细见下面tips</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-X</span>/--maxins &lt;int&gt; <span class="cm-comment"># 最大插入片段长度（默认500）。详细见下面tips</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--fr</span>/--rf/--ff &nbsp; &nbsp;<span class="cm-comment"># 定义两个 reads 的 预期相对方向</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># --fr:  R1 --&gt; &nbsp; &nbsp; &nbsp; &lt;-- R2</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># --rf:  R1 &lt;-- &nbsp; &nbsp; &nbsp; --&gt; R2</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># --ff:  R1 --&gt; &nbsp; &nbsp; &nbsp; --&gt; R2</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; <span class="cm-attribute">--no-mixed</span> &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># paired-end reads必须同时成功比对，否则丢弃该对read</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; <span class="cm-attribute">--no-discordant</span> &nbsp; &nbsp;<span class="cm-comment"># 仅输出符合 -I/-X 长度范围和 --fr/--rf/--ff 方向的配对。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; <span class="cm-attribute">--dovetail</span> &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 允许 部分重叠或延伸 的 reads 被视为有效配对</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; <span class="cm-attribute">--no-contain</span> &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 如果一个 read 的比对区域完全包含另一个 read 的比对区域，则不算有效配对</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; <span class="cm-attribute">--no-overlap</span> &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 如果两个 reads 的比对区域有任何重叠，则不算有效配对</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==输出==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-S</span> &lt;sam&gt; &nbsp; &nbsp;<span class="cm-comment"># 输出SAM格式的文件。如果未指定，结果输出到标准输出。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-t</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 打印各阶段耗时</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--un</span> &lt;path&gt; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 将未比对的单端读序列写入指定路径。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--al</span> &lt;path&gt; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 将至少比对过一次的单端读序列写入指定路径。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--un-conc</span> &lt;path&gt; &nbsp;<span class="cm-comment"># 输出非一致配对的reads</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--al-conc</span> &lt;path&gt; &nbsp;<span class="cm-comment"># 输出一致配对的reads</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--no-unal</span> &nbsp; <span class="cm-comment"># suppress SAM records for unaligned reads</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--no-head</span> &nbsp; <span class="cm-comment"># suppress header lines, i.e. lines starting with @</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--no-sq</span> &nbsp; &nbsp; <span class="cm-comment"># suppress @SQ header lines</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--rg-id</span> &lt;text&gt; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 设置SAM的@RG ID(记录测序数据的信息)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--rg</span> &lt;text&gt; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 添加@RG字段（需配合--rg-id）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;  &amp;&gt; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 将命令的标准输出和错误输出重定向到指定的日志文件</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==性能优化与扩展== </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-p</span>/--threads &lt;int&gt; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment">#指定使用多少线程进行比对（默认1）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--mm</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 使用内存映射索引（多进程共享）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 当你运行多个 bowtie2 进程时，它们可以共享索引数据的内存映射副本，而不是每个进程都单独载入一份索引，节省了内存并加快启动速度</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--reorder</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 强制SAM输出顺序与输入一致</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--seed</span> &lt;int&gt; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 随机数种子（默认0）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--non-deterministic</span> &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 使用非确定性随机种子</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 3060px;"></div><div class="CodeMirror-gutters" style="display: none; height: 3060px;"></div></div></div></pre><p>&nbsp;</p><div class="md-alert md-alert-important important"><p><span class='md-alert-text md-alert-text-important'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M0 1.75C0 .784.784 0 1.75 0h12.5C15.216 0 16 .784 16 1.75v9.5A1.75 1.75 0 0 1 14.25 13H8.06l-2.573 2.573A1.458 1.458 0 0 1 3 14.543V13H1.75A1.75 1.75 0 0 1 0 11.25Zm1.75-.25a.25.25 0 0 0-.25.25v9.5c0 .138.112.25.25.25h2a.75.75 0 0 1 .75.75v2.19l2.72-2.72a.749.749 0 0 1 .53-.22h6.5a.25.25 0 0 0 .25-.25v-9.5a.25.25 0 0 0-.25-.25Zm7 2.25v2.5a.75.75 0 0 1-1.5 0v-2.5a.75.75 0 0 1 1.5 0ZM9 9a1 1 0 1 1-2 0 1 1 0 0 1 2 0Z"></path></svg>Important</span><br></p><p><span>对于经过</span><strong><span>trimmed read</span></strong><span>，建议使用</span><code>--local</code><span>。</span><code>--local</code><span>模式允许比对器在read的一部分上寻找匹配，而不要求整个read都参与allignment。这非常适合</span><strong><span>trimmed read</span></strong><span>.</span></p></div><p>&nbsp;</p><div class="md-alert md-alert-tip tip"><p><span class='md-alert-text md-alert-text-tip'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M8 1.5c-2.363 0-4 1.69-4 3.75 0 .984.424 1.625.984 2.304l.214.253c.223.264.47.556.673.848.284.411.537.896.621 1.49a.75.75 0 0 1-1.484.211c-.04-.282-.163-.547-.37-.847a8.456 8.456 0 0 0-.542-.68c-.084-.1-.173-.205-.268-.32C3.201 7.75 2.5 6.766 2.5 5.25 2.5 2.31 4.863 0 8 0s5.5 2.31 5.5 5.25c0 1.516-.701 2.5-1.328 3.259-.095.115-.184.22-.268.319-.207.245-.383.453-.541.681-.208.3-.33.565-.37.847a.751.751 0 0 1-1.485-.212c.084-.593.337-1.078.621-1.489.203-.292.45-.584.673-.848.075-.088.147-.173.213-.253.561-.679.985-1.32.985-2.304 0-2.06-1.637-3.75-4-3.75ZM5.75 12h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5ZM6 15.25a.75.75 0 0 1 .75-.75h2.5a.75.75 0 0 1 0 1.5h-2.5a.75.75 0 0 1-.75-.75Z"></path></svg>Tip</span><br></p><p><strong><span>插入片段长度？</span></strong><code>-I</code><span> </span><code>-X</code><span>  Deafult </span><code>0</code><span> </span><code>500</code></p><ul><li><p><strong><span>背景</span></strong><span> 在paired-end 中，每个paired-end reads来自同一个DNA片段的两端。理论上，这两个read在参考基因组上比对的位置应该有一个合理的间隔，称为</span><strong><span>插入片段长度</span></strong></p><p><img src="./picture/079.png" width="800"/></p></li><li><p><strong><span>实际应用</span></strong></p><ul><li><p><span>选择适当的</span><strong><span>最小插入片段长度</span></strong><span>可以帮助排除那些可能由于测序或库构建过程中产生的异常情况。例如，在</span><strong><span>DNA片段化</span></strong><span>过程中，有时可能会产生</span><strong><span>非常短</span></strong><span>的read，而这些片段可能导致两个read之间的间隔太小</span></p></li><li><p><span>选择适当的</span><strong><span>最大插入片段长度</span></strong><span>可以帮助排除那些比对到不合理远的区域的read。例如，如果成对read的间隔过大，这可能表明这些read并非来自同一片段，而是由于重复序列或其他原因错误地配对在一起</span></p></li></ul></li></ul><p>&nbsp;</p><ul><li><p><strong><span>parameter suggestion</span></strong></p><ul><li><p><span>对于</span><strong><span>25x25 PE</span></strong><span>，</span><code>-I</code><span>可设置为</span><code>10</code><span> </span><code>-X</code><span>可设置为</span><code>700</code></p></li><li><p><span>对于长读长read，不建议设置</span><code>-I</code></p></li></ul></li></ul></div><p>&nbsp;</p><div class="md-alert md-alert-tip tip"><p><span class='md-alert-text md-alert-text-tip'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M8 1.5c-2.363 0-4 1.69-4 3.75 0 .984.424 1.625.984 2.304l.214.253c.223.264.47.556.673.848.284.411.537.896.621 1.49a.75.75 0 0 1-1.484.211c-.04-.282-.163-.547-.37-.847a8.456 8.456 0 0 0-.542-.68c-.084-.1-.173-.205-.268-.32C3.201 7.75 2.5 6.766 2.5 5.25 2.5 2.31 4.863 0 8 0s5.5 2.31 5.5 5.25c0 1.516-.701 2.5-1.328 3.259-.095.115-.184.22-.268.319-.207.245-.383.453-.541.681-.208.3-.33.565-.37.847a.751.751 0 0 1-1.485-.212c.084-.593.337-1.078.621-1.489.203-.292.45-.584.673-.848.075-.088.147-.173.213-.253.561-.679.985-1.32.985-2.304 0-2.06-1.637-3.75-4-3.75ZM5.75 12h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5ZM6 15.25a.75.75 0 0 1 .75-.75h2.5a.75.75 0 0 1 0 1.5h-2.5a.75.75 0 0 1-.75-.75Z"></path></svg>Tip</span><br></p><p><strong><span>允许非ATCG的最大比例</span></strong><code>--n-ceil</code><span>    Deafult: </span><code>L,0,0.15</code><span> </span></p><ul><li><p><span>其值由三部分组成：</span><code>&lt;类型&gt;,&lt;最小值&gt;,&lt;最大值&gt;</code></p><figure class='table-figure'><table><thead><tr><th style='text-align:left;' ><span>类型</span></th><th style='text-align:left;' ><span>含义</span></th><th style='text-align:left;' ><span>示例</span></th></tr></thead><tbody><tr><td style='text-align:left;' ><code>L</code></td><td style='text-align:left;' ><span>read长度（Length）</span></td><td style='text-align:left;' ><strong><code>L,0,0.15</code></strong><span>=允许 </span><code>0%-15%</code><span>个非 A/T/C/G</span></td></tr><tr><td style='text-align:left;' ><code>C</code></td><td style='text-align:left;' ><span>固定数值（Constant）</span></td><td style='text-align:left;' ><code>C,0,5</code><span> = 允许 </span><code>0-5</code><span> 个非 A/T/C/G</span></td></tr><tr><td style='text-align:left;' ><code>S</code></td><td style='text-align:left;' ><span>分段函数（Step）</span></td><td style='text-align:left;' ><code>S,1,0.1</code><span> = 每 </span><code>1</code><span>bp 允许 </span><code>0.1</code><span> 个</span></td></tr></tbody></table></figure></li></ul><p>&nbsp;</p></div><p>&nbsp;</p><p><strong><span>Example</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==单端测序==</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">bowtie2 \ </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-attribute">-p</span> <span class="cm-number">12</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-attribute">-x</span>  /mnt/e/test/index/ucsu.hg19/hg19 \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-attribute">-U</span> IP_H3K27ac_Vehicle_1_trimmed.fq.gz \ </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-attribute">-S</span> IP_H3K27ac_Vehicle.sam</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==双端测序==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">bowtie2 \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-attribute">-1</span> trim_galore/KChIP2-230929S_230929S_R1_val_1.fq.gz <span class="cm-attribute">-2</span> trim_galore/KChIP2-230929S_230929S_R2_val_2.fq.gz\</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-attribute">-x</span>  ../index/bowtie2/Rnor_6.0/Rnor_6.0\ &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># index location</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-attribute">-p</span> <span class="cm-number">10</span> \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 核心数</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-attribute">-S</span> bowtie2/KCHIIP2.sam \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 指定输出SAM文件的路径 </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-attribute">--end-to-end</span> \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 全局比对模式 &nbsp; </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-attribute">--very-sensitive</span> \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 启用非常灵敏的参数设置</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-attribute">--no-mixed</span> \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># paired-end reads必须同时成功比对，否则丢弃该对read</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-attribute">--no-discordant</span> \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 仅输出符合 -I/-X 长度范围和 --fr/--rf/--ff 方向的配对。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-attribute">--phred33</span> \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 指定质量得分的编码格式为Phred+33 &nbsp; </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-attribute">-I</span> <span class="cm-number">10</span> \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 设置paired-end reads之间的最小插入片段长度为10 bp &nbsp; &nbsp; </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-attribute">-X</span> <span class="cm-number">700</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 设置paired-end reads之间的最大插入片段长度为700 bp</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 459px;"></div><div class="CodeMirror-gutters" style="display: none; height: 459px;"></div></div></div></pre><div class="md-alert md-alert-caution caution"><p><span class='md-alert-text md-alert-text-caution'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M4.47.22A.749.749 0 0 1 5 0h6c.199 0 .389.079.53.22l4.25 4.25c.141.14.22.331.22.53v6a.749.749 0 0 1-.22.53l-4.25 4.25A.749.749 0 0 1 11 16H5a.749.749 0 0 1-.53-.22L.22 11.53A.749.749 0 0 1 0 11V5c0-.199.079-.389.22-.53Zm.84 1.28L1.5 5.31v5.38l3.81 3.81h5.38l3.81-3.81V5.31L10.69 1.5ZM8 4a.75.75 0 0 1 .75.75v3.5a.75.75 0 0 1-1.5 0v-3.5A.75.75 0 0 1 8 4Zm0 8a1 1 0 1 1 0-2 1 1 0 0 1 0 2Z"></path></svg>Caution</span><br></p><p><strong><span>比对率大于75%才比较可以</span></strong></p></div><div class="md-alert md-alert-important important"><p><span class='md-alert-text md-alert-text-important'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M0 1.75C0 .784.784 0 1.75 0h12.5C15.216 0 16 .784 16 1.75v9.5A1.75 1.75 0 0 1 14.25 13H8.06l-2.573 2.573A1.458 1.458 0 0 1 3 14.543V13H1.75A1.75 1.75 0 0 1 0 11.25Zm1.75-.25a.25.25 0 0 0-.25.25v9.5c0 .138.112.25.25.25h2a.75.75 0 0 1 .75.75v2.19l2.72-2.72a.749.749 0 0 1 .53-.22h6.5a.25.25 0 0 0 .25-.25v-9.5a.25.25 0 0 0-.25-.25Zm7 2.25v2.5a.75.75 0 0 1-1.5 0v-2.5a.75.75 0 0 1 1.5 0ZM9 9a1 1 0 1 1-2 0 1 1 0 0 1 2 0Z"></path></svg>Important</span><br></p><p><span>有时候相同物种不同版本的index其比对率也有很大差别，如果比对率过低建议更换index试一下</span></p></div><p>&nbsp;</p><hr /><p>&nbsp;</p><h2 id='use-the-ecoli-genome-to-normalize-data-optionalcuttagstep1'><span>Use the E.coli genome to normalize data (optional/CUT&amp;Tag)(step1)</span></h2><ul><li><p><strong><span>背景</span></strong><span>：</span></p><p><span>在</span><strong><span>CUT&amp;Tag</span></strong><span>实验中，</span><strong><span>Tn5的转座酶</span></strong><span>复合物带有细菌来源的DNA（如E.coli的DNA），由于</span><strong><span>pA-Tn5</span></strong><span>转座酶通常是在细菌（如E.coli）中表达和纯化的。在这个过程中，不可避免地会有一部分细菌的DNA残留</span></p></li><li><p><strong><span>非特异性标签化</span></strong><span>：</span></p><p><span>在实验中，</span><strong><span>pA-Tn5</span></strong><span>蛋白带有的E.coli DNA可能会被非特异性地标签化。因此，在测序数据中可能会发现一部分read是与E. coli基因组比对的，而不是与目标基因组比对的。</span></p></li><li><p><strong><span>恒定的E.coli DNA量</span></strong><span>：</span></p><p><span>由于在实验中加入的</span><strong><span>pA-Tn5</span></strong><span>蛋白的量是恒定的，因此每个反应中的E. coli DNA量也是固定的。这意味着E. coliread可以作为一个内参（spike-in control）来进行数据的标准化处理。</span></p></li></ul><p>&nbsp;</p><ol start='' ><li><p><strong><span>比对E.coli Index</span></strong><span> </span></p><p><span>下载对应版本的E.coli 的fasta文件 ：</span><a href='https://www.ncbi.nlm.nih.gov/datasets/genome/?taxon=562' target='_blank' class='url'>https://www.ncbi.nlm.nih.gov/datasets/genome/?taxon=562</a></p><div class="md-alert md-alert-tip tip"><p><span class='md-alert-text md-alert-text-tip'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M8 1.5c-2.363 0-4 1.69-4 3.75 0 .984.424 1.625.984 2.304l.214.253c.223.264.47.556.673.848.284.411.537.896.621 1.49a.75.75 0 0 1-1.484.211c-.04-.282-.163-.547-.37-.847a8.456 8.456 0 0 0-.542-.68c-.084-.1-.173-.205-.268-.32C3.201 7.75 2.5 6.766 2.5 5.25 2.5 2.31 4.863 0 8 0s5.5 2.31 5.5 5.25c0 1.516-.701 2.5-1.328 3.259-.095.115-.184.22-.268.319-.207.245-.383.453-.541.681-.208.3-.33.565-.37.847a.751.751 0 0 1-1.485-.212c.084-.593.337-1.078.621-1.489.203-.292.45-.584.673-.848.075-.088.147-.173.213-.253.561-.679.985-1.32.985-2.304 0-2.06-1.637-3.75-4-3.75ZM5.75 12h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5ZM6 15.25a.75.75 0 0 1 .75-.75h2.5a.75.75 0 0 1 0 1.5h-2.5a.75.75 0 0 1-.75-.75Z"></path></svg>Tip</span><br></p><p><span>如果不确定版本，可以下载前两个</span></p></div></li><li><p><strong><span>将fasta文件转换成index</span></strong></p><p><code>bowtie2-build Ecoil-ASM584v2.fasta Ecoil-ASM584v2/genome</code></p></li><li><p><strong><span>比对E.coli Index</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">bowtie2 </span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; <span class="cm-attribute">--end-to-end</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment">#全局比对模式</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; <span class="cm-attribute">--local</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment">#局部比对模式</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; <span class="cm-attribute">--very-sensitive</span> &nbsp; &nbsp; &nbsp;<span class="cm-comment">#启用非常灵敏的参数设置</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; <span class="cm-attribute">--no-mixed</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment">#paired-end reads必须同时成功比对，否则丢弃该对读段</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; <span class="cm-attribute">--no-discordant</span> &nbsp; &nbsp; &nbsp; <span class="cm-comment">#paired-end reads必须在合理的距离和并且方向一致内比对，否则丢弃该对读段</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; <span class="cm-attribute">--phred33</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment">#指定质量得分的编码格式为Phred+33</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; <span class="cm-attribute">-I</span> <span class="cm-number">10</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment">#设置paired-end reads之间的最小插入片段长度为10 bp</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; <span class="cm-attribute">-X</span> <span class="cm-number">700</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment">#设置paired-end reads之间的最大插入片段长度为700 bp</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; <span class="cm-attribute">-p</span> &lt;int&gt; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment">#核心数</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; <span class="cm-attribute">-x</span> &lt;Ecoil-index&gt; </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; <span class="cm-attribute">-1</span> &lt;R1.fastq.gz&gt; </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; <span class="cm-attribute">-2</span> &lt;R2.fastq.gz&gt; </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; <span class="cm-attribute">-S</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment">#指定输出比对Ecoil的SAM文件的路径 </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &amp;&gt; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment">#将命令的标准输出和错误输出重定向到指定的日志文件</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 在与E.coil比对时候加上以xia2种参数，提供了额外的限制条件，确保了比对的准确性。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; <span class="cm-attribute">--no-overlap</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment">#防止paired-end reads比对到重叠的位置</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; <span class="cm-attribute">--no-dovetail</span> &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment">#防止paired-end reads在比对时出现交叉现象</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 525px;"></div><div class="CodeMirror-gutters" style="display: none; height: 525px;"></div></div></div></pre><p>&nbsp;</p><p><strong><span>Example</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">bowtie2 \</span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-attribute">-1</span> trim_galore/KChIP2-230929S_230929S_R1_val_1.fq.gz <span class="cm-attribute">-2</span> trim_galore/KChIP2-230929S_230929S_R2_val_2.fq.gz\</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-attribute">-x</span>  ../index/bowtie2/Ecoil-ASM584v2/genome\</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-attribute">-p</span> <span class="cm-number">10</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-attribute">-S</span> bowtie2/Ecoil/KCHIIP2.sam \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-attribute">--local</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-attribute">--very-sensitive</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-attribute">--no-mixed</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-attribute">--no-discordant</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-attribute">--phred33</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-attribute">-I</span> <span class="cm-number">10</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-attribute">-X</span> <span class="cm-number">700</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-attribute">--no-overlap</span>\</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-attribute">--no-dovetail</span> </span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 306px;"></div><div class="CodeMirror-gutters" style="display: none; height: 306px;"></div></div></div></pre></li></ol><p>&nbsp;</p><p><a href='#use-the-ecoli-genome-to-normalize-data-optionalcuttagstep2'><span>Click here to view step 2</span></a></p><p>&nbsp;</p><hr /><h1 id='samtools'><span>samtools</span></h1><ul><li><p><span>samtools是一个用于操作sam和bam文件的工具合集。能够实现二进制查看、格式转换、排序及合并等功能，结合sam格式中的flag、tag等信息，还可以完成比对结果的统计汇总</span></p></li><li><p><span>Software installation: </span><code>conda install -c bioconda samtools</code></p></li></ul><p>&nbsp;</p><hr /><h2 id='samtools-sort'><span>Samtools sort</span></h2><ul><li><p><code>samtools sort</code><span>可以将sam转化为BAM或CRAM，并进行排序</span></p></li><li><p><code>samtools sort</code><span>可以对文件就行自定义排序</span></p></li><li><p><a href='https://www.htslib.org/doc/1.14/samtools-sort.html'><span>samtools sort</span></a></p></li></ul><p>&nbsp;</p><p><strong><span>parameters</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; samtools <span class="cm-builtin">sort</span> <span class="cm-comment">#输入文件名</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==输出选项==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-o</span> <span class="cm-comment">#指定输出文件的名称</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-O</span> <span class="cm-comment">#指定输出文件的格式，如 SAM、BAM 或 CRAM</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-l</span> &lt;int&gt;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment">#设置输出文件的压缩级别，范围从 0 到 9。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment">#0 表示不压缩，9 表示最高压缩级别。压缩级别越高，输出文件的大小越小，但压缩所需时间更长</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==排序选项，默认使用的是按参考基因组位置（coordinate）进行排序==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-n</span> &nbsp;<span class="cm-comment">#按读取名称对文件进行排序</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-t</span> &lt;TAG&gt;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 根据指定的 TAG 字段的值对文件进行排序。使用位置作为次级排序标准，如果设置了 -n，则使用读取名称作为次级标准。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 例如，-t NM 可以根据比对不匹配数 (NM tag) 排序。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; <span class="cm-attribute">-T</span> &lt;PREFIX&gt;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; <span class="cm-comment">#指定排序过程中使用的临时文件的前缀，临时文件会命名为 PREFIX.nnnn.bam。排序完成后，这些临时文件会被自动删除</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; <span class="cm-attribute">--reference</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; <span class="cm-comment">#指定参考序列的 FASTA 文件。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; <span class="cm-comment">#如果你在输出 CRAM 文件时没有指定参考序列，某些操作可能会失败或输出数据不完整。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==性能优化==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-m</span> &lt;int&gt;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment">#设置每个线程可使用的最大内存量，默认为 768MB。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment">#支持使用 K（千字节）、M（兆字节）、G（千兆字节）作为后缀。例如，-m 2G </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-</span>@ &lt;int&gt; <span class="cm-comment">#指定使用的线程数量,默认值为 1，即只使用一个线程</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 852px;"></div><div class="CodeMirror-gutters" style="display: none; height: 852px;"></div></div></div></pre><p>&nbsp;</p><p><strong><span>Example</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation"><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">samtools <span class="cm-builtin">sort</span>  Input_H3K27ac_XY108.sam \</span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-attribute">-O</span> bam \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-attribute">-</span>@ <span class="cm-number">10</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-attribute">-o</span> Input_H3K27ac_XY108.bam</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 87px;"></div><div class="CodeMirror-gutters" style="display: none; height: 87px;"></div></div></div></pre><p>&nbsp;</p><hr /><h2 id='samtools-idxstats------view-the-alignment-with-chromosome'><span>Samtools idxstats——view the alignment with chromosome</span></h2><p><span>samtools idxstats 命令可以显示 BAM 文件中所有染色体（参考序列）的名称、长度，以及每个染色体上比对的片段数量。</span></p><ul><li><p><span>首先，需要确保 BAM 文件已经建立了索引文件（.bai），如果没有，可以先生成索引</span></p><p><code>samtools index</code><span> </span><code>&lt;in.bam/sam/cram&gt;</code><span>  </span></p></li><li><p><span>然后使用以下命令查看染色体信息</span></p><p><code>samtools idxstats</code><span> </span><code>&lt;in.bam/sam/cram&gt;</code></p></li><li><p><span>输出将类似如下内容</span></p><figure class='table-figure'><table><thead><tr><th><span>Chromosome</span></th><th><span>Length</span></th><th><span>Mapped Reads</span></th><th><span>Unmapped Reads</span></th></tr></thead><tbody><tr><td><span>chr1</span></td><td><span>248956422</span></td><td><span>10000</span></td><td><span>200</span></td></tr><tr><td><span>chr2</span></td><td><span>242193529</span></td><td><span>8000</span></td><td><span>150</span></td></tr><tr><td><span>chrX</span></td><td><span>156040895</span></td><td><span>5000</span></td><td><span>100</span></td></tr><tr><td><span>chrM</span></td><td><span>16569</span></td><td><span>200</span></td><td><span>50</span></td></tr><tr><td><span>*</span></td><td><span>0</span></td><td><span>0</span></td><td><span>0</span></td></tr></tbody></table></figure><ul><li><p><strong><span>Chromosome</span></strong><span>：染色体名称</span></p></li><li><p><strong><span>Length</span></strong><span>：染色体长度</span></p></li><li><p><strong><span>Mapped Reads</span></strong><span>：比对到该染色体的已比对序列数量</span></p></li><li><p><strong><span>Unmapped Reads</span></strong><span>：表示这些序列在比对时没有成功比对到该染色体，但可能被比对到其他染色体</span></p></li><li><p><span>*</span><strong><span>行</span></strong><span>：在比对过程中没有成功比对到参考基因组的任何染色体，contig或scaffold的reads总数</span></p></li></ul></li></ul><p>&nbsp;</p><hr /><h2 id='samtools-flagstat------view-the-alignment-results'><span>Samtools flagstat——view the alignment results</span></h2><ul><li><p><span>比对结束后，需要了解比对结果的情况，可以采用</span><code>samtools flagstat</code></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation"><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">samtools flagstat &lt;in.bam/sam/cram&gt; &gt;&gt; historydata.txt</span></pre></div></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 22px;"></div><div class="CodeMirror-gutters" style="display: none; height: 22px;"></div></div></div></pre></li></ul><ul><li><p><span>the explanation of alignment results</span></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-number">14608455</span> <span class="cm-operator">+</span> <span class="cm-number">0</span> <span class="cm-keyword">in</span> total (QC-passed reads <span class="cm-operator">+</span> QC-failed reads) &nbsp; &nbsp; <span class="cm-comment">## reads总数</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-number">37967</span> <span class="cm-operator">+</span> <span class="cm-number">0</span> secondary &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment">## 出现比对到参考基因组多个位置的reads数</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-number">0</span> <span class="cm-operator">+</span> <span class="cm-number">0</span> supplementary &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment">## 可能存在嵌合的reads数</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-number">0</span> <span class="cm-operator">+</span> <span class="cm-number">0</span> duplicates &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment">## 重复的reads数</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-number">14590894</span> <span class="cm-operator">+</span> <span class="cm-number">0</span> mapped (99.88% : N/A) &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment">## 比对到参考基因组上的reads数</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-number">14570488</span> <span class="cm-operator">+</span> <span class="cm-number">0</span> paired <span class="cm-keyword">in</span> sequencing &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment">## 属于PE read的reads总数。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-number">7285244</span> <span class="cm-operator">+</span> <span class="cm-number">0</span> read1 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment">## PE read中Read 1 的reads 总数。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-number">7285244</span> <span class="cm-operator">+</span> <span class="cm-number">0</span> read2 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment">## PE read中Read 2 的reads 总数。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-number">14507068</span> <span class="cm-operator">+</span> <span class="cm-number">0</span> properly paired (99.56% : N/A) &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment">## 完美比对的reads总数。PE两端reads比对到同一条序列，且根据比对结果推断的插入片段大小符合设置的阈值。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-number">14551500</span> <span class="cm-operator">+</span> <span class="cm-number">0</span> with itself and mate mapped &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment">## PE两端reads都比对上参考序列的reads总数。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-number">1427</span> <span class="cm-operator">+</span> <span class="cm-number">0</span> singletons (0.01% : N/A) &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment">## PE两端reads，其中一端比上，另一端没比上的reads总数。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-number">26260</span> <span class="cm-operator">+</span> <span class="cm-number">0</span> with mate mapped to a different chr &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment">## PE read中，两端分别比对到两条不同的序列的reads总数。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-number">17346</span> <span class="cm-operator">+</span> <span class="cm-number">0</span> with mate mapped to a different chr (mapQ&gt;<span class="cm-operator">=</span><span class="cm-number">5</span>) &nbsp; &nbsp; &nbsp; <span class="cm-comment">## PE read中，两端分别比对到两条不同的序列，且mapQ&gt;=5的reads总数。</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 284px;"></div><div class="CodeMirror-gutters" style="display: none; height: 284px;"></div></div></div></pre><div class="md-alert md-alert-important important"><p><span class='md-alert-text md-alert-text-important'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M0 1.75C0 .784.784 0 1.75 0h12.5C15.216 0 16 .784 16 1.75v9.5A1.75 1.75 0 0 1 14.25 13H8.06l-2.573 2.573A1.458 1.458 0 0 1 3 14.543V13H1.75A1.75 1.75 0 0 1 0 11.25Zm1.75-.25a.25.25 0 0 0-.25.25v9.5c0 .138.112.25.25.25h2a.75.75 0 0 1 .75.75v2.19l2.72-2.72a.749.749 0 0 1 .53-.22h6.5a.25.25 0 0 0 .25-.25v-9.5a.25.25 0 0 0-.25-.25Zm7 2.25v2.5a.75.75 0 0 1-1.5 0v-2.5a.75.75 0 0 1 1.5 0ZM9 9a1 1 0 1 1-2 0 1 1 0 0 1 2 0Z"></path></svg>Important</span><br></p><p><strong><span>测序深度 = reads总数</span></strong></p></div><div class="md-alert md-alert-tip tip"><p><span class='md-alert-text md-alert-text-tip'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M8 1.5c-2.363 0-4 1.69-4 3.75 0 .984.424 1.625.984 2.304l.214.253c.223.264.47.556.673.848.284.411.537.896.621 1.49a.75.75 0 0 1-1.484.211c-.04-.282-.163-.547-.37-.847a8.456 8.456 0 0 0-.542-.68c-.084-.1-.173-.205-.268-.32C3.201 7.75 2.5 6.766 2.5 5.25 2.5 2.31 4.863 0 8 0s5.5 2.31 5.5 5.25c0 1.516-.701 2.5-1.328 3.259-.095.115-.184.22-.268.319-.207.245-.383.453-.541.681-.208.3-.33.565-.37.847a.751.751 0 0 1-1.485-.212c.084-.593.337-1.078.621-1.489.203-.292.45-.584.673-.848.075-.088.147-.173.213-.253.561-.679.985-1.32.985-2.304 0-2.06-1.637-3.75-4-3.75ZM5.75 12h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5ZM6 15.25a.75.75 0 0 1 .75-.75h2.5a.75.75 0 0 1 0 1.5h-2.5a.75.75 0 0 1-.75-.75Z"></path></svg>Tip</span><br></p><p><code>+0</code><span> 表示没有任何reads因相关原因而被标记为未通过质量控制（QC）的reads</span></p></div></li></ul><p>&nbsp;</p><hr /><h1 id='convert-sam-to-bamcram'><span>Convert SAM to BAM/CRAM</span></h1><p><span>We use the </span><a href='#samtools-sort'><span>Samtools sort</span></a><span> to convert SAM file to BAM/CRAM file ,and sort the file by coordinate.</span></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation"><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">samtools <span class="cm-builtin">sort</span>  Input_H3K27ac_XY108.sam \</span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-attribute">-O</span> bam \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-attribute">-</span>@ <span class="cm-number">10</span> \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-attribute">-o</span> Input_H3K27ac_XY108.bam</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 87px;"></div><div class="CodeMirror-gutters" style="display: none; height: 87px;"></div></div></div></pre><div class="md-alert md-alert-caution caution"><p><span class='md-alert-text md-alert-text-caution'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M4.47.22A.749.749 0 0 1 5 0h6c.199 0 .389.079.53.22l4.25 4.25c.141.14.22.331.22.53v6a.749.749 0 0 1-.22.53l-4.25 4.25A.749.749 0 0 1 11 16H5a.749.749 0 0 1-.53-.22L.22 11.53A.749.749 0 0 1 0 11V5c0-.199.079-.389.22-.53Zm.84 1.28L1.5 5.31v5.38l3.81 3.81h5.38l3.81-3.81V5.31L10.69 1.5ZM8 4a.75.75 0 0 1 .75.75v3.5a.75.75 0 0 1-1.5 0v-3.5A.75.75 0 0 1 8 4Zm0 8a1 1 0 1 1 0-2 1 1 0 0 1 0 2Z"></path></svg>Caution</span><br></p><p><span>建议按coordinate进行排序，在remove duplicates 中的Picard MarkDuplicates需要按照coordinate排序bam文件</span></p></div><p>&nbsp;</p><hr /><h1 id='remove-duplicates'><span>Remove duplicates</span></h1><h2 id='remove-duplicates-theory'><span>Remove duplicates theory</span></h2><ul><li><p><strong><span>PCR duplicates</span></strong><span>：不同的序列在进行PCR扩增时，扩增的倍数应该是相同的。但是由于聚合酶的偏好性，PCR扩增次数过多的情况下，会导致一些序列持续扩增，而另一些序列扩增到一定程度后便不再进行，也就是我们常说的PCR偏好性。</span></p></li><li><p><strong><span>Optical duplicates</span></strong><span>：测序仪的光学传感器错误地将单个扩增的簇检测为多个簇。</span></p></li><li><p><strong><span>PCR duplicates</span></strong><span>的比例会因测序量的增加而增加，</span><strong><span>Optical duplicates</span></strong><span>是随机分布的。</span></p><table>
  <tr>
    <td><img src="./picture/038.png" style="width:650px;" /></td>
    <td><img src="./picture/039.png" style="width:750px;" /></td>
  </tr>
</table></li></ul><p>&nbsp;</p><p><strong><span>Should we remove duplicates?</span></strong></p><ul><li><p><strong><span>RNA-seq</span></strong><span>时，我们所得到的reads覆盖度通常是不均匀的（由于不同的基因表达水平不同），如果我们不加区分地remove duplicates，可能会导致大量重要信息的丢失   </span></p><ul><li><p><span>RNA-seq由于其建库起始量一般都很高所以不需要去除重复</span></p></li><li><p><span>RNA-seq数据中经常会出现某些基因的表达量十分高，因此这些reads的比对位置往往高度一致，这种“重复”是生物学本身的体现，而非技术噪音</span></p></li></ul></li><li><p><strong><span>ChIP-seq</span></strong><span>中，由于起始量不高，且没有那种富集程度很高的位点，因此通常需要考虑去除PCR重复。</span></p></li><li><p><strong><span>call SNP</span></strong><span>中，起始量一般都高（因为要保证测序深度），此外由于PCR扩增会导致一些序列复制错误，这将严重影响SNP位点识别的置信度。因此一般需要remove duplicates。</span></p></li><li><p><span>利用</span><strong><span>reads mapping的均匀程度</span></strong><span>判断是否具有重复。</span></p><ul><li><p><span>若富集位点周围的reads均匀覆盖，那么没有重复；</span></p></li><li><p><span>若富集位点周围覆盖度不均匀，某些区域猛然升高，那么很有可能需要进行PCR去重复</span></p></li></ul></li></ul><p>&nbsp;</p><p><strong><span>Key point</span></strong></p><ul><li><p><span>起始量很多时，不需要去重复</span></p></li><li><p><span>扩增数很少时，15个cycle以内，不需要去重复</span></p></li><li><p><span>双端测序由于其两个reads的位置矫正，有助于去除PCR重复</span></p></li><li><p><span>reads长度越长，越容易识别真正的PCR重复</span></p></li></ul><p>&nbsp;</p><p><strong><span>Theory</span></strong></p><p><span>测序所得到的reads是由于超声波或者酶切断裂得到的，因此这些reads比对到基因组上的位置是</span><strong><span>完全随机的</span></strong><span>。那么两个reads</span><strong><span>比对到相同位置</span></strong><span>的概率是非常低的。如果两个reads比对情况相同或者极其相似，则很有可能是由于PCR重复所导致的</span></p><p>&nbsp;</p><hr /><h2 id='patterned-flowcell'><span>Patterned Flowcell?</span></h2><h3 id='传统未图案化流动池non-patterned-flowcell'><span>传统未图案化流动池(Non-Patterned Flowcell)</span></h3><ul><li><p><span>在</span><strong><span>Non-Patterned Flowcell</span></strong><span>中，DNA 片段随机地附着在流动池的表面上，并形成测序cluster。这种随机分布可能导致一些区域的cluster密度过高，而其他区域的cluster密度过低，从而影响测序的均匀性和数据产出</span></p></li><li><p><span>在</span><strong><span>Non-Patterned Flowcell</span></strong><span>中，每个 lane 通常被分为</span><strong><span>较少数量的 tiles</span></strong><span>，tile 数量有限，编号也较小</span></p><p>&nbsp;</p></li></ul><h3 id='图案化流动池patterned-flowcell'><span>图案化流动池(Patterned Flowcell)</span></h3><ul><li><p><strong><span>Patterned Flowcell</span></strong><span>的表面覆盖着一层规则排列的纳米井（nanowells），这些nanowells的作用是控制 DNA 簇的生长位置。每个nanowells只能容纳一个测序cluster，因此cluster的分布更加均匀。</span></p></li><li><p><span>由于每个nanowells都可以独立形成一个测序cluster，这种设计大大增加了流动池上的有效cluster数，显著提高了测序的通量。</span></p></li><li><p><strong><span>Patterned Flowcell</span></strong><span>的设计减少了cluster之间的相互干扰（cluster间竞争和cluster重叠），从而提高了数据的准确性和质量。</span></p></li></ul><table>
  <tr>
    <td><img src="./picture/040.png" style="width:650px;" /></td>
    <td><img src="./picture/041.png" style="width:650px;" /></td>
  </tr>
</table><p>&nbsp;</p><h3 id='patterned-flowcell与光学重复检测的关系'><span>Patterned Flowcell与光学重复检测的关系</span></h3><p><span>由于</span><strong><span>Patterned Flowcell</span></strong><span>中的测序cluster位置是预先确定的，因此光学重复（optical duplicates）在这种设计下变得更容易识别和控制。</span><code>Optical Duplicate Pixel Distance</code><span> 的默认值在图案化流动池中通常设置得更高（如 2500），以适应这种新型设计</span></p><p>&nbsp;</p><h3 id='如何查看自己数据是否使用patterned-flowcell'><span>如何查看自己数据是否使用Patterned Flowcell？</span></h3><p><strong><span>查看FASTQ文件中@行信息来推断</span></strong><span> </span><code>zcat &lt;xxx.fastq.gz&gt; | grep &quot;^@&quot; | head -n 5</code></p><ul><li><p><strong><span>Patterned Flowcell</span></strong><span> 通常用于 HiSeq X、NovaSeq 系列、以及部分 MiSeq 和 NextSeq 仪器</span></p></li><li><p><span>Tile 编号很大（例如 1101，1201），并且X坐标值也很大，那么有可能使用的是 </span><strong><span>Patterned Flowcell</span></strong><span>。</span></p><p><img src="./picture/042.png"></p></li></ul><p>&nbsp;</p><hr /><h2 id='software-selection-of-remove-duplicates'><span>Software selection of remove duplicates</span></h2><h3 id='samtools-rmdup'><span>samtools rmdup</span></h3><ul><li><p><span>如果多个reads具有相同的比对位置时，</span><code>samtools rmdup</code><span>将它们标记为duplicates，然后去除重复，通常只保留第一个识别到的reads。</span></p></li><li><p><span>该方法对于以下两种情况，有很好的去除效果：</span></p><ul><li><p><span>一些来源于相同 DNA 分子的 reads 由于测序错误可能在序列上存在差异，但它们的比对位置一致</span></p></li><li><p><span>比对错误导致不同的序列比对到相同的位置（可能性不大）</span></p></li></ul></li><li><p><strong><span>Shortcoming</span></strong><span>：由于</span><code>samtools rmdup</code><span>去重只考虑reads比对上的起始终止位置，不考虑比对情况，这种去重有时会导致测序信息的丢失。</span></p></li></ul><p>&nbsp;</p><h3 id='picard-markduplicates'><span>Picard MarkDuplicates</span></h3><ul><li><p><span>与</span><code>samtools rmdup</code><span>不同的是，</span><code>Picard MarkDuplicates</code><span>仅仅是对duplicates做一个标记，只在需要的时候对reads进行去重。而</span><code>samtools rmdup</code><span>则是直接将其识别出来的重复reads去掉。</span></p></li><li><p><code>Picard MarkDuplicates</code><span>另一个不同之处在于它不仅考虑reads的比对位置，还会考虑其中的插入错配等情况（即会利用sam/bam文件中的CIGAR值），甚至reads的tail，lane以及flowcell。</span></p></li><li><p><code>Picard MarkDuplicates</code><span>主要考虑reads的5&#39;端的比对位置，其可以从一定程度上认为，5&#39;端的位置，方向，以及碱基比对情况相同，</span><code>Picard MarkDuplicates</code><span>就将这些reads中碱基</span><strong><span>比对值Q&gt;15</span></strong><span>的看作是</span><strong><span>best pair</span></strong><span>而其他的reads则当作是duplicate reads。甚至当reads的长度不同时，</span><code>Picard MarkDuplicates</code><span>依然利用上述原理进行去重。</span></p><div class="md-alert md-alert-tip tip"><p><span class='md-alert-text md-alert-text-tip'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M8 1.5c-2.363 0-4 1.69-4 3.75 0 .984.424 1.625.984 2.304l.214.253c.223.264.47.556.673.848.284.411.537.896.621 1.49a.75.75 0 0 1-1.484.211c-.04-.282-.163-.547-.37-.847a8.456 8.456 0 0 0-.542-.68c-.084-.1-.173-.205-.268-.32C3.201 7.75 2.5 6.766 2.5 5.25 2.5 2.31 4.863 0 8 0s5.5 2.31 5.5 5.25c0 1.516-.701 2.5-1.328 3.259-.095.115-.184.22-.268.319-.207.245-.383.453-.541.681-.208.3-.33.565-.37.847a.751.751 0 0 1-1.485-.212c.084-.593.337-1.078.621-1.489.203-.292.45-.584.673-.848.075-.088.147-.173.213-.253.561-.679.985-1.32.985-2.304 0-2.06-1.637-3.75-4-3.75ZM5.75 12h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5ZM6 15.25a.75.75 0 0 1 .75-.75h2.5a.75.75 0 0 1 0 1.5h-2.5a.75.75 0 0 1-.75-.75Z"></path></svg>Tip</span><br></p><p><code>Picard MarkDuplicates</code><span>中，reads的5&#39;端信息更为重要.若duplicates是PCR重复，那么它们的序列不一定完全相同。但是由于PCR扩增时，酶的前进方向是5&#39;——3&#39;方向，PCR重复序列中5&#39;端的部分相似的可能性更高。</span></p></div></li></ul><p>&nbsp;</p><hr /><h2 id='picard-markduplicates-2'><span>Picard MarkDuplicates</span></h2><ul><li><p><a href='https://gatk.broadinstitute.org/hc/en-us/articles/360037052812-MarkDuplicates-Picard'><span>Picard MarkDuplicates</span></a></p></li><li><p><code>Picard MarkDuplicates</code><span> ,它会对 </span><strong><span>reads 做一种临时排序（sort）</span></strong><span>，以便找出哪些是重复的（duplicates）</span></p></li><li><p><strong><span>Software installation</span></strong><span> </span><code>conda install -c bioconda picard</code></p></li><li><p><strong><span>需要准备按基因组位置（coordinate）进行排序的Bam file</span></strong></p></li></ul><p>&nbsp;</p><h3 id='picard-markduplicates-parameters'><span>Picard MarkDuplicates parameters</span></h3><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">picard MarkDuplicates</span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==基本设置==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-I</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 输入一个或多个（排序后）SAM 或 BAM 文件</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-O</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 输出文件（去除重复后的 BAM 文件）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-M</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 输出的度量文件（报告重复情况）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--REMOVE_DUPLICATES</span> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 如果设置为 true，会完全移除重复的 reads(dafult:false)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">#  ==其他输入设置==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--arguments_file</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 可指定一个或多个参数文件，代替手动在命令行写参数，文件格式见下 </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--REFERENCE_SEQUENCE</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 提供参考基因组序列的路径(deafult:null)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--ASSUME_SORT_ORDER</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 如果不是null,则强制程序假设输入文件是指定的排序方式，而忽略 header中的排序声明 </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># optional parameters:unsorted、queryname、coordinate、duplicate、unknown 或 null(deafult)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--GA4GH_CLIENT_SECRETS</span> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 提供 GA4GH 客户端密钥的 JSON 文件路径，用于与 Google Genomics API 交互。 (deafult:client_secrets.json)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--VALIDATION_STRINGENCY</span> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 控制当程序在读取 BAM/SAM 文件时遇到格式错误或不规范内容时的行为。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># STRICT(deafult):一旦发现格式错误，立即停止运行并报错。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># LENIENT:报出警告(warning)，但不会终止程序。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># SILENT:完全忽略格式错误，不报错也不报警。(适合提高性能)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==去重策略== </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; <span class="cm-attribute">--DUPLICATE_SCORING_STRATEGY</span> &nbsp; <span class="cm-comment"># 判断哪个 read 被保留用的非重复的策略 </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># optional parameters：# SUM_OF_BASE_QUALITIES(deafult)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># TOTAL_MAPPED_REFERENCE_LENGTH(read在参考基因组上比对的总长度)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># RANDOM </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==光学重复检测选项==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--REMOVE_SEQUENCING_DUPLICATES</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 如果设置为 true，则删除光学重复。deafult: falue</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 如果设为 true，即使 REMOVE_DUPLICATES 为 false，也会去除这类重复 </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--READ_NAME_REGE</span> <span class="cm-string">'&lt;your regex&gt;'</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 在一些定制化平台或非标准 read 名字格式中，使用自定义 regex 解析 reads 名称，从而提取出 tile 坐标、x、y 坐标，用于估算光学重复率。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 如果不想做光学重复估算，可以将其设为 null</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--OPTICAL_DUPLICATE_PIXEL_DISTANCE</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 选项定义了2个clusters之间的最大偏移量（以像素为单位），超过此距离的cluster将不被视为光学重复。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 对于未设计过的 Non-Patterned Flowcell（例如早期版本的平台），默认值 100 是适用的。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 新的设计过的 Patterned Flowcell，光学重复的检测要求更高，因此建议将此值设置为 2500</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--MAX_OPTICAL_DUPLICATE_SET_SIZE</span><span class="cm-operator">=</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 用于确定最大光学重复集大小的阈值。如果一组重复 reads 超过这个数目，程序会尝试识别哪些是光学重复。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 请注意，如果设置值过高且确实遇到非常大的重复read集，它将严重影响工具的运行时间。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 要完全禁用此检查，请将值设置为 -1。(deafult:300000)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> <span class="cm-comment"># ==Header(HD)标签== &nbsp; &nbsp; </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; <span class="cm-attribute">--COMMENT</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 添加注释（comment）到 BAM HD标签中，</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 可以添加多个注释，只需要多次使用 --COMMENT</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># example:--COMMENT "This is a comment about the experiment."</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==duplicate type (DT)标签==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--CLEAR_DT</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment">#从输入的 SAM 记录中清除 DT 标签。deafult: true</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--TAGGING_POLICY</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 决定如何在 DT 可选属性中记录重复类型。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># optional parameters：DontTag（不标记,deafult）、OpticalOnly（仅标记光学重复）、All（标记所有重复）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--TAG_DUPLICATE_SET_MEMBERS</span> &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 如果read属于重复集，则添加两个标签 （deafult: false）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 第一个标签 DS 表示重复集的大小。最小的 DS 值为 2，表示两个read映射到参考基因组的相同位置，其中一个被标记为重复。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 第二个标签 DI 表示重复集的唯一标识符。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==Barcode 或 UMI==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># SAM/BAM文件中 BC(Barcode)或 10X Genomics 的 BX(Cell Barcode + UMI) 标签包含用于标识不同样本或分子库的 barcode 信息。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># barcode通常会在测序过程中被添加到每个 DNA 片段的两端，允许区分不同样本或分子。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--BARCODE_TAG</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 指定用于区分 barcode 的标签(deafult:null)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--READ_ONE_BARCODE_TAG</span> &nbsp; &nbsp; <span class="cm-comment"># 为第一条read指定barcode 的 标签(deafult:null)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--READ_TWO_BARCODE_TAG</span> &nbsp; &nbsp; <span class="cm-comment"># 为第二条read指定barcode 的 标签(deafult:null)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 双链 UMI 通常由两个相等长度的字符串组成，这两个字符串由一个连字符（例如 "ATC-GTC"）分隔</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 双链 UMI 通常储存在 RX (Read-specific UMI) 或 MI (Molecular Identifier)标签中</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--MOLECULAR_IDENTIFIER_TAG</span> <span class="cm-comment"># 指定用于区分 UMI 的标签(deafult:null)。使用此选项时，必须将 BARCODE_TAG 选项设置为非空值,deafult:null</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--DUPLEX_UMI</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 双链 UMI 通常由两个相等长度的字符串组成，这两个字符串由一个连字符（例如 "ATC-GTC"）分隔</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 双链 UMI 通常储存在 RX (Read-specific UMI) 或 MI (Molecular Identifier)标签中</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 启用后，工具会自动识别 RX或 MI 标签，根据双链 UMI 来区分重复reads。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==Program(PG)标签==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" class="cm-tab-wrap-hack" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--PROGRAM_GROUP_COMMAND_LINE</span> &nbsp;<span class="cm-comment">#<span class="cm-tab" role="presentation" cm-text="	"> </span>生成 PG 标签时，使用的命令行参数。(deafult:程序会自动生成这个值)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--PROGRAM_GROUP_NAME</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 生成 PG 标签中PN(program name)(deafult:MarkDuplicates)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--PROGRAM_RECORD_ID</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 生成 PG 标签中ID(program ID)，如果设置为 null 则不会生成该标签(deafult:MarkDuplicates)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--PROGRAM_GROUP_VERSION</span> &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 生成 PG 标签中VN(verson)。如果不指定，程序会自动检测当前版本</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--ADD_PG_TAG_TO_READS</span> &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 是否为每个 read 添加 PG 标签(deafult:true)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==输出压缩设置==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--COMPRESSION_LEVEL</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 输出压缩文件（例如 BAM 文件）的压缩级别，0 是最小压缩，9 是最大压缩(deafult:5)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--USE_JDK_DEFLATER</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 是否使用  Java 内置(JDK) 提供的 Deflater 压缩工具，而非默认的 Intel Deflater。(deafult:false)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--USE_JDK_INFLATER</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 是否使用 Java 内置(JDK) 提供的 Inflater 解压工具，而非默认的 Intel Inflater。(deafult:false)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==日志输出设置==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--QUIET</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 抑制日志中大部分标准输出，仅保留错误信息。 (deafult:false)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--VERBOSITY</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 设置日志输出的详细程度，</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># optional parameters:INFO(deafult)、 DEBUG 或 ERROR</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==其他输出设置== &nbsp; &nbsp;</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--CREATE_INDEX</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 如果输出是 BAM 文件，是否自动为其生成 .bai 索引文件。 (deafult:false)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--CREATE_MD5_FILE</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 是否生成 .MD5 校验文件。 (deafult:false)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--TMP_DIR</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 临时存储工作的目录 &nbsp; &nbsp;</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==性能优化== </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--SORTING_COLLECTION_SIZE_RATIO</span> &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 控制在排序过程中，Picard 使用内存的比例来缓存记录，超过这个比例后就会写入临时文件进行磁盘排序。(deafult:0.25) </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 设置太低，速度慢</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 设置太高，可能会 导致内存溢出（OutOfMemoryError）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--MAX_FILE_HANDLES_FOR_READ_ENDS_MAP</span> &nbsp;<span class="cm-comment"># 允许同时打开的文件句柄(每个read的数据块可能需要一个文件句柄来存储)数量 (deafult:8000)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 当内存不足，遇到“too many open files”错误，可以适当调小 &nbsp;</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--MAX_RECORDS_IN_RAM</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 控制写入文件时，RAM 中能存储的最大记录数。超出该数目会将数据写入磁盘。 (deafult:500000)</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 2994px;"></div><div class="CodeMirror-gutters" style="display: none; height: 2994px;"></div></div></div></pre><p>&nbsp;</p><div class="md-alert md-alert-tip tip"><p><span class='md-alert-text md-alert-text-tip'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M8 1.5c-2.363 0-4 1.69-4 3.75 0 .984.424 1.625.984 2.304l.214.253c.223.264.47.556.673.848.284.411.537.896.621 1.49a.75.75 0 0 1-1.484.211c-.04-.282-.163-.547-.37-.847a8.456 8.456 0 0 0-.542-.68c-.084-.1-.173-.205-.268-.32C3.201 7.75 2.5 6.766 2.5 5.25 2.5 2.31 4.863 0 8 0s5.5 2.31 5.5 5.25c0 1.516-.701 2.5-1.328 3.259-.095.115-.184.22-.268.319-.207.245-.383.453-.541.681-.208.3-.33.565-.37.847a.751.751 0 0 1-1.485-.212c.084-.593.337-1.078.621-1.489.203-.292.45-.584.673-.848.075-.088.147-.173.213-.253.561-.679.985-1.32.985-2.304 0-2.06-1.637-3.75-4-3.75ZM5.75 12h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5ZM6 15.25a.75.75 0 0 1 .75-.75h2.5a.75.75 0 0 1 0 1.5h-2.5a.75.75 0 0 1-.75-.75Z"></path></svg>Tip</span><br></p><p><code>--arguments_file</code><span>可指定一个或多个参数文件，代替手动在命令行写参数，其文件格式如下</span></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang=""><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang=""><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">-I sorted_input.bam</span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">-O deduplicated_output.bam</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">-M output.metrics.txt</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">--REMOVE_DUPLICATES true</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">--OPTICAL_DUPLICATE_PIXEL_DISTANCE 2500</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 109px;"></div><div class="CodeMirror-gutters" style="display: none; height: 109px;"></div></div></div></pre></div><p>&nbsp;</p><hr /><h3 id='remove-duplicates-strategy'><span>Remove duplicates strategy</span></h3><p><strong><span>策略:先查看重复数，再考虑是否去除重复</span></strong></p><ol start='' ><li><p><strong><span>Check the duplicates</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">picard MarkDuplicates \</span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-I</span> sorted_input.bam \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-O</span> deduplicated_output.bam \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-M</span> output.metrics.txt \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--REMOVE_DUPLICATES</span> <span class="cm-atom">false</span> \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 只标记不去除重复 &nbsp; </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--OPTICAL_DUPLICATE_PIXEL_DISTANCE</span> <span class="cm-number">2500</span> &nbsp; &nbsp;</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 131px;"></div><div class="CodeMirror-gutters" style="display: none; height: 131px;"></div></div></div></pre><p>&nbsp;</p></li><li><p><strong><span>Check the </span><code>output.metrics</code></strong></p><p><code>Picard MarkDuplicates</code><span>会生成一个</span><code>output.metrics.txt</code><span>，来查看我们的重复情况</span></p><figure class='table-figure'><table><thead><tr><th><span>LIBRARY</span></th><th><span>UNPAIRED_READS_EXAMINED</span></th><th><span>READ_PAIRS_EXAMINED</span></th><th><span>SECONDARY_OR_SUPPLEMENTARY_RDS</span></th><th><span>UNMAPPED_READS</span></th><th><span>UNPAIRED_READ_DUPLICATES</span></th><th><span>READ_PAIR_DUPLICATES</span></th><th><span>READ_PAIR_OPTICAL_DUPLICATES</span></th><th><span>PERCENT_DUPLICATION</span></th><th><span>ESTIMATED_LIBRARY_SIZE</span></th></tr></thead><tbody><tr><td><span>Unknown Library</span></td><td><span>0</span></td><td><span>64747726</span></td><td><span>0</span></td><td><span>4910198</span></td><td><span>0</span></td><td><span>62515854</span></td><td><span>2919405</span></td><td><span>0.96553</span></td><td><span>2231872</span></td></tr></tbody></table></figure><ul><li><p><strong><span>LIBRARY</span></strong><span>：文库的名称。</span></p></li><li><p><strong><span>UNPAIRED_READS_EXAMINED</span></strong><span>：检查的单端（unpaired）reads数量。在这个例子中，这个值是 0，表示所有reads都是paired-end reads。</span></p></li><li><p><strong><span>READ_PAIRS_EXAMINED</span></strong><span>：检查paired-end reads的数量。</span></p></li><li><p><strong><span>SECONDARY_OR_SUPPLEMENTARY_RDS</span></strong><span>：二次或补充（secondary or supplementary）reads的数量。这些reads通常不用于重复率计算。</span></p></li></ul><p>&nbsp;</p><div class="md-alert md-alert-tip tip"><p><span class='md-alert-text md-alert-text-tip'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M8 1.5c-2.363 0-4 1.69-4 3.75 0 .984.424 1.625.984 2.304l.214.253c.223.264.47.556.673.848.284.411.537.896.621 1.49a.75.75 0 0 1-1.484.211c-.04-.282-.163-.547-.37-.847a8.456 8.456 0 0 0-.542-.68c-.084-.1-.173-.205-.268-.32C3.201 7.75 2.5 6.766 2.5 5.25 2.5 2.31 4.863 0 8 0s5.5 2.31 5.5 5.25c0 1.516-.701 2.5-1.328 3.259-.095.115-.184.22-.268.319-.207.245-.383.453-.541.681-.208.3-.33.565-.37.847a.751.751 0 0 1-1.485-.212c.084-.593.337-1.078.621-1.489.203-.292.45-.584.673-.848.075-.088.147-.173.213-.253.561-.679.985-1.32.985-2.304 0-2.06-1.637-3.75-4-3.75ZM5.75 12h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5ZM6 15.25a.75.75 0 0 1 .75-.75h2.5a.75.75 0 0 1 0 1.5h-2.5a.75.75 0 0 1-.75-.75Z"></path></svg>Tip</span><br></p><ul><li><p><strong><span>Secondary Reads</span></strong><span>：是指在比对过程中，一个reads可能比对到多个位置（即存在多重比对）。在这种情况下，比对工具会选择一个主要的比对位置（通常是得分最高的那个），并将其他位置标记为次要比对（</span><strong><span>secondary alignment</span></strong><span>）。这些次要比对位置的reads就被称为 </span><strong><span>Secondary Reads</span></strong></p></li><li><p><strong><span>Supplementary Reads</span></strong><span>：是指在比对过程中，一个长的reads（或 read pair）由于结构变异或其他复杂原因，比对到多个不连续的位置。为了完整描述这种情况，原始的读数会被拆分成几段，每段的比对结果都保留，并被标记为 </span><strong><span>Supplementary Reads</span></strong><span>。</span></p></li></ul></div><ul><li><p><strong><span>UNMAPPED_READS</span></strong><span>：未比对的reads数量。</span></p></li><li><p><strong><span>UNPAIRED_READ_DUPLICATES</span></strong><span>：单端（unpaired）reads中的重复读数数量。</span></p></li><li><p><strong><span>READ_PAIR_DUPLICATES</span></strong><span>：paired-end reads中的重复reads数量。</span></p></li><li><p><strong><span>READ_PAIR_OPTICAL_DUPLICATES</span></strong><span>：由于光学重复配对reads数量。</span></p></li><li><p><strong><span>PERCENT_DUPLICATION</span></strong><span>：重复reads所占的百分比。</span></p></li><li><p><strong><span>ESTIMATED_LIBRARY_SIZE</span></strong><span>：估计的文库大小，即根据观察到的重复率估算的原始分子数目。</span></p><div class="md-alert md-alert-tip tip"><p><span class='md-alert-text md-alert-text-tip'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M8 1.5c-2.363 0-4 1.69-4 3.75 0 .984.424 1.625.984 2.304l.214.253c.223.264.47.556.673.848.284.411.537.896.621 1.49a.75.75 0 0 1-1.484.211c-.04-.282-.163-.547-.37-.847a8.456 8.456 0 0 0-.542-.68c-.084-.1-.173-.205-.268-.32C3.201 7.75 2.5 6.766 2.5 5.25 2.5 2.31 4.863 0 8 0s5.5 2.31 5.5 5.25c0 1.516-.701 2.5-1.328 3.259-.095.115-.184.22-.268.319-.207.245-.383.453-.541.681-.208.3-.33.565-.37.847a.751.751 0 0 1-1.485-.212c.084-.593.337-1.078.621-1.489.203-.292.45-.584.673-.848.075-.088.147-.173.213-.253.561-.679.985-1.32.985-2.304 0-2.06-1.637-3.75-4-3.75ZM5.75 12h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5ZM6 15.25a.75.75 0 0 1 .75-.75h2.5a.75.75 0 0 1 0 1.5h-2.5a.75.75 0 0 1-.75-.75Z"></path></svg>Tip</span><br></p><p><strong><span>Estimated library size ?</span></strong></p><p><span>Estimated Library Size 是根据重复读数的比例估算的原始分子数目。常见的估算公式如下：</span></p><div contenteditable="false" spellcheck="false" class="mathjax-block md-end-block md-math-block md-rawblock" id="mathjax-n2488" cid="n2488" mdtype="math_block" data-math-tag-before="0" data-math-tag-after="0" data-math-labels="[]"><div class="md-math-container"><mjx-container class="MathJax" jax="SVG" display="true" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="35.874ex" height="5.326ex" role="img" focusable="false" viewBox="0 -1381 15856.4 2354.2" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -2.202ex;"><defs><path id="MJX-4-TEX-N-45" d="M128 619Q121 626 117 628T101 631T58 634H25V680H597V676Q599 670 611 560T625 444V440H585V444Q584 447 582 465Q578 500 570 526T553 571T528 601T498 619T457 629T411 633T353 634Q266 634 251 633T233 622Q233 622 233 621Q232 619 232 497V376H286Q359 378 377 385Q413 401 416 469Q416 471 416 473V493H456V213H416V233Q415 268 408 288T383 317T349 328T297 330Q290 330 286 330H232V196V114Q232 57 237 52Q243 47 289 47H340H391Q428 47 452 50T505 62T552 92T584 146Q594 172 599 200T607 247T612 270V273H652V270Q651 267 632 137T610 3V0H25V46H58Q100 47 109 49T128 61V619Z"></path><path id="MJX-4-TEX-N-73" d="M295 316Q295 356 268 385T190 414Q154 414 128 401Q98 382 98 349Q97 344 98 336T114 312T157 287Q175 282 201 278T245 269T277 256Q294 248 310 236T342 195T359 133Q359 71 321 31T198 -10H190Q138 -10 94 26L86 19L77 10Q71 4 65 -1L54 -11H46H42Q39 -11 33 -5V74V132Q33 153 35 157T45 162H54Q66 162 70 158T75 146T82 119T101 77Q136 26 198 26Q295 26 295 104Q295 133 277 151Q257 175 194 187T111 210Q75 227 54 256T33 318Q33 357 50 384T93 424T143 442T187 447H198Q238 447 268 432L283 424L292 431Q302 440 314 448H322H326Q329 448 335 442V310L329 304H301Q295 310 295 316Z"></path><path id="MJX-4-TEX-N-74" d="M27 422Q80 426 109 478T141 600V615H181V431H316V385H181V241Q182 116 182 100T189 68Q203 29 238 29Q282 29 292 100Q293 108 293 146V181H333V146V134Q333 57 291 17Q264 -10 221 -10Q187 -10 162 2T124 33T105 68T98 100Q97 107 97 248V385H18V422H27Z"></path><path id="MJX-4-TEX-N-69" d="M69 609Q69 637 87 653T131 669Q154 667 171 652T188 609Q188 579 171 564T129 549Q104 549 87 564T69 609ZM247 0Q232 3 143 3Q132 3 106 3T56 1L34 0H26V46H42Q70 46 91 49Q100 53 102 60T104 102V205V293Q104 345 102 359T88 378Q74 385 41 385H30V408Q30 431 32 431L42 432Q52 433 70 434T106 436Q123 437 142 438T171 441T182 442H185V62Q190 52 197 50T232 46H255V0H247Z"></path><path id="MJX-4-TEX-N-6D" d="M41 46H55Q94 46 102 60V68Q102 77 102 91T102 122T103 161T103 203Q103 234 103 269T102 328V351Q99 370 88 376T43 385H25V408Q25 431 27 431L37 432Q47 433 65 434T102 436Q119 437 138 438T167 441T178 442H181V402Q181 364 182 364T187 369T199 384T218 402T247 421T285 437Q305 442 336 442Q351 442 364 440T387 434T406 426T421 417T432 406T441 395T448 384T452 374T455 366L457 361L460 365Q463 369 466 373T475 384T488 397T503 410T523 422T546 432T572 439T603 442Q729 442 740 329Q741 322 741 190V104Q741 66 743 59T754 49Q775 46 803 46H819V0H811L788 1Q764 2 737 2T699 3Q596 3 587 0H579V46H595Q656 46 656 62Q657 64 657 200Q656 335 655 343Q649 371 635 385T611 402T585 404Q540 404 506 370Q479 343 472 315T464 232V168V108Q464 78 465 68T468 55T477 49Q498 46 526 46H542V0H534L510 1Q487 2 460 2T422 3Q319 3 310 0H302V46H318Q379 46 379 62Q380 64 380 200Q379 335 378 343Q372 371 358 385T334 402T308 404Q263 404 229 370Q202 343 195 315T187 232V168V108Q187 78 188 68T191 55T200 49Q221 46 249 46H265V0H257L234 1Q210 2 183 2T145 3Q42 3 33 0H25V46H41Z"></path><path id="MJX-4-TEX-N-61" d="M137 305T115 305T78 320T63 359Q63 394 97 421T218 448Q291 448 336 416T396 340Q401 326 401 309T402 194V124Q402 76 407 58T428 40Q443 40 448 56T453 109V145H493V106Q492 66 490 59Q481 29 455 12T400 -6T353 12T329 54V58L327 55Q325 52 322 49T314 40T302 29T287 17T269 6T247 -2T221 -8T190 -11Q130 -11 82 20T34 107Q34 128 41 147T68 188T116 225T194 253T304 268H318V290Q318 324 312 340Q290 411 215 411Q197 411 181 410T156 406T148 403Q170 388 170 359Q170 334 154 320ZM126 106Q126 75 150 51T209 26Q247 26 276 49T315 109Q317 116 318 175Q318 233 317 233Q309 233 296 232T251 223T193 203T147 166T126 106Z"></path><path id="MJX-4-TEX-N-65" d="M28 218Q28 273 48 318T98 391T163 433T229 448Q282 448 320 430T378 380T406 316T415 245Q415 238 408 231H126V216Q126 68 226 36Q246 30 270 30Q312 30 342 62Q359 79 369 104L379 128Q382 131 395 131H398Q415 131 415 121Q415 117 412 108Q393 53 349 21T250 -11Q155 -11 92 58T28 218ZM333 275Q322 403 238 411H236Q228 411 220 410T195 402T166 381T143 340T127 274V267H333V275Z"></path><path id="MJX-4-TEX-N-64" d="M376 495Q376 511 376 535T377 568Q377 613 367 624T316 637H298V660Q298 683 300 683L310 684Q320 685 339 686T376 688Q393 689 413 690T443 693T454 694H457V390Q457 84 458 81Q461 61 472 55T517 46H535V0Q533 0 459 -5T380 -11H373V44L365 37Q307 -11 235 -11Q158 -11 96 50T34 215Q34 315 97 378T244 442Q319 442 376 393V495ZM373 342Q328 405 260 405Q211 405 173 369Q146 341 139 305T131 211Q131 155 138 120T173 59Q203 26 251 26Q322 26 373 103V342Z"></path><path id="MJX-4-TEX-N-20" d=""></path><path id="MJX-4-TEX-N-4C" d="M128 622Q121 629 117 631T101 634T58 637H25V683H36Q48 680 182 680Q324 680 348 683H360V637H333Q273 637 258 635T233 622L232 342V129Q232 57 237 52Q243 47 313 47Q384 47 410 53Q470 70 498 110T536 221Q536 226 537 238T540 261T542 272T562 273H582V268Q580 265 568 137T554 5V0H25V46H58Q100 47 109 49T128 61V622Z"></path><path id="MJX-4-TEX-N-62" d="M307 -11Q234 -11 168 55L158 37Q156 34 153 28T147 17T143 10L138 1L118 0H98V298Q98 599 97 603Q94 622 83 628T38 637H20V660Q20 683 22 683L32 684Q42 685 61 686T98 688Q115 689 135 690T165 693T176 694H179V543Q179 391 180 391L183 394Q186 397 192 401T207 411T228 421T254 431T286 439T323 442Q401 442 461 379T522 216Q522 115 458 52T307 -11ZM182 98Q182 97 187 90T196 79T206 67T218 55T233 44T250 35T271 29T295 26Q330 26 363 46T412 113Q424 148 424 212Q424 287 412 323Q385 405 300 405Q270 405 239 390T188 347L182 339V98Z"></path><path id="MJX-4-TEX-N-72" d="M36 46H50Q89 46 97 60V68Q97 77 97 91T98 122T98 161T98 203Q98 234 98 269T98 328L97 351Q94 370 83 376T38 385H20V408Q20 431 22 431L32 432Q42 433 60 434T96 436Q112 437 131 438T160 441T171 442H174V373Q213 441 271 441H277Q322 441 343 419T364 373Q364 352 351 337T313 322Q288 322 276 338T263 372Q263 381 265 388T270 400T273 405Q271 407 250 401Q234 393 226 386Q179 341 179 207V154Q179 141 179 127T179 101T180 81T180 66V61Q181 59 183 57T188 54T193 51T200 49T207 48T216 47T225 47T235 46T245 46H276V0H267Q249 3 140 3Q37 3 28 0H20V46H36Z"></path><path id="MJX-4-TEX-N-79" d="M69 -66Q91 -66 104 -80T118 -116Q118 -134 109 -145T91 -160Q84 -163 97 -166Q104 -168 111 -168Q131 -168 148 -159T175 -138T197 -106T213 -75T225 -43L242 0L170 183Q150 233 125 297Q101 358 96 368T80 381Q79 382 78 382Q66 385 34 385H19V431H26L46 430Q65 430 88 429T122 428Q129 428 142 428T171 429T200 430T224 430L233 431H241V385H232Q183 385 185 366L286 112Q286 113 332 227L376 341V350Q376 365 366 373T348 383T334 385H331V431H337H344Q351 431 361 431T382 430T405 429T422 429Q477 429 503 431H508V385H497Q441 380 422 345Q420 343 378 235T289 9T227 -131Q180 -204 113 -204Q69 -204 44 -177T19 -116Q19 -89 35 -78T69 -66Z"></path><path id="MJX-4-TEX-N-53" d="M55 507Q55 590 112 647T243 704H257Q342 704 405 641L426 672Q431 679 436 687T446 700L449 704Q450 704 453 704T459 705H463Q466 705 472 699V462L466 456H448Q437 456 435 459T430 479Q413 605 329 646Q292 662 254 662Q201 662 168 626T135 542Q135 508 152 480T200 435Q210 431 286 412T370 389Q427 367 463 314T500 191Q500 110 448 45T301 -21Q245 -21 201 -4T140 27L122 41Q118 36 107 21T87 -7T78 -21Q76 -22 68 -22H64Q61 -22 55 -16V101Q55 220 56 222Q58 227 76 227H89Q95 221 95 214Q95 182 105 151T139 90T205 42T305 24Q352 24 386 62T420 155Q420 198 398 233T340 281Q284 295 266 300Q261 301 239 306T206 314T174 325T141 343T112 367T85 402Q55 451 55 507Z"></path><path id="MJX-4-TEX-N-7A" d="M42 263Q44 270 48 345T53 423V431H393Q399 425 399 415Q399 403 398 402L381 378Q364 355 331 309T265 220L134 41L182 40H206Q254 40 283 46T331 77Q352 105 359 185L361 201Q361 202 381 202H401V196Q401 195 393 103T384 6V0H209L34 1L31 3Q28 8 28 17Q28 30 29 31T160 210T294 394H236Q169 393 152 388Q127 382 113 367Q89 344 82 264V255H42V263Z"></path><path id="MJX-4-TEX-N-3D" d="M56 347Q56 360 70 367H707Q722 359 722 347Q722 336 708 328L390 327H72Q56 332 56 347ZM56 153Q56 168 72 173H708Q722 163 722 153Q722 140 707 133H70Q56 140 56 153Z"></path><path id="MJX-4-TEX-I-1D436" d="M50 252Q50 367 117 473T286 641T490 704Q580 704 633 653Q642 643 648 636T656 626L657 623Q660 623 684 649Q691 655 699 663T715 679T725 690L740 705H746Q760 705 760 698Q760 694 728 561Q692 422 692 421Q690 416 687 415T669 413H653Q647 419 647 422Q647 423 648 429T650 449T651 481Q651 552 619 605T510 659Q484 659 454 652T382 628T299 572T226 479Q194 422 175 346T156 222Q156 108 232 58Q280 24 350 24Q441 24 512 92T606 240Q610 253 612 255T628 257Q648 257 648 248Q648 243 647 239Q618 132 523 55T319 -22Q206 -22 128 53T50 252Z"></path><path id="MJX-4-TEX-N-31" d="M213 578L200 573Q186 568 160 563T102 556H83V602H102Q149 604 189 617T245 641T273 663Q275 666 285 666Q294 666 302 660V361L303 61Q310 54 315 52T339 48T401 46H427V0H416Q395 3 257 3Q121 3 100 0H88V46H114Q136 46 152 46T177 47T193 50T201 52T207 57T213 61V578Z"></path><path id="MJX-4-TEX-N-D7" d="M630 29Q630 9 609 9Q604 9 587 25T493 118L389 222L284 117Q178 13 175 11Q171 9 168 9Q160 9 154 15T147 29Q147 36 161 51T255 146L359 250L255 354Q174 435 161 449T147 471Q147 480 153 485T168 490Q173 490 175 489Q178 487 284 383L389 278L493 382Q570 459 587 475T609 491Q630 491 630 471Q630 464 620 453T522 355L418 250L522 145Q606 61 618 48T630 29Z"></path><path id="MJX-4-TEX-N-30" d="M96 585Q152 666 249 666Q297 666 345 640T423 548Q460 465 460 320Q460 165 417 83Q397 41 362 16T301 -15T250 -22Q224 -22 198 -16T137 16T82 83Q39 165 39 320Q39 494 96 585ZM321 597Q291 629 250 629Q208 629 178 597Q153 571 145 525T137 333Q137 175 145 125T181 46Q209 16 250 16Q290 16 318 46Q347 76 354 130T362 333Q362 478 354 524T321 597Z"></path><path id="MJX-4-TEX-N-2212" d="M84 237T84 250T98 270H679Q694 262 694 250T679 230H98Q84 237 84 250Z"></path><path id="MJX-4-TEX-I-1D45C" d="M201 -11Q126 -11 80 38T34 156Q34 221 64 279T146 380Q222 441 301 441Q333 441 341 440Q354 437 367 433T402 417T438 387T464 338T476 268Q476 161 390 75T201 -11ZM121 120Q121 70 147 48T206 26Q250 26 289 58T351 142Q360 163 374 216T388 308Q388 352 370 375Q346 405 306 405Q243 405 195 347Q158 303 140 230T121 120Z"></path><path id="MJX-4-TEX-I-1D45D" d="M23 287Q24 290 25 295T30 317T40 348T55 381T75 411T101 433T134 442Q209 442 230 378L240 387Q302 442 358 442Q423 442 460 395T497 281Q497 173 421 82T249 -10Q227 -10 210 -4Q199 1 187 11T168 28L161 36Q160 35 139 -51T118 -138Q118 -144 126 -145T163 -148H188Q194 -155 194 -157T191 -175Q188 -187 185 -190T172 -194Q170 -194 161 -194T127 -193T65 -192Q-5 -192 -24 -194H-32Q-39 -187 -39 -183Q-37 -156 -26 -148H-6Q28 -147 33 -136Q36 -130 94 103T155 350Q156 355 156 364Q156 405 131 405Q109 405 94 377T71 316T59 280Q57 278 43 278H29Q23 284 23 287ZM178 102Q200 26 252 26Q282 26 310 49T356 107Q374 141 392 215T411 325V331Q411 405 350 405Q339 405 328 402T306 393T286 380T269 365T254 350T243 336T235 326L232 322Q232 321 229 308T218 264T204 212Q178 106 178 102Z"></path><path id="MJX-4-TEX-I-1D461" d="M26 385Q19 392 19 395Q19 399 22 411T27 425Q29 430 36 430T87 431H140L159 511Q162 522 166 540T173 566T179 586T187 603T197 615T211 624T229 626Q247 625 254 615T261 596Q261 589 252 549T232 470L222 433Q222 431 272 431H323Q330 424 330 420Q330 398 317 385H210L174 240Q135 80 135 68Q135 26 162 26Q197 26 230 60T283 144Q285 150 288 151T303 153H307Q322 153 322 145Q322 142 319 133Q314 117 301 95T267 48T216 6T155 -11Q125 -11 98 4T59 56Q57 64 57 83V101L92 241Q127 382 128 383Q128 385 77 385H26Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="mtext"><use data-c="45" xlink:href="#MJX-4-TEX-N-45"></use><use data-c="73" xlink:href="#MJX-4-TEX-N-73" transform="translate(681,0)"></use><use data-c="74" xlink:href="#MJX-4-TEX-N-74" transform="translate(1075,0)"></use><use data-c="69" xlink:href="#MJX-4-TEX-N-69" transform="translate(1464,0)"></use><use data-c="6D" xlink:href="#MJX-4-TEX-N-6D" transform="translate(1742,0)"></use><use data-c="61" xlink:href="#MJX-4-TEX-N-61" transform="translate(2575,0)"></use><use data-c="74" xlink:href="#MJX-4-TEX-N-74" transform="translate(3075,0)"></use><use data-c="65" xlink:href="#MJX-4-TEX-N-65" transform="translate(3464,0)"></use><use data-c="64" xlink:href="#MJX-4-TEX-N-64" transform="translate(3908,0)"></use><use data-c="20" xlink:href="#MJX-4-TEX-N-20" transform="translate(4464,0)"></use><use data-c="4C" xlink:href="#MJX-4-TEX-N-4C" transform="translate(4714,0)"></use><use data-c="69" xlink:href="#MJX-4-TEX-N-69" transform="translate(5339,0)"></use><use data-c="62" xlink:href="#MJX-4-TEX-N-62" transform="translate(5617,0)"></use><use data-c="72" xlink:href="#MJX-4-TEX-N-72" transform="translate(6173,0)"></use><use data-c="61" xlink:href="#MJX-4-TEX-N-61" transform="translate(6565,0)"></use><use data-c="72" xlink:href="#MJX-4-TEX-N-72" transform="translate(7065,0)"></use><use data-c="79" xlink:href="#MJX-4-TEX-N-79" transform="translate(7457,0)"></use><use data-c="20" xlink:href="#MJX-4-TEX-N-20" transform="translate(7985,0)"></use><use data-c="53" xlink:href="#MJX-4-TEX-N-53" transform="translate(8235,0)"></use><use data-c="69" xlink:href="#MJX-4-TEX-N-69" transform="translate(8791,0)"></use><use data-c="7A" xlink:href="#MJX-4-TEX-N-7A" transform="translate(9069,0)"></use><use data-c="65" xlink:href="#MJX-4-TEX-N-65" transform="translate(9513,0)"></use></g><g data-mml-node="mo" transform="translate(10234.8,0)"><use data-c="3D" xlink:href="#MJX-4-TEX-N-3D"></use></g><g data-mml-node="mfrac" transform="translate(11290.6,0)"><g data-mml-node="mrow" transform="translate(520.2,676)"><g data-mml-node="msub"><g data-mml-node="mi"><use data-c="1D436" xlink:href="#MJX-4-TEX-I-1D436"></use></g><g data-mml-node="mn" transform="translate(748,-150) scale(0.707)"><use data-c="31" xlink:href="#MJX-4-TEX-N-31"></use></g></g><g data-mml-node="mo" transform="translate(1373.8,0)"><use data-c="D7" xlink:href="#MJX-4-TEX-N-D7"></use></g><g data-mml-node="msub" transform="translate(2374,0)"><g data-mml-node="mi"><use data-c="1D436" xlink:href="#MJX-4-TEX-I-1D436"></use></g><g data-mml-node="mn" transform="translate(748,-150) scale(0.707)"><use data-c="30" xlink:href="#MJX-4-TEX-N-30"></use></g></g></g><g data-mml-node="mrow" transform="translate(220,-686)"><g data-mml-node="msub"><g data-mml-node="mi"><use data-c="1D436" xlink:href="#MJX-4-TEX-I-1D436"></use></g><g data-mml-node="mn" transform="translate(748,-150) scale(0.707)"><use data-c="31" xlink:href="#MJX-4-TEX-N-31"></use></g></g><g data-mml-node="mo" transform="translate(1373.8,0)"><use data-c="2212" xlink:href="#MJX-4-TEX-N-2212"></use></g><g data-mml-node="msub" transform="translate(2374,0)"><g data-mml-node="mi"><use data-c="1D436" xlink:href="#MJX-4-TEX-I-1D436"></use></g><g data-mml-node="TeXAtom" transform="translate(748,-150) scale(0.707)" data-mjx-texclass="ORD"><g data-mml-node="mi"><use data-c="1D45C" xlink:href="#MJX-4-TEX-I-1D45C"></use></g><g data-mml-node="mi" transform="translate(485,0)"><use data-c="1D45D" xlink:href="#MJX-4-TEX-I-1D45D"></use></g><g data-mml-node="mi" transform="translate(988,0)"><use data-c="1D461" xlink:href="#MJX-4-TEX-I-1D461"></use></g></g></g></g><rect width="4325.9" height="60" x="120" y="220"></rect></g></g></g></svg><mjx-assistive-mml unselectable="on" display="block"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><mtext>Estimated Library Size</mtext><mo>=</mo><mfrac><mrow><msub><mi>C</mi><mn>1</mn></msub><mo>×</mo><msub><mi>C</mi><mn>0</mn></msub></mrow><mrow><msub><mi>C</mi><mn>1</mn></msub><mo>−</mo><msub><mi>C</mi><mrow data-mjx-texclass="ORD"><mi>o</mi><mi>p</mi><mi>t</mi></mrow></msub></mrow></mfrac></math></mjx-assistive-mml></mjx-container></div></div><p><mjx-container class="MathJax" jax="SVG" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="2.605ex" height="1.934ex" role="img" focusable="false" viewBox="0 -705 1151.6 855" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -0.339ex;"><defs><path id="MJX-30-TEX-I-1D436" d="M50 252Q50 367 117 473T286 641T490 704Q580 704 633 653Q642 643 648 636T656 626L657 623Q660 623 684 649Q691 655 699 663T715 679T725 690L740 705H746Q760 705 760 698Q760 694 728 561Q692 422 692 421Q690 416 687 415T669 413H653Q647 419 647 422Q647 423 648 429T650 449T651 481Q651 552 619 605T510 659Q484 659 454 652T382 628T299 572T226 479Q194 422 175 346T156 222Q156 108 232 58Q280 24 350 24Q441 24 512 92T606 240Q610 253 612 255T628 257Q648 257 648 248Q648 243 647 239Q618 132 523 55T319 -22Q206 -22 128 53T50 252Z"></path><path id="MJX-30-TEX-N-31" d="M213 578L200 573Q186 568 160 563T102 556H83V602H102Q149 604 189 617T245 641T273 663Q275 666 285 666Q294 666 302 660V361L303 61Q310 54 315 52T339 48T401 46H427V0H416Q395 3 257 3Q121 3 100 0H88V46H114Q136 46 152 46T177 47T193 50T201 52T207 57T213 61V578Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="msub"><g data-mml-node="mi"><use data-c="1D436" xlink:href="#MJX-30-TEX-I-1D436"></use></g><g data-mml-node="mn" transform="translate(748,-150) scale(0.707)"><use data-c="31" xlink:href="#MJX-30-TEX-N-31"></use></g></g></g></g></svg><mjx-assistive-mml unselectable="on" display="inline"><math xmlns="http://www.w3.org/1998/Math/MathML"><msub><mi>C</mi><mn>1</mn></msub></math></mjx-assistive-mml></mjx-container><script type="math/tex"> C_1 </script><span> 是所有配对 reads 的总数（包括重复的）。</span></p><p><mjx-container class="MathJax" jax="SVG" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="2.605ex" height="1.97ex" role="img" focusable="false" viewBox="0 -705 1151.6 870.6" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -0.375ex;"><defs><path id="MJX-31-TEX-I-1D436" d="M50 252Q50 367 117 473T286 641T490 704Q580 704 633 653Q642 643 648 636T656 626L657 623Q660 623 684 649Q691 655 699 663T715 679T725 690L740 705H746Q760 705 760 698Q760 694 728 561Q692 422 692 421Q690 416 687 415T669 413H653Q647 419 647 422Q647 423 648 429T650 449T651 481Q651 552 619 605T510 659Q484 659 454 652T382 628T299 572T226 479Q194 422 175 346T156 222Q156 108 232 58Q280 24 350 24Q441 24 512 92T606 240Q610 253 612 255T628 257Q648 257 648 248Q648 243 647 239Q618 132 523 55T319 -22Q206 -22 128 53T50 252Z"></path><path id="MJX-31-TEX-N-30" d="M96 585Q152 666 249 666Q297 666 345 640T423 548Q460 465 460 320Q460 165 417 83Q397 41 362 16T301 -15T250 -22Q224 -22 198 -16T137 16T82 83Q39 165 39 320Q39 494 96 585ZM321 597Q291 629 250 629Q208 629 178 597Q153 571 145 525T137 333Q137 175 145 125T181 46Q209 16 250 16Q290 16 318 46Q347 76 354 130T362 333Q362 478 354 524T321 597Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="msub"><g data-mml-node="mi"><use data-c="1D436" xlink:href="#MJX-31-TEX-I-1D436"></use></g><g data-mml-node="mn" transform="translate(748,-150) scale(0.707)"><use data-c="30" xlink:href="#MJX-31-TEX-N-30"></use></g></g></g></g></svg><mjx-assistive-mml unselectable="on" display="inline"><math xmlns="http://www.w3.org/1998/Math/MathML"><msub><mi>C</mi><mn>0</mn></msub></math></mjx-assistive-mml></mjx-container><script type="math/tex"> C_0 </script><span> 是唯一的配对 reads 数目（非重复的）。</span></p><p><mjx-container class="MathJax" jax="SVG" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="3.964ex" height="2.245ex" role="img" focusable="false" viewBox="0 -705 1751.9 992.2" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -0.65ex;"><defs><path id="MJX-32-TEX-I-1D436" d="M50 252Q50 367 117 473T286 641T490 704Q580 704 633 653Q642 643 648 636T656 626L657 623Q660 623 684 649Q691 655 699 663T715 679T725 690L740 705H746Q760 705 760 698Q760 694 728 561Q692 422 692 421Q690 416 687 415T669 413H653Q647 419 647 422Q647 423 648 429T650 449T651 481Q651 552 619 605T510 659Q484 659 454 652T382 628T299 572T226 479Q194 422 175 346T156 222Q156 108 232 58Q280 24 350 24Q441 24 512 92T606 240Q610 253 612 255T628 257Q648 257 648 248Q648 243 647 239Q618 132 523 55T319 -22Q206 -22 128 53T50 252Z"></path><path id="MJX-32-TEX-I-1D45C" d="M201 -11Q126 -11 80 38T34 156Q34 221 64 279T146 380Q222 441 301 441Q333 441 341 440Q354 437 367 433T402 417T438 387T464 338T476 268Q476 161 390 75T201 -11ZM121 120Q121 70 147 48T206 26Q250 26 289 58T351 142Q360 163 374 216T388 308Q388 352 370 375Q346 405 306 405Q243 405 195 347Q158 303 140 230T121 120Z"></path><path id="MJX-32-TEX-I-1D45D" d="M23 287Q24 290 25 295T30 317T40 348T55 381T75 411T101 433T134 442Q209 442 230 378L240 387Q302 442 358 442Q423 442 460 395T497 281Q497 173 421 82T249 -10Q227 -10 210 -4Q199 1 187 11T168 28L161 36Q160 35 139 -51T118 -138Q118 -144 126 -145T163 -148H188Q194 -155 194 -157T191 -175Q188 -187 185 -190T172 -194Q170 -194 161 -194T127 -193T65 -192Q-5 -192 -24 -194H-32Q-39 -187 -39 -183Q-37 -156 -26 -148H-6Q28 -147 33 -136Q36 -130 94 103T155 350Q156 355 156 364Q156 405 131 405Q109 405 94 377T71 316T59 280Q57 278 43 278H29Q23 284 23 287ZM178 102Q200 26 252 26Q282 26 310 49T356 107Q374 141 392 215T411 325V331Q411 405 350 405Q339 405 328 402T306 393T286 380T269 365T254 350T243 336T235 326L232 322Q232 321 229 308T218 264T204 212Q178 106 178 102Z"></path><path id="MJX-32-TEX-I-1D461" d="M26 385Q19 392 19 395Q19 399 22 411T27 425Q29 430 36 430T87 431H140L159 511Q162 522 166 540T173 566T179 586T187 603T197 615T211 624T229 626Q247 625 254 615T261 596Q261 589 252 549T232 470L222 433Q222 431 272 431H323Q330 424 330 420Q330 398 317 385H210L174 240Q135 80 135 68Q135 26 162 26Q197 26 230 60T283 144Q285 150 288 151T303 153H307Q322 153 322 145Q322 142 319 133Q314 117 301 95T267 48T216 6T155 -11Q125 -11 98 4T59 56Q57 64 57 83V101L92 241Q127 382 128 383Q128 385 77 385H26Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="msub"><g data-mml-node="mi"><use data-c="1D436" xlink:href="#MJX-32-TEX-I-1D436"></use></g><g data-mml-node="TeXAtom" transform="translate(748,-150) scale(0.707)" data-mjx-texclass="ORD"><g data-mml-node="mi"><use data-c="1D45C" xlink:href="#MJX-32-TEX-I-1D45C"></use></g><g data-mml-node="mi" transform="translate(485,0)"><use data-c="1D45D" xlink:href="#MJX-32-TEX-I-1D45D"></use></g><g data-mml-node="mi" transform="translate(988,0)"><use data-c="1D461" xlink:href="#MJX-32-TEX-I-1D461"></use></g></g></g></g></g></svg><mjx-assistive-mml unselectable="on" display="inline"><math xmlns="http://www.w3.org/1998/Math/MathML"><msub><mi>C</mi><mrow data-mjx-texclass="ORD"><mi>o</mi><mi>p</mi><mi>t</mi></mrow></msub></math></mjx-assistive-mml></mjx-container><script type="math/tex"> C_{opt} </script><span> 是光学重复的 reads 数目。</span></p><p>&nbsp;</p><ul><li><p><span>The estimated library sizes are </span><strong><span>proportional</span></strong><span> to the abundance of </span><strong><span>the targeted epitope</span></strong><span> and to the </span><strong><span>quality of the antibody</span></strong><span> used</span></p></li><li><p><span>The estimated library sizes of </span><strong><span>IgG sample</span></strong><span>s are expected to be very low.</span></p></li></ul></div></li></ul><p>&nbsp;</p></li><li><p><strong><span>Remove duplicates when we need</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">picard MarkDuplicates \</span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-I</span> sorted_input.bam \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-O</span> deduplicated_output.bam \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-M</span> output.metrics.txt \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--REMOVE_DUPLICATES</span> <span class="cm-atom">true</span> \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--OPTICAL_DUPLICATE_PIXEL_DISTANCE</span> <span class="cm-number">2500</span> \ &nbsp; <span class="cm-comment"># 定义了2个clusters之间的最大偏移量（以像素为单位），超过此距离的cluster将不被视为光学重复</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--CREATE_INDEX</span> <span class="cm-atom">true</span> &nbsp;</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 153px;"></div><div class="CodeMirror-gutters" style="display: none; height: 153px;"></div></div></div></pre><div class="md-alert md-alert-important important"><p><span class='md-alert-text md-alert-text-important'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M0 1.75C0 .784.784 0 1.75 0h12.5C15.216 0 16 .784 16 1.75v9.5A1.75 1.75 0 0 1 14.25 13H8.06l-2.573 2.573A1.458 1.458 0 0 1 3 14.543V13H1.75A1.75 1.75 0 0 1 0 11.25Zm1.75-.25a.25.25 0 0 0-.25.25v9.5c0 .138.112.25.25.25h2a.75.75 0 0 1 .75.75v2.19l2.72-2.72a.749.749 0 0 1 .53-.22h6.5a.25.25 0 0 0 .25-.25v-9.5a.25.25 0 0 0-.25-.25Zm7 2.25v2.5a.75.75 0 0 1-1.5 0v-2.5a.75.75 0 0 1 1.5 0ZM9 9a1 1 0 1 1-2 0 1 1 0 0 1 2 0Z"></path></svg>Important</span><br></p><p><strong><span>Sometimes the duplicates rate of IgG is relatively high ???</span></strong></p><ul><li><p><span>因为 IgG 是一种非特异性的抗体，在实验中不与目标蛋白或特定的 DNA 片段结合。</span></p></li><li><p><span>因此，IgG 对照样本中大部分的测序 reads 是由非特异性结合引起的</span></p></li><li><p><span>在下游分析之前，从 IgG 数据集中移除重复样本是合适的。</span></p></li></ul></div></li></ol><p>&nbsp;</p><hr /><h2 id='remove-duplicates-in-cuttag'><span>Remove duplicates in CUT&amp;Tag</span></h2><ul><li><p><span>在 </span><strong><span>CUT&amp;Tag</span></strong><span>实验中，DNA被转座酶（Tn5）切割并标记，形成包含抗体靶向区域的短片段。这些短片段通常会映射到染色体的特定位置</span></p><p><span>但由于DNA的可及性及Tn5转座酶的</span><strong><span>偏好性</span></strong><span>，具有相同起始和终止位置的片段是常见的。</span></p><p><img src="./picture/011.png"  width="500"/></p></li><li><p><span>在实践中，我们发现</span><strong><span>高质量的 CUT&amp;Tag</span></strong><span> 数据集的明显</span><strong><span>重复率很低</span></strong><span>，即使是明显的 &quot;重复 &quot;片段也可能是真实的片段。 因此，我们不建议删除重复片段。</span></p></li></ul><div class="md-alert md-alert-important important"><p><span class='md-alert-text md-alert-text-important'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M0 1.75C0 .784.784 0 1.75 0h12.5C15.216 0 16 .784 16 1.75v9.5A1.75 1.75 0 0 1 14.25 13H8.06l-2.573 2.573A1.458 1.458 0 0 1 3 14.543V13H1.75A1.75 1.75 0 0 1 0 11.25Zm1.75-.25a.25.25 0 0 0-.25.25v9.5c0 .138.112.25.25.25h2a.75.75 0 0 1 .75.75v2.19l2.72-2.72a.749.749 0 0 1 .53-.22h6.5a.25.25 0 0 0 .25-.25v-9.5a.25.25 0 0 0-.25-.25Zm7 2.25v2.5a.75.75 0 0 1-1.5 0v-2.5a.75.75 0 0 1 1.5 0ZM9 9a1 1 0 1 1-2 0 1 1 0 0 1 2 0Z"></path></svg>Important</span><br></p><p><strong><span>Should duplicate reads been removed in CUT&amp;Run data ???</span></strong></p><ul><li><p><span>如果重复率很低，比如</span><strong><span>20-30%</span></strong><span>，则不需要删除它们。</span></p></li><li><p><span>然而，如果这个比率很高，大约有</span><strong><span>80-100%</span></strong><span>。您需要检查是否在实验阶段过度测序，或者是否有其他潜在问题导致实验失败。</span></p></li><li><p><span>如果一切看起来都很好，在继续下游分析之前，您应该考虑remove duplicates</span></p></li></ul></div><p>&nbsp;</p><hr /><h1 id='assess-mapped-fragment-size-distributionoptial'><span>Assess mapped fragment size distribution(optial)</span></h1><h2 id='what-is-the-fragment-'><span>What is the fragment ?</span></h2><ul><li><p><span>read1 和 read2 比对到参考基因组后，</span><strong><span>从 read1 的 5&#39; 端到 read2 的 3&#39; 端之间</span></strong><span>（无论方向），就是 </span><strong><span>fragment</span></strong><span> </span></p></li><li><p><span>在 BAM 文件中，第9列（</span><code>TLEN</code><span>，即 Template Length）记录了 </span><strong><span>read pair 所对应的 fragment length</span></strong></p></li><li><p><span>read1 的 TLEN 是正值，read2 的 TLEN 是负值，用 </span><code>abs($9)</code><span> 取其绝对值来表示</span><strong><span>fragment length</span></strong></p></li><li><p><strong><span>fragment Count</span></strong><span> 则对</span><strong><span>fragment length</span></strong><span>出现的次数</span><strong><span>除以2</span></strong></p><p><span>双端测序中, 每对 reads 都记录了 </span><strong><span>fragment length</span></strong><span>，且出现了两次，除以 2得到真正的</span><strong><span>fragment count</span></strong></p></li></ul><p>&nbsp;</p><h2 id='tn5-转座酶偏好性'><span>Tn5 转座酶偏好性</span></h2><ul><li><p><span>以</span><strong><span>组蛋白修饰为靶标</span></strong><span>的 CUT&amp;Tag  反应主要产生</span><strong><span>核糖体长度（约 180 bp）</span></strong><span>或</span><strong><span>该长度倍数</span></strong><span>的片段</span></p></li><li><p><span>以</span><strong><span>转录因子为靶标</span></strong><span>的 CUT&amp;Tag 反应主要产生</span><strong><span>核糖体大小的片段</span></strong><span>和</span><strong><span>数量不等的较短片段</span></strong><span>，分别来自邻近的核糖体和因子结合位点</span></p><p><img src="./picture/044.png"></p></li><li><p><strong><span>10-bp sawtooth periodicity</span></strong></p><p><span>核小体上的 DNA 是以约10个碱基对为一组的周期性缠绕在组蛋白上的。由于核小体结构的这种规律性，</span><strong><span>Tn5 转座酶</span></strong><span>在核小体表面切割 DNA 时，会产生一种锯齿状周期性（Sawtooth Periodicity）现象</span></p><div contenteditable="false" spellcheck="false" class="mathjax-block md-end-block md-math-block md-rawblock" id="mathjax-n2557" cid="n2557" mdtype="math_block" data-math-tag-before="0" data-math-tag-after="0" data-math-labels="[]"><div class="md-math-container"><mjx-container class="MathJax" jax="SVG" display="true" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="20.372ex" height="2.262ex" role="img" focusable="false" viewBox="0 -750 9004.6 1000" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -0.566ex;"><defs><path id="MJX-5-TEX-N-31" d="M213 578L200 573Q186 568 160 563T102 556H83V602H102Q149 604 189 617T245 641T273 663Q275 666 285 666Q294 666 302 660V361L303 61Q310 54 315 52T339 48T401 46H427V0H416Q395 3 257 3Q121 3 100 0H88V46H114Q136 46 152 46T177 47T193 50T201 52T207 57T213 61V578Z"></path><path id="MJX-5-TEX-N-34" d="M462 0Q444 3 333 3Q217 3 199 0H190V46H221Q241 46 248 46T265 48T279 53T286 61Q287 63 287 115V165H28V211L179 442Q332 674 334 675Q336 677 355 677H373L379 671V211H471V165H379V114Q379 73 379 66T385 54Q393 47 442 46H471V0H462ZM293 211V545L74 212L183 211H293Z"></path><path id="MJX-5-TEX-N-37" d="M55 458Q56 460 72 567L88 674Q88 676 108 676H128V672Q128 662 143 655T195 646T364 644H485V605L417 512Q408 500 387 472T360 435T339 403T319 367T305 330T292 284T284 230T278 162T275 80Q275 66 275 52T274 28V19Q270 2 255 -10T221 -22Q210 -22 200 -19T179 0T168 40Q168 198 265 368Q285 400 349 489L395 552H302Q128 552 119 546Q113 543 108 522T98 479L95 458V455H55V458Z"></path><path id="MJX-5-TEX-N-2F" d="M423 750Q432 750 438 744T444 730Q444 725 271 248T92 -240Q85 -250 75 -250Q68 -250 62 -245T56 -231Q56 -221 230 257T407 740Q411 750 423 750Z"></path><path id="MJX-5-TEX-N-2E" d="M78 60Q78 84 95 102T138 120Q162 120 180 104T199 61Q199 36 182 18T139 0T96 17T78 60Z"></path><path id="MJX-5-TEX-N-35" d="M164 157Q164 133 148 117T109 101H102Q148 22 224 22Q294 22 326 82Q345 115 345 210Q345 313 318 349Q292 382 260 382H254Q176 382 136 314Q132 307 129 306T114 304Q97 304 95 310Q93 314 93 485V614Q93 664 98 664Q100 666 102 666Q103 666 123 658T178 642T253 634Q324 634 389 662Q397 666 402 666Q410 666 410 648V635Q328 538 205 538Q174 538 149 544L139 546V374Q158 388 169 396T205 412T256 420Q337 420 393 355T449 201Q449 109 385 44T229 -22Q148 -22 99 32T50 154Q50 178 61 192T84 210T107 214Q132 214 148 197T164 157Z"></path><path id="MJX-5-TEX-N-38" d="M70 417T70 494T124 618T248 666Q319 666 374 624T429 515Q429 485 418 459T392 417T361 389T335 371T324 363L338 354Q352 344 366 334T382 323Q457 264 457 174Q457 95 399 37T249 -22Q159 -22 101 29T43 155Q43 263 172 335L154 348Q133 361 127 368Q70 417 70 494ZM286 386L292 390Q298 394 301 396T311 403T323 413T334 425T345 438T355 454T364 471T369 491T371 513Q371 556 342 586T275 624Q268 625 242 625Q201 625 165 599T128 534Q128 511 141 492T167 463T217 431Q224 426 228 424L286 386ZM250 21Q308 21 350 55T392 137Q392 154 387 169T375 194T353 216T330 234T301 253T274 270Q260 279 244 289T218 306L210 311Q204 311 181 294T133 239T107 157Q107 98 150 60T250 21Z"></path><path id="MJX-5-TEX-N-3D" d="M56 347Q56 360 70 367H707Q722 359 722 347Q722 336 708 328L390 327H72Q56 332 56 347ZM56 153Q56 168 72 173H708Q722 163 722 153Q722 140 707 133H70Q56 140 56 153Z"></path><path id="MJX-5-TEX-N-30" d="M96 585Q152 666 249 666Q297 666 345 640T423 548Q460 465 460 320Q460 165 417 83Q397 41 362 16T301 -15T250 -22Q224 -22 198 -16T137 16T82 83Q39 165 39 320Q39 494 96 585ZM321 597Q291 629 250 629Q208 629 178 597Q153 571 145 525T137 333Q137 175 145 125T181 46Q209 16 250 16Q290 16 318 46Q347 76 354 130T362 333Q362 478 354 524T321 597Z"></path><path id="MJX-5-TEX-I-1D44F" d="M73 647Q73 657 77 670T89 683Q90 683 161 688T234 694Q246 694 246 685T212 542Q204 508 195 472T180 418L176 399Q176 396 182 402Q231 442 283 442Q345 442 383 396T422 280Q422 169 343 79T173 -11Q123 -11 82 27T40 150V159Q40 180 48 217T97 414Q147 611 147 623T109 637Q104 637 101 637H96Q86 637 83 637T76 640T73 647ZM336 325V331Q336 405 275 405Q258 405 240 397T207 376T181 352T163 330L157 322L136 236Q114 150 114 114Q114 66 138 42Q154 26 178 26Q211 26 245 58Q270 81 285 114T318 219Q336 291 336 325Z"></path><path id="MJX-5-TEX-I-1D44E" d="M33 157Q33 258 109 349T280 441Q331 441 370 392Q386 422 416 422Q429 422 439 414T449 394Q449 381 412 234T374 68Q374 43 381 35T402 26Q411 27 422 35Q443 55 463 131Q469 151 473 152Q475 153 483 153H487Q506 153 506 144Q506 138 501 117T481 63T449 13Q436 0 417 -8Q409 -10 393 -10Q359 -10 336 5T306 36L300 51Q299 52 296 50Q294 48 292 46Q233 -10 172 -10Q117 -10 75 30T33 157ZM351 328Q351 334 346 350T323 385T277 405Q242 405 210 374T160 293Q131 214 119 129Q119 126 119 118T118 106Q118 61 136 44T179 26Q217 26 254 59T298 110Q300 114 325 217T351 328Z"></path><path id="MJX-5-TEX-I-1D460" d="M131 289Q131 321 147 354T203 415T300 442Q362 442 390 415T419 355Q419 323 402 308T364 292Q351 292 340 300T328 326Q328 342 337 354T354 372T367 378Q368 378 368 379Q368 382 361 388T336 399T297 405Q249 405 227 379T204 326Q204 301 223 291T278 274T330 259Q396 230 396 163Q396 135 385 107T352 51T289 7T195 -10Q118 -10 86 19T53 87Q53 126 74 143T118 160Q133 160 146 151T160 120Q160 94 142 76T111 58Q109 57 108 57T107 55Q108 52 115 47T146 34T201 27Q237 27 263 38T301 66T318 97T323 122Q323 150 302 164T254 181T195 196T148 231Q131 256 131 289Z"></path><path id="MJX-5-TEX-I-1D452" d="M39 168Q39 225 58 272T107 350T174 402T244 433T307 442H310Q355 442 388 420T421 355Q421 265 310 237Q261 224 176 223Q139 223 138 221Q138 219 132 186T125 128Q125 81 146 54T209 26T302 45T394 111Q403 121 406 121Q410 121 419 112T429 98T420 82T390 55T344 24T281 -1T205 -11Q126 -11 83 42T39 168ZM373 353Q367 405 305 405Q272 405 244 391T199 357T170 316T154 280T149 261Q149 260 169 260Q282 260 327 284T373 353Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="mn"><use data-c="31" xlink:href="#MJX-5-TEX-N-31"></use><use data-c="34" xlink:href="#MJX-5-TEX-N-34" transform="translate(500,0)"></use><use data-c="37" xlink:href="#MJX-5-TEX-N-37" transform="translate(1000,0)"></use></g><g data-mml-node="TeXAtom" data-mjx-texclass="ORD" transform="translate(1500,0)"><g data-mml-node="mo"><use data-c="2F" xlink:href="#MJX-5-TEX-N-2F"></use></g></g><g data-mml-node="mn" transform="translate(2000,0)"><use data-c="31" xlink:href="#MJX-5-TEX-N-31"></use><use data-c="2E" xlink:href="#MJX-5-TEX-N-2E" transform="translate(500,0)"></use><use data-c="37" xlink:href="#MJX-5-TEX-N-37" transform="translate(778,0)"></use><use data-c="35" xlink:href="#MJX-5-TEX-N-35" transform="translate(1278,0)"></use></g><g data-mml-node="TeXAtom" data-mjx-texclass="ORD" transform="translate(3778,0)"><g data-mml-node="mo"><use data-c="2F" xlink:href="#MJX-5-TEX-N-2F"></use></g></g><g data-mml-node="mn" transform="translate(4278,0)"><use data-c="38" xlink:href="#MJX-5-TEX-N-38"></use></g><g data-mml-node="mo" transform="translate(5055.8,0)"><use data-c="3D" xlink:href="#MJX-5-TEX-N-3D"></use></g><g data-mml-node="mn" transform="translate(6111.6,0)"><use data-c="31" xlink:href="#MJX-5-TEX-N-31"></use><use data-c="30" xlink:href="#MJX-5-TEX-N-30" transform="translate(500,0)"></use></g><g data-mml-node="mi" transform="translate(7111.6,0)"><use data-c="1D44F" xlink:href="#MJX-5-TEX-I-1D44F"></use></g><g data-mml-node="mi" transform="translate(7540.6,0)"><use data-c="1D44E" xlink:href="#MJX-5-TEX-I-1D44E"></use></g><g data-mml-node="mi" transform="translate(8069.6,0)"><use data-c="1D460" xlink:href="#MJX-5-TEX-I-1D460"></use></g><g data-mml-node="mi" transform="translate(8538.6,0)"><use data-c="1D452" xlink:href="#MJX-5-TEX-I-1D452"></use></g></g></g></svg><mjx-assistive-mml unselectable="on" display="block"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><mn>147</mn><mrow data-mjx-texclass="ORD"><mo>/</mo></mrow><mn>1.75</mn><mrow data-mjx-texclass="ORD"><mo>/</mo></mrow><mn>8</mn><mo>=</mo><mn>10</mn><mi>b</mi><mi>a</mi><mi>s</mi><mi>e</mi></math></mjx-assistive-mml></mjx-container></div></div><ul><li><p><mjx-container class="MathJax" jax="SVG" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="4.023ex" height="1.579ex" role="img" focusable="false" viewBox="0 -676 1778 698" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -0.05ex;"><defs><path id="MJX-33-TEX-N-31" d="M213 578L200 573Q186 568 160 563T102 556H83V602H102Q149 604 189 617T245 641T273 663Q275 666 285 666Q294 666 302 660V361L303 61Q310 54 315 52T339 48T401 46H427V0H416Q395 3 257 3Q121 3 100 0H88V46H114Q136 46 152 46T177 47T193 50T201 52T207 57T213 61V578Z"></path><path id="MJX-33-TEX-N-2E" d="M78 60Q78 84 95 102T138 120Q162 120 180 104T199 61Q199 36 182 18T139 0T96 17T78 60Z"></path><path id="MJX-33-TEX-N-37" d="M55 458Q56 460 72 567L88 674Q88 676 108 676H128V672Q128 662 143 655T195 646T364 644H485V605L417 512Q408 500 387 472T360 435T339 403T319 367T305 330T292 284T284 230T278 162T275 80Q275 66 275 52T274 28V19Q270 2 255 -10T221 -22Q210 -22 200 -19T179 0T168 40Q168 198 265 368Q285 400 349 489L395 552H302Q128 552 119 546Q113 543 108 522T98 479L95 458V455H55V458Z"></path><path id="MJX-33-TEX-N-35" d="M164 157Q164 133 148 117T109 101H102Q148 22 224 22Q294 22 326 82Q345 115 345 210Q345 313 318 349Q292 382 260 382H254Q176 382 136 314Q132 307 129 306T114 304Q97 304 95 310Q93 314 93 485V614Q93 664 98 664Q100 666 102 666Q103 666 123 658T178 642T253 634Q324 634 389 662Q397 666 402 666Q410 666 410 648V635Q328 538 205 538Q174 538 149 544L139 546V374Q158 388 169 396T205 412T256 420Q337 420 393 355T449 201Q449 109 385 44T229 -22Q148 -22 99 32T50 154Q50 178 61 192T84 210T107 214Q132 214 148 197T164 157Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="mn"><use data-c="31" xlink:href="#MJX-33-TEX-N-31"></use><use data-c="2E" xlink:href="#MJX-33-TEX-N-2E" transform="translate(500,0)"></use><use data-c="37" xlink:href="#MJX-33-TEX-N-37" transform="translate(778,0)"></use><use data-c="35" xlink:href="#MJX-33-TEX-N-35" transform="translate(1278,0)"></use></g></g></g></svg><mjx-assistive-mml unselectable="on" display="inline"><math xmlns="http://www.w3.org/1998/Math/MathML"><mn>1.75</mn></math></mjx-assistive-mml></mjx-container><script type="math/tex">1.75</script><span>：DNA绕核小体1.75圈</span></p></li><li><p><mjx-container class="MathJax" jax="SVG" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="1.131ex" height="1.557ex" role="img" focusable="false" viewBox="0 -666 500 688" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -0.05ex;"><defs><path id="MJX-34-TEX-N-38" d="M70 417T70 494T124 618T248 666Q319 666 374 624T429 515Q429 485 418 459T392 417T361 389T335 371T324 363L338 354Q352 344 366 334T382 323Q457 264 457 174Q457 95 399 37T249 -22Q159 -22 101 29T43 155Q43 263 172 335L154 348Q133 361 127 368Q70 417 70 494ZM286 386L292 390Q298 394 301 396T311 403T323 413T334 425T345 438T355 454T364 471T369 491T371 513Q371 556 342 586T275 624Q268 625 242 625Q201 625 165 599T128 534Q128 511 141 492T167 463T217 431Q224 426 228 424L286 386ZM250 21Q308 21 350 55T392 137Q392 154 387 169T375 194T353 216T330 234T301 253T274 270Q260 279 244 289T218 306L210 311Q204 311 181 294T133 239T107 157Q107 98 150 60T250 21Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="mn"><use data-c="38" xlink:href="#MJX-34-TEX-N-38"></use></g></g></g></svg><mjx-assistive-mml unselectable="on" display="inline"><math xmlns="http://www.w3.org/1998/Math/MathML"><mn>8</mn></math></mjx-assistive-mml></mjx-container><script type="math/tex">8</script><span>：每个核小体由8个组蛋白组成</span></p><p>&nbsp;</p></li></ul><p><span>在片段长度的绘图中，随着片段长度的增加，可以看到大约每隔10 bp 就有一个小峰值。这种锯齿状模式是成功的 </span><strong><span>ATAC-seq / CUT&amp;Tag</span></strong><span> 实验的特征之一，表明 </span><strong><span>Tn5 转座酶</span></strong><span>切割主要发生在核小体表面，并且具有高度的结构依赖性。</span></p><figure class='table-figure'><table><thead><tr><th><span>项目</span></th><th><span>ATAC-seq / CUT&amp;Tag</span></th><th><span>ChIP-seq</span></th></tr></thead><tbody><tr><td><span>酶类型</span></td><td><span>Tn5 转座酶（直接切 DNA）</span></td><td><span>超声破碎或酶消化</span></td></tr><tr><td><span>切割机制</span></td><td><span>定位性强、与染色质结构直接相关</span></td><td><span>非定向，破碎随机</span></td></tr><tr><td><span>是否与核小体周期有关</span></td><td><span>是，有强烈的10-bp sawtooth periodicity</span></td><td><span>否，不显示10-bp sawtooth periodicity</span></td></tr></tbody></table></figure><p><img src="./picture/045.png" width="700"/></p></li></ul><p>&nbsp;</p><ul><li><p><strong><span>The smaller fragments (50-100 bp)</span></strong></p><ul><li><p><span>Tn5它也可以在核小体表面暴露的 DNA 上进行和核小体之间的链接区域切割。这可能会产生较小的 DNA 片段。</span></p></li><li><p><span>这些</span><strong><span>小DNA 片段可能不是背景</span></strong><span>。</span></p></li></ul></li></ul><p>&nbsp;</p><h2 id='extract-fragment-length-and-count'><span>Extract fragment length and count</span></h2><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># -F 0x04参数过滤掉未比对的read。</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 0x04 表示未比对的标记</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">samtools view <span class="cm-attribute">-F</span> 0x04 &lt;in.bam/sam/cram&gt;| \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 使用 awk 工具处理输出，以制表符（\t）为字段分隔符</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 定义了一个 abs 函数：提取第9列（fragment length），并取其绝对值</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-builtin">awk</span> <span class="cm-attribute">-F</span><span class="cm-string">'\t'</span> <span class="cm-string">'function abs(x){return ((x &lt; 0.0) ? -x : x)} {print abs($9)}'</span>| \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 将片段长度的输出排序，然后使用 uniq -c 统计每个fragment出现的次数</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-builtin">sort</span> | uniq <span class="cm-attribute">-c</span> | \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 再次使用 awk，将输出格式化为制表符分隔的格式。$2 是fragment length</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># $1 是该fragment出现的次数。这里将fragment次数除以2得到 fragment count</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-builtin">awk</span> <span class="cm-attribute">-v</span> <span class="cm-def">OFS</span><span class="cm-operator">=</span><span class="cm-string">"\t"</span> <span class="cm-string">'{print $2, $1/2}'</span> &gt; fragmentLen.txt</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 372px;"></div><div class="CodeMirror-gutters" style="display: none; height: 372px;"></div></div></div></pre><p>&nbsp;</p><h2 id='draw-picture'><span>Draw picture</span></h2><h3 id='python'><span>python</span></h3><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="python" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="python"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">import</span> <span class="cm-variable">pandas</span> <span class="cm-keyword">as</span> <span class="cm-variable">pd</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">import</span> <span class="cm-variable">matplotlib</span>.<span class="cm-property">pyplot</span> <span class="cm-keyword">as</span> <span class="cm-variable">plt</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">from</span> <span class="cm-variable">matplotlib</span>.<span class="cm-property">font_manager</span> <span class="cm-keyword">import</span> <span class="cm-variable">FontProperties</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 设置字体属性（假设你使用的是 Windows 系统）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">font</span> <span class="cm-operator">=</span> <span class="cm-variable">FontProperties</span>(<span class="cm-variable">fname</span><span class="cm-operator">=</span><span class="cm-string">'C:/Windows/Fonts/simhei.ttf'</span>) &nbsp;<span class="cm-comment"># 黑体</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 读取数据</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">df</span> <span class="cm-operator">=</span> <span class="cm-variable">pd</span>.<span class="cm-property">read_csv</span>(<span class="cm-string">'fragmentLen.txt'</span>, <span class="cm-variable">sep</span><span class="cm-operator">=</span><span class="cm-string">'\t'</span>, <span class="cm-variable">header</span><span class="cm-operator">=</span><span class="cm-keyword">None</span>, <span class="cm-variable">names</span><span class="cm-operator">=</span>[<span class="cm-string">'Fragment_Length'</span>, <span class="cm-string">'Frequency'</span>])</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 使用数据频数扩展为实际数据点</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">data</span> <span class="cm-operator">=</span> <span class="cm-variable">df</span>.<span class="cm-property">loc</span>[<span class="cm-variable">df</span>.<span class="cm-property">index</span>.<span class="cm-property">repeat</span>(<span class="cm-variable">df</span>[<span class="cm-string">'Frequency'</span>])][<span class="cm-string">'Fragment_Length'</span>]</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 创建子图</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">fig</span>, <span class="cm-variable">axs</span> <span class="cm-operator">=</span> <span class="cm-variable">plt</span>.<span class="cm-property">subplots</span>(<span class="cm-number">1</span>, <span class="cm-number">2</span>, <span class="cm-variable">figsize</span><span class="cm-operator">=</span>(<span class="cm-number">12</span>, <span class="cm-number">4</span>)) &nbsp;<span class="cm-comment"># 1行 2列的子图</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 绘制片段长度频数图</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">axs</span>[<span class="cm-number">0</span>].<span class="cm-property">bar</span>(<span class="cm-variable">df</span>[<span class="cm-string">'Fragment_Length'</span>], <span class="cm-variable">df</span>[<span class="cm-string">'Frequency'</span>], <span class="cm-variable">color</span><span class="cm-operator">=</span><span class="cm-string">'skyblue'</span>, <span class="cm-variable">width</span><span class="cm-operator">=</span><span class="cm-number">0.8</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">axs</span>[<span class="cm-number">0</span>].<span class="cm-property">set_xlabel</span>(<span class="cm-string">'fragment length'</span>, <span class="cm-variable">fontproperties</span><span class="cm-operator">=</span><span class="cm-variable">font</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">axs</span>[<span class="cm-number">0</span>].<span class="cm-property">set_ylabel</span>(<span class="cm-string">'fragment Count'</span>, <span class="cm-variable">fontproperties</span><span class="cm-operator">=</span><span class="cm-variable">font</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">axs</span>[<span class="cm-number">0</span>].<span class="cm-property">set_title</span>(<span class="cm-string">'Frequency chart of fragment length'</span>, <span class="cm-variable">fontproperties</span><span class="cm-operator">=</span><span class="cm-variable">font</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">axs</span>[<span class="cm-number">0</span>].<span class="cm-property">grid</span>(<span class="cm-keyword">True</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 缩窄 x 和 y 轴的范围</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">axs</span>[<span class="cm-number">0</span>].<span class="cm-property">set_xlim</span>(<span class="cm-variable">df</span>[<span class="cm-string">'Fragment_Length'</span>].<span class="cm-property">min</span>() <span class="cm-operator">-</span> <span class="cm-number">10</span>, <span class="cm-variable">df</span>[<span class="cm-string">'Fragment_Length'</span>].<span class="cm-property">max</span>() <span class="cm-operator">+</span> <span class="cm-number">10</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">axs</span>[<span class="cm-number">0</span>].<span class="cm-property">set_ylim</span>(<span class="cm-number">0</span>, <span class="cm-variable">df</span>[<span class="cm-string">'Frequency'</span>].<span class="cm-property">max</span>() <span class="cm-operator">*</span> <span class="cm-number">1.1</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 绘制频数分布图（直方图）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">axs</span>[<span class="cm-number">1</span>].<span class="cm-property">hist</span>(<span class="cm-variable">data</span>, <span class="cm-variable">bins</span><span class="cm-operator">=</span><span class="cm-number">30</span>, <span class="cm-variable">color</span><span class="cm-operator">=</span><span class="cm-string">'skyblue'</span>, <span class="cm-variable">edgecolor</span><span class="cm-operator">=</span><span class="cm-string">'black'</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">axs</span>[<span class="cm-number">1</span>].<span class="cm-property">set_xlabel</span>(<span class="cm-string">'fragment length'</span>, <span class="cm-variable">fontproperties</span><span class="cm-operator">=</span><span class="cm-variable">font</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">axs</span>[<span class="cm-number">1</span>].<span class="cm-property">set_ylabel</span>(<span class="cm-string">'fragment Count'</span>, <span class="cm-variable">fontproperties</span><span class="cm-operator">=</span><span class="cm-variable">font</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">axs</span>[<span class="cm-number">1</span>].<span class="cm-property">set_title</span>(<span class="cm-string">'Frequency distribution chart of fragment length'</span>, <span class="cm-variable">fontproperties</span><span class="cm-operator">=</span><span class="cm-variable">font</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">axs</span>[<span class="cm-number">1</span>].<span class="cm-property">grid</span>(<span class="cm-keyword">True</span>) <span class="cm-comment"># 启用网格</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 调整子图布局</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">plt</span>.<span class="cm-property">tight_layout</span>()</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">plt</span>.<span class="cm-property">show</span>()</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 983px;"></div><div class="CodeMirror-gutters" style="display: none; height: 983px;"></div></div></div></pre><p>&nbsp;</p><hr /><h1 id='alignment-filtering'><span>Alignment filtering</span></h1><h2 id='alignment-filtering-with-mapq-optinal'><span>Alignment filtering with MAPQ (optinal)</span></h2><h3 id='filtering-with-mapq'><span>Filtering with MAPQ</span></h3><p><strong><span>Some projects may require more stringent filtering on the alignment quality score</span></strong></p><ul><li><p><strong><span>MAPQ（Mapping Quality）</span></strong><span>是衡量 allignment 质量的分数，表示某个比对结果被错误映射的概率。计算公式为：</span></p><div contenteditable="false" spellcheck="false" class="mathjax-block md-end-block md-math-block md-rawblock" id="mathjax-n2608" cid="n2608" mdtype="math_block" data-math-tag-before="0" data-math-tag-after="0" data-math-labels="[]"><div class="md-math-container"><mjx-container class="MathJax" jax="SVG" display="true" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="68.251ex" height="2.278ex" role="img" focusable="false" viewBox="0 -750 30166.9 1006.9" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -0.581ex;"><defs><path id="MJX-6-TEX-I-1D440" d="M289 629Q289 635 232 637Q208 637 201 638T194 648Q194 649 196 659Q197 662 198 666T199 671T201 676T203 679T207 681T212 683T220 683T232 684Q238 684 262 684T307 683Q386 683 398 683T414 678Q415 674 451 396L487 117L510 154Q534 190 574 254T662 394Q837 673 839 675Q840 676 842 678T846 681L852 683H948Q965 683 988 683T1017 684Q1051 684 1051 673Q1051 668 1048 656T1045 643Q1041 637 1008 637Q968 636 957 634T939 623Q936 618 867 340T797 59Q797 55 798 54T805 50T822 48T855 46H886Q892 37 892 35Q892 19 885 5Q880 0 869 0Q864 0 828 1T736 2Q675 2 644 2T609 1Q592 1 592 11Q592 13 594 25Q598 41 602 43T625 46Q652 46 685 49Q699 52 704 61Q706 65 742 207T813 490T848 631L654 322Q458 10 453 5Q451 4 449 3Q444 0 433 0Q418 0 415 7Q413 11 374 317L335 624L267 354Q200 88 200 79Q206 46 272 46H282Q288 41 289 37T286 19Q282 3 278 1Q274 0 267 0Q265 0 255 0T221 1T157 2Q127 2 95 1T58 0Q43 0 39 2T35 11Q35 13 38 25T43 40Q45 46 65 46Q135 46 154 86Q158 92 223 354T289 629Z"></path><path id="MJX-6-TEX-I-1D434" d="M208 74Q208 50 254 46Q272 46 272 35Q272 34 270 22Q267 8 264 4T251 0Q249 0 239 0T205 1T141 2Q70 2 50 0H42Q35 7 35 11Q37 38 48 46H62Q132 49 164 96Q170 102 345 401T523 704Q530 716 547 716H555H572Q578 707 578 706L606 383Q634 60 636 57Q641 46 701 46Q726 46 726 36Q726 34 723 22Q720 7 718 4T704 0Q701 0 690 0T651 1T578 2Q484 2 455 0H443Q437 6 437 9T439 27Q443 40 445 43L449 46H469Q523 49 533 63L521 213H283L249 155Q208 86 208 74ZM516 260Q516 271 504 416T490 562L463 519Q447 492 400 412L310 260L413 259Q516 259 516 260Z"></path><path id="MJX-6-TEX-I-1D443" d="M287 628Q287 635 230 637Q206 637 199 638T192 648Q192 649 194 659Q200 679 203 681T397 683Q587 682 600 680Q664 669 707 631T751 530Q751 453 685 389Q616 321 507 303Q500 302 402 301H307L277 182Q247 66 247 59Q247 55 248 54T255 50T272 48T305 46H336Q342 37 342 35Q342 19 335 5Q330 0 319 0Q316 0 282 1T182 2Q120 2 87 2T51 1Q33 1 33 11Q33 13 36 25Q40 41 44 43T67 46Q94 46 127 49Q141 52 146 61Q149 65 218 339T287 628ZM645 554Q645 567 643 575T634 597T609 619T560 635Q553 636 480 637Q463 637 445 637T416 636T404 636Q391 635 386 627Q384 621 367 550T332 412T314 344Q314 342 395 342H407H430Q542 342 590 392Q617 419 631 471T645 554Z"></path><path id="MJX-6-TEX-I-1D444" d="M399 -80Q399 -47 400 -30T402 -11V-7L387 -11Q341 -22 303 -22Q208 -22 138 35T51 201Q50 209 50 244Q50 346 98 438T227 601Q351 704 476 704Q514 704 524 703Q621 689 680 617T740 435Q740 255 592 107Q529 47 461 16L444 8V3Q444 2 449 -24T470 -66T516 -82Q551 -82 583 -60T625 -3Q631 11 638 11Q647 11 649 2Q649 -6 639 -34T611 -100T557 -165T481 -194Q399 -194 399 -87V-80ZM636 468Q636 523 621 564T580 625T530 655T477 665Q429 665 379 640Q277 591 215 464T153 216Q153 110 207 59Q231 38 236 38V46Q236 86 269 120T347 155Q372 155 390 144T417 114T429 82T435 55L448 64Q512 108 557 185T619 334T636 468ZM314 18Q362 18 404 39L403 49Q399 104 366 115Q354 117 347 117Q344 117 341 117T337 118Q317 118 296 98T274 52Q274 18 314 18Z"></path><path id="MJX-6-TEX-N-28" d="M94 250Q94 319 104 381T127 488T164 576T202 643T244 695T277 729T302 750H315H319Q333 750 333 741Q333 738 316 720T275 667T226 581T184 443T167 250T184 58T225 -81T274 -167T316 -220T333 -241Q333 -250 318 -250H315H302L274 -226Q180 -141 137 -14T94 250Z"></path><path id="MJX-6-TEX-I-1D465" d="M52 289Q59 331 106 386T222 442Q257 442 286 424T329 379Q371 442 430 442Q467 442 494 420T522 361Q522 332 508 314T481 292T458 288Q439 288 427 299T415 328Q415 374 465 391Q454 404 425 404Q412 404 406 402Q368 386 350 336Q290 115 290 78Q290 50 306 38T341 26Q378 26 414 59T463 140Q466 150 469 151T485 153H489Q504 153 504 145Q504 144 502 134Q486 77 440 33T333 -11Q263 -11 227 52Q186 -10 133 -10H127Q78 -10 57 16T35 71Q35 103 54 123T99 143Q142 143 142 101Q142 81 130 66T107 46T94 41L91 40Q91 39 97 36T113 29T132 26Q168 26 194 71Q203 87 217 139T245 247T261 313Q266 340 266 352Q266 380 251 392T217 404Q177 404 142 372T93 290Q91 281 88 280T72 278H58Q52 284 52 289Z"></path><path id="MJX-6-TEX-N-29" d="M60 749L64 750Q69 750 74 750H86L114 726Q208 641 251 514T294 250Q294 182 284 119T261 12T224 -76T186 -143T145 -194T113 -227T90 -246Q87 -249 86 -250H74Q66 -250 63 -250T58 -247T55 -238Q56 -237 66 -225Q221 -64 221 250T66 725Q56 737 55 738Q55 746 60 749Z"></path><path id="MJX-6-TEX-N-3D" d="M56 347Q56 360 70 367H707Q722 359 722 347Q722 336 708 328L390 327H72Q56 332 56 347ZM56 153Q56 168 72 173H708Q722 163 722 153Q722 140 707 133H70Q56 140 56 153Z"></path><path id="MJX-6-TEX-N-2212" d="M84 237T84 250T98 270H679Q694 262 694 250T679 230H98Q84 237 84 250Z"></path><path id="MJX-6-TEX-N-31" d="M213 578L200 573Q186 568 160 563T102 556H83V602H102Q149 604 189 617T245 641T273 663Q275 666 285 666Q294 666 302 660V361L303 61Q310 54 315 52T339 48T401 46H427V0H416Q395 3 257 3Q121 3 100 0H88V46H114Q136 46 152 46T177 47T193 50T201 52T207 57T213 61V578Z"></path><path id="MJX-6-TEX-N-30" d="M96 585Q152 666 249 666Q297 666 345 640T423 548Q460 465 460 320Q460 165 417 83Q397 41 362 16T301 -15T250 -22Q224 -22 198 -16T137 16T82 83Q39 165 39 320Q39 494 96 585ZM321 597Q291 629 250 629Q208 629 178 597Q153 571 145 525T137 333Q137 175 145 125T181 46Q209 16 250 16Q290 16 318 46Q347 76 354 130T362 333Q362 478 354 524T321 597Z"></path><path id="MJX-6-TEX-N-D7" d="M630 29Q630 9 609 9Q604 9 587 25T493 118L389 222L284 117Q178 13 175 11Q171 9 168 9Q160 9 154 15T147 29Q147 36 161 51T255 146L359 250L255 354Q174 435 161 449T147 471Q147 480 153 485T168 490Q173 490 175 489Q178 487 284 383L389 278L493 382Q570 459 587 475T609 491Q630 491 630 471Q630 464 620 453T522 355L418 250L522 145Q606 61 618 48T630 29Z"></path><path id="MJX-6-TEX-N-6C" d="M42 46H56Q95 46 103 60V68Q103 77 103 91T103 124T104 167T104 217T104 272T104 329Q104 366 104 407T104 482T104 542T103 586T103 603Q100 622 89 628T44 637H26V660Q26 683 28 683L38 684Q48 685 67 686T104 688Q121 689 141 690T171 693T182 694H185V379Q185 62 186 60Q190 52 198 49Q219 46 247 46H263V0H255L232 1Q209 2 183 2T145 3T107 3T57 1L34 0H26V46H42Z"></path><path id="MJX-6-TEX-N-6F" d="M28 214Q28 309 93 378T250 448Q340 448 405 380T471 215Q471 120 407 55T250 -10Q153 -10 91 57T28 214ZM250 30Q372 30 372 193V225V250Q372 272 371 288T364 326T348 362T317 390T268 410Q263 411 252 411Q222 411 195 399Q152 377 139 338T126 246V226Q126 130 145 91Q177 30 250 30Z"></path><path id="MJX-6-TEX-N-67" d="M329 409Q373 453 429 453Q459 453 472 434T485 396Q485 382 476 371T449 360Q416 360 412 390Q410 404 415 411Q415 412 416 414V415Q388 412 363 393Q355 388 355 386Q355 385 359 381T368 369T379 351T388 325T392 292Q392 230 343 187T222 143Q172 143 123 171Q112 153 112 133Q112 98 138 81Q147 75 155 75T227 73Q311 72 335 67Q396 58 431 26Q470 -13 470 -72Q470 -139 392 -175Q332 -206 250 -206Q167 -206 107 -175Q29 -140 29 -75Q29 -39 50 -15T92 18L103 24Q67 55 67 108Q67 155 96 193Q52 237 52 292Q52 355 102 398T223 442Q274 442 318 416L329 409ZM299 343Q294 371 273 387T221 404Q192 404 171 388T145 343Q142 326 142 292Q142 248 149 227T179 192Q196 182 222 182Q244 182 260 189T283 207T294 227T299 242Q302 258 302 292T299 343ZM403 -75Q403 -50 389 -34T348 -11T299 -2T245 0H218Q151 0 138 -6Q118 -15 107 -34T95 -74Q95 -84 101 -97T122 -127T170 -155T250 -167Q319 -167 361 -139T403 -75Z"></path><path id="MJX-6-TEX-N-2061" d=""></path><path id="MJX-6-TEX-N-A0" d=""></path><path id="MJX-6-TEX-N-69" d="M69 609Q69 637 87 653T131 669Q154 667 171 652T188 609Q188 579 171 564T129 549Q104 549 87 564T69 609ZM247 0Q232 3 143 3Q132 3 106 3T56 1L34 0H26V46H42Q70 46 91 49Q100 53 102 60T104 102V205V293Q104 345 102 359T88 378Q74 385 41 385H30V408Q30 431 32 431L42 432Q52 433 70 434T106 436Q123 437 142 438T171 441T182 442H185V62Q190 52 197 50T232 46H255V0H247Z"></path><path id="MJX-6-TEX-N-73" d="M295 316Q295 356 268 385T190 414Q154 414 128 401Q98 382 98 349Q97 344 98 336T114 312T157 287Q175 282 201 278T245 269T277 256Q294 248 310 236T342 195T359 133Q359 71 321 31T198 -10H190Q138 -10 94 26L86 19L77 10Q71 4 65 -1L54 -11H46H42Q39 -11 33 -5V74V132Q33 153 35 157T45 162H54Q66 162 70 158T75 146T82 119T101 77Q136 26 198 26Q295 26 295 104Q295 133 277 151Q257 175 194 187T111 210Q75 227 54 256T33 318Q33 357 50 384T93 424T143 442T187 447H198Q238 447 268 432L283 424L292 431Q302 440 314 448H322H326Q329 448 335 442V310L329 304H301Q295 310 295 316Z"></path><path id="MJX-6-TEX-N-20" d=""></path><path id="MJX-6-TEX-N-6D" d="M41 46H55Q94 46 102 60V68Q102 77 102 91T102 122T103 161T103 203Q103 234 103 269T102 328V351Q99 370 88 376T43 385H25V408Q25 431 27 431L37 432Q47 433 65 434T102 436Q119 437 138 438T167 441T178 442H181V402Q181 364 182 364T187 369T199 384T218 402T247 421T285 437Q305 442 336 442Q351 442 364 440T387 434T406 426T421 417T432 406T441 395T448 384T452 374T455 366L457 361L460 365Q463 369 466 373T475 384T488 397T503 410T523 422T546 432T572 439T603 442Q729 442 740 329Q741 322 741 190V104Q741 66 743 59T754 49Q775 46 803 46H819V0H811L788 1Q764 2 737 2T699 3Q596 3 587 0H579V46H595Q656 46 656 62Q657 64 657 200Q656 335 655 343Q649 371 635 385T611 402T585 404Q540 404 506 370Q479 343 472 315T464 232V168V108Q464 78 465 68T468 55T477 49Q498 46 526 46H542V0H534L510 1Q487 2 460 2T422 3Q319 3 310 0H302V46H318Q379 46 379 62Q380 64 380 200Q379 335 378 343Q372 371 358 385T334 402T308 404Q263 404 229 370Q202 343 195 315T187 232V168V108Q187 78 188 68T191 55T200 49Q221 46 249 46H265V0H257L234 1Q210 2 183 2T145 3Q42 3 33 0H25V46H41Z"></path><path id="MJX-6-TEX-N-61" d="M137 305T115 305T78 320T63 359Q63 394 97 421T218 448Q291 448 336 416T396 340Q401 326 401 309T402 194V124Q402 76 407 58T428 40Q443 40 448 56T453 109V145H493V106Q492 66 490 59Q481 29 455 12T400 -6T353 12T329 54V58L327 55Q325 52 322 49T314 40T302 29T287 17T269 6T247 -2T221 -8T190 -11Q130 -11 82 20T34 107Q34 128 41 147T68 188T116 225T194 253T304 268H318V290Q318 324 312 340Q290 411 215 411Q197 411 181 410T156 406T148 403Q170 388 170 359Q170 334 154 320ZM126 106Q126 75 150 51T209 26Q247 26 276 49T315 109Q317 116 318 175Q318 233 317 233Q309 233 296 232T251 223T193 203T147 166T126 106Z"></path><path id="MJX-6-TEX-N-70" d="M36 -148H50Q89 -148 97 -134V-126Q97 -119 97 -107T97 -77T98 -38T98 6T98 55T98 106Q98 140 98 177T98 243T98 296T97 335T97 351Q94 370 83 376T38 385H20V408Q20 431 22 431L32 432Q42 433 61 434T98 436Q115 437 135 438T165 441T176 442H179V416L180 390L188 397Q247 441 326 441Q407 441 464 377T522 216Q522 115 457 52T310 -11Q242 -11 190 33L182 40V-45V-101Q182 -128 184 -134T195 -145Q216 -148 244 -148H260V-194H252L228 -193Q205 -192 178 -192T140 -191Q37 -191 28 -194H20V-148H36ZM424 218Q424 292 390 347T305 402Q234 402 182 337V98Q222 26 294 26Q345 26 384 80T424 218Z"></path><path id="MJX-6-TEX-N-65" d="M28 218Q28 273 48 318T98 391T163 433T229 448Q282 448 320 430T378 380T406 316T415 245Q415 238 408 231H126V216Q126 68 226 36Q246 30 270 30Q312 30 342 62Q359 79 369 104L379 128Q382 131 395 131H398Q415 131 415 121Q415 117 412 108Q393 53 349 21T250 -11Q155 -11 92 58T28 218ZM333 275Q322 403 238 411H236Q228 411 220 410T195 402T166 381T143 340T127 274V267H333V275Z"></path><path id="MJX-6-TEX-N-64" d="M376 495Q376 511 376 535T377 568Q377 613 367 624T316 637H298V660Q298 683 300 683L310 684Q320 685 339 686T376 688Q393 689 413 690T443 693T454 694H457V390Q457 84 458 81Q461 61 472 55T517 46H535V0Q533 0 459 -5T380 -11H373V44L365 37Q307 -11 235 -11Q158 -11 96 50T34 215Q34 315 97 378T244 442Q319 442 376 393V495ZM373 342Q328 405 260 405Q211 405 173 369Q146 341 139 305T131 211Q131 155 138 120T173 59Q203 26 251 26Q322 26 373 103V342Z"></path><path id="MJX-6-TEX-N-77" d="M90 368Q84 378 76 380T40 385H18V431H24L43 430Q62 430 84 429T116 428Q206 428 221 431H229V385H215Q177 383 177 368Q177 367 221 239L265 113L339 328L333 345Q323 374 316 379Q308 384 278 385H258V431H264Q270 428 348 428Q439 428 454 431H461V385H452Q404 385 404 369Q404 366 418 324T449 234T481 143L496 100L537 219Q579 341 579 347Q579 363 564 373T530 385H522V431H529Q541 428 624 428Q692 428 698 431H703V385H697Q696 385 691 385T682 384Q635 377 619 334L559 161Q546 124 528 71Q508 12 503 1T487 -11H479Q460 -11 456 -4Q455 -3 407 133L361 267Q359 263 266 -4Q261 -11 243 -11H238Q225 -11 220 -3L90 368Z"></path><path id="MJX-6-TEX-N-72" d="M36 46H50Q89 46 97 60V68Q97 77 97 91T98 122T98 161T98 203Q98 234 98 269T98 328L97 351Q94 370 83 376T38 385H20V408Q20 431 22 431L32 432Q42 433 60 434T96 436Q112 437 131 438T160 441T171 442H174V373Q213 441 271 441H277Q322 441 343 419T364 373Q364 352 351 337T313 322Q288 322 276 338T263 372Q263 381 265 388T270 400T273 405Q271 407 250 401Q234 393 226 386Q179 341 179 207V154Q179 141 179 127T179 101T180 81T180 66V61Q181 59 183 57T188 54T193 51T200 49T207 48T216 47T225 47T235 46T245 46H276V0H267Q249 3 140 3Q37 3 28 0H20V46H36Z"></path><path id="MJX-6-TEX-N-6E" d="M41 46H55Q94 46 102 60V68Q102 77 102 91T102 122T103 161T103 203Q103 234 103 269T102 328V351Q99 370 88 376T43 385H25V408Q25 431 27 431L37 432Q47 433 65 434T102 436Q119 437 138 438T167 441T178 442H181V402Q181 364 182 364T187 369T199 384T218 402T247 421T285 437Q305 442 336 442Q450 438 463 329Q464 322 464 190V104Q464 66 466 59T477 49Q498 46 526 46H542V0H534L510 1Q487 2 460 2T422 3Q319 3 310 0H302V46H318Q379 46 379 62Q380 64 380 200Q379 335 378 343Q372 371 358 385T334 402T308 404Q263 404 229 370Q202 343 195 315T187 232V168V108Q187 78 188 68T191 55T200 49Q221 46 249 46H265V0H257L234 1Q210 2 183 2T145 3Q42 3 33 0H25V46H41Z"></path><path id="MJX-6-TEX-N-79" d="M69 -66Q91 -66 104 -80T118 -116Q118 -134 109 -145T91 -160Q84 -163 97 -166Q104 -168 111 -168Q131 -168 148 -159T175 -138T197 -106T213 -75T225 -43L242 0L170 183Q150 233 125 297Q101 358 96 368T80 381Q79 382 78 382Q66 385 34 385H19V431H26L46 430Q65 430 88 429T122 428Q129 428 142 428T171 429T200 430T224 430L233 431H241V385H232Q183 385 185 366L286 112Q286 113 332 227L376 341V350Q376 365 366 373T348 383T334 385H331V431H337H344Q351 431 361 431T382 430T405 429T422 429Q477 429 503 431H508V385H497Q441 380 422 345Q420 343 378 235T289 9T227 -131Q180 -204 113 -204Q69 -204 44 -177T19 -116Q19 -89 35 -78T69 -66Z"></path><path id="MJX-6-TEX-I-1D45D" d="M23 287Q24 290 25 295T30 317T40 348T55 381T75 411T101 433T134 442Q209 442 230 378L240 387Q302 442 358 442Q423 442 460 395T497 281Q497 173 421 82T249 -10Q227 -10 210 -4Q199 1 187 11T168 28L161 36Q160 35 139 -51T118 -138Q118 -144 126 -145T163 -148H188Q194 -155 194 -157T191 -175Q188 -187 185 -190T172 -194Q170 -194 161 -194T127 -193T65 -192Q-5 -192 -24 -194H-32Q-39 -187 -39 -183Q-37 -156 -26 -148H-6Q28 -147 33 -136Q36 -130 94 103T155 350Q156 355 156 364Q156 405 131 405Q109 405 94 377T71 316T59 280Q57 278 43 278H29Q23 284 23 287ZM178 102Q200 26 252 26Q282 26 310 49T356 107Q374 141 392 215T411 325V331Q411 405 350 405Q339 405 328 402T306 393T286 380T269 365T254 350T243 336T235 326L232 322Q232 321 229 308T218 264T204 212Q178 106 178 102Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="mi"><use data-c="1D440" xlink:href="#MJX-6-TEX-I-1D440"></use></g><g data-mml-node="mi" transform="translate(1051,0)"><use data-c="1D434" xlink:href="#MJX-6-TEX-I-1D434"></use></g><g data-mml-node="mi" transform="translate(1801,0)"><use data-c="1D443" xlink:href="#MJX-6-TEX-I-1D443"></use></g><g data-mml-node="mi" transform="translate(2552,0)"><use data-c="1D444" xlink:href="#MJX-6-TEX-I-1D444"></use></g><g data-mml-node="mo" transform="translate(3343,0)"><use data-c="28" xlink:href="#MJX-6-TEX-N-28"></use></g><g data-mml-node="mi" transform="translate(3732,0)"><use data-c="1D465" xlink:href="#MJX-6-TEX-I-1D465"></use></g><g data-mml-node="mo" transform="translate(4304,0)"><use data-c="29" xlink:href="#MJX-6-TEX-N-29"></use></g><g data-mml-node="mo" transform="translate(4970.8,0)"><use data-c="3D" xlink:href="#MJX-6-TEX-N-3D"></use></g><g data-mml-node="mo" transform="translate(6026.6,0)"><use data-c="2212" xlink:href="#MJX-6-TEX-N-2212"></use></g><g data-mml-node="mn" transform="translate(6804.6,0)"><use data-c="31" xlink:href="#MJX-6-TEX-N-31"></use><use data-c="30" xlink:href="#MJX-6-TEX-N-30" transform="translate(500,0)"></use></g><g data-mml-node="mo" transform="translate(8026.8,0)"><use data-c="D7" xlink:href="#MJX-6-TEX-N-D7"></use></g><g data-mml-node="msub" transform="translate(9027,0)"><g data-mml-node="mi"><use data-c="6C" xlink:href="#MJX-6-TEX-N-6C"></use><use data-c="6F" xlink:href="#MJX-6-TEX-N-6F" transform="translate(278,0)"></use><use data-c="67" xlink:href="#MJX-6-TEX-N-67" transform="translate(778,0)"></use></g><g data-mml-node="TeXAtom" transform="translate(1311,-241.4) scale(0.707)" data-mjx-texclass="ORD"><g data-mml-node="mn"><use data-c="31" xlink:href="#MJX-6-TEX-N-31"></use><use data-c="30" xlink:href="#MJX-6-TEX-N-30" transform="translate(500,0)"></use></g></g></g><g data-mml-node="mo" transform="translate(11095.1,0)"><use data-c="2061" xlink:href="#MJX-6-TEX-N-2061"></use></g><g data-mml-node="mrow" transform="translate(11261.8,0)"><g data-mml-node="mo"><use data-c="28" xlink:href="#MJX-6-TEX-N-28"></use></g><g data-mml-node="mi" transform="translate(389,0)"><use data-c="1D443" xlink:href="#MJX-6-TEX-I-1D443"></use></g><g data-mml-node="mo" transform="translate(1140,0)"><use data-c="28" xlink:href="#MJX-6-TEX-N-28"></use></g><g data-mml-node="mi" transform="translate(1529,0)"><use data-c="1D465" xlink:href="#MJX-6-TEX-I-1D465"></use></g><g data-mml-node="mtext" transform="translate(2101,0)"><use data-c="A0" xlink:href="#MJX-6-TEX-N-A0"></use><use data-c="69" xlink:href="#MJX-6-TEX-N-69" transform="translate(250,0)"></use><use data-c="73" xlink:href="#MJX-6-TEX-N-73" transform="translate(528,0)"></use><use data-c="20" xlink:href="#MJX-6-TEX-N-20" transform="translate(922,0)"></use><use data-c="6D" xlink:href="#MJX-6-TEX-N-6D" transform="translate(1172,0)"></use><use data-c="61" xlink:href="#MJX-6-TEX-N-61" transform="translate(2005,0)"></use><use data-c="70" xlink:href="#MJX-6-TEX-N-70" transform="translate(2505,0)"></use><use data-c="70" xlink:href="#MJX-6-TEX-N-70" transform="translate(3061,0)"></use><use data-c="65" xlink:href="#MJX-6-TEX-N-65" transform="translate(3617,0)"></use><use data-c="64" xlink:href="#MJX-6-TEX-N-64" transform="translate(4061,0)"></use><use data-c="20" xlink:href="#MJX-6-TEX-N-20" transform="translate(4617,0)"></use><use data-c="77" xlink:href="#MJX-6-TEX-N-77" transform="translate(4867,0)"></use><use data-c="72" xlink:href="#MJX-6-TEX-N-72" transform="translate(5589,0)"></use><use data-c="6F" xlink:href="#MJX-6-TEX-N-6F" transform="translate(5981,0)"></use><use data-c="6E" xlink:href="#MJX-6-TEX-N-6E" transform="translate(6481,0)"></use><use data-c="67" xlink:href="#MJX-6-TEX-N-67" transform="translate(7037,0)"></use><use data-c="6C" xlink:href="#MJX-6-TEX-N-6C" transform="translate(7537,0)"></use><use data-c="79" xlink:href="#MJX-6-TEX-N-79" transform="translate(7815,0)"></use></g><g data-mml-node="mo" transform="translate(10444,0)"><use data-c="29" xlink:href="#MJX-6-TEX-N-29"></use></g><g data-mml-node="mo" transform="translate(10833,0)"><use data-c="29" xlink:href="#MJX-6-TEX-N-29"></use></g></g><g data-mml-node="mo" transform="translate(22761.6,0)"><use data-c="3D" xlink:href="#MJX-6-TEX-N-3D"></use></g><g data-mml-node="mo" transform="translate(23817.3,0)"><use data-c="2212" xlink:href="#MJX-6-TEX-N-2212"></use></g><g data-mml-node="mn" transform="translate(24595.3,0)"><use data-c="31" xlink:href="#MJX-6-TEX-N-31"></use><use data-c="30" xlink:href="#MJX-6-TEX-N-30" transform="translate(500,0)"></use></g><g data-mml-node="mo" transform="translate(25817.6,0)"><use data-c="D7" xlink:href="#MJX-6-TEX-N-D7"></use></g><g data-mml-node="msub" transform="translate(26817.8,0)"><g data-mml-node="mi"><use data-c="6C" xlink:href="#MJX-6-TEX-N-6C"></use><use data-c="6F" xlink:href="#MJX-6-TEX-N-6F" transform="translate(278,0)"></use><use data-c="67" xlink:href="#MJX-6-TEX-N-67" transform="translate(778,0)"></use></g><g data-mml-node="TeXAtom" transform="translate(1311,-241.4) scale(0.707)" data-mjx-texclass="ORD"><g data-mml-node="mn"><use data-c="31" xlink:href="#MJX-6-TEX-N-31"></use><use data-c="30" xlink:href="#MJX-6-TEX-N-30" transform="translate(500,0)"></use></g></g></g><g data-mml-node="mo" transform="translate(28885.9,0)"><use data-c="2061" xlink:href="#MJX-6-TEX-N-2061"></use></g><g data-mml-node="mo" transform="translate(28885.9,0)"><use data-c="28" xlink:href="#MJX-6-TEX-N-28"></use></g><g data-mml-node="mi" transform="translate(29274.9,0)"><use data-c="1D45D" xlink:href="#MJX-6-TEX-I-1D45D"></use></g><g data-mml-node="mo" transform="translate(29777.9,0)"><use data-c="29" xlink:href="#MJX-6-TEX-N-29"></use></g></g></g></svg><mjx-assistive-mml unselectable="on" display="block"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><mi>M</mi><mi>A</mi><mi>P</mi><mi>Q</mi><mo stretchy="false">(</mo><mi>x</mi><mo stretchy="false">)</mo><mo>=</mo><mo>−</mo><mn>10</mn><mo>×</mo><msub><mi>log</mi><mrow data-mjx-texclass="ORD"><mn>10</mn></mrow></msub><mo data-mjx-texclass="NONE">⁡</mo><mrow data-mjx-texclass="INNER"><mo data-mjx-texclass="OPEN">(</mo><mi>P</mi><mo stretchy="false">(</mo><mi>x</mi><mtext>&nbsp;is mapped wrongly</mtext><mo stretchy="false">)</mo><mo data-mjx-texclass="CLOSE">)</mo></mrow><mo>=</mo><mo>−</mo><mn>10</mn><mo>×</mo><msub><mi>log</mi><mrow data-mjx-texclass="ORD"><mn>10</mn></mrow></msub><mo data-mjx-texclass="NONE">⁡</mo><mo stretchy="false">(</mo><mi>p</mi><mo stretchy="false">)</mo></math></mjx-assistive-mml></mjx-container></div></div><p><span>MAPQ 值通常范围在 0 到 37、40 或 42 之间，这取决于所使用的比对工具。值越高，表示比对的置信度越高</span></p></li></ul><p>&nbsp;</p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation"><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">samtools view <span class="cm-attribute">-q</span> &lt;int&gt; &lt;before.bam/sam/cram&gt; <span class="cm-attribute">-o</span> &lt;after.bam/sam/cram&gt;</span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># int为minQualityScore</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 推荐为2，将比对质量分数小于 2 的比对结果过滤掉</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 66px;"></div><div class="CodeMirror-gutters" style="display: none; height: 66px;"></div></div></div></pre><p>&nbsp;</p><h3 id='view-the-reads-before-and-after-filtering'><span>View the reads before and after filtering</span></h3><ul><li><p><span>创建一个bashrc脚本</span><code>vim filter_contrast.bashrc</code></p></li></ul><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">#在bashrc配置脚本中复制粘贴</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">#定义一个函数</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">filter_contrast() {</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-def">before</span><span class="cm-operator">=</span><span class="cm-def">$1</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-def">after</span><span class="cm-operator">=</span><span class="cm-def">$2</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-def">output_file</span><span class="cm-operator">=</span><span class="cm-def">$3</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-builtin">echo</span> <span class="cm-string">"</span><span class="cm-def">$before</span><span class="cm-string">"</span> &gt;&gt; <span class="cm-string">"</span><span class="cm-def">$output_file</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;  samtools flagstat <span class="cm-string">"</span><span class="cm-def">$before</span><span class="cm-string">"</span> &gt;&gt; <span class="cm-string">"</span><span class="cm-def">$output_file</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-builtin">echo</span> <span class="cm-string">"--------------------"</span> &gt;&gt; <span class="cm-string">"</span><span class="cm-def">$output_file</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-builtin">echo</span> <span class="cm-string">"reads with chromosome:</span><span class="cm-def">$before</span><span class="cm-string">"</span> &gt;&gt; <span class="cm-string">"</span><span class="cm-def">$output_file</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># samtools idxstats用于查看每条染色体的read数 &nbsp; &nbsp;</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;  samtools idxstats <span class="cm-string">"</span><span class="cm-def">$before</span><span class="cm-string">"</span> &gt;&gt; <span class="cm-string">"</span><span class="cm-def">$output_file</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-builtin">echo</span> <span class="cm-string">"--------------------"</span> &gt;&gt; <span class="cm-string">"</span><span class="cm-def">$output_file</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-builtin">echo</span> <span class="cm-string">"</span><span class="cm-def">$after</span><span class="cm-string">"</span> &gt;&gt; <span class="cm-string">"</span><span class="cm-def">$output_file</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;  samtools flagstat <span class="cm-string">"</span><span class="cm-def">$after</span><span class="cm-string">"</span> &gt;&gt; <span class="cm-string">"</span><span class="cm-def">$output_file</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-builtin">echo</span> <span class="cm-string">"--------------------"</span> &gt;&gt; <span class="cm-string">"</span><span class="cm-def">$output_file</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-builtin">echo</span> <span class="cm-string">"reads with chromosome:</span><span class="cm-def">$after</span><span class="cm-string">"</span> &gt;&gt; <span class="cm-string">"</span><span class="cm-def">$output_file</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;  samtools idxstats <span class="cm-string">"</span><span class="cm-def">$after</span><span class="cm-string">"</span> &gt;&gt; <span class="cm-string">"</span><span class="cm-def">$output_file</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-builtin">echo</span> <span class="cm-string">"--------------------"</span> &gt;&gt; <span class="cm-string">"</span><span class="cm-def">$output_file</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">}</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 503px;"></div><div class="CodeMirror-gutters" style="display: none; height: 503px;"></div></div></div></pre><ul><li><p><span>在终端使配置脚本生效 </span><code>source filter_contrast.bashrc</code></p></li></ul><ul><li><p><span>在终端运行函数 </span><code>filter_contrast &lt;before.bam/sam/cram&gt; &lt;after.bam/sam/cram&gt; &lt;output_file.txt&gt;</code></p></li></ul><p>&nbsp;</p><hr /><h2 id='alignment-filtering-with-mapped-read-pairs'><span>Alignment filtering with Mapped Read Pairs</span></h2><p><strong><span>注意： 我们需要过滤掉对未必对的reads。</span></strong></p><ul><li><p><span>否则，在生成的bed文件会出现 </span><code>.</code><strong><span>的染色体条目</span></strong><span>（未比对的reads）。</span></p></li><li><p><span>在将bed文件转换成bedgraph时，因为 </span><code>.</code><strong><span>的染色体条目</span></strong><span>与NCBI下载的chrom.sizes里的染色体名称对应不上，会出现报错。</span></p></li><li><p><span>可以通过bam文件生成的chrom.sizes解决上面的问题，把未必对的reads也输出到 Bedgraph 文件，</span><a href='#chromsizes'><span>详情点此</span></a></p></li></ul><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation"><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># -F 0x04：-F 选项用于排除标志为 0x04 的reads，这表示排除未比对的reads。仅保留那些已成功比对到参考基因组的reads。</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">samtools view <span class="cm-attribute">-F</span> 0x04 &lt;before.bam/sam/cram&gt; <span class="cm-attribute">-o</span> &lt;after.bam/sam/cram&gt;</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 44px;"></div><div class="CodeMirror-gutters" style="display: none; height: 44px;"></div></div></div></pre><p>&nbsp;</p><hr /><h1 id='bedtools'><span>Bedtools</span></h1><h2 id='bedtools-introduction'><span>Bedtools introduction</span></h2><p><img src="./picture/046.png" width="150"/></p><p><span>Bedtools（</span><a href='https://bedtools.readthedocs.io/en/latest/index.html'><span>bedtools: a powerful toolset for genome arithmetic — bedtools 2.31.0 documentation</span></a><span>） 被誉为基因组分析中的“瑞士军刀”，提供了一系列多功能的工具，用于操作和分析基因组区间。其最广泛使用的工具可以进行</span><strong><span>基因组算术</span></strong><span>操作，即在基因组上执行集合理论相关的操作。</span></p><p>&nbsp;</p><p><strong><span>主要功能</span></strong></p><ul><li><p><strong><span>交集 (Intersect):</span></strong><span>查找不同基因组数据集之间的重叠区间。</span></p></li><li><p><strong><span>合并 (Merge):</span></strong><span>将重叠的区间合并成单一区间。</span></p></li><li><p><strong><span>计数 (Count):</span></strong><span>统计与其他区间重叠的区间数量。</span></p></li><li><p><strong><span>互补 (Complement):</span></strong><span>识别基因组中不与任何给定数据集的区间重叠的区域。</span></p></li><li><p><strong><span>随机打乱 (Shuffle):</span></strong><span>随机排列基因组上的区间，以创建对照数据集。</span></p></li></ul><p>&nbsp;</p><p><strong><span>支持的文件格式</span></strong></p><ul><li><p><strong><span>BAM</span></strong><span>: 用于比对序列的二进制文件格式。</span></p></li><li><p><strong><span>BED</span></strong><span>: 用于注释轨道的文本文件格式。</span></p></li><li><p><strong><span>GFF/GTF</span></strong><span>: 用于基因注释的文件格式。</span></p></li><li><p><strong><span>VCF</span></strong><span>: 用于变异检测数据的文件格式。</span></p></li></ul><p>&nbsp;</p><p><strong><span>Bedtools installation</span></strong><span> </span><code>conda install -c bioconda bedtools</code></p><p>&nbsp;</p><hr /><h2 id='bedtools-genomecov'><span>Bedtools genomecov</span></h2><ul><li><p><a href='https://bedtools.readthedocs.io/en/latest/content/tools/genomecov.html'><span>Bedtools genomecov</span></a></p></li><li><p><span>使用</span><code>bedtools genomecov</code><span>可以计算基因组水平上的reads </span><strong><span>Coverage</span></strong><span>，并生成各种各样的文件（包括</span><code>Bedgraph</code><span>）</span></p></li></ul><p><img src="./picture/049.png" width="700"/></p><p>&nbsp;</p><p><strong><span>parameter</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;  bedtools genomecov </span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==输入选项==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-i</span> &lt;bed/gff/vcf&gt; &nbsp;<span class="cm-comment"># 输入文件，注意BED 文件必须按染色体进行排序</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-ibam</span> &lt;bam&gt; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 输入文件是 BAM 格式。注意 BAM 文件必须按照位置排序才能正确处理</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-g</span> &lt;genome_file&gt; &nbsp;<span class="cm-comment"># 提供染色体长度信息文件(chrom.sizes)路径，不使用 -ibam 选项时必须提供。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==输出选项==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-bg</span> </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 生成BedGraph格式（适用于基因组浏览器等工具）输出</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-bga</span>：</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 与-bg相似，但会输出覆盖度为 0的区间。这可以帮助用户快速提取覆盖度为 0 的基因组区域</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-d</span> &nbsp; </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 报告每个基因组位置的覆盖深度，使用1-based</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 默认情况下，genomecov 报告的是一个覆盖度直方图，而使用 -d 会输出每个位置的详细深度信息</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-dz</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 类似于 -d，但使用0-based </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-split</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 对BAM 或 BED12 文件中的split区间进行独立处理。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 在 RNA-Seq 数据中，经常会有 spliced reads（即 read 被分割成多个片段），使用此选项可以分别计算每个片段的覆盖度</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 对于 BAM 文件，它使用 CIGAR 字符串中的 "N" 和 "D" 操作来推断覆盖区间</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 对于 BED12 文件，它使用 BlockCount、BlockStarts 和 BlockEnds 字段</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-max</span> &lt;int&gt;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 将覆盖度大于等于指定最大值的所有位置合并为一个区间。与 -d 和 -bg 一起使用时无效</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==特殊选项==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-scale</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 按常数因子缩放覆盖度。每个覆盖度值将乘以此因子。这个选项常用于标准化覆盖度数据。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-ignoreD</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 在 BAM 文件中忽略局部缺失（CIGAR 字符串中的 "D" 操作），计算覆盖度时不会考虑这些删除位点</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-strand</span> &lt;<span class="cm-operator">+</span> or <span class="cm-attribute">-</span>&gt;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 仅计算特定链上的覆盖度。对于 BED 文件，文件需要至少包含 6 列，链信息位于第 6 列</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-pc</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 计算paired-end的覆盖度，仅适用于 BAM 文件</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-fs</span> &lt;int&gt;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 强制使用指定的片段长度而不是 read 长度。仅适用于 BAM 文件</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-du</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 对于paired-end，将两个配对的 read 看作是同一条链上的数据（用于链特异性数据分析）。仅适用于 BAM 文件</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-5</span> <span class="cm-comment"># 只会统计每个 read 在 5' 端第一个碱基的converage</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-3</span> <span class="cm-comment"># 只会统计每个 read 在 3' 端第一个碱基的converage</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 适用于需要关注 read 起点或终点的分析场景，例如转录起始位点 (TSS) 或终止位点 (TES) 的converage</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==上传UCSC基因组浏览器选项==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; <span class="cm-attribute">-trackline</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; <span class="cm-comment"># 在输出的 BedGraph 文件的第一行添加一个 UCSC 基因组浏览器的 track line 定义，以便将生成的文件直接上传到 UCSC 基因组浏览器作为自定义轨迹</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; <span class="cm-comment"># 注意：如果输出文件中包含了 track line，不能直接将该文件转换为 BigWig 文件，除非手动删除第一行</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; <span class="cm-attribute">-trackopts</span> <span class="cm-string">'name="My Track" visibility=2 color=255,30,30'</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; <span class="cm-comment"># 在第一行输出中添加的 track line参数，用于自定义 UCSC 基因组浏览器的轨迹定义。</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 1464px;"></div><div class="CodeMirror-gutters" style="display: none; height: 1464px;"></div></div></div></pre><p>&nbsp;</p><div class="md-alert md-alert-caution caution"><p><span class='md-alert-text md-alert-text-caution'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M4.47.22A.749.749 0 0 1 5 0h6c.199 0 .389.079.53.22l4.25 4.25c.141.14.22.331.22.53v6a.749.749 0 0 1-.22.53l-4.25 4.25A.749.749 0 0 1 11 16H5a.749.749 0 0 1-.53-.22L.22 11.53A.749.749 0 0 1 0 11V5c0-.199.079-.389.22-.53Zm.84 1.28L1.5 5.31v5.38l3.81 3.81h5.38l3.81-3.81V5.31L10.69 1.5ZM8 4a.75.75 0 0 1 .75.75v3.5a.75.75 0 0 1-1.5 0v-3.5A.75.75 0 0 1 8 4Zm0 8a1 1 0 1 1 0-2 1 1 0 0 1 0 2Z"></path></svg>Caution</span><br></p><ul><li><p><span>输入的 BED 文件必须按染色体进行排序</span></p></li><li><p><span>输入格式为 BED 文件，则需要提供染色体长度信息文件(</span><a href='#chromsizes'><span>chrom.sizes</span></a><span>)。</span><a href='#chromsizes'><span>chrom.sizes文件详细点此</span></a></p></li><li><p><span>输入的chrom.sizes文件与BED文件染色体格式必须要相同</span></p></li><li><p><span>输入的 BAM 文件必须按位置排序</span></p></li></ul></div><p>&nbsp;</p><p><strong><span>Example</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-meta">#!/bin/bash</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 输入样本名</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-def">sample</span><span class="cm-operator">=</span><span class="cm-string">"sample"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 提供染色体长度信息文件(chrom.sizes)路径。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-def">chromSize</span><span class="cm-operator">=</span><span class="cm-string">"&lt;xxx.chrom.sizes&gt;"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 定义比对到主基因组的文件路径</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-def">primary_alignment</span><span class="cm-operator">=</span><span class="cm-string">"</span><span class="cm-def">${sample}</span><span class="cm-string">.fragments.bed"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 将bed转换成bedtools</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">bedtools genomecov <span class="cm-attribute">-bg</span> &nbsp;<span class="cm-attribute">-i</span> <span class="cm-def">$primary_alignment</span> <span class="cm-attribute">-g</span> <span class="cm-def">$chromSize</span> &gt; <span class="cm-def">${sample}</span>.fragments.bedgraph</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 284px;"></div><div class="CodeMirror-gutters" style="display: none; height: 284px;"></div></div></div></pre><p>&nbsp;</p><hr /><h2 id='bedtools-intersect'><span>Bedtools intersect</span></h2><ul><li><p><a href='https://bedtools.readthedocs.io/en/latest/content/tools/intersect.html'><span>bedtools intersect</span></a><span>: This command is used to find overlapping features between two datasets</span></p></li><li><p><span>if you have </span><strong><span>large file to intersect</span></strong><span> ,suggesting you sort your file ,and add </span><code>-sorted</code><span> option in the </span><code>bedtools intersect</code><span> to reduce memory, Otherwise Out of Memory will kill the program.  </span></p></li></ul><p><img src="./picture/054.png" width="900"/></p><p><strong><span>Parameters</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">bedtools intersect</span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==input option==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-a</span> &lt;A&gt; &nbsp; &nbsp;<span class="cm-comment"># Specifies the file "A", which could be in BED/BAM/GFF/VCF format. Each feature in A is compared with B to look for overlaps</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-b</span> &lt;B&gt; &nbsp; &nbsp;<span class="cm-comment"># Specifies the file "B", which can be one or more BED/BAM/GFF/VCF files. You can pass multiple files or use wildcards (e.g., *.bed)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==output option==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-ubam</span> &nbsp; &nbsp; &nbsp;<span class="cm-comment"># Outputs the BAM file as uncompressed. (Default:compressed BAM)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-bed</span> &nbsp; &nbsp; &nbsp; <span class="cm-comment"># When using BAM input for "A", this option outputs the results in BED format. (Default:BAM format)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-header</span> &nbsp; &nbsp;<span class="cm-comment"># Print the header from the A file to results</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-names</span> &nbsp; &nbsp; <span class="cm-comment"># When using multiple databases (-b),  provide an alias for each that will appear instead of a file ID in DB record.</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-filenames</span> <span class="cm-comment"># When using multiple databases (-b), show each complete filename instead of a file ID in DB record.</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-sortout</span> &nbsp; <span class="cm-comment"># Sorts the output hits from B for each record in A. This can be useful when you're using multiple databases in -b.</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==Reporting Options==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-wa</span><span class="cm-tab" role="presentation" cm-text="	"> </span> <span class="cm-comment"># Write the original entry in A for each overlap.</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-wb</span><span class="cm-tab" role="presentation" cm-text="	"> </span> <span class="cm-comment"># Write the original entry in B for each overlap. (Restricted by -f and -r)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-loj</span> <span class="cm-comment"># meaning that each feature in A will be reported with its overlaps in B</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># If no overlaps are found, report a NULL feature for B.</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-wo</span> &nbsp;<span class="cm-comment"># Write the original A and B entries plus the number of base pairs of overlap between the two features. </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># Only A features with overlap are reported. (Restricted by -f and -r). &nbsp; </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-wao</span> <span class="cm-comment"># Similar to -wo, but also reports A features with no overlap. (Restricted by -f and -r)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-u</span> &nbsp; <span class="cm-comment"># Writes the original entry from A once if any overlap is found with B. (Restricted by -f and -r)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-v</span> &nbsp; <span class="cm-comment"># Only reports those entries in A that have no overlap with B. (This is the opposite of -u)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==Counting Options==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-c</span> <span class="cm-comment"># For each feature in A, reports the number of overlaps found in B. (Restricted -f, -F, -r, and -s)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-C</span> <span class="cm-comment"># For each feature in A, it reports the number of overlaps with each B file on separate lines. (Restricted -f, -F, -r, and -s)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==Overlap Requirements==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-f</span> <span class="cm-comment"># Minimum overlap required as a fraction of A. (Default: is 1E-9 (1bp))</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-F</span> <span class="cm-comment"># Minimum overlap required as a fraction of B. (Default: 1E-9 (1bp))</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-r</span> <span class="cm-comment"># Requires reciprocal overlap between A and B, meaning that both must satisfy the minimum overlap fractions.</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; <span class="cm-comment"># For example, if -f is 0.90 and -r is used, this requires that B overlaps at least 90% of A and vice versa.</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-e</span> <span class="cm-comment"># Requires that the overlap fraction be satisfied for A or B </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; <span class="cm-comment"># In other words, if -e is used with -f 0.90 and -F 0.10 this requires that either 90% of A is covered OR 10% of B is covered. </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; <span class="cm-comment"># Without -e, both fractions would have to be satisfied.</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==Strand-Specific(+-链) Options==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-s</span> <span class="cm-comment"># Enforces "strandedness", meaning that only overlaps on the same strand are considered. (Default: strand is ignored)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-S</span> <span class="cm-comment"># Enforces "opposite strandedness", meaning only overlaps on the opposite strand are reported.</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==Special Options for BAM Files==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-split</span> <span class="cm-comment"># Treat “split” BAM (having an “N” CIGAR(Skipped region)operation) or BED12 entries as distinct BED intervals</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==performance optimization==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-sorted</span> <span class="cm-comment"># For very large B files, invokes a "sweeping" algorithm that requires the input to be sorted by chromosome and position </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># (e.g., sort -k1,1 -k2,2n).</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># This reduces memory usage for large files. &nbsp;</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-g</span> &lt;genome file&gt; <span class="cm-comment"># pecifies a genome file that defines the expected chromosome order for use with the -sorted option.</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-nobuf</span> <span class="cm-comment"># Disables buffered output, meaning each line is printed immediately. </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># This can slow down large outputs but is useful when output is processed line-by-line.</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-iobuf</span> &lt;size&gt; <span class="cm-comment"># Sets the size of the read buffer (in bytes). </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># Supports suffixes like K, M, or G . </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># Currently has no effect with compressed files.</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 1770px;"></div><div class="CodeMirror-gutters" style="display: none; height: 1770px;"></div></div></div></pre><p>&nbsp;</p><hr /><h2 id='bedtools-bamtobed'><span>Bedtools bamtobed</span></h2><p><span>## </span></p><p><a href='https://bedtools.readthedocs.io/en/latest/content/tools/bedtobam.html'><span>bedtools bamtobed</span></a><span> 用于将 BAM 文件转换为 BED 格式。</span></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation"><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">bedtools bamtobed <span class="cm-attribute">-i</span> sample.bam &gt; sample.bed <span class="cm-attribute">-bedpe</span></span></pre></div></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 22px;"></div><div class="CodeMirror-gutters" style="display: none; height: 22px;"></div></div></div></pre><ul><li><p><span>默认的输出格式为</span><strong><span>BED6</span></strong><span>格式 ，如果为paired-end reads，则把paired ends 分开几行输出，而不在一行</span></p></li><li><p><span>如果为paired-end reads，可以选择 </span><code>-bedpe</code><span>来方便后面的过率，对于生成的 BEDPE 文件。</span></p></li><li><p><code>-mate1</code><span>：当使用 </span><code>-bedpe</code><span>时，</span><code>-mate1</code><span> 参数确保输出的 BEDPE 条目中的第一对片段（第一个 &quot;block&quot;）始终是 read 1</span></p><p><span>不使用</span><code>-mate1</code><span>：</span></p><ul><li><p><strong><span>第一个片段（block1)</span></strong><span> 通常对应于在基因组中较小的起始位置的片段（即，Start1）</span></p></li><li><p><strong><span>第二个片段（block2)</span></strong><span> 则对应于较大的起始位置的片段（即，Start2）</span></p></li></ul></li><li><p><code>-bed12</code><span>：输出为BED12格式</span></p></li><li><p><code>-color</code><span>：指定用于 BED12 格式中的颜色，使用 R,G,B 字符串，默认颜色为 (255,0,0)，即红色，在可视化时，此选项可以用来定义显示的颜色</span></p></li><li><p><code>-cigar</code><span>：将 CIGAR 字符串添加为 BED 文件的第七列。</span></p></li><li><p><code>-split</code><span>：如果 BAM 文件中包含有“分割”的比对（例如，跨多个外显子或重复区域的比对），此选项会将这些比对拆分成多个单独的 BED 条目</span></p></li><li><p><code>-splitD</code><span>：用于将 BAM 文件中的比对按照 N 和 D 的 CIGAR 操作符进行拆分，并将每个拆分后的片段作为单独的 BED 记录</span></p></li><li><p><code>-tag</code><span>：使用其他数值类型的 BAM 标签作为 BED 文件中的得分。默认情况下，BED 文件使用比对质量作为得分。不允许与 BEDPE 格式一起使用</span></p></li><li><p><code>-ed</code><span>： 参数用于将 BAM 文件中的编辑距离（即 NM 标签）（即比对过程中引入的错配、插入和删除的总数）作为生成的 BED 文件的得分列的值。</span></p></li></ul><p>&nbsp;</p><hr /><h1 id='convert-bam-to-bed-and-bed-operate'><span>Convert BAM to BED and BED Operate</span></h1><h2 id='convert-bam-to-bed-with-bedtools'><span>Convert BAM to BED with bedtools</span></h2><p><a href='#bedtools-bamtobed'><span>bedtools bamtobed</span></a><span> 用于将 BAM 文件转换为 BED 格式。</span></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation"><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">bedtools bamtobed <span class="cm-attribute">-i</span> sample.bam &gt; sample.bed <span class="cm-attribute">-bedpe</span></span></pre></div></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 22px;"></div><div class="CodeMirror-gutters" style="display: none; height: 22px;"></div></div></div></pre><p>&nbsp;</p><hr /><h2 id='keep-the-read-pairs-that-are-on-the-same-chromosome-and-fragment-length-less-than-1000bpoptial'><span>Keep the read pairs that are on the same chromosome and fragment length less than 1000bp（optial）</span></h2><ul><li><p><span>对于</span><strong><span>Paired-End</span></strong><span>需进一步过滤并保留位于相同染色体上且fragment长度小于 1000bp 的read pairs</span></p></li><li><p><code>bowtie2</code><span>设置以下parameter可以逃过此步：</span></p><p><code>--no-discordant</code><span> together with </span><code>-I</code><span> </span><code>10</code><span> </span><code>-X</code><span> </span><code>700</code><span> ensures that the fragment is 10-700bp long which will also exclude read pairs that are on the different chromosome.</span></p></li><li><p><strong><span>Code</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation"><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-builtin">awk</span> <span class="cm-string">'$1==$4 &amp;&amp; $6-$2 &lt; 1000 {print $0}'</span> sample.bed &gt; sample.clean.bed </span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># $1==$4：表示过滤掉那些不在同一染色体上的读取对（即第一列和第四列的染色体名称必须相同）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># &amp;&amp; 的作用是 仅当前一个命令成功执行（即返回状态码为 0）时，才会执行下一个命令。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># $6-$2 &lt; 1000：过滤条件，表示保留片段长度小于 1000bp 的读取对。第六列和第二列分别代表片段末端和起始位置，计算差值即为片段长度。</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 87px;"></div><div class="CodeMirror-gutters" style="display: none; height: 87px;"></div></div></div></pre><p>&nbsp;</p></li></ul><hr /><h2 id='extract-the-fragment-related-columns'><span>Extract the fragment related columns</span></h2><p><span>我们可以提取</span><strong><span>fragment</span></strong><span>从</span><strong><span>bedpe</span></strong><span>文件中</span></p><p><img src="./picture/080.png" width="900"/></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-builtin">cut</span> <span class="cm-attribute">-f</span> <span class="cm-number">1</span>,2,6 sample.bed | \</span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># cut 命令用于从文件中提取指定列。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># -f 1,2,6 表示提取第 1、2 和 6 列</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">#第 1 列：代表染色体（Chromosome）。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">#第 2 列：代表较小的起始位置的片段（Start1）。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">#第 6 列：代表较大的起始位置的片段（End2）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-builtin">sort</span> <span class="cm-attribute">-k1</span>,1 <span class="cm-attribute">-k2</span>,2n <span class="cm-attribute">-k3</span>,3n &gt; sample.fragments.bed</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">#这个命令将对数据进行多级排序：</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">#第一步：按照第 1 列（-k1,1）进行排序。这一步按字母顺序进行。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">#第二步：在第 1 列相同的情况下，按第 2 列（-k2,2n）进行排序。这一步按数字大小排序，确保数值上较小的排在前面。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">#第三步：如果第 1 列和第 2 列都相同，再按第 3 列（-k3,3n）进行排序，也按数字大小排序。</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 284px;"></div><div class="CodeMirror-gutters" style="display: none; height: 284px;"></div></div></div></pre><div class="md-alert md-alert-warning warning"><p><span class='md-alert-text md-alert-text-warning'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M6.457 1.047c.659-1.234 2.427-1.234 3.086 0l6.082 11.378A1.75 1.75 0 0 1 14.082 15H1.918a1.75 1.75 0 0 1-1.543-2.575Zm1.763.707a.25.25 0 0 0-.44 0L1.698 13.132a.25.25 0 0 0 .22.368h12.164a.25.25 0 0 0 .22-.368Zm.53 3.996v2.5a.75.75 0 0 1-1.5 0v-2.5a.75.75 0 0 1 1.5 0ZM9 11a1 1 0 1 1-2 0 1 1 0 0 1 2 0Z"></path></svg>Warning</span><br></p><p><strong><span>这里输入的bed格式为bedpe格式，否则第1，2，6列不对应，需要根据情况更改。</span></strong></p></div><p>&nbsp;</p><hr /><h2 id='evaluate-the-reproducibility-of-replicate-samplesoptial'><span>Evaluate the reproducibility of replicate samples(optial)</span></h2><p><span>为了评估不同重复样本之间以及不同条件下的重现性：首先将基因组分成</span><strong><span>500 bp</span></strong><span> 的窗口（bin），然后计算每个窗口中</span><strong><span>read count的 log2 转换值</span></strong><span>。接着，计算每对重复样本之间的 </span><strong><span>Pearson 相关性</span></strong><span>，并通过层次聚类的方式展示。</span></p><p>&nbsp;</p><ol start='' ><li><p><strong><span>计算每个窗口的read count</span></strong></p><p><strong><span>如何计算每个窗口的read count？</span></strong><span>计算每个read中点所在的窗口中心位置</span></p><ul><li><p><span>计算每个 Read 的中点</span></p><div contenteditable="false" spellcheck="false" class="mathjax-block md-end-block md-math-block md-rawblock" id="mathjax-n2779" cid="n2779" mdtype="math_block" data-math-tag-before="0" data-math-tag-after="0" data-math-labels="[]"><div class="md-math-container"><mjx-container class="MathJax" jax="SVG" display="true" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="26.115ex" height="4.652ex" role="img" focusable="false" viewBox="0 -1370 11542.9 2056" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -1.552ex;"><defs><path id="MJX-7-TEX-N-6D" d="M41 46H55Q94 46 102 60V68Q102 77 102 91T102 122T103 161T103 203Q103 234 103 269T102 328V351Q99 370 88 376T43 385H25V408Q25 431 27 431L37 432Q47 433 65 434T102 436Q119 437 138 438T167 441T178 442H181V402Q181 364 182 364T187 369T199 384T218 402T247 421T285 437Q305 442 336 442Q351 442 364 440T387 434T406 426T421 417T432 406T441 395T448 384T452 374T455 366L457 361L460 365Q463 369 466 373T475 384T488 397T503 410T523 422T546 432T572 439T603 442Q729 442 740 329Q741 322 741 190V104Q741 66 743 59T754 49Q775 46 803 46H819V0H811L788 1Q764 2 737 2T699 3Q596 3 587 0H579V46H595Q656 46 656 62Q657 64 657 200Q656 335 655 343Q649 371 635 385T611 402T585 404Q540 404 506 370Q479 343 472 315T464 232V168V108Q464 78 465 68T468 55T477 49Q498 46 526 46H542V0H534L510 1Q487 2 460 2T422 3Q319 3 310 0H302V46H318Q379 46 379 62Q380 64 380 200Q379 335 378 343Q372 371 358 385T334 402T308 404Q263 404 229 370Q202 343 195 315T187 232V168V108Q187 78 188 68T191 55T200 49Q221 46 249 46H265V0H257L234 1Q210 2 183 2T145 3Q42 3 33 0H25V46H41Z"></path><path id="MJX-7-TEX-N-69" d="M69 609Q69 637 87 653T131 669Q154 667 171 652T188 609Q188 579 171 564T129 549Q104 549 87 564T69 609ZM247 0Q232 3 143 3Q132 3 106 3T56 1L34 0H26V46H42Q70 46 91 49Q100 53 102 60T104 102V205V293Q104 345 102 359T88 378Q74 385 41 385H30V408Q30 431 32 431L42 432Q52 433 70 434T106 436Q123 437 142 438T171 441T182 442H185V62Q190 52 197 50T232 46H255V0H247Z"></path><path id="MJX-7-TEX-N-64" d="M376 495Q376 511 376 535T377 568Q377 613 367 624T316 637H298V660Q298 683 300 683L310 684Q320 685 339 686T376 688Q393 689 413 690T443 693T454 694H457V390Q457 84 458 81Q461 61 472 55T517 46H535V0Q533 0 459 -5T380 -11H373V44L365 37Q307 -11 235 -11Q158 -11 96 50T34 215Q34 315 97 378T244 442Q319 442 376 393V495ZM373 342Q328 405 260 405Q211 405 173 369Q146 341 139 305T131 211Q131 155 138 120T173 59Q203 26 251 26Q322 26 373 103V342Z"></path><path id="MJX-7-TEX-N-70" d="M36 -148H50Q89 -148 97 -134V-126Q97 -119 97 -107T97 -77T98 -38T98 6T98 55T98 106Q98 140 98 177T98 243T98 296T97 335T97 351Q94 370 83 376T38 385H20V408Q20 431 22 431L32 432Q42 433 61 434T98 436Q115 437 135 438T165 441T176 442H179V416L180 390L188 397Q247 441 326 441Q407 441 464 377T522 216Q522 115 457 52T310 -11Q242 -11 190 33L182 40V-45V-101Q182 -128 184 -134T195 -145Q216 -148 244 -148H260V-194H252L228 -193Q205 -192 178 -192T140 -191Q37 -191 28 -194H20V-148H36ZM424 218Q424 292 390 347T305 402Q234 402 182 337V98Q222 26 294 26Q345 26 384 80T424 218Z"></path><path id="MJX-7-TEX-N-6F" d="M28 214Q28 309 93 378T250 448Q340 448 405 380T471 215Q471 120 407 55T250 -10Q153 -10 91 57T28 214ZM250 30Q372 30 372 193V225V250Q372 272 371 288T364 326T348 362T317 390T268 410Q263 411 252 411Q222 411 195 399Q152 377 139 338T126 246V226Q126 130 145 91Q177 30 250 30Z"></path><path id="MJX-7-TEX-N-6E" d="M41 46H55Q94 46 102 60V68Q102 77 102 91T102 122T103 161T103 203Q103 234 103 269T102 328V351Q99 370 88 376T43 385H25V408Q25 431 27 431L37 432Q47 433 65 434T102 436Q119 437 138 438T167 441T178 442H181V402Q181 364 182 364T187 369T199 384T218 402T247 421T285 437Q305 442 336 442Q450 438 463 329Q464 322 464 190V104Q464 66 466 59T477 49Q498 46 526 46H542V0H534L510 1Q487 2 460 2T422 3Q319 3 310 0H302V46H318Q379 46 379 62Q380 64 380 200Q379 335 378 343Q372 371 358 385T334 402T308 404Q263 404 229 370Q202 343 195 315T187 232V168V108Q187 78 188 68T191 55T200 49Q221 46 249 46H265V0H257L234 1Q210 2 183 2T145 3Q42 3 33 0H25V46H41Z"></path><path id="MJX-7-TEX-N-74" d="M27 422Q80 426 109 478T141 600V615H181V431H316V385H181V241Q182 116 182 100T189 68Q203 29 238 29Q282 29 292 100Q293 108 293 146V181H333V146V134Q333 57 291 17Q264 -10 221 -10Q187 -10 162 2T124 33T105 68T98 100Q97 107 97 248V385H18V422H27Z"></path><path id="MJX-7-TEX-I-1D456" d="M184 600Q184 624 203 642T247 661Q265 661 277 649T290 619Q290 596 270 577T226 557Q211 557 198 567T184 600ZM21 287Q21 295 30 318T54 369T98 420T158 442Q197 442 223 419T250 357Q250 340 236 301T196 196T154 83Q149 61 149 51Q149 26 166 26Q175 26 185 29T208 43T235 78T260 137Q263 149 265 151T282 153Q302 153 302 143Q302 135 293 112T268 61T223 11T161 -11Q129 -11 102 10T74 74Q74 91 79 106T122 220Q160 321 166 341T173 380Q173 404 156 404H154Q124 404 99 371T61 287Q60 286 59 284T58 281T56 279T53 278T49 278T41 278H27Q21 284 21 287Z"></path><path id="MJX-7-TEX-N-3D" d="M56 347Q56 360 70 367H707Q722 359 722 347Q722 336 708 328L390 327H72Q56 332 56 347ZM56 153Q56 168 72 173H708Q722 163 722 153Q722 140 707 133H70Q56 140 56 153Z"></path><path id="MJX-7-TEX-N-73" d="M295 316Q295 356 268 385T190 414Q154 414 128 401Q98 382 98 349Q97 344 98 336T114 312T157 287Q175 282 201 278T245 269T277 256Q294 248 310 236T342 195T359 133Q359 71 321 31T198 -10H190Q138 -10 94 26L86 19L77 10Q71 4 65 -1L54 -11H46H42Q39 -11 33 -5V74V132Q33 153 35 157T45 162H54Q66 162 70 158T75 146T82 119T101 77Q136 26 198 26Q295 26 295 104Q295 133 277 151Q257 175 194 187T111 210Q75 227 54 256T33 318Q33 357 50 384T93 424T143 442T187 447H198Q238 447 268 432L283 424L292 431Q302 440 314 448H322H326Q329 448 335 442V310L329 304H301Q295 310 295 316Z"></path><path id="MJX-7-TEX-N-61" d="M137 305T115 305T78 320T63 359Q63 394 97 421T218 448Q291 448 336 416T396 340Q401 326 401 309T402 194V124Q402 76 407 58T428 40Q443 40 448 56T453 109V145H493V106Q492 66 490 59Q481 29 455 12T400 -6T353 12T329 54V58L327 55Q325 52 322 49T314 40T302 29T287 17T269 6T247 -2T221 -8T190 -11Q130 -11 82 20T34 107Q34 128 41 147T68 188T116 225T194 253T304 268H318V290Q318 324 312 340Q290 411 215 411Q197 411 181 410T156 406T148 403Q170 388 170 359Q170 334 154 320ZM126 106Q126 75 150 51T209 26Q247 26 276 49T315 109Q317 116 318 175Q318 233 317 233Q309 233 296 232T251 223T193 203T147 166T126 106Z"></path><path id="MJX-7-TEX-N-72" d="M36 46H50Q89 46 97 60V68Q97 77 97 91T98 122T98 161T98 203Q98 234 98 269T98 328L97 351Q94 370 83 376T38 385H20V408Q20 431 22 431L32 432Q42 433 60 434T96 436Q112 437 131 438T160 441T171 442H174V373Q213 441 271 441H277Q322 441 343 419T364 373Q364 352 351 337T313 322Q288 322 276 338T263 372Q263 381 265 388T270 400T273 405Q271 407 250 401Q234 393 226 386Q179 341 179 207V154Q179 141 179 127T179 101T180 81T180 66V61Q181 59 183 57T188 54T193 51T200 49T207 48T216 47T225 47T235 46T245 46H276V0H267Q249 3 140 3Q37 3 28 0H20V46H36Z"></path><path id="MJX-7-TEX-N-2B" d="M56 237T56 250T70 270H369V420L370 570Q380 583 389 583Q402 583 409 568V270H707Q722 262 722 250T707 230H409V-68Q401 -82 391 -82H389H387Q375 -82 369 -68V230H70Q56 237 56 250Z"></path><path id="MJX-7-TEX-N-65" d="M28 218Q28 273 48 318T98 391T163 433T229 448Q282 448 320 430T378 380T406 316T415 245Q415 238 408 231H126V216Q126 68 226 36Q246 30 270 30Q312 30 342 62Q359 79 369 104L379 128Q382 131 395 131H398Q415 131 415 121Q415 117 412 108Q393 53 349 21T250 -11Q155 -11 92 58T28 218ZM333 275Q322 403 238 411H236Q228 411 220 410T195 402T166 381T143 340T127 274V267H333V275Z"></path><path id="MJX-7-TEX-N-32" d="M109 429Q82 429 66 447T50 491Q50 562 103 614T235 666Q326 666 387 610T449 465Q449 422 429 383T381 315T301 241Q265 210 201 149L142 93L218 92Q375 92 385 97Q392 99 409 186V189H449V186Q448 183 436 95T421 3V0H50V19V31Q50 38 56 46T86 81Q115 113 136 137Q145 147 170 174T204 211T233 244T261 278T284 308T305 340T320 369T333 401T340 431T343 464Q343 527 309 573T212 619Q179 619 154 602T119 569T109 550Q109 549 114 549Q132 549 151 535T170 489Q170 464 154 447T109 429Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="msub"><g data-mml-node="mtext"><use data-c="6D" xlink:href="#MJX-7-TEX-N-6D"></use><use data-c="69" xlink:href="#MJX-7-TEX-N-69" transform="translate(833,0)"></use><use data-c="64" xlink:href="#MJX-7-TEX-N-64" transform="translate(1111,0)"></use><use data-c="70" xlink:href="#MJX-7-TEX-N-70" transform="translate(1667,0)"></use><use data-c="6F" xlink:href="#MJX-7-TEX-N-6F" transform="translate(2223,0)"></use><use data-c="69" xlink:href="#MJX-7-TEX-N-69" transform="translate(2723,0)"></use><use data-c="6E" xlink:href="#MJX-7-TEX-N-6E" transform="translate(3001,0)"></use><use data-c="74" xlink:href="#MJX-7-TEX-N-74" transform="translate(3557,0)"></use></g><g data-mml-node="mi" transform="translate(3979,-229.4) scale(0.707)"><use data-c="1D456" xlink:href="#MJX-7-TEX-I-1D456"></use></g></g><g data-mml-node="mo" transform="translate(4550.7,0)"><use data-c="3D" xlink:href="#MJX-7-TEX-N-3D"></use></g><g data-mml-node="mfrac" transform="translate(5606.5,0)"><g data-mml-node="mrow" transform="translate(220,676)"><g data-mml-node="msub"><g data-mml-node="mtext"><use data-c="73" xlink:href="#MJX-7-TEX-N-73"></use><use data-c="74" xlink:href="#MJX-7-TEX-N-74" transform="translate(394,0)"></use><use data-c="61" xlink:href="#MJX-7-TEX-N-61" transform="translate(783,0)"></use><use data-c="72" xlink:href="#MJX-7-TEX-N-72" transform="translate(1283,0)"></use><use data-c="74" xlink:href="#MJX-7-TEX-N-74" transform="translate(1675,0)"></use></g><g data-mml-node="mi" transform="translate(2097,-150) scale(0.707)"><use data-c="1D456" xlink:href="#MJX-7-TEX-I-1D456"></use></g></g><g data-mml-node="mo" transform="translate(2613.2,0)"><use data-c="2B" xlink:href="#MJX-7-TEX-N-2B"></use></g><g data-mml-node="msub" transform="translate(3613.4,0)"><g data-mml-node="mtext"><use data-c="65" xlink:href="#MJX-7-TEX-N-65"></use><use data-c="6E" xlink:href="#MJX-7-TEX-N-6E" transform="translate(444,0)"></use><use data-c="64" xlink:href="#MJX-7-TEX-N-64" transform="translate(1000,0)"></use></g><g data-mml-node="mi" transform="translate(1589,-150) scale(0.707)"><use data-c="1D456" xlink:href="#MJX-7-TEX-I-1D456"></use></g></g></g><g data-mml-node="mn" transform="translate(2718.2,-686)"><use data-c="32" xlink:href="#MJX-7-TEX-N-32"></use></g><rect width="5696.3" height="60" x="120" y="220"></rect></g></g></g></svg><mjx-assistive-mml unselectable="on" display="block"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><msub><mtext>midpoint</mtext><mi>i</mi></msub><mo>=</mo><mfrac><mrow><msub><mtext>start</mtext><mi>i</mi></msub><mo>+</mo><msub><mtext>end</mtext><mi>i</mi></msub></mrow><mn>2</mn></mfrac></math></mjx-assistive-mml></mjx-container></div></div></li><li><p><span>计算每个中点所在的窗口索引</span></p><div contenteditable="false" spellcheck="false" class="mathjax-block md-end-block md-math-block md-rawblock" id="mathjax-n2782" cid="n2782" mdtype="math_block" data-math-tag-before="0" data-math-tag-after="0" data-math-labels="[]"><div class="md-math-container"><mjx-container class="MathJax" jax="SVG" display="true" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="31.076ex" height="5.428ex" role="img" focusable="false" viewBox="0 -1449.5 13735.5 2399" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -2.148ex;"><defs><path id="MJX-8-TEX-N-77" d="M90 368Q84 378 76 380T40 385H18V431H24L43 430Q62 430 84 429T116 428Q206 428 221 431H229V385H215Q177 383 177 368Q177 367 221 239L265 113L339 328L333 345Q323 374 316 379Q308 384 278 385H258V431H264Q270 428 348 428Q439 428 454 431H461V385H452Q404 385 404 369Q404 366 418 324T449 234T481 143L496 100L537 219Q579 341 579 347Q579 363 564 373T530 385H522V431H529Q541 428 624 428Q692 428 698 431H703V385H697Q696 385 691 385T682 384Q635 377 619 334L559 161Q546 124 528 71Q508 12 503 1T487 -11H479Q460 -11 456 -4Q455 -3 407 133L361 267Q359 263 266 -4Q261 -11 243 -11H238Q225 -11 220 -3L90 368Z"></path><path id="MJX-8-TEX-N-69" d="M69 609Q69 637 87 653T131 669Q154 667 171 652T188 609Q188 579 171 564T129 549Q104 549 87 564T69 609ZM247 0Q232 3 143 3Q132 3 106 3T56 1L34 0H26V46H42Q70 46 91 49Q100 53 102 60T104 102V205V293Q104 345 102 359T88 378Q74 385 41 385H30V408Q30 431 32 431L42 432Q52 433 70 434T106 436Q123 437 142 438T171 441T182 442H185V62Q190 52 197 50T232 46H255V0H247Z"></path><path id="MJX-8-TEX-N-6E" d="M41 46H55Q94 46 102 60V68Q102 77 102 91T102 122T103 161T103 203Q103 234 103 269T102 328V351Q99 370 88 376T43 385H25V408Q25 431 27 431L37 432Q47 433 65 434T102 436Q119 437 138 438T167 441T178 442H181V402Q181 364 182 364T187 369T199 384T218 402T247 421T285 437Q305 442 336 442Q450 438 463 329Q464 322 464 190V104Q464 66 466 59T477 49Q498 46 526 46H542V0H534L510 1Q487 2 460 2T422 3Q319 3 310 0H302V46H318Q379 46 379 62Q380 64 380 200Q379 335 378 343Q372 371 358 385T334 402T308 404Q263 404 229 370Q202 343 195 315T187 232V168V108Q187 78 188 68T191 55T200 49Q221 46 249 46H265V0H257L234 1Q210 2 183 2T145 3Q42 3 33 0H25V46H41Z"></path><path id="MJX-8-TEX-N-64" d="M376 495Q376 511 376 535T377 568Q377 613 367 624T316 637H298V660Q298 683 300 683L310 684Q320 685 339 686T376 688Q393 689 413 690T443 693T454 694H457V390Q457 84 458 81Q461 61 472 55T517 46H535V0Q533 0 459 -5T380 -11H373V44L365 37Q307 -11 235 -11Q158 -11 96 50T34 215Q34 315 97 378T244 442Q319 442 376 393V495ZM373 342Q328 405 260 405Q211 405 173 369Q146 341 139 305T131 211Q131 155 138 120T173 59Q203 26 251 26Q322 26 373 103V342Z"></path><path id="MJX-8-TEX-N-6F" d="M28 214Q28 309 93 378T250 448Q340 448 405 380T471 215Q471 120 407 55T250 -10Q153 -10 91 57T28 214ZM250 30Q372 30 372 193V225V250Q372 272 371 288T364 326T348 362T317 390T268 410Q263 411 252 411Q222 411 195 399Q152 377 139 338T126 246V226Q126 130 145 91Q177 30 250 30Z"></path><path id="MJX-8-TEX-N-5F" d="M0 -62V-25H499V-62H0Z"></path><path id="MJX-8-TEX-N-65" d="M28 218Q28 273 48 318T98 391T163 433T229 448Q282 448 320 430T378 380T406 316T415 245Q415 238 408 231H126V216Q126 68 226 36Q246 30 270 30Q312 30 342 62Q359 79 369 104L379 128Q382 131 395 131H398Q415 131 415 121Q415 117 412 108Q393 53 349 21T250 -11Q155 -11 92 58T28 218ZM333 275Q322 403 238 411H236Q228 411 220 410T195 402T166 381T143 340T127 274V267H333V275Z"></path><path id="MJX-8-TEX-N-78" d="M201 0Q189 3 102 3Q26 3 17 0H11V46H25Q48 47 67 52T96 61T121 78T139 96T160 122T180 150L226 210L168 288Q159 301 149 315T133 336T122 351T113 363T107 370T100 376T94 379T88 381T80 383Q74 383 44 385H16V431H23Q59 429 126 429Q219 429 229 431H237V385Q201 381 201 369Q201 367 211 353T239 315T268 274L272 270L297 304Q329 345 329 358Q329 364 327 369T322 376T317 380T310 384L307 385H302V431H309Q324 428 408 428Q487 428 493 431H499V385H492Q443 385 411 368Q394 360 377 341T312 257L296 236L358 151Q424 61 429 57T446 50Q464 46 499 46H516V0H510H502Q494 1 482 1T457 2T432 2T414 3Q403 3 377 3T327 1L304 0H295V46H298Q309 46 320 51T331 63Q331 65 291 120L250 175Q249 174 219 133T185 88Q181 83 181 74Q181 63 188 55T206 46Q208 46 208 23V0H201Z"></path><path id="MJX-8-TEX-I-1D456" d="M184 600Q184 624 203 642T247 661Q265 661 277 649T290 619Q290 596 270 577T226 557Q211 557 198 567T184 600ZM21 287Q21 295 30 318T54 369T98 420T158 442Q197 442 223 419T250 357Q250 340 236 301T196 196T154 83Q149 61 149 51Q149 26 166 26Q175 26 185 29T208 43T235 78T260 137Q263 149 265 151T282 153Q302 153 302 143Q302 135 293 112T268 61T223 11T161 -11Q129 -11 102 10T74 74Q74 91 79 106T122 220Q160 321 166 341T173 380Q173 404 156 404H154Q124 404 99 371T61 287Q60 286 59 284T58 281T56 279T53 278T49 278T41 278H27Q21 284 21 287Z"></path><path id="MJX-8-TEX-N-3D" d="M56 347Q56 360 70 367H707Q722 359 722 347Q722 336 708 328L390 327H72Q56 332 56 347ZM56 153Q56 168 72 173H708Q722 163 722 153Q722 140 707 133H70Q56 140 56 153Z"></path><path id="MJX-8-TEX-S3-230A" d="M246 -949V1450H308V-887H571V-949H246Z"></path><path id="MJX-8-TEX-N-6D" d="M41 46H55Q94 46 102 60V68Q102 77 102 91T102 122T103 161T103 203Q103 234 103 269T102 328V351Q99 370 88 376T43 385H25V408Q25 431 27 431L37 432Q47 433 65 434T102 436Q119 437 138 438T167 441T178 442H181V402Q181 364 182 364T187 369T199 384T218 402T247 421T285 437Q305 442 336 442Q351 442 364 440T387 434T406 426T421 417T432 406T441 395T448 384T452 374T455 366L457 361L460 365Q463 369 466 373T475 384T488 397T503 410T523 422T546 432T572 439T603 442Q729 442 740 329Q741 322 741 190V104Q741 66 743 59T754 49Q775 46 803 46H819V0H811L788 1Q764 2 737 2T699 3Q596 3 587 0H579V46H595Q656 46 656 62Q657 64 657 200Q656 335 655 343Q649 371 635 385T611 402T585 404Q540 404 506 370Q479 343 472 315T464 232V168V108Q464 78 465 68T468 55T477 49Q498 46 526 46H542V0H534L510 1Q487 2 460 2T422 3Q319 3 310 0H302V46H318Q379 46 379 62Q380 64 380 200Q379 335 378 343Q372 371 358 385T334 402T308 404Q263 404 229 370Q202 343 195 315T187 232V168V108Q187 78 188 68T191 55T200 49Q221 46 249 46H265V0H257L234 1Q210 2 183 2T145 3Q42 3 33 0H25V46H41Z"></path><path id="MJX-8-TEX-N-70" d="M36 -148H50Q89 -148 97 -134V-126Q97 -119 97 -107T97 -77T98 -38T98 6T98 55T98 106Q98 140 98 177T98 243T98 296T97 335T97 351Q94 370 83 376T38 385H20V408Q20 431 22 431L32 432Q42 433 61 434T98 436Q115 437 135 438T165 441T176 442H179V416L180 390L188 397Q247 441 326 441Q407 441 464 377T522 216Q522 115 457 52T310 -11Q242 -11 190 33L182 40V-45V-101Q182 -128 184 -134T195 -145Q216 -148 244 -148H260V-194H252L228 -193Q205 -192 178 -192T140 -191Q37 -191 28 -194H20V-148H36ZM424 218Q424 292 390 347T305 402Q234 402 182 337V98Q222 26 294 26Q345 26 384 80T424 218Z"></path><path id="MJX-8-TEX-N-74" d="M27 422Q80 426 109 478T141 600V615H181V431H316V385H181V241Q182 116 182 100T189 68Q203 29 238 29Q282 29 292 100Q293 108 293 146V181H333V146V134Q333 57 291 17Q264 -10 221 -10Q187 -10 162 2T124 33T105 68T98 100Q97 107 97 248V385H18V422H27Z"></path><path id="MJX-8-TEX-N-62" d="M307 -11Q234 -11 168 55L158 37Q156 34 153 28T147 17T143 10L138 1L118 0H98V298Q98 599 97 603Q94 622 83 628T38 637H20V660Q20 683 22 683L32 684Q42 685 61 686T98 688Q115 689 135 690T165 693T176 694H179V543Q179 391 180 391L183 394Q186 397 192 401T207 411T228 421T254 431T286 439T323 442Q401 442 461 379T522 216Q522 115 458 52T307 -11ZM182 98Q182 97 187 90T196 79T206 67T218 55T233 44T250 35T271 29T295 26Q330 26 363 46T412 113Q424 148 424 212Q424 287 412 323Q385 405 300 405Q270 405 239 390T188 347L182 339V98Z"></path><path id="MJX-8-TEX-N-4C" d="M128 622Q121 629 117 631T101 634T58 637H25V683H36Q48 680 182 680Q324 680 348 683H360V637H333Q273 637 258 635T233 622L232 342V129Q232 57 237 52Q243 47 313 47Q384 47 410 53Q470 70 498 110T536 221Q536 226 537 238T540 261T542 272T562 273H582V268Q580 265 568 137T554 5V0H25V46H58Q100 47 109 49T128 61V622Z"></path><path id="MJX-8-TEX-S3-230B" d="M274 -887V1450H336V-949H11V-887H274Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="msub"><g data-mml-node="mtext"><use data-c="77" xlink:href="#MJX-8-TEX-N-77"></use><use data-c="69" xlink:href="#MJX-8-TEX-N-69" transform="translate(722,0)"></use><use data-c="6E" xlink:href="#MJX-8-TEX-N-6E" transform="translate(1000,0)"></use><use data-c="64" xlink:href="#MJX-8-TEX-N-64" transform="translate(1556,0)"></use><use data-c="6F" xlink:href="#MJX-8-TEX-N-6F" transform="translate(2112,0)"></use><use data-c="77" xlink:href="#MJX-8-TEX-N-77" transform="translate(2612,0)"></use><use data-c="5F" xlink:href="#MJX-8-TEX-N-5F" transform="translate(3334,0)"></use><use data-c="69" xlink:href="#MJX-8-TEX-N-69" transform="translate(3834,0)"></use><use data-c="6E" xlink:href="#MJX-8-TEX-N-6E" transform="translate(4112,0)"></use><use data-c="64" xlink:href="#MJX-8-TEX-N-64" transform="translate(4668,0)"></use><use data-c="65" xlink:href="#MJX-8-TEX-N-65" transform="translate(5224,0)"></use><use data-c="78" xlink:href="#MJX-8-TEX-N-78" transform="translate(5668,0)"></use></g><g data-mml-node="mi" transform="translate(6229,-150) scale(0.707)"><use data-c="1D456" xlink:href="#MJX-8-TEX-I-1D456"></use></g></g><g data-mml-node="mo" transform="translate(6800.7,0)"><use data-c="3D" xlink:href="#MJX-8-TEX-N-3D"></use></g><g data-mml-node="mrow" transform="translate(7856.5,0)"><g data-mml-node="mo" transform="translate(0 -0.5)"><use data-c="230A" xlink:href="#MJX-8-TEX-S3-230A"></use></g><g data-mml-node="mfrac" transform="translate(583,0)"><g data-mml-node="msub" transform="translate(220,697.1)"><g data-mml-node="mtext"><use data-c="6D" xlink:href="#MJX-8-TEX-N-6D"></use><use data-c="69" xlink:href="#MJX-8-TEX-N-69" transform="translate(833,0)"></use><use data-c="64" xlink:href="#MJX-8-TEX-N-64" transform="translate(1111,0)"></use><use data-c="70" xlink:href="#MJX-8-TEX-N-70" transform="translate(1667,0)"></use><use data-c="6F" xlink:href="#MJX-8-TEX-N-6F" transform="translate(2223,0)"></use><use data-c="69" xlink:href="#MJX-8-TEX-N-69" transform="translate(2723,0)"></use><use data-c="6E" xlink:href="#MJX-8-TEX-N-6E" transform="translate(3001,0)"></use><use data-c="74" xlink:href="#MJX-8-TEX-N-74" transform="translate(3557,0)"></use></g><g data-mml-node="mi" transform="translate(3979,-229.4) scale(0.707)"><use data-c="1D456" xlink:href="#MJX-8-TEX-I-1D456"></use></g></g><g data-mml-node="mtext" transform="translate(849,-686)"><use data-c="62" xlink:href="#MJX-8-TEX-N-62"></use><use data-c="69" xlink:href="#MJX-8-TEX-N-69" transform="translate(556,0)"></use><use data-c="6E" xlink:href="#MJX-8-TEX-N-6E" transform="translate(834,0)"></use><use data-c="4C" xlink:href="#MJX-8-TEX-N-4C" transform="translate(1390,0)"></use><use data-c="65" xlink:href="#MJX-8-TEX-N-65" transform="translate(2015,0)"></use><use data-c="6E" xlink:href="#MJX-8-TEX-N-6E" transform="translate(2459,0)"></use></g><rect width="4473" height="60" x="120" y="220"></rect></g><g data-mml-node="mo" transform="translate(5296,0) translate(0 -0.5)"><use data-c="230B" xlink:href="#MJX-8-TEX-S3-230B"></use></g></g></g></g></svg><mjx-assistive-mml unselectable="on" display="block"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><msub><mtext>window_index</mtext><mi>i</mi></msub><mo>=</mo><mrow data-mjx-texclass="INNER"><mo data-mjx-texclass="OPEN">⌊</mo><mfrac><msub><mtext>midpoint</mtext><mi>i</mi></msub><mtext>binLen</mtext></mfrac><mo data-mjx-texclass="CLOSE">⌋</mo></mrow></math></mjx-assistive-mml></mjx-container></div></div><p><mjx-container class="MathJax" jax="SVG" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="2.638ex" height="2.262ex" role="img" focusable="false" viewBox="0 -750 1166 1000" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -0.566ex;"><defs><path id="MJX-35-TEX-N-230A" d="M174 734Q174 735 175 737T177 740T180 744T184 747T189 749T196 750Q206 748 214 735V-210H310H373Q401 -210 411 -213T422 -230T411 -247T369 -251Q362 -251 338 -251T298 -250H190Q178 -246 174 -234V734Z"></path><path id="MJX-35-TEX-N-22C5" d="M78 250Q78 274 95 292T138 310Q162 310 180 294T199 251Q199 226 182 208T139 190T96 207T78 250Z"></path><path id="MJX-35-TEX-N-230B" d="M229 734Q229 735 230 737T232 740T235 744T239 747T244 749T251 750Q262 748 269 735V-235Q266 -240 256 -249L147 -250H77Q43 -250 32 -247T21 -230T32 -213T72 -209Q79 -209 99 -209T133 -210H229V734Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="mrow"><g data-mml-node="mo"><use data-c="230A" xlink:href="#MJX-35-TEX-N-230A"></use></g><g data-mml-node="mo" transform="translate(444,0)"><use data-c="22C5" xlink:href="#MJX-35-TEX-N-22C5"></use></g><g data-mml-node="mo" transform="translate(722,0)"><use data-c="230B" xlink:href="#MJX-35-TEX-N-230B"></use></g></g></g></g></svg><mjx-assistive-mml unselectable="on" display="inline"><math xmlns="http://www.w3.org/1998/Math/MathML"><mrow data-mjx-texclass="INNER"><mo data-mjx-texclass="OPEN">⌊</mo><mo>⋅</mo><mo data-mjx-texclass="CLOSE">⌋</mo></mrow></math></mjx-assistive-mml></mjx-container><script type="math/tex">\left\lfloor \cdot \right\rfloor</script><span> 表示向下取整（floor）</span></p></li><li><p><span>计算该索引窗口的起始位置</span></p><div contenteditable="false" spellcheck="false" class="mathjax-block md-end-block md-math-block md-rawblock" id="mathjax-n2786" cid="n2786" mdtype="math_block" data-math-tag-before="0" data-math-tag-after="0" data-math-labels="[]"><div class="md-math-container"><mjx-container class="MathJax" jax="SVG" display="true" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="41.445ex" height="1.927ex" role="img" focusable="false" viewBox="0 -694 18318.9 851.8" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -0.357ex;"><defs><path id="MJX-9-TEX-N-77" d="M90 368Q84 378 76 380T40 385H18V431H24L43 430Q62 430 84 429T116 428Q206 428 221 431H229V385H215Q177 383 177 368Q177 367 221 239L265 113L339 328L333 345Q323 374 316 379Q308 384 278 385H258V431H264Q270 428 348 428Q439 428 454 431H461V385H452Q404 385 404 369Q404 366 418 324T449 234T481 143L496 100L537 219Q579 341 579 347Q579 363 564 373T530 385H522V431H529Q541 428 624 428Q692 428 698 431H703V385H697Q696 385 691 385T682 384Q635 377 619 334L559 161Q546 124 528 71Q508 12 503 1T487 -11H479Q460 -11 456 -4Q455 -3 407 133L361 267Q359 263 266 -4Q261 -11 243 -11H238Q225 -11 220 -3L90 368Z"></path><path id="MJX-9-TEX-N-69" d="M69 609Q69 637 87 653T131 669Q154 667 171 652T188 609Q188 579 171 564T129 549Q104 549 87 564T69 609ZM247 0Q232 3 143 3Q132 3 106 3T56 1L34 0H26V46H42Q70 46 91 49Q100 53 102 60T104 102V205V293Q104 345 102 359T88 378Q74 385 41 385H30V408Q30 431 32 431L42 432Q52 433 70 434T106 436Q123 437 142 438T171 441T182 442H185V62Q190 52 197 50T232 46H255V0H247Z"></path><path id="MJX-9-TEX-N-6E" d="M41 46H55Q94 46 102 60V68Q102 77 102 91T102 122T103 161T103 203Q103 234 103 269T102 328V351Q99 370 88 376T43 385H25V408Q25 431 27 431L37 432Q47 433 65 434T102 436Q119 437 138 438T167 441T178 442H181V402Q181 364 182 364T187 369T199 384T218 402T247 421T285 437Q305 442 336 442Q450 438 463 329Q464 322 464 190V104Q464 66 466 59T477 49Q498 46 526 46H542V0H534L510 1Q487 2 460 2T422 3Q319 3 310 0H302V46H318Q379 46 379 62Q380 64 380 200Q379 335 378 343Q372 371 358 385T334 402T308 404Q263 404 229 370Q202 343 195 315T187 232V168V108Q187 78 188 68T191 55T200 49Q221 46 249 46H265V0H257L234 1Q210 2 183 2T145 3Q42 3 33 0H25V46H41Z"></path><path id="MJX-9-TEX-N-64" d="M376 495Q376 511 376 535T377 568Q377 613 367 624T316 637H298V660Q298 683 300 683L310 684Q320 685 339 686T376 688Q393 689 413 690T443 693T454 694H457V390Q457 84 458 81Q461 61 472 55T517 46H535V0Q533 0 459 -5T380 -11H373V44L365 37Q307 -11 235 -11Q158 -11 96 50T34 215Q34 315 97 378T244 442Q319 442 376 393V495ZM373 342Q328 405 260 405Q211 405 173 369Q146 341 139 305T131 211Q131 155 138 120T173 59Q203 26 251 26Q322 26 373 103V342Z"></path><path id="MJX-9-TEX-N-6F" d="M28 214Q28 309 93 378T250 448Q340 448 405 380T471 215Q471 120 407 55T250 -10Q153 -10 91 57T28 214ZM250 30Q372 30 372 193V225V250Q372 272 371 288T364 326T348 362T317 390T268 410Q263 411 252 411Q222 411 195 399Q152 377 139 338T126 246V226Q126 130 145 91Q177 30 250 30Z"></path><path id="MJX-9-TEX-N-5F" d="M0 -62V-25H499V-62H0Z"></path><path id="MJX-9-TEX-N-73" d="M295 316Q295 356 268 385T190 414Q154 414 128 401Q98 382 98 349Q97 344 98 336T114 312T157 287Q175 282 201 278T245 269T277 256Q294 248 310 236T342 195T359 133Q359 71 321 31T198 -10H190Q138 -10 94 26L86 19L77 10Q71 4 65 -1L54 -11H46H42Q39 -11 33 -5V74V132Q33 153 35 157T45 162H54Q66 162 70 158T75 146T82 119T101 77Q136 26 198 26Q295 26 295 104Q295 133 277 151Q257 175 194 187T111 210Q75 227 54 256T33 318Q33 357 50 384T93 424T143 442T187 447H198Q238 447 268 432L283 424L292 431Q302 440 314 448H322H326Q329 448 335 442V310L329 304H301Q295 310 295 316Z"></path><path id="MJX-9-TEX-N-74" d="M27 422Q80 426 109 478T141 600V615H181V431H316V385H181V241Q182 116 182 100T189 68Q203 29 238 29Q282 29 292 100Q293 108 293 146V181H333V146V134Q333 57 291 17Q264 -10 221 -10Q187 -10 162 2T124 33T105 68T98 100Q97 107 97 248V385H18V422H27Z"></path><path id="MJX-9-TEX-N-61" d="M137 305T115 305T78 320T63 359Q63 394 97 421T218 448Q291 448 336 416T396 340Q401 326 401 309T402 194V124Q402 76 407 58T428 40Q443 40 448 56T453 109V145H493V106Q492 66 490 59Q481 29 455 12T400 -6T353 12T329 54V58L327 55Q325 52 322 49T314 40T302 29T287 17T269 6T247 -2T221 -8T190 -11Q130 -11 82 20T34 107Q34 128 41 147T68 188T116 225T194 253T304 268H318V290Q318 324 312 340Q290 411 215 411Q197 411 181 410T156 406T148 403Q170 388 170 359Q170 334 154 320ZM126 106Q126 75 150 51T209 26Q247 26 276 49T315 109Q317 116 318 175Q318 233 317 233Q309 233 296 232T251 223T193 203T147 166T126 106Z"></path><path id="MJX-9-TEX-N-72" d="M36 46H50Q89 46 97 60V68Q97 77 97 91T98 122T98 161T98 203Q98 234 98 269T98 328L97 351Q94 370 83 376T38 385H20V408Q20 431 22 431L32 432Q42 433 60 434T96 436Q112 437 131 438T160 441T171 442H174V373Q213 441 271 441H277Q322 441 343 419T364 373Q364 352 351 337T313 322Q288 322 276 338T263 372Q263 381 265 388T270 400T273 405Q271 407 250 401Q234 393 226 386Q179 341 179 207V154Q179 141 179 127T179 101T180 81T180 66V61Q181 59 183 57T188 54T193 51T200 49T207 48T216 47T225 47T235 46T245 46H276V0H267Q249 3 140 3Q37 3 28 0H20V46H36Z"></path><path id="MJX-9-TEX-I-1D456" d="M184 600Q184 624 203 642T247 661Q265 661 277 649T290 619Q290 596 270 577T226 557Q211 557 198 567T184 600ZM21 287Q21 295 30 318T54 369T98 420T158 442Q197 442 223 419T250 357Q250 340 236 301T196 196T154 83Q149 61 149 51Q149 26 166 26Q175 26 185 29T208 43T235 78T260 137Q263 149 265 151T282 153Q302 153 302 143Q302 135 293 112T268 61T223 11T161 -11Q129 -11 102 10T74 74Q74 91 79 106T122 220Q160 321 166 341T173 380Q173 404 156 404H154Q124 404 99 371T61 287Q60 286 59 284T58 281T56 279T53 278T49 278T41 278H27Q21 284 21 287Z"></path><path id="MJX-9-TEX-N-3D" d="M56 347Q56 360 70 367H707Q722 359 722 347Q722 336 708 328L390 327H72Q56 332 56 347ZM56 153Q56 168 72 173H708Q722 163 722 153Q722 140 707 133H70Q56 140 56 153Z"></path><path id="MJX-9-TEX-N-65" d="M28 218Q28 273 48 318T98 391T163 433T229 448Q282 448 320 430T378 380T406 316T415 245Q415 238 408 231H126V216Q126 68 226 36Q246 30 270 30Q312 30 342 62Q359 79 369 104L379 128Q382 131 395 131H398Q415 131 415 121Q415 117 412 108Q393 53 349 21T250 -11Q155 -11 92 58T28 218ZM333 275Q322 403 238 411H236Q228 411 220 410T195 402T166 381T143 340T127 274V267H333V275Z"></path><path id="MJX-9-TEX-N-78" d="M201 0Q189 3 102 3Q26 3 17 0H11V46H25Q48 47 67 52T96 61T121 78T139 96T160 122T180 150L226 210L168 288Q159 301 149 315T133 336T122 351T113 363T107 370T100 376T94 379T88 381T80 383Q74 383 44 385H16V431H23Q59 429 126 429Q219 429 229 431H237V385Q201 381 201 369Q201 367 211 353T239 315T268 274L272 270L297 304Q329 345 329 358Q329 364 327 369T322 376T317 380T310 384L307 385H302V431H309Q324 428 408 428Q487 428 493 431H499V385H492Q443 385 411 368Q394 360 377 341T312 257L296 236L358 151Q424 61 429 57T446 50Q464 46 499 46H516V0H510H502Q494 1 482 1T457 2T432 2T414 3Q403 3 377 3T327 1L304 0H295V46H298Q309 46 320 51T331 63Q331 65 291 120L250 175Q249 174 219 133T185 88Q181 83 181 74Q181 63 188 55T206 46Q208 46 208 23V0H201Z"></path><path id="MJX-9-TEX-N-D7" d="M630 29Q630 9 609 9Q604 9 587 25T493 118L389 222L284 117Q178 13 175 11Q171 9 168 9Q160 9 154 15T147 29Q147 36 161 51T255 146L359 250L255 354Q174 435 161 449T147 471Q147 480 153 485T168 490Q173 490 175 489Q178 487 284 383L389 278L493 382Q570 459 587 475T609 491Q630 491 630 471Q630 464 620 453T522 355L418 250L522 145Q606 61 618 48T630 29Z"></path><path id="MJX-9-TEX-N-62" d="M307 -11Q234 -11 168 55L158 37Q156 34 153 28T147 17T143 10L138 1L118 0H98V298Q98 599 97 603Q94 622 83 628T38 637H20V660Q20 683 22 683L32 684Q42 685 61 686T98 688Q115 689 135 690T165 693T176 694H179V543Q179 391 180 391L183 394Q186 397 192 401T207 411T228 421T254 431T286 439T323 442Q401 442 461 379T522 216Q522 115 458 52T307 -11ZM182 98Q182 97 187 90T196 79T206 67T218 55T233 44T250 35T271 29T295 26Q330 26 363 46T412 113Q424 148 424 212Q424 287 412 323Q385 405 300 405Q270 405 239 390T188 347L182 339V98Z"></path><path id="MJX-9-TEX-N-4C" d="M128 622Q121 629 117 631T101 634T58 637H25V683H36Q48 680 182 680Q324 680 348 683H360V637H333Q273 637 258 635T233 622L232 342V129Q232 57 237 52Q243 47 313 47Q384 47 410 53Q470 70 498 110T536 221Q536 226 537 238T540 261T542 272T562 273H582V268Q580 265 568 137T554 5V0H25V46H58Q100 47 109 49T128 61V622Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="msub"><g data-mml-node="mtext"><use data-c="77" xlink:href="#MJX-9-TEX-N-77"></use><use data-c="69" xlink:href="#MJX-9-TEX-N-69" transform="translate(722,0)"></use><use data-c="6E" xlink:href="#MJX-9-TEX-N-6E" transform="translate(1000,0)"></use><use data-c="64" xlink:href="#MJX-9-TEX-N-64" transform="translate(1556,0)"></use><use data-c="6F" xlink:href="#MJX-9-TEX-N-6F" transform="translate(2112,0)"></use><use data-c="77" xlink:href="#MJX-9-TEX-N-77" transform="translate(2612,0)"></use><use data-c="5F" xlink:href="#MJX-9-TEX-N-5F" transform="translate(3334,0)"></use><use data-c="73" xlink:href="#MJX-9-TEX-N-73" transform="translate(3834,0)"></use><use data-c="74" xlink:href="#MJX-9-TEX-N-74" transform="translate(4228,0)"></use><use data-c="61" xlink:href="#MJX-9-TEX-N-61" transform="translate(4617,0)"></use><use data-c="72" xlink:href="#MJX-9-TEX-N-72" transform="translate(5117,0)"></use><use data-c="74" xlink:href="#MJX-9-TEX-N-74" transform="translate(5509,0)"></use></g><g data-mml-node="mi" transform="translate(5931,-150) scale(0.707)"><use data-c="1D456" xlink:href="#MJX-9-TEX-I-1D456"></use></g></g><g data-mml-node="mo" transform="translate(6502.7,0)"><use data-c="3D" xlink:href="#MJX-9-TEX-N-3D"></use></g><g data-mml-node="msub" transform="translate(7558.5,0)"><g data-mml-node="mtext"><use data-c="77" xlink:href="#MJX-9-TEX-N-77"></use><use data-c="69" xlink:href="#MJX-9-TEX-N-69" transform="translate(722,0)"></use><use data-c="6E" xlink:href="#MJX-9-TEX-N-6E" transform="translate(1000,0)"></use><use data-c="64" xlink:href="#MJX-9-TEX-N-64" transform="translate(1556,0)"></use><use data-c="6F" xlink:href="#MJX-9-TEX-N-6F" transform="translate(2112,0)"></use><use data-c="77" xlink:href="#MJX-9-TEX-N-77" transform="translate(2612,0)"></use><use data-c="5F" xlink:href="#MJX-9-TEX-N-5F" transform="translate(3334,0)"></use><use data-c="69" xlink:href="#MJX-9-TEX-N-69" transform="translate(3834,0)"></use><use data-c="6E" xlink:href="#MJX-9-TEX-N-6E" transform="translate(4112,0)"></use><use data-c="64" xlink:href="#MJX-9-TEX-N-64" transform="translate(4668,0)"></use><use data-c="65" xlink:href="#MJX-9-TEX-N-65" transform="translate(5224,0)"></use><use data-c="78" xlink:href="#MJX-9-TEX-N-78" transform="translate(5668,0)"></use></g><g data-mml-node="mi" transform="translate(6229,-150) scale(0.707)"><use data-c="1D456" xlink:href="#MJX-9-TEX-I-1D456"></use></g></g><g data-mml-node="mo" transform="translate(14303.7,0)"><use data-c="D7" xlink:href="#MJX-9-TEX-N-D7"></use></g><g data-mml-node="mtext" transform="translate(15303.9,0)"><use data-c="62" xlink:href="#MJX-9-TEX-N-62"></use><use data-c="69" xlink:href="#MJX-9-TEX-N-69" transform="translate(556,0)"></use><use data-c="6E" xlink:href="#MJX-9-TEX-N-6E" transform="translate(834,0)"></use><use data-c="4C" xlink:href="#MJX-9-TEX-N-4C" transform="translate(1390,0)"></use><use data-c="65" xlink:href="#MJX-9-TEX-N-65" transform="translate(2015,0)"></use><use data-c="6E" xlink:href="#MJX-9-TEX-N-6E" transform="translate(2459,0)"></use></g></g></g></svg><mjx-assistive-mml unselectable="on" display="block"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><msub><mtext>window_start</mtext><mi>i</mi></msub><mo>=</mo><msub><mtext>window_index</mtext><mi>i</mi></msub><mo>×</mo><mtext>binLen</mtext></math></mjx-assistive-mml></mjx-container></div></div></li><li><p><span>计算该索引窗口的中心位置</span></p><div contenteditable="false" spellcheck="false" class="mathjax-block md-end-block md-math-block md-rawblock" id="mathjax-n2789" cid="n2789" mdtype="math_block" data-math-tag-before="0" data-math-tag-after="0" data-math-labels="[]"><div class="md-math-container"><mjx-container class="MathJax" jax="SVG" display="true" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="44.267ex" height="4.652ex" role="img" focusable="false" viewBox="0 -1370 19565.9 2056" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -1.552ex;"><defs><path id="MJX-10-TEX-N-24" d="M162 187Q162 164 146 149T109 133H103V130Q108 115 115 105Q122 92 131 82T150 64T170 52T190 44T206 40T220 37L227 36V313Q190 320 162 335Q116 358 86 404T55 508Q55 567 85 614T165 685Q186 696 225 704H227V750H273V704L286 703Q369 690 413 631Q441 588 444 531Q444 514 443 509Q439 490 425 479T391 468Q368 468 353 483T337 522Q337 546 353 560T390 575L394 576V578Q386 599 372 614T342 637T314 649T288 656L273 658V408L288 405Q329 394 355 376Q396 348 420 300T444 199Q444 130 408 76T313 1Q286 -9 276 -9H273V-56H227V-10H221Q202 -6 193 -4T155 11T108 41T74 94T55 176V182Q55 227 95 238Q103 240 108 240Q129 240 145 226T162 187ZM225 657Q219 657 204 651T169 632T135 594T121 538Q121 512 131 491T156 457T187 435T213 423T227 420V539Q227 657 225 657ZM378 169Q378 230 339 265T274 301Q273 301 273 169V37Q324 50 351 87T378 169Z"></path><path id="MJX-10-TEX-N-77" d="M90 368Q84 378 76 380T40 385H18V431H24L43 430Q62 430 84 429T116 428Q206 428 221 431H229V385H215Q177 383 177 368Q177 367 221 239L265 113L339 328L333 345Q323 374 316 379Q308 384 278 385H258V431H264Q270 428 348 428Q439 428 454 431H461V385H452Q404 385 404 369Q404 366 418 324T449 234T481 143L496 100L537 219Q579 341 579 347Q579 363 564 373T530 385H522V431H529Q541 428 624 428Q692 428 698 431H703V385H697Q696 385 691 385T682 384Q635 377 619 334L559 161Q546 124 528 71Q508 12 503 1T487 -11H479Q460 -11 456 -4Q455 -3 407 133L361 267Q359 263 266 -4Q261 -11 243 -11H238Q225 -11 220 -3L90 368Z"></path><path id="MJX-10-TEX-N-69" d="M69 609Q69 637 87 653T131 669Q154 667 171 652T188 609Q188 579 171 564T129 549Q104 549 87 564T69 609ZM247 0Q232 3 143 3Q132 3 106 3T56 1L34 0H26V46H42Q70 46 91 49Q100 53 102 60T104 102V205V293Q104 345 102 359T88 378Q74 385 41 385H30V408Q30 431 32 431L42 432Q52 433 70 434T106 436Q123 437 142 438T171 441T182 442H185V62Q190 52 197 50T232 46H255V0H247Z"></path><path id="MJX-10-TEX-N-6E" d="M41 46H55Q94 46 102 60V68Q102 77 102 91T102 122T103 161T103 203Q103 234 103 269T102 328V351Q99 370 88 376T43 385H25V408Q25 431 27 431L37 432Q47 433 65 434T102 436Q119 437 138 438T167 441T178 442H181V402Q181 364 182 364T187 369T199 384T218 402T247 421T285 437Q305 442 336 442Q450 438 463 329Q464 322 464 190V104Q464 66 466 59T477 49Q498 46 526 46H542V0H534L510 1Q487 2 460 2T422 3Q319 3 310 0H302V46H318Q379 46 379 62Q380 64 380 200Q379 335 378 343Q372 371 358 385T334 402T308 404Q263 404 229 370Q202 343 195 315T187 232V168V108Q187 78 188 68T191 55T200 49Q221 46 249 46H265V0H257L234 1Q210 2 183 2T145 3Q42 3 33 0H25V46H41Z"></path><path id="MJX-10-TEX-N-64" d="M376 495Q376 511 376 535T377 568Q377 613 367 624T316 637H298V660Q298 683 300 683L310 684Q320 685 339 686T376 688Q393 689 413 690T443 693T454 694H457V390Q457 84 458 81Q461 61 472 55T517 46H535V0Q533 0 459 -5T380 -11H373V44L365 37Q307 -11 235 -11Q158 -11 96 50T34 215Q34 315 97 378T244 442Q319 442 376 393V495ZM373 342Q328 405 260 405Q211 405 173 369Q146 341 139 305T131 211Q131 155 138 120T173 59Q203 26 251 26Q322 26 373 103V342Z"></path><path id="MJX-10-TEX-N-6F" d="M28 214Q28 309 93 378T250 448Q340 448 405 380T471 215Q471 120 407 55T250 -10Q153 -10 91 57T28 214ZM250 30Q372 30 372 193V225V250Q372 272 371 288T364 326T348 362T317 390T268 410Q263 411 252 411Q222 411 195 399Q152 377 139 338T126 246V226Q126 130 145 91Q177 30 250 30Z"></path><path id="MJX-10-TEX-N-5F" d="M0 -62V-25H499V-62H0Z"></path><path id="MJX-10-TEX-N-63" d="M370 305T349 305T313 320T297 358Q297 381 312 396Q317 401 317 402T307 404Q281 408 258 408Q209 408 178 376Q131 329 131 219Q131 137 162 90Q203 29 272 29Q313 29 338 55T374 117Q376 125 379 127T395 129H409Q415 123 415 120Q415 116 411 104T395 71T366 33T318 2T249 -11Q163 -11 99 53T34 214Q34 318 99 383T250 448T370 421T404 357Q404 334 387 320Z"></path><path id="MJX-10-TEX-N-65" d="M28 218Q28 273 48 318T98 391T163 433T229 448Q282 448 320 430T378 380T406 316T415 245Q415 238 408 231H126V216Q126 68 226 36Q246 30 270 30Q312 30 342 62Q359 79 369 104L379 128Q382 131 395 131H398Q415 131 415 121Q415 117 412 108Q393 53 349 21T250 -11Q155 -11 92 58T28 218ZM333 275Q322 403 238 411H236Q228 411 220 410T195 402T166 381T143 340T127 274V267H333V275Z"></path><path id="MJX-10-TEX-N-74" d="M27 422Q80 426 109 478T141 600V615H181V431H316V385H181V241Q182 116 182 100T189 68Q203 29 238 29Q282 29 292 100Q293 108 293 146V181H333V146V134Q333 57 291 17Q264 -10 221 -10Q187 -10 162 2T124 33T105 68T98 100Q97 107 97 248V385H18V422H27Z"></path><path id="MJX-10-TEX-N-72" d="M36 46H50Q89 46 97 60V68Q97 77 97 91T98 122T98 161T98 203Q98 234 98 269T98 328L97 351Q94 370 83 376T38 385H20V408Q20 431 22 431L32 432Q42 433 60 434T96 436Q112 437 131 438T160 441T171 442H174V373Q213 441 271 441H277Q322 441 343 419T364 373Q364 352 351 337T313 322Q288 322 276 338T263 372Q263 381 265 388T270 400T273 405Q271 407 250 401Q234 393 226 386Q179 341 179 207V154Q179 141 179 127T179 101T180 81T180 66V61Q181 59 183 57T188 54T193 51T200 49T207 48T216 47T225 47T235 46T245 46H276V0H267Q249 3 140 3Q37 3 28 0H20V46H36Z"></path><path id="MJX-10-TEX-I-1D456" d="M184 600Q184 624 203 642T247 661Q265 661 277 649T290 619Q290 596 270 577T226 557Q211 557 198 567T184 600ZM21 287Q21 295 30 318T54 369T98 420T158 442Q197 442 223 419T250 357Q250 340 236 301T196 196T154 83Q149 61 149 51Q149 26 166 26Q175 26 185 29T208 43T235 78T260 137Q263 149 265 151T282 153Q302 153 302 143Q302 135 293 112T268 61T223 11T161 -11Q129 -11 102 10T74 74Q74 91 79 106T122 220Q160 321 166 341T173 380Q173 404 156 404H154Q124 404 99 371T61 287Q60 286 59 284T58 281T56 279T53 278T49 278T41 278H27Q21 284 21 287Z"></path><path id="MJX-10-TEX-N-3D" d="M56 347Q56 360 70 367H707Q722 359 722 347Q722 336 708 328L390 327H72Q56 332 56 347ZM56 153Q56 168 72 173H708Q722 163 722 153Q722 140 707 133H70Q56 140 56 153Z"></path><path id="MJX-10-TEX-N-73" d="M295 316Q295 356 268 385T190 414Q154 414 128 401Q98 382 98 349Q97 344 98 336T114 312T157 287Q175 282 201 278T245 269T277 256Q294 248 310 236T342 195T359 133Q359 71 321 31T198 -10H190Q138 -10 94 26L86 19L77 10Q71 4 65 -1L54 -11H46H42Q39 -11 33 -5V74V132Q33 153 35 157T45 162H54Q66 162 70 158T75 146T82 119T101 77Q136 26 198 26Q295 26 295 104Q295 133 277 151Q257 175 194 187T111 210Q75 227 54 256T33 318Q33 357 50 384T93 424T143 442T187 447H198Q238 447 268 432L283 424L292 431Q302 440 314 448H322H326Q329 448 335 442V310L329 304H301Q295 310 295 316Z"></path><path id="MJX-10-TEX-N-61" d="M137 305T115 305T78 320T63 359Q63 394 97 421T218 448Q291 448 336 416T396 340Q401 326 401 309T402 194V124Q402 76 407 58T428 40Q443 40 448 56T453 109V145H493V106Q492 66 490 59Q481 29 455 12T400 -6T353 12T329 54V58L327 55Q325 52 322 49T314 40T302 29T287 17T269 6T247 -2T221 -8T190 -11Q130 -11 82 20T34 107Q34 128 41 147T68 188T116 225T194 253T304 268H318V290Q318 324 312 340Q290 411 215 411Q197 411 181 410T156 406T148 403Q170 388 170 359Q170 334 154 320ZM126 106Q126 75 150 51T209 26Q247 26 276 49T315 109Q317 116 318 175Q318 233 317 233Q309 233 296 232T251 223T193 203T147 166T126 106Z"></path><path id="MJX-10-TEX-N-2B" d="M56 237T56 250T70 270H369V420L370 570Q380 583 389 583Q402 583 409 568V270H707Q722 262 722 250T707 230H409V-68Q401 -82 391 -82H389H387Q375 -82 369 -68V230H70Q56 237 56 250Z"></path><path id="MJX-10-TEX-N-62" d="M307 -11Q234 -11 168 55L158 37Q156 34 153 28T147 17T143 10L138 1L118 0H98V298Q98 599 97 603Q94 622 83 628T38 637H20V660Q20 683 22 683L32 684Q42 685 61 686T98 688Q115 689 135 690T165 693T176 694H179V543Q179 391 180 391L183 394Q186 397 192 401T207 411T228 421T254 431T286 439T323 442Q401 442 461 379T522 216Q522 115 458 52T307 -11ZM182 98Q182 97 187 90T196 79T206 67T218 55T233 44T250 35T271 29T295 26Q330 26 363 46T412 113Q424 148 424 212Q424 287 412 323Q385 405 300 405Q270 405 239 390T188 347L182 339V98Z"></path><path id="MJX-10-TEX-N-4C" d="M128 622Q121 629 117 631T101 634T58 637H25V683H36Q48 680 182 680Q324 680 348 683H360V637H333Q273 637 258 635T233 622L232 342V129Q232 57 237 52Q243 47 313 47Q384 47 410 53Q470 70 498 110T536 221Q536 226 537 238T540 261T542 272T562 273H582V268Q580 265 568 137T554 5V0H25V46H58Q100 47 109 49T128 61V622Z"></path><path id="MJX-10-TEX-N-32" d="M109 429Q82 429 66 447T50 491Q50 562 103 614T235 666Q326 666 387 610T449 465Q449 422 429 383T381 315T301 241Q265 210 201 149L142 93L218 92Q375 92 385 97Q392 99 409 186V189H449V186Q448 183 436 95T421 3V0H50V19V31Q50 38 56 46T86 81Q115 113 136 137Q145 147 170 174T204 211T233 244T261 278T284 308T305 340T320 369T333 401T340 431T343 464Q343 527 309 573T212 619Q179 619 154 602T119 569T109 550Q109 549 114 549Q132 549 151 535T170 489Q170 464 154 447T109 429Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="TeXAtom" data-mjx-texclass="ORD"><g data-mml-node="mo"><use data-c="24" xlink:href="#MJX-10-TEX-N-24"></use></g></g><g data-mml-node="msub" transform="translate(500,0)"><g data-mml-node="mtext"><use data-c="77" xlink:href="#MJX-10-TEX-N-77"></use><use data-c="69" xlink:href="#MJX-10-TEX-N-69" transform="translate(722,0)"></use><use data-c="6E" xlink:href="#MJX-10-TEX-N-6E" transform="translate(1000,0)"></use><use data-c="64" xlink:href="#MJX-10-TEX-N-64" transform="translate(1556,0)"></use><use data-c="6F" xlink:href="#MJX-10-TEX-N-6F" transform="translate(2112,0)"></use><use data-c="77" xlink:href="#MJX-10-TEX-N-77" transform="translate(2612,0)"></use><use data-c="5F" xlink:href="#MJX-10-TEX-N-5F" transform="translate(3334,0)"></use><use data-c="63" xlink:href="#MJX-10-TEX-N-63" transform="translate(3834,0)"></use><use data-c="65" xlink:href="#MJX-10-TEX-N-65" transform="translate(4278,0)"></use><use data-c="6E" xlink:href="#MJX-10-TEX-N-6E" transform="translate(4722,0)"></use><use data-c="74" xlink:href="#MJX-10-TEX-N-74" transform="translate(5278,0)"></use><use data-c="65" xlink:href="#MJX-10-TEX-N-65" transform="translate(5667,0)"></use><use data-c="72" xlink:href="#MJX-10-TEX-N-72" transform="translate(6111,0)"></use></g><g data-mml-node="mi" transform="translate(6536,-150) scale(0.707)"><use data-c="1D456" xlink:href="#MJX-10-TEX-I-1D456"></use></g></g><g data-mml-node="mo" transform="translate(7607.7,0)"><use data-c="3D" xlink:href="#MJX-10-TEX-N-3D"></use></g><g data-mml-node="msub" transform="translate(8663.5,0)"><g data-mml-node="mtext"><use data-c="77" xlink:href="#MJX-10-TEX-N-77"></use><use data-c="69" xlink:href="#MJX-10-TEX-N-69" transform="translate(722,0)"></use><use data-c="6E" xlink:href="#MJX-10-TEX-N-6E" transform="translate(1000,0)"></use><use data-c="64" xlink:href="#MJX-10-TEX-N-64" transform="translate(1556,0)"></use><use data-c="6F" xlink:href="#MJX-10-TEX-N-6F" transform="translate(2112,0)"></use><use data-c="77" xlink:href="#MJX-10-TEX-N-77" transform="translate(2612,0)"></use><use data-c="5F" xlink:href="#MJX-10-TEX-N-5F" transform="translate(3334,0)"></use><use data-c="73" xlink:href="#MJX-10-TEX-N-73" transform="translate(3834,0)"></use><use data-c="74" xlink:href="#MJX-10-TEX-N-74" transform="translate(4228,0)"></use><use data-c="61" xlink:href="#MJX-10-TEX-N-61" transform="translate(4617,0)"></use><use data-c="72" xlink:href="#MJX-10-TEX-N-72" transform="translate(5117,0)"></use><use data-c="74" xlink:href="#MJX-10-TEX-N-74" transform="translate(5509,0)"></use></g><g data-mml-node="mi" transform="translate(5931,-150) scale(0.707)"><use data-c="1D456" xlink:href="#MJX-10-TEX-I-1D456"></use></g></g><g data-mml-node="mo" transform="translate(15110.7,0)"><use data-c="2B" xlink:href="#MJX-10-TEX-N-2B"></use></g><g data-mml-node="mfrac" transform="translate(16110.9,0)"><g data-mml-node="mtext" transform="translate(220,676)"><use data-c="62" xlink:href="#MJX-10-TEX-N-62"></use><use data-c="69" xlink:href="#MJX-10-TEX-N-69" transform="translate(556,0)"></use><use data-c="6E" xlink:href="#MJX-10-TEX-N-6E" transform="translate(834,0)"></use><use data-c="4C" xlink:href="#MJX-10-TEX-N-4C" transform="translate(1390,0)"></use><use data-c="65" xlink:href="#MJX-10-TEX-N-65" transform="translate(2015,0)"></use><use data-c="6E" xlink:href="#MJX-10-TEX-N-6E" transform="translate(2459,0)"></use></g><g data-mml-node="mn" transform="translate(1477.5,-686)"><use data-c="32" xlink:href="#MJX-10-TEX-N-32"></use></g><rect width="3215" height="60" x="120" y="220"></rect></g></g></g></svg><mjx-assistive-mml unselectable="on" display="block"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><mrow data-mjx-texclass="ORD"><mo>$</mo></mrow><msub><mtext>window_center</mtext><mi>i</mi></msub><mo>=</mo><msub><mtext>window_start</mtext><mi>i</mi></msub><mo>+</mo><mfrac><mtext>binLen</mtext><mn>2</mn></mfrac></math></mjx-assistive-mml></mjx-container></div></div></li><li><p><span>把上面公式合在一起</span></p><div contenteditable="false" spellcheck="false" class="mathjax-block md-end-block md-math-block md-rawblock" id="mathjax-n2792" cid="n2792" mdtype="math_block" data-math-tag-before="0" data-math-tag-after="0" data-math-labels="[]"><div class="md-math-container"><mjx-container class="MathJax" jax="SVG" display="true" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="55.187ex" height="6.785ex" role="img" focusable="false" viewBox="0 -1749.5 24392.6 2999" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -2.827ex;"><defs><path id="MJX-11-TEX-N-77" d="M90 368Q84 378 76 380T40 385H18V431H24L43 430Q62 430 84 429T116 428Q206 428 221 431H229V385H215Q177 383 177 368Q177 367 221 239L265 113L339 328L333 345Q323 374 316 379Q308 384 278 385H258V431H264Q270 428 348 428Q439 428 454 431H461V385H452Q404 385 404 369Q404 366 418 324T449 234T481 143L496 100L537 219Q579 341 579 347Q579 363 564 373T530 385H522V431H529Q541 428 624 428Q692 428 698 431H703V385H697Q696 385 691 385T682 384Q635 377 619 334L559 161Q546 124 528 71Q508 12 503 1T487 -11H479Q460 -11 456 -4Q455 -3 407 133L361 267Q359 263 266 -4Q261 -11 243 -11H238Q225 -11 220 -3L90 368Z"></path><path id="MJX-11-TEX-N-69" d="M69 609Q69 637 87 653T131 669Q154 667 171 652T188 609Q188 579 171 564T129 549Q104 549 87 564T69 609ZM247 0Q232 3 143 3Q132 3 106 3T56 1L34 0H26V46H42Q70 46 91 49Q100 53 102 60T104 102V205V293Q104 345 102 359T88 378Q74 385 41 385H30V408Q30 431 32 431L42 432Q52 433 70 434T106 436Q123 437 142 438T171 441T182 442H185V62Q190 52 197 50T232 46H255V0H247Z"></path><path id="MJX-11-TEX-N-6E" d="M41 46H55Q94 46 102 60V68Q102 77 102 91T102 122T103 161T103 203Q103 234 103 269T102 328V351Q99 370 88 376T43 385H25V408Q25 431 27 431L37 432Q47 433 65 434T102 436Q119 437 138 438T167 441T178 442H181V402Q181 364 182 364T187 369T199 384T218 402T247 421T285 437Q305 442 336 442Q450 438 463 329Q464 322 464 190V104Q464 66 466 59T477 49Q498 46 526 46H542V0H534L510 1Q487 2 460 2T422 3Q319 3 310 0H302V46H318Q379 46 379 62Q380 64 380 200Q379 335 378 343Q372 371 358 385T334 402T308 404Q263 404 229 370Q202 343 195 315T187 232V168V108Q187 78 188 68T191 55T200 49Q221 46 249 46H265V0H257L234 1Q210 2 183 2T145 3Q42 3 33 0H25V46H41Z"></path><path id="MJX-11-TEX-N-64" d="M376 495Q376 511 376 535T377 568Q377 613 367 624T316 637H298V660Q298 683 300 683L310 684Q320 685 339 686T376 688Q393 689 413 690T443 693T454 694H457V390Q457 84 458 81Q461 61 472 55T517 46H535V0Q533 0 459 -5T380 -11H373V44L365 37Q307 -11 235 -11Q158 -11 96 50T34 215Q34 315 97 378T244 442Q319 442 376 393V495ZM373 342Q328 405 260 405Q211 405 173 369Q146 341 139 305T131 211Q131 155 138 120T173 59Q203 26 251 26Q322 26 373 103V342Z"></path><path id="MJX-11-TEX-N-6F" d="M28 214Q28 309 93 378T250 448Q340 448 405 380T471 215Q471 120 407 55T250 -10Q153 -10 91 57T28 214ZM250 30Q372 30 372 193V225V250Q372 272 371 288T364 326T348 362T317 390T268 410Q263 411 252 411Q222 411 195 399Q152 377 139 338T126 246V226Q126 130 145 91Q177 30 250 30Z"></path><path id="MJX-11-TEX-N-5F" d="M0 -62V-25H499V-62H0Z"></path><path id="MJX-11-TEX-N-63" d="M370 305T349 305T313 320T297 358Q297 381 312 396Q317 401 317 402T307 404Q281 408 258 408Q209 408 178 376Q131 329 131 219Q131 137 162 90Q203 29 272 29Q313 29 338 55T374 117Q376 125 379 127T395 129H409Q415 123 415 120Q415 116 411 104T395 71T366 33T318 2T249 -11Q163 -11 99 53T34 214Q34 318 99 383T250 448T370 421T404 357Q404 334 387 320Z"></path><path id="MJX-11-TEX-N-65" d="M28 218Q28 273 48 318T98 391T163 433T229 448Q282 448 320 430T378 380T406 316T415 245Q415 238 408 231H126V216Q126 68 226 36Q246 30 270 30Q312 30 342 62Q359 79 369 104L379 128Q382 131 395 131H398Q415 131 415 121Q415 117 412 108Q393 53 349 21T250 -11Q155 -11 92 58T28 218ZM333 275Q322 403 238 411H236Q228 411 220 410T195 402T166 381T143 340T127 274V267H333V275Z"></path><path id="MJX-11-TEX-N-74" d="M27 422Q80 426 109 478T141 600V615H181V431H316V385H181V241Q182 116 182 100T189 68Q203 29 238 29Q282 29 292 100Q293 108 293 146V181H333V146V134Q333 57 291 17Q264 -10 221 -10Q187 -10 162 2T124 33T105 68T98 100Q97 107 97 248V385H18V422H27Z"></path><path id="MJX-11-TEX-N-72" d="M36 46H50Q89 46 97 60V68Q97 77 97 91T98 122T98 161T98 203Q98 234 98 269T98 328L97 351Q94 370 83 376T38 385H20V408Q20 431 22 431L32 432Q42 433 60 434T96 436Q112 437 131 438T160 441T171 442H174V373Q213 441 271 441H277Q322 441 343 419T364 373Q364 352 351 337T313 322Q288 322 276 338T263 372Q263 381 265 388T270 400T273 405Q271 407 250 401Q234 393 226 386Q179 341 179 207V154Q179 141 179 127T179 101T180 81T180 66V61Q181 59 183 57T188 54T193 51T200 49T207 48T216 47T225 47T235 46T245 46H276V0H267Q249 3 140 3Q37 3 28 0H20V46H36Z"></path><path id="MJX-11-TEX-I-1D456" d="M184 600Q184 624 203 642T247 661Q265 661 277 649T290 619Q290 596 270 577T226 557Q211 557 198 567T184 600ZM21 287Q21 295 30 318T54 369T98 420T158 442Q197 442 223 419T250 357Q250 340 236 301T196 196T154 83Q149 61 149 51Q149 26 166 26Q175 26 185 29T208 43T235 78T260 137Q263 149 265 151T282 153Q302 153 302 143Q302 135 293 112T268 61T223 11T161 -11Q129 -11 102 10T74 74Q74 91 79 106T122 220Q160 321 166 341T173 380Q173 404 156 404H154Q124 404 99 371T61 287Q60 286 59 284T58 281T56 279T53 278T49 278T41 278H27Q21 284 21 287Z"></path><path id="MJX-11-TEX-N-3D" d="M56 347Q56 360 70 367H707Q722 359 722 347Q722 336 708 328L390 327H72Q56 332 56 347ZM56 153Q56 168 72 173H708Q722 163 722 153Q722 140 707 133H70Q56 140 56 153Z"></path><path id="MJX-11-TEX-S4-28" d="M758 -1237T758 -1240T752 -1249H736Q718 -1249 717 -1248Q711 -1245 672 -1199Q237 -706 237 251T672 1700Q697 1730 716 1749Q718 1750 735 1750H752Q758 1744 758 1741Q758 1737 740 1713T689 1644T619 1537T540 1380T463 1176Q348 802 348 251Q348 -242 441 -599T744 -1218Q758 -1237 758 -1240Z"></path><path id="MJX-11-TEX-S4-230A" d="M269 -1249V1750H342V-1176H633V-1249H269Z"></path><path id="MJX-11-TEX-N-73" d="M295 316Q295 356 268 385T190 414Q154 414 128 401Q98 382 98 349Q97 344 98 336T114 312T157 287Q175 282 201 278T245 269T277 256Q294 248 310 236T342 195T359 133Q359 71 321 31T198 -10H190Q138 -10 94 26L86 19L77 10Q71 4 65 -1L54 -11H46H42Q39 -11 33 -5V74V132Q33 153 35 157T45 162H54Q66 162 70 158T75 146T82 119T101 77Q136 26 198 26Q295 26 295 104Q295 133 277 151Q257 175 194 187T111 210Q75 227 54 256T33 318Q33 357 50 384T93 424T143 442T187 447H198Q238 447 268 432L283 424L292 431Q302 440 314 448H322H326Q329 448 335 442V310L329 304H301Q295 310 295 316Z"></path><path id="MJX-11-TEX-N-61" d="M137 305T115 305T78 320T63 359Q63 394 97 421T218 448Q291 448 336 416T396 340Q401 326 401 309T402 194V124Q402 76 407 58T428 40Q443 40 448 56T453 109V145H493V106Q492 66 490 59Q481 29 455 12T400 -6T353 12T329 54V58L327 55Q325 52 322 49T314 40T302 29T287 17T269 6T247 -2T221 -8T190 -11Q130 -11 82 20T34 107Q34 128 41 147T68 188T116 225T194 253T304 268H318V290Q318 324 312 340Q290 411 215 411Q197 411 181 410T156 406T148 403Q170 388 170 359Q170 334 154 320ZM126 106Q126 75 150 51T209 26Q247 26 276 49T315 109Q317 116 318 175Q318 233 317 233Q309 233 296 232T251 223T193 203T147 166T126 106Z"></path><path id="MJX-11-TEX-N-2B" d="M56 237T56 250T70 270H369V420L370 570Q380 583 389 583Q402 583 409 568V270H707Q722 262 722 250T707 230H409V-68Q401 -82 391 -82H389H387Q375 -82 369 -68V230H70Q56 237 56 250Z"></path><path id="MJX-11-TEX-N-32" d="M109 429Q82 429 66 447T50 491Q50 562 103 614T235 666Q326 666 387 610T449 465Q449 422 429 383T381 315T301 241Q265 210 201 149L142 93L218 92Q375 92 385 97Q392 99 409 186V189H449V186Q448 183 436 95T421 3V0H50V19V31Q50 38 56 46T86 81Q115 113 136 137Q145 147 170 174T204 211T233 244T261 278T284 308T305 340T320 369T333 401T340 431T343 464Q343 527 309 573T212 619Q179 619 154 602T119 569T109 550Q109 549 114 549Q132 549 151 535T170 489Q170 464 154 447T109 429Z"></path><path id="MJX-11-TEX-N-62" d="M307 -11Q234 -11 168 55L158 37Q156 34 153 28T147 17T143 10L138 1L118 0H98V298Q98 599 97 603Q94 622 83 628T38 637H20V660Q20 683 22 683L32 684Q42 685 61 686T98 688Q115 689 135 690T165 693T176 694H179V543Q179 391 180 391L183 394Q186 397 192 401T207 411T228 421T254 431T286 439T323 442Q401 442 461 379T522 216Q522 115 458 52T307 -11ZM182 98Q182 97 187 90T196 79T206 67T218 55T233 44T250 35T271 29T295 26Q330 26 363 46T412 113Q424 148 424 212Q424 287 412 323Q385 405 300 405Q270 405 239 390T188 347L182 339V98Z"></path><path id="MJX-11-TEX-N-4C" d="M128 622Q121 629 117 631T101 634T58 637H25V683H36Q48 680 182 680Q324 680 348 683H360V637H333Q273 637 258 635T233 622L232 342V129Q232 57 237 52Q243 47 313 47Q384 47 410 53Q470 70 498 110T536 221Q536 226 537 238T540 261T542 272T562 273H582V268Q580 265 568 137T554 5V0H25V46H58Q100 47 109 49T128 61V622Z"></path><path id="MJX-11-TEX-S4-230B" d="M296 -1176V1750H369V-1249H5V-1176H296Z"></path><path id="MJX-11-TEX-N-D7" d="M630 29Q630 9 609 9Q604 9 587 25T493 118L389 222L284 117Q178 13 175 11Q171 9 168 9Q160 9 154 15T147 29Q147 36 161 51T255 146L359 250L255 354Q174 435 161 449T147 471Q147 480 153 485T168 490Q173 490 175 489Q178 487 284 383L389 278L493 382Q570 459 587 475T609 491Q630 491 630 471Q630 464 620 453T522 355L418 250L522 145Q606 61 618 48T630 29Z"></path><path id="MJX-11-TEX-S4-29" d="M33 1741Q33 1750 51 1750H60H65Q73 1750 81 1743T119 1700Q554 1207 554 251Q554 -707 119 -1199Q76 -1250 66 -1250Q65 -1250 62 -1250T56 -1249Q55 -1249 53 -1249T49 -1250Q33 -1250 33 -1239Q33 -1236 50 -1214T98 -1150T163 -1052T238 -910T311 -727Q443 -335 443 251Q443 402 436 532T405 831T339 1142T224 1438T50 1716Q33 1737 33 1741Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="mtable"><g data-mml-node="mtr"><g data-mml-node="mtd"><g data-mml-node="msub"><g data-mml-node="mtext"><use data-c="77" xlink:href="#MJX-11-TEX-N-77"></use><use data-c="69" xlink:href="#MJX-11-TEX-N-69" transform="translate(722,0)"></use><use data-c="6E" xlink:href="#MJX-11-TEX-N-6E" transform="translate(1000,0)"></use><use data-c="64" xlink:href="#MJX-11-TEX-N-64" transform="translate(1556,0)"></use><use data-c="6F" xlink:href="#MJX-11-TEX-N-6F" transform="translate(2112,0)"></use><use data-c="77" xlink:href="#MJX-11-TEX-N-77" transform="translate(2612,0)"></use><use data-c="5F" xlink:href="#MJX-11-TEX-N-5F" transform="translate(3334,0)"></use><use data-c="63" xlink:href="#MJX-11-TEX-N-63" transform="translate(3834,0)"></use><use data-c="65" xlink:href="#MJX-11-TEX-N-65" transform="translate(4278,0)"></use><use data-c="6E" xlink:href="#MJX-11-TEX-N-6E" transform="translate(4722,0)"></use><use data-c="74" xlink:href="#MJX-11-TEX-N-74" transform="translate(5278,0)"></use><use data-c="65" xlink:href="#MJX-11-TEX-N-65" transform="translate(5667,0)"></use><use data-c="72" xlink:href="#MJX-11-TEX-N-72" transform="translate(6111,0)"></use></g><g data-mml-node="mi" transform="translate(6536,-150) scale(0.707)"><use data-c="1D456" xlink:href="#MJX-11-TEX-I-1D456"></use></g></g></g><g data-mml-node="mtd" transform="translate(6830,0)"><g data-mml-node="mi"></g><g data-mml-node="mo" transform="translate(277.8,0)"><use data-c="3D" xlink:href="#MJX-11-TEX-N-3D"></use></g><g data-mml-node="mrow" transform="translate(1333.6,0)"><g data-mml-node="mo" transform="translate(0 -0.5)"><use data-c="28" xlink:href="#MJX-11-TEX-S4-28"></use></g><g data-mml-node="mrow" transform="translate(792,0)"><g data-mml-node="mo" transform="translate(0 -0.5)"><use data-c="230A" xlink:href="#MJX-11-TEX-S4-230A"></use></g><g data-mml-node="mfrac" transform="translate(639,0)"><g data-mml-node="mfrac" transform="translate(220,805)"><g data-mml-node="mrow" transform="translate(220,451.6) scale(0.707)"><g data-mml-node="msub"><g data-mml-node="mtext"><use data-c="73" xlink:href="#MJX-11-TEX-N-73"></use><use data-c="74" xlink:href="#MJX-11-TEX-N-74" transform="translate(394,0)"></use><use data-c="61" xlink:href="#MJX-11-TEX-N-61" transform="translate(783,0)"></use><use data-c="72" xlink:href="#MJX-11-TEX-N-72" transform="translate(1283,0)"></use><use data-c="74" xlink:href="#MJX-11-TEX-N-74" transform="translate(1675,0)"></use></g><g data-mml-node="mi" transform="translate(2097,-150) scale(0.707)"><use data-c="1D456" xlink:href="#MJX-11-TEX-I-1D456"></use></g></g><g data-mml-node="mo" transform="translate(2391,0)"><use data-c="2B" xlink:href="#MJX-11-TEX-N-2B"></use></g><g data-mml-node="msub" transform="translate(3169,0)"><g data-mml-node="mtext"><use data-c="65" xlink:href="#MJX-11-TEX-N-65"></use><use data-c="6E" xlink:href="#MJX-11-TEX-N-6E" transform="translate(444,0)"></use><use data-c="64" xlink:href="#MJX-11-TEX-N-64" transform="translate(1000,0)"></use></g><g data-mml-node="mi" transform="translate(1589,-150) scale(0.707)"><use data-c="1D456" xlink:href="#MJX-11-TEX-I-1D456"></use></g></g></g><g data-mml-node="mn" transform="translate(1829.3,-345) scale(0.707)"><use data-c="32" xlink:href="#MJX-11-TEX-N-32"></use></g><rect width="3772.2" height="60" x="120" y="220"></rect></g><g data-mml-node="mtext" transform="translate(718.6,-686)"><use data-c="62" xlink:href="#MJX-11-TEX-N-62"></use><use data-c="69" xlink:href="#MJX-11-TEX-N-69" transform="translate(556,0)"></use><use data-c="6E" xlink:href="#MJX-11-TEX-N-6E" transform="translate(834,0)"></use><use data-c="4C" xlink:href="#MJX-11-TEX-N-4C" transform="translate(1390,0)"></use><use data-c="65" xlink:href="#MJX-11-TEX-N-65" transform="translate(2015,0)"></use><use data-c="6E" xlink:href="#MJX-11-TEX-N-6E" transform="translate(2459,0)"></use></g><rect width="4212.2" height="60" x="120" y="220"></rect></g><g data-mml-node="mo" transform="translate(5091.2,0) translate(0 -0.5)"><use data-c="230B" xlink:href="#MJX-11-TEX-S4-230B"></use></g></g><g data-mml-node="mo" transform="translate(6744.5,0)"><use data-c="D7" xlink:href="#MJX-11-TEX-N-D7"></use></g><g data-mml-node="mtext" transform="translate(7744.7,0)"><use data-c="62" xlink:href="#MJX-11-TEX-N-62"></use><use data-c="69" xlink:href="#MJX-11-TEX-N-69" transform="translate(556,0)"></use><use data-c="6E" xlink:href="#MJX-11-TEX-N-6E" transform="translate(834,0)"></use><use data-c="4C" xlink:href="#MJX-11-TEX-N-4C" transform="translate(1390,0)"></use><use data-c="65" xlink:href="#MJX-11-TEX-N-65" transform="translate(2015,0)"></use><use data-c="6E" xlink:href="#MJX-11-TEX-N-6E" transform="translate(2459,0)"></use></g><g data-mml-node="mo" transform="translate(10759.7,0) translate(0 -0.5)"><use data-c="29" xlink:href="#MJX-11-TEX-S4-29"></use></g></g><g data-mml-node="mo" transform="translate(13107.5,0)"><use data-c="2B" xlink:href="#MJX-11-TEX-N-2B"></use></g><g data-mml-node="mfrac" transform="translate(14107.7,0)"><g data-mml-node="mtext" transform="translate(220,676)"><use data-c="62" xlink:href="#MJX-11-TEX-N-62"></use><use data-c="69" xlink:href="#MJX-11-TEX-N-69" transform="translate(556,0)"></use><use data-c="6E" xlink:href="#MJX-11-TEX-N-6E" transform="translate(834,0)"></use><use data-c="4C" xlink:href="#MJX-11-TEX-N-4C" transform="translate(1390,0)"></use><use data-c="65" xlink:href="#MJX-11-TEX-N-65" transform="translate(2015,0)"></use><use data-c="6E" xlink:href="#MJX-11-TEX-N-6E" transform="translate(2459,0)"></use></g><g data-mml-node="mn" transform="translate(1477.5,-686)"><use data-c="32" xlink:href="#MJX-11-TEX-N-32"></use></g><rect width="3215" height="60" x="120" y="220"></rect></g></g></g></g></g></g></svg><mjx-assistive-mml unselectable="on" display="block"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><mtable displaystyle="true" columnalign="right left" columnspacing="0em" rowspacing="3pt"><mtr><mtd><msub><mtext>window_center</mtext><mi>i</mi></msub></mtd><mtd><mi></mi><mo>=</mo><mrow data-mjx-texclass="INNER"><mo data-mjx-texclass="OPEN">(</mo><mrow data-mjx-texclass="INNER"><mo data-mjx-texclass="OPEN">⌊</mo><mfrac><mfrac><mrow><msub><mtext>start</mtext><mi>i</mi></msub><mo>+</mo><msub><mtext>end</mtext><mi>i</mi></msub></mrow><mn>2</mn></mfrac><mtext>binLen</mtext></mfrac><mo data-mjx-texclass="CLOSE">⌋</mo></mrow><mo>×</mo><mtext>binLen</mtext><mo data-mjx-texclass="CLOSE">)</mo></mrow><mo>+</mo><mfrac><mtext>binLen</mtext><mn>2</mn></mfrac></mtd></mtr></mtable></math></mjx-assistive-mml></mjx-container></div></div></li></ul><ul><li><p><span>计算该索引窗口的中心位置</span></p></li></ul><div contenteditable="false" spellcheck="false" class="mathjax-block md-end-block md-math-block md-rawblock" id="mathjax-n2796" cid="n2796" mdtype="math_block" data-math-tag-before="0" data-math-tag-after="0" data-math-labels="[]"><div class="md-math-container"><mjx-container class="MathJax" jax="SVG" display="true" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="44.267ex" height="4.652ex" role="img" focusable="false" viewBox="0 -1370 19565.9 2056" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -1.552ex;"><defs><path id="MJX-12-TEX-N-24" d="M162 187Q162 164 146 149T109 133H103V130Q108 115 115 105Q122 92 131 82T150 64T170 52T190 44T206 40T220 37L227 36V313Q190 320 162 335Q116 358 86 404T55 508Q55 567 85 614T165 685Q186 696 225 704H227V750H273V704L286 703Q369 690 413 631Q441 588 444 531Q444 514 443 509Q439 490 425 479T391 468Q368 468 353 483T337 522Q337 546 353 560T390 575L394 576V578Q386 599 372 614T342 637T314 649T288 656L273 658V408L288 405Q329 394 355 376Q396 348 420 300T444 199Q444 130 408 76T313 1Q286 -9 276 -9H273V-56H227V-10H221Q202 -6 193 -4T155 11T108 41T74 94T55 176V182Q55 227 95 238Q103 240 108 240Q129 240 145 226T162 187ZM225 657Q219 657 204 651T169 632T135 594T121 538Q121 512 131 491T156 457T187 435T213 423T227 420V539Q227 657 225 657ZM378 169Q378 230 339 265T274 301Q273 301 273 169V37Q324 50 351 87T378 169Z"></path><path id="MJX-12-TEX-N-77" d="M90 368Q84 378 76 380T40 385H18V431H24L43 430Q62 430 84 429T116 428Q206 428 221 431H229V385H215Q177 383 177 368Q177 367 221 239L265 113L339 328L333 345Q323 374 316 379Q308 384 278 385H258V431H264Q270 428 348 428Q439 428 454 431H461V385H452Q404 385 404 369Q404 366 418 324T449 234T481 143L496 100L537 219Q579 341 579 347Q579 363 564 373T530 385H522V431H529Q541 428 624 428Q692 428 698 431H703V385H697Q696 385 691 385T682 384Q635 377 619 334L559 161Q546 124 528 71Q508 12 503 1T487 -11H479Q460 -11 456 -4Q455 -3 407 133L361 267Q359 263 266 -4Q261 -11 243 -11H238Q225 -11 220 -3L90 368Z"></path><path id="MJX-12-TEX-N-69" d="M69 609Q69 637 87 653T131 669Q154 667 171 652T188 609Q188 579 171 564T129 549Q104 549 87 564T69 609ZM247 0Q232 3 143 3Q132 3 106 3T56 1L34 0H26V46H42Q70 46 91 49Q100 53 102 60T104 102V205V293Q104 345 102 359T88 378Q74 385 41 385H30V408Q30 431 32 431L42 432Q52 433 70 434T106 436Q123 437 142 438T171 441T182 442H185V62Q190 52 197 50T232 46H255V0H247Z"></path><path id="MJX-12-TEX-N-6E" d="M41 46H55Q94 46 102 60V68Q102 77 102 91T102 122T103 161T103 203Q103 234 103 269T102 328V351Q99 370 88 376T43 385H25V408Q25 431 27 431L37 432Q47 433 65 434T102 436Q119 437 138 438T167 441T178 442H181V402Q181 364 182 364T187 369T199 384T218 402T247 421T285 437Q305 442 336 442Q450 438 463 329Q464 322 464 190V104Q464 66 466 59T477 49Q498 46 526 46H542V0H534L510 1Q487 2 460 2T422 3Q319 3 310 0H302V46H318Q379 46 379 62Q380 64 380 200Q379 335 378 343Q372 371 358 385T334 402T308 404Q263 404 229 370Q202 343 195 315T187 232V168V108Q187 78 188 68T191 55T200 49Q221 46 249 46H265V0H257L234 1Q210 2 183 2T145 3Q42 3 33 0H25V46H41Z"></path><path id="MJX-12-TEX-N-64" d="M376 495Q376 511 376 535T377 568Q377 613 367 624T316 637H298V660Q298 683 300 683L310 684Q320 685 339 686T376 688Q393 689 413 690T443 693T454 694H457V390Q457 84 458 81Q461 61 472 55T517 46H535V0Q533 0 459 -5T380 -11H373V44L365 37Q307 -11 235 -11Q158 -11 96 50T34 215Q34 315 97 378T244 442Q319 442 376 393V495ZM373 342Q328 405 260 405Q211 405 173 369Q146 341 139 305T131 211Q131 155 138 120T173 59Q203 26 251 26Q322 26 373 103V342Z"></path><path id="MJX-12-TEX-N-6F" d="M28 214Q28 309 93 378T250 448Q340 448 405 380T471 215Q471 120 407 55T250 -10Q153 -10 91 57T28 214ZM250 30Q372 30 372 193V225V250Q372 272 371 288T364 326T348 362T317 390T268 410Q263 411 252 411Q222 411 195 399Q152 377 139 338T126 246V226Q126 130 145 91Q177 30 250 30Z"></path><path id="MJX-12-TEX-N-5F" d="M0 -62V-25H499V-62H0Z"></path><path id="MJX-12-TEX-N-63" d="M370 305T349 305T313 320T297 358Q297 381 312 396Q317 401 317 402T307 404Q281 408 258 408Q209 408 178 376Q131 329 131 219Q131 137 162 90Q203 29 272 29Q313 29 338 55T374 117Q376 125 379 127T395 129H409Q415 123 415 120Q415 116 411 104T395 71T366 33T318 2T249 -11Q163 -11 99 53T34 214Q34 318 99 383T250 448T370 421T404 357Q404 334 387 320Z"></path><path id="MJX-12-TEX-N-65" d="M28 218Q28 273 48 318T98 391T163 433T229 448Q282 448 320 430T378 380T406 316T415 245Q415 238 408 231H126V216Q126 68 226 36Q246 30 270 30Q312 30 342 62Q359 79 369 104L379 128Q382 131 395 131H398Q415 131 415 121Q415 117 412 108Q393 53 349 21T250 -11Q155 -11 92 58T28 218ZM333 275Q322 403 238 411H236Q228 411 220 410T195 402T166 381T143 340T127 274V267H333V275Z"></path><path id="MJX-12-TEX-N-74" d="M27 422Q80 426 109 478T141 600V615H181V431H316V385H181V241Q182 116 182 100T189 68Q203 29 238 29Q282 29 292 100Q293 108 293 146V181H333V146V134Q333 57 291 17Q264 -10 221 -10Q187 -10 162 2T124 33T105 68T98 100Q97 107 97 248V385H18V422H27Z"></path><path id="MJX-12-TEX-N-72" d="M36 46H50Q89 46 97 60V68Q97 77 97 91T98 122T98 161T98 203Q98 234 98 269T98 328L97 351Q94 370 83 376T38 385H20V408Q20 431 22 431L32 432Q42 433 60 434T96 436Q112 437 131 438T160 441T171 442H174V373Q213 441 271 441H277Q322 441 343 419T364 373Q364 352 351 337T313 322Q288 322 276 338T263 372Q263 381 265 388T270 400T273 405Q271 407 250 401Q234 393 226 386Q179 341 179 207V154Q179 141 179 127T179 101T180 81T180 66V61Q181 59 183 57T188 54T193 51T200 49T207 48T216 47T225 47T235 46T245 46H276V0H267Q249 3 140 3Q37 3 28 0H20V46H36Z"></path><path id="MJX-12-TEX-I-1D456" d="M184 600Q184 624 203 642T247 661Q265 661 277 649T290 619Q290 596 270 577T226 557Q211 557 198 567T184 600ZM21 287Q21 295 30 318T54 369T98 420T158 442Q197 442 223 419T250 357Q250 340 236 301T196 196T154 83Q149 61 149 51Q149 26 166 26Q175 26 185 29T208 43T235 78T260 137Q263 149 265 151T282 153Q302 153 302 143Q302 135 293 112T268 61T223 11T161 -11Q129 -11 102 10T74 74Q74 91 79 106T122 220Q160 321 166 341T173 380Q173 404 156 404H154Q124 404 99 371T61 287Q60 286 59 284T58 281T56 279T53 278T49 278T41 278H27Q21 284 21 287Z"></path><path id="MJX-12-TEX-N-3D" d="M56 347Q56 360 70 367H707Q722 359 722 347Q722 336 708 328L390 327H72Q56 332 56 347ZM56 153Q56 168 72 173H708Q722 163 722 153Q722 140 707 133H70Q56 140 56 153Z"></path><path id="MJX-12-TEX-N-73" d="M295 316Q295 356 268 385T190 414Q154 414 128 401Q98 382 98 349Q97 344 98 336T114 312T157 287Q175 282 201 278T245 269T277 256Q294 248 310 236T342 195T359 133Q359 71 321 31T198 -10H190Q138 -10 94 26L86 19L77 10Q71 4 65 -1L54 -11H46H42Q39 -11 33 -5V74V132Q33 153 35 157T45 162H54Q66 162 70 158T75 146T82 119T101 77Q136 26 198 26Q295 26 295 104Q295 133 277 151Q257 175 194 187T111 210Q75 227 54 256T33 318Q33 357 50 384T93 424T143 442T187 447H198Q238 447 268 432L283 424L292 431Q302 440 314 448H322H326Q329 448 335 442V310L329 304H301Q295 310 295 316Z"></path><path id="MJX-12-TEX-N-61" d="M137 305T115 305T78 320T63 359Q63 394 97 421T218 448Q291 448 336 416T396 340Q401 326 401 309T402 194V124Q402 76 407 58T428 40Q443 40 448 56T453 109V145H493V106Q492 66 490 59Q481 29 455 12T400 -6T353 12T329 54V58L327 55Q325 52 322 49T314 40T302 29T287 17T269 6T247 -2T221 -8T190 -11Q130 -11 82 20T34 107Q34 128 41 147T68 188T116 225T194 253T304 268H318V290Q318 324 312 340Q290 411 215 411Q197 411 181 410T156 406T148 403Q170 388 170 359Q170 334 154 320ZM126 106Q126 75 150 51T209 26Q247 26 276 49T315 109Q317 116 318 175Q318 233 317 233Q309 233 296 232T251 223T193 203T147 166T126 106Z"></path><path id="MJX-12-TEX-N-2B" d="M56 237T56 250T70 270H369V420L370 570Q380 583 389 583Q402 583 409 568V270H707Q722 262 722 250T707 230H409V-68Q401 -82 391 -82H389H387Q375 -82 369 -68V230H70Q56 237 56 250Z"></path><path id="MJX-12-TEX-N-62" d="M307 -11Q234 -11 168 55L158 37Q156 34 153 28T147 17T143 10L138 1L118 0H98V298Q98 599 97 603Q94 622 83 628T38 637H20V660Q20 683 22 683L32 684Q42 685 61 686T98 688Q115 689 135 690T165 693T176 694H179V543Q179 391 180 391L183 394Q186 397 192 401T207 411T228 421T254 431T286 439T323 442Q401 442 461 379T522 216Q522 115 458 52T307 -11ZM182 98Q182 97 187 90T196 79T206 67T218 55T233 44T250 35T271 29T295 26Q330 26 363 46T412 113Q424 148 424 212Q424 287 412 323Q385 405 300 405Q270 405 239 390T188 347L182 339V98Z"></path><path id="MJX-12-TEX-N-4C" d="M128 622Q121 629 117 631T101 634T58 637H25V683H36Q48 680 182 680Q324 680 348 683H360V637H333Q273 637 258 635T233 622L232 342V129Q232 57 237 52Q243 47 313 47Q384 47 410 53Q470 70 498 110T536 221Q536 226 537 238T540 261T542 272T562 273H582V268Q580 265 568 137T554 5V0H25V46H58Q100 47 109 49T128 61V622Z"></path><path id="MJX-12-TEX-N-32" d="M109 429Q82 429 66 447T50 491Q50 562 103 614T235 666Q326 666 387 610T449 465Q449 422 429 383T381 315T301 241Q265 210 201 149L142 93L218 92Q375 92 385 97Q392 99 409 186V189H449V186Q448 183 436 95T421 3V0H50V19V31Q50 38 56 46T86 81Q115 113 136 137Q145 147 170 174T204 211T233 244T261 278T284 308T305 340T320 369T333 401T340 431T343 464Q343 527 309 573T212 619Q179 619 154 602T119 569T109 550Q109 549 114 549Q132 549 151 535T170 489Q170 464 154 447T109 429Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="TeXAtom" data-mjx-texclass="ORD"><g data-mml-node="mo"><use data-c="24" xlink:href="#MJX-12-TEX-N-24"></use></g></g><g data-mml-node="msub" transform="translate(500,0)"><g data-mml-node="mtext"><use data-c="77" xlink:href="#MJX-12-TEX-N-77"></use><use data-c="69" xlink:href="#MJX-12-TEX-N-69" transform="translate(722,0)"></use><use data-c="6E" xlink:href="#MJX-12-TEX-N-6E" transform="translate(1000,0)"></use><use data-c="64" xlink:href="#MJX-12-TEX-N-64" transform="translate(1556,0)"></use><use data-c="6F" xlink:href="#MJX-12-TEX-N-6F" transform="translate(2112,0)"></use><use data-c="77" xlink:href="#MJX-12-TEX-N-77" transform="translate(2612,0)"></use><use data-c="5F" xlink:href="#MJX-12-TEX-N-5F" transform="translate(3334,0)"></use><use data-c="63" xlink:href="#MJX-12-TEX-N-63" transform="translate(3834,0)"></use><use data-c="65" xlink:href="#MJX-12-TEX-N-65" transform="translate(4278,0)"></use><use data-c="6E" xlink:href="#MJX-12-TEX-N-6E" transform="translate(4722,0)"></use><use data-c="74" xlink:href="#MJX-12-TEX-N-74" transform="translate(5278,0)"></use><use data-c="65" xlink:href="#MJX-12-TEX-N-65" transform="translate(5667,0)"></use><use data-c="72" xlink:href="#MJX-12-TEX-N-72" transform="translate(6111,0)"></use></g><g data-mml-node="mi" transform="translate(6536,-150) scale(0.707)"><use data-c="1D456" xlink:href="#MJX-12-TEX-I-1D456"></use></g></g><g data-mml-node="mo" transform="translate(7607.7,0)"><use data-c="3D" xlink:href="#MJX-12-TEX-N-3D"></use></g><g data-mml-node="msub" transform="translate(8663.5,0)"><g data-mml-node="mtext"><use data-c="77" xlink:href="#MJX-12-TEX-N-77"></use><use data-c="69" xlink:href="#MJX-12-TEX-N-69" transform="translate(722,0)"></use><use data-c="6E" xlink:href="#MJX-12-TEX-N-6E" transform="translate(1000,0)"></use><use data-c="64" xlink:href="#MJX-12-TEX-N-64" transform="translate(1556,0)"></use><use data-c="6F" xlink:href="#MJX-12-TEX-N-6F" transform="translate(2112,0)"></use><use data-c="77" xlink:href="#MJX-12-TEX-N-77" transform="translate(2612,0)"></use><use data-c="5F" xlink:href="#MJX-12-TEX-N-5F" transform="translate(3334,0)"></use><use data-c="73" xlink:href="#MJX-12-TEX-N-73" transform="translate(3834,0)"></use><use data-c="74" xlink:href="#MJX-12-TEX-N-74" transform="translate(4228,0)"></use><use data-c="61" xlink:href="#MJX-12-TEX-N-61" transform="translate(4617,0)"></use><use data-c="72" xlink:href="#MJX-12-TEX-N-72" transform="translate(5117,0)"></use><use data-c="74" xlink:href="#MJX-12-TEX-N-74" transform="translate(5509,0)"></use></g><g data-mml-node="mi" transform="translate(5931,-150) scale(0.707)"><use data-c="1D456" xlink:href="#MJX-12-TEX-I-1D456"></use></g></g><g data-mml-node="mo" transform="translate(15110.7,0)"><use data-c="2B" xlink:href="#MJX-12-TEX-N-2B"></use></g><g data-mml-node="mfrac" transform="translate(16110.9,0)"><g data-mml-node="mtext" transform="translate(220,676)"><use data-c="62" xlink:href="#MJX-12-TEX-N-62"></use><use data-c="69" xlink:href="#MJX-12-TEX-N-69" transform="translate(556,0)"></use><use data-c="6E" xlink:href="#MJX-12-TEX-N-6E" transform="translate(834,0)"></use><use data-c="4C" xlink:href="#MJX-12-TEX-N-4C" transform="translate(1390,0)"></use><use data-c="65" xlink:href="#MJX-12-TEX-N-65" transform="translate(2015,0)"></use><use data-c="6E" xlink:href="#MJX-12-TEX-N-6E" transform="translate(2459,0)"></use></g><g data-mml-node="mn" transform="translate(1477.5,-686)"><use data-c="32" xlink:href="#MJX-12-TEX-N-32"></use></g><rect width="3215" height="60" x="120" y="220"></rect></g></g></g></svg><mjx-assistive-mml unselectable="on" display="block"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><mrow data-mjx-texclass="ORD"><mo>$</mo></mrow><msub><mtext>window_center</mtext><mi>i</mi></msub><mo>=</mo><msub><mtext>window_start</mtext><mi>i</mi></msub><mo>+</mo><mfrac><mtext>binLen</mtext><mn>2</mn></mfrac></math></mjx-assistive-mml></mjx-container></div></div></li></ol><ul><li><p><span>对同一条染色体上相同的</span><mjx-container class="MathJax" jax="SVG" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="15.452ex" height="1.927ex" role="img" focusable="false" viewBox="0 -694 6830 851.8" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -0.357ex;"><defs><path id="MJX-36-TEX-N-77" d="M90 368Q84 378 76 380T40 385H18V431H24L43 430Q62 430 84 429T116 428Q206 428 221 431H229V385H215Q177 383 177 368Q177 367 221 239L265 113L339 328L333 345Q323 374 316 379Q308 384 278 385H258V431H264Q270 428 348 428Q439 428 454 431H461V385H452Q404 385 404 369Q404 366 418 324T449 234T481 143L496 100L537 219Q579 341 579 347Q579 363 564 373T530 385H522V431H529Q541 428 624 428Q692 428 698 431H703V385H697Q696 385 691 385T682 384Q635 377 619 334L559 161Q546 124 528 71Q508 12 503 1T487 -11H479Q460 -11 456 -4Q455 -3 407 133L361 267Q359 263 266 -4Q261 -11 243 -11H238Q225 -11 220 -3L90 368Z"></path><path id="MJX-36-TEX-N-69" d="M69 609Q69 637 87 653T131 669Q154 667 171 652T188 609Q188 579 171 564T129 549Q104 549 87 564T69 609ZM247 0Q232 3 143 3Q132 3 106 3T56 1L34 0H26V46H42Q70 46 91 49Q100 53 102 60T104 102V205V293Q104 345 102 359T88 378Q74 385 41 385H30V408Q30 431 32 431L42 432Q52 433 70 434T106 436Q123 437 142 438T171 441T182 442H185V62Q190 52 197 50T232 46H255V0H247Z"></path><path id="MJX-36-TEX-N-6E" d="M41 46H55Q94 46 102 60V68Q102 77 102 91T102 122T103 161T103 203Q103 234 103 269T102 328V351Q99 370 88 376T43 385H25V408Q25 431 27 431L37 432Q47 433 65 434T102 436Q119 437 138 438T167 441T178 442H181V402Q181 364 182 364T187 369T199 384T218 402T247 421T285 437Q305 442 336 442Q450 438 463 329Q464 322 464 190V104Q464 66 466 59T477 49Q498 46 526 46H542V0H534L510 1Q487 2 460 2T422 3Q319 3 310 0H302V46H318Q379 46 379 62Q380 64 380 200Q379 335 378 343Q372 371 358 385T334 402T308 404Q263 404 229 370Q202 343 195 315T187 232V168V108Q187 78 188 68T191 55T200 49Q221 46 249 46H265V0H257L234 1Q210 2 183 2T145 3Q42 3 33 0H25V46H41Z"></path><path id="MJX-36-TEX-N-64" d="M376 495Q376 511 376 535T377 568Q377 613 367 624T316 637H298V660Q298 683 300 683L310 684Q320 685 339 686T376 688Q393 689 413 690T443 693T454 694H457V390Q457 84 458 81Q461 61 472 55T517 46H535V0Q533 0 459 -5T380 -11H373V44L365 37Q307 -11 235 -11Q158 -11 96 50T34 215Q34 315 97 378T244 442Q319 442 376 393V495ZM373 342Q328 405 260 405Q211 405 173 369Q146 341 139 305T131 211Q131 155 138 120T173 59Q203 26 251 26Q322 26 373 103V342Z"></path><path id="MJX-36-TEX-N-6F" d="M28 214Q28 309 93 378T250 448Q340 448 405 380T471 215Q471 120 407 55T250 -10Q153 -10 91 57T28 214ZM250 30Q372 30 372 193V225V250Q372 272 371 288T364 326T348 362T317 390T268 410Q263 411 252 411Q222 411 195 399Q152 377 139 338T126 246V226Q126 130 145 91Q177 30 250 30Z"></path><path id="MJX-36-TEX-N-5F" d="M0 -62V-25H499V-62H0Z"></path><path id="MJX-36-TEX-N-63" d="M370 305T349 305T313 320T297 358Q297 381 312 396Q317 401 317 402T307 404Q281 408 258 408Q209 408 178 376Q131 329 131 219Q131 137 162 90Q203 29 272 29Q313 29 338 55T374 117Q376 125 379 127T395 129H409Q415 123 415 120Q415 116 411 104T395 71T366 33T318 2T249 -11Q163 -11 99 53T34 214Q34 318 99 383T250 448T370 421T404 357Q404 334 387 320Z"></path><path id="MJX-36-TEX-N-65" d="M28 218Q28 273 48 318T98 391T163 433T229 448Q282 448 320 430T378 380T406 316T415 245Q415 238 408 231H126V216Q126 68 226 36Q246 30 270 30Q312 30 342 62Q359 79 369 104L379 128Q382 131 395 131H398Q415 131 415 121Q415 117 412 108Q393 53 349 21T250 -11Q155 -11 92 58T28 218ZM333 275Q322 403 238 411H236Q228 411 220 410T195 402T166 381T143 340T127 274V267H333V275Z"></path><path id="MJX-36-TEX-N-74" d="M27 422Q80 426 109 478T141 600V615H181V431H316V385H181V241Q182 116 182 100T189 68Q203 29 238 29Q282 29 292 100Q293 108 293 146V181H333V146V134Q333 57 291 17Q264 -10 221 -10Q187 -10 162 2T124 33T105 68T98 100Q97 107 97 248V385H18V422H27Z"></path><path id="MJX-36-TEX-N-72" d="M36 46H50Q89 46 97 60V68Q97 77 97 91T98 122T98 161T98 203Q98 234 98 269T98 328L97 351Q94 370 83 376T38 385H20V408Q20 431 22 431L32 432Q42 433 60 434T96 436Q112 437 131 438T160 441T171 442H174V373Q213 441 271 441H277Q322 441 343 419T364 373Q364 352 351 337T313 322Q288 322 276 338T263 372Q263 381 265 388T270 400T273 405Q271 407 250 401Q234 393 226 386Q179 341 179 207V154Q179 141 179 127T179 101T180 81T180 66V61Q181 59 183 57T188 54T193 51T200 49T207 48T216 47T225 47T235 46T245 46H276V0H267Q249 3 140 3Q37 3 28 0H20V46H36Z"></path><path id="MJX-36-TEX-I-1D456" d="M184 600Q184 624 203 642T247 661Q265 661 277 649T290 619Q290 596 270 577T226 557Q211 557 198 567T184 600ZM21 287Q21 295 30 318T54 369T98 420T158 442Q197 442 223 419T250 357Q250 340 236 301T196 196T154 83Q149 61 149 51Q149 26 166 26Q175 26 185 29T208 43T235 78T260 137Q263 149 265 151T282 153Q302 153 302 143Q302 135 293 112T268 61T223 11T161 -11Q129 -11 102 10T74 74Q74 91 79 106T122 220Q160 321 166 341T173 380Q173 404 156 404H154Q124 404 99 371T61 287Q60 286 59 284T58 281T56 279T53 278T49 278T41 278H27Q21 284 21 287Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="msub"><g data-mml-node="mtext"><use data-c="77" xlink:href="#MJX-36-TEX-N-77"></use><use data-c="69" xlink:href="#MJX-36-TEX-N-69" transform="translate(722,0)"></use><use data-c="6E" xlink:href="#MJX-36-TEX-N-6E" transform="translate(1000,0)"></use><use data-c="64" xlink:href="#MJX-36-TEX-N-64" transform="translate(1556,0)"></use><use data-c="6F" xlink:href="#MJX-36-TEX-N-6F" transform="translate(2112,0)"></use><use data-c="77" xlink:href="#MJX-36-TEX-N-77" transform="translate(2612,0)"></use><use data-c="5F" xlink:href="#MJX-36-TEX-N-5F" transform="translate(3334,0)"></use><use data-c="63" xlink:href="#MJX-36-TEX-N-63" transform="translate(3834,0)"></use><use data-c="65" xlink:href="#MJX-36-TEX-N-65" transform="translate(4278,0)"></use><use data-c="6E" xlink:href="#MJX-36-TEX-N-6E" transform="translate(4722,0)"></use><use data-c="74" xlink:href="#MJX-36-TEX-N-74" transform="translate(5278,0)"></use><use data-c="65" xlink:href="#MJX-36-TEX-N-65" transform="translate(5667,0)"></use><use data-c="72" xlink:href="#MJX-36-TEX-N-72" transform="translate(6111,0)"></use></g><g data-mml-node="mi" transform="translate(6536,-150) scale(0.707)"><use data-c="1D456" xlink:href="#MJX-36-TEX-I-1D456"></use></g></g></g></g></svg><mjx-assistive-mml unselectable="on" display="inline"><math xmlns="http://www.w3.org/1998/Math/MathML"><msub><mtext>window_center</mtext><mi>i</mi></msub></math></mjx-assistive-mml></mjx-container><script type="math/tex">\text{window\_center}_i</script><span> 计数为该窗口的read count</span></p><p>&nbsp;</p></li></ul><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">#定义窗口大小为 500 bp。</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-def">binLen</span><span class="cm-operator">=</span><span class="cm-number">500</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">#计算每个read中点所在的窗口中心位置。 $1是染色体位置， $2是片段的起始位置，$3是终止位置。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-builtin">awk</span> <span class="cm-attribute">-v</span> <span class="cm-def">w</span><span class="cm-operator">=</span><span class="cm-def">$binLen</span> <span class="cm-string">'{print $1, int(($2 + $3)/(2*w))*w + w/2}'</span> sample.fragments.bed | \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 对片段的染色体位置（k1）和所在的窗口中心位置（k2）进行排序。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-builtin">sort</span> <span class="cm-attribute">-k1</span>,1V <span class="cm-attribute">-k2</span>,2n | \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 去除重复，并对重复行进行计数（每个窗口中有多少个片段），计数放在第一列</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">uniq <span class="cm-attribute">-c</span> | \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 按（\t）分割输出染色体位置($2),窗口中心位置($3),所在计数($1)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-builtin">awk</span> <span class="cm-attribute">-v</span> <span class="cm-def">OFS</span><span class="cm-operator">=</span><span class="cm-string">"\t"</span> <span class="cm-string">'{print $2, $3, $1}'</span>| \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 对染色体位置(k1)和窗口中心位置(k2)进行排序并输出</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-builtin">sort</span> <span class="cm-attribute">-k1</span>,1V <span class="cm-attribute">-k2</span>,2n &gt; sample.fragmentsCount.bed</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 372px;"></div><div class="CodeMirror-gutters" style="display: none; height: 372px;"></div></div></div></pre><p><span>   </span></p><ol start='2' ><li><p><strong><span>Pearson相关性matrix</span></strong></p><ul><li><p><strong><span>Pearson 相关性系数</span></strong><span>（Pearson Correlation Coefficient, PCC）用于衡量两个变量之间的线性相关性，值的范围在 -1 到 1 之间：</span></p><ul><li><p><strong><span>1</span></strong><span> 表示完全正相关，两个变量同步变化。</span></p></li><li><p><strong><span>-1</span></strong><span> 表示完全负相关，一个变量增加时，另一个减少。</span></p></li><li><p><strong><span>0</span></strong><span> 表示无线性相关性。</span></p></li></ul></li></ul><p>&nbsp;</p><ul><li><p><strong><span>Pearson计算公式</span></strong></p></li></ul><div contenteditable="false" spellcheck="false" class="mathjax-block md-end-block md-math-block md-rawblock" id="mathjax-n2821" cid="n2821" mdtype="math_block" data-math-tag-before="0" data-math-tag-after="0" data-math-labels="[]"><div class="md-math-container"><mjx-container class="MathJax" jax="SVG" display="true" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="35.409ex" height="7.986ex" role="img" focusable="false" viewBox="0 -1710 15650.9 3530" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -4.118ex;"><defs><path id="MJX-13-TEX-I-1D45F" d="M21 287Q22 290 23 295T28 317T38 348T53 381T73 411T99 433T132 442Q161 442 183 430T214 408T225 388Q227 382 228 382T236 389Q284 441 347 441H350Q398 441 422 400Q430 381 430 363Q430 333 417 315T391 292T366 288Q346 288 334 299T322 328Q322 376 378 392Q356 405 342 405Q286 405 239 331Q229 315 224 298T190 165Q156 25 151 16Q138 -11 108 -11Q95 -11 87 -5T76 7T74 17Q74 30 114 189T154 366Q154 405 128 405Q107 405 92 377T68 316T57 280Q55 278 41 278H27Q21 284 21 287Z"></path><path id="MJX-13-TEX-N-3D" d="M56 347Q56 360 70 367H707Q722 359 722 347Q722 336 708 328L390 327H72Q56 332 56 347ZM56 153Q56 168 72 173H708Q722 163 722 153Q722 140 707 133H70Q56 140 56 153Z"></path><path id="MJX-13-TEX-SO-2211" d="M61 748Q64 750 489 750H913L954 640Q965 609 976 579T993 533T999 516H979L959 517Q936 579 886 621T777 682Q724 700 655 705T436 710H319Q183 710 183 709Q186 706 348 484T511 259Q517 250 513 244L490 216Q466 188 420 134T330 27L149 -187Q149 -188 362 -188Q388 -188 436 -188T506 -189Q679 -189 778 -162T936 -43Q946 -27 959 6H999L913 -249L489 -250Q65 -250 62 -248Q56 -246 56 -239Q56 -234 118 -161Q186 -81 245 -11L428 206Q428 207 242 462L57 717L56 728Q56 744 61 748Z"></path><path id="MJX-13-TEX-N-28" d="M94 250Q94 319 104 381T127 488T164 576T202 643T244 695T277 729T302 750H315H319Q333 750 333 741Q333 738 316 720T275 667T226 581T184 443T167 250T184 58T225 -81T274 -167T316 -220T333 -241Q333 -250 318 -250H315H302L274 -226Q180 -141 137 -14T94 250Z"></path><path id="MJX-13-TEX-I-1D44B" d="M42 0H40Q26 0 26 11Q26 15 29 27Q33 41 36 43T55 46Q141 49 190 98Q200 108 306 224T411 342Q302 620 297 625Q288 636 234 637H206Q200 643 200 645T202 664Q206 677 212 683H226Q260 681 347 681Q380 681 408 681T453 682T473 682Q490 682 490 671Q490 670 488 658Q484 643 481 640T465 637Q434 634 411 620L488 426L541 485Q646 598 646 610Q646 628 622 635Q617 635 609 637Q594 637 594 648Q594 650 596 664Q600 677 606 683H618Q619 683 643 683T697 681T738 680Q828 680 837 683H845Q852 676 852 672Q850 647 840 637H824Q790 636 763 628T722 611T698 593L687 584Q687 585 592 480L505 384Q505 383 536 304T601 142T638 56Q648 47 699 46Q734 46 734 37Q734 35 732 23Q728 7 725 4T711 1Q708 1 678 1T589 2Q528 2 496 2T461 1Q444 1 444 10Q444 11 446 25Q448 35 450 39T455 44T464 46T480 47T506 54Q523 62 523 64Q522 64 476 181L429 299Q241 95 236 84Q232 76 232 72Q232 53 261 47Q262 47 267 47T273 46Q276 46 277 46T280 45T283 42T284 35Q284 26 282 19Q279 6 276 4T261 1Q258 1 243 1T201 2T142 2Q64 2 42 0Z"></path><path id="MJX-13-TEX-I-1D456" d="M184 600Q184 624 203 642T247 661Q265 661 277 649T290 619Q290 596 270 577T226 557Q211 557 198 567T184 600ZM21 287Q21 295 30 318T54 369T98 420T158 442Q197 442 223 419T250 357Q250 340 236 301T196 196T154 83Q149 61 149 51Q149 26 166 26Q175 26 185 29T208 43T235 78T260 137Q263 149 265 151T282 153Q302 153 302 143Q302 135 293 112T268 61T223 11T161 -11Q129 -11 102 10T74 74Q74 91 79 106T122 220Q160 321 166 341T173 380Q173 404 156 404H154Q124 404 99 371T61 287Q60 286 59 284T58 281T56 279T53 278T49 278T41 278H27Q21 284 21 287Z"></path><path id="MJX-13-TEX-N-2212" d="M84 237T84 250T98 270H679Q694 262 694 250T679 230H98Q84 237 84 250Z"></path><path id="MJX-13-TEX-S4-2013" d="M0 248V285H499V248H0Z"></path><path id="MJX-13-TEX-N-29" d="M60 749L64 750Q69 750 74 750H86L114 726Q208 641 251 514T294 250Q294 182 284 119T261 12T224 -76T186 -143T145 -194T113 -227T90 -246Q87 -249 86 -250H74Q66 -250 63 -250T58 -247T55 -238Q56 -237 66 -225Q221 -64 221 250T66 725Q56 737 55 738Q55 746 60 749Z"></path><path id="MJX-13-TEX-I-1D44C" d="M66 637Q54 637 49 637T39 638T32 641T30 647T33 664T42 682Q44 683 56 683Q104 680 165 680Q288 680 306 683H316Q322 677 322 674T320 656Q316 643 310 637H298Q242 637 242 624Q242 619 292 477T343 333L346 336Q350 340 358 349T379 373T411 410T454 461Q546 568 561 587T577 618Q577 634 545 637Q528 637 528 647Q528 649 530 661Q533 676 535 679T549 683Q551 683 578 682T657 680Q684 680 713 681T746 682Q763 682 763 673Q763 669 760 657T755 643Q753 637 734 637Q662 632 617 587Q608 578 477 424L348 273L322 169Q295 62 295 57Q295 46 363 46Q379 46 384 45T390 35Q390 33 388 23Q384 6 382 4T366 1Q361 1 324 1T232 2Q170 2 138 2T102 1Q84 1 84 9Q84 14 87 24Q88 27 89 30T90 35T91 39T93 42T96 44T101 45T107 45T116 46T129 46Q168 47 180 50T198 63Q201 68 227 171L252 274L129 623Q128 624 127 625T125 627T122 629T118 631T113 633T105 634T96 635T83 636T66 637Z"></path><path id="MJX-13-TEX-LO-221A" d="M1001 1150Q1017 1150 1020 1132Q1020 1127 741 244L460 -643Q453 -650 436 -650H424Q423 -647 423 -645T421 -640T419 -631T415 -617T408 -594T399 -560T385 -512T367 -448T343 -364T312 -259L203 119L138 41L111 67L212 188L264 248L472 -474L983 1140Q988 1150 1001 1150Z"></path><path id="MJX-13-TEX-N-32" d="M109 429Q82 429 66 447T50 491Q50 562 103 614T235 666Q326 666 387 610T449 465Q449 422 429 383T381 315T301 241Q265 210 201 149L142 93L218 92Q375 92 385 97Q392 99 409 186V189H449V186Q448 183 436 95T421 3V0H50V19V31Q50 38 56 46T86 81Q115 113 136 137Q145 147 170 174T204 211T233 244T261 278T284 308T305 340T320 369T333 401T340 431T343 464Q343 527 309 573T212 619Q179 619 154 602T119 569T109 550Q109 549 114 549Q132 549 151 535T170 489Q170 464 154 447T109 429Z"></path><path id="MJX-13-TEX-N-22C5" d="M78 250Q78 274 95 292T138 310Q162 310 180 294T199 251Q199 226 182 208T139 190T96 207T78 250Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="mi"><use data-c="1D45F" xlink:href="#MJX-13-TEX-I-1D45F"></use></g><g data-mml-node="mo" transform="translate(728.8,0)"><use data-c="3D" xlink:href="#MJX-13-TEX-N-3D"></use></g><g data-mml-node="mfrac" transform="translate(1784.6,0)"><g data-mml-node="TeXAtom" data-mjx-texclass="ORD" transform="translate(2565.8,710)"><g data-mml-node="mo"><use data-c="2211" xlink:href="#MJX-13-TEX-SO-2211"></use></g><g data-mml-node="mo" transform="translate(1056,0)"><use data-c="28" xlink:href="#MJX-13-TEX-N-28"></use></g><g data-mml-node="msub" transform="translate(1445,0)"><g data-mml-node="mi"><use data-c="1D44B" xlink:href="#MJX-13-TEX-I-1D44B"></use></g><g data-mml-node="mi" transform="translate(861,-150) scale(0.707)"><use data-c="1D456" xlink:href="#MJX-13-TEX-I-1D456"></use></g></g><g data-mml-node="mo" transform="translate(2822.2,0)"><use data-c="2212" xlink:href="#MJX-13-TEX-N-2212"></use></g><g data-mml-node="mover" transform="translate(3822.4,0)"><g data-mml-node="mi"><use data-c="1D44B" xlink:href="#MJX-13-TEX-I-1D44B"></use></g><g data-mml-node="mo" transform="translate(0,615)"><svg width="852" height="237" x="0" y="148" viewBox="213 148 852 237"><use data-c="2013" xlink:href="#MJX-13-TEX-S4-2013" transform="scale(2.556,1)"></use></svg></g></g><g data-mml-node="mo" transform="translate(4674.4,0)"><use data-c="29" xlink:href="#MJX-13-TEX-N-29"></use></g><g data-mml-node="mo" transform="translate(5063.4,0)"><use data-c="28" xlink:href="#MJX-13-TEX-N-28"></use></g><g data-mml-node="msub" transform="translate(5452.4,0)"><g data-mml-node="mi"><use data-c="1D44C" xlink:href="#MJX-13-TEX-I-1D44C"></use></g><g data-mml-node="mi" transform="translate(614,-150) scale(0.707)"><use data-c="1D456" xlink:href="#MJX-13-TEX-I-1D456"></use></g></g><g data-mml-node="mo" transform="translate(6582.6,0)"><use data-c="2212" xlink:href="#MJX-13-TEX-N-2212"></use></g><g data-mml-node="mover" transform="translate(7582.8,0)"><g data-mml-node="mi"><use data-c="1D44C" xlink:href="#MJX-13-TEX-I-1D44C"></use></g><g data-mml-node="mo" transform="translate(0,615)"><svg width="763" height="237" x="0" y="148" viewBox="190.7 148 763 237"><use data-c="2013" xlink:href="#MJX-13-TEX-S4-2013" transform="scale(2.289,1)"></use></svg></g></g><g data-mml-node="mo" transform="translate(8345.8,0)"><use data-c="29" xlink:href="#MJX-13-TEX-N-29"></use></g></g><g data-mml-node="TeXAtom" data-mjx-texclass="ORD" transform="translate(220,-1362.5)"><g data-mml-node="msqrt"><g transform="translate(1020,0)"><g data-mml-node="mo"><use data-c="2211" xlink:href="#MJX-13-TEX-SO-2211"></use></g><g data-mml-node="mo" transform="translate(1056,0)"><use data-c="28" xlink:href="#MJX-13-TEX-N-28"></use></g><g data-mml-node="msub" transform="translate(1445,0)"><g data-mml-node="mi"><use data-c="1D44B" xlink:href="#MJX-13-TEX-I-1D44B"></use></g><g data-mml-node="mi" transform="translate(861,-150) scale(0.707)"><use data-c="1D456" xlink:href="#MJX-13-TEX-I-1D456"></use></g></g><g data-mml-node="mo" transform="translate(2822.2,0)"><use data-c="2212" xlink:href="#MJX-13-TEX-N-2212"></use></g><g data-mml-node="mover" transform="translate(3822.4,0)"><g data-mml-node="mi"><use data-c="1D44B" xlink:href="#MJX-13-TEX-I-1D44B"></use></g><g data-mml-node="mo" transform="translate(0,615)"><svg width="852" height="237" x="0" y="148" viewBox="213 148 852 237"><use data-c="2013" xlink:href="#MJX-13-TEX-S4-2013" transform="scale(2.556,1)"></use></svg></g></g><g data-mml-node="msup" transform="translate(4674.4,0)"><g data-mml-node="mo"><use data-c="29" xlink:href="#MJX-13-TEX-N-29"></use></g><g data-mml-node="mn" transform="translate(422,289) scale(0.707)"><use data-c="32" xlink:href="#MJX-13-TEX-N-32"></use></g></g></g><g data-mml-node="mo" transform="translate(0,192.5)"><use data-c="221A" xlink:href="#MJX-13-TEX-LO-221A"></use></g><rect width="5499.9" height="60" x="1020" y="1282.5"></rect></g><g data-mml-node="mo" transform="translate(6742.2,0)"><use data-c="22C5" xlink:href="#MJX-13-TEX-N-22C5"></use></g><g data-mml-node="msqrt" transform="translate(7242.4,0)"><g transform="translate(1020,0)"><g data-mml-node="mo"><use data-c="2211" xlink:href="#MJX-13-TEX-SO-2211"></use></g><g data-mml-node="mo" transform="translate(1056,0)"><use data-c="28" xlink:href="#MJX-13-TEX-N-28"></use></g><g data-mml-node="msub" transform="translate(1445,0)"><g data-mml-node="mi"><use data-c="1D44C" xlink:href="#MJX-13-TEX-I-1D44C"></use></g><g data-mml-node="mi" transform="translate(614,-150) scale(0.707)"><use data-c="1D456" xlink:href="#MJX-13-TEX-I-1D456"></use></g></g><g data-mml-node="mo" transform="translate(2575.2,0)"><use data-c="2212" xlink:href="#MJX-13-TEX-N-2212"></use></g><g data-mml-node="mover" transform="translate(3575.4,0)"><g data-mml-node="mi"><use data-c="1D44C" xlink:href="#MJX-13-TEX-I-1D44C"></use></g><g data-mml-node="mo" transform="translate(0,615)"><svg width="763" height="237" x="0" y="148" viewBox="190.7 148 763 237"><use data-c="2013" xlink:href="#MJX-13-TEX-S4-2013" transform="scale(2.289,1)"></use></svg></g></g><g data-mml-node="msup" transform="translate(4338.4,0)"><g data-mml-node="mo"><use data-c="29" xlink:href="#MJX-13-TEX-N-29"></use></g><g data-mml-node="mn" transform="translate(422,289) scale(0.707)"><use data-c="32" xlink:href="#MJX-13-TEX-N-32"></use></g></g></g><g data-mml-node="mo" transform="translate(0,192.5)"><use data-c="221A" xlink:href="#MJX-13-TEX-LO-221A"></use></g><rect width="5163.9" height="60" x="1020" y="1282.5"></rect></g></g><rect width="13626.3" height="60" x="120" y="220"></rect></g></g></g></svg><mjx-assistive-mml unselectable="on" display="block"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><mi>r</mi><mo>=</mo><mfrac><mrow data-mjx-texclass="ORD"><mo data-mjx-texclass="OP">∑</mo><mo stretchy="false">(</mo><msub><mi>X</mi><mi>i</mi></msub><mo>−</mo><mover><mi>X</mi><mo accent="true">―</mo></mover><mo stretchy="false">)</mo><mo stretchy="false">(</mo><msub><mi>Y</mi><mi>i</mi></msub><mo>−</mo><mover><mi>Y</mi><mo accent="true">―</mo></mover><mo stretchy="false">)</mo></mrow><mrow data-mjx-texclass="ORD"><msqrt><mo data-mjx-texclass="OP">∑</mo><mo stretchy="false">(</mo><msub><mi>X</mi><mi>i</mi></msub><mo>−</mo><mover><mi>X</mi><mo accent="true">―</mo></mover><msup><mo stretchy="false">)</mo><mn>2</mn></msup></msqrt><mo>⋅</mo><msqrt><mo data-mjx-texclass="OP">∑</mo><mo stretchy="false">(</mo><msub><mi>Y</mi><mi>i</mi></msub><mo>−</mo><mover><mi>Y</mi><mo accent="true">―</mo></mover><msup><mo stretchy="false">)</mo><mn>2</mn></msup></msqrt></mrow></mfrac></math></mjx-assistive-mml></mjx-container></div></div><p><mjx-container class="MathJax" jax="SVG" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="2.613ex" height="1.902ex" role="img" focusable="false" viewBox="0 -683 1155 840.8" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -0.357ex;"><defs><path id="MJX-37-TEX-I-1D44B" d="M42 0H40Q26 0 26 11Q26 15 29 27Q33 41 36 43T55 46Q141 49 190 98Q200 108 306 224T411 342Q302 620 297 625Q288 636 234 637H206Q200 643 200 645T202 664Q206 677 212 683H226Q260 681 347 681Q380 681 408 681T453 682T473 682Q490 682 490 671Q490 670 488 658Q484 643 481 640T465 637Q434 634 411 620L488 426L541 485Q646 598 646 610Q646 628 622 635Q617 635 609 637Q594 637 594 648Q594 650 596 664Q600 677 606 683H618Q619 683 643 683T697 681T738 680Q828 680 837 683H845Q852 676 852 672Q850 647 840 637H824Q790 636 763 628T722 611T698 593L687 584Q687 585 592 480L505 384Q505 383 536 304T601 142T638 56Q648 47 699 46Q734 46 734 37Q734 35 732 23Q728 7 725 4T711 1Q708 1 678 1T589 2Q528 2 496 2T461 1Q444 1 444 10Q444 11 446 25Q448 35 450 39T455 44T464 46T480 47T506 54Q523 62 523 64Q522 64 476 181L429 299Q241 95 236 84Q232 76 232 72Q232 53 261 47Q262 47 267 47T273 46Q276 46 277 46T280 45T283 42T284 35Q284 26 282 19Q279 6 276 4T261 1Q258 1 243 1T201 2T142 2Q64 2 42 0Z"></path><path id="MJX-37-TEX-I-1D456" d="M184 600Q184 624 203 642T247 661Q265 661 277 649T290 619Q290 596 270 577T226 557Q211 557 198 567T184 600ZM21 287Q21 295 30 318T54 369T98 420T158 442Q197 442 223 419T250 357Q250 340 236 301T196 196T154 83Q149 61 149 51Q149 26 166 26Q175 26 185 29T208 43T235 78T260 137Q263 149 265 151T282 153Q302 153 302 143Q302 135 293 112T268 61T223 11T161 -11Q129 -11 102 10T74 74Q74 91 79 106T122 220Q160 321 166 341T173 380Q173 404 156 404H154Q124 404 99 371T61 287Q60 286 59 284T58 281T56 279T53 278T49 278T41 278H27Q21 284 21 287Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="msub"><g data-mml-node="mi"><use data-c="1D44B" xlink:href="#MJX-37-TEX-I-1D44B"></use></g><g data-mml-node="mi" transform="translate(861,-150) scale(0.707)"><use data-c="1D456" xlink:href="#MJX-37-TEX-I-1D456"></use></g></g></g></g></svg><mjx-assistive-mml unselectable="on" display="inline"><math xmlns="http://www.w3.org/1998/Math/MathML"><msub><mi>X</mi><mi>i</mi></msub></math></mjx-assistive-mml></mjx-container><script type="math/tex"> X_i </script><span> 和 </span><mjx-container class="MathJax" jax="SVG" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="2.054ex" height="1.902ex" role="img" focusable="false" viewBox="0 -683 908 840.8" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -0.357ex;"><defs><path id="MJX-38-TEX-I-1D44C" d="M66 637Q54 637 49 637T39 638T32 641T30 647T33 664T42 682Q44 683 56 683Q104 680 165 680Q288 680 306 683H316Q322 677 322 674T320 656Q316 643 310 637H298Q242 637 242 624Q242 619 292 477T343 333L346 336Q350 340 358 349T379 373T411 410T454 461Q546 568 561 587T577 618Q577 634 545 637Q528 637 528 647Q528 649 530 661Q533 676 535 679T549 683Q551 683 578 682T657 680Q684 680 713 681T746 682Q763 682 763 673Q763 669 760 657T755 643Q753 637 734 637Q662 632 617 587Q608 578 477 424L348 273L322 169Q295 62 295 57Q295 46 363 46Q379 46 384 45T390 35Q390 33 388 23Q384 6 382 4T366 1Q361 1 324 1T232 2Q170 2 138 2T102 1Q84 1 84 9Q84 14 87 24Q88 27 89 30T90 35T91 39T93 42T96 44T101 45T107 45T116 46T129 46Q168 47 180 50T198 63Q201 68 227 171L252 274L129 623Q128 624 127 625T125 627T122 629T118 631T113 633T105 634T96 635T83 636T66 637Z"></path><path id="MJX-38-TEX-I-1D456" d="M184 600Q184 624 203 642T247 661Q265 661 277 649T290 619Q290 596 270 577T226 557Q211 557 198 567T184 600ZM21 287Q21 295 30 318T54 369T98 420T158 442Q197 442 223 419T250 357Q250 340 236 301T196 196T154 83Q149 61 149 51Q149 26 166 26Q175 26 185 29T208 43T235 78T260 137Q263 149 265 151T282 153Q302 153 302 143Q302 135 293 112T268 61T223 11T161 -11Q129 -11 102 10T74 74Q74 91 79 106T122 220Q160 321 166 341T173 380Q173 404 156 404H154Q124 404 99 371T61 287Q60 286 59 284T58 281T56 279T53 278T49 278T41 278H27Q21 284 21 287Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="msub"><g data-mml-node="mi"><use data-c="1D44C" xlink:href="#MJX-38-TEX-I-1D44C"></use></g><g data-mml-node="mi" transform="translate(614,-150) scale(0.707)"><use data-c="1D456" xlink:href="#MJX-38-TEX-I-1D456"></use></g></g></g></g></svg><mjx-assistive-mml unselectable="on" display="inline"><math xmlns="http://www.w3.org/1998/Math/MathML"><msub><mi>Y</mi><mi>i</mi></msub></math></mjx-assistive-mml></mjx-container><script type="math/tex"> Y_i </script><span> 是两个变量的值。</span></p><p><mjx-container class="MathJax" jax="SVG" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="1.928ex" height="2.262ex" role="img" focusable="false" viewBox="0 -1000 852 1000" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: 0px;"><defs><path id="MJX-39-TEX-I-1D44B" d="M42 0H40Q26 0 26 11Q26 15 29 27Q33 41 36 43T55 46Q141 49 190 98Q200 108 306 224T411 342Q302 620 297 625Q288 636 234 637H206Q200 643 200 645T202 664Q206 677 212 683H226Q260 681 347 681Q380 681 408 681T453 682T473 682Q490 682 490 671Q490 670 488 658Q484 643 481 640T465 637Q434 634 411 620L488 426L541 485Q646 598 646 610Q646 628 622 635Q617 635 609 637Q594 637 594 648Q594 650 596 664Q600 677 606 683H618Q619 683 643 683T697 681T738 680Q828 680 837 683H845Q852 676 852 672Q850 647 840 637H824Q790 636 763 628T722 611T698 593L687 584Q687 585 592 480L505 384Q505 383 536 304T601 142T638 56Q648 47 699 46Q734 46 734 37Q734 35 732 23Q728 7 725 4T711 1Q708 1 678 1T589 2Q528 2 496 2T461 1Q444 1 444 10Q444 11 446 25Q448 35 450 39T455 44T464 46T480 47T506 54Q523 62 523 64Q522 64 476 181L429 299Q241 95 236 84Q232 76 232 72Q232 53 261 47Q262 47 267 47T273 46Q276 46 277 46T280 45T283 42T284 35Q284 26 282 19Q279 6 276 4T261 1Q258 1 243 1T201 2T142 2Q64 2 42 0Z"></path><path id="MJX-39-TEX-S4-2013" d="M0 248V285H499V248H0Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="mover"><g data-mml-node="mi"><use data-c="1D44B" xlink:href="#MJX-39-TEX-I-1D44B"></use></g><g data-mml-node="mo" transform="translate(0,615)"><svg width="852" height="237" x="0" y="148" viewBox="213 148 852 237"><use data-c="2013" xlink:href="#MJX-39-TEX-S4-2013" transform="scale(2.556,1)"></use></svg></g></g></g></g></svg><mjx-assistive-mml unselectable="on" display="inline"><math xmlns="http://www.w3.org/1998/Math/MathML"><mover><mi>X</mi><mo accent="true">―</mo></mover></math></mjx-assistive-mml></mjx-container><script type="math/tex"> \overline{X} </script><span> 和 </span><mjx-container class="MathJax" jax="SVG" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="1.726ex" height="2.262ex" role="img" focusable="false" viewBox="0 -1000 763 1000" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: 0px;"><defs><path id="MJX-40-TEX-I-1D44C" d="M66 637Q54 637 49 637T39 638T32 641T30 647T33 664T42 682Q44 683 56 683Q104 680 165 680Q288 680 306 683H316Q322 677 322 674T320 656Q316 643 310 637H298Q242 637 242 624Q242 619 292 477T343 333L346 336Q350 340 358 349T379 373T411 410T454 461Q546 568 561 587T577 618Q577 634 545 637Q528 637 528 647Q528 649 530 661Q533 676 535 679T549 683Q551 683 578 682T657 680Q684 680 713 681T746 682Q763 682 763 673Q763 669 760 657T755 643Q753 637 734 637Q662 632 617 587Q608 578 477 424L348 273L322 169Q295 62 295 57Q295 46 363 46Q379 46 384 45T390 35Q390 33 388 23Q384 6 382 4T366 1Q361 1 324 1T232 2Q170 2 138 2T102 1Q84 1 84 9Q84 14 87 24Q88 27 89 30T90 35T91 39T93 42T96 44T101 45T107 45T116 46T129 46Q168 47 180 50T198 63Q201 68 227 171L252 274L129 623Q128 624 127 625T125 627T122 629T118 631T113 633T105 634T96 635T83 636T66 637Z"></path><path id="MJX-40-TEX-S4-2013" d="M0 248V285H499V248H0Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="mover"><g data-mml-node="mi"><use data-c="1D44C" xlink:href="#MJX-40-TEX-I-1D44C"></use></g><g data-mml-node="mo" transform="translate(0,615)"><svg width="763" height="237" x="0" y="148" viewBox="190.7 148 763 237"><use data-c="2013" xlink:href="#MJX-40-TEX-S4-2013" transform="scale(2.289,1)"></use></svg></g></g></g></g></svg><mjx-assistive-mml unselectable="on" display="inline"><math xmlns="http://www.w3.org/1998/Math/MathML"><mover><mi>Y</mi><mo accent="true">―</mo></mover></math></mjx-assistive-mml></mjx-container><script type="math/tex"> \overline{Y} </script><span> 是两个变量的均值。</span></p><p>&nbsp;</p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="python" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="python"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">import</span> <span class="cm-variable">pandas</span> <span class="cm-keyword">as</span> <span class="cm-variable">pd</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">import</span> <span class="cm-variable">numpy</span> <span class="cm-keyword">as</span> <span class="cm-variable">np</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">import</span> <span class="cm-variable">seaborn</span> <span class="cm-keyword">as</span> <span class="cm-variable">sns</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">import</span> <span class="cm-variable">matplotlib</span>.<span class="cm-property">pyplot</span> <span class="cm-keyword">as</span> <span class="cm-variable">plt</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">from</span> <span class="cm-variable">scipy</span>.<span class="cm-property">stats</span> <span class="cm-keyword">import</span> <span class="cm-variable">pearsonr</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">import</span> <span class="cm-variable">matplotlib</span>.<span class="cm-property">colors</span> <span class="cm-keyword">as</span> <span class="cm-variable">mcolors</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 读取并合并多个样本的片段计数数据</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">frag_count</span> <span class="cm-operator">=</span> <span class="cm-keyword">None</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 输入样本名</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">sample_list</span><span class="cm-operator">=</span>[<span class="cm-string">'KCHIP2'</span>,<span class="cm-string">'KCHIP2-OE'</span>]</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 遍历样本列表,并把所有样本都整合在一个表格内frag_count</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">for</span> <span class="cm-variable">sample</span> <span class="cm-keyword">in</span> <span class="cm-variable">sample_list</span>: &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-variable">frag_count_tmp</span> <span class="cm-operator">=</span> <span class="cm-variable">pd</span>.<span class="cm-property">read_csv</span>(<span class="cm-string">f"./fragmentsCount/</span>{<span class="cm-variable">sample</span>}<span class="cm-string">.fragmentsCount.bed"</span>, </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-variable">sep</span><span class="cm-operator">=</span><span class="cm-string">"\t"</span>, <span class="cm-variable">header</span><span class="cm-operator">=</span><span class="cm-keyword">None</span>, </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-variable">names</span><span class="cm-operator">=</span>[<span class="cm-string">"chrom"</span>, <span class="cm-string">"bin"</span>, <span class="cm-variable">sample</span>],</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-variable">dtype</span><span class="cm-operator">=</span>{<span class="cm-string">"chrom"</span>: <span class="cm-builtin">str</span>, <span class="cm-string">"bin"</span>: <span class="cm-builtin">int</span>, <span class="cm-variable">sample</span>: <span class="cm-builtin">int</span>})</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-keyword">if</span> <span class="cm-variable">frag_count</span> <span class="cm-keyword">is</span> <span class="cm-keyword">None</span>:</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-variable">frag_count</span> <span class="cm-operator">=</span> <span class="cm-variable">frag_count_tmp</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-keyword">else</span>:</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-variable">frag_count</span> <span class="cm-operator">=</span> <span class="cm-variable">pd</span>.<span class="cm-property">merge</span>(<span class="cm-variable">frag_count</span>, <span class="cm-variable">frag_count_tmp</span>, <span class="cm-variable">on</span><span class="cm-operator">=</span>[<span class="cm-string">"chrom"</span>, <span class="cm-string">"bin"</span>], <span class="cm-variable">how</span><span class="cm-operator">=</span><span class="cm-string">"outer"</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">frag_count</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 634px;"></div><div class="CodeMirror-gutters" style="display: none; height: 634px;"></div></div></div></pre><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="python"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="python"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 进行log2(x + 1)转换</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">frag_count_data</span> <span class="cm-operator">=</span> <span class="cm-variable">frag_count</span>.<span class="cm-property">drop</span>(<span class="cm-variable">columns</span><span class="cm-operator">=</span>[<span class="cm-string">"chrom"</span>, <span class="cm-string">"bin"</span>]).<span class="cm-property">apply</span>(<span class="cm-keyword">lambda</span> <span class="cm-variable">x</span>: <span class="cm-variable">np</span>.<span class="cm-property">log2</span>(<span class="cm-variable">x</span> <span class="cm-operator">+</span> <span class="cm-number">1</span>)) &nbsp;<span class="cm-comment"># 添加1以防止取log时出现负无穷</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 计算Pearson相关系数矩阵</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">corr_matrix</span> <span class="cm-operator">=</span> <span class="cm-variable">frag_count_data</span>.<span class="cm-property">corr</span>(<span class="cm-variable">method</span><span class="cm-operator">=</span><span class="cm-string">'pearson'</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">corr_matrix</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 175px;"></div><div class="CodeMirror-gutters" style="display: none; height: 175px;"></div></div></div></pre><p>&nbsp;</p></li><li><p><strong><span>Pearson相关性热图</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="python" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="python"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 创建热图</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">plt</span>.<span class="cm-property">figure</span>(<span class="cm-variable">figsize</span><span class="cm-operator">=</span>(<span class="cm-number">6</span>, <span class="cm-number">5</span>))</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 定义颜色映射：从深蓝到白色再到深红</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">cmap</span> <span class="cm-operator">=</span> <span class="cm-variable">mcolors</span>.<span class="cm-property">LinearSegmentedColormap</span>.<span class="cm-property">from_list</span>(<span class="cm-string">'custom_cmap'</span>, [<span class="cm-string">"midnightblue"</span>, <span class="cm-string">"white"</span>, <span class="cm-string">"darkred"</span>])</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 使用 seaborn 绘制热图</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">sns</span>.<span class="cm-property">heatmap</span>(<span class="cm-variable">corr_matrix</span>, </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-variable">annot</span><span class="cm-operator">=</span><span class="cm-keyword">True</span>, &nbsp;<span class="cm-comment">#表示在单元格内显示相关系数数值</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-variable">fmt</span><span class="cm-operator">=</span><span class="cm-string">".2f"</span>, &nbsp; <span class="cm-comment">#控制显示数值的格式。".2f" 表示保留两位小数</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-variable">cmap</span><span class="cm-operator">=</span><span class="cm-variable">cmap</span>, &nbsp; <span class="cm-comment">#颜色</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-variable">linewidths</span><span class="cm-operator">=</span><span class="cm-number">0.5</span>, </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-variable">linecolor</span><span class="cm-operator">=</span><span class="cm-string">'darkgray'</span>, </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-variable">cbar_kws</span><span class="cm-operator">=</span>{<span class="cm-string">"shrink"</span>: <span class="cm-number">0.8</span>}) &nbsp;<span class="cm-comment">#用于传递与颜色条的参数。这里的 {"shrink": .8} 颜色条的高度将缩小为原来的 80%</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 设置热图标题</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">plt</span>.<span class="cm-property">title</span>(<span class="cm-string">"Correlation Matrix Heatmap"</span>, <span class="cm-variable">fontsize</span><span class="cm-operator">=</span><span class="cm-number">16</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 设置x轴和y轴的标签旋转角度</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">plt</span>.<span class="cm-property">xticks</span>(<span class="cm-variable">rotation</span><span class="cm-operator">=</span><span class="cm-number">90</span>, <span class="cm-variable">fontsize</span><span class="cm-operator">=</span><span class="cm-number">10</span>, <span class="cm-variable">color</span><span class="cm-operator">=</span><span class="cm-string">'darkred'</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">plt</span>.<span class="cm-property">yticks</span>(<span class="cm-variable">fontsize</span><span class="cm-operator">=</span><span class="cm-number">10</span>, <span class="cm-variable">color</span><span class="cm-operator">=</span><span class="cm-string">'darkred'</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 显示图像</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">plt</span>.<span class="cm-property">tight_layout</span>()</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">plt</span>.<span class="cm-property">show</span>()</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 546px;"></div><div class="CodeMirror-gutters" style="display: none; height: 546px;"></div></div></div></pre><p>&nbsp;</p></li></ol><hr /><h1 id='convert-bed-to-bedgraph'><span>Convert BED to Bedgraph</span></h1><h2 id='convert-bed-to-bedgraph-2'><span>Convert BED to Bedgraph</span></h2><ul><li><p><span>在</span><strong><span>CUT&amp;Tag</span></strong><span>和</span><strong><span>CUT&amp;RUN</span></strong><span>中的</span><code>SEACR</code><span>的</span><strong><span>peak calling</span></strong><span>需要Bedgraph文件作为输入</span></p></li><li><p><span>使用</span><code>bedtools genomecov</code><span>将BED文件转换为Bedgraph </span><a href='#bedtools-genomecov'><span>详见点此</span></a></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-meta">#!/bin/bash</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 输入样本名</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-def">sample</span><span class="cm-operator">=</span><span class="cm-string">"sample"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 提供染色体长度信息文件(chrom.sizes)路径。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-def">chromSize</span><span class="cm-operator">=</span><span class="cm-string">"&lt;xxx.chrom.sizes&gt;"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 定义比对到主基因组的文件路径</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-def">primary_alignment</span><span class="cm-operator">=</span><span class="cm-string">"</span><span class="cm-def">${sample}</span><span class="cm-string">.fragments.bed"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 将bed转换成bedtools</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">bedtools genomecov <span class="cm-attribute">-bg</span> &nbsp;<span class="cm-attribute">-i</span> <span class="cm-def">$primary_alignment</span> <span class="cm-attribute">-g</span> <span class="cm-def">$chromSize</span> &gt; <span class="cm-def">${sample}</span>.fragments.bedgraph</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 284px;"></div><div class="CodeMirror-gutters" style="display: none; height: 284px;"></div></div></div></pre><p>&nbsp;</p></li></ul><hr /><h2 id='use-the-ecoli-genome-to-normalize-data-optionalcuttagstep2'><span>Use the E.coli genome to normalize data (optional/CUT&amp;Tag)(step2)</span></h2><ul><li><p><a href='#use-the-ecoli-genome-to-normalize-data-optionalcuttagstep1'><span>Click here to view step 1</span></a></p></li><li><p><span>Using tools：</span><a href='#bedtools-genomecov'><span>Bedtools genomecov</span></a></p></li></ul><p>&nbsp;</p><p><strong><span>Step 2 Spike-in calibration</span></strong><span> </span></p><p><span>假设：对于每个使用相同数量细胞的样本，比对到主基因组和 E. coli 基因组的片段比例应该是相同的。</span></p><ul><li><p><span>通过对E. coli 基因组作为内参对主基因组进行</span><strong><span>calibration</span></strong></p></li><li><p><span>缩放因子 </span><mjx-container class="MathJax" jax="SVG" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="1.459ex" height="1.645ex" role="img" focusable="false" viewBox="0 -705 645 727" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -0.05ex;"><defs><path id="MJX-41-TEX-I-1D446" d="M308 24Q367 24 416 76T466 197Q466 260 414 284Q308 311 278 321T236 341Q176 383 176 462Q176 523 208 573T273 648Q302 673 343 688T407 704H418H425Q521 704 564 640Q565 640 577 653T603 682T623 704Q624 704 627 704T632 705Q645 705 645 698T617 577T585 459T569 456Q549 456 549 465Q549 471 550 475Q550 478 551 494T553 520Q553 554 544 579T526 616T501 641Q465 662 419 662Q362 662 313 616T263 510Q263 480 278 458T319 427Q323 425 389 408T456 390Q490 379 522 342T554 242Q554 216 546 186Q541 164 528 137T492 78T426 18T332 -20Q320 -22 298 -22Q199 -22 144 33L134 44L106 13Q83 -14 78 -18T65 -22Q52 -22 52 -14Q52 -11 110 221Q112 227 130 227H143Q149 221 149 216Q149 214 148 207T144 186T142 153Q144 114 160 87T203 47T255 29T308 24Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="mi"><use data-c="1D446" xlink:href="#MJX-41-TEX-I-1D446"></use></g></g></g></svg><mjx-assistive-mml unselectable="on" display="inline"><math xmlns="http://www.w3.org/1998/Math/MathML"><mi>S</mi></math></mjx-assistive-mml></mjx-container><script type="math/tex"> S </script><span>的定义：为了避免归一化数据中的小数，使用一个常数 </span><mjx-container class="MathJax" jax="SVG" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="1.719ex" height="1.645ex" role="img" focusable="false" viewBox="0 -705 760 727" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -0.05ex;"><defs><path id="MJX-42-TEX-I-1D436" d="M50 252Q50 367 117 473T286 641T490 704Q580 704 633 653Q642 643 648 636T656 626L657 623Q660 623 684 649Q691 655 699 663T715 679T725 690L740 705H746Q760 705 760 698Q760 694 728 561Q692 422 692 421Q690 416 687 415T669 413H653Q647 419 647 422Q647 423 648 429T650 449T651 481Q651 552 619 605T510 659Q484 659 454 652T382 628T299 572T226 479Q194 422 175 346T156 222Q156 108 232 58Q280 24 350 24Q441 24 512 92T606 240Q610 253 612 255T628 257Q648 257 648 248Q648 243 647 239Q618 132 523 55T319 -22Q206 -22 128 53T50 252Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="mi"><use data-c="1D436" xlink:href="#MJX-42-TEX-I-1D436"></use></g></g></g></svg><mjx-assistive-mml unselectable="on" display="inline"><math xmlns="http://www.w3.org/1998/Math/MathML"><mi>𝐶</mi></math></mjx-assistive-mml></mjx-container><script type="math/tex">𝐶</script><span>（通常是 10,000）来定义的</span></p><div contenteditable="false" spellcheck="false" class="mathjax-block md-end-block md-math-block md-rawblock" id="mathjax-n2857" cid="n2857" mdtype="math_block" data-math-tag-before="0" data-math-tag-after="0" data-math-labels="[]"><div class="md-math-container"><mjx-container class="MathJax" jax="SVG" display="true" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="32.494ex" height="5.183ex" role="img" focusable="false" viewBox="0 -1381 14362.6 2291" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -2.059ex;"><defs><path id="MJX-14-TEX-I-1D446" d="M308 24Q367 24 416 76T466 197Q466 260 414 284Q308 311 278 321T236 341Q176 383 176 462Q176 523 208 573T273 648Q302 673 343 688T407 704H418H425Q521 704 564 640Q565 640 577 653T603 682T623 704Q624 704 627 704T632 705Q645 705 645 698T617 577T585 459T569 456Q549 456 549 465Q549 471 550 475Q550 478 551 494T553 520Q553 554 544 579T526 616T501 641Q465 662 419 662Q362 662 313 616T263 510Q263 480 278 458T319 427Q323 425 389 408T456 390Q490 379 522 342T554 242Q554 216 546 186Q541 164 528 137T492 78T426 18T332 -20Q320 -22 298 -22Q199 -22 144 33L134 44L106 13Q83 -14 78 -18T65 -22Q52 -22 52 -14Q52 -11 110 221Q112 227 130 227H143Q149 221 149 216Q149 214 148 207T144 186T142 153Q144 114 160 87T203 47T255 29T308 24Z"></path><path id="MJX-14-TEX-N-3D" d="M56 347Q56 360 70 367H707Q722 359 722 347Q722 336 708 328L390 327H72Q56 332 56 347ZM56 153Q56 168 72 173H708Q722 163 722 153Q722 140 707 133H70Q56 140 56 153Z"></path><path id="MJX-14-TEX-I-1D436" d="M50 252Q50 367 117 473T286 641T490 704Q580 704 633 653Q642 643 648 636T656 626L657 623Q660 623 684 649Q691 655 699 663T715 679T725 690L740 705H746Q760 705 760 698Q760 694 728 561Q692 422 692 421Q690 416 687 415T669 413H653Q647 419 647 422Q647 423 648 429T650 449T651 481Q651 552 619 605T510 659Q484 659 454 652T382 628T299 572T226 479Q194 422 175 346T156 222Q156 108 232 58Q280 24 350 24Q441 24 512 92T606 240Q610 253 612 255T628 257Q648 257 648 248Q648 243 647 239Q618 132 523 55T319 -22Q206 -22 128 53T50 252Z"></path><path id="MJX-14-TEX-N-A0" d=""></path><path id="MJX-14-TEX-I-1D438" d="M492 213Q472 213 472 226Q472 230 477 250T482 285Q482 316 461 323T364 330H312Q311 328 277 192T243 52Q243 48 254 48T334 46Q428 46 458 48T518 61Q567 77 599 117T670 248Q680 270 683 272Q690 274 698 274Q718 274 718 261Q613 7 608 2Q605 0 322 0H133Q31 0 31 11Q31 13 34 25Q38 41 42 43T65 46Q92 46 125 49Q139 52 144 61Q146 66 215 342T285 622Q285 629 281 629Q273 632 228 634H197Q191 640 191 642T193 659Q197 676 203 680H757Q764 676 764 669Q764 664 751 557T737 447Q735 440 717 440H705Q698 445 698 453L701 476Q704 500 704 528Q704 558 697 578T678 609T643 625T596 632T532 634H485Q397 633 392 631Q388 629 386 622Q385 619 355 499T324 377Q347 376 372 376H398Q464 376 489 391T534 472Q538 488 540 490T557 493Q562 493 565 493T570 492T572 491T574 487T577 483L544 351Q511 218 508 216Q505 213 492 213Z"></path><path id="MJX-14-TEX-N-2E" d="M78 60Q78 84 95 102T138 120Q162 120 180 104T199 61Q199 36 182 18T139 0T96 17T78 60Z"></path><path id="MJX-14-TEX-I-1D450" d="M34 159Q34 268 120 355T306 442Q362 442 394 418T427 355Q427 326 408 306T360 285Q341 285 330 295T319 325T330 359T352 380T366 386H367Q367 388 361 392T340 400T306 404Q276 404 249 390Q228 381 206 359Q162 315 142 235T121 119Q121 73 147 50Q169 26 205 26H209Q321 26 394 111Q403 121 406 121Q410 121 419 112T429 98T420 83T391 55T346 25T282 0T202 -11Q127 -11 81 37T34 159Z"></path><path id="MJX-14-TEX-I-1D45C" d="M201 -11Q126 -11 80 38T34 156Q34 221 64 279T146 380Q222 441 301 441Q333 441 341 440Q354 437 367 433T402 417T438 387T464 338T476 268Q476 161 390 75T201 -11ZM121 120Q121 70 147 48T206 26Q250 26 289 58T351 142Q360 163 374 216T388 308Q388 352 370 375Q346 405 306 405Q243 405 195 347Q158 303 140 230T121 120Z"></path><path id="MJX-14-TEX-I-1D459" d="M117 59Q117 26 142 26Q179 26 205 131Q211 151 215 152Q217 153 225 153H229Q238 153 241 153T246 151T248 144Q247 138 245 128T234 90T214 43T183 6T137 -11Q101 -11 70 11T38 85Q38 97 39 102L104 360Q167 615 167 623Q167 626 166 628T162 632T157 634T149 635T141 636T132 637T122 637Q112 637 109 637T101 638T95 641T94 647Q94 649 96 661Q101 680 107 682T179 688Q194 689 213 690T243 693T254 694Q266 694 266 686Q266 675 193 386T118 83Q118 81 118 75T117 65V59Z"></path><path id="MJX-14-TEX-I-1D456" d="M184 600Q184 624 203 642T247 661Q265 661 277 649T290 619Q290 596 270 577T226 557Q211 557 198 567T184 600ZM21 287Q21 295 30 318T54 369T98 420T158 442Q197 442 223 419T250 357Q250 340 236 301T196 196T154 83Q149 61 149 51Q149 26 166 26Q175 26 185 29T208 43T235 78T260 137Q263 149 265 151T282 153Q302 153 302 143Q302 135 293 112T268 61T223 11T161 -11Q129 -11 102 10T74 74Q74 91 79 106T122 220Q160 321 166 341T173 380Q173 404 156 404H154Q124 404 99 371T61 287Q60 286 59 284T58 281T56 279T53 278T49 278T41 278H27Q21 284 21 287Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="mi"><use data-c="1D446" xlink:href="#MJX-14-TEX-I-1D446"></use></g><g data-mml-node="mo" transform="translate(922.8,0)"><use data-c="3D" xlink:href="#MJX-14-TEX-N-3D"></use></g><g data-mml-node="mfrac" transform="translate(1978.6,0)"><g data-mml-node="mi" transform="translate(5812,676)"><use data-c="1D436" xlink:href="#MJX-14-TEX-I-1D436"></use></g><g data-mml-node="mrow" transform="translate(220,-710)"><g data-mml-node="mtext"><text data-variant="normal" transform="scale(1,-1)" font-size="852.2px" font-family="serif">比</text><text data-variant="normal" transform="translate(867.4,0) scale(1,-1)" font-size="852.2px" font-family="serif">对</text><text data-variant="normal" transform="translate(1734.9,0) scale(1,-1)" font-size="852.2px" font-family="serif">到</text><use data-c="A0" xlink:href="#MJX-14-TEX-N-A0" transform="translate(2602.3,0)"></use></g><g data-mml-node="mi" transform="translate(2852.3,0)"><use data-c="1D438" xlink:href="#MJX-14-TEX-I-1D438"></use></g><g data-mml-node="mo" transform="translate(3616.3,0)"><use data-c="2E" xlink:href="#MJX-14-TEX-N-2E"></use></g><g data-mml-node="mi" transform="translate(4061,0)"><use data-c="1D450" xlink:href="#MJX-14-TEX-I-1D450"></use></g><g data-mml-node="mi" transform="translate(4494,0)"><use data-c="1D45C" xlink:href="#MJX-14-TEX-I-1D45C"></use></g><g data-mml-node="mi" transform="translate(4979,0)"><use data-c="1D459" xlink:href="#MJX-14-TEX-I-1D459"></use></g><g data-mml-node="mi" transform="translate(5277,0)"><use data-c="1D456" xlink:href="#MJX-14-TEX-I-1D456"></use></g><g data-mml-node="mtext" transform="translate(5622,0)"><use data-c="A0" xlink:href="#MJX-14-TEX-N-A0"></use><text data-variant="normal" transform="translate(250,0) scale(1,-1)" font-size="852.2px" font-family="serif">基</text><text data-variant="normal" transform="translate(1117.4,0) scale(1,-1)" font-size="852.2px" font-family="serif">因</text><text data-variant="normal" transform="translate(1984.9,0) scale(1,-1)" font-size="852.2px" font-family="serif">组</text><text data-variant="normal" transform="translate(2852.3,0) scale(1,-1)" font-size="852.2px" font-family="serif">的</text><text data-variant="normal" transform="translate(3719.7,0) scale(1,-1)" font-size="852.2px" font-family="serif">片</text><text data-variant="normal" transform="translate(4587.2,0) scale(1,-1)" font-size="852.2px" font-family="serif">段</text><text data-variant="normal" transform="translate(5454.6,0) scale(1,-1)" font-size="852.2px" font-family="serif">数</text></g></g><rect width="12144" height="60" x="120" y="220"></rect></g></g></g></svg><mjx-assistive-mml unselectable="on" display="block"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><mi>S</mi><mo>=</mo><mfrac><mi>C</mi><mrow><mtext>比对到&nbsp;</mtext><mi>E</mi><mo>.</mo><mi>c</mi><mi>o</mi><mi>l</mi><mi>i</mi><mtext>&nbsp;基因组的片段数</mtext></mrow></mfrac></math></mjx-assistive-mml></mjx-container></div></div></li><li><p><span>归一化的计算公式为：</span></p><div contenteditable="false" spellcheck="false" class="mathjax-block md-end-block md-math-block md-rawblock" id="mathjax-n2860" cid="n2860" mdtype="math_block" data-math-tag-before="0" data-math-tag-after="0" data-math-labels="[]"><div class="md-math-container"><mjx-container class="MathJax" jax="SVG" display="true" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="54.217ex" height="2.262ex" role="img" focusable="false" viewBox="0 -750 23964 1000" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -0.566ex;"><defs><path id="MJX-15-TEX-N-4E" d="M42 46Q74 48 94 56T118 69T128 86V634H124Q114 637 52 637H25V683H232L235 680Q237 679 322 554T493 303L578 178V598Q572 608 568 613T544 627T492 637H475V683H483Q498 680 600 680Q706 680 715 683H724V637H707Q634 633 622 598L621 302V6L614 0H600Q585 0 582 3T481 150T282 443T171 605V345L172 86Q183 50 257 46H274V0H265Q250 3 150 3Q48 3 33 0H25V46H42Z"></path><path id="MJX-15-TEX-N-6F" d="M28 214Q28 309 93 378T250 448Q340 448 405 380T471 215Q471 120 407 55T250 -10Q153 -10 91 57T28 214ZM250 30Q372 30 372 193V225V250Q372 272 371 288T364 326T348 362T317 390T268 410Q263 411 252 411Q222 411 195 399Q152 377 139 338T126 246V226Q126 130 145 91Q177 30 250 30Z"></path><path id="MJX-15-TEX-N-72" d="M36 46H50Q89 46 97 60V68Q97 77 97 91T98 122T98 161T98 203Q98 234 98 269T98 328L97 351Q94 370 83 376T38 385H20V408Q20 431 22 431L32 432Q42 433 60 434T96 436Q112 437 131 438T160 441T171 442H174V373Q213 441 271 441H277Q322 441 343 419T364 373Q364 352 351 337T313 322Q288 322 276 338T263 372Q263 381 265 388T270 400T273 405Q271 407 250 401Q234 393 226 386Q179 341 179 207V154Q179 141 179 127T179 101T180 81T180 66V61Q181 59 183 57T188 54T193 51T200 49T207 48T216 47T225 47T235 46T245 46H276V0H267Q249 3 140 3Q37 3 28 0H20V46H36Z"></path><path id="MJX-15-TEX-N-6D" d="M41 46H55Q94 46 102 60V68Q102 77 102 91T102 122T103 161T103 203Q103 234 103 269T102 328V351Q99 370 88 376T43 385H25V408Q25 431 27 431L37 432Q47 433 65 434T102 436Q119 437 138 438T167 441T178 442H181V402Q181 364 182 364T187 369T199 384T218 402T247 421T285 437Q305 442 336 442Q351 442 364 440T387 434T406 426T421 417T432 406T441 395T448 384T452 374T455 366L457 361L460 365Q463 369 466 373T475 384T488 397T503 410T523 422T546 432T572 439T603 442Q729 442 740 329Q741 322 741 190V104Q741 66 743 59T754 49Q775 46 803 46H819V0H811L788 1Q764 2 737 2T699 3Q596 3 587 0H579V46H595Q656 46 656 62Q657 64 657 200Q656 335 655 343Q649 371 635 385T611 402T585 404Q540 404 506 370Q479 343 472 315T464 232V168V108Q464 78 465 68T468 55T477 49Q498 46 526 46H542V0H534L510 1Q487 2 460 2T422 3Q319 3 310 0H302V46H318Q379 46 379 62Q380 64 380 200Q379 335 378 343Q372 371 358 385T334 402T308 404Q263 404 229 370Q202 343 195 315T187 232V168V108Q187 78 188 68T191 55T200 49Q221 46 249 46H265V0H257L234 1Q210 2 183 2T145 3Q42 3 33 0H25V46H41Z"></path><path id="MJX-15-TEX-N-61" d="M137 305T115 305T78 320T63 359Q63 394 97 421T218 448Q291 448 336 416T396 340Q401 326 401 309T402 194V124Q402 76 407 58T428 40Q443 40 448 56T453 109V145H493V106Q492 66 490 59Q481 29 455 12T400 -6T353 12T329 54V58L327 55Q325 52 322 49T314 40T302 29T287 17T269 6T247 -2T221 -8T190 -11Q130 -11 82 20T34 107Q34 128 41 147T68 188T116 225T194 253T304 268H318V290Q318 324 312 340Q290 411 215 411Q197 411 181 410T156 406T148 403Q170 388 170 359Q170 334 154 320ZM126 106Q126 75 150 51T209 26Q247 26 276 49T315 109Q317 116 318 175Q318 233 317 233Q309 233 296 232T251 223T193 203T147 166T126 106Z"></path><path id="MJX-15-TEX-N-6C" d="M42 46H56Q95 46 103 60V68Q103 77 103 91T103 124T104 167T104 217T104 272T104 329Q104 366 104 407T104 482T104 542T103 586T103 603Q100 622 89 628T44 637H26V660Q26 683 28 683L38 684Q48 685 67 686T104 688Q121 689 141 690T171 693T182 694H185V379Q185 62 186 60Q190 52 198 49Q219 46 247 46H263V0H255L232 1Q209 2 183 2T145 3T107 3T57 1L34 0H26V46H42Z"></path><path id="MJX-15-TEX-N-69" d="M69 609Q69 637 87 653T131 669Q154 667 171 652T188 609Q188 579 171 564T129 549Q104 549 87 564T69 609ZM247 0Q232 3 143 3Q132 3 106 3T56 1L34 0H26V46H42Q70 46 91 49Q100 53 102 60T104 102V205V293Q104 345 102 359T88 378Q74 385 41 385H30V408Q30 431 32 431L42 432Q52 433 70 434T106 436Q123 437 142 438T171 441T182 442H185V62Q190 52 197 50T232 46H255V0H247Z"></path><path id="MJX-15-TEX-N-7A" d="M42 263Q44 270 48 345T53 423V431H393Q399 425 399 415Q399 403 398 402L381 378Q364 355 331 309T265 220L134 41L182 40H206Q254 40 283 46T331 77Q352 105 359 185L361 201Q361 202 381 202H401V196Q401 195 393 103T384 6V0H209L34 1L31 3Q28 8 28 17Q28 30 29 31T160 210T294 394H236Q169 393 152 388Q127 382 113 367Q89 344 82 264V255H42V263Z"></path><path id="MJX-15-TEX-N-65" d="M28 218Q28 273 48 318T98 391T163 433T229 448Q282 448 320 430T378 380T406 316T415 245Q415 238 408 231H126V216Q126 68 226 36Q246 30 270 30Q312 30 342 62Q359 79 369 104L379 128Q382 131 395 131H398Q415 131 415 121Q415 117 412 108Q393 53 349 21T250 -11Q155 -11 92 58T28 218ZM333 275Q322 403 238 411H236Q228 411 220 410T195 402T166 381T143 340T127 274V267H333V275Z"></path><path id="MJX-15-TEX-N-64" d="M376 495Q376 511 376 535T377 568Q377 613 367 624T316 637H298V660Q298 683 300 683L310 684Q320 685 339 686T376 688Q393 689 413 690T443 693T454 694H457V390Q457 84 458 81Q461 61 472 55T517 46H535V0Q533 0 459 -5T380 -11H373V44L365 37Q307 -11 235 -11Q158 -11 96 50T34 215Q34 315 97 378T244 442Q319 442 376 393V495ZM373 342Q328 405 260 405Q211 405 173 369Q146 341 139 305T131 211Q131 155 138 120T173 59Q203 26 251 26Q322 26 373 103V342Z"></path><path id="MJX-15-TEX-N-20" d=""></path><path id="MJX-15-TEX-N-63" d="M370 305T349 305T313 320T297 358Q297 381 312 396Q317 401 317 402T307 404Q281 408 258 408Q209 408 178 376Q131 329 131 219Q131 137 162 90Q203 29 272 29Q313 29 338 55T374 117Q376 125 379 127T395 129H409Q415 123 415 120Q415 116 411 104T395 71T366 33T318 2T249 -11Q163 -11 99 53T34 214Q34 318 99 383T250 448T370 421T404 357Q404 334 387 320Z"></path><path id="MJX-15-TEX-N-76" d="M338 431Q344 429 422 429Q479 429 503 431H508V385H497Q439 381 423 345Q421 341 356 172T288 -2Q283 -11 263 -11Q244 -11 239 -2Q99 359 98 364Q93 378 82 381T43 385H19V431H25L33 430Q41 430 53 430T79 430T104 429T122 428Q217 428 232 431H240V385H226Q187 384 184 370Q184 366 235 234L286 102L377 341V349Q377 363 367 372T349 383T335 385H331V431H338Z"></path><path id="MJX-15-TEX-N-67" d="M329 409Q373 453 429 453Q459 453 472 434T485 396Q485 382 476 371T449 360Q416 360 412 390Q410 404 415 411Q415 412 416 414V415Q388 412 363 393Q355 388 355 386Q355 385 359 381T368 369T379 351T388 325T392 292Q392 230 343 187T222 143Q172 143 123 171Q112 153 112 133Q112 98 138 81Q147 75 155 75T227 73Q311 72 335 67Q396 58 431 26Q470 -13 470 -72Q470 -139 392 -175Q332 -206 250 -206Q167 -206 107 -175Q29 -140 29 -75Q29 -39 50 -15T92 18L103 24Q67 55 67 108Q67 155 96 193Q52 237 52 292Q52 355 102 398T223 442Q274 442 318 416L329 409ZM299 343Q294 371 273 387T221 404Q192 404 171 388T145 343Q142 326 142 292Q142 248 149 227T179 192Q196 182 222 182Q244 182 260 189T283 207T294 227T299 242Q302 258 302 292T299 343ZM403 -75Q403 -50 389 -34T348 -11T299 -2T245 0H218Q151 0 138 -6Q118 -15 107 -34T95 -74Q95 -84 101 -97T122 -127T170 -155T250 -167Q319 -167 361 -139T403 -75Z"></path><path id="MJX-15-TEX-N-3D" d="M56 347Q56 360 70 367H707Q722 359 722 347Q722 336 708 328L390 327H72Q56 332 56 347ZM56 153Q56 168 72 173H708Q722 163 722 153Q722 140 707 133H70Q56 140 56 153Z"></path><path id="MJX-15-TEX-N-28" d="M94 250Q94 319 104 381T127 488T164 576T202 643T244 695T277 729T302 750H315H319Q333 750 333 741Q333 738 316 720T275 667T226 581T184 443T167 250T184 58T225 -81T274 -167T316 -220T333 -241Q333 -250 318 -250H315H302L274 -226Q180 -141 137 -14T94 250Z"></path><path id="MJX-15-TEX-N-70" d="M36 -148H50Q89 -148 97 -134V-126Q97 -119 97 -107T97 -77T98 -38T98 6T98 55T98 106Q98 140 98 177T98 243T98 296T97 335T97 351Q94 370 83 376T38 385H20V408Q20 431 22 431L32 432Q42 433 61 434T98 436Q115 437 135 438T165 441T176 442H179V416L180 390L188 397Q247 441 326 441Q407 441 464 377T522 216Q522 115 457 52T310 -11Q242 -11 190 33L182 40V-45V-101Q182 -128 184 -134T195 -145Q216 -148 244 -148H260V-194H252L228 -193Q205 -192 178 -192T140 -191Q37 -191 28 -194H20V-148H36ZM424 218Q424 292 390 347T305 402Q234 402 182 337V98Q222 26 294 26Q345 26 384 80T424 218Z"></path><path id="MJX-15-TEX-N-79" d="M69 -66Q91 -66 104 -80T118 -116Q118 -134 109 -145T91 -160Q84 -163 97 -166Q104 -168 111 -168Q131 -168 148 -159T175 -138T197 -106T213 -75T225 -43L242 0L170 183Q150 233 125 297Q101 358 96 368T80 381Q79 382 78 382Q66 385 34 385H19V431H26L46 430Q65 430 88 429T122 428Q129 428 142 428T171 429T200 430T224 430L233 431H241V385H232Q183 385 185 366L286 112Q286 113 332 227L376 341V350Q376 365 366 373T348 383T334 385H331V431H337H344Q351 431 361 431T382 430T405 429T422 429Q477 429 503 431H508V385H497Q441 380 422 345Q420 343 378 235T289 9T227 -131Q180 -204 113 -204Q69 -204 44 -177T19 -116Q19 -89 35 -78T69 -66Z"></path><path id="MJX-15-TEX-N-6E" d="M41 46H55Q94 46 102 60V68Q102 77 102 91T102 122T103 161T103 203Q103 234 103 269T102 328V351Q99 370 88 376T43 385H25V408Q25 431 27 431L37 432Q47 433 65 434T102 436Q119 437 138 438T167 441T178 442H181V402Q181 364 182 364T187 369T199 384T218 402T247 421T285 437Q305 442 336 442Q450 438 463 329Q464 322 464 190V104Q464 66 466 59T477 49Q498 46 526 46H542V0H534L510 1Q487 2 460 2T422 3Q319 3 310 0H302V46H318Q379 46 379 62Q380 64 380 200Q379 335 378 343Q372 371 358 385T334 402T308 404Q263 404 229 370Q202 343 195 315T187 232V168V108Q187 78 188 68T191 55T200 49Q221 46 249 46H265V0H257L234 1Q210 2 183 2T145 3Q42 3 33 0H25V46H41Z"></path><path id="MJX-15-TEX-N-29" d="M60 749L64 750Q69 750 74 750H86L114 726Q208 641 251 514T294 250Q294 182 284 119T261 12T224 -76T186 -143T145 -194T113 -227T90 -246Q87 -249 86 -250H74Q66 -250 63 -250T58 -247T55 -238Q56 -237 66 -225Q221 -64 221 250T66 725Q56 737 55 738Q55 746 60 749Z"></path><path id="MJX-15-TEX-N-D7" d="M630 29Q630 9 609 9Q604 9 587 25T493 118L389 222L284 117Q178 13 175 11Q171 9 168 9Q160 9 154 15T147 29Q147 36 161 51T255 146L359 250L255 354Q174 435 161 449T147 471Q147 480 153 485T168 490Q173 490 175 489Q178 487 284 383L389 278L493 382Q570 459 587 475T609 491Q630 491 630 471Q630 464 620 453T522 355L418 250L522 145Q606 61 618 48T630 29Z"></path><path id="MJX-15-TEX-I-1D446" d="M308 24Q367 24 416 76T466 197Q466 260 414 284Q308 311 278 321T236 341Q176 383 176 462Q176 523 208 573T273 648Q302 673 343 688T407 704H418H425Q521 704 564 640Q565 640 577 653T603 682T623 704Q624 704 627 704T632 705Q645 705 645 698T617 577T585 459T569 456Q549 456 549 465Q549 471 550 475Q550 478 551 494T553 520Q553 554 544 579T526 616T501 641Q465 662 419 662Q362 662 313 616T263 510Q263 480 278 458T319 427Q323 425 389 408T456 390Q490 379 522 342T554 242Q554 216 546 186Q541 164 528 137T492 78T426 18T332 -20Q320 -22 298 -22Q199 -22 144 33L134 44L106 13Q83 -14 78 -18T65 -22Q52 -22 52 -14Q52 -11 110 221Q112 227 130 227H143Q149 221 149 216Q149 214 148 207T144 186T142 153Q144 114 160 87T203 47T255 29T308 24Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="mtext"><use data-c="4E" xlink:href="#MJX-15-TEX-N-4E"></use><use data-c="6F" xlink:href="#MJX-15-TEX-N-6F" transform="translate(750,0)"></use><use data-c="72" xlink:href="#MJX-15-TEX-N-72" transform="translate(1250,0)"></use><use data-c="6D" xlink:href="#MJX-15-TEX-N-6D" transform="translate(1642,0)"></use><use data-c="61" xlink:href="#MJX-15-TEX-N-61" transform="translate(2475,0)"></use><use data-c="6C" xlink:href="#MJX-15-TEX-N-6C" transform="translate(2975,0)"></use><use data-c="69" xlink:href="#MJX-15-TEX-N-69" transform="translate(3253,0)"></use><use data-c="7A" xlink:href="#MJX-15-TEX-N-7A" transform="translate(3531,0)"></use><use data-c="65" xlink:href="#MJX-15-TEX-N-65" transform="translate(3975,0)"></use><use data-c="64" xlink:href="#MJX-15-TEX-N-64" transform="translate(4419,0)"></use><use data-c="20" xlink:href="#MJX-15-TEX-N-20" transform="translate(4975,0)"></use><use data-c="63" xlink:href="#MJX-15-TEX-N-63" transform="translate(5225,0)"></use><use data-c="6F" xlink:href="#MJX-15-TEX-N-6F" transform="translate(5669,0)"></use><use data-c="76" xlink:href="#MJX-15-TEX-N-76" transform="translate(6169,0)"></use><use data-c="65" xlink:href="#MJX-15-TEX-N-65" transform="translate(6697,0)"></use><use data-c="72" xlink:href="#MJX-15-TEX-N-72" transform="translate(7141,0)"></use><use data-c="61" xlink:href="#MJX-15-TEX-N-61" transform="translate(7533,0)"></use><use data-c="67" xlink:href="#MJX-15-TEX-N-67" transform="translate(8033,0)"></use><use data-c="65" xlink:href="#MJX-15-TEX-N-65" transform="translate(8533,0)"></use></g><g data-mml-node="mo" transform="translate(9254.8,0)"><use data-c="3D" xlink:href="#MJX-15-TEX-N-3D"></use></g><g data-mml-node="mo" transform="translate(10310.6,0)"><use data-c="28" xlink:href="#MJX-15-TEX-N-28"></use></g><g data-mml-node="mtext" transform="translate(10699.6,0)"><use data-c="70" xlink:href="#MJX-15-TEX-N-70"></use><use data-c="72" xlink:href="#MJX-15-TEX-N-72" transform="translate(556,0)"></use><use data-c="69" xlink:href="#MJX-15-TEX-N-69" transform="translate(948,0)"></use><use data-c="6D" xlink:href="#MJX-15-TEX-N-6D" transform="translate(1226,0)"></use><use data-c="61" xlink:href="#MJX-15-TEX-N-61" transform="translate(2059,0)"></use><use data-c="72" xlink:href="#MJX-15-TEX-N-72" transform="translate(2559,0)"></use><use data-c="79" xlink:href="#MJX-15-TEX-N-79" transform="translate(2951,0)"></use><use data-c="20" xlink:href="#MJX-15-TEX-N-20" transform="translate(3479,0)"></use><use data-c="67" xlink:href="#MJX-15-TEX-N-67" transform="translate(3729,0)"></use><use data-c="65" xlink:href="#MJX-15-TEX-N-65" transform="translate(4229,0)"></use><use data-c="6E" xlink:href="#MJX-15-TEX-N-6E" transform="translate(4673,0)"></use><use data-c="6F" xlink:href="#MJX-15-TEX-N-6F" transform="translate(5229,0)"></use><use data-c="6D" xlink:href="#MJX-15-TEX-N-6D" transform="translate(5729,0)"></use><use data-c="65" xlink:href="#MJX-15-TEX-N-65" transform="translate(6562,0)"></use><use data-c="20" xlink:href="#MJX-15-TEX-N-20" transform="translate(7006,0)"></use><use data-c="63" xlink:href="#MJX-15-TEX-N-63" transform="translate(7256,0)"></use><use data-c="6F" xlink:href="#MJX-15-TEX-N-6F" transform="translate(7700,0)"></use><use data-c="76" xlink:href="#MJX-15-TEX-N-76" transform="translate(8200,0)"></use><use data-c="65" xlink:href="#MJX-15-TEX-N-65" transform="translate(8728,0)"></use><use data-c="72" xlink:href="#MJX-15-TEX-N-72" transform="translate(9172,0)"></use><use data-c="61" xlink:href="#MJX-15-TEX-N-61" transform="translate(9564,0)"></use><use data-c="67" xlink:href="#MJX-15-TEX-N-67" transform="translate(10064,0)"></use><use data-c="65" xlink:href="#MJX-15-TEX-N-65" transform="translate(10564,0)"></use></g><g data-mml-node="mo" transform="translate(21707.6,0)"><use data-c="29" xlink:href="#MJX-15-TEX-N-29"></use></g><g data-mml-node="mo" transform="translate(22318.8,0)"><use data-c="D7" xlink:href="#MJX-15-TEX-N-D7"></use></g><g data-mml-node="mi" transform="translate(23319,0)"><use data-c="1D446" xlink:href="#MJX-15-TEX-I-1D446"></use></g></g></g></svg><mjx-assistive-mml unselectable="on" display="block"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><mtext>Normalized coverage</mtext><mo>=</mo><mo stretchy="false">(</mo><mtext>primary genome coverage</mtext><mo stretchy="false">)</mo><mo>×</mo><mi>S</mi></math></mjx-assistive-mml></mjx-container></div></div></li></ul><p>&nbsp;</p><p><strong><span>calibration process</span></strong></p><ul><li><p><span>根据比对到Ecoil的reads计算测序深度,其输出到</span><code>sample.Ecoil.seqDepth</code><span>文件</span></p></li><li><p><span>计算缩放因子</span></p></li><li><p><span>使用</span><a href='#bedtools-genomecov'><span>Bedtools genomecov</span></a><span>工具对文件执行 normalization</span></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-meta">#!/bin/bash</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 输入样本名</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-def">sample</span><span class="cm-operator">=</span><span class="cm-string">"sample"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 提供染色体长度信息文件(chrom.sizes)路径。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-def">chromSize</span><span class="cm-operator">=</span><span class="cm-string">"&lt;xxx.chrom.sizes&gt;"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 定义比对到E.COli的文件路径</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-def">Ecoil_alignment</span><span class="cm-operator">=</span><span class="cm-string">"Ecoil/&lt;sample.bam/sam/cram&gt;"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 定义比对到主基因组的文件路径</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-def">primary_alignment</span><span class="cm-operator">=</span><span class="cm-string">"</span><span class="cm-def">${sample}</span><span class="cm-string">.fragments.bed"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 创建bedgraph文件夹用于输出</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-builtin">mkdir</span> <span class="cm-attribute">-p</span> ./bedgraph</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># samtools view查看比对成功的reads（-F 0x04）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># wc -l 执行后返回的行数，即比对成功的 reads 数量</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-def">seqDepthDouble</span><span class="cm-operator">=</span><span class="cm-quote">$(samtools view -F 0x04 </span><span class="cm-def">$Ecoil_aligement</span><span class="cm-quote"> | wc -l)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 测序深度等于 reads 数量除以 2（假设是双端测序）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-def">seqDepth</span><span class="cm-operator">=</span><span class="cm-quote">$((seqDepthDouble / 2))</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 输出测序深度到sample.Ecoil.seqDepth文件</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-builtin">echo</span> <span class="cm-def">$seqDepth</span> &gt; Ecoil/<span class="cm-def">${sample}</span>.Ecoil.seqDepth</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 检查seqDepth（测序深度）是否大于 1，大于1后进行以下操作</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">if</span> [[ <span class="cm-string">"</span><span class="cm-def">$seqDepth</span><span class="cm-string">"</span> <span class="cm-attribute">-gt</span> <span class="cm-string">"1"</span> ]]; <span class="cm-keyword">then</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 计算缩放因子S</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># bc -l 会加载数学库，从而允许进行浮点数运算并支持一些数学函数（如正弦、余弦、对数等）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 此外，它还会将默认的小数位数设为 20 位，这样可以进行更精确的浮点计算。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-def">scale_factor</span><span class="cm-operator">=</span><span class="cm-quote">$(echo "10000 / </span><span class="cm-def">$seqDepth</span><span class="cm-quote">" | bc -l)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 使用echo把缩放因子S打印到终端</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-builtin">echo</span> <span class="cm-string">"Scaling factor for </span><span class="cm-def">$sample</span><span class="cm-string"> is: </span><span class="cm-def">$scale_factor</span><span class="cm-string">!"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 使用bedtools生成归一化的bedgraph文件 </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;  bedtools genomecov <span class="cm-attribute">-bg</span> <span class="cm-attribute">-scale</span> <span class="cm-def">$scale_factor</span> <span class="cm-attribute">-i</span> <span class="cm-def">$primary_alignment</span> <span class="cm-attribute">-g</span> <span class="cm-def">$chromSize</span> &gt; bedgraph/<span class="cm-def">${sample}</span>.fragments.normalized.bedgraph</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">fi</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 940px;"></div><div class="CodeMirror-gutters" style="display: none; height: 940px;"></div></div></div></pre></li></ul><p>&nbsp;</p><hr /><h1 id='peak-calling'><span>Peak calling</span></h1><h2 id='macs2'><span>macs2</span></h2><ul><li><p><a href='https://github.com/macs3-project/MACS/wiki'><span>Home · macs3-project/MACS Wiki</span></a></p></li><li><p><strong><span>Software installation:</span></strong><span> </span><code>conda install -y macs2</code></p><p><span>	</span></p></li></ul><hr /><h2 id='seacrfor-cuttag-and-cutrun'><span>SEACR(For CUT&amp;Tag and CUT&amp;RUN)</span></h2><ul><li><p><span>在ChIP-seq数据分析中，常用的方法是</span><strong><span>peak calling</span></strong><span>，通过比较目标ChIP-seq信号与背景噪音来识别基因组中信号富集的区域。这类算法通常采用泊松分布或负二项分布模型来衡量信号与背景的差异。然而，由于ChIP-seq实验通常测序较深，背景噪音较高，因此大多数ChIP-seq峰值识别算法优化时更注重高灵敏度，以便从噪音中区分出信号。</span></p></li><li><p><span>CUT&amp;Tag和CUT&amp;RUN 的数据背景噪音比ChIP-seq低得多。该数据的低背景和较少的读取深度使得传统的峰值识别算法容易受到误报的影响，从而降低了精确性</span></p></li><li><p><span>针对CUT&amp;Tag和CUT&amp;RUN，</span><strong><span>peak calling</span></strong><span>算法需要具备较高的特异性，因此使用</span><strong><span>稀疏富集分析工具（SEACR）</span></strong></p></li></ul><ul><li><p><code>SEACR</code><span> requires </span><strong><span>bedGraph</span></strong><span> files from </span><strong><span>paired-end</span></strong><span> sequencing as input</span></p></li><li><p><span>If you have </span><strong><span>normalized fragment counts with the E. coli read count</span></strong><span>, we set the normalization option of SEACR to </span><strong><span>non</span></strong></p><p><span>Otherwise, the </span><strong><span>norm</span></strong><span> is recommended. </span></p></li></ul><p>&nbsp;</p><h3 id='seacr-theory'><span>SEACR Theory</span></h3><ul><li><p><a href='https://epigeneticsandchromatin.biomedcentral.com/articles/10.1186/s13072-019-0287-4'><span>SEACR</span></a></p></li><li><p><span>首先将来自目标抗体和 IgG 对照实验数据解析为</span><strong><span>signal blocks</span></strong><span>(连续、非零读数的reads)</span></p></li><li><p><span>The signal in each block is calculated by </span><strong><span>summing read counts</span></strong></p></li><li><p><span>通过绘制</span><strong><span>目标signal blocks</span></strong><span>与 </span><strong><span>IgG signal blocks</span></strong><span>的比例图，确定</span><strong><span>目标signal blocks</span></strong><span>与</span><strong><span>IgG signal blocks</span></strong><span>比例达到最大值的阈值；然后过滤掉未达到阈值的</span><strong><span>目标signal blocks</span></strong></p></li><li><p><span>同时，SEACR还过滤掉与IgG信号重叠的目标signal blocks，以去除可能的假阳性信号。</span></p></li></ul><p><img src="./picture/047.png" width="700"/></p><p>&nbsp;</p><h3 id='seacr-选项'><span>SEACR 选项</span></h3><p><span>为了简化用户的分析流程，SEACR 仅提供了两个主要的选项</span></p><ul><li><p><span>IgG数据集或全局阈值：</span></p><ul><li><p><strong><span>IgG数据集</span></strong><span>：默认选项，用户可以提供 IgG作为对照样本数据集，用于去除背景噪声。</span><strong><span>通常建议使用这种方法</span></strong><span>。</span></p></li><li><p><strong><span>全局阈值</span></strong><span>：如果用户没有 IgG 对照样本，SEACR 允许使用全局阈值进行峰值调用。</span></p></li></ul></li><li><p><span>stringent pattern and relaxed pattern：</span></p><ul><li><p><strong><span>stringent</span></strong><span>：默认选项，它设定一个阈值，使得能够保留最大比例的目标信号块，而对照 IgG 信号块尽可能被排除。这种模式强调高精度，通常用于信号较为强烈的实验。</span></p></li><li><p><strong><span>relaxed</span></strong><span>：这种模式的阈值较为灵活，设定在目标信号比例曲线的最高点和“膝点”（曲线开始急剧变化的地方）之间的中间位置。这样可以检测到更多的峰值，但可能包含更多噪声。</span></p></li></ul><p><img src="./picture/053.png" width="700"/></p></li></ul><p>&nbsp;</p><h3 id='seacr安装与用法'><span>SEACR安装与用法</span></h3><ul><li><p><a href='https://github.com/FredHutch/SEACR'><span>SEACR GitHub</span></a></p></li><li><p><span>SEACR是一个sh脚本，其1.3版本下载：</span><code>wget https://raw.githubusercontent.com/FredHutch/SEACR/master/SEACR_1.3.sh</code></p></li><li><p><span>SEACR需要R的支持，其1.3的R脚本下载：</span><code>wget https://github.com/FredHutch/SEACR/blob/master/SEACR_1.3.R</code></p></li><li><p><strong><span>其SEACR的R脚本需要和SEACR的sh脚本需要放在同一文件下才能正常运行</span></strong></p></li><li><p><span>SEACR运行：</span><code>bash SEACR_1.3.sh &lt;signalFile&gt; &lt;controlFile&gt; &lt;norm&gt; &lt;stringency&gt; &lt;outputPrefix&gt;</code></p><ul><li><p><code>signalFile</code><span>: 实验样本的 BEDgraph 文件。</span></p></li><li><p><code>controlFile</code><span>:</span></p></li><li><p><span>有对照组：对照样本的 BEDgraph 文件</span></p></li><li><p><span>无对照组：使用介于 </span><code>0</code><span> 和 </span><code>1</code><span> 之间的全局阈值 </span><code>n</code><span>，选择信号最高的前 n% 的区域作为峰值。（推荐</span><code>0.01</code><span>）</span></p></li></ul></li><li><p><strong><code>norm</code></strong><span>: 归一化处理，有两个选项：</span></p><ul><li><p><code>non</code><span>: 不进行归一化处理。</span></p></li><li><p><code>norm</code><span>: 进行归一化处理，并从信号中减去对照噪声。</span></p></li><li><p><code>stringency</code><span>: 指定严格性阈值，两个选项：</span></p></li><li><p><code>relaxed</code><span>: 更宽松的峰调用标准，允许检测到更多峰。</span></p></li><li><p><code>stringent</code><span>: 严格的标准，检测到的峰数量相对更少，但精确度更高。</span></p></li></ul></li><li><p><code>outputPrefix</code><span>: 输出文件的前缀，结果将输出为 .bed 格式。</span></p></li></ul><p>&nbsp;</p><p><strong><span>Example</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-meta">#!/bin/bash</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># creat file to store resuts</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-builtin">mkdir</span> <span class="cm-attribute">-p</span> SEACR</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># path of SEACR</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-def">seacr</span><span class="cm-operator">=</span><span class="cm-string">"SEACR_1.3.sh"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># samples name</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-def">samples</span><span class="cm-operator">=</span>(KCHIP2 KCHIP2-OE)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># path of controlFile</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-def">controlFile</span><span class="cm-operator">=</span><span class="cm-string">"</span><span class="cm-def">${samples[0]}</span><span class="cm-string">.fragments.bedgraph"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># define Top peaks threshold</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-def">peaks</span><span class="cm-operator">=</span>(0.01 <span class="cm-number">0</span>.05)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># traversa samples</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">for</span> sample <span class="cm-keyword">in</span> <span class="cm-string">"</span><span class="cm-def">${samples[@]}</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">do</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; <span class="cm-comment"># path of samples</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; <span class="cm-def">signalFile</span><span class="cm-operator">=</span><span class="cm-string">"</span><span class="cm-def">${sample}</span><span class="cm-string">.fragments.bedgraph"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; <span class="cm-comment"># traversa peaks</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; <span class="cm-keyword">for</span> peak <span class="cm-keyword">in</span> <span class="cm-string">"</span><span class="cm-def">${peaks[@]}</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; <span class="cm-keyword">do</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; <span class="cm-comment"># top peaks</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; <span class="cm-builtin">bash</span> <span class="cm-string">"</span><span class="cm-def">$seacr</span><span class="cm-string">"</span> <span class="cm-string">"</span><span class="cm-def">$signalFile</span><span class="cm-string">"</span> <span class="cm-string">"</span><span class="cm-def">$peak</span><span class="cm-string">"</span> norm stringent <span class="cm-string">"SEACR/</span><span class="cm-def">${sample}</span><span class="cm-string">_seacr_top</span><span class="cm-def">${peak}</span><span class="cm-string">.peaks"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; <span class="cm-comment"># Report with green text</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; <span class="cm-builtin">echo</span> <span class="cm-attribute">-e</span> <span class="cm-string">"\e[32m</span><span class="cm-def">${signalFile}</span><span class="cm-string">_top</span><span class="cm-def">${peak}</span><span class="cm-string"> is done\e[0m"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-keyword">done</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">done</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># KCHIP2 as background to calculate KCHIP2-OE peaks</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-builtin">bash</span> <span class="cm-string">"</span><span class="cm-def">$seacr</span><span class="cm-string">"</span> <span class="cm-string">"</span><span class="cm-def">${samples[1]}</span><span class="cm-string">.fragments.bedgraph"</span> <span class="cm-string">"</span><span class="cm-def">$controlFile</span><span class="cm-string">"</span> norm stringent <span class="cm-string">"SEACR/</span><span class="cm-def">${samples[1]}</span><span class="cm-string">_seacr_</span><span class="cm-def">${samples[0]}</span><span class="cm-string">ascontrol.peaks"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Report final result with green text</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-builtin">echo</span> <span class="cm-attribute">-e</span> <span class="cm-string">"\e[32mBackground peak calculation for </span><span class="cm-def">${samples[1]}</span><span class="cm-string"> using </span><span class="cm-def">${samples[0]}</span><span class="cm-string"> as control is done.\e[0m"</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 918px;"></div><div class="CodeMirror-gutters" style="display: none; height: 918px;"></div></div></div></pre><p>&nbsp;</p><h3 id='seacr网页版'><span>SEACR网页版</span></h3><ul><li><p><span>作者开发了 </span><a href='http://seacr.fredhutch.org'><span>SEACR 网络服务器</span></a><span>用于随时随地地分析CUT&amp;Tag和CUT&amp;RUN 数据。 </span></p></li><li><p><span>该网络服务器接受最大 500 Mb 的bedgraph文件作为输入，用户可以切换所需的参数（归一化或非归一化模式、严格或宽松模式），在运行过程中报告进度，并提供可下载的包含结果的 BED 文件链接。 </span></p><p><img src="./picture/048.png" width="900"/></p></li></ul><p>&nbsp;</p><h3 id='structure-of-seacr-bed-file'><span>Structure of SEACR bed file</span></h3><p><strong><span>Example SEACR BED Entry：</span></strong></p><figure class='table-figure'><table><thead><tr><th><span>Chromosome</span></th><th><span>Start</span></th><th><span>End</span></th><th><span>total signal</span></th><th><span>max signal</span></th><th><span>max signal region</span></th></tr></thead><tbody><tr><td><span>1</span></td><td><span>216897094</span></td><td><span>216897762</span></td><td><span>816</span></td><td><span>2</span></td><td><span>1:216897348-216897496</span></td></tr><tr><td><span>1</span></td><td><span>217018188</span></td><td><span>217018946</span></td><td><span>919</span></td><td><span>2</span></td><td><span>1:217018442-217018890</span></td></tr><tr><td><span>1</span></td><td><span>217197126</span></td><td><span>217198022</span></td><td><span>902</span></td><td><span>2</span></td><td><span>1:217197517-217197523</span></td></tr></tbody></table></figure><p>&nbsp;</p><p><strong><span>SEACR BED File Structure:</span></strong></p><figure class='table-figure'><table><thead><tr><th><span>Column Number</span></th><th><span>Field</span></th><th><span>Description</span></th></tr></thead><tbody><tr><td><span>1</span></td><td><span>Chromosome</span></td><td><span>The chromosome where the peak is located (e.g., </span><code>1</code><span>, </span><code>chrX</code><span>)</span></td></tr><tr><td><span>2</span></td><td><span>Start</span></td><td><span>Start position of the peak (0-based, inclusive)</span></td></tr><tr><td><span>3</span></td><td><span>End</span></td><td><span>End position of the peak (0-based, exclusive)</span></td></tr><tr><td><span>4</span></td><td><span>total signal</span></td><td><span>Total signal contained within denoted coordinates</span></td></tr><tr><td><span>5</span></td><td><span>max signal</span></td><td><span>Maximum bedgraph signal attained at any base pair within denoted coordinates</span></td></tr><tr><td><span>6</span></td><td><span>max signal region</span></td><td><span>region representing the farthest upstream and farthest downstream bases within the denoted coordinates that are represented by the maximum bedgraph signal, formatted as </span><code>chr:start-end</code></td></tr></tbody></table></figure><p>&nbsp;</p><hr /><h2 id='number-of-peaks'><span>Number of peaks</span></h2><p><strong><span>通过读取生成的bed文件，获取不同样本，不同peak类型的数目</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="python" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="python"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">import</span> <span class="cm-variable">pandas</span> <span class="cm-keyword">as</span> <span class="cm-variable">pd</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">import</span> <span class="cm-variable">os</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Define the sample list</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">sampleList</span> <span class="cm-operator">=</span> [<span class="cm-string">"KCHIP2"</span>,<span class="cm-string">"KCHIP2-OE"</span>]</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Define the types of peaks to analyze</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">peakType</span> <span class="cm-operator">=</span> [<span class="cm-string">"top0.01"</span>, <span class="cm-string">"top0.05"</span>, <span class="cm-string">"KCHIP2ascontrol"</span>]</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Initialize empty lists to store peak count (peakN), peak width</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">peakN</span> <span class="cm-operator">=</span> []</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">peakWidth</span> <span class="cm-operator">=</span> []</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Traverse over the sample list</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">for</span> <span class="cm-variable">sample</span> <span class="cm-keyword">in</span> <span class="cm-variable">sampleList</span>:</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># Traverse over the peakType</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-keyword">for</span> <span class="cm-variable">ptype</span> <span class="cm-keyword">in</span> <span class="cm-variable">peakType</span>:</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># Construct the peak file path </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-variable">file_path</span> <span class="cm-operator">=</span> <span class="cm-variable">os</span>.<span class="cm-property">path</span>.<span class="cm-property">join</span>(<span class="cm-string">"SEACR"</span>, <span class="cm-string">f"</span>{<span class="cm-variable">sample</span>}<span class="cm-string">_seacr_</span>{<span class="cm-variable">ptype</span>}<span class="cm-string">.peaks.stringent.bed"</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># Check if the file exists to avoid errors</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-keyword">if</span> <span class="cm-variable">os</span>.<span class="cm-property">path</span>.<span class="cm-property">exists</span>(<span class="cm-variable">file_path</span>):</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># Read the peak file (without headers) </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-variable">peakInfo</span> <span class="cm-operator">=</span> <span class="cm-variable">pd</span>.<span class="cm-property">read_csv</span>(<span class="cm-variable">file_path</span>, <span class="cm-variable">sep</span><span class="cm-operator">=</span><span class="cm-string">"\t"</span>, <span class="cm-variable">header</span><span class="cm-operator">=</span><span class="cm-keyword">None</span>, <span class="cm-variable">low_memory</span><span class="cm-operator">=</span><span class="cm-keyword">False</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># calculate the width of each peak</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># v3(end)-v2(start)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-variable">peakInfo</span>[<span class="cm-string">'width'</span>] <span class="cm-operator">=</span> <span class="cm-builtin">abs</span>(<span class="cm-variable">peakInfo</span>[<span class="cm-number">2</span>] <span class="cm-operator">-</span> <span class="cm-variable">peakInfo</span>[<span class="cm-number">1</span>]) &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># Store the peak count and type information</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-variable">peakN</span>.<span class="cm-property">append</span>({</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-string">"sample"</span>: <span class="cm-variable">sample</span>, &nbsp;<span class="cm-comment"># sample type</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-string">"peakType"</span>: <span class="cm-variable">ptype</span>, &nbsp;<span class="cm-comment"># Peak type (control or top0.01)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-string">"peakN"</span>: <span class="cm-variable">peakInfo</span>.<span class="cm-property">shape</span>[<span class="cm-number">0</span>] &nbsp;<span class="cm-comment"># Number of peaks (rows in the dataframe)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;  })</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># Store the peak width information</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-keyword">for</span> <span class="cm-variable">width</span> <span class="cm-keyword">in</span> <span class="cm-variable">peakInfo</span>[<span class="cm-string">'width'</span>]:</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-variable">peakWidth</span>.<span class="cm-property">append</span>({</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-string">"sample"</span>: <span class="cm-variable">sample</span>, &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-string">"peakType"</span>: <span class="cm-variable">ptype</span>, &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-string">"width"</span>: <span class="cm-variable">width</span> &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;  })</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-keyword">else</span>:</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 输出文件未找到的消息</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-builtin">print</span>(<span class="cm-string">f"File not found: </span>{<span class="cm-variable">file_path</span>}<span class="cm-string">"</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Convert the peakN and peakWidth lists into pandas DataFrames for further analysis or saving</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">peakN_df</span> <span class="cm-operator">=</span> <span class="cm-variable">pd</span>.<span class="cm-property">DataFrame</span>(<span class="cm-variable">peakN</span>)[[<span class="cm-string">'sample'</span>, <span class="cm-string">'peakType'</span>, <span class="cm-string">'peakN'</span>]]</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">peakWidth_df</span> <span class="cm-operator">=</span> <span class="cm-variable">pd</span>.<span class="cm-property">DataFrame</span>(<span class="cm-variable">peakWidth</span>)</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 1180px;"></div><div class="CodeMirror-gutters" style="display: none; height: 1180px;"></div></div></div></pre><p>&nbsp;</p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="python"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="python"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation"><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># check the peak count</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">peakN_df</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 44px;"></div><div class="CodeMirror-gutters" style="display: none; height: 44px;"></div></div></div></pre><figure class='table-figure'><table><thead><tr><th style='text-align:center;' >&nbsp;</th><th style='text-align:center;' ><span>sample</span></th><th style='text-align:center;' ><span>peakType</span></th><th style='text-align:center;' ><span>peakN</span></th></tr></thead><tbody><tr><td style='text-align:center;' ><span>0</span></td><td style='text-align:center;' ><span>KCHIP2</span></td><td style='text-align:center;' ><span>top0.01</span></td><td style='text-align:center;' ><span>9294</span></td></tr><tr><td style='text-align:center;' ><span>1</span></td><td style='text-align:center;' ><span>KCHIP2</span></td><td style='text-align:center;' ><span>top0.05</span></td><td style='text-align:center;' ><span>46527</span></td></tr><tr><td style='text-align:center;' ><span>2</span></td><td style='text-align:center;' ><span>KCHIP2-OE</span></td><td style='text-align:center;' ><span>top0.01</span></td><td style='text-align:center;' ><span>34989</span></td></tr><tr><td style='text-align:center;' ><span>3</span></td><td style='text-align:center;' ><span>KCHIP2-OE</span></td><td style='text-align:center;' ><span>top0.05</span></td><td style='text-align:center;' ><span>175082</span></td></tr><tr><td style='text-align:center;' ><span>4</span></td><td style='text-align:center;' ><span>KCHIP2-OE</span></td><td style='text-align:center;' ><span>KCHIP2ascontrol</span></td><td style='text-align:center;' ><span>7089</span></td></tr></tbody></table></figure><p>&nbsp;</p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="python"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="python"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation"><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># check the peak widrh</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">peakWidth_df</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 44px;"></div><div class="CodeMirror-gutters" style="display: none; height: 44px;"></div></div></div></pre><figure class='table-figure'><table><thead><tr><th style='text-align:right;' >&nbsp;</th><th style='text-align:right;' ><span>sample</span></th><th style='text-align:right;' ><span>peakType</span></th><th style='text-align:right;' ><span>width</span></th></tr></thead><tbody><tr><td style='text-align:right;' ><span>0</span></td><td style='text-align:right;' ><span>KCHIP2</span></td><td style='text-align:right;' ><span>top0.01</span></td><td style='text-align:right;' ><span>597</span></td></tr><tr><td style='text-align:right;' ><span>1</span></td><td style='text-align:right;' ><span>KCHIP2</span></td><td style='text-align:right;' ><span>top0.01</span></td><td style='text-align:right;' ><span>690</span></td></tr><tr><td style='text-align:right;' ><span>2</span></td><td style='text-align:right;' ><span>KCHIP2</span></td><td style='text-align:right;' ><span>top0.01</span></td><td style='text-align:right;' ><span>848</span></td></tr><tr><td style='text-align:right;' ><span>3</span></td><td style='text-align:right;' ><span>KCHIP2</span></td><td style='text-align:right;' ><span>top0.01</span></td><td style='text-align:right;' ><span>797</span></td></tr><tr><td style='text-align:right;' ><span>4</span></td><td style='text-align:right;' ><span>KCHIP2</span></td><td style='text-align:right;' ><span>top0.01</span></td><td style='text-align:right;' ><span>952</span></td></tr><tr><td style='text-align:right;' ><span>...</span></td><td style='text-align:right;' ><span>...</span></td><td style='text-align:right;' ><span>...</span></td><td style='text-align:right;' ><span>...</span></td></tr><tr><td style='text-align:right;' ><span>272976</span></td><td style='text-align:right;' ><span>KCHIP2-OE</span></td><td style='text-align:right;' ><span>KCHIP2ascontrol</span></td><td style='text-align:right;' ><span>1007</span></td></tr><tr><td style='text-align:right;' ><span>272977</span></td><td style='text-align:right;' ><span>KCHIP2-OE</span></td><td style='text-align:right;' ><span>KCHIP2ascontrol</span></td><td style='text-align:right;' ><span>420</span></td></tr><tr><td style='text-align:right;' ><span>272978</span></td><td style='text-align:right;' ><span>KCHIP2-OE</span></td><td style='text-align:right;' ><span>KCHIP2ascontrol</span></td><td style='text-align:right;' ><span>615</span></td></tr><tr><td style='text-align:right;' ><span>272979</span></td><td style='text-align:right;' ><span>KCHIP2-OE</span></td><td style='text-align:right;' ><span>KCHIP2ascontrol</span></td><td style='text-align:right;' ><span>487</span></td></tr><tr><td style='text-align:right;' ><span>272980</span></td><td style='text-align:right;' ><span>KCHIP2-OE</span></td><td style='text-align:right;' ><span>KCHIP2ascontrol</span></td><td style='text-align:right;' ><span>847</span></td></tr></tbody></table></figure><p><span>272981 rows × 3 columns</span></p><p>&nbsp;</p><p><strong><span>Visualization of peak count</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="python" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="python"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">import</span> <span class="cm-variable">pandas</span> <span class="cm-keyword">as</span> <span class="cm-variable">pd</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">import</span> <span class="cm-variable">seaborn</span> <span class="cm-keyword">as</span> <span class="cm-variable">sns</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">import</span> <span class="cm-variable">matplotlib</span>.<span class="cm-property">pyplot</span> <span class="cm-keyword">as</span> <span class="cm-variable">plt</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Create the FacetGrid to separate plots by 'peakType'</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">g</span> <span class="cm-operator">=</span> <span class="cm-variable">sns</span>.<span class="cm-property">FacetGrid</span>(<span class="cm-variable">peakN_df</span>, <span class="cm-variable">col</span><span class="cm-operator">=</span><span class="cm-string">'peakType'</span>, <span class="cm-variable">height</span><span class="cm-operator">=</span><span class="cm-number">5</span>, <span class="cm-variable">aspect</span><span class="cm-operator">=</span><span class="cm-number">1</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Create a barplot for each subplot with sample on x-axis and peakN on y-axis</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">g</span>.<span class="cm-property">map</span>(<span class="cm-variable">sns</span>.<span class="cm-property">barplot</span>, <span class="cm-string">'sample'</span>, <span class="cm-string">'peakN'</span>, <span class="cm-variable">order</span><span class="cm-operator">=</span><span class="cm-variable">peakN_df</span>[<span class="cm-string">'sample'</span>].<span class="cm-property">unique</span>(), <span class="cm-variable">hue</span><span class="cm-operator">=</span><span class="cm-variable">peakN_df</span>[<span class="cm-string">'sample'</span>], <span class="cm-variable">palette</span><span class="cm-operator">=</span><span class="cm-string">"viridis"</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Add labels and title for the entire plot</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">g</span>.<span class="cm-property">set_axis_labels</span>(<span class="cm-string">'Sample'</span>, <span class="cm-string">'Peak Count (peakN)'</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">g</span>.<span class="cm-property">fig</span>.<span class="cm-property">suptitle</span>(<span class="cm-string">'Peak Counts per Sample for Different Peak Types'</span>, <span class="cm-variable">fontsize</span><span class="cm-operator">=</span><span class="cm-number">16</span>, <span class="cm-variable">y</span><span class="cm-operator">=</span><span class="cm-number">1.05</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Show the plot</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">plt</span>.<span class="cm-property">show</span>()</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 415px;"></div><div class="CodeMirror-gutters" style="display: none; height: 415px;"></div></div></div></pre><p>&nbsp;</p><p><strong><span>Visualization of peak width</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="python" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="python"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">import</span> <span class="cm-variable">pandas</span> <span class="cm-keyword">as</span> <span class="cm-variable">pd</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">import</span> <span class="cm-variable">seaborn</span> <span class="cm-keyword">as</span> <span class="cm-variable">sns</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">import</span> <span class="cm-variable">matplotlib</span>.<span class="cm-property">pyplot</span> <span class="cm-keyword">as</span> <span class="cm-variable">plt</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Create the FacetGrid to separate plots by 'peakType'</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">g</span> <span class="cm-operator">=</span> <span class="cm-variable">sns</span>.<span class="cm-property">FacetGrid</span>(<span class="cm-variable">peakWidth_df</span>, <span class="cm-variable">col</span><span class="cm-operator">=</span><span class="cm-string">'peakType'</span>, <span class="cm-variable">height</span><span class="cm-operator">=</span><span class="cm-number">5</span>, <span class="cm-variable">aspect</span><span class="cm-operator">=</span><span class="cm-number">1</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Create a barplot for each subplot with sample on x-axis and peakN on y-axis</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">g</span>.<span class="cm-property">map</span>(<span class="cm-variable">sns</span>.<span class="cm-property">violinplot</span>, <span class="cm-string">'sample'</span>, <span class="cm-string">'width'</span>, <span class="cm-variable">order</span><span class="cm-operator">=</span><span class="cm-variable">peakWidth_df</span>[<span class="cm-string">'sample'</span>].<span class="cm-property">unique</span>(), <span class="cm-variable">hue</span><span class="cm-operator">=</span><span class="cm-variable">peakWidth_df</span>[<span class="cm-string">'sample'</span>], <span class="cm-variable">palette</span><span class="cm-operator">=</span><span class="cm-string">"viridis"</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Add labels and title for the entire plot</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">g</span>.<span class="cm-property">set_axis_labels</span>(<span class="cm-string">'Sample'</span>, <span class="cm-string">'peak width'</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">g</span>.<span class="cm-property">fig</span>.<span class="cm-property">suptitle</span>(<span class="cm-string">'Peak Counts per Sample for Different Peak Types'</span>, <span class="cm-variable">fontsize</span><span class="cm-operator">=</span><span class="cm-number">16</span>, <span class="cm-variable">y</span><span class="cm-operator">=</span><span class="cm-number">1.05</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Show the plot</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">plt</span>.<span class="cm-property">show</span>()</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 350px;"></div><div class="CodeMirror-gutters" style="display: none; height: 350px;"></div></div></div></pre><p>&nbsp;</p><hr /><h2 id='comparing-the-same-peaks-on-replicate-sample-data-setsoptial'><span>Comparing the same peaks on replicate sample data sets(optial)</span></h2><ul><li><p><span>Using tools: </span><a href='#bedtools-intersect'><span>Bedtools intersect</span></a></p></li></ul><p>&nbsp;</p><p><strong><span>Comparing the same peaks on replicate sample data</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-meta">#!/bin/bash</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">###################per setting####################</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Define variables</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-def">samples</span><span class="cm-operator">=</span>(<span class="cm-string">"K27me3"</span> <span class="cm-string">"K4me3"</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-def">repL</span><span class="cm-operator">=</span>(<span class="cm-string">"rep1"</span> <span class="cm-string">"rep2"</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-def">peakType</span><span class="cm-operator">=</span>(<span class="cm-string">"control"</span> <span class="cm-string">"top0.01"</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Define your bed path</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-def">bedpath</span><span class="cm-operator">=</span><span class="cm-string">"SERCA"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Output file</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-def">output_file</span><span class="cm-operator">=</span><span class="cm-string">"output/peak_reproducibility.txt"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">#####################################################</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Initialize an array for peak overlap information</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">declare <span class="cm-attribute">-A</span> peakOverlap</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Clear the output file before writing new data</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">&gt; <span class="cm-string">"</span><span class="cm-def">$output_file</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Create a header for the table and write it to the output file</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-builtin">echo</span> <span class="cm-string">"Sample,PeakType,TotalPeaks,ReproduciblePeaks,ReproducibilityRate"</span> &gt; <span class="cm-string">"</span><span class="cm-def">$output_file</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Loop over peak types and samples</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">for</span> type <span class="cm-keyword">in</span> <span class="cm-string">"</span><span class="cm-def">${peakType[@]}</span><span class="cm-string">"</span>; <span class="cm-keyword">do</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-keyword">for</span> sample <span class="cm-keyword">in</span> <span class="cm-string">"</span><span class="cm-def">${samples[@]}</span><span class="cm-string">"</span>; <span class="cm-keyword">do</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># Initialize an empty file to hold overlapping peaks</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-def">overlap_file</span><span class="cm-operator">=</span><span class="cm-string">"overlap_</span><span class="cm-def">${sample}</span><span class="cm-string">_</span><span class="cm-def">${type}</span><span class="cm-string">.bed"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-builtin">rm</span> <span class="cm-attribute">-f</span> <span class="cm-string">"</span><span class="cm-def">$overlap_file</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># Initialize the first replicate for comparison</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-def">first_rep</span><span class="cm-operator">=</span><span class="cm-string">"</span><span class="cm-def">${bedpath}</span><span class="cm-string">/</span><span class="cm-def">${sample}</span><span class="cm-string">_rep1_seacr_</span><span class="cm-def">${type}</span><span class="cm-string">.peaks.stringent.bed"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># Check if the first replicate file exists</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-keyword">if</span> [[ ! <span class="cm-attribute">-f</span> <span class="cm-string">"</span><span class="cm-def">$first_rep</span><span class="cm-string">"</span> ]]; <span class="cm-keyword">then</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-builtin">echo</span> <span class="cm-string">"First replicate not found: </span><span class="cm-def">$first_rep</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;  continue</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-keyword">fi</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># copy the first_rep file and save it as the overlap file</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-builtin">cp</span> <span class="cm-string">"</span><span class="cm-def">$first_rep</span><span class="cm-string">"</span> <span class="cm-string">"</span><span class="cm-def">$overlap_file</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># Loop over the remaining replicates</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-keyword">for</span> rep <span class="cm-keyword">in</span> <span class="cm-string">"</span><span class="cm-def">${repL[@]:1}</span><span class="cm-string">"</span>; <span class="cm-keyword">do</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-def">peak_file</span><span class="cm-operator">=</span><span class="cm-string">"</span><span class="cm-def">${bedpath}</span><span class="cm-string">/</span><span class="cm-def">${sample}</span><span class="cm-string">_</span><span class="cm-def">${rep}</span><span class="cm-string">_seacr_</span><span class="cm-def">${type}</span><span class="cm-string">.peaks.stringent.bed"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># Check if the replicate file exists</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-keyword">if</span> [[ ! <span class="cm-attribute">-f</span> <span class="cm-string">"</span><span class="cm-def">$peak_file</span><span class="cm-string">"</span> ]]; <span class="cm-keyword">then</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-builtin">echo</span> <span class="cm-string">"Replicate file not found: </span><span class="cm-def">$peak_file</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;  continue</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-keyword">fi</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># Find the overlapping peaks between replicates using bedtools</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;  bedtools intersect <span class="cm-attribute">-a</span> <span class="cm-string">"</span><span class="cm-def">$overlap_file</span><span class="cm-string">"</span> <span class="cm-attribute">-b</span> <span class="cm-string">"</span><span class="cm-def">$peak_file</span><span class="cm-string">"</span> &gt; tmp_overlap.bed</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># Replace overlap_file with the updated overlapping peaks</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-builtin">mv</span> tmp_overlap.bed <span class="cm-string">"</span><span class="cm-def">$overlap_file</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-keyword">done</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># Count the number of overlapping peaks</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-def">overlap_count</span><span class="cm-operator">=</span><span class="cm-quote">$(wc -l &lt; "</span><span class="cm-def">$overlap_file</span><span class="cm-quote">")</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># Calculate the total peaks in the first replicate (to get peakN)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-def">total_peaks</span><span class="cm-operator">=</span><span class="cm-quote">$(wc -l &lt; "</span><span class="cm-def">$first_rep</span><span class="cm-quote">")</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># Calculate peak reproducibility rate</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-def">reproducibility_rate</span><span class="cm-operator">=</span><span class="cm-quote">$(echo "scale=2; (</span><span class="cm-def">$overlap_count</span><span class="cm-quote"> / </span><span class="cm-def">$total_peaks</span><span class="cm-quote">) * 100" | bc)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># Store the result as a row in the CSV file</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-builtin">echo</span> <span class="cm-string">"</span><span class="cm-def">$sample</span><span class="cm-string">,</span><span class="cm-def">$type</span><span class="cm-string">,</span><span class="cm-def">$total_peaks</span><span class="cm-string">,</span><span class="cm-def">$overlap_count</span><span class="cm-string">,</span><span class="cm-def">$reproducibility_rate</span><span class="cm-string">%"</span> &gt;&gt; <span class="cm-string">"</span><span class="cm-def">$output_file</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-keyword">done</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">done</span> </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Print the final reproducibility rates for all samples and peak types</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-builtin">echo</span> <span class="cm-string">"Peak reproducibility results have been saved to </span><span class="cm-def">$output_file</span><span class="cm-string">."</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 1705px;"></div><div class="CodeMirror-gutters" style="display: none; height: 1705px;"></div></div></div></pre><p>&nbsp;</p><p><strong><span>Check the output</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="python"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="python"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">import</span> <span class="cm-variable">pandas</span> <span class="cm-keyword">as</span> <span class="cm-variable">pd</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">from</span> <span class="cm-variable">tabulate</span> <span class="cm-keyword">import</span> <span class="cm-variable">tabulate</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Read the output_file</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">df</span> <span class="cm-operator">=</span> <span class="cm-variable">pd</span>.<span class="cm-property">read_csv</span>(<span class="cm-string">"./peak_reproducibility.txt"</span>, <span class="cm-variable">sep</span><span class="cm-operator">=</span><span class="cm-string">'\t'</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Display the table </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-builtin">print</span>(<span class="cm-variable">df</span>)</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 175px;"></div><div class="CodeMirror-gutters" style="display: none; height: 175px;"></div></div></div></pre><p>&nbsp;</p><hr /><h2 id='calculates-frips-fraction-of-reads-in-peaks'><span>Calculates FRiPs (Fraction of Reads in Peaks)</span></h2><ul><li><p><strong><span>FRiPs</span></strong><span> is a measure of how many sequencing reads fall within peak regions.</span></p></li><li><p><span>A higher </span><strong><span>FRiP score</span></strong><span> indicates a better signal-to-noise ratio in the data, which is often used to assess the quality of ChIP-seq or CUT&amp;Tag experiments.</span></p></li><li><p><span>we can use tool with </span><a href='#bedtools-intersect'><span>Bedtools intersect</span></a><span> to calcuate FRiP</span></p><ul><li><p><span>A BED file containing the peak regions</span></p></li><li><p><span>A BAM file containing your sequencing reads</span></p></li><li><p><span>calcuate the Number of Reads in Peaks :</span></p><p><code>bedtools intersect -a &quot;$bed_file&quot; -b &quot;$bam_file&quot;| awk &#39;{sum += $NF} END {print sum}</code></p></li></ul></li></ul><ul><li><p><mjx-container class="MathJax" jax="SVG" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="27.479ex" height="2.835ex" role="img" focusable="false" viewBox="0 -892.5 12145.8 1253.1" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -0.816ex;"><defs><path id="MJX-43-TEX-N-46" d="M128 619Q121 626 117 628T101 631T58 634H25V680H582V676Q584 670 596 560T610 444V440H570V444Q563 493 561 501Q555 538 543 563T516 601T477 622T431 631T374 633H334H286Q252 633 244 631T233 621Q232 619 232 490V363H284Q287 363 303 363T327 364T349 367T372 373T389 385Q407 403 410 459V480H450V200H410V221Q407 276 389 296Q381 303 371 307T348 313T327 316T303 317T284 317H232V189L233 61Q240 54 245 52T270 48T333 46H360V0H348Q324 3 182 3Q51 3 36 0H25V46H58Q100 47 109 49T128 61V619Z"></path><path id="MJX-43-TEX-N-52" d="M130 622Q123 629 119 631T103 634T60 637H27V683H202H236H300Q376 683 417 677T500 648Q595 600 609 517Q610 512 610 501Q610 468 594 439T556 392T511 361T472 343L456 338Q459 335 467 332Q497 316 516 298T545 254T559 211T568 155T578 94Q588 46 602 31T640 16H645Q660 16 674 32T692 87Q692 98 696 101T712 105T728 103T732 90Q732 59 716 27T672 -16Q656 -22 630 -22Q481 -16 458 90Q456 101 456 163T449 246Q430 304 373 320L363 322L297 323H231V192L232 61Q238 51 249 49T301 46H334V0H323Q302 3 181 3Q59 3 38 0H27V46H60Q102 47 111 49T130 61V622ZM491 499V509Q491 527 490 539T481 570T462 601T424 623T362 636Q360 636 340 636T304 637H283Q238 637 234 628Q231 624 231 492V360H289Q390 360 434 378T489 456Q491 467 491 499Z"></path><path id="MJX-43-TEX-N-69" d="M69 609Q69 637 87 653T131 669Q154 667 171 652T188 609Q188 579 171 564T129 549Q104 549 87 564T69 609ZM247 0Q232 3 143 3Q132 3 106 3T56 1L34 0H26V46H42Q70 46 91 49Q100 53 102 60T104 102V205V293Q104 345 102 359T88 378Q74 385 41 385H30V408Q30 431 32 431L42 432Q52 433 70 434T106 436Q123 437 142 438T171 441T182 442H185V62Q190 52 197 50T232 46H255V0H247Z"></path><path id="MJX-43-TEX-N-50" d="M130 622Q123 629 119 631T103 634T60 637H27V683H214Q237 683 276 683T331 684Q419 684 471 671T567 616Q624 563 624 489Q624 421 573 372T451 307Q429 302 328 301H234V181Q234 62 237 58Q245 47 304 46H337V0H326Q305 3 182 3Q47 3 38 0H27V46H60Q102 47 111 49T130 61V622ZM507 488Q507 514 506 528T500 564T483 597T450 620T397 635Q385 637 307 637H286Q237 637 234 628Q231 624 231 483V342H302H339Q390 342 423 349T481 382Q507 411 507 488Z"></path><path id="MJX-43-TEX-N-3D" d="M56 347Q56 360 70 367H707Q722 359 722 347Q722 336 708 328L390 327H72Q56 332 56 347ZM56 153Q56 168 72 173H708Q722 163 722 153Q722 140 707 133H70Q56 140 56 153Z"></path><path id="MJX-43-TEX-N-4E" d="M42 46Q74 48 94 56T118 69T128 86V634H124Q114 637 52 637H25V683H232L235 680Q237 679 322 554T493 303L578 178V598Q572 608 568 613T544 627T492 637H475V683H483Q498 680 600 680Q706 680 715 683H724V637H707Q634 633 622 598L621 302V6L614 0H600Q585 0 582 3T481 150T282 443T171 605V345L172 86Q183 50 257 46H274V0H265Q250 3 150 3Q48 3 33 0H25V46H42Z"></path><path id="MJX-43-TEX-N-75" d="M383 58Q327 -10 256 -10H249Q124 -10 105 89Q104 96 103 226Q102 335 102 348T96 369Q86 385 36 385H25V408Q25 431 27 431L38 432Q48 433 67 434T105 436Q122 437 142 438T172 441T184 442H187V261Q188 77 190 64Q193 49 204 40Q224 26 264 26Q290 26 311 35T343 58T363 90T375 120T379 144Q379 145 379 161T380 201T380 248V315Q380 361 370 372T320 385H302V431Q304 431 378 436T457 442H464V264Q464 84 465 81Q468 61 479 55T524 46H542V0Q540 0 467 -5T390 -11H383V58Z"></path><path id="MJX-43-TEX-N-6D" d="M41 46H55Q94 46 102 60V68Q102 77 102 91T102 122T103 161T103 203Q103 234 103 269T102 328V351Q99 370 88 376T43 385H25V408Q25 431 27 431L37 432Q47 433 65 434T102 436Q119 437 138 438T167 441T178 442H181V402Q181 364 182 364T187 369T199 384T218 402T247 421T285 437Q305 442 336 442Q351 442 364 440T387 434T406 426T421 417T432 406T441 395T448 384T452 374T455 366L457 361L460 365Q463 369 466 373T475 384T488 397T503 410T523 422T546 432T572 439T603 442Q729 442 740 329Q741 322 741 190V104Q741 66 743 59T754 49Q775 46 803 46H819V0H811L788 1Q764 2 737 2T699 3Q596 3 587 0H579V46H595Q656 46 656 62Q657 64 657 200Q656 335 655 343Q649 371 635 385T611 402T585 404Q540 404 506 370Q479 343 472 315T464 232V168V108Q464 78 465 68T468 55T477 49Q498 46 526 46H542V0H534L510 1Q487 2 460 2T422 3Q319 3 310 0H302V46H318Q379 46 379 62Q380 64 380 200Q379 335 378 343Q372 371 358 385T334 402T308 404Q263 404 229 370Q202 343 195 315T187 232V168V108Q187 78 188 68T191 55T200 49Q221 46 249 46H265V0H257L234 1Q210 2 183 2T145 3Q42 3 33 0H25V46H41Z"></path><path id="MJX-43-TEX-N-62" d="M307 -11Q234 -11 168 55L158 37Q156 34 153 28T147 17T143 10L138 1L118 0H98V298Q98 599 97 603Q94 622 83 628T38 637H20V660Q20 683 22 683L32 684Q42 685 61 686T98 688Q115 689 135 690T165 693T176 694H179V543Q179 391 180 391L183 394Q186 397 192 401T207 411T228 421T254 431T286 439T323 442Q401 442 461 379T522 216Q522 115 458 52T307 -11ZM182 98Q182 97 187 90T196 79T206 67T218 55T233 44T250 35T271 29T295 26Q330 26 363 46T412 113Q424 148 424 212Q424 287 412 323Q385 405 300 405Q270 405 239 390T188 347L182 339V98Z"></path><path id="MJX-43-TEX-N-65" d="M28 218Q28 273 48 318T98 391T163 433T229 448Q282 448 320 430T378 380T406 316T415 245Q415 238 408 231H126V216Q126 68 226 36Q246 30 270 30Q312 30 342 62Q359 79 369 104L379 128Q382 131 395 131H398Q415 131 415 121Q415 117 412 108Q393 53 349 21T250 -11Q155 -11 92 58T28 218ZM333 275Q322 403 238 411H236Q228 411 220 410T195 402T166 381T143 340T127 274V267H333V275Z"></path><path id="MJX-43-TEX-N-72" d="M36 46H50Q89 46 97 60V68Q97 77 97 91T98 122T98 161T98 203Q98 234 98 269T98 328L97 351Q94 370 83 376T38 385H20V408Q20 431 22 431L32 432Q42 433 60 434T96 436Q112 437 131 438T160 441T171 442H174V373Q213 441 271 441H277Q322 441 343 419T364 373Q364 352 351 337T313 322Q288 322 276 338T263 372Q263 381 265 388T270 400T273 405Q271 407 250 401Q234 393 226 386Q179 341 179 207V154Q179 141 179 127T179 101T180 81T180 66V61Q181 59 183 57T188 54T193 51T200 49T207 48T216 47T225 47T235 46T245 46H276V0H267Q249 3 140 3Q37 3 28 0H20V46H36Z"></path><path id="MJX-43-TEX-N-20" d=""></path><path id="MJX-43-TEX-N-6F" d="M28 214Q28 309 93 378T250 448Q340 448 405 380T471 215Q471 120 407 55T250 -10Q153 -10 91 57T28 214ZM250 30Q372 30 372 193V225V250Q372 272 371 288T364 326T348 362T317 390T268 410Q263 411 252 411Q222 411 195 399Q152 377 139 338T126 246V226Q126 130 145 91Q177 30 250 30Z"></path><path id="MJX-43-TEX-N-66" d="M273 0Q255 3 146 3Q43 3 34 0H26V46H42Q70 46 91 49Q99 52 103 60Q104 62 104 224V385H33V431H104V497L105 564L107 574Q126 639 171 668T266 704Q267 704 275 704T289 705Q330 702 351 679T372 627Q372 604 358 590T321 576T284 590T270 627Q270 647 288 667H284Q280 668 273 668Q245 668 223 647T189 592Q183 572 182 497V431H293V385H185V225Q185 63 186 61T189 57T194 54T199 51T206 49T213 48T222 47T231 47T241 46T251 46H282V0H273Z"></path><path id="MJX-43-TEX-N-61" d="M137 305T115 305T78 320T63 359Q63 394 97 421T218 448Q291 448 336 416T396 340Q401 326 401 309T402 194V124Q402 76 407 58T428 40Q443 40 448 56T453 109V145H493V106Q492 66 490 59Q481 29 455 12T400 -6T353 12T329 54V58L327 55Q325 52 322 49T314 40T302 29T287 17T269 6T247 -2T221 -8T190 -11Q130 -11 82 20T34 107Q34 128 41 147T68 188T116 225T194 253T304 268H318V290Q318 324 312 340Q290 411 215 411Q197 411 181 410T156 406T148 403Q170 388 170 359Q170 334 154 320ZM126 106Q126 75 150 51T209 26Q247 26 276 49T315 109Q317 116 318 175Q318 233 317 233Q309 233 296 232T251 223T193 203T147 166T126 106Z"></path><path id="MJX-43-TEX-N-64" d="M376 495Q376 511 376 535T377 568Q377 613 367 624T316 637H298V660Q298 683 300 683L310 684Q320 685 339 686T376 688Q393 689 413 690T443 693T454 694H457V390Q457 84 458 81Q461 61 472 55T517 46H535V0Q533 0 459 -5T380 -11H373V44L365 37Q307 -11 235 -11Q158 -11 96 50T34 215Q34 315 97 378T244 442Q319 442 376 393V495ZM373 342Q328 405 260 405Q211 405 173 369Q146 341 139 305T131 211Q131 155 138 120T173 59Q203 26 251 26Q322 26 373 103V342Z"></path><path id="MJX-43-TEX-N-73" d="M295 316Q295 356 268 385T190 414Q154 414 128 401Q98 382 98 349Q97 344 98 336T114 312T157 287Q175 282 201 278T245 269T277 256Q294 248 310 236T342 195T359 133Q359 71 321 31T198 -10H190Q138 -10 94 26L86 19L77 10Q71 4 65 -1L54 -11H46H42Q39 -11 33 -5V74V132Q33 153 35 157T45 162H54Q66 162 70 158T75 146T82 119T101 77Q136 26 198 26Q295 26 295 104Q295 133 277 151Q257 175 194 187T111 210Q75 227 54 256T33 318Q33 357 50 384T93 424T143 442T187 447H198Q238 447 268 432L283 424L292 431Q302 440 314 448H322H326Q329 448 335 442V310L329 304H301Q295 310 295 316Z"></path><path id="MJX-43-TEX-N-6E" d="M41 46H55Q94 46 102 60V68Q102 77 102 91T102 122T103 161T103 203Q103 234 103 269T102 328V351Q99 370 88 376T43 385H25V408Q25 431 27 431L37 432Q47 433 65 434T102 436Q119 437 138 438T167 441T178 442H181V402Q181 364 182 364T187 369T199 384T218 402T247 421T285 437Q305 442 336 442Q450 438 463 329Q464 322 464 190V104Q464 66 466 59T477 49Q498 46 526 46H542V0H534L510 1Q487 2 460 2T422 3Q319 3 310 0H302V46H318Q379 46 379 62Q380 64 380 200Q379 335 378 343Q372 371 358 385T334 402T308 404Q263 404 229 370Q202 343 195 315T187 232V168V108Q187 78 188 68T191 55T200 49Q221 46 249 46H265V0H257L234 1Q210 2 183 2T145 3Q42 3 33 0H25V46H41Z"></path><path id="MJX-43-TEX-N-6B" d="M36 46H50Q89 46 97 60V68Q97 77 97 91T97 124T98 167T98 217T98 272T98 329Q98 366 98 407T98 482T98 542T97 586T97 603Q94 622 83 628T38 637H20V660Q20 683 22 683L32 684Q42 685 61 686T98 688Q115 689 135 690T165 693T176 694H179V463L180 233L240 287Q300 341 304 347Q310 356 310 364Q310 383 289 385H284V431H293Q308 428 412 428Q475 428 484 431H489V385H476Q407 380 360 341Q286 278 286 274Q286 273 349 181T420 79Q434 60 451 53T500 46H511V0H505Q496 3 418 3Q322 3 307 0H299V46H306Q330 48 330 65Q330 72 326 79Q323 84 276 153T228 222L176 176V120V84Q176 65 178 59T189 49Q210 46 238 46H254V0H246Q231 3 137 3T28 0H20V46H36Z"></path><path id="MJX-43-TEX-N-54" d="M36 443Q37 448 46 558T55 671V677H666V671Q667 666 676 556T685 443V437H645V443Q645 445 642 478T631 544T610 593Q593 614 555 625Q534 630 478 630H451H443Q417 630 414 618Q413 616 413 339V63Q420 53 439 50T528 46H558V0H545L361 3Q186 1 177 0H164V46H194Q264 46 283 49T309 63V339V550Q309 620 304 625T271 630H244H224Q154 630 119 601Q101 585 93 554T81 486T76 443V437H36V443Z"></path><path id="MJX-43-TEX-N-74" d="M27 422Q80 426 109 478T141 600V615H181V431H316V385H181V241Q182 116 182 100T189 68Q203 29 238 29Q282 29 292 100Q293 108 293 146V181H333V146V134Q333 57 291 17Q264 -10 221 -10Q187 -10 162 2T124 33T105 68T98 100Q97 107 97 248V385H18V422H27Z"></path><path id="MJX-43-TEX-N-6C" d="M42 46H56Q95 46 103 60V68Q103 77 103 91T103 124T104 167T104 217T104 272T104 329Q104 366 104 407T104 482T104 542T103 586T103 603Q100 622 89 628T44 637H26V660Q26 683 28 683L38 684Q48 685 67 686T104 688Q121 689 141 690T171 693T182 694H185V379Q185 62 186 60Q190 52 198 49Q219 46 247 46H263V0H255L232 1Q209 2 183 2T145 3T107 3T57 1L34 0H26V46H42Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="mtext"><use data-c="46" xlink:href="#MJX-43-TEX-N-46"></use><use data-c="52" xlink:href="#MJX-43-TEX-N-52" transform="translate(653,0)"></use><use data-c="69" xlink:href="#MJX-43-TEX-N-69" transform="translate(1389,0)"></use><use data-c="50" xlink:href="#MJX-43-TEX-N-50" transform="translate(1667,0)"></use></g><g data-mml-node="mo" transform="translate(2625.8,0)"><use data-c="3D" xlink:href="#MJX-43-TEX-N-3D"></use></g><g data-mml-node="mfrac" transform="translate(3681.6,0)"><g data-mml-node="mtext" transform="translate(220,394) scale(0.707)"><use data-c="4E" xlink:href="#MJX-43-TEX-N-4E"></use><use data-c="75" xlink:href="#MJX-43-TEX-N-75" transform="translate(750,0)"></use><use data-c="6D" xlink:href="#MJX-43-TEX-N-6D" transform="translate(1306,0)"></use><use data-c="62" xlink:href="#MJX-43-TEX-N-62" transform="translate(2139,0)"></use><use data-c="65" xlink:href="#MJX-43-TEX-N-65" transform="translate(2695,0)"></use><use data-c="72" xlink:href="#MJX-43-TEX-N-72" transform="translate(3139,0)"></use><use data-c="20" xlink:href="#MJX-43-TEX-N-20" transform="translate(3531,0)"></use><use data-c="6F" xlink:href="#MJX-43-TEX-N-6F" transform="translate(3781,0)"></use><use data-c="66" xlink:href="#MJX-43-TEX-N-66" transform="translate(4281,0)"></use><use data-c="20" xlink:href="#MJX-43-TEX-N-20" transform="translate(4587,0)"></use><use data-c="52" xlink:href="#MJX-43-TEX-N-52" transform="translate(4837,0)"></use><use data-c="65" xlink:href="#MJX-43-TEX-N-65" transform="translate(5573,0)"></use><use data-c="61" xlink:href="#MJX-43-TEX-N-61" transform="translate(6017,0)"></use><use data-c="64" xlink:href="#MJX-43-TEX-N-64" transform="translate(6517,0)"></use><use data-c="73" xlink:href="#MJX-43-TEX-N-73" transform="translate(7073,0)"></use><use data-c="20" xlink:href="#MJX-43-TEX-N-20" transform="translate(7467,0)"></use><use data-c="69" xlink:href="#MJX-43-TEX-N-69" transform="translate(7717,0)"></use><use data-c="6E" xlink:href="#MJX-43-TEX-N-6E" transform="translate(7995,0)"></use><use data-c="20" xlink:href="#MJX-43-TEX-N-20" transform="translate(8551,0)"></use><use data-c="50" xlink:href="#MJX-43-TEX-N-50" transform="translate(8801,0)"></use><use data-c="65" xlink:href="#MJX-43-TEX-N-65" transform="translate(9482,0)"></use><use data-c="61" xlink:href="#MJX-43-TEX-N-61" transform="translate(9926,0)"></use><use data-c="6B" xlink:href="#MJX-43-TEX-N-6B" transform="translate(10426,0)"></use><use data-c="73" xlink:href="#MJX-43-TEX-N-73" transform="translate(10954,0)"></use></g><g data-mml-node="mtext" transform="translate(659.1,-345) scale(0.707)"><use data-c="54" xlink:href="#MJX-43-TEX-N-54"></use><use data-c="6F" xlink:href="#MJX-43-TEX-N-6F" transform="translate(722,0)"></use><use data-c="74" xlink:href="#MJX-43-TEX-N-74" transform="translate(1222,0)"></use><use data-c="61" xlink:href="#MJX-43-TEX-N-61" transform="translate(1611,0)"></use><use data-c="6C" xlink:href="#MJX-43-TEX-N-6C" transform="translate(2111,0)"></use><use data-c="20" xlink:href="#MJX-43-TEX-N-20" transform="translate(2389,0)"></use><use data-c="4E" xlink:href="#MJX-43-TEX-N-4E" transform="translate(2639,0)"></use><use data-c="75" xlink:href="#MJX-43-TEX-N-75" transform="translate(3389,0)"></use><use data-c="6D" xlink:href="#MJX-43-TEX-N-6D" transform="translate(3945,0)"></use><use data-c="62" xlink:href="#MJX-43-TEX-N-62" transform="translate(4778,0)"></use><use data-c="65" xlink:href="#MJX-43-TEX-N-65" transform="translate(5334,0)"></use><use data-c="72" xlink:href="#MJX-43-TEX-N-72" transform="translate(5778,0)"></use><use data-c="20" xlink:href="#MJX-43-TEX-N-20" transform="translate(6170,0)"></use><use data-c="6F" xlink:href="#MJX-43-TEX-N-6F" transform="translate(6420,0)"></use><use data-c="66" xlink:href="#MJX-43-TEX-N-66" transform="translate(6920,0)"></use><use data-c="20" xlink:href="#MJX-43-TEX-N-20" transform="translate(7226,0)"></use><use data-c="52" xlink:href="#MJX-43-TEX-N-52" transform="translate(7476,0)"></use><use data-c="65" xlink:href="#MJX-43-TEX-N-65" transform="translate(8212,0)"></use><use data-c="61" xlink:href="#MJX-43-TEX-N-61" transform="translate(8656,0)"></use><use data-c="64" xlink:href="#MJX-43-TEX-N-64" transform="translate(9156,0)"></use><use data-c="73" xlink:href="#MJX-43-TEX-N-73" transform="translate(9712,0)"></use></g><rect width="8224.2" height="60" x="120" y="220"></rect></g></g></g></svg><mjx-assistive-mml unselectable="on" display="inline"><math xmlns="http://www.w3.org/1998/Math/MathML"><mtext>FRiP</mtext><mo>=</mo><mfrac><mtext>Number of Reads in Peaks</mtext><mtext>Total Number of Reads</mtext></mfrac></math></mjx-assistive-mml></mjx-container><script type="math/tex">\text{FRiP} = \frac{\text{Number of Reads in Peaks}}{\text{Total Number of Reads}}</script></p></li></ul><ul><li><p><strong><span>Ensure your bed file chromosome order consistent with your bam file !!!</span></strong></p></li></ul><p>&nbsp;</p><h3 id='sorting-file-procedure'><span>Sorting file procedure</span></h3><p><strong><span>1 Check your bam file order</span></strong></p><ul><li><p><span>check your bam file order : </span><code>samtools view -H &lt;bam&gt;</code></p><p><img src="./picture/055.png" width="200"/></p></li><li><p><span>This BAM file&#39;s sorting mode is currently set to </span><strong><span>coordinate (SO:coordinate in the header)</span></strong></p></li><li><p><span>However, the order of chromosomes (shown in the @SQ lines) follows the </span><strong><span>lexicographical order</span></strong><span>, meaning that chromosomes 10, 11, 12, etc., come before chromosome 1, 2, 3, etc. This is because </span><strong><span>lexicographical sorting</span></strong><span> treats numbers as strings, where &quot;10&quot; comes before &quot;2&quot;</span></p></li></ul><p>&nbsp;</p><p><strong><span>2.Check your bed file order</span></strong></p><ul><li><p><span>check your bed file order :</span><code>cut -f1 &lt;bed&gt; | uniq</code></p><p><img src="./picture/056.png" width="50"/></p></li></ul><ul><li><p><span>you found your bed file chromosome order not consistent with your bam file</span></p></li><li><p><span>you can fix this problem with sorrting your bed file with </span><strong><span>custom order(bam file order)</span></strong></p></li></ul><p>&nbsp;</p><p><strong><span>3 Create a Custom Order File</span></strong></p><p><img src="./picture/055.png" width="180"/></p><ul><li><p><span>First, create a custom chromosome order file that matches the order you specified. Let&#39;s call this file chromosome_order.txt. </span></p></li><li><p><span>The code will be:</span></p><p><code>samtools view -H &lt;bam&gt; | grep &#39;^@SQ&#39; | awk &#39;{print $2}&#39; | cut -d&#39;:&#39; -f2 &gt; chromosome_order.txt</code></p><ul><li><p><code>samtools view -H &lt;bam&gt;</code><span> :see head information of bam file</span></p></li><li><p><code>grep &#39;^@SQ&#39;</code><span> :Filter out all lines starting with @ SQ</span></p></li><li><p><code>awk &#39;{print $2}&#39;</code><span> :awk is used to print the second field of each line(SN:</span><chromosome><span>)</span></p></li><li><p><code>cut -d&#39;:&#39; -f2</code><span> : cut is used to further process the output by splitting each line on the colon (:) and extracting the second part</span></p></li></ul></li></ul><p>&nbsp;</p><p><strong><span>4.sort your bed file with custom order</span></strong></p><ul><li><p><span>sort your bed file with custom order</span></p><p><code>awk &#39;NR==FNR {order[$1]=NR; next} {print order[$1], $0}&#39; chromosome_order.txt &lt;unsorted.bed&gt; | sort -k1,1n -k3,3n | cut -d&#39; &#39; -f2- &gt; &lt;sorted.bed&gt;</code></p><p>&nbsp;</p><ul><li><p><code>NR==FNR</code><span> :</span></p><ul><li><p><code>NR</code><span> : is the total number of input records read</span></p></li><li><p><code>FNR</code><span> : is the number of records read from the </span><strong><span>current file</span></strong><span>.</span></p></li><li><p><span>This means that as long as we are reading the first file, </span><code>order[$1]=NR</code><span> is executed</span></p><p>&nbsp;</p></li></ul></li><li><p><code>order[$1]=NR</code><span> : This creates an associative array called </span><strong><span>order</span></strong><span> where the key is the chromosome name (from the first column of chromosome_order.txt)</span></p><ul><li><p><span>It creates an order mapping:</span></p></li><li><p><code>chr1</code><span> → 1</span></p></li><li><p><code>chr2</code><span> → 2</span></p></li><li><p><code>chrX</code><span> → 3</span></p></li><li><p><code>chrY</code><span> → 4</span></p></li></ul></li></ul><p><span> </span></p><ul><li><p><code>next</code><span>: This skips the remaining actions for the current line to the next line. </span></p></li><li><p><code>print order[$1], $0}</code><span> : For lines in the second file (unsorted.bed),  this prints the chromosome rank (from the </span><strong><span>order</span></strong><span> array) followed by the entire line (</span><code>$0</code><span> represents the whole line)</span></p></li></ul></li></ul><figure class='table-figure'><table><thead><tr><th><span>order</span></th><th><span>Chromosome</span></th><th><span>Start</span></th><th><span>End</span></th><th><span>Peak</span></th></tr></thead><tbody><tr><td><span>3</span></td><td><span>chrX</span></td><td><span>100</span></td><td><span>200</span></td><td><span>peak1</span></td></tr><tr><td><span>1</span></td><td><span>chr1</span></td><td><span>150</span></td><td><span>250</span></td><td><span>peak2</span></td></tr><tr><td><span>2</span></td><td><span>chr2</span></td><td><span>50</span></td><td><span>150</span></td><td><span>peak3</span></td></tr></tbody></table></figure><ul><li><p><code>sort -k1,1n -k3,3n</code><span> :sort by </span><strong><span>order</span></strong><span> array ,subsort by the start position of the peak</span></p></li><li><p><code>cut -d&#39; &#39; -f2-</code><span> :</span></p><ul><li><p><code>-d&#39; &#39;</code><span> : Sets the delimiter as a space.</span></p></li><li><p><code>-f2-</code><span>  : This specifies that we want to extract fields from the second one onward. Essentially, it removes the first field (</span><strong><span>order</span></strong><span>) from the output.</span></p></li></ul></li></ul><p>&nbsp;</p><h3 id='code-with-sorting-file-and-calcuating-frips'><span>Code with Sorting file and calcuating FRIPs </span></h3><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-meta">#!/bin/bash</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">###################per setting####################</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Define the variables</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-def">sampleMarks</span><span class="cm-operator">=</span>(<span class="cm-string">"KCHIP2"</span> <span class="cm-string">"KCHIP2-OE"</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-def">peakType</span><span class="cm-operator">=</span>(<span class="cm-string">"top0.01"</span> <span class="cm-string">"top0.05"</span> <span class="cm-string">"KCHIP2ascontrol"</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Define the threads</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-def">threads</span><span class="cm-operator">=</span><span class="cm-number">10</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Define your peak file directory</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-def">bedDir</span><span class="cm-operator">=</span><span class="cm-string">"SERCA"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># BAM directory</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-def">bamDir</span><span class="cm-operator">=</span><span class="cm-string">"deduplicated.bam"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Output file</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-def">output_file</span><span class="cm-operator">=</span><span class="cm-string">"output/FRiPs.txt"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">#####################################################</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Initialize an empty list to store the number of reads falling within peaks</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">declare <span class="cm-attribute">-A</span> FRIPsData</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Clear the output file before writing new data</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">&gt; <span class="cm-string">"</span><span class="cm-def">$output_file</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Define the function to get fragment counts in peak regions</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">get_fragment_counts() {</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;  local <span class="cm-def">bam_file</span><span class="cm-operator">=</span><span class="cm-def">$1</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;  local <span class="cm-def">peaks_bed</span><span class="cm-operator">=</span><span class="cm-def">$2</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># Use bedtools intersect to find reads that overlap with peaks &nbsp; &nbsp; &nbsp;</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-def">overlap_counts</span><span class="cm-operator">=</span><span class="cm-quote">$(bedtools intersect -sorted -c -a "</span><span class="cm-def">$peaks_bed</span><span class="cm-quote">" -b "</span><span class="cm-def">$bam_file</span><span class="cm-quote">" | awk '{sum += </span><span class="cm-def">$NF</span><span class="cm-quote">} END {print sum}')</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># Return the total count of reads overlapping peaks</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-builtin">echo</span> <span class="cm-string">"</span><span class="cm-def">$overlap_counts</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">}</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># create a tem file to store temporary data</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-builtin">mkdir</span> <span class="cm-attribute">-p</span> tmp</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Loop over the sampleMarks and peakType</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">for</span> sample <span class="cm-keyword">in</span> <span class="cm-string">"</span><span class="cm-def">${sampleMarks[@]}</span><span class="cm-string">"</span>; <span class="cm-keyword">do</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-keyword">for</span> peak <span class="cm-keyword">in</span> <span class="cm-string">"</span><span class="cm-def">${peakType[@]}</span><span class="cm-string">"</span>; <span class="cm-keyword">do</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># Construct the path of the peak information file (BED format)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-def">peak_file</span><span class="cm-operator">=</span><span class="cm-string">"</span><span class="cm-def">${bedDir}</span><span class="cm-string">/</span><span class="cm-def">${sample}</span><span class="cm-string">_seacr_</span><span class="cm-def">${peak}</span><span class="cm-string">.peaks.stringent.bed"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-def">sort_peak_file</span><span class="cm-operator">=</span><span class="cm-string">"tmp/</span><span class="cm-def">${sample}</span><span class="cm-string">_seacr_</span><span class="cm-def">${peak}</span><span class="cm-string">.peaks.stringent.sort.bed"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># Check if the peak file exists</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-keyword">if</span> [[ <span class="cm-attribute">-f</span> <span class="cm-string">"</span><span class="cm-def">$peak_file</span><span class="cm-string">"</span> ]]; <span class="cm-keyword">then</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># BAM file path</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-def">bam_file</span><span class="cm-operator">=</span><span class="cm-string">"</span><span class="cm-def">${bamDir}</span><span class="cm-string">/filter</span><span class="cm-def">${sample}</span><span class="cm-string">deduplicated.bam"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-def">sort_bam_file</span><span class="cm-operator">=</span><span class="cm-string">"tmp/filter</span><span class="cm-def">${sample}</span><span class="cm-string">deduplicated.sort.bam"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># Check if the BAM file exists</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-keyword">if</span> [[ <span class="cm-attribute">-f</span> <span class="cm-string">"</span><span class="cm-def">$bam_file</span><span class="cm-string">"</span> ]]; <span class="cm-keyword">then</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># Sort the BAM file (skip sorting if sorted already)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-keyword">if</span> [[ ! <span class="cm-attribute">-f</span> <span class="cm-string">"</span><span class="cm-def">$sort_bam_file</span><span class="cm-string">"</span> ]]; <span class="cm-keyword">then</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-builtin">echo</span> <span class="cm-string">"Sorting BAM file: </span><span class="cm-def">$bam_file</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;  samtools <span class="cm-builtin">sort</span> <span class="cm-attribute">-o</span> <span class="cm-string">"</span><span class="cm-def">$sort_bam_file</span><span class="cm-string">"</span> <span class="cm-string">"</span><span class="cm-def">$bam_file</span><span class="cm-string">"</span> <span class="cm-attribute">-</span>@ <span class="cm-string">"</span><span class="cm-def">$threads</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-keyword">fi</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># Create a Custom Order File(skip create if Custom Order already)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-def">Custom_Order</span><span class="cm-operator">=</span><span class="cm-string">"tmp/</span><span class="cm-def">${sample}</span><span class="cm-string">.chromosome_order.txt"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-keyword">if</span> [[ ! <span class="cm-attribute">-f</span> <span class="cm-string">"</span><span class="cm-def">$Custom_Order</span><span class="cm-string">"</span> ]]; <span class="cm-keyword">then</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-builtin">echo</span> <span class="cm-string">"Create a Custom Order File: </span><span class="cm-def">$Custom_Order</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;  samtools view <span class="cm-attribute">-H</span> <span class="cm-string">"</span><span class="cm-def">$sort_bam_file</span><span class="cm-string">"</span> | <span class="cm-builtin">grep</span> <span class="cm-string">'^@SQ'</span> | <span class="cm-builtin">awk</span> <span class="cm-string">'{print $2}'</span> | <span class="cm-builtin">cut</span> <span class="cm-attribute">-d</span><span class="cm-string">':'</span> <span class="cm-attribute">-f2</span> &gt; <span class="cm-string">"</span><span class="cm-def">$Custom_Order</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-keyword">fi</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># Sort the bed file (sort by Custom Order)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-builtin">echo</span> <span class="cm-string">"Sorting peak file: </span><span class="cm-def">$peak_file</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-builtin">awk</span> <span class="cm-string">'NR==FNR {order[$1]=NR; next} {print order[$1], $0}'</span> <span class="cm-string">"</span><span class="cm-def">$Custom_Order</span><span class="cm-string">"</span> <span class="cm-string">"</span><span class="cm-def">$peak_file</span><span class="cm-string">"</span> | <span class="cm-builtin">sort</span> <span class="cm-attribute">-k1</span>,1n <span class="cm-attribute">-k3</span>,3n | <span class="cm-builtin">cut</span> <span class="cm-attribute">-d</span><span class="cm-string">' '</span> <span class="cm-attribute">-f2-</span> &gt; <span class="cm-string">"</span><span class="cm-def">$sort_peak_file</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># Get the Number of Reads in Peaks</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-builtin">echo</span> <span class="cm-string">"Get the Number of Reads in Peaks: </span><span class="cm-def">$sort_peak_file</span><span class="cm-string"> and </span><span class="cm-def">$sort_bam_file</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-def">reads_in_peaks</span><span class="cm-operator">=</span><span class="cm-quote">$(get_fragment_counts "</span><span class="cm-def">$sort_bam_file</span><span class="cm-quote">" "</span><span class="cm-def">$sort_peak_file</span><span class="cm-quote">")</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># Get the Total Number of Reads</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-def">total_reads</span><span class="cm-operator">=</span><span class="cm-quote">$(samtools view -c "</span><span class="cm-def">$bam_file</span><span class="cm-quote">")</span> </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># calcuate the FRIP(Fraction of Reads in Peaks)(4 decimal places)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-def">frip</span><span class="cm-operator">=</span><span class="cm-quote">$(echo "scale=4; </span><span class="cm-def">$reads_in_peaks</span><span class="cm-quote"> / </span><span class="cm-def">$total_reads</span><span class="cm-quote">" | bc)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># Append the results to the inPeakData array</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;  FRIPsData[<span class="cm-string">"</span><span class="cm-def">$sample</span><span class="cm-string">,</span><span class="cm-def">$peak</span><span class="cm-string">"</span>]<span class="cm-operator">=</span><span class="cm-string">"Reads in Peaks: </span><span class="cm-def">$reads_in_peaks</span><span class="cm-string">, Total Reads: </span><span class="cm-def">$total_reads</span><span class="cm-string">, FRiP: </span><span class="cm-def">$frip</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-keyword">else</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-builtin">echo</span> <span class="cm-attribute">-e</span> <span class="cm-string">"\e[31mBAM file not found: </span><span class="cm-def">$peak_file</span><span class="cm-string">\e[0m"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-keyword">fi</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-keyword">else</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-builtin">echo</span> <span class="cm-attribute">-e</span> <span class="cm-string">"\e[31mPeak file not found: </span><span class="cm-def">$peak_file</span><span class="cm-string">\e[0m"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-keyword">fi</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-keyword">done</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">done</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Output the header of the table</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-builtin">echo</span> <span class="cm-attribute">-e</span> <span class="cm-string">"Sample\tPeak Type\tReads in Peaks\tTotal Reads\tFRiP"</span> &gt;&gt; <span class="cm-string">"</span><span class="cm-def">$output_file</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Output the results to the text file</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">for</span> key <span class="cm-keyword">in</span> <span class="cm-string">"</span><span class="cm-def">${!FRIPsData[@]}</span><span class="cm-string">"</span>; <span class="cm-keyword">do</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-def">sample</span><span class="cm-operator">=</span><span class="cm-def">${key%,*}</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-def">peak_type</span><span class="cm-operator">=</span><span class="cm-def">${key#*,}</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-def">reads_in_peaks</span><span class="cm-operator">=</span><span class="cm-def">${FRIPsData[$key]#*Reads in Peaks: }</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-def">total_reads</span><span class="cm-operator">=</span><span class="cm-def">${reads_in_peaks#*, Total Reads: }</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-def">frip</span><span class="cm-operator">=</span><span class="cm-def">${total_reads#*, FRiP: }</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># Clean up values</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-def">reads_in_peaks</span><span class="cm-operator">=</span><span class="cm-def">${reads_in_peaks%%,*}</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-def">total_reads</span><span class="cm-operator">=</span><span class="cm-def">${total_reads%%,*}</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-def">frip</span><span class="cm-operator">=</span><span class="cm-def">${frip%%,*}</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># Format the output as a table</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;  printf <span class="cm-string">"%s\t%s\t%s\t%s\t%s\n"</span> <span class="cm-string">"</span><span class="cm-def">$sample</span><span class="cm-string">"</span> <span class="cm-string">"</span><span class="cm-def">$peak_type</span><span class="cm-string">"</span> <span class="cm-string">"</span><span class="cm-def">$reads_in_peaks</span><span class="cm-string">"</span> <span class="cm-string">"</span><span class="cm-def">$total_reads</span><span class="cm-string">"</span> <span class="cm-string">"</span><span class="cm-def">$frip</span><span class="cm-string">"</span> &gt;&gt; <span class="cm-string">"</span><span class="cm-def">$output_file</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">done</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># remove tmp file</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-builtin">rm</span> <span class="cm-attribute">-r</span> tmp</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Confirmation message</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-builtin">echo</span> <span class="cm-string">"Results have been saved to </span><span class="cm-def">$output_file</span><span class="cm-string">"</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 2819px;"></div><div class="CodeMirror-gutters" style="display: none; height: 2819px;"></div></div></div></pre><p><span> </span></p><p><strong><span>Check the output</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="python"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="python"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">import</span> <span class="cm-variable">pandas</span> <span class="cm-keyword">as</span> <span class="cm-variable">pd</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">from</span> <span class="cm-variable">tabulate</span> <span class="cm-keyword">import</span> <span class="cm-variable">tabulate</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Read the output_file</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">df</span> <span class="cm-operator">=</span> <span class="cm-variable">pd</span>.<span class="cm-property">read_csv</span>(<span class="cm-string">"./FRiPs.txt"</span>, <span class="cm-variable">sep</span><span class="cm-operator">=</span><span class="cm-string">'\t'</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Display the table </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-builtin">print</span>(<span class="cm-variable">df</span>)</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 175px;"></div><div class="CodeMirror-gutters" style="display: none; height: 175px;"></div></div></div></pre><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="text"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="text"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;  Sample &nbsp; &nbsp; &nbsp;  Peak Type  Reads in Peaks  Total Reads &nbsp;  FRiP</span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">0 &nbsp; &nbsp; KCHIP2 &nbsp; &nbsp; &nbsp; &nbsp;  top0.05 &nbsp; &nbsp; &nbsp; &nbsp; 1388331 &nbsp; &nbsp;  4463744  0.3110</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">1 &nbsp; &nbsp; KCHIP2 &nbsp; &nbsp; &nbsp; &nbsp;  top0.01 &nbsp; &nbsp; &nbsp; &nbsp; 1115334 &nbsp; &nbsp;  4463744  0.2498</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">2  KCHIP2-OE  KCHIP2ascontrol &nbsp; &nbsp; &nbsp; &nbsp;  418309 &nbsp; &nbsp; 17177518  0.0243</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">3  KCHIP2-OE &nbsp; &nbsp; &nbsp; &nbsp;  top0.01 &nbsp; &nbsp; &nbsp; &nbsp; 1440794 &nbsp; &nbsp; 17177518  0.0838</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">4  KCHIP2-OE &nbsp; &nbsp; &nbsp; &nbsp;  top0.05 &nbsp; &nbsp; &nbsp; &nbsp; 3051352 &nbsp; &nbsp; 17177518  0.1776</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 131px;"></div><div class="CodeMirror-gutters" style="display: none; height: 131px;"></div></div></div></pre><p>&nbsp;</p><p><strong><span>Draw picture</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="python" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="python"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">import</span> <span class="cm-variable">pandas</span> <span class="cm-keyword">as</span> <span class="cm-variable">pd</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">import</span> <span class="cm-variable">matplotlib</span>.<span class="cm-property">pyplot</span> <span class="cm-keyword">as</span> <span class="cm-variable">plt</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">import</span> <span class="cm-variable">seaborn</span> <span class="cm-keyword">as</span> <span class="cm-variable">sns</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Create a box plot with seaborn</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">plt</span>.<span class="cm-property">figure</span>(<span class="cm-variable">figsize</span><span class="cm-operator">=</span>(<span class="cm-number">6</span>, <span class="cm-number">4</span>))</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Box plot of FRiP scores grouped by Sample </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">sns</span>.<span class="cm-property">boxplot</span>(<span class="cm-variable">x</span><span class="cm-operator">=</span><span class="cm-string">'Sample'</span>, <span class="cm-variable">y</span><span class="cm-operator">=</span><span class="cm-string">'FRiP'</span>, <span class="cm-variable">data</span><span class="cm-operator">=</span><span class="cm-variable">df</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Add titles and labels</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">plt</span>.<span class="cm-property">title</span>(<span class="cm-string">'Box Plot of FRiP Scores Grouped by Sample'</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">plt</span>.<span class="cm-property">xlabel</span>(<span class="cm-string">'Sample'</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">plt</span>.<span class="cm-property">ylabel</span>(<span class="cm-string">'FRiP Score'</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Display the plot</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">plt</span>.<span class="cm-property">tight_layout</span>()</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">plt</span>.<span class="cm-property">show</span>()</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 393px;"></div><div class="CodeMirror-gutters" style="display: none; height: 393px;"></div></div></div></pre><p>&nbsp;</p><hr /><h1 id='genome-browser'><span>Genome browser</span></h1><ul><li><p><span>Typically we are interested in visualizing a chromatin landscape in regions using a genome browser.</span></p></li><li><p><span>You can input </span><strong><span>bed</span></strong><span> and </span><strong><span>bedgraph</span></strong><span> file</span></p></li><li><p><a href='https://igv.org/doc/desktop/'><span>The Integrative Genomic Viewer(IGV)</span></a><span>provides a web app version and a local desktop version that is easy to use.</span></p></li><li><p><a href='https://genome.ucsc.edu/'><span>The UCSC Genome Browser</span></a><span> provides the most comprehensive supplementary genome information.</span></p></li></ul><p><img src="./picture/057.png" width="900"/></p><p>&nbsp;</p><hr /><h1 id='normalization-methods'><span>Normalization methods</span></h1><h2 id='rpkm'><span>RPKM</span></h2><p><strong><span>RPKM(Reads Per Kilobase Million)</span></strong></p><ul><li><p><strong><span>Definition</span></strong><span>: Normalizes read counts by both the sequencing depth (in millions of reads) and the length of each region (in kilobases).</span></p></li><li><p><strong><span>Purpose</span></strong><span>: Corrects for sequencing depth and gene/feature length, commonly used in </span><strong><span>RNA-seq</span></strong><span>.</span></p></li></ul><div contenteditable="false" spellcheck="false" class="mathjax-block md-end-block md-math-block md-rawblock" id="mathjax-n3335" cid="n3335" mdtype="math_block" data-math-tag-before="0" data-math-tag-after="0" data-math-labels="[]"><div class="md-math-container"><mjx-container class="MathJax" jax="SVG" display="true" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="69.645ex" height="5.656ex" role="img" focusable="false" viewBox="0 -1540 30783 2500" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -2.172ex;"><defs><path id="MJX-16-TEX-N-52" d="M130 622Q123 629 119 631T103 634T60 637H27V683H202H236H300Q376 683 417 677T500 648Q595 600 609 517Q610 512 610 501Q610 468 594 439T556 392T511 361T472 343L456 338Q459 335 467 332Q497 316 516 298T545 254T559 211T568 155T578 94Q588 46 602 31T640 16H645Q660 16 674 32T692 87Q692 98 696 101T712 105T728 103T732 90Q732 59 716 27T672 -16Q656 -22 630 -22Q481 -16 458 90Q456 101 456 163T449 246Q430 304 373 320L363 322L297 323H231V192L232 61Q238 51 249 49T301 46H334V0H323Q302 3 181 3Q59 3 38 0H27V46H60Q102 47 111 49T130 61V622ZM491 499V509Q491 527 490 539T481 570T462 601T424 623T362 636Q360 636 340 636T304 637H283Q238 637 234 628Q231 624 231 492V360H289Q390 360 434 378T489 456Q491 467 491 499Z"></path><path id="MJX-16-TEX-N-50" d="M130 622Q123 629 119 631T103 634T60 637H27V683H214Q237 683 276 683T331 684Q419 684 471 671T567 616Q624 563 624 489Q624 421 573 372T451 307Q429 302 328 301H234V181Q234 62 237 58Q245 47 304 46H337V0H326Q305 3 182 3Q47 3 38 0H27V46H60Q102 47 111 49T130 61V622ZM507 488Q507 514 506 528T500 564T483 597T450 620T397 635Q385 637 307 637H286Q237 637 234 628Q231 624 231 483V342H302H339Q390 342 423 349T481 382Q507 411 507 488Z"></path><path id="MJX-16-TEX-N-4B" d="M128 622Q121 629 117 631T101 634T58 637H25V683H36Q57 680 180 680Q315 680 324 683H335V637H313Q235 637 233 620Q232 618 232 462L233 307L379 449Q425 494 479 546Q518 584 524 591T531 607V608Q531 630 503 636Q501 636 498 636T493 637H489V683H499Q517 680 630 680Q704 680 716 683H722V637H708Q633 633 589 597Q584 592 495 506T406 419T515 254T631 80Q644 60 662 54T715 46H736V0H728Q719 3 615 3Q493 3 472 0H461V46H469Q515 46 515 72Q515 78 512 84L336 351Q332 348 278 296L232 251V156Q232 62 235 58Q243 47 302 46H335V0H324Q303 3 180 3Q45 3 36 0H25V46H58Q100 47 109 49T128 61V622Z"></path><path id="MJX-16-TEX-N-4D" d="M132 622Q125 629 121 631T105 634T62 637H29V683H135Q221 683 232 682T249 675Q250 674 354 398L458 124L562 398Q666 674 668 675Q671 681 683 682T781 683H887V637H854Q814 636 803 634T785 622V61Q791 51 802 49T854 46H887V0H876Q855 3 736 3Q605 3 596 0H585V46H618Q660 47 669 49T688 61V347Q688 424 688 461T688 546T688 613L687 632Q454 14 450 7Q446 1 430 1T410 7Q409 9 292 316L176 624V606Q175 588 175 543T175 463T175 356L176 86Q187 50 261 46H278V0H269Q254 3 154 3Q52 3 37 0H29V46H46Q78 48 98 56T122 69T132 86V622Z"></path><path id="MJX-16-TEX-N-3D" d="M56 347Q56 360 70 367H707Q722 359 722 347Q722 336 708 328L390 327H72Q56 332 56 347ZM56 153Q56 168 72 173H708Q722 163 722 153Q722 140 707 133H70Q56 140 56 153Z"></path><path id="MJX-16-TEX-N-72" d="M36 46H50Q89 46 97 60V68Q97 77 97 91T98 122T98 161T98 203Q98 234 98 269T98 328L97 351Q94 370 83 376T38 385H20V408Q20 431 22 431L32 432Q42 433 60 434T96 436Q112 437 131 438T160 441T171 442H174V373Q213 441 271 441H277Q322 441 343 419T364 373Q364 352 351 337T313 322Q288 322 276 338T263 372Q263 381 265 388T270 400T273 405Q271 407 250 401Q234 393 226 386Q179 341 179 207V154Q179 141 179 127T179 101T180 81T180 66V61Q181 59 183 57T188 54T193 51T200 49T207 48T216 47T225 47T235 46T245 46H276V0H267Q249 3 140 3Q37 3 28 0H20V46H36Z"></path><path id="MJX-16-TEX-N-65" d="M28 218Q28 273 48 318T98 391T163 433T229 448Q282 448 320 430T378 380T406 316T415 245Q415 238 408 231H126V216Q126 68 226 36Q246 30 270 30Q312 30 342 62Q359 79 369 104L379 128Q382 131 395 131H398Q415 131 415 121Q415 117 412 108Q393 53 349 21T250 -11Q155 -11 92 58T28 218ZM333 275Q322 403 238 411H236Q228 411 220 410T195 402T166 381T143 340T127 274V267H333V275Z"></path><path id="MJX-16-TEX-N-61" d="M137 305T115 305T78 320T63 359Q63 394 97 421T218 448Q291 448 336 416T396 340Q401 326 401 309T402 194V124Q402 76 407 58T428 40Q443 40 448 56T453 109V145H493V106Q492 66 490 59Q481 29 455 12T400 -6T353 12T329 54V58L327 55Q325 52 322 49T314 40T302 29T287 17T269 6T247 -2T221 -8T190 -11Q130 -11 82 20T34 107Q34 128 41 147T68 188T116 225T194 253T304 268H318V290Q318 324 312 340Q290 411 215 411Q197 411 181 410T156 406T148 403Q170 388 170 359Q170 334 154 320ZM126 106Q126 75 150 51T209 26Q247 26 276 49T315 109Q317 116 318 175Q318 233 317 233Q309 233 296 232T251 223T193 203T147 166T126 106Z"></path><path id="MJX-16-TEX-N-64" d="M376 495Q376 511 376 535T377 568Q377 613 367 624T316 637H298V660Q298 683 300 683L310 684Q320 685 339 686T376 688Q393 689 413 690T443 693T454 694H457V390Q457 84 458 81Q461 61 472 55T517 46H535V0Q533 0 459 -5T380 -11H373V44L365 37Q307 -11 235 -11Q158 -11 96 50T34 215Q34 315 97 378T244 442Q319 442 376 393V495ZM373 342Q328 405 260 405Q211 405 173 369Q146 341 139 305T131 211Q131 155 138 120T173 59Q203 26 251 26Q322 26 373 103V342Z"></path><path id="MJX-16-TEX-N-73" d="M295 316Q295 356 268 385T190 414Q154 414 128 401Q98 382 98 349Q97 344 98 336T114 312T157 287Q175 282 201 278T245 269T277 256Q294 248 310 236T342 195T359 133Q359 71 321 31T198 -10H190Q138 -10 94 26L86 19L77 10Q71 4 65 -1L54 -11H46H42Q39 -11 33 -5V74V132Q33 153 35 157T45 162H54Q66 162 70 158T75 146T82 119T101 77Q136 26 198 26Q295 26 295 104Q295 133 277 151Q257 175 194 187T111 210Q75 227 54 256T33 318Q33 357 50 384T93 424T143 442T187 447H198Q238 447 268 432L283 424L292 431Q302 440 314 448H322H326Q329 448 335 442V310L329 304H301Q295 310 295 316Z"></path><path id="MJX-16-TEX-N-20" d=""></path><path id="MJX-16-TEX-N-69" d="M69 609Q69 637 87 653T131 669Q154 667 171 652T188 609Q188 579 171 564T129 549Q104 549 87 564T69 609ZM247 0Q232 3 143 3Q132 3 106 3T56 1L34 0H26V46H42Q70 46 91 49Q100 53 102 60T104 102V205V293Q104 345 102 359T88 378Q74 385 41 385H30V408Q30 431 32 431L42 432Q52 433 70 434T106 436Q123 437 142 438T171 441T182 442H185V62Q190 52 197 50T232 46H255V0H247Z"></path><path id="MJX-16-TEX-N-6E" d="M41 46H55Q94 46 102 60V68Q102 77 102 91T102 122T103 161T103 203Q103 234 103 269T102 328V351Q99 370 88 376T43 385H25V408Q25 431 27 431L37 432Q47 433 65 434T102 436Q119 437 138 438T167 441T178 442H181V402Q181 364 182 364T187 369T199 384T218 402T247 421T285 437Q305 442 336 442Q450 438 463 329Q464 322 464 190V104Q464 66 466 59T477 49Q498 46 526 46H542V0H534L510 1Q487 2 460 2T422 3Q319 3 310 0H302V46H318Q379 46 379 62Q380 64 380 200Q379 335 378 343Q372 371 358 385T334 402T308 404Q263 404 229 370Q202 343 195 315T187 232V168V108Q187 78 188 68T191 55T200 49Q221 46 249 46H265V0H257L234 1Q210 2 183 2T145 3Q42 3 33 0H25V46H41Z"></path><path id="MJX-16-TEX-N-67" d="M329 409Q373 453 429 453Q459 453 472 434T485 396Q485 382 476 371T449 360Q416 360 412 390Q410 404 415 411Q415 412 416 414V415Q388 412 363 393Q355 388 355 386Q355 385 359 381T368 369T379 351T388 325T392 292Q392 230 343 187T222 143Q172 143 123 171Q112 153 112 133Q112 98 138 81Q147 75 155 75T227 73Q311 72 335 67Q396 58 431 26Q470 -13 470 -72Q470 -139 392 -175Q332 -206 250 -206Q167 -206 107 -175Q29 -140 29 -75Q29 -39 50 -15T92 18L103 24Q67 55 67 108Q67 155 96 193Q52 237 52 292Q52 355 102 398T223 442Q274 442 318 416L329 409ZM299 343Q294 371 273 387T221 404Q192 404 171 388T145 343Q142 326 142 292Q142 248 149 227T179 192Q196 182 222 182Q244 182 260 189T283 207T294 227T299 242Q302 258 302 292T299 343ZM403 -75Q403 -50 389 -34T348 -11T299 -2T245 0H218Q151 0 138 -6Q118 -15 107 -34T95 -74Q95 -84 101 -97T122 -127T170 -155T250 -167Q319 -167 361 -139T403 -75Z"></path><path id="MJX-16-TEX-N-6F" d="M28 214Q28 309 93 378T250 448Q340 448 405 380T471 215Q471 120 407 55T250 -10Q153 -10 91 57T28 214ZM250 30Q372 30 372 193V225V250Q372 272 371 288T364 326T348 362T317 390T268 410Q263 411 252 411Q222 411 195 399Q152 377 139 338T126 246V226Q126 130 145 91Q177 30 250 30Z"></path><path id="MJX-16-TEX-N-D7" d="M630 29Q630 9 609 9Q604 9 587 25T493 118L389 222L284 117Q178 13 175 11Q171 9 168 9Q160 9 154 15T147 29Q147 36 161 51T255 146L359 250L255 354Q174 435 161 449T147 471Q147 480 153 485T168 490Q173 490 175 489Q178 487 284 383L389 278L493 382Q570 459 587 475T609 491Q630 491 630 471Q630 464 620 453T522 355L418 250L522 145Q606 61 618 48T630 29Z"></path><path id="MJX-16-TEX-N-31" d="M213 578L200 573Q186 568 160 563T102 556H83V602H102Q149 604 189 617T245 641T273 663Q275 666 285 666Q294 666 302 660V361L303 61Q310 54 315 52T339 48T401 46H427V0H416Q395 3 257 3Q121 3 100 0H88V46H114Q136 46 152 46T177 47T193 50T201 52T207 57T213 61V578Z"></path><path id="MJX-16-TEX-N-30" d="M96 585Q152 666 249 666Q297 666 345 640T423 548Q460 465 460 320Q460 165 417 83Q397 41 362 16T301 -15T250 -22Q224 -22 198 -16T137 16T82 83Q39 165 39 320Q39 494 96 585ZM321 597Q291 629 250 629Q208 629 178 597Q153 571 145 525T137 333Q137 175 145 125T181 46Q209 16 250 16Q290 16 318 46Q347 76 354 130T362 333Q362 478 354 524T321 597Z"></path><path id="MJX-16-TEX-N-39" d="M352 287Q304 211 232 211Q154 211 104 270T44 396Q42 412 42 436V444Q42 537 111 606Q171 666 243 666Q245 666 249 666T257 665H261Q273 665 286 663T323 651T370 619T413 560Q456 472 456 334Q456 194 396 97Q361 41 312 10T208 -22Q147 -22 108 7T68 93T121 149Q143 149 158 135T173 96Q173 78 164 65T148 49T135 44L131 43Q131 41 138 37T164 27T206 22H212Q272 22 313 86Q352 142 352 280V287ZM244 248Q292 248 321 297T351 430Q351 508 343 542Q341 552 337 562T323 588T293 615T246 625Q208 625 181 598Q160 576 154 546T147 441Q147 358 152 329T172 282Q197 248 244 248Z"></path><path id="MJX-16-TEX-N-74" d="M27 422Q80 426 109 478T141 600V615H181V431H316V385H181V241Q182 116 182 100T189 68Q203 29 238 29Q282 29 292 100Q293 108 293 146V181H333V146V134Q333 57 291 17Q264 -10 221 -10Q187 -10 162 2T124 33T105 68T98 100Q97 107 97 248V385H18V422H27Z"></path><path id="MJX-16-TEX-N-6C" d="M42 46H56Q95 46 103 60V68Q103 77 103 91T103 124T104 167T104 217T104 272T104 329Q104 366 104 407T104 482T104 542T103 586T103 603Q100 622 89 628T44 637H26V660Q26 683 28 683L38 684Q48 685 67 686T104 688Q121 689 141 690T171 693T182 694H185V379Q185 62 186 60Q190 52 198 49Q219 46 247 46H263V0H255L232 1Q209 2 183 2T145 3T107 3T57 1L34 0H26V46H42Z"></path><path id="MJX-16-TEX-N-6D" d="M41 46H55Q94 46 102 60V68Q102 77 102 91T102 122T103 161T103 203Q103 234 103 269T102 328V351Q99 370 88 376T43 385H25V408Q25 431 27 431L37 432Q47 433 65 434T102 436Q119 437 138 438T167 441T178 442H181V402Q181 364 182 364T187 369T199 384T218 402T247 421T285 437Q305 442 336 442Q351 442 364 440T387 434T406 426T421 417T432 406T441 395T448 384T452 374T455 366L457 361L460 365Q463 369 466 373T475 384T488 397T503 410T523 422T546 432T572 439T603 442Q729 442 740 329Q741 322 741 190V104Q741 66 743 59T754 49Q775 46 803 46H819V0H811L788 1Q764 2 737 2T699 3Q596 3 587 0H579V46H595Q656 46 656 62Q657 64 657 200Q656 335 655 343Q649 371 635 385T611 402T585 404Q540 404 506 370Q479 343 472 315T464 232V168V108Q464 78 465 68T468 55T477 49Q498 46 526 46H542V0H534L510 1Q487 2 460 2T422 3Q319 3 310 0H302V46H318Q379 46 379 62Q380 64 380 200Q379 335 378 343Q372 371 358 385T334 402T308 404Q263 404 229 370Q202 343 195 315T187 232V168V108Q187 78 188 68T191 55T200 49Q221 46 249 46H265V0H257L234 1Q210 2 183 2T145 3Q42 3 33 0H25V46H41Z"></path><path id="MJX-16-TEX-N-70" d="M36 -148H50Q89 -148 97 -134V-126Q97 -119 97 -107T97 -77T98 -38T98 6T98 55T98 106Q98 140 98 177T98 243T98 296T97 335T97 351Q94 370 83 376T38 385H20V408Q20 431 22 431L32 432Q42 433 61 434T98 436Q115 437 135 438T165 441T176 442H179V416L180 390L188 397Q247 441 326 441Q407 441 464 377T522 216Q522 115 457 52T310 -11Q242 -11 190 33L182 40V-45V-101Q182 -128 184 -134T195 -145Q216 -148 244 -148H260V-194H252L228 -193Q205 -192 178 -192T140 -191Q37 -191 28 -194H20V-148H36ZM424 218Q424 292 390 347T305 402Q234 402 182 337V98Q222 26 294 26Q345 26 384 80T424 218Z"></path><path id="MJX-16-TEX-N-68" d="M41 46H55Q94 46 102 60V68Q102 77 102 91T102 124T102 167T103 217T103 272T103 329Q103 366 103 407T103 482T102 542T102 586T102 603Q99 622 88 628T43 637H25V660Q25 683 27 683L37 684Q47 685 66 686T103 688Q120 689 140 690T170 693T181 694H184V367Q244 442 328 442Q451 442 463 329Q464 322 464 190V104Q464 66 466 59T477 49Q498 46 526 46H542V0H534L510 1Q487 2 460 2T422 3Q319 3 310 0H302V46H318Q379 46 379 62Q380 64 380 200Q379 335 378 343Q372 371 358 385T334 402T308 404Q263 404 229 370Q202 343 195 315T187 232V168V108Q187 78 188 68T191 55T200 49Q221 46 249 46H265V0H257L234 1Q210 2 183 2T145 3Q42 3 33 0H25V46H41Z"></path><path id="MJX-16-TEX-N-28" d="M94 250Q94 319 104 381T127 488T164 576T202 643T244 695T277 729T302 750H315H319Q333 750 333 741Q333 738 316 720T275 667T226 581T184 443T167 250T184 58T225 -81T274 -167T316 -220T333 -241Q333 -250 318 -250H315H302L274 -226Q180 -141 137 -14T94 250Z"></path><path id="MJX-16-TEX-N-75" d="M383 58Q327 -10 256 -10H249Q124 -10 105 89Q104 96 103 226Q102 335 102 348T96 369Q86 385 36 385H25V408Q25 431 27 431L38 432Q48 433 67 434T105 436Q122 437 142 438T172 441T184 442H187V261Q188 77 190 64Q193 49 204 40Q224 26 264 26Q290 26 311 35T343 58T363 90T375 120T379 144Q379 145 379 161T380 201T380 248V315Q380 361 370 372T320 385H302V431Q304 431 378 436T457 442H464V264Q464 84 465 81Q468 61 479 55T524 46H542V0Q540 0 467 -5T390 -11H383V58Z"></path><path id="MJX-16-TEX-N-62" d="M307 -11Q234 -11 168 55L158 37Q156 34 153 28T147 17T143 10L138 1L118 0H98V298Q98 599 97 603Q94 622 83 628T38 637H20V660Q20 683 22 683L32 684Q42 685 61 686T98 688Q115 689 135 690T165 693T176 694H179V543Q179 391 180 391L183 394Q186 397 192 401T207 411T228 421T254 431T286 439T323 442Q401 442 461 379T522 216Q522 115 458 52T307 -11ZM182 98Q182 97 187 90T196 79T206 67T218 55T233 44T250 35T271 29T295 26Q330 26 363 46T412 113Q424 148 424 212Q424 287 412 323Q385 405 300 405Q270 405 239 390T188 347L182 339V98Z"></path><path id="MJX-16-TEX-N-29" d="M60 749L64 750Q69 750 74 750H86L114 726Q208 641 251 514T294 250Q294 182 284 119T261 12T224 -76T186 -143T145 -194T113 -227T90 -246Q87 -249 86 -250H74Q66 -250 63 -250T58 -247T55 -238Q56 -237 66 -225Q221 -64 221 250T66 725Q56 737 55 738Q55 746 60 749Z"></path><path id="MJX-16-TEX-N-A0" d=""></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="mtext"><use data-c="52" xlink:href="#MJX-16-TEX-N-52"></use><use data-c="50" xlink:href="#MJX-16-TEX-N-50" transform="translate(736,0)"></use><use data-c="4B" xlink:href="#MJX-16-TEX-N-4B" transform="translate(1417,0)"></use><use data-c="4D" xlink:href="#MJX-16-TEX-N-4D" transform="translate(2195,0)"></use></g><g data-mml-node="mo" transform="translate(3389.8,0)"><use data-c="3D" xlink:href="#MJX-16-TEX-N-3D"></use></g><g data-mml-node="mfrac" transform="translate(4445.6,0)"><g data-mml-node="mrow" transform="translate(8319.2,676)"><g data-mml-node="mtext"><use data-c="72" xlink:href="#MJX-16-TEX-N-72"></use><use data-c="65" xlink:href="#MJX-16-TEX-N-65" transform="translate(392,0)"></use><use data-c="61" xlink:href="#MJX-16-TEX-N-61" transform="translate(836,0)"></use><use data-c="64" xlink:href="#MJX-16-TEX-N-64" transform="translate(1336,0)"></use><use data-c="73" xlink:href="#MJX-16-TEX-N-73" transform="translate(1892,0)"></use><use data-c="20" xlink:href="#MJX-16-TEX-N-20" transform="translate(2286,0)"></use><use data-c="69" xlink:href="#MJX-16-TEX-N-69" transform="translate(2536,0)"></use><use data-c="6E" xlink:href="#MJX-16-TEX-N-6E" transform="translate(2814,0)"></use><use data-c="20" xlink:href="#MJX-16-TEX-N-20" transform="translate(3370,0)"></use><use data-c="61" xlink:href="#MJX-16-TEX-N-61" transform="translate(3620,0)"></use><use data-c="20" xlink:href="#MJX-16-TEX-N-20" transform="translate(4120,0)"></use><use data-c="72" xlink:href="#MJX-16-TEX-N-72" transform="translate(4370,0)"></use><use data-c="65" xlink:href="#MJX-16-TEX-N-65" transform="translate(4762,0)"></use><use data-c="67" xlink:href="#MJX-16-TEX-N-67" transform="translate(5206,0)"></use><use data-c="69" xlink:href="#MJX-16-TEX-N-69" transform="translate(5706,0)"></use><use data-c="6F" xlink:href="#MJX-16-TEX-N-6F" transform="translate(5984,0)"></use><use data-c="6E" xlink:href="#MJX-16-TEX-N-6E" transform="translate(6484,0)"></use></g><g data-mml-node="mo" transform="translate(7262.2,0)"><use data-c="D7" xlink:href="#MJX-16-TEX-N-D7"></use></g><g data-mml-node="msup" transform="translate(8262.4,0)"><g data-mml-node="mn"><use data-c="31" xlink:href="#MJX-16-TEX-N-31"></use><use data-c="30" xlink:href="#MJX-16-TEX-N-30" transform="translate(500,0)"></use></g><g data-mml-node="mn" transform="translate(1033,393.1) scale(0.707)"><use data-c="39" xlink:href="#MJX-16-TEX-N-39"></use></g></g></g><g data-mml-node="mrow" transform="translate(220,-710)"><g data-mml-node="mtext"><use data-c="74" xlink:href="#MJX-16-TEX-N-74"></use><use data-c="6F" xlink:href="#MJX-16-TEX-N-6F" transform="translate(389,0)"></use><use data-c="74" xlink:href="#MJX-16-TEX-N-74" transform="translate(889,0)"></use><use data-c="61" xlink:href="#MJX-16-TEX-N-61" transform="translate(1278,0)"></use><use data-c="6C" xlink:href="#MJX-16-TEX-N-6C" transform="translate(1778,0)"></use><use data-c="20" xlink:href="#MJX-16-TEX-N-20" transform="translate(2056,0)"></use><use data-c="6D" xlink:href="#MJX-16-TEX-N-6D" transform="translate(2306,0)"></use><use data-c="61" xlink:href="#MJX-16-TEX-N-61" transform="translate(3139,0)"></use><use data-c="70" xlink:href="#MJX-16-TEX-N-70" transform="translate(3639,0)"></use><use data-c="70" xlink:href="#MJX-16-TEX-N-70" transform="translate(4195,0)"></use><use data-c="65" xlink:href="#MJX-16-TEX-N-65" transform="translate(4751,0)"></use><use data-c="64" xlink:href="#MJX-16-TEX-N-64" transform="translate(5195,0)"></use><use data-c="20" xlink:href="#MJX-16-TEX-N-20" transform="translate(5751,0)"></use><use data-c="72" xlink:href="#MJX-16-TEX-N-72" transform="translate(6001,0)"></use><use data-c="65" xlink:href="#MJX-16-TEX-N-65" transform="translate(6393,0)"></use><use data-c="61" xlink:href="#MJX-16-TEX-N-61" transform="translate(6837,0)"></use><use data-c="64" xlink:href="#MJX-16-TEX-N-64" transform="translate(7337,0)"></use><use data-c="73" xlink:href="#MJX-16-TEX-N-73" transform="translate(7893,0)"></use></g><g data-mml-node="mo" transform="translate(8509.2,0)"><use data-c="D7" xlink:href="#MJX-16-TEX-N-D7"></use></g><g data-mml-node="mtext" transform="translate(9509.4,0)"><use data-c="72" xlink:href="#MJX-16-TEX-N-72"></use><use data-c="65" xlink:href="#MJX-16-TEX-N-65" transform="translate(392,0)"></use><use data-c="67" xlink:href="#MJX-16-TEX-N-67" transform="translate(836,0)"></use><use data-c="69" xlink:href="#MJX-16-TEX-N-69" transform="translate(1336,0)"></use><use data-c="6F" xlink:href="#MJX-16-TEX-N-6F" transform="translate(1614,0)"></use><use data-c="6E" xlink:href="#MJX-16-TEX-N-6E" transform="translate(2114,0)"></use><use data-c="20" xlink:href="#MJX-16-TEX-N-20" transform="translate(2670,0)"></use><use data-c="6C" xlink:href="#MJX-16-TEX-N-6C" transform="translate(2920,0)"></use><use data-c="65" xlink:href="#MJX-16-TEX-N-65" transform="translate(3198,0)"></use><use data-c="6E" xlink:href="#MJX-16-TEX-N-6E" transform="translate(3642,0)"></use><use data-c="67" xlink:href="#MJX-16-TEX-N-67" transform="translate(4198,0)"></use><use data-c="74" xlink:href="#MJX-16-TEX-N-74" transform="translate(4698,0)"></use><use data-c="68" xlink:href="#MJX-16-TEX-N-68" transform="translate(5087,0)"></use><use data-c="28" xlink:href="#MJX-16-TEX-N-28" transform="translate(5643,0)"></use><use data-c="6D" xlink:href="#MJX-16-TEX-N-6D" transform="translate(6032,0)"></use><use data-c="65" xlink:href="#MJX-16-TEX-N-65" transform="translate(6865,0)"></use><use data-c="61" xlink:href="#MJX-16-TEX-N-61" transform="translate(7309,0)"></use><use data-c="73" xlink:href="#MJX-16-TEX-N-73" transform="translate(7809,0)"></use><use data-c="75" xlink:href="#MJX-16-TEX-N-75" transform="translate(8203,0)"></use><use data-c="72" xlink:href="#MJX-16-TEX-N-72" transform="translate(8759,0)"></use><use data-c="65" xlink:href="#MJX-16-TEX-N-65" transform="translate(9151,0)"></use><use data-c="64" xlink:href="#MJX-16-TEX-N-64" transform="translate(9595,0)"></use><use data-c="20" xlink:href="#MJX-16-TEX-N-20" transform="translate(10151,0)"></use><use data-c="69" xlink:href="#MJX-16-TEX-N-69" transform="translate(10401,0)"></use><use data-c="6E" xlink:href="#MJX-16-TEX-N-6E" transform="translate(10679,0)"></use><use data-c="20" xlink:href="#MJX-16-TEX-N-20" transform="translate(11235,0)"></use><use data-c="62" xlink:href="#MJX-16-TEX-N-62" transform="translate(11485,0)"></use><use data-c="61" xlink:href="#MJX-16-TEX-N-61" transform="translate(12041,0)"></use><use data-c="73" xlink:href="#MJX-16-TEX-N-73" transform="translate(12541,0)"></use><use data-c="65" xlink:href="#MJX-16-TEX-N-65" transform="translate(12935,0)"></use><use data-c="20" xlink:href="#MJX-16-TEX-N-20" transform="translate(13379,0)"></use><use data-c="70" xlink:href="#MJX-16-TEX-N-70" transform="translate(13629,0)"></use><use data-c="61" xlink:href="#MJX-16-TEX-N-61" transform="translate(14185,0)"></use><use data-c="69" xlink:href="#MJX-16-TEX-N-69" transform="translate(14685,0)"></use><use data-c="72" xlink:href="#MJX-16-TEX-N-72" transform="translate(14963,0)"></use><use data-c="73" xlink:href="#MJX-16-TEX-N-73" transform="translate(15355,0)"></use><use data-c="29" xlink:href="#MJX-16-TEX-N-29" transform="translate(15749,0)"></use><use data-c="A0" xlink:href="#MJX-16-TEX-N-A0" transform="translate(16138,0)"></use></g></g><rect width="26097.4" height="60" x="120" y="220"></rect></g></g></g></svg><mjx-assistive-mml unselectable="on" display="block"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><mtext>RPKM</mtext><mo>=</mo><mfrac><mrow><mtext>reads in a region</mtext><mo>×</mo><msup><mn>10</mn><mn>9</mn></msup></mrow><mrow><mtext>total mapped reads</mtext><mo>×</mo><mtext>region length(measured in base pairs)&nbsp;</mtext></mrow></mfrac></math></mjx-assistive-mml></mjx-container></div></div><p>&nbsp;</p><hr /><h2 id='cpm'><span>CPM</span></h2><p><strong><span>CPM (Counts Per Million)</span></strong></p><ul><li><p><strong><span>Definition</span></strong><span>: Normalizes read counts based only on sequencing depth, ignoring region length.</span></p></li><li><p><strong><span>Purpose</span></strong><span>: Useful when </span><strong><span>comparing coverage</span></strong><span> across samples with different read counts, but </span><strong><span>region length is not a consideration.</span></strong></p></li></ul><div contenteditable="false" spellcheck="false" class="mathjax-block md-end-block md-math-block md-rawblock" id="mathjax-n3345" cid="n3345" mdtype="math_block" data-math-tag-before="0" data-math-tag-after="0" data-math-labels="[]"><div class="md-math-container"><mjx-container class="MathJax" jax="SVG" display="true" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="31.205ex" height="5.475ex" role="img" focusable="false" viewBox="0 -1540 13792.6 2420" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -1.991ex;"><defs><path id="MJX-17-TEX-N-43" d="M56 342Q56 428 89 500T174 615T283 681T391 705Q394 705 400 705T408 704Q499 704 569 636L582 624L612 663Q639 700 643 704Q644 704 647 704T653 705H657Q660 705 666 699V419L660 413H626Q620 419 619 430Q610 512 571 572T476 651Q457 658 426 658Q322 658 252 588Q173 509 173 342Q173 221 211 151Q232 111 263 84T328 45T384 29T428 24Q517 24 571 93T626 244Q626 251 632 257H660L666 251V236Q661 133 590 56T403 -21Q262 -21 159 83T56 342Z"></path><path id="MJX-17-TEX-N-50" d="M130 622Q123 629 119 631T103 634T60 637H27V683H214Q237 683 276 683T331 684Q419 684 471 671T567 616Q624 563 624 489Q624 421 573 372T451 307Q429 302 328 301H234V181Q234 62 237 58Q245 47 304 46H337V0H326Q305 3 182 3Q47 3 38 0H27V46H60Q102 47 111 49T130 61V622ZM507 488Q507 514 506 528T500 564T483 597T450 620T397 635Q385 637 307 637H286Q237 637 234 628Q231 624 231 483V342H302H339Q390 342 423 349T481 382Q507 411 507 488Z"></path><path id="MJX-17-TEX-N-4D" d="M132 622Q125 629 121 631T105 634T62 637H29V683H135Q221 683 232 682T249 675Q250 674 354 398L458 124L562 398Q666 674 668 675Q671 681 683 682T781 683H887V637H854Q814 636 803 634T785 622V61Q791 51 802 49T854 46H887V0H876Q855 3 736 3Q605 3 596 0H585V46H618Q660 47 669 49T688 61V347Q688 424 688 461T688 546T688 613L687 632Q454 14 450 7Q446 1 430 1T410 7Q409 9 292 316L176 624V606Q175 588 175 543T175 463T175 356L176 86Q187 50 261 46H278V0H269Q254 3 154 3Q52 3 37 0H29V46H46Q78 48 98 56T122 69T132 86V622Z"></path><path id="MJX-17-TEX-N-3D" d="M56 347Q56 360 70 367H707Q722 359 722 347Q722 336 708 328L390 327H72Q56 332 56 347ZM56 153Q56 168 72 173H708Q722 163 722 153Q722 140 707 133H70Q56 140 56 153Z"></path><path id="MJX-17-TEX-N-72" d="M36 46H50Q89 46 97 60V68Q97 77 97 91T98 122T98 161T98 203Q98 234 98 269T98 328L97 351Q94 370 83 376T38 385H20V408Q20 431 22 431L32 432Q42 433 60 434T96 436Q112 437 131 438T160 441T171 442H174V373Q213 441 271 441H277Q322 441 343 419T364 373Q364 352 351 337T313 322Q288 322 276 338T263 372Q263 381 265 388T270 400T273 405Q271 407 250 401Q234 393 226 386Q179 341 179 207V154Q179 141 179 127T179 101T180 81T180 66V61Q181 59 183 57T188 54T193 51T200 49T207 48T216 47T225 47T235 46T245 46H276V0H267Q249 3 140 3Q37 3 28 0H20V46H36Z"></path><path id="MJX-17-TEX-N-65" d="M28 218Q28 273 48 318T98 391T163 433T229 448Q282 448 320 430T378 380T406 316T415 245Q415 238 408 231H126V216Q126 68 226 36Q246 30 270 30Q312 30 342 62Q359 79 369 104L379 128Q382 131 395 131H398Q415 131 415 121Q415 117 412 108Q393 53 349 21T250 -11Q155 -11 92 58T28 218ZM333 275Q322 403 238 411H236Q228 411 220 410T195 402T166 381T143 340T127 274V267H333V275Z"></path><path id="MJX-17-TEX-N-61" d="M137 305T115 305T78 320T63 359Q63 394 97 421T218 448Q291 448 336 416T396 340Q401 326 401 309T402 194V124Q402 76 407 58T428 40Q443 40 448 56T453 109V145H493V106Q492 66 490 59Q481 29 455 12T400 -6T353 12T329 54V58L327 55Q325 52 322 49T314 40T302 29T287 17T269 6T247 -2T221 -8T190 -11Q130 -11 82 20T34 107Q34 128 41 147T68 188T116 225T194 253T304 268H318V290Q318 324 312 340Q290 411 215 411Q197 411 181 410T156 406T148 403Q170 388 170 359Q170 334 154 320ZM126 106Q126 75 150 51T209 26Q247 26 276 49T315 109Q317 116 318 175Q318 233 317 233Q309 233 296 232T251 223T193 203T147 166T126 106Z"></path><path id="MJX-17-TEX-N-64" d="M376 495Q376 511 376 535T377 568Q377 613 367 624T316 637H298V660Q298 683 300 683L310 684Q320 685 339 686T376 688Q393 689 413 690T443 693T454 694H457V390Q457 84 458 81Q461 61 472 55T517 46H535V0Q533 0 459 -5T380 -11H373V44L365 37Q307 -11 235 -11Q158 -11 96 50T34 215Q34 315 97 378T244 442Q319 442 376 393V495ZM373 342Q328 405 260 405Q211 405 173 369Q146 341 139 305T131 211Q131 155 138 120T173 59Q203 26 251 26Q322 26 373 103V342Z"></path><path id="MJX-17-TEX-N-73" d="M295 316Q295 356 268 385T190 414Q154 414 128 401Q98 382 98 349Q97 344 98 336T114 312T157 287Q175 282 201 278T245 269T277 256Q294 248 310 236T342 195T359 133Q359 71 321 31T198 -10H190Q138 -10 94 26L86 19L77 10Q71 4 65 -1L54 -11H46H42Q39 -11 33 -5V74V132Q33 153 35 157T45 162H54Q66 162 70 158T75 146T82 119T101 77Q136 26 198 26Q295 26 295 104Q295 133 277 151Q257 175 194 187T111 210Q75 227 54 256T33 318Q33 357 50 384T93 424T143 442T187 447H198Q238 447 268 432L283 424L292 431Q302 440 314 448H322H326Q329 448 335 442V310L329 304H301Q295 310 295 316Z"></path><path id="MJX-17-TEX-N-20" d=""></path><path id="MJX-17-TEX-N-69" d="M69 609Q69 637 87 653T131 669Q154 667 171 652T188 609Q188 579 171 564T129 549Q104 549 87 564T69 609ZM247 0Q232 3 143 3Q132 3 106 3T56 1L34 0H26V46H42Q70 46 91 49Q100 53 102 60T104 102V205V293Q104 345 102 359T88 378Q74 385 41 385H30V408Q30 431 32 431L42 432Q52 433 70 434T106 436Q123 437 142 438T171 441T182 442H185V62Q190 52 197 50T232 46H255V0H247Z"></path><path id="MJX-17-TEX-N-6E" d="M41 46H55Q94 46 102 60V68Q102 77 102 91T102 122T103 161T103 203Q103 234 103 269T102 328V351Q99 370 88 376T43 385H25V408Q25 431 27 431L37 432Q47 433 65 434T102 436Q119 437 138 438T167 441T178 442H181V402Q181 364 182 364T187 369T199 384T218 402T247 421T285 437Q305 442 336 442Q450 438 463 329Q464 322 464 190V104Q464 66 466 59T477 49Q498 46 526 46H542V0H534L510 1Q487 2 460 2T422 3Q319 3 310 0H302V46H318Q379 46 379 62Q380 64 380 200Q379 335 378 343Q372 371 358 385T334 402T308 404Q263 404 229 370Q202 343 195 315T187 232V168V108Q187 78 188 68T191 55T200 49Q221 46 249 46H265V0H257L234 1Q210 2 183 2T145 3Q42 3 33 0H25V46H41Z"></path><path id="MJX-17-TEX-N-67" d="M329 409Q373 453 429 453Q459 453 472 434T485 396Q485 382 476 371T449 360Q416 360 412 390Q410 404 415 411Q415 412 416 414V415Q388 412 363 393Q355 388 355 386Q355 385 359 381T368 369T379 351T388 325T392 292Q392 230 343 187T222 143Q172 143 123 171Q112 153 112 133Q112 98 138 81Q147 75 155 75T227 73Q311 72 335 67Q396 58 431 26Q470 -13 470 -72Q470 -139 392 -175Q332 -206 250 -206Q167 -206 107 -175Q29 -140 29 -75Q29 -39 50 -15T92 18L103 24Q67 55 67 108Q67 155 96 193Q52 237 52 292Q52 355 102 398T223 442Q274 442 318 416L329 409ZM299 343Q294 371 273 387T221 404Q192 404 171 388T145 343Q142 326 142 292Q142 248 149 227T179 192Q196 182 222 182Q244 182 260 189T283 207T294 227T299 242Q302 258 302 292T299 343ZM403 -75Q403 -50 389 -34T348 -11T299 -2T245 0H218Q151 0 138 -6Q118 -15 107 -34T95 -74Q95 -84 101 -97T122 -127T170 -155T250 -167Q319 -167 361 -139T403 -75Z"></path><path id="MJX-17-TEX-N-6F" d="M28 214Q28 309 93 378T250 448Q340 448 405 380T471 215Q471 120 407 55T250 -10Q153 -10 91 57T28 214ZM250 30Q372 30 372 193V225V250Q372 272 371 288T364 326T348 362T317 390T268 410Q263 411 252 411Q222 411 195 399Q152 377 139 338T126 246V226Q126 130 145 91Q177 30 250 30Z"></path><path id="MJX-17-TEX-N-D7" d="M630 29Q630 9 609 9Q604 9 587 25T493 118L389 222L284 117Q178 13 175 11Q171 9 168 9Q160 9 154 15T147 29Q147 36 161 51T255 146L359 250L255 354Q174 435 161 449T147 471Q147 480 153 485T168 490Q173 490 175 489Q178 487 284 383L389 278L493 382Q570 459 587 475T609 491Q630 491 630 471Q630 464 620 453T522 355L418 250L522 145Q606 61 618 48T630 29Z"></path><path id="MJX-17-TEX-N-31" d="M213 578L200 573Q186 568 160 563T102 556H83V602H102Q149 604 189 617T245 641T273 663Q275 666 285 666Q294 666 302 660V361L303 61Q310 54 315 52T339 48T401 46H427V0H416Q395 3 257 3Q121 3 100 0H88V46H114Q136 46 152 46T177 47T193 50T201 52T207 57T213 61V578Z"></path><path id="MJX-17-TEX-N-30" d="M96 585Q152 666 249 666Q297 666 345 640T423 548Q460 465 460 320Q460 165 417 83Q397 41 362 16T301 -15T250 -22Q224 -22 198 -16T137 16T82 83Q39 165 39 320Q39 494 96 585ZM321 597Q291 629 250 629Q208 629 178 597Q153 571 145 525T137 333Q137 175 145 125T181 46Q209 16 250 16Q290 16 318 46Q347 76 354 130T362 333Q362 478 354 524T321 597Z"></path><path id="MJX-17-TEX-N-36" d="M42 313Q42 476 123 571T303 666Q372 666 402 630T432 550Q432 525 418 510T379 495Q356 495 341 509T326 548Q326 592 373 601Q351 623 311 626Q240 626 194 566Q147 500 147 364L148 360Q153 366 156 373Q197 433 263 433H267Q313 433 348 414Q372 400 396 374T435 317Q456 268 456 210V192Q456 169 451 149Q440 90 387 34T253 -22Q225 -22 199 -14T143 16T92 75T56 172T42 313ZM257 397Q227 397 205 380T171 335T154 278T148 216Q148 133 160 97T198 39Q222 21 251 21Q302 21 329 59Q342 77 347 104T352 209Q352 289 347 316T329 361Q302 397 257 397Z"></path><path id="MJX-17-TEX-N-74" d="M27 422Q80 426 109 478T141 600V615H181V431H316V385H181V241Q182 116 182 100T189 68Q203 29 238 29Q282 29 292 100Q293 108 293 146V181H333V146V134Q333 57 291 17Q264 -10 221 -10Q187 -10 162 2T124 33T105 68T98 100Q97 107 97 248V385H18V422H27Z"></path><path id="MJX-17-TEX-N-6C" d="M42 46H56Q95 46 103 60V68Q103 77 103 91T103 124T104 167T104 217T104 272T104 329Q104 366 104 407T104 482T104 542T103 586T103 603Q100 622 89 628T44 637H26V660Q26 683 28 683L38 684Q48 685 67 686T104 688Q121 689 141 690T171 693T182 694H185V379Q185 62 186 60Q190 52 198 49Q219 46 247 46H263V0H255L232 1Q209 2 183 2T145 3T107 3T57 1L34 0H26V46H42Z"></path><path id="MJX-17-TEX-N-6D" d="M41 46H55Q94 46 102 60V68Q102 77 102 91T102 122T103 161T103 203Q103 234 103 269T102 328V351Q99 370 88 376T43 385H25V408Q25 431 27 431L37 432Q47 433 65 434T102 436Q119 437 138 438T167 441T178 442H181V402Q181 364 182 364T187 369T199 384T218 402T247 421T285 437Q305 442 336 442Q351 442 364 440T387 434T406 426T421 417T432 406T441 395T448 384T452 374T455 366L457 361L460 365Q463 369 466 373T475 384T488 397T503 410T523 422T546 432T572 439T603 442Q729 442 740 329Q741 322 741 190V104Q741 66 743 59T754 49Q775 46 803 46H819V0H811L788 1Q764 2 737 2T699 3Q596 3 587 0H579V46H595Q656 46 656 62Q657 64 657 200Q656 335 655 343Q649 371 635 385T611 402T585 404Q540 404 506 370Q479 343 472 315T464 232V168V108Q464 78 465 68T468 55T477 49Q498 46 526 46H542V0H534L510 1Q487 2 460 2T422 3Q319 3 310 0H302V46H318Q379 46 379 62Q380 64 380 200Q379 335 378 343Q372 371 358 385T334 402T308 404Q263 404 229 370Q202 343 195 315T187 232V168V108Q187 78 188 68T191 55T200 49Q221 46 249 46H265V0H257L234 1Q210 2 183 2T145 3Q42 3 33 0H25V46H41Z"></path><path id="MJX-17-TEX-N-70" d="M36 -148H50Q89 -148 97 -134V-126Q97 -119 97 -107T97 -77T98 -38T98 6T98 55T98 106Q98 140 98 177T98 243T98 296T97 335T97 351Q94 370 83 376T38 385H20V408Q20 431 22 431L32 432Q42 433 61 434T98 436Q115 437 135 438T165 441T176 442H179V416L180 390L188 397Q247 441 326 441Q407 441 464 377T522 216Q522 115 457 52T310 -11Q242 -11 190 33L182 40V-45V-101Q182 -128 184 -134T195 -145Q216 -148 244 -148H260V-194H252L228 -193Q205 -192 178 -192T140 -191Q37 -191 28 -194H20V-148H36ZM424 218Q424 292 390 347T305 402Q234 402 182 337V98Q222 26 294 26Q345 26 384 80T424 218Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="mtext"><use data-c="43" xlink:href="#MJX-17-TEX-N-43"></use><use data-c="50" xlink:href="#MJX-17-TEX-N-50" transform="translate(722,0)"></use><use data-c="4D" xlink:href="#MJX-17-TEX-N-4D" transform="translate(1403,0)"></use></g><g data-mml-node="mo" transform="translate(2597.8,0)"><use data-c="3D" xlink:href="#MJX-17-TEX-N-3D"></use></g><g data-mml-node="mfrac" transform="translate(3653.6,0)"><g data-mml-node="mrow" transform="translate(220,676)"><g data-mml-node="mtext"><use data-c="72" xlink:href="#MJX-17-TEX-N-72"></use><use data-c="65" xlink:href="#MJX-17-TEX-N-65" transform="translate(392,0)"></use><use data-c="61" xlink:href="#MJX-17-TEX-N-61" transform="translate(836,0)"></use><use data-c="64" xlink:href="#MJX-17-TEX-N-64" transform="translate(1336,0)"></use><use data-c="73" xlink:href="#MJX-17-TEX-N-73" transform="translate(1892,0)"></use><use data-c="20" xlink:href="#MJX-17-TEX-N-20" transform="translate(2286,0)"></use><use data-c="69" xlink:href="#MJX-17-TEX-N-69" transform="translate(2536,0)"></use><use data-c="6E" xlink:href="#MJX-17-TEX-N-6E" transform="translate(2814,0)"></use><use data-c="20" xlink:href="#MJX-17-TEX-N-20" transform="translate(3370,0)"></use><use data-c="61" xlink:href="#MJX-17-TEX-N-61" transform="translate(3620,0)"></use><use data-c="20" xlink:href="#MJX-17-TEX-N-20" transform="translate(4120,0)"></use><use data-c="72" xlink:href="#MJX-17-TEX-N-72" transform="translate(4370,0)"></use><use data-c="65" xlink:href="#MJX-17-TEX-N-65" transform="translate(4762,0)"></use><use data-c="67" xlink:href="#MJX-17-TEX-N-67" transform="translate(5206,0)"></use><use data-c="69" xlink:href="#MJX-17-TEX-N-69" transform="translate(5706,0)"></use><use data-c="6F" xlink:href="#MJX-17-TEX-N-6F" transform="translate(5984,0)"></use><use data-c="6E" xlink:href="#MJX-17-TEX-N-6E" transform="translate(6484,0)"></use></g><g data-mml-node="mo" transform="translate(7262.2,0)"><use data-c="D7" xlink:href="#MJX-17-TEX-N-D7"></use></g><g data-mml-node="msup" transform="translate(8262.4,0)"><g data-mml-node="mn"><use data-c="31" xlink:href="#MJX-17-TEX-N-31"></use><use data-c="30" xlink:href="#MJX-17-TEX-N-30" transform="translate(500,0)"></use></g><g data-mml-node="mn" transform="translate(1033,393.1) scale(0.707)"><use data-c="36" xlink:href="#MJX-17-TEX-N-36"></use></g></g></g><g data-mml-node="mtext" transform="translate(926,-686)"><use data-c="74" xlink:href="#MJX-17-TEX-N-74"></use><use data-c="6F" xlink:href="#MJX-17-TEX-N-6F" transform="translate(389,0)"></use><use data-c="74" xlink:href="#MJX-17-TEX-N-74" transform="translate(889,0)"></use><use data-c="61" xlink:href="#MJX-17-TEX-N-61" transform="translate(1278,0)"></use><use data-c="6C" xlink:href="#MJX-17-TEX-N-6C" transform="translate(1778,0)"></use><use data-c="20" xlink:href="#MJX-17-TEX-N-20" transform="translate(2056,0)"></use><use data-c="6D" xlink:href="#MJX-17-TEX-N-6D" transform="translate(2306,0)"></use><use data-c="61" xlink:href="#MJX-17-TEX-N-61" transform="translate(3139,0)"></use><use data-c="70" xlink:href="#MJX-17-TEX-N-70" transform="translate(3639,0)"></use><use data-c="70" xlink:href="#MJX-17-TEX-N-70" transform="translate(4195,0)"></use><use data-c="65" xlink:href="#MJX-17-TEX-N-65" transform="translate(4751,0)"></use><use data-c="64" xlink:href="#MJX-17-TEX-N-64" transform="translate(5195,0)"></use><use data-c="20" xlink:href="#MJX-17-TEX-N-20" transform="translate(5751,0)"></use><use data-c="72" xlink:href="#MJX-17-TEX-N-72" transform="translate(6001,0)"></use><use data-c="65" xlink:href="#MJX-17-TEX-N-65" transform="translate(6393,0)"></use><use data-c="61" xlink:href="#MJX-17-TEX-N-61" transform="translate(6837,0)"></use><use data-c="64" xlink:href="#MJX-17-TEX-N-64" transform="translate(7337,0)"></use><use data-c="73" xlink:href="#MJX-17-TEX-N-73" transform="translate(7893,0)"></use></g><rect width="9899" height="60" x="120" y="220"></rect></g></g></g></svg><mjx-assistive-mml unselectable="on" display="block"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><mtext>CPM</mtext><mo>=</mo><mfrac><mrow><mtext>reads in a region</mtext><mo>×</mo><msup><mn>10</mn><mn>6</mn></msup></mrow><mtext>total mapped reads</mtext></mfrac></math></mjx-assistive-mml></mjx-container></div></div><p>&nbsp;</p><hr /><h2 id='bpm'><span>BPM</span></h2><p><strong><span>BPM (Bins Per Million mapped reads)</span></strong></p><ul><li><p><strong><span>Definition</span></strong><span>: Similar to CPM but applies normalization across </span><strong><span>equal-size bins</span></strong><span> rather than distinct genomic regions.</span></p></li><li><p><strong><span>Purpose</span></strong><span>: Often used in whole-genome coverage analysis, where it’s important to normalize read counts across consistent bins.</span></p><ul><li><p><strong><span>equal-size bins</span></strong><span> ：dividing the entire genome into segments or windows of the same fixed length (e.g., 100 bp, 500 bp, or 1 kb)</span></p></li></ul></li></ul><div contenteditable="false" spellcheck="false" class="mathjax-block md-end-block md-math-block md-rawblock" id="mathjax-n3358" cid="n3358" mdtype="math_block" data-math-tag-before="0" data-math-tag-after="0" data-math-labels="[]"><div class="md-math-container"><mjx-container class="MathJax" jax="SVG" display="true" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="28.277ex" height="5.475ex" role="img" focusable="false" viewBox="0 -1540 12498.6 2420" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -1.991ex;"><defs><path id="MJX-18-TEX-N-42" d="M131 622Q124 629 120 631T104 634T61 637H28V683H229H267H346Q423 683 459 678T531 651Q574 627 599 590T624 512Q624 461 583 419T476 360L466 357Q539 348 595 302T651 187Q651 119 600 67T469 3Q456 1 242 0H28V46H61Q103 47 112 49T131 61V622ZM511 513Q511 560 485 594T416 636Q415 636 403 636T371 636T333 637Q266 637 251 636T232 628Q229 624 229 499V374H312L396 375L406 377Q410 378 417 380T442 393T474 417T499 456T511 513ZM537 188Q537 239 509 282T430 336L329 337H229V200V116Q229 57 234 52Q240 47 334 47H383Q425 47 443 53Q486 67 511 104T537 188Z"></path><path id="MJX-18-TEX-N-50" d="M130 622Q123 629 119 631T103 634T60 637H27V683H214Q237 683 276 683T331 684Q419 684 471 671T567 616Q624 563 624 489Q624 421 573 372T451 307Q429 302 328 301H234V181Q234 62 237 58Q245 47 304 46H337V0H326Q305 3 182 3Q47 3 38 0H27V46H60Q102 47 111 49T130 61V622ZM507 488Q507 514 506 528T500 564T483 597T450 620T397 635Q385 637 307 637H286Q237 637 234 628Q231 624 231 483V342H302H339Q390 342 423 349T481 382Q507 411 507 488Z"></path><path id="MJX-18-TEX-N-4D" d="M132 622Q125 629 121 631T105 634T62 637H29V683H135Q221 683 232 682T249 675Q250 674 354 398L458 124L562 398Q666 674 668 675Q671 681 683 682T781 683H887V637H854Q814 636 803 634T785 622V61Q791 51 802 49T854 46H887V0H876Q855 3 736 3Q605 3 596 0H585V46H618Q660 47 669 49T688 61V347Q688 424 688 461T688 546T688 613L687 632Q454 14 450 7Q446 1 430 1T410 7Q409 9 292 316L176 624V606Q175 588 175 543T175 463T175 356L176 86Q187 50 261 46H278V0H269Q254 3 154 3Q52 3 37 0H29V46H46Q78 48 98 56T122 69T132 86V622Z"></path><path id="MJX-18-TEX-N-3D" d="M56 347Q56 360 70 367H707Q722 359 722 347Q722 336 708 328L390 327H72Q56 332 56 347ZM56 153Q56 168 72 173H708Q722 163 722 153Q722 140 707 133H70Q56 140 56 153Z"></path><path id="MJX-18-TEX-N-72" d="M36 46H50Q89 46 97 60V68Q97 77 97 91T98 122T98 161T98 203Q98 234 98 269T98 328L97 351Q94 370 83 376T38 385H20V408Q20 431 22 431L32 432Q42 433 60 434T96 436Q112 437 131 438T160 441T171 442H174V373Q213 441 271 441H277Q322 441 343 419T364 373Q364 352 351 337T313 322Q288 322 276 338T263 372Q263 381 265 388T270 400T273 405Q271 407 250 401Q234 393 226 386Q179 341 179 207V154Q179 141 179 127T179 101T180 81T180 66V61Q181 59 183 57T188 54T193 51T200 49T207 48T216 47T225 47T235 46T245 46H276V0H267Q249 3 140 3Q37 3 28 0H20V46H36Z"></path><path id="MJX-18-TEX-N-65" d="M28 218Q28 273 48 318T98 391T163 433T229 448Q282 448 320 430T378 380T406 316T415 245Q415 238 408 231H126V216Q126 68 226 36Q246 30 270 30Q312 30 342 62Q359 79 369 104L379 128Q382 131 395 131H398Q415 131 415 121Q415 117 412 108Q393 53 349 21T250 -11Q155 -11 92 58T28 218ZM333 275Q322 403 238 411H236Q228 411 220 410T195 402T166 381T143 340T127 274V267H333V275Z"></path><path id="MJX-18-TEX-N-61" d="M137 305T115 305T78 320T63 359Q63 394 97 421T218 448Q291 448 336 416T396 340Q401 326 401 309T402 194V124Q402 76 407 58T428 40Q443 40 448 56T453 109V145H493V106Q492 66 490 59Q481 29 455 12T400 -6T353 12T329 54V58L327 55Q325 52 322 49T314 40T302 29T287 17T269 6T247 -2T221 -8T190 -11Q130 -11 82 20T34 107Q34 128 41 147T68 188T116 225T194 253T304 268H318V290Q318 324 312 340Q290 411 215 411Q197 411 181 410T156 406T148 403Q170 388 170 359Q170 334 154 320ZM126 106Q126 75 150 51T209 26Q247 26 276 49T315 109Q317 116 318 175Q318 233 317 233Q309 233 296 232T251 223T193 203T147 166T126 106Z"></path><path id="MJX-18-TEX-N-64" d="M376 495Q376 511 376 535T377 568Q377 613 367 624T316 637H298V660Q298 683 300 683L310 684Q320 685 339 686T376 688Q393 689 413 690T443 693T454 694H457V390Q457 84 458 81Q461 61 472 55T517 46H535V0Q533 0 459 -5T380 -11H373V44L365 37Q307 -11 235 -11Q158 -11 96 50T34 215Q34 315 97 378T244 442Q319 442 376 393V495ZM373 342Q328 405 260 405Q211 405 173 369Q146 341 139 305T131 211Q131 155 138 120T173 59Q203 26 251 26Q322 26 373 103V342Z"></path><path id="MJX-18-TEX-N-73" d="M295 316Q295 356 268 385T190 414Q154 414 128 401Q98 382 98 349Q97 344 98 336T114 312T157 287Q175 282 201 278T245 269T277 256Q294 248 310 236T342 195T359 133Q359 71 321 31T198 -10H190Q138 -10 94 26L86 19L77 10Q71 4 65 -1L54 -11H46H42Q39 -11 33 -5V74V132Q33 153 35 157T45 162H54Q66 162 70 158T75 146T82 119T101 77Q136 26 198 26Q295 26 295 104Q295 133 277 151Q257 175 194 187T111 210Q75 227 54 256T33 318Q33 357 50 384T93 424T143 442T187 447H198Q238 447 268 432L283 424L292 431Q302 440 314 448H322H326Q329 448 335 442V310L329 304H301Q295 310 295 316Z"></path><path id="MJX-18-TEX-N-20" d=""></path><path id="MJX-18-TEX-N-69" d="M69 609Q69 637 87 653T131 669Q154 667 171 652T188 609Q188 579 171 564T129 549Q104 549 87 564T69 609ZM247 0Q232 3 143 3Q132 3 106 3T56 1L34 0H26V46H42Q70 46 91 49Q100 53 102 60T104 102V205V293Q104 345 102 359T88 378Q74 385 41 385H30V408Q30 431 32 431L42 432Q52 433 70 434T106 436Q123 437 142 438T171 441T182 442H185V62Q190 52 197 50T232 46H255V0H247Z"></path><path id="MJX-18-TEX-N-6E" d="M41 46H55Q94 46 102 60V68Q102 77 102 91T102 122T103 161T103 203Q103 234 103 269T102 328V351Q99 370 88 376T43 385H25V408Q25 431 27 431L37 432Q47 433 65 434T102 436Q119 437 138 438T167 441T178 442H181V402Q181 364 182 364T187 369T199 384T218 402T247 421T285 437Q305 442 336 442Q450 438 463 329Q464 322 464 190V104Q464 66 466 59T477 49Q498 46 526 46H542V0H534L510 1Q487 2 460 2T422 3Q319 3 310 0H302V46H318Q379 46 379 62Q380 64 380 200Q379 335 378 343Q372 371 358 385T334 402T308 404Q263 404 229 370Q202 343 195 315T187 232V168V108Q187 78 188 68T191 55T200 49Q221 46 249 46H265V0H257L234 1Q210 2 183 2T145 3Q42 3 33 0H25V46H41Z"></path><path id="MJX-18-TEX-N-62" d="M307 -11Q234 -11 168 55L158 37Q156 34 153 28T147 17T143 10L138 1L118 0H98V298Q98 599 97 603Q94 622 83 628T38 637H20V660Q20 683 22 683L32 684Q42 685 61 686T98 688Q115 689 135 690T165 693T176 694H179V543Q179 391 180 391L183 394Q186 397 192 401T207 411T228 421T254 431T286 439T323 442Q401 442 461 379T522 216Q522 115 458 52T307 -11ZM182 98Q182 97 187 90T196 79T206 67T218 55T233 44T250 35T271 29T295 26Q330 26 363 46T412 113Q424 148 424 212Q424 287 412 323Q385 405 300 405Q270 405 239 390T188 347L182 339V98Z"></path><path id="MJX-18-TEX-N-D7" d="M630 29Q630 9 609 9Q604 9 587 25T493 118L389 222L284 117Q178 13 175 11Q171 9 168 9Q160 9 154 15T147 29Q147 36 161 51T255 146L359 250L255 354Q174 435 161 449T147 471Q147 480 153 485T168 490Q173 490 175 489Q178 487 284 383L389 278L493 382Q570 459 587 475T609 491Q630 491 630 471Q630 464 620 453T522 355L418 250L522 145Q606 61 618 48T630 29Z"></path><path id="MJX-18-TEX-N-31" d="M213 578L200 573Q186 568 160 563T102 556H83V602H102Q149 604 189 617T245 641T273 663Q275 666 285 666Q294 666 302 660V361L303 61Q310 54 315 52T339 48T401 46H427V0H416Q395 3 257 3Q121 3 100 0H88V46H114Q136 46 152 46T177 47T193 50T201 52T207 57T213 61V578Z"></path><path id="MJX-18-TEX-N-30" d="M96 585Q152 666 249 666Q297 666 345 640T423 548Q460 465 460 320Q460 165 417 83Q397 41 362 16T301 -15T250 -22Q224 -22 198 -16T137 16T82 83Q39 165 39 320Q39 494 96 585ZM321 597Q291 629 250 629Q208 629 178 597Q153 571 145 525T137 333Q137 175 145 125T181 46Q209 16 250 16Q290 16 318 46Q347 76 354 130T362 333Q362 478 354 524T321 597Z"></path><path id="MJX-18-TEX-N-36" d="M42 313Q42 476 123 571T303 666Q372 666 402 630T432 550Q432 525 418 510T379 495Q356 495 341 509T326 548Q326 592 373 601Q351 623 311 626Q240 626 194 566Q147 500 147 364L148 360Q153 366 156 373Q197 433 263 433H267Q313 433 348 414Q372 400 396 374T435 317Q456 268 456 210V192Q456 169 451 149Q440 90 387 34T253 -22Q225 -22 199 -14T143 16T92 75T56 172T42 313ZM257 397Q227 397 205 380T171 335T154 278T148 216Q148 133 160 97T198 39Q222 21 251 21Q302 21 329 59Q342 77 347 104T352 209Q352 289 347 316T329 361Q302 397 257 397Z"></path><path id="MJX-18-TEX-N-74" d="M27 422Q80 426 109 478T141 600V615H181V431H316V385H181V241Q182 116 182 100T189 68Q203 29 238 29Q282 29 292 100Q293 108 293 146V181H333V146V134Q333 57 291 17Q264 -10 221 -10Q187 -10 162 2T124 33T105 68T98 100Q97 107 97 248V385H18V422H27Z"></path><path id="MJX-18-TEX-N-6F" d="M28 214Q28 309 93 378T250 448Q340 448 405 380T471 215Q471 120 407 55T250 -10Q153 -10 91 57T28 214ZM250 30Q372 30 372 193V225V250Q372 272 371 288T364 326T348 362T317 390T268 410Q263 411 252 411Q222 411 195 399Q152 377 139 338T126 246V226Q126 130 145 91Q177 30 250 30Z"></path><path id="MJX-18-TEX-N-6C" d="M42 46H56Q95 46 103 60V68Q103 77 103 91T103 124T104 167T104 217T104 272T104 329Q104 366 104 407T104 482T104 542T103 586T103 603Q100 622 89 628T44 637H26V660Q26 683 28 683L38 684Q48 685 67 686T104 688Q121 689 141 690T171 693T182 694H185V379Q185 62 186 60Q190 52 198 49Q219 46 247 46H263V0H255L232 1Q209 2 183 2T145 3T107 3T57 1L34 0H26V46H42Z"></path><path id="MJX-18-TEX-N-6D" d="M41 46H55Q94 46 102 60V68Q102 77 102 91T102 122T103 161T103 203Q103 234 103 269T102 328V351Q99 370 88 376T43 385H25V408Q25 431 27 431L37 432Q47 433 65 434T102 436Q119 437 138 438T167 441T178 442H181V402Q181 364 182 364T187 369T199 384T218 402T247 421T285 437Q305 442 336 442Q351 442 364 440T387 434T406 426T421 417T432 406T441 395T448 384T452 374T455 366L457 361L460 365Q463 369 466 373T475 384T488 397T503 410T523 422T546 432T572 439T603 442Q729 442 740 329Q741 322 741 190V104Q741 66 743 59T754 49Q775 46 803 46H819V0H811L788 1Q764 2 737 2T699 3Q596 3 587 0H579V46H595Q656 46 656 62Q657 64 657 200Q656 335 655 343Q649 371 635 385T611 402T585 404Q540 404 506 370Q479 343 472 315T464 232V168V108Q464 78 465 68T468 55T477 49Q498 46 526 46H542V0H534L510 1Q487 2 460 2T422 3Q319 3 310 0H302V46H318Q379 46 379 62Q380 64 380 200Q379 335 378 343Q372 371 358 385T334 402T308 404Q263 404 229 370Q202 343 195 315T187 232V168V108Q187 78 188 68T191 55T200 49Q221 46 249 46H265V0H257L234 1Q210 2 183 2T145 3Q42 3 33 0H25V46H41Z"></path><path id="MJX-18-TEX-N-70" d="M36 -148H50Q89 -148 97 -134V-126Q97 -119 97 -107T97 -77T98 -38T98 6T98 55T98 106Q98 140 98 177T98 243T98 296T97 335T97 351Q94 370 83 376T38 385H20V408Q20 431 22 431L32 432Q42 433 61 434T98 436Q115 437 135 438T165 441T176 442H179V416L180 390L188 397Q247 441 326 441Q407 441 464 377T522 216Q522 115 457 52T310 -11Q242 -11 190 33L182 40V-45V-101Q182 -128 184 -134T195 -145Q216 -148 244 -148H260V-194H252L228 -193Q205 -192 178 -192T140 -191Q37 -191 28 -194H20V-148H36ZM424 218Q424 292 390 347T305 402Q234 402 182 337V98Q222 26 294 26Q345 26 384 80T424 218Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="mtext"><use data-c="42" xlink:href="#MJX-18-TEX-N-42"></use><use data-c="50" xlink:href="#MJX-18-TEX-N-50" transform="translate(708,0)"></use><use data-c="4D" xlink:href="#MJX-18-TEX-N-4D" transform="translate(1389,0)"></use></g><g data-mml-node="mo" transform="translate(2583.8,0)"><use data-c="3D" xlink:href="#MJX-18-TEX-N-3D"></use></g><g data-mml-node="mfrac" transform="translate(3639.6,0)"><g data-mml-node="mrow" transform="translate(220,676)"><g data-mml-node="mtext"><use data-c="72" xlink:href="#MJX-18-TEX-N-72"></use><use data-c="65" xlink:href="#MJX-18-TEX-N-65" transform="translate(392,0)"></use><use data-c="61" xlink:href="#MJX-18-TEX-N-61" transform="translate(836,0)"></use><use data-c="64" xlink:href="#MJX-18-TEX-N-64" transform="translate(1336,0)"></use><use data-c="73" xlink:href="#MJX-18-TEX-N-73" transform="translate(1892,0)"></use><use data-c="20" xlink:href="#MJX-18-TEX-N-20" transform="translate(2286,0)"></use><use data-c="69" xlink:href="#MJX-18-TEX-N-69" transform="translate(2536,0)"></use><use data-c="6E" xlink:href="#MJX-18-TEX-N-6E" transform="translate(2814,0)"></use><use data-c="20" xlink:href="#MJX-18-TEX-N-20" transform="translate(3370,0)"></use><use data-c="61" xlink:href="#MJX-18-TEX-N-61" transform="translate(3620,0)"></use><use data-c="20" xlink:href="#MJX-18-TEX-N-20" transform="translate(4120,0)"></use><use data-c="62" xlink:href="#MJX-18-TEX-N-62" transform="translate(4370,0)"></use><use data-c="69" xlink:href="#MJX-18-TEX-N-69" transform="translate(4926,0)"></use><use data-c="6E" xlink:href="#MJX-18-TEX-N-6E" transform="translate(5204,0)"></use></g><g data-mml-node="mo" transform="translate(5982.2,0)"><use data-c="D7" xlink:href="#MJX-18-TEX-N-D7"></use></g><g data-mml-node="msup" transform="translate(6982.4,0)"><g data-mml-node="mn"><use data-c="31" xlink:href="#MJX-18-TEX-N-31"></use><use data-c="30" xlink:href="#MJX-18-TEX-N-30" transform="translate(500,0)"></use></g><g data-mml-node="mn" transform="translate(1033,393.1) scale(0.707)"><use data-c="36" xlink:href="#MJX-18-TEX-N-36"></use></g></g></g><g data-mml-node="mtext" transform="translate(286,-686)"><use data-c="74" xlink:href="#MJX-18-TEX-N-74"></use><use data-c="6F" xlink:href="#MJX-18-TEX-N-6F" transform="translate(389,0)"></use><use data-c="74" xlink:href="#MJX-18-TEX-N-74" transform="translate(889,0)"></use><use data-c="61" xlink:href="#MJX-18-TEX-N-61" transform="translate(1278,0)"></use><use data-c="6C" xlink:href="#MJX-18-TEX-N-6C" transform="translate(1778,0)"></use><use data-c="20" xlink:href="#MJX-18-TEX-N-20" transform="translate(2056,0)"></use><use data-c="6D" xlink:href="#MJX-18-TEX-N-6D" transform="translate(2306,0)"></use><use data-c="61" xlink:href="#MJX-18-TEX-N-61" transform="translate(3139,0)"></use><use data-c="70" xlink:href="#MJX-18-TEX-N-70" transform="translate(3639,0)"></use><use data-c="70" xlink:href="#MJX-18-TEX-N-70" transform="translate(4195,0)"></use><use data-c="65" xlink:href="#MJX-18-TEX-N-65" transform="translate(4751,0)"></use><use data-c="64" xlink:href="#MJX-18-TEX-N-64" transform="translate(5195,0)"></use><use data-c="20" xlink:href="#MJX-18-TEX-N-20" transform="translate(5751,0)"></use><use data-c="72" xlink:href="#MJX-18-TEX-N-72" transform="translate(6001,0)"></use><use data-c="65" xlink:href="#MJX-18-TEX-N-65" transform="translate(6393,0)"></use><use data-c="61" xlink:href="#MJX-18-TEX-N-61" transform="translate(6837,0)"></use><use data-c="64" xlink:href="#MJX-18-TEX-N-64" transform="translate(7337,0)"></use><use data-c="73" xlink:href="#MJX-18-TEX-N-73" transform="translate(7893,0)"></use></g><rect width="8619" height="60" x="120" y="220"></rect></g></g></g></svg><mjx-assistive-mml unselectable="on" display="block"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><mtext>BPM</mtext><mo>=</mo><mfrac><mrow><mtext>reads in a bin</mtext><mo>×</mo><msup><mn>10</mn><mn>6</mn></msup></mrow><mtext>total mapped reads</mtext></mfrac></math></mjx-assistive-mml></mjx-container></div></div><p>&nbsp;</p><hr /><h2 id='rpgc'><span>RPGC</span></h2><p><strong><span>RPGC (Reads Per Genomic Content)</span></strong></p><ul><li><p><strong><span>Definition</span></strong><span>: Normalizes read counts based on the </span><strong><span>effective genome size</span></strong><span>, usually to adjust for mappability in the genome. It’s commonly used in </span><strong><span>ChIP-seq</span></strong><span> and </span><strong><span>ATAC-seq</span></strong><span>.</span></p><ul><li><p><strong><span>Effective Genome Size</span></strong><span>:This refers to the mappable portion of the genome. For example, the human genome is about 3 billion base pairs, but due to repetitive sequences, only a portion of it (~2.7 billion base pairs) is actually usable for most short-read alignments.</span></p></li><li><p><span>By using the effective genome size, you account only for the regions where reads can be uniquely mapped, which makes the normalization more accurate, especially for data like ChIP-seq or ATAC-seq.</span></p></li></ul></li><li><p><strong><span>Purpose</span></strong><span>: Useful for adjusting read counts by the mappable portion of the genome rather than total read count.</span></p></li></ul><div contenteditable="false" spellcheck="false" class="mathjax-block md-end-block md-math-block md-rawblock" id="mathjax-n3373" cid="n3373" mdtype="math_block" data-math-tag-before="0" data-math-tag-after="0" data-math-labels="[]"><div class="md-math-container"><mjx-container class="MathJax" jax="SVG" display="true" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="49.502ex" height="5.143ex" role="img" focusable="false" viewBox="0 -1381 21880 2273" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -2.018ex;"><defs><path id="MJX-19-TEX-N-52" d="M130 622Q123 629 119 631T103 634T60 637H27V683H202H236H300Q376 683 417 677T500 648Q595 600 609 517Q610 512 610 501Q610 468 594 439T556 392T511 361T472 343L456 338Q459 335 467 332Q497 316 516 298T545 254T559 211T568 155T578 94Q588 46 602 31T640 16H645Q660 16 674 32T692 87Q692 98 696 101T712 105T728 103T732 90Q732 59 716 27T672 -16Q656 -22 630 -22Q481 -16 458 90Q456 101 456 163T449 246Q430 304 373 320L363 322L297 323H231V192L232 61Q238 51 249 49T301 46H334V0H323Q302 3 181 3Q59 3 38 0H27V46H60Q102 47 111 49T130 61V622ZM491 499V509Q491 527 490 539T481 570T462 601T424 623T362 636Q360 636 340 636T304 637H283Q238 637 234 628Q231 624 231 492V360H289Q390 360 434 378T489 456Q491 467 491 499Z"></path><path id="MJX-19-TEX-N-50" d="M130 622Q123 629 119 631T103 634T60 637H27V683H214Q237 683 276 683T331 684Q419 684 471 671T567 616Q624 563 624 489Q624 421 573 372T451 307Q429 302 328 301H234V181Q234 62 237 58Q245 47 304 46H337V0H326Q305 3 182 3Q47 3 38 0H27V46H60Q102 47 111 49T130 61V622ZM507 488Q507 514 506 528T500 564T483 597T450 620T397 635Q385 637 307 637H286Q237 637 234 628Q231 624 231 483V342H302H339Q390 342 423 349T481 382Q507 411 507 488Z"></path><path id="MJX-19-TEX-N-43" d="M56 342Q56 428 89 500T174 615T283 681T391 705Q394 705 400 705T408 704Q499 704 569 636L582 624L612 663Q639 700 643 704Q644 704 647 704T653 705H657Q660 705 666 699V419L660 413H626Q620 419 619 430Q610 512 571 572T476 651Q457 658 426 658Q322 658 252 588Q173 509 173 342Q173 221 211 151Q232 111 263 84T328 45T384 29T428 24Q517 24 571 93T626 244Q626 251 632 257H660L666 251V236Q661 133 590 56T403 -21Q262 -21 159 83T56 342Z"></path><path id="MJX-19-TEX-N-47" d="M56 342Q56 428 89 500T174 615T283 681T391 705Q394 705 400 705T408 704Q499 704 569 636L582 624L612 663Q639 700 643 704Q644 704 647 704T653 705H657Q660 705 666 699V419L660 413H626Q620 419 619 430Q610 512 571 572T476 651Q457 658 426 658Q401 658 376 654T316 633T254 592T205 519T177 411Q173 369 173 335Q173 259 192 201T238 111T302 58T370 31T431 24Q478 24 513 45T559 100Q562 110 562 160V212Q561 213 557 216T551 220T542 223T526 225T502 226T463 227H437V273H449L609 270Q715 270 727 273H735V227H721Q674 227 668 215Q666 211 666 108V6Q660 0 657 0Q653 0 639 10Q617 25 600 42L587 54Q571 27 524 3T406 -22Q317 -22 238 22T108 151T56 342Z"></path><path id="MJX-19-TEX-N-3D" d="M56 347Q56 360 70 367H707Q722 359 722 347Q722 336 708 328L390 327H72Q56 332 56 347ZM56 153Q56 168 72 173H708Q722 163 722 153Q722 140 707 133H70Q56 140 56 153Z"></path><path id="MJX-19-TEX-N-72" d="M36 46H50Q89 46 97 60V68Q97 77 97 91T98 122T98 161T98 203Q98 234 98 269T98 328L97 351Q94 370 83 376T38 385H20V408Q20 431 22 431L32 432Q42 433 60 434T96 436Q112 437 131 438T160 441T171 442H174V373Q213 441 271 441H277Q322 441 343 419T364 373Q364 352 351 337T313 322Q288 322 276 338T263 372Q263 381 265 388T270 400T273 405Q271 407 250 401Q234 393 226 386Q179 341 179 207V154Q179 141 179 127T179 101T180 81T180 66V61Q181 59 183 57T188 54T193 51T200 49T207 48T216 47T225 47T235 46T245 46H276V0H267Q249 3 140 3Q37 3 28 0H20V46H36Z"></path><path id="MJX-19-TEX-N-65" d="M28 218Q28 273 48 318T98 391T163 433T229 448Q282 448 320 430T378 380T406 316T415 245Q415 238 408 231H126V216Q126 68 226 36Q246 30 270 30Q312 30 342 62Q359 79 369 104L379 128Q382 131 395 131H398Q415 131 415 121Q415 117 412 108Q393 53 349 21T250 -11Q155 -11 92 58T28 218ZM333 275Q322 403 238 411H236Q228 411 220 410T195 402T166 381T143 340T127 274V267H333V275Z"></path><path id="MJX-19-TEX-N-61" d="M137 305T115 305T78 320T63 359Q63 394 97 421T218 448Q291 448 336 416T396 340Q401 326 401 309T402 194V124Q402 76 407 58T428 40Q443 40 448 56T453 109V145H493V106Q492 66 490 59Q481 29 455 12T400 -6T353 12T329 54V58L327 55Q325 52 322 49T314 40T302 29T287 17T269 6T247 -2T221 -8T190 -11Q130 -11 82 20T34 107Q34 128 41 147T68 188T116 225T194 253T304 268H318V290Q318 324 312 340Q290 411 215 411Q197 411 181 410T156 406T148 403Q170 388 170 359Q170 334 154 320ZM126 106Q126 75 150 51T209 26Q247 26 276 49T315 109Q317 116 318 175Q318 233 317 233Q309 233 296 232T251 223T193 203T147 166T126 106Z"></path><path id="MJX-19-TEX-N-64" d="M376 495Q376 511 376 535T377 568Q377 613 367 624T316 637H298V660Q298 683 300 683L310 684Q320 685 339 686T376 688Q393 689 413 690T443 693T454 694H457V390Q457 84 458 81Q461 61 472 55T517 46H535V0Q533 0 459 -5T380 -11H373V44L365 37Q307 -11 235 -11Q158 -11 96 50T34 215Q34 315 97 378T244 442Q319 442 376 393V495ZM373 342Q328 405 260 405Q211 405 173 369Q146 341 139 305T131 211Q131 155 138 120T173 59Q203 26 251 26Q322 26 373 103V342Z"></path><path id="MJX-19-TEX-N-73" d="M295 316Q295 356 268 385T190 414Q154 414 128 401Q98 382 98 349Q97 344 98 336T114 312T157 287Q175 282 201 278T245 269T277 256Q294 248 310 236T342 195T359 133Q359 71 321 31T198 -10H190Q138 -10 94 26L86 19L77 10Q71 4 65 -1L54 -11H46H42Q39 -11 33 -5V74V132Q33 153 35 157T45 162H54Q66 162 70 158T75 146T82 119T101 77Q136 26 198 26Q295 26 295 104Q295 133 277 151Q257 175 194 187T111 210Q75 227 54 256T33 318Q33 357 50 384T93 424T143 442T187 447H198Q238 447 268 432L283 424L292 431Q302 440 314 448H322H326Q329 448 335 442V310L329 304H301Q295 310 295 316Z"></path><path id="MJX-19-TEX-N-20" d=""></path><path id="MJX-19-TEX-N-69" d="M69 609Q69 637 87 653T131 669Q154 667 171 652T188 609Q188 579 171 564T129 549Q104 549 87 564T69 609ZM247 0Q232 3 143 3Q132 3 106 3T56 1L34 0H26V46H42Q70 46 91 49Q100 53 102 60T104 102V205V293Q104 345 102 359T88 378Q74 385 41 385H30V408Q30 431 32 431L42 432Q52 433 70 434T106 436Q123 437 142 438T171 441T182 442H185V62Q190 52 197 50T232 46H255V0H247Z"></path><path id="MJX-19-TEX-N-6E" d="M41 46H55Q94 46 102 60V68Q102 77 102 91T102 122T103 161T103 203Q103 234 103 269T102 328V351Q99 370 88 376T43 385H25V408Q25 431 27 431L37 432Q47 433 65 434T102 436Q119 437 138 438T167 441T178 442H181V402Q181 364 182 364T187 369T199 384T218 402T247 421T285 437Q305 442 336 442Q450 438 463 329Q464 322 464 190V104Q464 66 466 59T477 49Q498 46 526 46H542V0H534L510 1Q487 2 460 2T422 3Q319 3 310 0H302V46H318Q379 46 379 62Q380 64 380 200Q379 335 378 343Q372 371 358 385T334 402T308 404Q263 404 229 370Q202 343 195 315T187 232V168V108Q187 78 188 68T191 55T200 49Q221 46 249 46H265V0H257L234 1Q210 2 183 2T145 3Q42 3 33 0H25V46H41Z"></path><path id="MJX-19-TEX-N-67" d="M329 409Q373 453 429 453Q459 453 472 434T485 396Q485 382 476 371T449 360Q416 360 412 390Q410 404 415 411Q415 412 416 414V415Q388 412 363 393Q355 388 355 386Q355 385 359 381T368 369T379 351T388 325T392 292Q392 230 343 187T222 143Q172 143 123 171Q112 153 112 133Q112 98 138 81Q147 75 155 75T227 73Q311 72 335 67Q396 58 431 26Q470 -13 470 -72Q470 -139 392 -175Q332 -206 250 -206Q167 -206 107 -175Q29 -140 29 -75Q29 -39 50 -15T92 18L103 24Q67 55 67 108Q67 155 96 193Q52 237 52 292Q52 355 102 398T223 442Q274 442 318 416L329 409ZM299 343Q294 371 273 387T221 404Q192 404 171 388T145 343Q142 326 142 292Q142 248 149 227T179 192Q196 182 222 182Q244 182 260 189T283 207T294 227T299 242Q302 258 302 292T299 343ZM403 -75Q403 -50 389 -34T348 -11T299 -2T245 0H218Q151 0 138 -6Q118 -15 107 -34T95 -74Q95 -84 101 -97T122 -127T170 -155T250 -167Q319 -167 361 -139T403 -75Z"></path><path id="MJX-19-TEX-N-6F" d="M28 214Q28 309 93 378T250 448Q340 448 405 380T471 215Q471 120 407 55T250 -10Q153 -10 91 57T28 214ZM250 30Q372 30 372 193V225V250Q372 272 371 288T364 326T348 362T317 390T268 410Q263 411 252 411Q222 411 195 399Q152 377 139 338T126 246V226Q126 130 145 91Q177 30 250 30Z"></path><path id="MJX-19-TEX-N-D7" d="M630 29Q630 9 609 9Q604 9 587 25T493 118L389 222L284 117Q178 13 175 11Q171 9 168 9Q160 9 154 15T147 29Q147 36 161 51T255 146L359 250L255 354Q174 435 161 449T147 471Q147 480 153 485T168 490Q173 490 175 489Q178 487 284 383L389 278L493 382Q570 459 587 475T609 491Q630 491 630 471Q630 464 620 453T522 355L418 250L522 145Q606 61 618 48T630 29Z"></path><path id="MJX-19-TEX-N-66" d="M273 0Q255 3 146 3Q43 3 34 0H26V46H42Q70 46 91 49Q99 52 103 60Q104 62 104 224V385H33V431H104V497L105 564L107 574Q126 639 171 668T266 704Q267 704 275 704T289 705Q330 702 351 679T372 627Q372 604 358 590T321 576T284 590T270 627Q270 647 288 667H284Q280 668 273 668Q245 668 223 647T189 592Q183 572 182 497V431H293V385H185V225Q185 63 186 61T189 57T194 54T199 51T206 49T213 48T222 47T231 47T241 46T251 46H282V0H273Z"></path><path id="MJX-19-TEX-N-63" d="M370 305T349 305T313 320T297 358Q297 381 312 396Q317 401 317 402T307 404Q281 408 258 408Q209 408 178 376Q131 329 131 219Q131 137 162 90Q203 29 272 29Q313 29 338 55T374 117Q376 125 379 127T395 129H409Q415 123 415 120Q415 116 411 104T395 71T366 33T318 2T249 -11Q163 -11 99 53T34 214Q34 318 99 383T250 448T370 421T404 357Q404 334 387 320Z"></path><path id="MJX-19-TEX-N-74" d="M27 422Q80 426 109 478T141 600V615H181V431H316V385H181V241Q182 116 182 100T189 68Q203 29 238 29Q282 29 292 100Q293 108 293 146V181H333V146V134Q333 57 291 17Q264 -10 221 -10Q187 -10 162 2T124 33T105 68T98 100Q97 107 97 248V385H18V422H27Z"></path><path id="MJX-19-TEX-N-76" d="M338 431Q344 429 422 429Q479 429 503 431H508V385H497Q439 381 423 345Q421 341 356 172T288 -2Q283 -11 263 -11Q244 -11 239 -2Q99 359 98 364Q93 378 82 381T43 385H19V431H25L33 430Q41 430 53 430T79 430T104 429T122 428Q217 428 232 431H240V385H226Q187 384 184 370Q184 366 235 234L286 102L377 341V349Q377 363 367 372T349 383T335 385H331V431H338Z"></path><path id="MJX-19-TEX-N-6D" d="M41 46H55Q94 46 102 60V68Q102 77 102 91T102 122T103 161T103 203Q103 234 103 269T102 328V351Q99 370 88 376T43 385H25V408Q25 431 27 431L37 432Q47 433 65 434T102 436Q119 437 138 438T167 441T178 442H181V402Q181 364 182 364T187 369T199 384T218 402T247 421T285 437Q305 442 336 442Q351 442 364 440T387 434T406 426T421 417T432 406T441 395T448 384T452 374T455 366L457 361L460 365Q463 369 466 373T475 384T488 397T503 410T523 422T546 432T572 439T603 442Q729 442 740 329Q741 322 741 190V104Q741 66 743 59T754 49Q775 46 803 46H819V0H811L788 1Q764 2 737 2T699 3Q596 3 587 0H579V46H595Q656 46 656 62Q657 64 657 200Q656 335 655 343Q649 371 635 385T611 402T585 404Q540 404 506 370Q479 343 472 315T464 232V168V108Q464 78 465 68T468 55T477 49Q498 46 526 46H542V0H534L510 1Q487 2 460 2T422 3Q319 3 310 0H302V46H318Q379 46 379 62Q380 64 380 200Q379 335 378 343Q372 371 358 385T334 402T308 404Q263 404 229 370Q202 343 195 315T187 232V168V108Q187 78 188 68T191 55T200 49Q221 46 249 46H265V0H257L234 1Q210 2 183 2T145 3Q42 3 33 0H25V46H41Z"></path><path id="MJX-19-TEX-N-7A" d="M42 263Q44 270 48 345T53 423V431H393Q399 425 399 415Q399 403 398 402L381 378Q364 355 331 309T265 220L134 41L182 40H206Q254 40 283 46T331 77Q352 105 359 185L361 201Q361 202 381 202H401V196Q401 195 393 103T384 6V0H209L34 1L31 3Q28 8 28 17Q28 30 29 31T160 210T294 394H236Q169 393 152 388Q127 382 113 367Q89 344 82 264V255H42V263Z"></path><path id="MJX-19-TEX-N-6C" d="M42 46H56Q95 46 103 60V68Q103 77 103 91T103 124T104 167T104 217T104 272T104 329Q104 366 104 407T104 482T104 542T103 586T103 603Q100 622 89 628T44 637H26V660Q26 683 28 683L38 684Q48 685 67 686T104 688Q121 689 141 690T171 693T182 694H185V379Q185 62 186 60Q190 52 198 49Q219 46 247 46H263V0H255L232 1Q209 2 183 2T145 3T107 3T57 1L34 0H26V46H42Z"></path><path id="MJX-19-TEX-N-70" d="M36 -148H50Q89 -148 97 -134V-126Q97 -119 97 -107T97 -77T98 -38T98 6T98 55T98 106Q98 140 98 177T98 243T98 296T97 335T97 351Q94 370 83 376T38 385H20V408Q20 431 22 431L32 432Q42 433 61 434T98 436Q115 437 135 438T165 441T176 442H179V416L180 390L188 397Q247 441 326 441Q407 441 464 377T522 216Q522 115 457 52T310 -11Q242 -11 190 33L182 40V-45V-101Q182 -128 184 -134T195 -145Q216 -148 244 -148H260V-194H252L228 -193Q205 -192 178 -192T140 -191Q37 -191 28 -194H20V-148H36ZM424 218Q424 292 390 347T305 402Q234 402 182 337V98Q222 26 294 26Q345 26 384 80T424 218Z"></path><path id="MJX-19-TEX-N-68" d="M41 46H55Q94 46 102 60V68Q102 77 102 91T102 124T102 167T103 217T103 272T103 329Q103 366 103 407T103 482T102 542T102 586T102 603Q99 622 88 628T43 637H25V660Q25 683 27 683L37 684Q47 685 66 686T103 688Q120 689 140 690T170 693T181 694H184V367Q244 442 328 442Q451 442 463 329Q464 322 464 190V104Q464 66 466 59T477 49Q498 46 526 46H542V0H534L510 1Q487 2 460 2T422 3Q319 3 310 0H302V46H318Q379 46 379 62Q380 64 380 200Q379 335 378 343Q372 371 358 385T334 402T308 404Q263 404 229 370Q202 343 195 315T187 232V168V108Q187 78 188 68T191 55T200 49Q221 46 249 46H265V0H257L234 1Q210 2 183 2T145 3Q42 3 33 0H25V46H41Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="mtext"><use data-c="52" xlink:href="#MJX-19-TEX-N-52"></use><use data-c="50" xlink:href="#MJX-19-TEX-N-50" transform="translate(736,0)"></use><use data-c="43" xlink:href="#MJX-19-TEX-N-43" transform="translate(1417,0)"></use><use data-c="47" xlink:href="#MJX-19-TEX-N-47" transform="translate(2139,0)"></use></g><g data-mml-node="mo" transform="translate(3201.8,0)"><use data-c="3D" xlink:href="#MJX-19-TEX-N-3D"></use></g><g data-mml-node="mfrac" transform="translate(4257.6,0)"><g data-mml-node="mrow" transform="translate(220,676)"><g data-mml-node="mtext"><use data-c="72" xlink:href="#MJX-19-TEX-N-72"></use><use data-c="65" xlink:href="#MJX-19-TEX-N-65" transform="translate(392,0)"></use><use data-c="61" xlink:href="#MJX-19-TEX-N-61" transform="translate(836,0)"></use><use data-c="64" xlink:href="#MJX-19-TEX-N-64" transform="translate(1336,0)"></use><use data-c="73" xlink:href="#MJX-19-TEX-N-73" transform="translate(1892,0)"></use><use data-c="20" xlink:href="#MJX-19-TEX-N-20" transform="translate(2286,0)"></use><use data-c="69" xlink:href="#MJX-19-TEX-N-69" transform="translate(2536,0)"></use><use data-c="6E" xlink:href="#MJX-19-TEX-N-6E" transform="translate(2814,0)"></use><use data-c="20" xlink:href="#MJX-19-TEX-N-20" transform="translate(3370,0)"></use><use data-c="61" xlink:href="#MJX-19-TEX-N-61" transform="translate(3620,0)"></use><use data-c="20" xlink:href="#MJX-19-TEX-N-20" transform="translate(4120,0)"></use><use data-c="72" xlink:href="#MJX-19-TEX-N-72" transform="translate(4370,0)"></use><use data-c="65" xlink:href="#MJX-19-TEX-N-65" transform="translate(4762,0)"></use><use data-c="67" xlink:href="#MJX-19-TEX-N-67" transform="translate(5206,0)"></use><use data-c="69" xlink:href="#MJX-19-TEX-N-69" transform="translate(5706,0)"></use><use data-c="6F" xlink:href="#MJX-19-TEX-N-6F" transform="translate(5984,0)"></use><use data-c="6E" xlink:href="#MJX-19-TEX-N-6E" transform="translate(6484,0)"></use></g><g data-mml-node="mo" transform="translate(7262.2,0)"><use data-c="D7" xlink:href="#MJX-19-TEX-N-D7"></use></g><g data-mml-node="mtext" transform="translate(8262.4,0)"><use data-c="65" xlink:href="#MJX-19-TEX-N-65"></use><use data-c="66" xlink:href="#MJX-19-TEX-N-66" transform="translate(444,0)"></use><use data-c="66" xlink:href="#MJX-19-TEX-N-66" transform="translate(750,0)"></use><use data-c="65" xlink:href="#MJX-19-TEX-N-65" transform="translate(1056,0)"></use><use data-c="63" xlink:href="#MJX-19-TEX-N-63" transform="translate(1500,0)"></use><use data-c="74" xlink:href="#MJX-19-TEX-N-74" transform="translate(1944,0)"></use><use data-c="69" xlink:href="#MJX-19-TEX-N-69" transform="translate(2333,0)"></use><use data-c="76" xlink:href="#MJX-19-TEX-N-76" transform="translate(2611,0)"></use><use data-c="65" xlink:href="#MJX-19-TEX-N-65" transform="translate(3139,0)"></use><use data-c="20" xlink:href="#MJX-19-TEX-N-20" transform="translate(3583,0)"></use><use data-c="67" xlink:href="#MJX-19-TEX-N-67" transform="translate(3833,0)"></use><use data-c="65" xlink:href="#MJX-19-TEX-N-65" transform="translate(4333,0)"></use><use data-c="6E" xlink:href="#MJX-19-TEX-N-6E" transform="translate(4777,0)"></use><use data-c="6F" xlink:href="#MJX-19-TEX-N-6F" transform="translate(5333,0)"></use><use data-c="6D" xlink:href="#MJX-19-TEX-N-6D" transform="translate(5833,0)"></use><use data-c="65" xlink:href="#MJX-19-TEX-N-65" transform="translate(6666,0)"></use><use data-c="20" xlink:href="#MJX-19-TEX-N-20" transform="translate(7110,0)"></use><use data-c="73" xlink:href="#MJX-19-TEX-N-73" transform="translate(7360,0)"></use><use data-c="69" xlink:href="#MJX-19-TEX-N-69" transform="translate(7754,0)"></use><use data-c="7A" xlink:href="#MJX-19-TEX-N-7A" transform="translate(8032,0)"></use><use data-c="65" xlink:href="#MJX-19-TEX-N-65" transform="translate(8476,0)"></use></g></g><g data-mml-node="mrow" transform="translate(1235,-686)"><g data-mml-node="mtext"><use data-c="74" xlink:href="#MJX-19-TEX-N-74"></use><use data-c="6F" xlink:href="#MJX-19-TEX-N-6F" transform="translate(389,0)"></use><use data-c="74" xlink:href="#MJX-19-TEX-N-74" transform="translate(889,0)"></use><use data-c="61" xlink:href="#MJX-19-TEX-N-61" transform="translate(1278,0)"></use><use data-c="6C" xlink:href="#MJX-19-TEX-N-6C" transform="translate(1778,0)"></use><use data-c="20" xlink:href="#MJX-19-TEX-N-20" transform="translate(2056,0)"></use><use data-c="6D" xlink:href="#MJX-19-TEX-N-6D" transform="translate(2306,0)"></use><use data-c="61" xlink:href="#MJX-19-TEX-N-61" transform="translate(3139,0)"></use><use data-c="70" xlink:href="#MJX-19-TEX-N-70" transform="translate(3639,0)"></use><use data-c="70" xlink:href="#MJX-19-TEX-N-70" transform="translate(4195,0)"></use><use data-c="65" xlink:href="#MJX-19-TEX-N-65" transform="translate(4751,0)"></use><use data-c="64" xlink:href="#MJX-19-TEX-N-64" transform="translate(5195,0)"></use><use data-c="20" xlink:href="#MJX-19-TEX-N-20" transform="translate(5751,0)"></use><use data-c="72" xlink:href="#MJX-19-TEX-N-72" transform="translate(6001,0)"></use><use data-c="65" xlink:href="#MJX-19-TEX-N-65" transform="translate(6393,0)"></use><use data-c="61" xlink:href="#MJX-19-TEX-N-61" transform="translate(6837,0)"></use><use data-c="64" xlink:href="#MJX-19-TEX-N-64" transform="translate(7337,0)"></use><use data-c="73" xlink:href="#MJX-19-TEX-N-73" transform="translate(7893,0)"></use></g><g data-mml-node="mo" transform="translate(8509.2,0)"><use data-c="D7" xlink:href="#MJX-19-TEX-N-D7"></use></g><g data-mml-node="mtext" transform="translate(9509.4,0)"><use data-c="72" xlink:href="#MJX-19-TEX-N-72"></use><use data-c="65" xlink:href="#MJX-19-TEX-N-65" transform="translate(392,0)"></use><use data-c="67" xlink:href="#MJX-19-TEX-N-67" transform="translate(836,0)"></use><use data-c="69" xlink:href="#MJX-19-TEX-N-69" transform="translate(1336,0)"></use><use data-c="6F" xlink:href="#MJX-19-TEX-N-6F" transform="translate(1614,0)"></use><use data-c="6E" xlink:href="#MJX-19-TEX-N-6E" transform="translate(2114,0)"></use><use data-c="20" xlink:href="#MJX-19-TEX-N-20" transform="translate(2670,0)"></use><use data-c="6C" xlink:href="#MJX-19-TEX-N-6C" transform="translate(2920,0)"></use><use data-c="65" xlink:href="#MJX-19-TEX-N-65" transform="translate(3198,0)"></use><use data-c="6E" xlink:href="#MJX-19-TEX-N-6E" transform="translate(3642,0)"></use><use data-c="67" xlink:href="#MJX-19-TEX-N-67" transform="translate(4198,0)"></use><use data-c="74" xlink:href="#MJX-19-TEX-N-74" transform="translate(4698,0)"></use><use data-c="68" xlink:href="#MJX-19-TEX-N-68" transform="translate(5087,0)"></use></g></g><rect width="17382.4" height="60" x="120" y="220"></rect></g></g></g></svg><mjx-assistive-mml unselectable="on" display="block"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><mtext>RPCG</mtext><mo>=</mo><mfrac><mrow><mtext>reads in a region</mtext><mo>×</mo><mtext>effective genome size</mtext></mrow><mrow><mtext>total mapped reads</mtext><mo>×</mo><mtext>region length</mtext></mrow></mfrac></math></mjx-assistive-mml></mjx-container></div></div><p>&nbsp;</p><hr /><h2 id='chipseqspikeinfree'><span>ChIPseqSpikeInFree</span></h2><h3 id='chipseqspikeinfree-perface'><span>ChIPseqSpikeInFree perface</span></h3><ul><li><p><a href='https://github.com/stjude/ChIPseqSpikeInFree'><span>ChIPseqSpikeInFree: A Spike-in Free ChIP-Seq Normalization Approach for Detecting Global Changes in Histone Modifications</span></a></p></li><li><p><strong><span>Traditional reads per million (RPM)</span></strong><span> normalization method is </span><strong><span>inappropriate</span></strong><span> for the evaluation of ChIP-seq data when the </span><strong><span>treatment or mutation has the global effect</span></strong></p><div contenteditable="false" spellcheck="false" class="mathjax-block md-end-block md-math-block md-rawblock" id="mathjax-n3383" cid="n3383" mdtype="math_block" data-math-tag-before="0" data-math-tag-after="0" data-math-labels="[]"><div class="md-math-container"><mjx-container class="MathJax" jax="SVG" display="true" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="44.628ex" height="5.115ex" role="img" focusable="false" viewBox="0 -1381 19725.6 2261" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -1.991ex;"><defs><path id="MJX-20-TEX-N-52" d="M130 622Q123 629 119 631T103 634T60 637H27V683H202H236H300Q376 683 417 677T500 648Q595 600 609 517Q610 512 610 501Q610 468 594 439T556 392T511 361T472 343L456 338Q459 335 467 332Q497 316 516 298T545 254T559 211T568 155T578 94Q588 46 602 31T640 16H645Q660 16 674 32T692 87Q692 98 696 101T712 105T728 103T732 90Q732 59 716 27T672 -16Q656 -22 630 -22Q481 -16 458 90Q456 101 456 163T449 246Q430 304 373 320L363 322L297 323H231V192L232 61Q238 51 249 49T301 46H334V0H323Q302 3 181 3Q59 3 38 0H27V46H60Q102 47 111 49T130 61V622ZM491 499V509Q491 527 490 539T481 570T462 601T424 623T362 636Q360 636 340 636T304 637H283Q238 637 234 628Q231 624 231 492V360H289Q390 360 434 378T489 456Q491 467 491 499Z"></path><path id="MJX-20-TEX-N-50" d="M130 622Q123 629 119 631T103 634T60 637H27V683H214Q237 683 276 683T331 684Q419 684 471 671T567 616Q624 563 624 489Q624 421 573 372T451 307Q429 302 328 301H234V181Q234 62 237 58Q245 47 304 46H337V0H326Q305 3 182 3Q47 3 38 0H27V46H60Q102 47 111 49T130 61V622ZM507 488Q507 514 506 528T500 564T483 597T450 620T397 635Q385 637 307 637H286Q237 637 234 628Q231 624 231 483V342H302H339Q390 342 423 349T481 382Q507 411 507 488Z"></path><path id="MJX-20-TEX-N-4D" d="M132 622Q125 629 121 631T105 634T62 637H29V683H135Q221 683 232 682T249 675Q250 674 354 398L458 124L562 398Q666 674 668 675Q671 681 683 682T781 683H887V637H854Q814 636 803 634T785 622V61Q791 51 802 49T854 46H887V0H876Q855 3 736 3Q605 3 596 0H585V46H618Q660 47 669 49T688 61V347Q688 424 688 461T688 546T688 613L687 632Q454 14 450 7Q446 1 430 1T410 7Q409 9 292 316L176 624V606Q175 588 175 543T175 463T175 356L176 86Q187 50 261 46H278V0H269Q254 3 154 3Q52 3 37 0H29V46H46Q78 48 98 56T122 69T132 86V622Z"></path><path id="MJX-20-TEX-N-3D" d="M56 347Q56 360 70 367H707Q722 359 722 347Q722 336 708 328L390 327H72Q56 332 56 347ZM56 153Q56 168 72 173H708Q722 163 722 153Q722 140 707 133H70Q56 140 56 153Z"></path><path id="MJX-20-TEX-N-61" d="M137 305T115 305T78 320T63 359Q63 394 97 421T218 448Q291 448 336 416T396 340Q401 326 401 309T402 194V124Q402 76 407 58T428 40Q443 40 448 56T453 109V145H493V106Q492 66 490 59Q481 29 455 12T400 -6T353 12T329 54V58L327 55Q325 52 322 49T314 40T302 29T287 17T269 6T247 -2T221 -8T190 -11Q130 -11 82 20T34 107Q34 128 41 147T68 188T116 225T194 253T304 268H318V290Q318 324 312 340Q290 411 215 411Q197 411 181 410T156 406T148 403Q170 388 170 359Q170 334 154 320ZM126 106Q126 75 150 51T209 26Q247 26 276 49T315 109Q317 116 318 175Q318 233 317 233Q309 233 296 232T251 223T193 203T147 166T126 106Z"></path><path id="MJX-20-TEX-N-77" d="M90 368Q84 378 76 380T40 385H18V431H24L43 430Q62 430 84 429T116 428Q206 428 221 431H229V385H215Q177 383 177 368Q177 367 221 239L265 113L339 328L333 345Q323 374 316 379Q308 384 278 385H258V431H264Q270 428 348 428Q439 428 454 431H461V385H452Q404 385 404 369Q404 366 418 324T449 234T481 143L496 100L537 219Q579 341 579 347Q579 363 564 373T530 385H522V431H529Q541 428 624 428Q692 428 698 431H703V385H697Q696 385 691 385T682 384Q635 377 619 334L559 161Q546 124 528 71Q508 12 503 1T487 -11H479Q460 -11 456 -4Q455 -3 407 133L361 267Q359 263 266 -4Q261 -11 243 -11H238Q225 -11 220 -3L90 368Z"></path><path id="MJX-20-TEX-N-20" d=""></path><path id="MJX-20-TEX-N-72" d="M36 46H50Q89 46 97 60V68Q97 77 97 91T98 122T98 161T98 203Q98 234 98 269T98 328L97 351Q94 370 83 376T38 385H20V408Q20 431 22 431L32 432Q42 433 60 434T96 436Q112 437 131 438T160 441T171 442H174V373Q213 441 271 441H277Q322 441 343 419T364 373Q364 352 351 337T313 322Q288 322 276 338T263 372Q263 381 265 388T270 400T273 405Q271 407 250 401Q234 393 226 386Q179 341 179 207V154Q179 141 179 127T179 101T180 81T180 66V61Q181 59 183 57T188 54T193 51T200 49T207 48T216 47T225 47T235 46T245 46H276V0H267Q249 3 140 3Q37 3 28 0H20V46H36Z"></path><path id="MJX-20-TEX-N-65" d="M28 218Q28 273 48 318T98 391T163 433T229 448Q282 448 320 430T378 380T406 316T415 245Q415 238 408 231H126V216Q126 68 226 36Q246 30 270 30Q312 30 342 62Q359 79 369 104L379 128Q382 131 395 131H398Q415 131 415 121Q415 117 412 108Q393 53 349 21T250 -11Q155 -11 92 58T28 218ZM333 275Q322 403 238 411H236Q228 411 220 410T195 402T166 381T143 340T127 274V267H333V275Z"></path><path id="MJX-20-TEX-N-64" d="M376 495Q376 511 376 535T377 568Q377 613 367 624T316 637H298V660Q298 683 300 683L310 684Q320 685 339 686T376 688Q393 689 413 690T443 693T454 694H457V390Q457 84 458 81Q461 61 472 55T517 46H535V0Q533 0 459 -5T380 -11H373V44L365 37Q307 -11 235 -11Q158 -11 96 50T34 215Q34 315 97 378T244 442Q319 442 376 393V495ZM373 342Q328 405 260 405Q211 405 173 369Q146 341 139 305T131 211Q131 155 138 120T173 59Q203 26 251 26Q322 26 373 103V342Z"></path><path id="MJX-20-TEX-N-63" d="M370 305T349 305T313 320T297 358Q297 381 312 396Q317 401 317 402T307 404Q281 408 258 408Q209 408 178 376Q131 329 131 219Q131 137 162 90Q203 29 272 29Q313 29 338 55T374 117Q376 125 379 127T395 129H409Q415 123 415 120Q415 116 411 104T395 71T366 33T318 2T249 -11Q163 -11 99 53T34 214Q34 318 99 383T250 448T370 421T404 357Q404 334 387 320Z"></path><path id="MJX-20-TEX-N-6F" d="M28 214Q28 309 93 378T250 448Q340 448 405 380T471 215Q471 120 407 55T250 -10Q153 -10 91 57T28 214ZM250 30Q372 30 372 193V225V250Q372 272 371 288T364 326T348 362T317 390T268 410Q263 411 252 411Q222 411 195 399Q152 377 139 338T126 246V226Q126 130 145 91Q177 30 250 30Z"></path><path id="MJX-20-TEX-N-75" d="M383 58Q327 -10 256 -10H249Q124 -10 105 89Q104 96 103 226Q102 335 102 348T96 369Q86 385 36 385H25V408Q25 431 27 431L38 432Q48 433 67 434T105 436Q122 437 142 438T172 441T184 442H187V261Q188 77 190 64Q193 49 204 40Q224 26 264 26Q290 26 311 35T343 58T363 90T375 120T379 144Q379 145 379 161T380 201T380 248V315Q380 361 370 372T320 385H302V431Q304 431 378 436T457 442H464V264Q464 84 465 81Q468 61 479 55T524 46H542V0Q540 0 467 -5T390 -11H383V58Z"></path><path id="MJX-20-TEX-N-6E" d="M41 46H55Q94 46 102 60V68Q102 77 102 91T102 122T103 161T103 203Q103 234 103 269T102 328V351Q99 370 88 376T43 385H25V408Q25 431 27 431L37 432Q47 433 65 434T102 436Q119 437 138 438T167 441T178 442H181V402Q181 364 182 364T187 369T199 384T218 402T247 421T285 437Q305 442 336 442Q450 438 463 329Q464 322 464 190V104Q464 66 466 59T477 49Q498 46 526 46H542V0H534L510 1Q487 2 460 2T422 3Q319 3 310 0H302V46H318Q379 46 379 62Q380 64 380 200Q379 335 378 343Q372 371 358 385T334 402T308 404Q263 404 229 370Q202 343 195 315T187 232V168V108Q187 78 188 68T191 55T200 49Q221 46 249 46H265V0H257L234 1Q210 2 183 2T145 3Q42 3 33 0H25V46H41Z"></path><path id="MJX-20-TEX-N-74" d="M27 422Q80 426 109 478T141 600V615H181V431H316V385H181V241Q182 116 182 100T189 68Q203 29 238 29Q282 29 292 100Q293 108 293 146V181H333V146V134Q333 57 291 17Q264 -10 221 -10Q187 -10 162 2T124 33T105 68T98 100Q97 107 97 248V385H18V422H27Z"></path><path id="MJX-20-TEX-N-66" d="M273 0Q255 3 146 3Q43 3 34 0H26V46H42Q70 46 91 49Q99 52 103 60Q104 62 104 224V385H33V431H104V497L105 564L107 574Q126 639 171 668T266 704Q267 704 275 704T289 705Q330 702 351 679T372 627Q372 604 358 590T321 576T284 590T270 627Q270 647 288 667H284Q280 668 273 668Q245 668 223 647T189 592Q183 572 182 497V431H293V385H185V225Q185 63 186 61T189 57T194 54T199 51T206 49T213 48T222 47T231 47T241 46T251 46H282V0H273Z"></path><path id="MJX-20-TEX-N-67" d="M329 409Q373 453 429 453Q459 453 472 434T485 396Q485 382 476 371T449 360Q416 360 412 390Q410 404 415 411Q415 412 416 414V415Q388 412 363 393Q355 388 355 386Q355 385 359 381T368 369T379 351T388 325T392 292Q392 230 343 187T222 143Q172 143 123 171Q112 153 112 133Q112 98 138 81Q147 75 155 75T227 73Q311 72 335 67Q396 58 431 26Q470 -13 470 -72Q470 -139 392 -175Q332 -206 250 -206Q167 -206 107 -175Q29 -140 29 -75Q29 -39 50 -15T92 18L103 24Q67 55 67 108Q67 155 96 193Q52 237 52 292Q52 355 102 398T223 442Q274 442 318 416L329 409ZM299 343Q294 371 273 387T221 404Q192 404 171 388T145 343Q142 326 142 292Q142 248 149 227T179 192Q196 182 222 182Q244 182 260 189T283 207T294 227T299 242Q302 258 302 292T299 343ZM403 -75Q403 -50 389 -34T348 -11T299 -2T245 0H218Q151 0 138 -6Q118 -15 107 -34T95 -74Q95 -84 101 -97T122 -127T170 -155T250 -167Q319 -167 361 -139T403 -75Z"></path><path id="MJX-20-TEX-N-69" d="M69 609Q69 637 87 653T131 669Q154 667 171 652T188 609Q188 579 171 564T129 549Q104 549 87 564T69 609ZM247 0Q232 3 143 3Q132 3 106 3T56 1L34 0H26V46H42Q70 46 91 49Q100 53 102 60T104 102V205V293Q104 345 102 359T88 378Q74 385 41 385H30V408Q30 431 32 431L42 432Q52 433 70 434T106 436Q123 437 142 438T171 441T182 442H185V62Q190 52 197 50T232 46H255V0H247Z"></path><path id="MJX-20-TEX-N-54" d="M36 443Q37 448 46 558T55 671V677H666V671Q667 666 676 556T685 443V437H645V443Q645 445 642 478T631 544T610 593Q593 614 555 625Q534 630 478 630H451H443Q417 630 414 618Q413 616 413 339V63Q420 53 439 50T528 46H558V0H545L361 3Q186 1 177 0H164V46H194Q264 46 283 49T309 63V339V550Q309 620 304 625T271 630H244H224Q154 630 119 601Q101 585 93 554T81 486T76 443V437H36V443Z"></path><path id="MJX-20-TEX-N-6C" d="M42 46H56Q95 46 103 60V68Q103 77 103 91T103 124T104 167T104 217T104 272T104 329Q104 366 104 407T104 482T104 542T103 586T103 603Q100 622 89 628T44 637H26V660Q26 683 28 683L38 684Q48 685 67 686T104 688Q121 689 141 690T171 693T182 694H185V379Q185 62 186 60Q190 52 198 49Q219 46 247 46H263V0H255L232 1Q209 2 183 2T145 3T107 3T57 1L34 0H26V46H42Z"></path><path id="MJX-20-TEX-N-6D" d="M41 46H55Q94 46 102 60V68Q102 77 102 91T102 122T103 161T103 203Q103 234 103 269T102 328V351Q99 370 88 376T43 385H25V408Q25 431 27 431L37 432Q47 433 65 434T102 436Q119 437 138 438T167 441T178 442H181V402Q181 364 182 364T187 369T199 384T218 402T247 421T285 437Q305 442 336 442Q351 442 364 440T387 434T406 426T421 417T432 406T441 395T448 384T452 374T455 366L457 361L460 365Q463 369 466 373T475 384T488 397T503 410T523 422T546 432T572 439T603 442Q729 442 740 329Q741 322 741 190V104Q741 66 743 59T754 49Q775 46 803 46H819V0H811L788 1Q764 2 737 2T699 3Q596 3 587 0H579V46H595Q656 46 656 62Q657 64 657 200Q656 335 655 343Q649 371 635 385T611 402T585 404Q540 404 506 370Q479 343 472 315T464 232V168V108Q464 78 465 68T468 55T477 49Q498 46 526 46H542V0H534L510 1Q487 2 460 2T422 3Q319 3 310 0H302V46H318Q379 46 379 62Q380 64 380 200Q379 335 378 343Q372 371 358 385T334 402T308 404Q263 404 229 370Q202 343 195 315T187 232V168V108Q187 78 188 68T191 55T200 49Q221 46 249 46H265V0H257L234 1Q210 2 183 2T145 3Q42 3 33 0H25V46H41Z"></path><path id="MJX-20-TEX-N-70" d="M36 -148H50Q89 -148 97 -134V-126Q97 -119 97 -107T97 -77T98 -38T98 6T98 55T98 106Q98 140 98 177T98 243T98 296T97 335T97 351Q94 370 83 376T38 385H20V408Q20 431 22 431L32 432Q42 433 61 434T98 436Q115 437 135 438T165 441T176 442H179V416L180 390L188 397Q247 441 326 441Q407 441 464 377T522 216Q522 115 457 52T310 -11Q242 -11 190 33L182 40V-45V-101Q182 -128 184 -134T195 -145Q216 -148 244 -148H260V-194H252L228 -193Q205 -192 178 -192T140 -191Q37 -191 28 -194H20V-148H36ZM424 218Q424 292 390 347T305 402Q234 402 182 337V98Q222 26 294 26Q345 26 384 80T424 218Z"></path><path id="MJX-20-TEX-N-73" d="M295 316Q295 356 268 385T190 414Q154 414 128 401Q98 382 98 349Q97 344 98 336T114 312T157 287Q175 282 201 278T245 269T277 256Q294 248 310 236T342 195T359 133Q359 71 321 31T198 -10H190Q138 -10 94 26L86 19L77 10Q71 4 65 -1L54 -11H46H42Q39 -11 33 -5V74V132Q33 153 35 157T45 162H54Q66 162 70 158T75 146T82 119T101 77Q136 26 198 26Q295 26 295 104Q295 133 277 151Q257 175 194 187T111 210Q75 227 54 256T33 318Q33 357 50 384T93 424T143 442T187 447H198Q238 447 268 432L283 424L292 431Q302 440 314 448H322H326Q329 448 335 442V310L329 304H301Q295 310 295 316Z"></path><path id="MJX-20-TEX-N-D7" d="M630 29Q630 9 609 9Q604 9 587 25T493 118L389 222L284 117Q178 13 175 11Q171 9 168 9Q160 9 154 15T147 29Q147 36 161 51T255 146L359 250L255 354Q174 435 161 449T147 471Q147 480 153 485T168 490Q173 490 175 489Q178 487 284 383L389 278L493 382Q570 459 587 475T609 491Q630 491 630 471Q630 464 620 453T522 355L418 250L522 145Q606 61 618 48T630 29Z"></path><path id="MJX-20-TEX-N-31" d="M213 578L200 573Q186 568 160 563T102 556H83V602H102Q149 604 189 617T245 641T273 663Q275 666 285 666Q294 666 302 660V361L303 61Q310 54 315 52T339 48T401 46H427V0H416Q395 3 257 3Q121 3 100 0H88V46H114Q136 46 152 46T177 47T193 50T201 52T207 57T213 61V578Z"></path><path id="MJX-20-TEX-N-30" d="M96 585Q152 666 249 666Q297 666 345 640T423 548Q460 465 460 320Q460 165 417 83Q397 41 362 16T301 -15T250 -22Q224 -22 198 -16T137 16T82 83Q39 165 39 320Q39 494 96 585ZM321 597Q291 629 250 629Q208 629 178 597Q153 571 145 525T137 333Q137 175 145 125T181 46Q209 16 250 16Q290 16 318 46Q347 76 354 130T362 333Q362 478 354 524T321 597Z"></path><path id="MJX-20-TEX-N-36" d="M42 313Q42 476 123 571T303 666Q372 666 402 630T432 550Q432 525 418 510T379 495Q356 495 341 509T326 548Q326 592 373 601Q351 623 311 626Q240 626 194 566Q147 500 147 364L148 360Q153 366 156 373Q197 433 263 433H267Q313 433 348 414Q372 400 396 374T435 317Q456 268 456 210V192Q456 169 451 149Q440 90 387 34T253 -22Q225 -22 199 -14T143 16T92 75T56 172T42 313ZM257 397Q227 397 205 380T171 335T154 278T148 216Q148 133 160 97T198 39Q222 21 251 21Q302 21 329 59Q342 77 347 104T352 209Q352 289 347 316T329 361Q302 397 257 397Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="mtext"><use data-c="52" xlink:href="#MJX-20-TEX-N-52"></use><use data-c="50" xlink:href="#MJX-20-TEX-N-50" transform="translate(736,0)"></use><use data-c="4D" xlink:href="#MJX-20-TEX-N-4D" transform="translate(1417,0)"></use></g><g data-mml-node="mo" transform="translate(2611.8,0)"><use data-c="3D" xlink:href="#MJX-20-TEX-N-3D"></use></g><g data-mml-node="mfrac" transform="translate(3667.6,0)"><g data-mml-node="mtext" transform="translate(1118,676)"><use data-c="52" xlink:href="#MJX-20-TEX-N-52"></use><use data-c="61" xlink:href="#MJX-20-TEX-N-61" transform="translate(736,0)"></use><use data-c="77" xlink:href="#MJX-20-TEX-N-77" transform="translate(1236,0)"></use><use data-c="20" xlink:href="#MJX-20-TEX-N-20" transform="translate(1958,0)"></use><use data-c="72" xlink:href="#MJX-20-TEX-N-72" transform="translate(2208,0)"></use><use data-c="65" xlink:href="#MJX-20-TEX-N-65" transform="translate(2600,0)"></use><use data-c="61" xlink:href="#MJX-20-TEX-N-61" transform="translate(3044,0)"></use><use data-c="64" xlink:href="#MJX-20-TEX-N-64" transform="translate(3544,0)"></use><use data-c="20" xlink:href="#MJX-20-TEX-N-20" transform="translate(4100,0)"></use><use data-c="63" xlink:href="#MJX-20-TEX-N-63" transform="translate(4350,0)"></use><use data-c="6F" xlink:href="#MJX-20-TEX-N-6F" transform="translate(4794,0)"></use><use data-c="75" xlink:href="#MJX-20-TEX-N-75" transform="translate(5294,0)"></use><use data-c="6E" xlink:href="#MJX-20-TEX-N-6E" transform="translate(5850,0)"></use><use data-c="74" xlink:href="#MJX-20-TEX-N-74" transform="translate(6406,0)"></use><use data-c="20" xlink:href="#MJX-20-TEX-N-20" transform="translate(6795,0)"></use><use data-c="66" xlink:href="#MJX-20-TEX-N-66" transform="translate(7045,0)"></use><use data-c="6F" xlink:href="#MJX-20-TEX-N-6F" transform="translate(7351,0)"></use><use data-c="72" xlink:href="#MJX-20-TEX-N-72" transform="translate(7851,0)"></use><use data-c="20" xlink:href="#MJX-20-TEX-N-20" transform="translate(8243,0)"></use><use data-c="72" xlink:href="#MJX-20-TEX-N-72" transform="translate(8493,0)"></use><use data-c="65" xlink:href="#MJX-20-TEX-N-65" transform="translate(8885,0)"></use><use data-c="67" xlink:href="#MJX-20-TEX-N-67" transform="translate(9329,0)"></use><use data-c="69" xlink:href="#MJX-20-TEX-N-69" transform="translate(9829,0)"></use><use data-c="6F" xlink:href="#MJX-20-TEX-N-6F" transform="translate(10107,0)"></use><use data-c="6E" xlink:href="#MJX-20-TEX-N-6E" transform="translate(10607,0)"></use></g><g data-mml-node="mtext" transform="translate(220,-686)"><use data-c="54" xlink:href="#MJX-20-TEX-N-54"></use><use data-c="6F" xlink:href="#MJX-20-TEX-N-6F" transform="translate(722,0)"></use><use data-c="74" xlink:href="#MJX-20-TEX-N-74" transform="translate(1222,0)"></use><use data-c="61" xlink:href="#MJX-20-TEX-N-61" transform="translate(1611,0)"></use><use data-c="6C" xlink:href="#MJX-20-TEX-N-6C" transform="translate(2111,0)"></use><use data-c="20" xlink:href="#MJX-20-TEX-N-20" transform="translate(2389,0)"></use><use data-c="6D" xlink:href="#MJX-20-TEX-N-6D" transform="translate(2639,0)"></use><use data-c="61" xlink:href="#MJX-20-TEX-N-61" transform="translate(3472,0)"></use><use data-c="70" xlink:href="#MJX-20-TEX-N-70" transform="translate(3972,0)"></use><use data-c="70" xlink:href="#MJX-20-TEX-N-70" transform="translate(4528,0)"></use><use data-c="65" xlink:href="#MJX-20-TEX-N-65" transform="translate(5084,0)"></use><use data-c="64" xlink:href="#MJX-20-TEX-N-64" transform="translate(5528,0)"></use><use data-c="20" xlink:href="#MJX-20-TEX-N-20" transform="translate(6084,0)"></use><use data-c="72" xlink:href="#MJX-20-TEX-N-72" transform="translate(6334,0)"></use><use data-c="65" xlink:href="#MJX-20-TEX-N-65" transform="translate(6726,0)"></use><use data-c="61" xlink:href="#MJX-20-TEX-N-61" transform="translate(7170,0)"></use><use data-c="64" xlink:href="#MJX-20-TEX-N-64" transform="translate(7670,0)"></use><use data-c="73" xlink:href="#MJX-20-TEX-N-73" transform="translate(8226,0)"></use><use data-c="20" xlink:href="#MJX-20-TEX-N-20" transform="translate(8620,0)"></use><use data-c="69" xlink:href="#MJX-20-TEX-N-69" transform="translate(8870,0)"></use><use data-c="6E" xlink:href="#MJX-20-TEX-N-6E" transform="translate(9148,0)"></use><use data-c="20" xlink:href="#MJX-20-TEX-N-20" transform="translate(9704,0)"></use><use data-c="73" xlink:href="#MJX-20-TEX-N-73" transform="translate(9954,0)"></use><use data-c="61" xlink:href="#MJX-20-TEX-N-61" transform="translate(10348,0)"></use><use data-c="6D" xlink:href="#MJX-20-TEX-N-6D" transform="translate(10848,0)"></use><use data-c="70" xlink:href="#MJX-20-TEX-N-70" transform="translate(11681,0)"></use><use data-c="6C" xlink:href="#MJX-20-TEX-N-6C" transform="translate(12237,0)"></use><use data-c="65" xlink:href="#MJX-20-TEX-N-65" transform="translate(12515,0)"></use></g><rect width="13159" height="60" x="120" y="220"></rect></g><g data-mml-node="mo" transform="translate(17288.8,0)"><use data-c="D7" xlink:href="#MJX-20-TEX-N-D7"></use></g><g data-mml-node="msup" transform="translate(18289,0)"><g data-mml-node="mn"><use data-c="31" xlink:href="#MJX-20-TEX-N-31"></use><use data-c="30" xlink:href="#MJX-20-TEX-N-30" transform="translate(500,0)"></use></g><g data-mml-node="mn" transform="translate(1033,413) scale(0.707)"><use data-c="36" xlink:href="#MJX-20-TEX-N-36"></use></g></g></g></g></svg><mjx-assistive-mml unselectable="on" display="block"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><mtext>RPM</mtext><mo>=</mo><mfrac><mtext>Raw read count for region</mtext><mtext>Total mapped reads in sample</mtext></mfrac><mo>×</mo><msup><mn>10</mn><mn>6</mn></msup></math></mjx-assistive-mml></mjx-container></div></div><div class="md-alert md-alert-tip tip"><p><span class='md-alert-text md-alert-text-tip'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M8 1.5c-2.363 0-4 1.69-4 3.75 0 .984.424 1.625.984 2.304l.214.253c.223.264.47.556.673.848.284.411.537.896.621 1.49a.75.75 0 0 1-1.484.211c-.04-.282-.163-.547-.37-.847a8.456 8.456 0 0 0-.542-.68c-.084-.1-.173-.205-.268-.32C3.201 7.75 2.5 6.766 2.5 5.25 2.5 2.31 4.863 0 8 0s5.5 2.31 5.5 5.25c0 1.516-.701 2.5-1.328 3.259-.095.115-.184.22-.268.319-.207.245-.383.453-.541.681-.208.3-.33.565-.37.847a.751.751 0 0 1-1.485-.212c.084-.593.337-1.078.621-1.489.203-.292.45-.584.673-.848.075-.088.147-.173.213-.253.561-.679.985-1.32.985-2.304 0-2.06-1.637-3.75-4-3.75ZM5.75 12h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5ZM6 15.25a.75.75 0 0 1 .75-.75h2.5a.75.75 0 0 1 0 1.5h-2.5a.75.75 0 0 1-.75-.75Z"></path></svg>Tip</span><br></p><p><strong><span>那为什么RPM不适用于“global effect”的 ChIP-seq 数据？</span></strong></p><p><span>在某些情况下，比如：药物处理、基因突变、某种转录因子失活/过表达， 这些干预可能导致 </span><strong><span>全基因组范围的组蛋白修饰或转录因子结合水平发生普遍改变</span></strong><span>，而不是仅仅局部变化</span></p><p>&nbsp;</p><p><span>假设你研究某个药物是否会提升 H3K27ac（组蛋白乙酰化）的总体水平：</span></p><figure class='table-figure'><table><thead><tr><th><span>Sample</span></th><th><span>总 ChIP read count</span></th><th><span>某区域的read count</span></th><th><span>RPM</span></th></tr></thead><tbody><tr><td><span>Control</span></td><td><span>10M</span></td><td><span>1000</span></td><td><span>100</span></td></tr><tr><td><span>Treated</span></td><td><span>20M (可能是global effect)</span></td><td><span>2000</span></td><td><span>100</span></td></tr></tbody></table></figure><p><span>你看到 RPM 一样，可能会以为这个区域没有变化，但实际上：</span></p><ul><li><p><span>总信号量上升了（表明全基因组水平有提升），</span></p></li><li><p><span>这个区域的原始读数也翻倍了！</span></p></li></ul><p><span>→ </span><strong><span>RPM 掩盖了全局效应，失去了比较价值！</span></strong></p></div><p>&nbsp;</p></li><li><p><span>However, most of the ChIP-seq studies have </span><strong><span>overlooked</span></strong><span> the </span><strong><span>normalization problem</span></strong><span> that have to </span><strong><span>be corrected with spike-in</span></strong><span>. A method that retrospectively renormalize data sets without spike-in is lacking. </span></p><div class="md-alert md-alert-tip tip"><p><span class='md-alert-text md-alert-text-tip'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M8 1.5c-2.363 0-4 1.69-4 3.75 0 .984.424 1.625.984 2.304l.214.253c.223.264.47.556.673.848.284.411.537.896.621 1.49a.75.75 0 0 1-1.484.211c-.04-.282-.163-.547-.37-.847a8.456 8.456 0 0 0-.542-.68c-.084-.1-.173-.205-.268-.32C3.201 7.75 2.5 6.766 2.5 5.25 2.5 2.31 4.863 0 8 0s5.5 2.31 5.5 5.25c0 1.516-.701 2.5-1.328 3.259-.095.115-.184.22-.268.319-.207.245-.383.453-.541.681-.208.3-.33.565-.37.847a.751.751 0 0 1-1.485-.212c.084-.593.337-1.078.621-1.489.203-.292.45-.584.673-.848.075-.088.147-.173.213-.253.561-.679.985-1.32.985-2.304 0-2.06-1.637-3.75-4-3.75ZM5.75 12h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5ZM6 15.25a.75.75 0 0 1 .75-.75h2.5a.75.75 0 0 1 0 1.5h-2.5a.75.75 0 0 1-.75-.75Z"></path></svg>Tip</span><br></p><p><strong><span>Spike-in </span></strong></p><p><span>是在样本中</span><strong><span>人为加入</span></strong><span>的一小部分</span><strong><span>外源性的（exogenous）DNA 或 RNA</span></strong><span>，通常来源于</span><strong><span>另一物种</span></strong><span>，用于在后续的数据分析中作为作为</span><strong><span>内参（internal control）</span></strong><span>，帮助进行数据标准化和校正</span></p></div><p>&nbsp;</p></li><li><p><code>ChIPseqSpikeInFree</code><span> 是一种新的 ChIP-seq 数据标准化方法，即可在不同条件和处理之间</span><strong><span>有效确定缩放因子</span></strong><span>，从而揭示</span><strong><span>组蛋白修饰</span></strong><span>的globel effect。且不依赖于</span><strong><span>外源性 spike-in DNA或峰值检测</span></strong><span>。</span></p></li></ul><ul><li><p><strong><span>How is scaling factor calculated?</span></strong></p><p><span>对于使用</span><strong><span>相同抗体</span></strong><span>进行 ChIP 的 </span><mjx-container class="MathJax" jax="SVG" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="1.357ex" height="1.025ex" role="img" focusable="false" viewBox="0 -442 600 453" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -0.025ex;"><defs><path id="MJX-44-TEX-I-1D45B" d="M21 287Q22 293 24 303T36 341T56 388T89 425T135 442Q171 442 195 424T225 390T231 369Q231 367 232 367L243 378Q304 442 382 442Q436 442 469 415T503 336T465 179T427 52Q427 26 444 26Q450 26 453 27Q482 32 505 65T540 145Q542 153 560 153Q580 153 580 145Q580 144 576 130Q568 101 554 73T508 17T439 -10Q392 -10 371 17T350 73Q350 92 386 193T423 345Q423 404 379 404H374Q288 404 229 303L222 291L189 157Q156 26 151 16Q138 -11 108 -11Q95 -11 87 -5T76 7T74 17Q74 30 112 180T152 343Q153 348 153 366Q153 405 129 405Q91 405 66 305Q60 285 60 284Q58 278 41 278H27Q21 284 21 287Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="mi"><use data-c="1D45B" xlink:href="#MJX-44-TEX-I-1D45B"></use></g></g></g></svg><mjx-assistive-mml unselectable="on" display="inline"><math xmlns="http://www.w3.org/1998/Math/MathML"><mi>n</mi></math></mjx-assistive-mml></mjx-container><script type="math/tex">n</script><span> 个样本，我们选择其中一个样本作为参考样本（其斜率为 </span><mjx-container class="MathJax" jax="SVG" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="2.19ex" height="2.034ex" role="img" focusable="false" viewBox="0 -705 967.9 899" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -0.439ex;"><defs><path id="MJX-45-TEX-I-1D6FD" d="M29 -194Q23 -188 23 -186Q23 -183 102 134T186 465Q208 533 243 584T309 658Q365 705 429 705H431Q493 705 533 667T573 570Q573 465 469 396L482 383Q533 332 533 252Q533 139 448 65T257 -10Q227 -10 203 -2T165 17T143 40T131 59T126 65L62 -188Q60 -194 42 -194H29ZM353 431Q392 431 427 419L432 422Q436 426 439 429T449 439T461 453T472 471T484 495T493 524T501 560Q503 569 503 593Q503 611 502 616Q487 667 426 667Q384 667 347 643T286 582T247 514T224 455Q219 439 186 308T152 168Q151 163 151 147Q151 99 173 68Q204 26 260 26Q302 26 349 51T425 137Q441 171 449 214T457 279Q457 337 422 372Q380 358 347 358H337Q258 358 258 389Q258 396 261 403Q275 431 353 431Z"></path><path id="MJX-45-TEX-I-1D45F" d="M21 287Q22 290 23 295T28 317T38 348T53 381T73 411T99 433T132 442Q161 442 183 430T214 408T225 388Q227 382 228 382T236 389Q284 441 347 441H350Q398 441 422 400Q430 381 430 363Q430 333 417 315T391 292T366 288Q346 288 334 299T322 328Q322 376 378 392Q356 405 342 405Q286 405 239 331Q229 315 224 298T190 165Q156 25 151 16Q138 -11 108 -11Q95 -11 87 -5T76 7T74 17Q74 30 114 189T154 366Q154 405 128 405Q107 405 92 377T68 316T57 280Q55 278 41 278H27Q21 284 21 287Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="msub"><g data-mml-node="mi"><use data-c="1D6FD" xlink:href="#MJX-45-TEX-I-1D6FD"></use></g><g data-mml-node="mi" transform="translate(599,-150) scale(0.707)"><use data-c="1D45F" xlink:href="#MJX-45-TEX-I-1D45F"></use></g></g></g></g></svg><mjx-assistive-mml unselectable="on" display="inline"><math xmlns="http://www.w3.org/1998/Math/MathML"><msub><mi>β</mi><mi>r</mi></msub></math></mjx-assistive-mml></mjx-container><script type="math/tex">\beta_r</script><span>）。</span><br/><span>然后，对于任意其他样本 </span><mjx-container class="MathJax" jax="SVG" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="0.781ex" height="1.52ex" role="img" focusable="false" viewBox="0 -661 345 672" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -0.025ex;"><defs><path id="MJX-46-TEX-I-1D456" d="M184 600Q184 624 203 642T247 661Q265 661 277 649T290 619Q290 596 270 577T226 557Q211 557 198 567T184 600ZM21 287Q21 295 30 318T54 369T98 420T158 442Q197 442 223 419T250 357Q250 340 236 301T196 196T154 83Q149 61 149 51Q149 26 166 26Q175 26 185 29T208 43T235 78T260 137Q263 149 265 151T282 153Q302 153 302 143Q302 135 293 112T268 61T223 11T161 -11Q129 -11 102 10T74 74Q74 91 79 106T122 220Q160 321 166 341T173 380Q173 404 156 404H154Q124 404 99 371T61 287Q60 286 59 284T58 281T56 279T53 278T49 278T41 278H27Q21 284 21 287Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="mi"><use data-c="1D456" xlink:href="#MJX-46-TEX-I-1D456"></use></g></g></g></svg><mjx-assistive-mml unselectable="on" display="inline"><math xmlns="http://www.w3.org/1998/Math/MathML"><mi>i</mi></math></mjx-assistive-mml></mjx-container><script type="math/tex">i</script><span>，我们根据该样本的斜率 </span><mjx-container class="MathJax" jax="SVG" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="2.02ex" height="2.034ex" role="img" focusable="false" viewBox="0 -705 893 899" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -0.439ex;"><defs><path id="MJX-47-TEX-I-1D6FD" d="M29 -194Q23 -188 23 -186Q23 -183 102 134T186 465Q208 533 243 584T309 658Q365 705 429 705H431Q493 705 533 667T573 570Q573 465 469 396L482 383Q533 332 533 252Q533 139 448 65T257 -10Q227 -10 203 -2T165 17T143 40T131 59T126 65L62 -188Q60 -194 42 -194H29ZM353 431Q392 431 427 419L432 422Q436 426 439 429T449 439T461 453T472 471T484 495T493 524T501 560Q503 569 503 593Q503 611 502 616Q487 667 426 667Q384 667 347 643T286 582T247 514T224 455Q219 439 186 308T152 168Q151 163 151 147Q151 99 173 68Q204 26 260 26Q302 26 349 51T425 137Q441 171 449 214T457 279Q457 337 422 372Q380 358 347 358H337Q258 358 258 389Q258 396 261 403Q275 431 353 431Z"></path><path id="MJX-47-TEX-I-1D456" d="M184 600Q184 624 203 642T247 661Q265 661 277 649T290 619Q290 596 270 577T226 557Q211 557 198 567T184 600ZM21 287Q21 295 30 318T54 369T98 420T158 442Q197 442 223 419T250 357Q250 340 236 301T196 196T154 83Q149 61 149 51Q149 26 166 26Q175 26 185 29T208 43T235 78T260 137Q263 149 265 151T282 153Q302 153 302 143Q302 135 293 112T268 61T223 11T161 -11Q129 -11 102 10T74 74Q74 91 79 106T122 220Q160 321 166 341T173 380Q173 404 156 404H154Q124 404 99 371T61 287Q60 286 59 284T58 281T56 279T53 278T49 278T41 278H27Q21 284 21 287Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="msub"><g data-mml-node="mi"><use data-c="1D6FD" xlink:href="#MJX-47-TEX-I-1D6FD"></use></g><g data-mml-node="mi" transform="translate(599,-150) scale(0.707)"><use data-c="1D456" xlink:href="#MJX-47-TEX-I-1D456"></use></g></g></g></g></svg><mjx-assistive-mml unselectable="on" display="inline"><math xmlns="http://www.w3.org/1998/Math/MathML"><msub><mi>β</mi><mi>i</mi></msub></math></mjx-assistive-mml></mjx-container><script type="math/tex">\beta_i</script><span> 计算其缩放因子 </span><mjx-container class="MathJax" jax="SVG" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="2.127ex" height="1.952ex" role="img" focusable="false" viewBox="0 -705 940 862.8" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -0.357ex;"><defs><path id="MJX-48-TEX-I-1D446" d="M308 24Q367 24 416 76T466 197Q466 260 414 284Q308 311 278 321T236 341Q176 383 176 462Q176 523 208 573T273 648Q302 673 343 688T407 704H418H425Q521 704 564 640Q565 640 577 653T603 682T623 704Q624 704 627 704T632 705Q645 705 645 698T617 577T585 459T569 456Q549 456 549 465Q549 471 550 475Q550 478 551 494T553 520Q553 554 544 579T526 616T501 641Q465 662 419 662Q362 662 313 616T263 510Q263 480 278 458T319 427Q323 425 389 408T456 390Q490 379 522 342T554 242Q554 216 546 186Q541 164 528 137T492 78T426 18T332 -20Q320 -22 298 -22Q199 -22 144 33L134 44L106 13Q83 -14 78 -18T65 -22Q52 -22 52 -14Q52 -11 110 221Q112 227 130 227H143Q149 221 149 216Q149 214 148 207T144 186T142 153Q144 114 160 87T203 47T255 29T308 24Z"></path><path id="MJX-48-TEX-I-1D456" d="M184 600Q184 624 203 642T247 661Q265 661 277 649T290 619Q290 596 270 577T226 557Q211 557 198 567T184 600ZM21 287Q21 295 30 318T54 369T98 420T158 442Q197 442 223 419T250 357Q250 340 236 301T196 196T154 83Q149 61 149 51Q149 26 166 26Q175 26 185 29T208 43T235 78T260 137Q263 149 265 151T282 153Q302 153 302 143Q302 135 293 112T268 61T223 11T161 -11Q129 -11 102 10T74 74Q74 91 79 106T122 220Q160 321 166 341T173 380Q173 404 156 404H154Q124 404 99 371T61 287Q60 286 59 284T58 281T56 279T53 278T49 278T41 278H27Q21 284 21 287Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="msub"><g data-mml-node="mi"><use data-c="1D446" xlink:href="#MJX-48-TEX-I-1D446"></use></g><g data-mml-node="mi" transform="translate(646,-150) scale(0.707)"><use data-c="1D456" xlink:href="#MJX-48-TEX-I-1D456"></use></g></g></g></g></svg><mjx-assistive-mml unselectable="on" display="inline"><math xmlns="http://www.w3.org/1998/Math/MathML"><msub><mi>S</mi><mi>i</mi></msub></math></mjx-assistive-mml></mjx-container><script type="math/tex">S_i</script><span>，公式如下：</span></p><div contenteditable="false" spellcheck="false" class="mathjax-block md-end-block md-math-block md-rawblock" id="mathjax-n3427" cid="n3427" mdtype="math_block" data-math-tag-before="0" data-math-tag-after="0" data-math-labels="[]"><div class="md-math-container"><mjx-container class="MathJax" jax="SVG" display="true" style="position: relative;"><svg xmlns="http://www.w3.org/2000/svg" width="8.329ex" height="5.115ex" role="img" focusable="false" viewBox="0 -1381 3681.4 2261" xmlns:xlink="http://www.w3.org/1999/xlink" aria-hidden="true" style="vertical-align: -1.991ex;"><defs><path id="MJX-21-TEX-I-1D446" d="M308 24Q367 24 416 76T466 197Q466 260 414 284Q308 311 278 321T236 341Q176 383 176 462Q176 523 208 573T273 648Q302 673 343 688T407 704H418H425Q521 704 564 640Q565 640 577 653T603 682T623 704Q624 704 627 704T632 705Q645 705 645 698T617 577T585 459T569 456Q549 456 549 465Q549 471 550 475Q550 478 551 494T553 520Q553 554 544 579T526 616T501 641Q465 662 419 662Q362 662 313 616T263 510Q263 480 278 458T319 427Q323 425 389 408T456 390Q490 379 522 342T554 242Q554 216 546 186Q541 164 528 137T492 78T426 18T332 -20Q320 -22 298 -22Q199 -22 144 33L134 44L106 13Q83 -14 78 -18T65 -22Q52 -22 52 -14Q52 -11 110 221Q112 227 130 227H143Q149 221 149 216Q149 214 148 207T144 186T142 153Q144 114 160 87T203 47T255 29T308 24Z"></path><path id="MJX-21-TEX-I-1D456" d="M184 600Q184 624 203 642T247 661Q265 661 277 649T290 619Q290 596 270 577T226 557Q211 557 198 567T184 600ZM21 287Q21 295 30 318T54 369T98 420T158 442Q197 442 223 419T250 357Q250 340 236 301T196 196T154 83Q149 61 149 51Q149 26 166 26Q175 26 185 29T208 43T235 78T260 137Q263 149 265 151T282 153Q302 153 302 143Q302 135 293 112T268 61T223 11T161 -11Q129 -11 102 10T74 74Q74 91 79 106T122 220Q160 321 166 341T173 380Q173 404 156 404H154Q124 404 99 371T61 287Q60 286 59 284T58 281T56 279T53 278T49 278T41 278H27Q21 284 21 287Z"></path><path id="MJX-21-TEX-N-3D" d="M56 347Q56 360 70 367H707Q722 359 722 347Q722 336 708 328L390 327H72Q56 332 56 347ZM56 153Q56 168 72 173H708Q722 163 722 153Q722 140 707 133H70Q56 140 56 153Z"></path><path id="MJX-21-TEX-I-1D6FD" d="M29 -194Q23 -188 23 -186Q23 -183 102 134T186 465Q208 533 243 584T309 658Q365 705 429 705H431Q493 705 533 667T573 570Q573 465 469 396L482 383Q533 332 533 252Q533 139 448 65T257 -10Q227 -10 203 -2T165 17T143 40T131 59T126 65L62 -188Q60 -194 42 -194H29ZM353 431Q392 431 427 419L432 422Q436 426 439 429T449 439T461 453T472 471T484 495T493 524T501 560Q503 569 503 593Q503 611 502 616Q487 667 426 667Q384 667 347 643T286 582T247 514T224 455Q219 439 186 308T152 168Q151 163 151 147Q151 99 173 68Q204 26 260 26Q302 26 349 51T425 137Q441 171 449 214T457 279Q457 337 422 372Q380 358 347 358H337Q258 358 258 389Q258 396 261 403Q275 431 353 431Z"></path><path id="MJX-21-TEX-I-1D45F" d="M21 287Q22 290 23 295T28 317T38 348T53 381T73 411T99 433T132 442Q161 442 183 430T214 408T225 388Q227 382 228 382T236 389Q284 441 347 441H350Q398 441 422 400Q430 381 430 363Q430 333 417 315T391 292T366 288Q346 288 334 299T322 328Q322 376 378 392Q356 405 342 405Q286 405 239 331Q229 315 224 298T190 165Q156 25 151 16Q138 -11 108 -11Q95 -11 87 -5T76 7T74 17Q74 30 114 189T154 366Q154 405 128 405Q107 405 92 377T68 316T57 280Q55 278 41 278H27Q21 284 21 287Z"></path></defs><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="msub"><g data-mml-node="mi"><use data-c="1D446" xlink:href="#MJX-21-TEX-I-1D446"></use></g><g data-mml-node="mi" transform="translate(646,-150) scale(0.707)"><use data-c="1D456" xlink:href="#MJX-21-TEX-I-1D456"></use></g></g><g data-mml-node="mo" transform="translate(1217.7,0)"><use data-c="3D" xlink:href="#MJX-21-TEX-N-3D"></use></g><g data-mml-node="mfrac" transform="translate(2273.5,0)"><g data-mml-node="msub" transform="translate(220,676)"><g data-mml-node="mi"><use data-c="1D6FD" xlink:href="#MJX-21-TEX-I-1D6FD"></use></g><g data-mml-node="mi" transform="translate(599,-150) scale(0.707)"><use data-c="1D45F" xlink:href="#MJX-21-TEX-I-1D45F"></use></g></g><g data-mml-node="msub" transform="translate(257.5,-686)"><g data-mml-node="mi"><use data-c="1D6FD" xlink:href="#MJX-21-TEX-I-1D6FD"></use></g><g data-mml-node="mi" transform="translate(599,-150) scale(0.707)"><use data-c="1D456" xlink:href="#MJX-21-TEX-I-1D456"></use></g></g><rect width="1167.9" height="60" x="120" y="220"></rect></g></g></g></svg><mjx-assistive-mml unselectable="on" display="block"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><msub><mi>S</mi><mi>i</mi></msub><mo>=</mo><mfrac><msub><mi>β</mi><mi>r</mi></msub><msub><mi>β</mi><mi>i</mi></msub></mfrac></math></mjx-assistive-mml></mjx-container></div></div><p><img src="./picture/085.png" width="900"/></p></li><li><p>&nbsp;</p></li></ul><p>&nbsp;</p><div class="md-alert md-alert-caution caution"><p><span class='md-alert-text md-alert-text-caution'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M4.47.22A.749.749 0 0 1 5 0h6c.199 0 .389.079.53.22l4.25 4.25c.141.14.22.331.22.53v6a.749.749 0 0 1-.22.53l-4.25 4.25A.749.749 0 0 1 11 16H5a.749.749 0 0 1-.53-.22L.22 11.53A.749.749 0 0 1 0 11V5c0-.199.079-.389.22-.53Zm.84 1.28L1.5 5.31v5.38l3.81 3.81h5.38l3.81-3.81V5.31L10.69 1.5ZM8 4a.75.75 0 0 1 .75.75v3.5a.75.75 0 0 1-1.5 0v-3.5A.75.75 0 0 1 8 4Zm0 8a1 1 0 1 1 0-2 1 1 0 0 1 0 2Z"></path></svg>Caution</span><br></p><ul><li><p><span>⚠️ 不建议</span><strong><span>盲目使用</span></strong><span> </span><code>ChIPseqSpikeInFree</code><span>：如果实验组和对照组之间 </span><strong><span>是否存在全局变化（如组蛋白修饰水平变化）</span></strong><span> 并不明确，那么应该通过其他生物学手段（例如 </span><strong><span>Western blot 检测组蛋白修饰水平</span></strong><span>）来验证这一前提，否则标准化的结果可能具有偏差或误导性</span></p></li><li><p><span>⚠️ </span><strong><span>转录因子 ChIP-seq</span></strong><span>不建议使用 </span><code>ChIPseqSpikeInFree</code><span>: </span></p><p><span>ChIPseqSpikeInFree 的核心假设是 </span></p><ul><li><p><strong><span>组蛋白修饰在全基因组上分布广泛</span></strong><span>，且有部分区域即使在globel effect时也相对“稳定&quot;</span></p></li><li><p><span>转录因子的 ChIP-seq 通常：</span></p><ul><li><p><span>binding site 很少，且非常局部化（不像 H3K27me3 那样大片区域修饰）</span></p></li><li><p><span>binding 容易受条件剧烈影响（比如某个 TF 只在刺激状态下结合）</span></p></li><li><p><span>没有“相对稳定”的参考区域可供校准</span></p></li></ul></li></ul></li></ul></div><p>&nbsp;</p><p>&nbsp;</p><h3 id='chipseqspikeinfree-code'><span>ChIPseqSpikeInFree code</span></h3><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="R" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="r"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># R包安装</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">#library(devtools)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">#install_github("stjude/ChIPseqSpikeInFree")</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># loading package</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">library</span>(<span class="cm-string">"ChIPseqSpikeInFree"</span>) &nbsp;<span class="cm-comment"># 加载 ChIPseqSpikeInFree 包，用于无 spike-in 的 ChIP-seq 归一化处理</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 设置 BAM 文件所在目录路径</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">bamDir</span> <span class="cm-operator">=</span> <span class="cm-variable">paste0</span>(<span class="cm-variable">projPath</span>, <span class="cm-string">"/alignment/bam"</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">metaData</span> <span class="cm-operator">=</span> <span class="cm-variable">c</span>() &nbsp;<span class="cm-comment"># 初始化存储元数据的对象</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 遍历每个样本名称（格式类似于 "H3K27ac_Treated"），从中提取抗体信息和组别信息</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">for</span>(<span class="cm-variable">hist</span> <span class="cm-keyword">in</span> <span class="cm-variable">sampleList</span>){</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;<span class="cm-variable">histInfo</span> <span class="cm-operator">=</span> <span class="cm-variable">strsplit</span>(<span class="cm-variable">hist</span>, <span class="cm-string">"_"</span>)[[<span class="cm-number">1</span>]] &nbsp;<span class="cm-comment"># 拆分样本名为抗体名和处理组，如 "H3K27ac_Treated" 拆成 "H3K27ac" 和 "Treated"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;<span class="cm-comment"># 构建一条样本数据，包含 BAM 文件名、抗体名、组别信息</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;<span class="cm-variable">metaData</span> <span class="cm-operator">=</span> <span class="cm-variable">data.frame</span>(</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-variable">ID</span> <span class="cm-arg-is">=</span> <span class="cm-variable">paste0</span>(<span class="cm-variable">hist</span>, <span class="cm-string">"_bowtie2.mapped.bam"</span>), &nbsp;<span class="cm-comment"># BAM 文件名</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-variable">ANTIBODY</span> <span class="cm-arg-is">=</span> <span class="cm-variable">histInfo</span>[<span class="cm-number">1</span>], &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 抗体类型（如 H3K27ac）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-variable">GROUP</span> <span class="cm-arg-is">=</span> <span class="cm-variable">histInfo</span>[<span class="cm-number">2</span>] &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 组别信息（如 Treated/Control）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">  ) <span class="cm-operator cm-variable-2">%&gt;%</span> <span class="cm-variable">rbind</span>(<span class="cm-variable">metaData</span>, <span class="cm-variable">.</span>) &nbsp;<span class="cm-comment"># 将该记录添加到元数据表中</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">}</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 将构建好的样本数据表保存为文本文件，用于后续分析</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">write.table</span>(</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;<span class="cm-variable">metaData</span>,</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;<span class="cm-variable">file</span> <span class="cm-arg-is">=</span> <span class="cm-variable">paste0</span>(<span class="cm-variable">projPath</span>, <span class="cm-string">"/alignment/ChIPseqSpikeInFree/metaData.txt"</span>),</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;<span class="cm-variable">row.names</span> <span class="cm-arg-is">=</span> <span class="cm-variable">FALSE</span>,</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;<span class="cm-builtin">quote</span> <span class="cm-arg-is">=</span> <span class="cm-variable">FALSE</span>,</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;<span class="cm-variable">sep</span> <span class="cm-arg-is">=</span> <span class="cm-string">"</span><span class="cm-string-2">\t</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 设置样本数据文件路径</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">metaFile</span> <span class="cm-operator">=</span> <span class="cm-variable">paste0</span>(<span class="cm-variable">projPath</span>, <span class="cm-string">"/alignment/ChIPseqSpikeInFree/metaData.txt"</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 构建 BAM 文件的完整路径向量</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">bams</span> <span class="cm-operator">=</span> <span class="cm-variable">paste0</span>(<span class="cm-variable">projPath</span>, <span class="cm-string">"/alignment/bam/"</span>, <span class="cm-variable">metaData</span><span class="cm-operator cm-dollar">$</span><span class="cm-variable">ID</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 运行 ChIPseqSpikeInFree 主函数进行无 spike-in 的标准化分析</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">ChIPseqSpikeInFree</span>(</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;<span class="cm-variable">bamFiles</span> <span class="cm-arg-is">=</span> <span class="cm-variable">bams</span>, &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 输入 BAM 文件路径</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;<span class="cm-variable">chromFile</span> <span class="cm-arg-is">=</span> <span class="cm-string">"hg38"</span>, &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 参考基因组——hg19, mm9, mm10, hg38 are included in the package.如果不纯在，可以自行提供.chrom.sizes文件</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;<span class="cm-variable">metaFile</span> <span class="cm-arg-is">=</span> <span class="cm-variable">metaFile</span>, &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 样本数据文件</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;<span class="cm-variable">prefix</span> <span class="cm-arg-is">=</span> <span class="cm-variable">paste0</span>(<span class="cm-variable">projPath</span>, <span class="cm-string">"/alignment/ChIPseqSpikeInFree/SpikeInFree_results"</span>) &nbsp;<span class="cm-comment"># 输出结果文件的前缀路径</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 1049px;"></div><div class="CodeMirror-gutters" style="display: none; height: 1049px;"></div></div></div></pre><ul><li><p><a href='https://github.com/stjude/ChIPseqSpikeInFree?tab=readme-ov-file#1-generate-sample_metatxt-tab-delimited-txt-file-with-header-line-as-follows'><span>metaData 格式</span></a></p><figure class='table-figure'><table><thead><tr><th><span>ID</span></th><th><span>ANTIBODY</span></th><th><span>GROUP</span></th></tr></thead><tbody><tr><td><span>WT-H3K27me3.rmdupq1.bam</span></td><td><span>H3K27me3</span></td><td><span>WT</span></td></tr><tr><td><span>K27M-H3K27me3.rmdupq1.bam</span></td><td><span>H3K27me3</span></td><td><span>K27M</span></td></tr><tr><td><span>WT-INPUT.rmdupq1.bam</span></td><td><span>INPUT</span></td><td><span>WT</span></td></tr><tr><td><span>K27M-INPUT.rmdupq1.bam</span></td><td><span>INPUT</span></td><td><span>K27M</span></td></tr></tbody></table></figure></li></ul><ul><li><p><a href='https://github.com/stjude/ChIPseqSpikeInFree#interpretation-of-scaling-factor-table'><span>Interpretation of the ChIPseqSpikeInFree output</span></a><span>.</span></p><figure class='table-figure'><table><thead><tr><th><span>ID</span></th><th><span>GROUP</span></th><th><span>ANTIBODY</span></th><th><span>COLOR</span></th><th><span>QC</span></th><th><span>SF</span></th><th><span>TURNS</span></th></tr></thead><tbody><tr><td><span>WT-H3K27me3.rmdupq1.bam</span></td><td><span>WT</span></td><td><span>H3K27me3</span></td><td><span>grey</span></td><td><span>pass</span></td><td><span>1</span></td><td><span>0.25,0.3817,6.1,0.9523</span></td></tr><tr><td><span>K27M-H3K27me3.rmdupq1.bam</span></td><td><span>K27M</span></td><td><span>H3K27me3</span></td><td><span>green</span></td><td><span>pass</span></td><td><span>5.47</span></td><td><span>0.2,0.3429,34.7,0.9585</span></td></tr><tr><td><span>WT-INPUT.rmdupq1.bam</span></td><td><span>WT</span></td><td><span>INPUT</span></td><td><span>black</span></td><td><span>fail: complete loss, input or poor enrichment</span></td><td><span>NA</span></td><td><span>0.2,0.1860,0.7,0.9396</span></td></tr><tr><td><span>K27M-INPUT.rmdupq1.bam</span></td><td><span>K27M</span></td><td><span>INPUT</span></td><td><span>grey</span></td><td><span>fail: complete loss, input or poor enrichment</span></td><td><span>NA</span></td><td><span>0.15,0.0675,0.35,0.8476</span></td></tr></tbody></table></figure><ul><li><p><strong><span>COLOR</span></strong><span>: defines color for each sample in cumulative distribution curve </span><code>*_distribution.pdf</code></p></li><li><p><strong><span>QC</span></strong><span>: quality control testing. QC failure indicates poor or no enrichment. 说明该样本可能没有富集出有效的组蛋白修饰区域</span></p></li><li><p><strong><span>SF</span></strong><span>: scaling factor. Only sample that </span><strong><span>passes QC</span></strong><span> will be given a </span><strong><span>SF</span></strong><span> and </span><strong><span>NA</span></strong><span> indicates sample with </span><strong><span>poor enrichment.</span></strong></p><ul><li><p><strong><span>较大的 SF 值</span></strong><span> 👉 表示该样本的 </span><strong><span>组蛋白修饰水平较低</span></strong><span>；</span></p></li><li><p><strong><span>较小的 SF 值</span></strong><span> 👉 表示该样本的 </span><strong><span>组蛋白修饰水平较高</span></strong></p></li></ul></li><li><p><strong><span>TURNS</span></strong><span>: 在累计分布图中，识别出两个关键点（Xa, Ya, Xb, Yb）坐标，用于基于斜率的缩放因子计算</span></p></li></ul></li><li><p><a href='https://github.com/stjude/ChIPseqSpikeInFree#how-to-use-chipseqspikein-scaling-factor'><span>How to use ChIPseqSpikeInFree scaling factor.</span></a></p></li></ul><p>&nbsp;</p><h3 id='chipseqspikeinfree-网页版'><span>ChIPseqSpikeInFree 网页版</span></h3><ul><li><p><span>To use the tool, you will need to create a </span><strong><span>DNAnexus</span></strong><span> account at </span><a href='https://platform.dnanexus.com/register?client_id=sjcloudplatform' target='_blank' class='url'>https://platform.dnanexus.com/register?client_id=sjcloudplatform</a><span>. </span></p><div class="md-alert md-alert-note note"><p><span class='md-alert-text md-alert-text-note'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M0 8a8 8 0 1 1 16 0A8 8 0 0 1 0 8Zm8-6.5a6.5 6.5 0 1 0 0 13 6.5 6.5 0 0 0 0-13ZM6.5 7.75A.75.75 0 0 1 7.25 7h1a.75.75 0 0 1 .75.75v2.75h.25a.75.75 0 0 1 0 1.5h-2a.75.75 0 0 1 0-1.5h.25v-2h-.25a.75.75 0 0 1-.75-.75ZM8 6a1 1 0 1 1 0-2 1 1 0 0 1 0 2Z"></path></svg>Note</span><br></p><p><span>🔬</span><strong><span>DNAnexus</span></strong><span> 是一个基于云计算的生物信息学平台，主要为 </span><strong><span>大规模基因组数据分析</span></strong><span> 提供支持，尤其在 </span><strong><span>临床测序、科研项目、药物研发等领域</span></strong><span> 得到广泛应用。</span></p><p><span>它将复杂的生信流程“打包”成易于使用的 </span><strong><span>图形化界面</span></strong><span> 和 </span><strong><span>工作流程（workflow）</span></strong><span>，允许用户上传数据、运行分析、管理结果，</span><strong><span>无需本地部署服务器</span></strong><span> 或自己写大量脚本。</span></p></div><p>&nbsp;</p></li><li><p><span>After logging in DNAnexus, you can create a project , upload your data to your project folder and you can run ChIPseqSpikeInFree [v1.2.3] at </span><a href='https://platform.dnanexus.com/app/ChIPseqSpikeInFree' target='_blank' class='url'>https://platform.dnanexus.com/app/ChIPseqSpikeInFree</a><span> and get results in an hour</span></p></li></ul><p>&nbsp;</p><hr /><h1 id='deeptools'><span>Deeptools</span></h1><p><img src="./picture/059.png" width="800"/>
<img src="./picture/060.png" width="800"/></p><ul><li><p><strong><span>deepTools</span></strong><span> is a suite of python tools particularly developed for the efficient analysis of high-throughput sequencing data, such as ChIP-seq, RNA-seq or MNase-seq</span></p></li><li><p><a href='https://test-argparse-readoc.readthedocs.io/en/latest/index.html'><span>deepTools manual</span></a></p></li><li><p><a href='https://usegalaxy.eu/'><span>deepTools Galaxy server</span></a><span> you use the deepTools within the familiar Galaxy framework without the need to master the command line</span></p></li></ul><p>&nbsp;</p><h2 id='deeptools-installation'><span>Deeptools installation</span></h2><ul><li><p><span>Requirements</span></p><ul><li><p><strong><span>Python 2.7 or Python 3.x</span></strong></p></li><li><p><span>numpy &gt;= 1.8.0</span></p></li><li><p><span>scipy &gt;= 0.17.0</span></p></li><li><p><span>py2bit &gt;= 0.1.0</span></p></li><li><p><span>pyBigWig &gt;= 0.2.1</span></p></li><li><p><span>pysam &gt;= 0.8</span></p></li><li><p><span>matplotlib &gt;= 1.4.0</span></p></li></ul></li><li><p><span>Install</span></p><p><code>conda install -c bioconda deeptools</code></p></li></ul><p>&nbsp;</p><h2 id='bamcoverage'><span>BamCoverage</span></h2><ul><li><p><a href='https://test-argparse-readoc.readthedocs.io/en/latest/content/tools/bamCoverage.html'><span>bamCoverage — deepTools 3.2.1 documentation</span></a></p></li><li><p><span>The </span><code>bamCoverage</code><span> tool, part of the deepTools suite</span></p></li><li><p><span>It’s designed for generating </span><strong><span>coverage tracks</span></strong><span> (bigWig or bedGraph files) from BAM files, for visualization in </span><a href='#genome-browser'><span>Genome browser</span></a></p><ul><li><p><strong><span>Coverage tracks</span></strong><span> visually represent the density of reads across genomic regions, often used in genomics to interpret the depth and intensity of coverage in different regions</span></p></li></ul></li></ul><p><img src="./picture/058.png" width="700"/></p><p>&nbsp;</p><p><strong><span>Parameters</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">bamCoverage</span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==General Parameters==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-b</span> &lt;BAM&gt; &nbsp;<span class="cm-comment"># Input BAM file.</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-o</span> &nbsp;<span class="cm-comment"># Output file name (commonly with .bigWig or .bw extension).</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--outFileFormat</span> <span class="cm-comment"># Format of the output file, typically bigwig or bedgraph.</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-P</span> <span class="cm-comment"># Number of processors to use for multithreading, speeding up processing on larger files,(Default: 1)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==Normalization and Scaling==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--effectiveGenomeSize</span> </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># The effective genome size is the portion of the genome that is mappable.</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># Large fractions of the genome are stretches of NNNN that should be discarded</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment">#  A table of values is available here: http://deeptools.readthedocs.io/en/latest/content/feature/effectiveGenomeSize.html</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--exactScaling</span> </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># bamCoverage calculates the scaling factor based on a sample of the reads, which is faster but less precise in some situations</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># Using exactScaling modifies this behavior to calculate the scaling factor based on all reads in the BAM file</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--normalizeUsing</span> <span class="cm-comment"># Specify normalization method. Options include:</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># RPKM (Reads Per Kilobase Million)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># CPM (Counts Per Million)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># BPM (Bins Per Million mapped reads)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># RPGC (Reads Per Genomic Content)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># None (no normalization)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># Default: None </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--ignoreForNormalization</span> </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># A list of space-delimited chromosome names containing those chromosomes that should be excluded for computing the normalization. </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># This is useful when considering samples with unequal coverage across chromosomes, like male samples. </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># An usage examples is --ignoreForNormalization chrX chrM</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==Coverage Computation==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--binSize</span> &lt;int bp&gt; </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># Bin size (in bp) to use for calculating coverage. Smaller values yield higher resolution but increase file size.</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># Default: 50 &nbsp; &nbsp;</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--region</span> &lt;CHR:START:END&gt; <span class="cm-comment"># Specify a specific genomic region to calculate coverage. &nbsp; </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--smoothLength</span> &lt;int bp&gt; <span class="cm-comment"># The smooth length defines a window, larger than the binSize, to average the number of reads.</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--skipNonCoveredRegions</span> <span class="cm-comment"># Exclude regions with zero coverage from the output file.</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==Read processing options==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--extendReads</span> &lt;int bp&gt; </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># This parameter allows the extension of reads to fragment size</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># NOTE:This feature is generally NOT recommended for spliced-read data, such as RNA-seq </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># *Single-end*: Reads that already exceed this fragment length will not be extended. </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># *Paired-end*: Reads with mates are always extended to match the fragment size defined by the two read mates. </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># mate reads that map too far apart (&gt;4x fragment length) or even map to different chromosomes are treated like single-end reads. &nbsp; &nbsp;</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--ignoreDuplicates</span> &nbsp; </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># If set, reads that have the same orientation and start position will be considered only once. </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># If reads are paired, the mate's position also has to coincide to ignore a read. &nbsp; &nbsp; </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--centerReads</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># adjusts the alignment position of each read to its central location, rather than using the read’s original start or end position &nbsp; &nbsp; </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># *Paired-end*: The read is centered based on the fragment size defined by the positions of the two mates in the pair.</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># *Single-end*: Centering requires a specified fragment length. &nbsp;</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># Be use in Sharp Peak Detection,like: ATAC-seq，chip-seq，CUT&amp;Tag。</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==Filtering Options== &nbsp;</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--minMappingQuality</span> &lt;int&gt;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># Minimum mapping quality score to include reads in the coverage calculation (useful for filtering out low-quality alignments)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--blackListFileName</span> &lt;bed&gt; </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># Specifies regions to exclude, useful for ignoring problematic regions like repetitive elements</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--minFragmentLength</span> &lt;int&gt;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># The minimum fragment length needed for read/pair inclusion. </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># This option is primarily useful in ATACseq experiments, for filtering mono-or di-nucleosome fragments</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># Default: 0 &nbsp; &nbsp; &nbsp; </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--maxFragmentLength</span> &lt;int&gt;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># The maximum fragment length needed for read/pair inclusion. </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># Default: 0 &nbsp; &nbsp;</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==SAM flag option==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--samFlagInclude</span> &lt;int&gt; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># This option includes reads that match a specified flag</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># For example, to get only reads that are the first mate, use a flag of 64 &nbsp; &nbsp;</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--samFlagExclude</span> &lt;int&gt; </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># This option excludes reads with a specified SAM flag &nbsp; &nbsp;</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># For example, to get only reads that map to the forward strand, use --samFlagExclude 16</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==special seq option==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--MNase</span> </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># Designed for MNase-seq, counts only the center 3 nucleotides of fragments between 130-200 bp for nucleosome positioning.</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># To over-write this, use the--minFragmentLength and --maxFragmentLength options, which will default to 130 and 200</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># *NOTE*: Requires paired-end data. A bin size of 1 is recommended. (default: False)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--Offset</span> &lt;int&gt; </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># Specifies the base offset from the start (or end) of each read for counting coverage</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># useful for methods like RiboSeq or GROseq.</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># A value of 1 indicates the first base of the alignment (taking alignmentorientation into account). </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># Likewise, a value of -1 is the last base of the alignment</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># If two values are specified, then they will be used to specify arange of positions</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--filterRNAstrand</span> &lt;forward,reverse&gt; </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># Filters reads based on the strand</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># This is particularly relevant for libraries prepared with strand-specific protocols, like dUTP-based library preparation</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 2426px;"></div><div class="CodeMirror-gutters" style="display: none; height: 2426px;"></div></div></div></pre><p>&nbsp;</p><hr /><h2 id='computematrix'><span>computeMatrix</span></h2><h3 id='computematrix-perface'><span>computeMatrix perface</span></h3><p><a href='https://test-argparse-readoc.readthedocs.io/en/latest/content/tools/computeMatrix.html'><span>computeMatrix — deepTools 3.2.1 documentation</span></a><span> </span></p><p><code>computeMatrix</code><span> 由两种主要使用模式组成</span></p><figure class='table-figure'><table><thead><tr><th><span>模式</span></th><th><span>说明</span></th></tr></thead><tbody><tr><td><code>computeMatrix scale-regions</code></td><td><span>将所有输入的 region 缩放成统一长度（无论原始region长短），然后在统一region上分析信号分布。</span><strong><span>适合比较整个区域（如整个基因体）上的信号模式</span></strong></td></tr><tr><td><code>computeMatrix reference-point</code></td><td><span>以每个region的一个特定点为中心（比如 TSS、TES），向上下游延伸一段距离，对这个范围内的信号进行统计。</span><strong><span>适合分析信号在某个特定位置的分布趋势</span></strong></td></tr></tbody></table></figure><p><img src="./picture/081.png" width="700"/></p><p>&nbsp;</p><p><span>📌 它最终会输出一个 </span><code>.gz</code><span> 的矩阵文件，可以直接用于热图 (</span><code>plotHeatmap</code><span>) 或轮廓图 (</span><code>plotProfile</code><span>) 的绘制。</span></p><p><img src="./picture/082.png" width="700"/></p><p>&nbsp;</p><hr /><h3 id='computematrix-scale-regions'><span>computeMatrix scale-regions</span></h3><p><code>computeMatrix scale-regions</code><span> 的工作流程是这样的：</span></p><ol start='' ><li><p><span>输入 </span><strong><span>bigWig</span></strong><span> 文件 (</span><code>-S</code><span>)，</span></p></li><li><p><span>输入 BED 或 GTF 文件 (</span><code>-R</code><span>)：这些是你关心的区域，通常是一组基因、peak、启动子等。每一行会作为一个</span><strong><span>region</span></strong><span>参与后续统计</span></p></li><li><p><code>--regionBodyLength</code><span> 将所有region</span><strong><span>缩放</span></strong><span>成统一长度</span></p></li><li><p><span>对缩放的region，deepTools 会按照</span><code>--binSize</code><span>进行 bin 的划分。</span></p></li><li><p><strong><span>信号值提取</span></strong><span>：对每个 bin，从 </span><strong><span>bigWig</span></strong><span> 文件提取出该 bin 信号值。组成一个</span><strong><span>Matrix（如下所示）</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="text"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="text"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; Bin1 &nbsp; Bin2 &nbsp; Bin3  ... &nbsp; BinN</span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">region1 &nbsp; 10 &nbsp; &nbsp; 12 &nbsp; &nbsp; 13 &nbsp; &nbsp; &nbsp; &nbsp; 15</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">region2 &nbsp; 0 &nbsp; &nbsp;  0 &nbsp; &nbsp;  0 &nbsp; &nbsp; &nbsp; &nbsp;  0</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">region3 &nbsp; 20 &nbsp; &nbsp; 23 &nbsp; &nbsp; 22 &nbsp; &nbsp; &nbsp; &nbsp; 25</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">...</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 109px;"></div><div class="CodeMirror-gutters" style="display: none; height: 109px;"></div></div></div></pre></li></ol><p>&nbsp;</p><div class="md-alert md-alert-tip tip"><p><span class='md-alert-text md-alert-text-tip'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M8 1.5c-2.363 0-4 1.69-4 3.75 0 .984.424 1.625.984 2.304l.214.253c.223.264.47.556.673.848.284.411.537.896.621 1.49a.75.75 0 0 1-1.484.211c-.04-.282-.163-.547-.37-.847a8.456 8.456 0 0 0-.542-.68c-.084-.1-.173-.205-.268-.32C3.201 7.75 2.5 6.766 2.5 5.25 2.5 2.31 4.863 0 8 0s5.5 2.31 5.5 5.25c0 1.516-.701 2.5-1.328 3.259-.095.115-.184.22-.268.319-.207.245-.383.453-.541.681-.208.3-.33.565-.37.847a.751.751 0 0 1-1.485-.212c.084-.593.337-1.078.621-1.489.203-.292.45-.584.673-.848.075-.088.147-.173.213-.253.561-.679.985-1.32.985-2.304 0-2.06-1.637-3.75-4-3.75ZM5.75 12h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5ZM6 15.25a.75.75 0 0 1 .75-.75h2.5a.75.75 0 0 1 0 1.5h-2.5a.75.75 0 0 1-.75-.75Z"></path></svg>Tip</span><br></p><p><strong><span>缩放原理</span></strong></p><p><span>原始 region 长度：1000bp 或 2000bp</span></p><p><code>--regionBodyLength 1500</code><span>（我们想把所有region统一“压缩/拉伸”为 1500bp 长）</span></p><p><code>--binSize 50</code><span>（缩放后区域每 50bp 一个 bin → 共 1500 / 50 = 30 个 bin）</span></p><ul><li><p><span>✅ 对于长度为 1000bp 的 region：</span></p><ul><li><p><span>每 1000/1500 ≈ </span><strong><span>0.667bp 映射到 1bp</span></strong></p></li><li><p><span>所以第一个 bin（0-50bp）在原始区域中对应 0–33bp（大约）</span></p></li><li><p><span>第二个 bin → 33–66bp，以此类推</span></p></li></ul></li></ul><p><strong><span>→ 整个 1000bp 被拉伸成 1500bp 的格式</span></strong></p></div><p>&nbsp;</p><p><strong><span>parameters</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">computeMatrix scale-regions</span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># == input file==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--regionsFileName</span> &nbsp; &nbsp;<span class="cm-comment"># File name or names, in BED or GTF format</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 每个文件会被当作一个 group 来处理</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" class="cm-tab-wrap-hack" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 也可以在一个 BED 文件内部用 # 把区域分group：BED文件中每遇到一个 #，就将前面的区域归为一个group。<span class="cm-tab" role="presentation" cm-text="	"> </span><span class="cm-tab" role="presentation" cm-text="	">    </span></span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--scoreFileName</span> &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 指定一个或多个 bigWig 文件</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--blackListFileName</span> &nbsp;<span class="cm-comment"># 指定一个 BED 文件，这个文件里列出了你想从分析中排除的region</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># == output file==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--outFileName</span><span class="cm-tab" role="presentation" cm-text="	">   </span> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 输出 .gz 矩阵文件，供 plotHeatmap 使用</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--outFileNameMatrix</span><span class="cm-tab" role="presentation" cm-text="	"> </span> &nbsp; &nbsp;<span class="cm-comment"># 输出可读的矩阵（tab 分隔）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--outFileSortedRegions</span><span class="cm-tab" role="presentation" cm-text="	">  </span><span class="cm-comment"># 排序后的region保存到一个 BED 文件中</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># == Matrix setting==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--binSize</span><span class="cm-tab" role="presentation" cm-text="	">   </span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 设置bin大小（默认10），决定矩阵的 resolution</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--averageTypeBins</span><span class="cm-tab" role="presentation" cm-text="	">   </span> &nbsp; <span class="cm-comment"># bin内信号统计量计算方式：mean(deafult), median, min, max, std, sum</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--scale</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 将所有信号值乘以一个缩放因子（scale factor）(Default: 1)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--missingDataAsZero</span><span class="cm-tab" role="presentation" cm-text="	"> </span> &nbsp; <span class="cm-comment"># 将 Na 数据视为0 (default: False)。在default下，Na在heatmap中为黑色显示</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--skipZeros</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 如果某个 region 的 所有 bin 都是 0，是否跳过该region。(default: False，不跳过)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--minThreshold</span> &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 如果某个 region中有任意一个 bin的数值小于或等于这个阈值，那么这个区域会被跳过。(Default: None)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--maxThreshold</span> &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 如果某个 region中有任意一个 bin的数值大于或等于这个阈值，那么这个区域会被跳过。(Default: None)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==region scale setting==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--regionBodyLength</span><span class="cm-tab" role="presentation" cm-text="	">  </span> &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 将所有region缩放成统一长度 (Default: 1000)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--beforeRegionStartLength</span><span class="cm-tab" role="presentation" cm-text="	">   </span> <span class="cm-comment"># 分析region向前延伸的长度 (Default: 0)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--afterRegionStartLength</span><span class="cm-tab" role="presentation" cm-text="	">    </span> <span class="cm-comment"># 分析region向后延伸的长度 (Default: 0)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--unscaled5prime</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 设置区域 5' 端保留多少碱基不进行缩放，适用于查看如 TSS 附近的真实信号(Default: 0)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--unscaled3prime</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 设置区域 3' 端保留多少碱基不进行缩放，适用于查看如 TES 附近的真实信号(Default: 0)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==region sort== &nbsp;</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--sortRegions</span><span class="cm-tab" role="presentation" cm-text="	">   </span> &nbsp; &nbsp;<span class="cm-comment"># 是否对region排序</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># optional parameters:descend（降序排列）、ascend（升序排列）、no （不排序，保持原始顺序）(deafult)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 使用 plotHeatmap 等工具来可视化结果，这些工具会自动覆盖排序设置，因此该参数在这些情况下没有效果</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; <span class="cm-attribute">--sortUsing</span><span class="cm-tab" role="presentation" cm-text="	">    </span> &nbsp; &nbsp;<span class="cm-comment"># 排序标准：mean,median,max,min,sum,region_length</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; <span class="cm-attribute">--sortUsingSamples</span><span class="cm-tab" role="presentation" cm-text="	"> </span>&lt;SORTUSINGSAMPLES&gt; &nbsp; <span class="cm-comment"># SORTUSINGSAMPLES是一个 样本编号列表，用于指定哪些样本将参与排序过程 (deafult:全部样本进行排序)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># == GTF / BED12 options ==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-tab" role="presentation" cm-text="	">    </span><span class="cm-attribute">--metagene</span> &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 不开启：以每个转录本从起始到终止（也就是 BED 的第2列到第3列）为整体region(default: False)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-tab" role="presentation" cm-text="	">    </span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment">#  开启： 则会把所有 exon 合并成一个连续region，然后在这个合并后的region上进行分析。适合用来看转录本结构下的富集情况</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; <span class="cm-attribute">--transcriptID</span> &nbsp; &nbsp;<span class="cm-comment"># 当你使用 GTF 文件时，这个参数指定哪种 feature 类型被视为transcripts (Default: transcript)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; <span class="cm-attribute">--exonID</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 当你使用 GTF 文件时，这个参数指定哪种 feature 类型被视为exon (Default: exon)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; <span class="cm-attribute">--transcript_id_designator</span> <span class="cm-comment"># GTF 格式的最后一列(attributes),那个key作为id(Default: transcript_id)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># == 文件标签，这些标签会用于后续的可视化，作为图列标签 ==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--samplesLabel</span> &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 为每个文件设置自定义的标签名称(default: None)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--smartLabels</span> &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 自动地把输入文件名去掉路径和扩展名，作为标签(default: False)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==Draw picture setting==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--startLabel</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 设置你在绘图中 region 起始位置的标签文字 (Default: TSS)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--endLabel</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 设置你在绘图中 region 结束位置的标签文字 (Default: TES)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==Log display==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-tab" role="presentation" cm-text="	">    </span><span class="cm-attribute">--quiet</span> &nbsp; &nbsp; <span class="cm-comment"># 让程序“安静”地运行，不显示任何警告或中间处理信息 (default: False)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-tab" role="presentation" cm-text="	">    </span><span class="cm-attribute">--verbose</span> &nbsp; <span class="cm-comment"># 让程序“话多”，显示非常详细的运行信息 (default: False)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==performance optimization==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--numberOfProcessors</span><span class="cm-tab" role="presentation" cm-text="	">    </span><span class="cm-comment"># 并行处理线程数</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># optional parameters:&lt;numb&gt;, max, max/2 &nbsp;</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># == DeepBlue parameters（仅适用于使用远程的 bedGraph/wig 文件）==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--deepBlueURL</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 指定 deepBlue 服务器的 URL 地址,不需要更改(default: http://deepblue.mpi-inf.mpg.de/xmlrpc)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--userKey</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 指定访问 deepBlue 数据时用的用户密钥，(default:anonymous_key，足够用于访问公开数据集)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--deepBlueTempDir</span> &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 指定 deepBlue 数据预加载过程中临时文件保存的位置(default: None)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--deepBlueKeepTemp</span> &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 保留从 deepBlue 下载生成的临时 bigWig 文件(default:False，不保留)</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 2142px;"></div><div class="CodeMirror-gutters" style="display: none; height: 2142px;"></div></div></div></pre><p>&nbsp;</p><p><strong><span>Example</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">computeMatrix scale-regions \</span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;<span class="cm-attribute">--regionsFileName</span> testFiles/genes.bed \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 输入region文件（BED 格式），定义你感兴趣的基因区域</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;<span class="cm-attribute">--scoreFileName</span> testFiles/log2ratio_H3K4Me3_chr19.bw  \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 输入信号文件（bigWig 格式）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;<span class="cm-attribute">--beforeRegionStartLength</span> <span class="cm-number">3000</span> <span class="cm-attribute">--afterRegionStartLength</span> <span class="cm-number">3000</span> \ &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 上游3000bp，下游3000bp</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;<span class="cm-attribute">--regionBodyLength</span> <span class="cm-number">5000</span> \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 将基因体统一缩放到5000bp</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;<span class="cm-attribute">--skipZeros</span> \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 跳过全为0的region</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;<span class="cm-attribute">--outFileName</span> matrix1_H3K4me3_l2r_TSS.gz \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 输出的压缩矩阵文件（供绘图使用）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;<span class="cm-attribute">--outFileNameMatrix</span> matrix2_multipleBW_l2r_twoGroups_scaled.tab \ &nbsp; <span class="cm-comment"># 输出可读的文本矩阵</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;<span class="cm-attribute">--outFileSortedRegions</span> regions2_multipleBW_l2r_twoGroups_genes.bed &nbsp;<span class="cm-comment"># 输出排序后的bed文件</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 197px;"></div><div class="CodeMirror-gutters" style="display: none; height: 197px;"></div></div></div></pre><p>&nbsp;</p><hr /><h3 id='computematrix-referencepoint'><span>computeMatrix referencePoint</span></h3><p><code>omputeMatrix reference-point</code><span> 的工作流程如下：</span></p><ol start='' ><li><p><strong><span>输入 bigWig 文件 (</span><code>-S</code><span>)</span></strong></p></li><li><p><strong><span>输入 BED 或 GTF 文件 (</span><code>-R</code><span>)</span></strong><span>：这些是你感兴趣的区域（如 TSS、peak 等），每一行定义一个 </span><strong><span>region</span></strong><span>。</span></p></li><li><p><strong><span>设定参考点 (</span><code>--referencePoint</code><span>)</span></strong></p><p><span>定义参考位置，如 </span><code>TSS</code><span>（转录起始点）、</span><code>TES</code><span>（转录终止点）或 </span><code>center</code><span>（区域中点）。所有信号的对齐都基于这个位置。</span></p></li><li><p><strong><span>定义 flanking 区域 (</span><code>--beforeRegionStartLength</code><span>, </span><code>--afterRegionStartLength</code><span>)</span></strong></p><p><code>--beforeRegionStartLength</code><span>：参考点</span><strong><span>上游</span></strong><span>多少 bp（before）</span>
<span> </span><code>--afterRegionStartLength</code><span>：参考点</span><strong><span>下游</span></strong><span>多少 bp（after）</span>
<span> 合起来就是对每个 region，提取 </span><code>referencePoint ± X bp</code><span> 的窗口。</span></p></li><li><p><strong><span>bin 划分 (</span><code>--binSize</code><span>)</span></strong></p><p><span>将上述窗口分割为固定大小的 bin。</span></p></li><li><p><strong><span>信号值提取</span></strong><span>:对每个 bin，从对应 bigWig 文件中提取信号强度，计算平均值，构成一个 </span><strong><span>Matrix（如下所示）</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="text"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="text"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; Bin1 &nbsp; Bin2 &nbsp; Bin3  ... &nbsp; BinN</span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">region1 &nbsp; 10 &nbsp; &nbsp; 12 &nbsp; &nbsp; 13 &nbsp; &nbsp; &nbsp; &nbsp; 15</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">region2 &nbsp; 0 &nbsp; &nbsp;  0 &nbsp; &nbsp;  0 &nbsp; &nbsp; &nbsp; &nbsp;  0</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">region3 &nbsp; 20 &nbsp; &nbsp; 23 &nbsp; &nbsp; 22 &nbsp; &nbsp; &nbsp; &nbsp; 25</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">...</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 109px;"></div><div class="CodeMirror-gutters" style="display: none; height: 109px;"></div></div></div></pre></li></ol><p>&nbsp;</p><p><strong><span>computeMatrix referencePoint parameter</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">computeMatrix reference-point</span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># == input file==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--regionsFileName</span> &nbsp; &nbsp;<span class="cm-comment"># File name or names, in BED or GTF format</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 每个文件会被当作一个 group 来处理</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" class="cm-tab-wrap-hack" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 也可以在一个 BED 文件内部用 # 把区域分group：BED文件中每遇到一个 #，就将前面的区域归为一个group。<span class="cm-tab" role="presentation" cm-text="	"> </span><span class="cm-tab" role="presentation" cm-text="	">    </span></span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--scoreFileName</span> &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 指定一个或多个 bigWig 文件</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--blackListFileName</span> &nbsp;<span class="cm-comment"># 指定一个 BED 文件，这个文件里列出了你想从分析中排除的region</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># == output file==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--outFileName</span><span class="cm-tab" role="presentation" cm-text="	">   </span> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 输出 .gz 矩阵文件，供 plotHeatmap 使用</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--outFileNameMatrix</span><span class="cm-tab" role="presentation" cm-text="	"> </span> &nbsp; &nbsp;<span class="cm-comment"># 输出可读的矩阵（tab 分隔）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--outFileSortedRegions</span><span class="cm-tab" role="presentation" cm-text="	">  </span><span class="cm-comment"># 排序后的region保存到一个 BED 文件中</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># == Matrix setting==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--binSize</span><span class="cm-tab" role="presentation" cm-text="	">   </span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 设置bin大小（默认10），决定矩阵的 resolution</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--averageTypeBins</span><span class="cm-tab" role="presentation" cm-text="	">   </span> &nbsp; <span class="cm-comment"># bin内信号统计量计算方式：mean(deafult), median, min, max, std, sum</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--scale</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 将所有信号值乘以一个缩放因子（scale factor）(Default: 1)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--missingDataAsZero</span><span class="cm-tab" role="presentation" cm-text="	"> </span> &nbsp; <span class="cm-comment"># 将 Na 数据视为0 (default: False)。在default下，Na在heatmap中为黑色显示</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--skipZeros</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 如果某个 region 的 所有 bin 都是 0，是否跳过该region。(default: False，不跳过)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--minThreshold</span> &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 如果某个 region中有任意一个 bin的数值小于或等于这个阈值，那么这个区域会被跳过。(Default: None)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--maxThreshold</span> &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 如果某个 region中有任意一个 bin的数值大于或等于这个阈值，那么这个区域会被跳过。(Default: None)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==reference-point setting==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--referencePoint</span><span class="cm-tab" role="presentation" cm-text="	">    </span> &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 指定参考点，TSS、TES或center(Default: TSS)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--beforeRegionStartLength</span><span class="cm-tab" role="presentation" cm-text="	">   </span> <span class="cm-comment"># 参考点上游提取长度，即在参考点前延伸多少 bp。(Default: 500)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--afterRegionStartLength</span><span class="cm-tab" role="presentation" cm-text="	">    </span> <span class="cm-comment"># 参考点下游提取长度，即在参考点后延伸多少 bp。(Default: 1500)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--nanAfterEnd</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># True，任何位于所定义的region结束后（即参考点 + afterRegionStartLength 距离后）的信号值会被标记为 NaN</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># False，即不会丢弃region结束后的信号数据。即便在绘图时，region结束后的数据仍会被显示 (deafult)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==region sort== &nbsp;</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--sortRegions</span><span class="cm-tab" role="presentation" cm-text="	">   </span> &nbsp; &nbsp;<span class="cm-comment"># 是否对region排序</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># optional parameters:descend（降序排列）、ascend（升序排列）、no （不排序，保持原始顺序）(deafult)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 使用 plotHeatmap 等工具来可视化结果，这些工具会自动覆盖排序设置，因此该参数在这些情况下没有效果</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; <span class="cm-attribute">--sortUsing</span><span class="cm-tab" role="presentation" cm-text="	">    </span> &nbsp; &nbsp;<span class="cm-comment"># 排序标准：mean,median,max,min,sum,region_length</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; <span class="cm-attribute">--sortUsingSamples</span><span class="cm-tab" role="presentation" cm-text="	"> </span>&lt;SORTUSINGSAMPLES&gt; &nbsp; <span class="cm-comment"># SORTUSINGSAMPLES是一个 样本编号列表，用于指定哪些样本将参与排序过程 (deafult:全部样本进行排序)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># == GTF / BED12 options ==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-tab" role="presentation" cm-text="	">    </span><span class="cm-attribute">--metagene</span> &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 不开启：以每个转录本从起始到终止（也就是 BED 的第2列到第3列）为整体region(default: False)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-tab" role="presentation" cm-text="	">    </span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment">#  开启： 则会把所有 exon 合并成一个连续region，然后在这个合并后的region上进行分析。适合用来看转录本结构下的富集情况</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; <span class="cm-attribute">--transcriptID</span> &nbsp; &nbsp;<span class="cm-comment"># 当你使用 GTF 文件时，这个参数指定哪种 feature 类型被视为transcripts (Default: transcript)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; <span class="cm-attribute">--exonID</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 当你使用 GTF 文件时，这个参数指定哪种 feature 类型被视为exon (Default: exon)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; <span class="cm-attribute">--transcript_id_designator</span> <span class="cm-comment"># GTF 格式的最后一列(attributes),那个key作为id(Default: transcript_id)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># == 文件标签，这些标签会用于后续的可视化，作为图列标签 ==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--samplesLabel</span> &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 为每个文件设置自定义的标签名称(default: None)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--smartLabels</span> &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 自动地把输入文件名去掉路径和扩展名，作为标签(default: False)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==Log display==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-tab" role="presentation" cm-text="	">    </span><span class="cm-attribute">--quiet</span> &nbsp; &nbsp; <span class="cm-comment"># 让程序“安静”地运行，不显示任何警告或中间处理信息 (default: False)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-tab" role="presentation" cm-text="	">    </span><span class="cm-attribute">--verbose</span> &nbsp; <span class="cm-comment"># 让程序“话多”，显示非常详细的运行信息 (default: False)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># ==performance optimization==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--numberOfProcessors</span><span class="cm-tab" role="presentation" cm-text="	">    </span><span class="cm-comment"># 并行处理线程数</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># optional parameters:&lt;numb&gt;, max, max/2 &nbsp;</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># == DeepBlue parameters（仅适用于使用远程的 bedGraph/wig 文件）==</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--deepBlueURL</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 指定 deepBlue 服务器的 URL 地址,不需要更改(default: http://deepblue.mpi-inf.mpg.de/xmlrpc)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--userKey</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 指定访问 deepBlue 数据时用的用户密钥，(default:anonymous_key，足够用于访问公开数据集)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--deepBlueTempDir</span> &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 指定 deepBlue 数据预加载过程中临时文件保存的位置(default: None)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--deepBlueKeepTemp</span> &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 保留从 deepBlue 下载生成的临时 bigWig 文件(default:False，不保留)</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 2011px;"></div><div class="CodeMirror-gutters" style="display: none; height: 2011px;"></div></div></div></pre><p>&nbsp;</p><p><strong><span>Example</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">computeMatrix reference-point </span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--referencePoint</span> TSS \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 参考点选择 TSS</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--beforeRegionStartLength</span> <span class="cm-number">3000</span> <span class="cm-attribute">--afterRegionStartLength</span> <span class="cm-number">3000</span> \ &nbsp; &nbsp; <span class="cm-comment"># 上游3000bp，下游3000bp</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--regionsFileName</span> testFiles/genes.bed \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 输入region文件（BED 格式），定义你感兴趣的基因区域</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--scoreFileName</span> testFiles/log2ratio_H3K4Me3_chr19.bw  \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 输入信号文件（bigWig 格式）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--skipZeros</span> \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 跳过全为0的region </span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--outFileName</span> matrix1_H3K4me3_l2r_TSS.gz \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 输出的压缩矩阵文件（供绘图使用）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--outFileSortedRegions</span> regions1_H3K4me3_l2r_genes.bed &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 输出排序后的bed文件</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 175px;"></div><div class="CodeMirror-gutters" style="display: none; height: 175px;"></div></div></div></pre><p>&nbsp;</p><hr /><h2 id='plotheatmap'><span>plotHeatmap</span></h2><ul><li><p><a href='https://test-argparse-readoc.readthedocs.io/en/latest/content/tools/plotHeatmap.html'><span>plotHeatmap — deepTools 3.2.1 documentation</span></a></p></li><li><p><code>plotHeatmap</code><span> does not change the values that </span><code>computeMatrix</code><span> calculated, it simply translates them into </span><strong><span>heatmaps</span></strong><span> and </span><strong><span>summary plots</span></strong></p></li><li><p><span>输入使用 </span><code>computeMatrix</code><span> 生成</span><strong><span>.gz 矩阵文件</span></strong></p></li><li><p><code>plotHeatmap</code><span> 默认由3个图构成summary plot (上方)+ heatmap(下方)+ colorba(右方)</span></p><p><img src="./picture/083.png" width="500"/></p></li></ul><p>&nbsp;</p><p>&nbsp;</p><p><strong><span>📌 必填参数（Required arguments）</span></strong></p><figure class='table-figure'><table><thead><tr><th><span>参数</span></th><th><span>含义</span></th></tr></thead><tbody><tr><td><code>--matrixFile</code><span> 或 </span><code>-m</code></td><td><span>输入使用 </span><code>computeMatrix</code><span> 工具生成的矩阵文件。</span><strong><span>这是绘图的基础数据。</span></strong></td></tr><tr><td><code>--outFileName</code><span> 或 </span><code>-out</code><span>, </span><code>-o</code></td><td><span>输出图片的文件名。根据文件名后缀自动决定格式（如 </span><code>.png</code><span>, </span><code>.pdf</code><span>, </span><code>.svg</code><span> 等）</span></td></tr></tbody></table></figure><p>&nbsp;</p><p><strong><span>📤 输出选项（Output options）</span></strong></p><figure class='table-figure'><table><thead><tr><th><span>参数</span></th><th><span>含义</span></th></tr></thead><tbody><tr><td><code>--outFileSortedRegions</code></td><td><span>输出排序后区域的 BED 文件，可用于之后再用相同排序绘图。</span></td></tr><tr><td><code>--outFileNameMatrix</code></td><td><span>输出heatmap下方实际使用的矩阵数据（</span><code>.gz</code><span> 格式），便于重复使用或自定义分析。</span></td></tr><tr><td><code>--interpolationMethod</code></td><td><span>插值方式，可让图像在列数很多时更平滑。可选：</span><code>nearest</code><span>, </span><code>bilinear</code><span>, </span><code>bicubic</code><span>, </span><code>gaussian</code><span> 等。</span><a href='https://matplotlib.org/2.0.2/examples/images_contours_and_fields/interpolation_methods.html'><span>详见点此</span></a><span> (default: auto)</span></td></tr><tr><td><code>--dpi</code></td><td><span>设置图片分辨率（单位为 DPI）(default: 200)</span></td></tr></tbody></table></figure><p>&nbsp;</p><p><strong><span>🔗 聚类参数（Clustering）</span></strong></p><figure class='table-figure'><table><thead><tr><th><span>参数</span></th><th><span>含义</span></th></tr></thead><tbody><tr><td><code>--kmeans</code></td><td><span>使用 K-means 算法对区域进行聚类，值为聚类个数（如 </span><code>--kmeans 3</code><span> 表示聚 3 类）。(default: None)</span></td></tr><tr><td><code>--hclust</code></td><td><span>使用层次聚类（hierarchical clustering）进行聚类，值为聚类个数，使用 </span><code>ward linkage</code><span> 方法来构建聚类树。(default: None)</span></td></tr><tr><td><code>--silhouette</code></td><td><span>若使用聚类，可计算每个区域的 </span><strong><span>silhouette</span></strong><span> 分数来评估聚类效果。可写入 BED 文件末列。(default: False)</span></td></tr></tbody></table></figure><p>&nbsp;</p><div class="md-alert md-alert-tip tip"><p><span class='md-alert-text md-alert-text-tip'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M8 1.5c-2.363 0-4 1.69-4 3.75 0 .984.424 1.625.984 2.304l.214.253c.223.264.47.556.673.848.284.411.537.896.621 1.49a.75.75 0 0 1-1.484.211c-.04-.282-.163-.547-.37-.847a8.456 8.456 0 0 0-.542-.68c-.084-.1-.173-.205-.268-.32C3.201 7.75 2.5 6.766 2.5 5.25 2.5 2.31 4.863 0 8 0s5.5 2.31 5.5 5.25c0 1.516-.701 2.5-1.328 3.259-.095.115-.184.22-.268.319-.207.245-.383.453-.541.681-.208.3-.33.565-.37.847a.751.751 0 0 1-1.485-.212c.084-.593.337-1.078.621-1.489.203-.292.45-.584.673-.848.075-.088.147-.173.213-.253.561-.679.985-1.32.985-2.304 0-2.06-1.637-3.75-4-3.75ZM5.75 12h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5ZM6 15.25a.75.75 0 0 1 .75-.75h2.5a.75.75 0 0 1 0 1.5h-2.5a.75.75 0 0 1-.75-.75Z"></path></svg>Tip</span><br></p><p><strong><span>K-means 聚类与层次聚类？</span></strong></p><figure class='table-figure'><table><thead><tr><th><span>方法</span></th><th><span>适合场景</span></th></tr></thead><tbody><tr><td><span>层次聚类</span></td><td><span>小规模数据（区域数量 &lt; 1000），追求树状结构关系</span></td></tr><tr><td><span>K-means 聚类</span></td><td><span>中大规模数据，速度更快，效果可接受</span></td></tr></tbody></table></figure></div><p>&nbsp;</p><p>&nbsp;</p><p><strong><span>🧩排序选项（sort arguments）</span></strong></p><figure class='table-figure'><table><thead><tr><th><strong><span>参数</span></strong></th><th><span>含义</span></th></tr></thead><tbody><tr><td><code>--sortRegions</code></td><td><span>排序区域的方式，默认是 </span><code>descend</code><span>（按均值从高到低），也可设为 </span><code>ascend</code><span>, </span><code>no</code><span>, </span><code>keep</code><span>。</span></td></tr><tr><td><code>--sortUsing</code></td><td><span>选择用于排序的统计方法：</span><code>mean</code><span>, </span><code>median</code><span>, </span><code>max</code><span>, </span><code>min</code><span>, </span><code>sum</code><span>, </span><code>region_length</code><span>。</span></td></tr><tr><td><code>--sortUsingSamples</code></td><td><span>指定用于排序的样本索引（如 </span><code>--sortUsingSamples 1 3</code><span>）(default: None)</span></td></tr><tr><td><code>--clusterUsingSamples</code></td><td><span>指定用于聚类的样本编号（默认使用所有样本）。</span></td></tr></tbody></table></figure><p>&nbsp;</p><p><span>🎨</span><strong><span>颜色选项（color arguments)</span></strong></p><figure class='table-figure'><table><thead><tr><th><span>参数</span></th><th><span>含义</span></th></tr></thead><tbody><tr><td><code>--missingDataColor</code></td><td><span>指定缺失数据的颜色，默认是黑色（black）。可用灰度值</span><code>0–1</code><span>或颜色名 </span><code>#hex</code><span>。</span></td></tr><tr><td><code>--colorMap</code></td><td><span>指定heatmap使用的颜色图谱（colormap）</span><a href='https://matplotlib.org/2.0.2/users/colormaps.html'><span>Choosing Colormaps </span></a></td></tr><tr><td><code>--alpha</code></td><td><span>heatmap透明度，0 表示全透明，1 表示完全不透明 (default: 1.0)</span></td></tr><tr><td><code>--colorList</code></td><td><span>自定义颜色过渡列表，例如 </span><code>--colorList black,yellow,blue</code><span>，会覆盖 </span><code>--colorMap</code><span> 设置。</span></td></tr><tr><td><code>--colorNumber</code></td><td><span>用于指定颜色过渡的级数，仅在设置 </span><code>--colorList</code><span> 时有效。</span></td></tr></tbody></table></figure><p>&nbsp;</p><p><span>🎢</span><strong><span>标签与坐标轴选项（Label and axis argument）</span></strong></p><figure class='table-figure'><table><thead><tr><th><span>参数</span></th><th><span>含义</span></th></tr></thead><tbody><tr><td><code>--xAxisLabel XAXISLABEL</code></td><td><span>设置 x 轴标签的描述。默认值为 &quot;gene distance (bp)&quot;。</span></td></tr><tr><td><code>--startLabel STARTLABEL</code></td><td><span>仅在 </span><code>scale-regions</code><span> 模式下使用，设置区域起始位置的标签。默认为 TSS 。</span></td></tr><tr><td><code>--endLabel ENDLABEL</code></td><td><span>仅在 </span><code>scale-regions</code><span> 模式下使用，设置区域结束位置的标签。默认为 TES 。</span></td></tr><tr><td><code>--refPointLabel REFPOINTLABEL</code></td><td><span>仅在 </span><code>reference-point</code><span> 模式下使用，设置参考点的标签。默认为所选参考点。</span></td></tr><tr><td><code>--labelRotation LABEL_ROTATION</code></td><td><span>设置 x 轴标签的旋转角度，正值为逆时针旋转。(default: 0.0)</span></td></tr><tr><td><code>--regionsLabel REGIONSLABEL</code></td><td><span>设置heatmap 中展示的区域标签，如果有多个区域，需用空格分隔。(default: None)</span></td></tr><tr><td><code>--samplesLabel SAMPLESLABEL</code></td><td><span>设置展示的样本标签，默认使用文件名。如果标签包含空格，需用引号括起来。(default: None)</span></td></tr><tr><td><code>--plotTitle PLOTTITLE</code></td><td><span>设置图表的标题。(default: None)</span></td></tr><tr><td><code>--yAxisLabel YAXISLABEL</code></td><td><span>设置 y 轴标签。(default: None)</span></td></tr><tr><td><code>--yMin YMIN</code></td><td><span>设置 y 轴的最小值，可以为多个值，为每个图表单独设置。(default: None)</span></td></tr><tr><td><code>--yMax YMAX</code></td><td><span>设置 y 轴的最大值，可以为多个值，为每个图表单独设置。(default: None)</span></td></tr><tr><td><code>--legendLocation</code></td><td><span>设置图例的位置，常用的位置选项包括 </span><code>best</code><span>, </span><code>upper-right</code><span>, </span><code>lower-left</code><span>, </span><code>none</code><span>等。默认值为 </span><code>best</code><span>。</span></td></tr></tbody></table></figure><p>&nbsp;</p><p><span>🔧 </span><strong><span>其他常用选项（Optional arguments）</span></strong></p><figure class='table-figure'><table><thead><tr><th><span>参数</span></th><th><span>含义</span></th></tr></thead><tbody><tr><td><code>--plotType</code></td><td><span>设置 summary plot 的样式，如 </span><code>lines</code><span>（deafult）（绘制平均曲线），</span><code>fill</code><span>（用半透明颜色填充曲线下方区域）, </span><code>se</code><span>（显示标准误差区域）, </span><code>std</code><span>（显示标准差区域）</span></td></tr><tr><td><code>--linesAtTickMarks</code></td><td><span>在heatmap上从 </span><strong><span>所有 tick marks（坐标轴上的刻度线）</span></strong><span> 处，向下绘制 </span><strong><span>虚线（dashed lines）</span></strong><span>(default: False)</span></td></tr><tr><td><code>--averageTypeSummaryPlot</code></td><td><span>设置 summary plot 的统计方式，默认是 </span><code>mean</code><span>，也可用 </span><code>median</code><span>, </span><code>max</code><span>, </span><code>min</code><span>, </span><code>std</code><span>, </span><code>sum</code><span> 等。</span></td></tr><tr><td><code>--zMin</code><span>, </span><code>--zMax</code></td><td><span>分别设置heatmap显示的最小和最大数值。支持多个值用于多图绘制。支持 </span><code>&quot;auto&quot;</code><span> 自动选择百分位数。(default: None)</span></td></tr><tr><td><code>--heatmapHeight</code><span>, </span><code>--heatmapWidth</code></td><td><span>图像尺寸，单位为 cm。默认是 28 cm × 4 cm。</span></td></tr><tr><td><code>--whatToShow</code></td><td><span>指定图像显示哪些内容，(deafult: summary plot + heatmap+ colorbar）。</span></td></tr><tr><td><code>--boxAroundHeatmaps</code></td><td><span>是否在heatmap周围加黑框，默认是 </span><code>yes</code><span>，可设为 </span><code>no</code><span> 取消。</span></td></tr><tr><td><code>--perGroup</code></td><td><span>控制是否按组别绘制所有样本的数据。默认为按样本绘制所有组别数据(False)。</span></td></tr><tr><td><code>--verbose</code></td><td><span>启用此选项时，显示警告信息和其他附加信息。(default: False)</span></td></tr></tbody></table></figure><p>&nbsp;</p><p><strong><span>Example</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation"><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">plotHeatmap <span class="cm-attribute">-m</span> matrix.mat.gz \</span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp;<span class="cm-attribute">-out</span> ExampleHeatmap1.png \</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 44px;"></div><div class="CodeMirror-gutters" style="display: none; height: 44px;"></div></div></div></pre><p>&nbsp;</p><hr /><h2 id='deepblue'><span>DeepBlue </span></h2><ul><li><p><a href='https://github.com/MPIIComputationalEpigenetics/DeepBlue?tab=readme-ov-file'><span>MPIIComputationalEpigenetics/DeepBlue: DeepBlue Epigenomic Data Server</span></a><span>（这个好像挂掉了）</span></p></li><li><p><span>DeepBlue 是一个生物信息学的在线平台，提供了多种基因组数据和生物标志物的数据集。用户可以在 DeepBlue 上访问、查询和下载不同类型的基因组数据，通常包括类似 </span><strong><span>bigWig</span></strong><span> 和 </span><strong><span>BED</span></strong><span> 格式的文件，这些文件存储了各类组学数据（如 ChIP-seq、RNA-seq 等）的信号。</span></p></li><li><p><span>如果你不想下载数据到本地，可以直接使用 </span><code>computeMatrix</code><span> 或 </span><code>plotProfile</code><span> 来读取和分析 deepBlue 上的数据</span></p></li></ul><p>&nbsp;</p><hr /><h1 id='convert-bam-to-bigwig-and-bigwig-operate'><span>Convert Bam to BigWIG and BigWIG Operate</span></h1><h2 id='convert-bam-to-bigwig'><span>Convert Bam to BigWIG </span></h2><ul><li><p><span>We use </span><a href='#samtools-sort'><span>Samtools sort</span></a><span> to sort the BAM file</span></p></li><li><p><span>We use </span><a href='#bamcoverage'><span>BamCoverage of deeptools</span></a><span> to convert BAM to BigWIG</span></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-meta">#!/bin/bash</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># create bigwig folder</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-builtin">mkdir</span> <span class="cm-attribute">-p</span> ./bigwig</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Define the threads</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-def">threads</span><span class="cm-operator">=</span><span class="cm-number">10</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Define the variables</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-def">samples</span><span class="cm-operator">=</span>(<span class="cm-string">"KCHIP2"</span> <span class="cm-string">"KCHIP2-OE"</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># BAM directory</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-def">bigwigDir</span><span class="cm-operator">=</span><span class="cm-string">"bigwig"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># BAM directory</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-def">bamDir</span><span class="cm-operator">=</span><span class="cm-string">"deduplicated.bam"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Loop over the samples</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">for</span> sample <span class="cm-keyword">in</span> <span class="cm-string">"</span><span class="cm-def">${samples[@]}</span><span class="cm-string">"</span>; <span class="cm-keyword">do</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># BAM file path</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-def">bam_file</span><span class="cm-operator">=</span><span class="cm-string">"</span><span class="cm-def">${bamDir}</span><span class="cm-string">/filter</span><span class="cm-def">${sample}</span><span class="cm-string">deduplicated.bam"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-def">sort_bam_file</span><span class="cm-operator">=</span><span class="cm-string">"</span><span class="cm-def">${bamDir}</span><span class="cm-string">/filter</span><span class="cm-def">${sample}</span><span class="cm-string">deduplicated.sort.bam"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># Sort the BAM file (skip sorting if sorted already)</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-keyword">if</span> [[ ! <span class="cm-attribute">-f</span> <span class="cm-string">"</span><span class="cm-def">$sort_bam_file</span><span class="cm-string">"</span> ]]; <span class="cm-keyword">then</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-builtin">echo</span> <span class="cm-string">"Sorting BAM file: </span><span class="cm-def">$bam_file</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp;  samtools <span class="cm-builtin">sort</span> <span class="cm-attribute">-o</span> <span class="cm-string">"</span><span class="cm-def">$sort_bam_file</span><span class="cm-string">"</span> <span class="cm-string">"</span><span class="cm-def">$bam_file</span><span class="cm-string">"</span> <span class="cm-attribute">-</span>@ <span class="cm-string">"</span><span class="cm-def">$threads</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-keyword">fi</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># create index</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;  samtools index <span class="cm-string">"</span><span class="cm-def">$sort_bam_file</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># bigwig file path</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-def">bigwig_file</span><span class="cm-operator">=</span><span class="cm-string">"</span><span class="cm-def">${bigwigDir}</span><span class="cm-string">/filter</span><span class="cm-def">${sample}</span><span class="cm-string">deduplicated.sort.bigwig"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># convert BAM TO bigwig</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;  bamCoverage <span class="cm-attribute">-b</span> <span class="cm-string">"</span><span class="cm-def">$sort_bam_file</span><span class="cm-string">"</span> <span class="cm-attribute">-o</span> <span class="cm-string">"</span><span class="cm-def">$bigwig_file</span><span class="cm-string">"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">done</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 830px;"></div><div class="CodeMirror-gutters" style="display: none; height: 830px;"></div></div></div></pre><p>&nbsp;</p></li></ul><hr /><h2 id='heatmap-visualization'><span>Heatmap visualization</span></h2><ul><li><p><span>We are also interested in looking at chromatin features at a list of annotated sites(such as promoters sites)</span></p></li><li><p><span>For example,  We can use the </span><a href='#computematrix'><span>computeMatrix</span></a><span> and </span><a href='#plotheatmap'><span>plotHeatmap</span></a><span> functions from deepTools to generate the heatmap of gene promoters</span></p></li></ul><p>&nbsp;</p><h3 id='heatmap-over-transcription-units'><span>Heatmap over transcription units</span></h3><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 设置并行处理的核心数</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-def">cores</span><span class="cm-operator">=</span><span class="cm-number">8</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Step 1: 使用 computeMatrix 工具来生成信号矩阵，适用于可视化染色质修饰在基因区域的分布模式</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">computeMatrix scale-regions \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-S</span> <span class="cm-def">$projPath</span>/alignment/bigwig/K27me3_rep1_raw.bw \ &nbsp; &nbsp; <span class="cm-comment"># 输入 bigWig 信号文件1（H3K27me3 replicate 1）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; <span class="cm-def">$projPath</span>/alignment/bigwig/K27me3_rep2_raw.bw \ &nbsp; &nbsp; <span class="cm-comment"># 输入 bigWig 信号文件2（H3K27me3 replicate 2）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; <span class="cm-def">$projPath</span>/alignment/bigwig/K4me3_rep1_raw.bw \ &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 输入 bigWig 信号文件3（H3K4me3 replicate 1）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; <span class="cm-def">$projPath</span>/alignment/bigwig/K4me3_rep2_raw.bw \ &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 输入 bigWig 信号文件4（H3K4me3 replicate 2）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-R</span> <span class="cm-def">$projPath</span>/data/hg38_gene/hg38_gene.tsv \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 输入基因注释文件（TSV 格式，包含区域信息）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--beforeRegionStartLength</span> <span class="cm-number">3000</span> \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 参考区域上游延伸长度：3000bp</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--regionBodyLength</span> <span class="cm-number">5000</span> \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 区域本体长度（基因主体长度）：5000bp</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--afterRegionStartLength</span> <span class="cm-number">3000</span> \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 参考区域下游延伸长度：3000bp</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--skipZeros</span> \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 跳过所有 bin 都为 0 的区域（可去除无信号区域）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-o</span> <span class="cm-def">$projPath</span>/data/hg38_gene/matrix_gene.mat.gz \ &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 输出的压缩矩阵文件（供 plotHeatmap 使用）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-p</span> <span class="cm-def">$cores</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 指定并行处理使用的核心数</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># Step 2: 使用 plotHeatmap 可视化 computeMatrix 生成的信号矩阵</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">plotHeatmap \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-m</span> <span class="cm-def">$projPath</span>/data/hg38_gene/matrix_gene.mat.gz \ &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 输入 computeMatrix 生成的矩阵文件</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-out</span> <span class="cm-def">$projPath</span>/data/hg38_gene/Histone_gene.png \ &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 输出的热图图片文件（PNG格式）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--sortUsing</span> sum &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 根据每个区域的信号总和进行排序（从低到高）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 525px;"></div><div class="CodeMirror-gutters" style="display: none; height: 525px;"></div></div></div></pre><p>&nbsp;</p><h3 id='heatmap-on-cuttag-peaks'><span>Heatmap on CUT&amp;Tag peaks</span></h3><ul><li><p><span>We use the </span><strong><span>midpoint of the signal block</span></strong><span> returned from </span><code>SEACR</code><span> to align signals in heatmaps. </span></p></li><li><p><span>The sixth column of the </span><a href='#structure-of-seacr-bed-file'><span>SEACR output</span></a><span> is an entry in the form </span><code>chr:start-end</code><span> that represents  the region with the maximum signal of the region. </span></p></li><li><p><span>We first generate a new </span><strong><span>bed file containing this midpoint information</span></strong><span> in column 6 and use </span><a href='#computematrix'><span>computeMatrix</span></a><span> and  </span><a href='#plotheatmap'><span>plotHeatmap</span></a><span> of </span><code>deeptools</code><span> for the heatmap visualization.</span></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="bash" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="bash"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 使用 awk 从 SEACR 输出的 peak 文件中提取 summit 区域</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-builtin">awk</span> <span class="cm-string">'{</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;  split(<span class="cm-def">$6</span>, summit, <span class="cm-string">":"</span>); &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 将第6列内容按冒号分割，如 "chr1:12345-12500"，得到 summit[1]="chr1"，summit[2]="12345-12500"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;  split(summit[2], region, <span class="cm-string">"-"</span>); &nbsp; &nbsp; <span class="cm-comment"># 将 summit 的区段部分按 - 分割，region[1]=12345，region[2]=12500</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;  print summit[1]<span class="cm-string">"\t"</span>region[1]<span class="cm-string">"\t"</span>region[2] &nbsp; <span class="cm-comment"># 输出格式为 BED：chr\tstart\tend</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">}<span class="cm-string">' $projPath/peakCalling/SEACR/${histName}_${repName}_seacr_control.peaks.stringent.bed \</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-string">&gt; $projPath/peakCalling/SEACR/${histName}_${repName}_seacr_control.peaks.summitRegion.bed</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 输出为一个 BED 文件，包含 peak summit 区域（用于后续画图的区域定义）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 使用 computeMatrix 构建以 peak summit 为中心的信号矩阵</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">computeMatrix reference-point \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-S</span> <span class="cm-def">$projPath</span>/alignment/bigwig/<span class="cm-def">${histName}</span>_<span class="cm-def">${repName}</span>_raw.bw \ &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 输入 bigWig 信号文件（某组 ChIP-seq 数据）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-R</span> <span class="cm-def">$projPath</span>/peakCalling/SEACR/<span class="cm-def">${histName}</span>_<span class="cm-def">${repName}</span>_seacr_control.peaks.summitRegion.bed \ <span class="cm-comment"># 输入以 summit 为中心的 BED 区域</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--skipZeros</span> \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 跳过所有 bin 都为 0 的区域，减少无意义背景</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-o</span> <span class="cm-def">$projPath</span>/peakCalling/SEACR/<span class="cm-def">${histName}</span>_<span class="cm-def">${repName}</span>_SEACR.mat.gz \ <span class="cm-comment"># 输出压缩信号矩阵文件</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-p</span> <span class="cm-def">$cores</span> \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 并行线程数（提高运行速度）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-a</span> <span class="cm-number">3000</span> \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 在参考点（summit）下游延伸 3000bp</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-b</span> <span class="cm-number">3000</span> \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># 在参考点（summit）上游延伸 3000bp</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--referencePoint</span> center &nbsp; &nbsp;<span class="cm-comment"># 设置参考点为区域中心（即 summit 的中心）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 使用 plotHeatmap 对上面生成的矩阵进行可视化</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">plotHeatmap \</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-m</span> <span class="cm-def">$projPath</span>/peakCalling/SEACR/<span class="cm-def">${histName}</span>_SEACR.mat.gz \ &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 输入信号矩阵文件</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">-out</span> <span class="cm-def">$projPath</span>/peakCalling/SEACR/<span class="cm-def">${histName}</span>_SEACR_heatmap.png \ <span class="cm-comment"># 输出热图图像文件（PNG）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--sortUsing</span> sum \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 按每个 peak 区域的总信号排序</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--startLabel</span> <span class="cm-string">"Peak Start"</span> \ &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 设置热图 X 轴左侧标签</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--endLabel</span> <span class="cm-string">"Peak End"</span> \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 设置热图 X 轴右侧标签</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--xAxisLabel</span> <span class="cm-string">""</span> \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span class="cm-comment"># 取消默认 X 轴标签（可避免显示“Distance from center”等）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--regionsLabel</span> <span class="cm-string">"Peaks"</span> \ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-comment"># Y轴区域名称，统一标注为“Peaks”</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-attribute">--samplesLabel</span> <span class="cm-string">"</span><span class="cm-def">${histName}</span><span class="cm-string"> </span><span class="cm-def">${repName}</span><span class="cm-string">"</span> &nbsp;<span class="cm-comment"># 热图图例标签，显示为样本名称（例如 H3K27me3 rep1）</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 699px;"></div><div class="CodeMirror-gutters" style="display: none; height: 699px;"></div></div></div></pre></li></ul><p>&nbsp;</p><hr /><h1 id='differential-analysis'><span>Differential analysis</span></h1><h2 id='deseq2'><span>DESeq2</span></h2><h3 id='deseq2-perface'><span>DEseq2 perface</span></h3><ul><li><p><span>DESeq2: </span><a href='https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8'><span>Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2</span></a></p></li><li><p><span>其通过估计高通量测序实验中计数数据的方差与均值之间的依赖关系，并基于</span><strong><span>负二项分布</span></strong><span>模型进行差异表达分析。</span></p></li></ul><p>&nbsp;</p><div class="md-alert md-alert-tip tip"><p><span class='md-alert-text md-alert-text-tip'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M8 1.5c-2.363 0-4 1.69-4 3.75 0 .984.424 1.625.984 2.304l.214.253c.223.264.47.556.673.848.284.411.537.896.621 1.49a.75.75 0 0 1-1.484.211c-.04-.282-.163-.547-.37-.847a8.456 8.456 0 0 0-.542-.68c-.084-.1-.173-.205-.268-.32C3.201 7.75 2.5 6.766 2.5 5.25 2.5 2.31 4.863 0 8 0s5.5 2.31 5.5 5.25c0 1.516-.701 2.5-1.328 3.259-.095.115-.184.22-.268.319-.207.245-.383.453-.541.681-.208.3-.33.565-.37.847a.751.751 0 0 1-1.485-.212c.084-.593.337-1.078.621-1.489.203-.292.45-.584.673-.848.075-.088.147-.173.213-.253.561-.679.985-1.32.985-2.304 0-2.06-1.637-3.75-4-3.75ZM5.75 12h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5ZM6 15.25a.75.75 0 0 1 .75-.75h2.5a.75.75 0 0 1 0 1.5h-2.5a.75.75 0 0 1-.75-.75Z"></path></svg>Tip</span><br></p><p><strong><span>Negative Binomial vs Poisson Distribution (μ = 20)</span></strong></p><p><span>下图展示了在相同均值 (μ = 20) 下，</span><strong><span>负二项分布</span></strong><span>和</span><strong><span>泊松分布</span></strong><span>的概率质量函数（PMF）比较：</span></p><p><img src="./picture/084.png" width="700"/></p><p><span>🔍 </span><strong><span>观察说明：</span></strong></p><ul><li><p><strong><span>Poisson 分布</span></strong><span>（蓝色虚线）假设：</span></p><ul><li><p><span>方差 = 均值（Var = μ）</span></p></li><li><p><span>因此其分布较窄，集中在均值附近。</span></p></li></ul></li><li><p><strong><span>Negative Binomial 分布</span></strong><span>（红色实线）假设：</span></p><ul><li><p><span>方差 &gt; 均值（Var = μ + α·μ²）</span></p></li><li><p><span>更能描述 RNA-seq 等高通量测序数据中常见的“过度离散”（overdispersion）现象，尾部更长，整体更宽。</span></p></li></ul></li></ul></div><p>&nbsp;</p><h3 id='code'><span>Code</span></h3><p><strong><span>loading package</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="R"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="r"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation"><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">library</span>(<span class="cm-variable">DESeq2</span>)</span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">library</span>(<span class="cm-variable">Rsamtools</span>) &nbsp;<span class="cm-comment"># 用于读取BAM文件</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">library</span>(<span class="cm-variable">GenomicAlignments</span>) &nbsp;<span class="cm-comment"># 用于计算peak区域的counts</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 66px;"></div><div class="CodeMirror-gutters" style="display: none; height: 66px;"></div></div></div></pre><p>&nbsp;</p><p><strong><span>Create a master peak list merging all the peaks called for each sample</span></strong></p><div class="md-alert md-alert-tip tip"><p><span class='md-alert-text md-alert-text-tip'><svg viewBox="0 0 16 16" version="1.1" width="1em" height="1em" aria-hidden="true"><path d="M8 1.5c-2.363 0-4 1.69-4 3.75 0 .984.424 1.625.984 2.304l.214.253c.223.264.47.556.673.848.284.411.537.896.621 1.49a.75.75 0 0 1-1.484.211c-.04-.282-.163-.547-.37-.847a8.456 8.456 0 0 0-.542-.68c-.084-.1-.173-.205-.268-.32C3.201 7.75 2.5 6.766 2.5 5.25 2.5 2.31 4.863 0 8 0s5.5 2.31 5.5 5.25c0 1.516-.701 2.5-1.328 3.259-.095.115-.184.22-.268.319-.207.245-.383.453-.541.681-.208.3-.33.565-.37.847a.751.751 0 0 1-1.485-.212c.084-.593.337-1.078.621-1.489.203-.292.45-.584.673-.848.075-.088.147-.173.213-.253.561-.679.985-1.32.985-2.304 0-2.06-1.637-3.75-4-3.75ZM5.75 12h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1 0-1.5ZM6 15.25a.75.75 0 0 1 .75-.75h2.5a.75.75 0 0 1 0 1.5h-2.5a.75.75 0 0 1-.75-.75Z"></path></svg>Tip</span><br></p><p><strong><span>master peak</span></strong></p><p><span>将样本中重叠或相邻的 peak 区域合并成一个更大的 peak 区域</span></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="ASCII" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="ascii"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">SAMPLES:</span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">───────────────────────────────────────────────────────────────────────</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">Sample 1:  ██████████ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; ██████████ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; █████ &nbsp; &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; [100-200] &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; [300-400] &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; [700-800] &nbsp; </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">Sample 2: &nbsp;  ██████████ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;  ██████████ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; [150-250] &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;  [350-450] &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">Sample 3: &nbsp; &nbsp;  ████ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;  ███████████ &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; [180-220] &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;  [380-500] &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">MASTER PEAK AFTER reduce():</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">───────────────────────────────────────────────────────────────────────</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp;  ██████████████ &nbsp; &nbsp; &nbsp;  ███████████████ &nbsp; &nbsp; &nbsp;  █████ &nbsp; &nbsp; &nbsp; &nbsp; </span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; [100-250] &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;  [300-500] &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;  [700-800] &nbsp; &nbsp; &nbsp; </span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 350px;"></div><div class="CodeMirror-gutters" style="display: none; height: 350px;"></div></div></div></pre><p>&nbsp;</p></div><p>&nbsp;</p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="R" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="r"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 加载必要的包</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">library</span>(<span class="cm-variable">DESeq2</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">library</span>(<span class="cm-variable">Rsamtools</span>) &nbsp;<span class="cm-comment"># 用于读取BAM文件</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">library</span>(<span class="cm-variable">GenomicAlignments</span>) &nbsp;<span class="cm-comment"># 用于计算peak区域的counts</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 初始化一个空的 GRanges 对象，用于保存所有 peak 区域</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">mPeak</span> <span class="cm-operator">=</span> <span class="cm-variable">GRanges</span>()</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">## 遍历每一个组学标记（hist）和其对应的重复（rep）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">## 从每个对应的 SEACR peak 文件中读取 peak 区域，并构建 GRanges 对象</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment">## 将所有 peak 合并进 mPeak 中</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">for</span>(<span class="cm-variable">hist</span> <span class="cm-keyword">in</span> <span class="cm-variable">histL</span>){</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;<span class="cm-keyword">for</span>(<span class="cm-variable">rep</span> <span class="cm-keyword">in</span> <span class="cm-variable">repL</span>){</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 读取 peak 文件（SEACR 输出的文件）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-variable">peakRes</span> <span class="cm-operator">=</span> <span class="cm-variable">read.table</span>(<span class="cm-variable">paste0</span>(<span class="cm-variable">projPath</span>, <span class="cm-string">"/peakCalling/SEACR/"</span>, <span class="cm-variable">hist</span>, <span class="cm-string">"_"</span>, <span class="cm-variable">rep</span>, <span class="cm-string">"_seacr_control.peaks.stringent.bed"</span>), <span class="cm-variable">header</span> <span class="cm-operator">=</span> <span class="cm-variable">FALSE</span>, <span class="cm-variable">fill</span> <span class="cm-operator">=</span> <span class="cm-variable">TRUE</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 构建 GRanges 对象，包含染色体名称、起始和终止位置，链信息设置为通配 '*'</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 并将其追加到 mPeak 中</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-variable">mPeak</span> <span class="cm-operator">=</span> <span class="cm-variable">GRanges</span>(<span class="cm-variable">seqnames</span> <span class="cm-operator">=</span> <span class="cm-variable">peakRes</span><span class="cm-operator cm-dollar">$</span><span class="cm-variable">V1</span>,</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-variable">IRanges</span>(<span class="cm-variable">start</span> <span class="cm-operator">=</span> <span class="cm-variable">peakRes</span><span class="cm-operator cm-dollar">$</span><span class="cm-variable">V2</span>, <span class="cm-variable">end</span> <span class="cm-operator">=</span> <span class="cm-variable">peakRes</span><span class="cm-operator cm-dollar">$</span><span class="cm-variable">V3</span>),</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-variable">strand</span> <span class="cm-operator">=</span> <span class="cm-string">"*"</span>) <span class="cm-operator cm-variable-2">%&gt;%</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-variable">append</span>(<span class="cm-variable">mPeak</span>, <span class="cm-variable">.</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">  }</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">}</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 使用 reduce() 函数将所有 peak 区域合并为 master peak 区域（即合并重叠或相邻区域）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">masterPeak</span> <span class="cm-operator">=</span> <span class="cm-variable">reduce</span>(<span class="cm-variable">mPeak</span>)</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 612px;"></div><div class="CodeMirror-gutters" style="display: none; height: 612px;"></div></div></div></pre><p>&nbsp;</p><p><strong><span>Get the fragment counts for each peak in the master peak list.</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="R" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="r"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 定义包含 BAM 文件的目录（项目路径拼接 "/alignment/bam"）</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">bamDir</span> <span class="cm-operator">=</span> <span class="cm-variable">paste0</span>(<span class="cm-variable">projPath</span>, <span class="cm-string">"/alignment/bam"</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 初始化一个矩阵来存储计数值（行表示 peaks，列表示样本）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">countMat</span> <span class="cm-operator">=</span> <span class="cm-variable">matrix</span>(<span class="cm-atom">NA</span>, <span class="cm-variable">length</span>(<span class="cm-variable">masterPeak</span>), <span class="cm-variable">length</span>(<span class="cm-variable">histL</span>) <span class="cm-operator">*</span> <span class="cm-variable">length</span>(<span class="cm-variable">repL</span>))</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 遍历 histL（组蛋白标记）和 repL（重复实验）组合</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">i</span> <span class="cm-operator">=</span> <span class="cm-number">1</span> &nbsp;<span class="cm-comment"># 初始化计数器，用于在 countMat 中的列索引</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-keyword">for</span>(<span class="cm-variable">hist</span> <span class="cm-keyword">in</span> <span class="cm-variable">histL</span>) {</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp;<span class="cm-keyword">for</span>(<span class="cm-variable">rep</span> <span class="cm-keyword">in</span> <span class="cm-variable">repL</span>) {</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 构造 BAM 文件路径，包含组蛋白标记和重复实验</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-variable">bamFile</span> <span class="cm-operator">=</span> <span class="cm-variable">paste0</span>(<span class="cm-variable">bamDir</span>, <span class="cm-string">"/"</span>, <span class="cm-variable">hist</span>, <span class="cm-string">"_"</span>, <span class="cm-variable">rep</span>, <span class="cm-string">"_bowtie2.mapped.bam"</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 从 BAM 文件中提取区域（由 masterPeak 定义）上的片段计数</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 参数说明：</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># - paired = TRUE: 数据为Paired-end</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># - by_rg = FALSE: 不按 read group 进行分组（根据 BAM 文件结构选择是否使用）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># - format = "bam": 输入文件格式为 BAM</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-variable">fragment_counts</span> <span class="cm-operator cm-arrow">&lt;-</span> <span class="cm-variable">getCounts</span>(<span class="cm-variable">bamFile</span>, <span class="cm-variable">masterPeak</span>, <span class="cm-variable">paired</span> <span class="cm-operator">=</span> <span class="cm-variable">TRUE</span>, <span class="cm-variable">by_rg</span> <span class="cm-operator">=</span> <span class="cm-variable">FALSE</span>, <span class="cm-variable">format</span> <span class="cm-operator">=</span> <span class="cm-string">"bam"</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 将提取的计数值存储到 countMat 矩阵中</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-variable">countMat</span>[, <span class="cm-variable">i</span>] <span class="cm-operator">=</span> <span class="cm-variable">counts</span>(<span class="cm-variable">fragment_counts</span>)[,<span class="cm-number">1</span>]</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-comment"># 增加列索引，用于处理下一个样本</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp;<span class="cm-variable">i</span> <span class="cm-operator">=</span> <span class="cm-variable">i</span> <span class="cm-operator">+</span> <span class="cm-number">1</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">  }</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">}</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 给 countMat 矩阵的列名命名，列名由组蛋白标记和重复实验组成</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 示例：histL = c("H3K27ac", "H3K4me3"), repL = c("rep1", "rep2")</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 结果列名将是 "H3K27ac_rep1", "H3K27ac_rep2" 等等</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">colnames</span>(<span class="cm-variable">countMat</span>) <span class="cm-operator">=</span> <span class="cm-variable">paste</span>(<span class="cm-variable">rep</span>(<span class="cm-variable">histL</span>, <span class="cm-number">2</span>), <span class="cm-variable">rep</span>(<span class="cm-variable">repL</span>, <span class="cm-variable">each</span> <span class="cm-operator">=</span> <span class="cm-number">2</span>), <span class="cm-variable">sep</span> <span class="cm-operator">=</span> <span class="cm-string">"_"</span>)</span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 721px;"></div><div class="CodeMirror-gutters" style="display: none; height: 721px;"></div></div></div></pre><p>&nbsp;</p><p><strong><span>Sequencing depth normalization and differential enriched peaks detection</span></strong></p><pre class="md-fences md-end-block ty-contain-cm modeLoaded" spellcheck="false" lang="R" style="break-inside: unset;"><div class="CodeMirror cm-s-inner cm-s-null-scroll CodeMirror-wrap" lang="r"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 8.9275px; left: 3.99001px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 0px; margin-bottom: 0px; border-right-width: 0px; padding-right: 0px; padding-bottom: 0px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><pre><span>xxxxxxxxxx</span></pre></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-code" role="presentation" style=""><div class="CodeMirror-activeline" style="position: relative;"><div class="CodeMirror-activeline-background CodeMirror-linebackground"></div><div class="CodeMirror-gutter-background CodeMirror-activeline-gutter" style="left: 0px; width: 0px;"></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 选择那些总计数大于 5 的行，即去除低表达基因</span></span></pre></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">selectR</span> <span class="cm-operator">=</span> <span class="cm-variable">which</span>(<span class="cm-variable">rowSums</span>(<span class="cm-variable">countMat</span>) <span class="cm-operator">&gt;</span> <span class="cm-number">5</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 从 countMat 中选择符合条件的行，得到一个新的数据集 dataS</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">dataS</span> <span class="cm-operator">=</span> <span class="cm-variable">countMat</span>[<span class="cm-variable">selectR</span>,]</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 创建一个 factor 变量 'condition'，表示不同的实验条件（组蛋白标记）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">condition</span> <span class="cm-operator">=</span> <span class="cm-variable">factor</span>(<span class="cm-variable">rep</span>(<span class="cm-variable">histL</span>, <span class="cm-variable">each</span> <span class="cm-operator">=</span> <span class="cm-variable">length</span>(<span class="cm-variable">repL</span>)))</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 使用 DESeq2 创建一个 DESeqDataSet 对象，数据来自 dataS，条件来自 'condition'</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 'design = ~ condition' 表示实验设计是基于条件（组蛋白标记）来分析</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">dds</span> <span class="cm-operator">=</span> <span class="cm-variable">DESeqDataSetFromMatrix</span>(<span class="cm-variable">countData</span> <span class="cm-operator">=</span> <span class="cm-variable">dataS</span>,</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-variable">colData</span> <span class="cm-operator">=</span> <span class="cm-variable">DataFrame</span>(<span class="cm-variable">condition</span>),</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span class="cm-variable">design</span> <span class="cm-operator">=</span> <span class="cm-operator">~</span> <span class="cm-variable">condition</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 使用 DESeq2 进行差异分析，得到一个 DESeq 结果对象 DDS</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">DDS</span> <span class="cm-operator">=</span> <span class="cm-variable">DESeq</span>(<span class="cm-variable">dds</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 获取标准化后的计数数据，标准化是基于测序深度进行的</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">normDDS</span> <span class="cm-operator">=</span> <span class="cm-variable">counts</span>(<span class="cm-variable">DDS</span>, <span class="cm-variable">normalized</span> <span class="cm-operator">=</span> <span class="cm-variable">TRUE</span>) &nbsp;<span class="cm-comment"># normalization with respect to the sequencing depth</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 给标准化后的计数数据添加列名后缀 "_norm"</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">colnames</span>(<span class="cm-variable">normDDS</span>) <span class="cm-operator">=</span> <span class="cm-variable">paste0</span>(<span class="cm-variable">colnames</span>(<span class="cm-variable">normDDS</span>), <span class="cm-string">"_norm"</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 获取差异分析的结果，设置独立过滤为 FALSE，指定备择假设为 "greaterAbs"（表达水平显著增加）</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">res</span> <span class="cm-operator">=</span> <span class="cm-variable">results</span>(<span class="cm-variable">DDS</span>, <span class="cm-variable">independentFiltering</span> <span class="cm-operator">=</span> <span class="cm-variable">FALSE</span>, <span class="cm-variable">altHypothesis</span> <span class="cm-operator">=</span> <span class="cm-string">"greaterAbs"</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 将原始数据、标准化计数数据和差异分析结果合并为一个新的数据框 countMatDiff</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">countMatDiff</span> <span class="cm-operator">=</span> <span class="cm-variable">cbind</span>(<span class="cm-variable">dataS</span>, <span class="cm-variable">normDDS</span>, <span class="cm-variable">res</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-comment"># 显示合并后的数据框的前几行</span></span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-variable">head</span>(<span class="cm-variable">countMatDiff</span>)</span></pre><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="" cm-zwsp="">
</span></span></pre></div></div></div></div></div><div style="position: absolute; height: 0px; width: 1px; border-bottom: 0px solid transparent; top: 721px;"></div><div class="CodeMirror-gutters" style="display: none; height: 721px;"></div></div></div></pre><p>&nbsp;</p><ul><li><p><span>DESeq2 requires the input matrix should be </span><strong><span>un-normalized counts</span></strong><span> </span></p></li><li><p><span>-</span><span> </span><strong><span>countMatDiff</span></strong><span> summarizes the differential analysis results: </span></p></li><li><p><span>-</span><span> First 4 columns: raw reads counts after filtering the peak regions with low counts</span></p></li><li><p><span>-</span><span> Second 4 columns: normalized read counts eliminating library size difference.</span></p></li><li><p><span>-</span><span> Remaining columns: differential detection results</span></p></li></ul><p>&nbsp;</p><hr /><p>&nbsp;</p></div></div>

<script>(function(){function e(e,n,i){document.addEventListener(e,function(e){if(!e.defaultPrevented)for(var t=e.target;t&&t!=this;t=t.parentNode)if(t.matches(n)){!1===i.call(t,e)&&(e.preventDefault(),e.stopPropagation());break}},!1)}var t=document.body.parentElement,i=[],r=null,o=document.body.classList.contains("typora-export-collapse-outline");function a(){return t.scrollTop}e("click",".outline-expander",function(e){var t=this.closest(".outline-item-wrapper").classList;return t.contains("outline-item-open")?t.remove("outline-item-open"):t.add("outline-item-open"),u(),!1}),e("click",".outline-item",function(e){var t=this.querySelector(".outline-label");location.hash="#"+t.getAttribute("href"),o&&((t=this.closest(".outline-item-wrapper").classList).contains("outline-item-open")||t.add("outline-item-open"),d(),t.add("outline-item-active"))});function s(){var e=a();r=null;for(var t=0;t<i.length&&i[t][1]-e<60;t++)r=i[t]}function n(){c=setTimeout(function(){var n;i=[],n=a(),document.querySelector("#write").querySelectorAll("h1, h2, h3, h4, h5, h6").forEach(e=>{var t=e.getAttribute("id");i.push([t,n+e.getBoundingClientRect().y])}),s(),u()},300)}var l,c,d=function(){document.querySelectorAll(".outline-item-active").forEach(e=>e.classList.remove("outline-item-active")),document.querySelectorAll(".outline-item-single.outline-item-open").forEach(e=>e.classList.remove("outline-item-open"))},u=function(){if(r&&(d(),t=document.querySelector('.outline-label[href="#'+(CSS.escape?CSS.escape(r[0]):r[0])+'"]')))if(o){var e=t.closest(".outline-item-open>ul>.outline-item-wrapper");if(e)e.classList.add("outline-item-active");else{for(var t,n=(t=t.closest(".outline-item-wrapper")).parentElement.closest(".outline-item-wrapper");n;)n=(t=n).parentElement.closest(".outline-item-wrapper");t.classList.add("outline-item-active")}}else t.closest(".outline-item-wrapper").classList.add("outline-item-active")};window.addEventListener("scroll",function(e){l&&clearTimeout(l),l=setTimeout(function(){s(),u()},300)});window.addEventListener("resize",function(e){c&&clearTimeout(c),n()}),n()})();</script></body>
</html>